Index: src/depot_gui.cpp
===================================================================
--- src/depot_gui.cpp	(revisi¢n: 17194)
+++ src/depot_gui.cpp	(copia de trabajo)
@@ -22,6 +22,8 @@
 #include "tilehighlight_func.h"
 #include "window_gui.h"
 #include "vehiclelist.h"
+#include "openttd.h"
+#include "cheat_type.h"

 #include "table/strings.h"
 #include "table/sprites.h"
@@ -553,7 +555,7 @@
 	 */
 	void HandleCloneVehClick(const Vehicle *v)
 	{
-		if (v == NULL || !IsCompanyBuildableVehicleType(v)) return;
+		if (v == NULL || !IsCompanyBuildableVehicleType(v) || (_pause_mode != PM_UNPAUSED && !_cheats.build_in_pause.value)) return;

 		if (!v->IsPrimaryVehicle()) {
 			v = v->First();
