# HG changeset patch
# Parent b2d7c694dfef44a4a1a89ed8f0869e050b0306da

diff -r b2d7c694dfef src/order_gui.cpp
--- a/src/order_gui.cpp	Wed Oct 09 18:45:17 2013 +0200
+++ b/src/order_gui.cpp	Wed Oct 09 18:45:21 2013 +0200
@@ -1275,7 +1275,15 @@
 				if (this->GetWidget<NWidgetLeaf>(widget)->ButtonHit(pt)) {
 					this->OrderClick_Goto();
 				} else {
-					ShowDropDownMenu(this, this->vehicle->type == VEH_AIRCRAFT ? _order_goto_dropdown_aircraft : _order_goto_dropdown, 0, WID_O_GOTO, 0, 0);
+					int sel;
+					switch (this->goto_type) {
+						case OPOS_NONE:        sel = -1; break;
+						case OPOS_GOTO:        sel =  0; break;
+						case OPOS_CONDITIONAL: sel =  2; break;
+						case OPOS_SHARE:       sel =  3; break;
+						default: NOT_REACHED();
+					}
+					ShowDropDownMenu(this, this->vehicle->type == VEH_AIRCRAFT ? _order_goto_dropdown_aircraft : _order_goto_dropdown, sel, WID_O_GOTO, 0, 0);
 				}
 				break;
 
