diff -r e5575895a889 bin/data/opntitle.dat
Binary file bin/data/opntitle.dat has changed
diff -r e5575895a889 config.lib
--- a/config.lib	Sat Apr 10 02:17:18 2010 +0000
+++ b/config.lib	Sun Apr 11 22:07:04 2010 +0200
@@ -3169,6 +3169,7 @@
 	echo "  --with-midi=midi               define which midi-player to use"
 	echo "  --with-midi-arg=arg            define which args to use for the"
 	echo "                                 midi-player"
+	echo "  --with-libtimidity             enables libtimidity support"
 	echo "  --with-allegrol[=allegro-config]"
 	echo "                                 enables Allegro video driver support"
 	echo "  --with-cocoa                   enables COCOA video driver (OSX ONLY)"
