Index: configure
===================================================================
--- configure	(revision 16412)
+++ configure	(working copy)
@@ -22,11 +22,11 @@
 fi
 # Find out where configure is (in what dir)
 ROOT_DIR="`dirname $0`"
-ROOT_DIR="`cd $ROOT_DIR && pwd`"
+ROOT_DIR="`cd $ROOT_DIR && (pwd -W 2>/dev/null || pwd 2>/dev/null)`"
 
 check_path_characters
 
-PWD="`pwd`"
+PWD="`pwd -W 2>/dev/null || pwd 2>/dev/null`"
 PREFIX="$PWD/bin"
 
 . $ROOT_DIR/config.lib
Index: Makefile.src.in
===================================================================
--- Makefile.src.in	(revision 16412)
+++ Makefile.src.in	(working copy)
@@ -190,7 +190,7 @@
 
 # Convert x:/... paths to /x/... for mingw
 ifeq ($(OS), MINGW)
-	@cat Makefile.dep.tmp | sed 's@\([a-zA-Z]\):\/@\/\1\/@g' > Makefile.dep.tmp.mingw
+	@cat Makefile.dep.tmp | sed 's@/\([a-zA-Z]\):\/@\/\1\/@g' > Makefile.dep.tmp.mingw
 	@cp Makefile.dep.tmp.mingw Makefile.dep.tmp
 	@rm -f Makefile.dep.tmp.mingw
 endif
@@ -200,7 +200,7 @@
 	@$(AWK) '                           \
 	/^# DO NOT/ { print $$0 ; next}     \
 	/^#/ {next}                         \
-	/:/ {                               \
+	/: / {                               \
 		left = NF - 1;                    \
 		for (n = 2; n <= NF; n++) {       \
 			if (match($$n, "^$(ROOT_DIR)") == 0) { \
