diff --git a/src/rail_cmd.cpp b/src/rail_cmd.cpp
index ffa42de..0c7b8f9 100644
--- a/src/rail_cmd.cpp
+++ b/src/rail_cmd.cpp
@@ -97,7 +97,7 @@ void ResolveRailTypeGUISprites(RailtypeInfo *rti)
 	};
 
 	for (SignalType type = SIGTYPE_NORMAL; type < SIGTYPE_END; type = (SignalType)(type + 1)) {
-		for (SignalVariant var = SIG_ELECTRIC; var <= SIG_SEMAPHORE; var = (SignalVariant)(var + 1)) {
+		for (SignalVariant var = SIG_ELECTRIC; var < SIG_END; var = (SignalVariant)(var + 1)) {
 			SpriteID red   = GetCustomSignalSprite(rti, INVALID_TILE, type, var, SIGNAL_STATE_RED, true);
 			SpriteID green = GetCustomSignalSprite(rti, INVALID_TILE, type, var, SIGNAL_STATE_GREEN, true);
 			rti->gui_sprites.signals[type][var][0] = (red != 0)   ? red + SIGNAL_TO_SOUTH   : _signal_lookup[var][type];
diff --git a/src/signal_type.h b/src/signal_type.h
index c7d0607..7292c8e 100644
--- a/src/signal_type.h
+++ b/src/signal_type.h
@@ -18,6 +18,8 @@
 enum SignalVariant {
 	SIG_ELECTRIC  = 0, ///< Light signal
 	SIG_SEMAPHORE = 1, ///< Old-fashioned semaphore signal
+
+	SIG_END,
 };
 
 
