Index: src/newgrf_debug_gui.cpp
===================================================================
--- src/newgrf_debug_gui.cpp	(revision 27171)
+++ src/newgrf_debug_gui.cpp	(working copy)
@@ -863,10 +863,10 @@
 
 				/* And draw only the part within the sprite area */
 				SubSprite subspr = {
-					spr->x_offs + (spr->width  - UnScaleGUI(width) ) / 2 + 1,
-					spr->y_offs + (spr->height - UnScaleGUI(height)) / 2 + 1,
-					spr->x_offs + (spr->width  + UnScaleGUI(width) ) / 2 - 1,
-					spr->y_offs + (spr->height + UnScaleGUI(height)) / 2 - 1,
+					spr->x_offs + (spr->width  - ScaleByZoom(width,  ZOOM_LVL_GUI)) / 2 + 1,
+					spr->y_offs + (spr->height - ScaleByZoom(height, ZOOM_LVL_GUI)) / 2 + 1,
+					spr->x_offs + (spr->width  + ScaleByZoom(width,  ZOOM_LVL_GUI)) / 2 - 1,
+					spr->y_offs + (spr->height + ScaleByZoom(height, ZOOM_LVL_GUI)) / 2 - 1,
 				};
 
 				DrawSprite(this->current_sprite, PAL_NONE, x, y, &subspr, ZOOM_LVL_GUI);
