Index: network_gui.c
===================================================================
--- network_gui.c	(r‚vision 4814)
+++ network_gui.c	(copie de travail)
@@ -192,6 +192,7 @@
 	uint i;
 
 	if (!(nqld->l.flags & VL_RESORT)) return;
+	if (nqld->sort_list == NULL) return;
 
 	_internal_sort_order = nqld->l.flags & VL_DESC;
 	qsort(nqld->sort_list, nqld->l.list_length, sizeof(nqld->sort_list[0]), _ngame_sorter[nqld->l.sort_type]);
