diff -r fce8aff98a0c src/widget.cpp
--- a/src/widget.cpp	Thu Jun 04 16:57:39 2009 +0200
+++ b/src/widget.cpp	Sat Jun 06 14:56:49 2009 +0200
@@ -1428,6 +1428,7 @@
 				num_changing_childs--;
 				if (hor_step > 1) increment -= increment % hor_step;
 				child_wid->current_x = child_wid->smallest_x + increment;
+				additional_length -= increment;
 				continue;
 			}
 			next_biggest_stepsize = max(next_biggest_stepsize, hor_step);
@@ -1560,6 +1561,7 @@
 				num_changing_childs--;
 				if (vert_step > 1) increment -= increment % vert_step;
 				child_wid->current_y = child_wid->smallest_y + increment;
+				additional_length -= increment;
 				continue;
 			}
 			next_biggest_stepsize = max(next_biggest_stepsize, vert_step);
