Index: main_gui.c
===================================================================
--- main_gui.c	(revision 6323)
+++ main_gui.c	(working copy)
@@ -865,16 +865,16 @@
 				break;
 		}
 
-		assert(wt);
-
-		// update the toolbar button too
-		CLRBIT(wt->disabled_state, button);
-		CLRBIT(wt->disabled_state, button + 1);
-		switch (vp->zoom) {
-			case 0: SETBIT(wt->disabled_state, button); break;
-			case 2: SETBIT(wt->disabled_state, button + 1); break;
+		if (wt) {
+			// update the toolbar button too
+			CLRBIT(wt->disabled_state, button);
+			CLRBIT(wt->disabled_state, button + 1);
+			switch (vp->zoom) {
+				case 0: SETBIT(wt->disabled_state, button); break;
+				case 2: SETBIT(wt->disabled_state, button + 1); break;
+			}
+			SetWindowDirty(wt);
 		}
-		SetWindowDirty(wt);
 	}
 
 	return true;
Index: genworld.c
===================================================================
--- genworld.c	(revision 6323)
+++ genworld.c	(working copy)
@@ -260,14 +260,12 @@
 		ShowGenerateWorldProgress();
 	}
 
+	/* Hide vital windows, because we don't allow to use them */
+	if (_gw.thread != NULL) HideVitalWindows();
+
 	/* Zoom out and center on the map (is pretty ;)) */
 	if (FindWindowById(WC_MAIN_WINDOW, 0) != NULL) {
 		while (DoZoomInOutWindow(ZOOM_OUT, FindWindowById(WC_MAIN_WINDOW, 0) ) ) {}
 		ScrollMainWindowToTile(TileXY(MapSizeX() / 2, MapSizeY() / 2));
 	}
-
-	/* Hide vital windows, because we don't allow to use them */
-	/* XXX -- Ideal it is done after ShowGenerateWorldProgress, but stupid
-	 * enough, DoZoomInOutWindow _needs_ the toolbar to exist... */
-	if (_gw.thread != NULL) HideVitalWindows();
 }
