Index: src/dock_gui.cpp
===================================================================
--- src/dock_gui.cpp	(revision 21000)
+++ src/dock_gui.cpp	(working copy)
@@ -260,10 +260,6 @@
 	{
 		if (pt.x != -1) {
 			switch (select_proc) {
-				case DDSP_BUILD_BRIDGE:
-					if (!_settings_client.gui.persistent_buildingtools) ResetObjectToPlace();
-					DoCommandP(end_tile, start_tile, TRANSPORT_WATER << 15, CMD_BUILD_BRIDGE | CMD_MSG(STR_ERROR_CAN_T_BUILD_AQUEDUCT_HERE), CcBuildBridge);
-
 				case DDSP_DEMOLISH_AREA:
 					GUIPlaceProcDragXY(select_proc, start_tile, end_tile);
 					break;
