Index: src/pbs.cpp
===================================================================
--- src/pbs.cpp	(revision 14849)
+++ src/pbs.cpp	(working copy)
@@ -194,7 +194,8 @@
 	 * end and use that as the tile for the loop test because the track
 	 * follower will skip to there in the actual loop as well. */
 	if (IsRailwayStationTile(start_tile) && ft.Follow(start_tile, start_trackdir) && ft.m_tiles_skipped > 0) {
-		start_tile = ft.m_new_tile;
+		start_tile = ft.m_old_tile;
+		start_trackdir = ft.m_old_td;
 	}
 
 	while (ft.Follow(tile, trackdir)) {
