Index: config.lib
===================================================================
--- config.lib	(revision 16599)
+++ config.lib	(working copy)
@@ -1643,24 +1643,22 @@
 }
 
 check_osx_sdk() {
-cat > tmp.osx.cpp << EOF
-#include <AvailabilityMacros.h>
-#if !defined(MAC_OS_X_VERSION_10_5)
-# error "Need newer SDK"
-#endif
+cat > tmp.osx.m << EOF
+#include <Cocoa/Cocoa.h>
 int main() {
+	kCGBitmapByteOrder32Host;
 	return 0;
 }
 EOF
-	execute="$cxx_host $CFLAGS -E tmp.osx.cpp -o - 2>&1"
+	execute="$cc_host $CFLAGS tmp.osx.m -framework Cocoa -o tmp.osx 2>&1"
 	eval $execute > /dev/null
 	ret=$?
 	log 2 "executing $execute"
 	log 2 "  exit code $ret"
-	rm -f tmp.osx.cpp
+	rm -f tmp.osx.m tmp.osx
 	if [ "$ret" != "0" ]; then
-		log 1 "I couldn't detect any XCode >= 2.5 on your system"
-		log 1 "please install/upgrade your XCode"
+		log 1 "Your system SDK is probably too old"
+		log 1 "Please install/upgrade your Xcode to >= 2.5"
 
 		exit 1
 	fi
