Index: src/news_gui.cpp
===================================================================
--- src/news_gui.cpp	(revision 16483)
+++ src/news_gui.cpp	(working copy)
@@ -223,7 +230,7 @@
 				DrawString(2, this->width - 1, 1, STR_SILVER_CROSS);
 
 				SetDParam(0, this->ni->date);
-				DrawString(2, this->width - 1, 1, STR_DATE_LONG_SMALL, TC_FROMSTRING, SA_RIGHT);
+				DrawString(2, this->width - 2, 1, STR_DATE_LONG_SMALL, TC_FROMSTRING, SA_RIGHT);
 
 				if (!(this->ni->flags & NF_VIEWPORT)) {
 					CopyInDParam(0, this->ni->params, lengthof(this->ni->params));
