diff --git a/src/os/macosx/macos.mm b/src/os/macosx/macos.mm
--- a/src/os/macosx/macos.mm
+++ b/src/os/macosx/macos.mm
@@ -88,7 +88,7 @@
 	if (system) {
 		ShowMacDialog("OpenTTD has encountered an error", buf, "Quit");
 	} else {
-		ShowMacDialog(buf, "See the readme for more info.\nMost likely you are missing files from the original TTD.", "Quit");
+		ShowMacDialog(buf, "See the readme for more info.", "Quit");
 	}
 }
 
