Index: src/landscape.cpp
===================================================================
--- src/landscape.cpp	(revision 11397)
+++ src/landscape.cpp	(working copy)
@@ -658,12 +658,24 @@
 				case SLOPE_E:
 				case SLOPE_S:
 				case SLOPE_W:
+					MakeShore(tile);
+					break;
 				case SLOPE_NW:
+					if (GetTileSlope(TileAddByDiagDir(tile, DIAGDIR_SE), NULL) != SLOPE_SE)
+						MakeShore(tile);
+						break;
 				case SLOPE_SW:
+					if (GetTileSlope(TileAddByDiagDir(tile, DIAGDIR_NE), NULL) != SLOPE_NE)
+						MakeShore(tile);
+						break;
 				case SLOPE_SE:
+					if (GetTileSlope(TileAddByDiagDir(tile, DIAGDIR_NW), NULL) != SLOPE_NW)
+						MakeShore(tile);
+						break;
 				case SLOPE_NE:
-					MakeShore(tile);
-					break;
+					if (GetTileSlope(TileAddByDiagDir(tile, DIAGDIR_SW), NULL) != SLOPE_SW)
+						MakeShore(tile);
+						break;
 
 				default:
 					break;
