Index: lang/english.txt
===================================================================
--- lang/english.txt	(revision 16404)
+++ lang/english.txt	(working copy)
@@ -564,6 +564,7 @@
 STR_NEWS_MESSAGE_TYPE_CHANGES_OF_CARGO_ACCEPTANCE               :{YELLOW}Changes to cargo acceptance
 STR_NEWS_MESSAGE_TYPE_SUBSIDIES                                 :{YELLOW}Subsidies
 STR_NEWS_MESSAGE_TYPE_GENERAL_INFORMATION                       :{YELLOW}General information
+STR_NEWS_MESSAGE_TYPE_ADVICE_VEH_OLD                            :{YELLOW}Advice on vehicle age
 STR_MESSAGES_ALL                                                :{YELLOW}Set all message types to: Off / Summary/ Full
 STR_MESSAGE_SOUND                                               :{YELLOW}Play sound for summarised news messages
 STR_ERROR_TOO_FAR_FROM_PREVIOUS_DESTINATION                     :{WHITE}...too far from previous destination
Index: news_gui.cpp
===================================================================
--- news_gui.cpp	(revision 16404)
+++ news_gui.cpp	(working copy)
@@ -121,6 +121,7 @@
 	{ NT_INDUSTRY_OTHER,   NM_THIN,     NF_VIEWPORT|NF_TILE,    NULL                    }, ///< NS_INDUSTRY_OTHER
 	{ NT_INDUSTRY_NOBODY,  NM_THIN,     NF_VIEWPORT|NF_TILE,    NULL                    }, ///< NS_INDUSTRY_NOBODY
 	{ NT_ADVICE,           NM_SMALL,    NF_VIEWPORT|NF_VEHICLE, NULL                    }, ///< NS_ADVICE
+	{ NT_AGE,              NM_SMALL,    NF_VIEWPORT|NF_VEHICLE, NULL                    }, ///< NS_AGE
 	{ NT_NEW_VEHICLES,     NM_NORMAL,   NF_NONE,                DrawNewsNewVehicleAvail }, ///< NS_NEW_VEHICLES
 	{ NT_ACCEPTANCE,       NM_SMALL,    NF_VIEWPORT|NF_TILE,    NULL                    }, ///< NS_ACCEPTANCE
 	{ NT_SUBSIDIES,        NM_NORMAL,   NF_TILE|NF_TILE2,       NULL                    }, ///< NS_SUBSIDIES
@@ -145,6 +146,7 @@
 	NewsTypeData("production_other",  30, SND_BEGIN,       STR_NEWS_MESSAGE_TYPE_INDUSTRY_CHANGES_SERVED_BY_OTHER   ),  ///< NT_INDUSTRY_OTHER
 	NewsTypeData("production_nobody", 30, SND_BEGIN,       STR_NEWS_MESSAGE_TYPE_INDUSTRY_CHANGES_UNSERVED          ),  ///< NT_INDUSTRY_NOBODY
 	NewsTypeData("advice",           150, SND_BEGIN,       STR_NEWS_MESSAGE_TYPE_ADVICE_INFORMATION_ON_COMPANY      ),  ///< NT_ADVICE
+	NewsTypeData("age",              150, SND_BEGIN,       STR_NEWS_MESSAGE_TYPE_ADVICE_VEH_OLD                     ),  ///< NT_AGE
 	NewsTypeData("new_vehicles",      30, SND_1E_OOOOH,    STR_NEWS_MESSAGE_TYPE_NEW_VEHICLES                       ),  ///< NT_NEW_VEHICLES
 	NewsTypeData("acceptance",        90, SND_BEGIN,       STR_NEWS_MESSAGE_TYPE_CHANGES_OF_CARGO_ACCEPTANCE        ),  ///< NT_ACCEPTANCE
 	NewsTypeData("subsidies",        180, SND_BEGIN,       STR_NEWS_MESSAGE_TYPE_SUBSIDIES                          ),  ///< NT_SUBSIDIES
@@ -1027,6 +1029,7 @@
 NEWS_SETTINGS_LINE(28, NT_INDUSTRY_OTHER),
 NEWS_SETTINGS_LINE(28, NT_INDUSTRY_NOBODY),
 NEWS_SETTINGS_LINE(28, NT_ADVICE),
+NEWS_SETTINGS_LINE(28, NT_AGE),
 NEWS_SETTINGS_LINE(28, NT_NEW_VEHICLES),
 NEWS_SETTINGS_LINE(28, NT_ACCEPTANCE),
 NEWS_SETTINGS_LINE(28, NT_SUBSIDIES),
Index: news_type.h
===================================================================
--- news_type.h	(revision 16404)
+++ news_type.h	(working copy)
@@ -25,6 +25,7 @@
 	NT_INDUSTRY_OTHER,  ///< Production changes of industry serviced by competitor(s)
 	NT_INDUSTRY_NOBODY, ///< Other industry production changes
 	NT_ADVICE,          ///< Bits of news about vehicles of the company
+	NT_AGE,             ///< Bits of news about old vehicles of the company
 	NT_NEW_VEHICLES,    ///< New vehicle has become available
 	NT_ACCEPTANCE,      ///< A type of cargo is (no longer) accepted
 	NT_SUBSIDIES,       ///< News about subsidies (announcements, expirations, acceptance)
@@ -51,6 +52,7 @@
 	NS_INDUSTRY_OTHER,   ///< NT_INDUSTRY_OTHER
 	NS_INDUSTRY_NOBODY,  ///< NT_INDUSTRY_NOBODY
 	NS_ADVICE,           ///< NT_ADVICE
+	NS_AGE,              ///< NT_AGE
 	NS_NEW_VEHICLES,     ///< NT_NEW_VEHICLES
 	NS_ACCEPTANCE,       ///< NT_ACCEPTANCE
 	NS_SUBSIDIES,        ///< NT_SUBSIDIES
Index: vehicle.cpp
===================================================================
--- vehicle.cpp	(revision 16404)
+++ vehicle.cpp	(working copy)
@@ -922,7 +922,7 @@
 	}
 
 	SetDParam(0, v->index);
-	AddNewsItem(str, NS_ADVICE, v->index, 0);
+	AddNewsItem(str, NS_AGE, v->index, 0);
 }
 
 /**
