diff -r c5027f9e757d src/newgrf_gui.cpp
--- a/src/newgrf_gui.cpp	Mon May 24 09:23:21 2010 +0200
+++ b/src/newgrf_gui.cpp	Mon May 24 09:23:37 2010 +0200
@@ -736,6 +736,9 @@
 				this->avails.ForceRebuild();
 				this->InvalidateData(2);
 			}
+		} else if (widget == SNGRFS_AVAIL_LIST && this->active_sel != NULL) { // Remove active NewGRF file by dragging it over available list
+			Point pt_dummy = {-1, -1};
+			this->OnClick(pt_dummy, SNGRFS_REMOVE, 1);
 		}
 
 		ResetObjectToPlace();
