Index: src/airport_gui.cpp
===================================================================
--- src/airport_gui.cpp	(revision 12917)
+++ src/airport_gui.cpp	(working copy)
@@ -190,7 +190,7 @@
 			 * XXX TODO : all airports should be held in arrays, with all relevant data.
 			 * This should be part of newgrf-airports, i suppose
 			 */
-			for (i = 0; i < BAW_AIRPORT_COUNT; i++) w->SetWidgetDisabledState(i + BAW_SMALL_AIRPORT, !HasBit(avail_airports, i));
+			for (i = 0; i <= BAW_AIRPORT_COUNT; i++) w->SetWidgetDisabledState(i + BAW_SMALL_AIRPORT, !HasBit(avail_airports, i));
 
 			/* select default the coverage area to 'Off' (16) */
 			airport = GetAirport(_selected_airport_type);
