Index: src/bridge_gui.cpp
===================================================================
--- src/bridge_gui.cpp	(revision 10167)
+++ src/bridge_gui.cpp	(working copy)
@@ -44,10 +44,10 @@
 		case WE_PAINT:
 			DrawWindowWidgets(w);
 
-			for (uint i = 0; i < 4 && i + w->vscroll.pos < _bridgedata.count; i++) {
-				const Bridge *b = &_bridge[_bridgedata.indexes[i + w->vscroll.pos]];
+			for (uint i = 0; i < _bridgedata.count; i++) {
+				const Bridge *b = &_bridge[_bridgedata.indexes[i]];
 
-				SetDParam(2, _bridgedata.costs[i + w->vscroll.pos]);
+				SetDParam(2, _bridgedata.costs[i]);
 				SetDParam(1, b->speed * 10 / 16);
 				SetDParam(0, b->material);
 				DrawSprite(b->sprite, b->pal, 3, 15 + i * 22);
@@ -69,47 +69,29 @@
 		case WE_CLICK:
 			if (e->we.click.widget == 2) {
 				uint ind = ((int)e->we.click.pt.y - 14) / 22;
-				if (ind < 4 && (ind += w->vscroll.pos) < _bridgedata.count)
+				if (ind < _bridgedata.count)
 					BuildBridge(w, ind);
 			}
 			break;
 	}
 }
 
-static const Widget _build_bridge_widgets[] = {
-{   WWT_CLOSEBOX,   RESIZE_NONE,     7,     0,    10,     0,    13, STR_00C5,                    STR_018B_CLOSE_WINDOW},
-{    WWT_CAPTION,   RESIZE_NONE,     7,    11,   199,     0,    13, STR_100D_SELECT_RAIL_BRIDGE, STR_018C_WINDOW_TITLE_DRAG_THIS},
-{     WWT_MATRIX,   RESIZE_NONE,     7,     0,   187,    14,   101, 0x401,                       STR_101F_BRIDGE_SELECTION_CLICK},
-{  WWT_SCROLLBAR,   RESIZE_NONE,     7,   188,   199,    14,   101, 0x0,                         STR_0190_SCROLL_BAR_SCROLLS_LIST},
+// members set to -1 are handled by ShowBuildBridgeWindow
+static Widget _build_bridge_widgets[] = {
+{   WWT_CLOSEBOX,   RESIZE_NONE,     7,     0,    10,     0,    13, STR_00C5,  STR_018B_CLOSE_WINDOW},
+{    WWT_CAPTION,   RESIZE_NONE,     7,    11,   199,     0,    13, -1,        STR_018C_WINDOW_TITLE_DRAG_THIS},
+{     WWT_MATRIX,   RESIZE_NONE,     7,     0,   199,    14,    -1, -1,        STR_101F_BRIDGE_SELECTION_CLICK},
 {   WIDGETS_END},
 };
 
-static const WindowDesc _build_bridge_desc = {
-	WDP_AUTO, WDP_AUTO, 200, 102,
+static WindowDesc _build_bridge_desc = {
+	WDP_AUTO, WDP_AUTO, 200, -1,
 	WC_BUILD_BRIDGE, WC_BUILD_TOOLBAR,
 	WDF_STD_TOOLTIPS | WDF_STD_BTN | WDF_DEF_WIDGET,
 	_build_bridge_widgets,
 	BuildBridgeWndProc
 };
 
-
-static const Widget _build_road_bridge_widgets[] = {
-{   WWT_CLOSEBOX,   RESIZE_NONE,     7,     0,    10,     0,    13, STR_00C5,                    STR_018B_CLOSE_WINDOW},
-{    WWT_CAPTION,   RESIZE_NONE,     7,    11,   199,     0,    13, STR_1803_SELECT_ROAD_BRIDGE, STR_018C_WINDOW_TITLE_DRAG_THIS},
-{     WWT_MATRIX,   RESIZE_NONE,     7,     0,   187,    14,   101, 0x401,                       STR_101F_BRIDGE_SELECTION_CLICK},
-{  WWT_SCROLLBAR,   RESIZE_NONE,     7,   188,   199,    14,   101, 0x0,                         STR_0190_SCROLL_BAR_SCROLLS_LIST},
-{   WIDGETS_END},
-};
-
-static const WindowDesc _build_road_bridge_desc = {
-	WDP_AUTO, WDP_AUTO, 200, 102,
-	WC_BUILD_BRIDGE, WC_BUILD_TOOLBAR,
-	WDF_STD_TOOLTIPS | WDF_STD_BTN | WDF_DEF_WIDGET,
-	_build_road_bridge_widgets,
-	BuildBridgeWndProc
-};
-
-
 void ShowBuildBridgeWindow(TileIndex start, TileIndex end, byte bridge_type)
 {
 	uint j = 0;
@@ -155,9 +137,12 @@
 	_bridgedata.count = j;
 
 	if (j != 0) {
-		Window *w = AllocateWindowDesc((_bridgedata.type & 0x80) ? &_build_road_bridge_desc : &_build_bridge_desc);
-		w->vscroll.cap = 4;
-		w->vscroll.count = (byte)j;
+		_build_bridge_widgets[1].data = (_bridgedata.type & 0x80) ? STR_1803_SELECT_ROAD_BRIDGE : STR_100D_SELECT_RAIL_BRIDGE;
+		_build_bridge_widgets[2].data = (j << 8) | 0x1; // draw j x 1 grid
+		_build_bridge_widgets[2].bottom = 13 + 22*j;
+		_build_bridge_desc.height = 14 + 22*j;
+
+		AllocateWindowDesc(&_build_bridge_desc);
 	} else {
 		ShowErrorMessage(errmsg, STR_5015_CAN_T_BUILD_BRIDGE_HERE, TileX(end) * TILE_SIZE, TileY(end) * TILE_SIZE);
 	}
