Index: src/town_cmd.cpp
===================================================================
--- src/town_cmd.cpp	(revision 18709)
+++ src/town_cmd.cpp	(working copy)
@@ -2164,8 +2164,10 @@
 			/* 1x1 house checks are already done */
 		}
 
+		uint8 random_data = Random();
+
 		if (HasBit(hs->callback_mask, CBM_HOUSE_ALLOW_CONSTRUCTION)) {
-			uint16 callback_res = GetHouseCallback(CBID_HOUSE_ALLOW_CONSTRUCTION, 0, 0, house, t, tile);
+			uint16 callback_res = GetHouseCallback(CBID_HOUSE_ALLOW_CONSTRUCTION, random_data, 0, house, t, tile);
 			if (callback_res != CALLBACK_FAILED && GB(callback_res, 0, 8) == 0) continue;
 		}
 
@@ -2191,7 +2193,7 @@
 			}
 		}
 
-		MakeTownHouse(tile, t, construction_counter, construction_stage, house, Random());
+		MakeTownHouse(tile, t, construction_counter, construction_stage, house, random_data);
 
 		return true;
 	}
