Index: src/squirrel_std.cpp
===================================================================
--- src/squirrel_std.cpp	(revision 12066)
+++ src/squirrel_std.cpp	(working copy)
@@ -22,7 +22,7 @@
 {
 	SQInteger top = sq_gettop(vm);
 	const SQChar *filename;
-	SQChar *real_filename;
+	SQChar real_filename[MAX_CHAR];
 
 	sq_getstring(vm, 2, &filename);
 
@@ -33,7 +33,10 @@
 		DEBUG(misc, 0, "[squirrel] Couldn't detect the script-name of the 'require'-caller; this should never happen!");
 		return SQ_ERROR;
 	}
-	real_filename = scstrdup(si.source);
+	// why there is't any scstrcpy ?!
+	real_filename[0] = '\0';
+	scstrcat(real_filename, si.source);
+
 	/* Keep the dir, remove the rest */
 	SQChar *s = scstrrchr(real_filename, PATHSEPCHAR);
 	if (s != NULL) {
