Index: src/lang/english.txt
===================================================================
--- src/lang/english.txt	(revision 15832)
+++ src/lang/english.txt	(working copy)
@@ -561,6 +561,7 @@
 STR_020D_CHANGES_OF_CARGO_ACCEPTANCE                            :{YELLOW}Changes to cargo acceptance
 STR_020E_SUBSIDIES                                              :{YELLOW}Subsidies
 STR_020F_GENERAL_INFORMATION                                    :{YELLOW}General information
+STR_ADVICE_IS_LOST                                              :{YELLOW}Advise when a train is lost
 STR_MESSAGES_ALL                                                :{YELLOW}Set all message types to: Off / Summary/ Full
 STR_MESSAGE_SOUND                                               :{YELLOW}Play sound for summarised news messages
 STR_0210_TOO_FAR_FROM_PREVIOUS_DESTINATIO                       :{WHITE}...too far from previous destination
Index: src/news_gui.cpp
===================================================================
--- src/news_gui.cpp	(revision 15832)
+++ src/news_gui.cpp	(working copy)
@@ -121,6 +121,7 @@
 	{ NT_INDUSTRY_OTHER,   NM_THIN,     NF_VIEWPORT|NF_TILE,    NULL                    }, ///< NS_INDUSTRY_OTHER
 	{ NT_INDUSTRY_NOBODY,  NM_THIN,     NF_VIEWPORT|NF_TILE,    NULL                    }, ///< NS_INDUSTRY_NOBODY
 	{ NT_ADVICE,           NM_SMALL,    NF_VIEWPORT|NF_VEHICLE, NULL                    }, ///< NS_ADVICE
+	{ NT_LOST,             NM_SMALL,    NF_VIEWPORT|NF_VEHICLE, NULL                    }, ///< NS_LOST
 	{ NT_NEW_VEHICLES,     NM_NORMAL,   NF_NONE,                DrawNewsNewVehicleAvail }, ///< NS_NEW_VEHICLES
 	{ NT_ACCEPTANCE,       NM_SMALL,    NF_VIEWPORT|NF_TILE,    NULL                    }, ///< NS_ACCEPTANCE
 	{ NT_SUBSIDIES,        NM_NORMAL,   NF_TILE|NF_TILE2,       NULL                    }, ///< NS_SUBSIDIES
@@ -145,6 +146,7 @@
 	{ "production_other",  30, SND_BEGIN,       ND_FULL },  ///< NT_INDUSTRY_OTHER
 	{ "production_nobody", 30, SND_BEGIN,       ND_FULL },  ///< NT_INDUSTRY_NOBODY
 	{ "advice",           150, SND_BEGIN,       ND_FULL },  ///< NT_ADVICE
+	{ "lost",             150, SND_BEGIN,       ND_FULL },  ///< NT_LOST
 	{ "new_vehicles",      30, SND_1E_OOOOH,    ND_FULL },  ///< NT_NEW_VEHICLES
 	{ "acceptance",        90, SND_BEGIN,       ND_FULL },  ///< NT_ACCEPTANCE
 	{ "subsidies",        180, SND_BEGIN,       ND_FULL },  ///< NT_SUBSIDIES
@@ -1000,6 +1002,7 @@
 NEWS_SETTINGS_LINE(26, NT_INDUSTRY_OTHER,   STR_INDUSTRY_CHANGES_SERVED_BY_OTHER),
 NEWS_SETTINGS_LINE(26, NT_INDUSTRY_NOBODY,  STR_OTHER_INDUSTRY_PRODUCTION_CHANGES),
 NEWS_SETTINGS_LINE(26, NT_ADVICE,           STR_020B_ADVICE_INFORMATION_ON_COMPANY),
+NEWS_SETTINGS_LINE(26, NT_LOST,             STR_ADVICE_IS_LOST),
 NEWS_SETTINGS_LINE(26, NT_NEW_VEHICLES,     STR_020C_NEW_VEHICLES),
 NEWS_SETTINGS_LINE(26, NT_ACCEPTANCE,       STR_020D_CHANGES_OF_CARGO_ACCEPTANCE),
 NEWS_SETTINGS_LINE(26, NT_SUBSIDIES,        STR_020E_SUBSIDIES),
Index: src/news_type.h
===================================================================
--- src/news_type.h	(revision 15832)
+++ src/news_type.h	(working copy)
@@ -24,6 +24,7 @@
 	NT_INDUSTRY_OTHER,  ///< Production changes of industry serviced by competitor(s)
 	NT_INDUSTRY_NOBODY, ///< Other industry production changes
 	NT_ADVICE,          ///< Bits of news about vehicles of the company
+	NT_LOST,            ///< A train is lost
 	NT_NEW_VEHICLES,    ///< New vehicle has become available
 	NT_ACCEPTANCE,      ///< A type of cargo is (no longer) accepted
 	NT_SUBSIDIES,       ///< News about subsidies (announcements, expirations, acceptance)
@@ -50,6 +51,7 @@
 	NS_INDUSTRY_OTHER,   ///< NT_INDUSTRY_OTHER
 	NS_INDUSTRY_NOBODY,  ///< NT_INDUSTRY_NOBODY
 	NS_ADVICE,           ///< NT_ADVICE
+	NS_LOST,             ///< NT_LOST
 	NS_NEW_VEHICLES,     ///< NT_NEW_VEHICLES
 	NS_ACCEPTANCE,       ///< NT_ACCEPTANCE
 	NS_SUBSIDIES,        ///< NT_SUBSIDIES
Index: src/train_cmd.cpp
===================================================================
--- src/train_cmd.cpp	(revision 15832)
+++ src/train_cmd.cpp	(working copy)
@@ -2942,7 +2942,7 @@
 					SetDParam(0, v->index);
 					AddNewsItem(
 						STR_TRAIN_IS_LOST,
-						NS_ADVICE,
+						NS_LOST,
 						v->index,
 						0);
 				}
