Index: src/town_cmd.cpp
===================================================================
--- src/town_cmd.cpp	(Revision 18169)
+++ src/town_cmd.cpp	(Arbeitskopie)
@@ -881,29 +881,37 @@
 	/* Optimise only X-junctions */
 	if (rcmd != ROAD_ALL) return rcmd;
 
-	RoadBits rb_template;
+	const Slope tileh = GetTileSlope(tile, NULL);
 
-	switch (GetTileSlope(tile, NULL)) {
-		default:       rb_template = ROAD_ALL; break;
-		case SLOPE_W:  rb_template = ROAD_NW | ROAD_SW; break;
-		case SLOPE_SW: rb_template = ROAD_Y  | ROAD_SW; break;
-		case SLOPE_S:  rb_template = ROAD_SW | ROAD_SE; break;
-		case SLOPE_SE: rb_template = ROAD_X  | ROAD_SE; break;
-		case SLOPE_E:  rb_template = ROAD_SE | ROAD_NE; break;
-		case SLOPE_NE: rb_template = ROAD_Y  | ROAD_NE; break;
-		case SLOPE_N:  rb_template = ROAD_NE | ROAD_NW; break;
-		case SLOPE_NW: rb_template = ROAD_X  | ROAD_NW; break;
-		case SLOPE_STEEP_W:
-		case SLOPE_STEEP_S:
-		case SLOPE_STEEP_E:
-		case SLOPE_STEEP_N:
-			rb_template = ROAD_NONE;
-			break;
+	if (!IsSteepSlope(tileh)) {
+		/* Maximum RoadBits for each slope */
+		static const RoadBits templates[15] = {
+			ROAD_ALL,          // SLOPE_FLAT
+			ROAD_NW | ROAD_SW, // SLOPE_W
+			ROAD_SW | ROAD_SE, // SLOPE_S
+
+			ROAD_Y | ROAD_SW,  // SLOPE_SW
+			ROAD_SE | ROAD_NE, // SLOPE_E
+			ROAD_ALL,          // SLOPE_EW
+
+			ROAD_X | ROAD_SE,  // SLOPE_SE
+			ROAD_ALL,          // SLOPE_WSE
+			ROAD_NE | ROAD_NW, // SLOPE_N
+
+			ROAD_X | ROAD_NW,  // SLOPE_NW
+			ROAD_ALL,          // SLOPE_NS
+			ROAD_ALL,          // SLOPE_ENW
+
+			ROAD_Y | ROAD_NE,  // SLOPE_NE
+			ROAD_ALL,          // SLOPE_SEN
+			ROAD_ALL           // SLOPE_NWS
+		};
+
+		/* Stop if the template is compatible to the growth dir */
+		if (DiagDirToRoadBits(ReverseDiagDir(dir)) & templates[tileh]) return templates[tileh];
 	}
 
-	/* Stop if the template is compatible to the growth dir */
-	if (DiagDirToRoadBits(ReverseDiagDir(dir)) & rb_template) return rb_template;
-	/* If not generate a straight road in the direction of the growth */
+	/* Generate a straight road in the direction of the growth */
 	return DiagDirToRoadBits(dir) | DiagDirToRoadBits(ReverseDiagDir(dir));
 }
 
