diff -r 693edc4a4071 src/rail_gui.cpp
--- a/src/rail_gui.cpp	Tue Jan 13 21:32:47 2009 +0000
+++ b/src/rail_gui.cpp	Tue Jan 13 22:47:59 2009 +0100
@@ -702,6 +702,7 @@ struct BuildRailToolbarWindow : Window {
 	{
 		/* no dragging if you have pressed the convert button */
 		if (FindWindowById(WC_BUILD_SIGNAL, 0) != NULL && _convert_signal_button && this->IsWidgetLowered(RTW_BUILD_SIGNALS)) return;
+		if (_ctrl_pressed) DeleteWindowById(WC_SELECT_STATION, 0);
 
 		VpSelectTilesWithMethod(pt.x, pt.y, select_method);
 	}
