Index: src/fileio.cpp
===================================================================
--- src/fileio.cpp	(revision 26081)
+++ src/fileio.cpp	(working copy)
@@ -761,7 +761,7 @@
 
 	TarHeader th;
 	char buf[sizeof(th.name) + 1], *end;
-	char name[sizeof(th.prefix) + 1 + sizeof(th.name) + 1] = "";
+	char name[sizeof(th.prefix) + 1 + sizeof(th.name) + 1];
 	char link[sizeof(th.linkname) + 1];
 	char dest[sizeof(th.prefix) + 1 + sizeof(th.name) + 1 + 1 + sizeof(th.linkname) + 1];
 	size_t num = 0, pos = 0;
@@ -784,6 +784,8 @@
 			return false;
 		}
 
+		name[0] = '\0';
+
 		/* The prefix contains the directory-name */
 		if (th.prefix[0] != '\0') {
 			ttd_strlcpy(name, th.prefix, lengthof(name));
