Index: src/industry_cmd.cpp
===================================================================
--- src/industry_cmd.cpp	(revision 27826)
+++ src/industry_cmd.cpp	(working copy)
@@ -1383,7 +1383,7 @@
 		}
 
 		if (gfx == GFX_WATERTILE_SPECIALCHECK) {
-			if (!IsTileType(cur_tile, MP_WATER) ||
+			if (!IsTileType(cur_tile, MP_WATER) && !IsBuoyTile(cur_tile) ||
 					!IsTileFlat(cur_tile)) {
 				return_cmd_error(STR_ERROR_SITE_UNSUITABLE);
 			}
