Index: src/ai/api/ai_vehicle.cpp
===================================================================
--- src/ai/api/ai_vehicle.cpp	(revision 16063)
+++ src/ai/api/ai_vehicle.cpp	(working copy)
@@ -157,7 +157,7 @@
 {
 	EnforcePrecondition(false, IsValidVehicle(vehicle_id));
 
-	return AIObject::DoCommand(0, vehicle_id, 0, GetCmdSendToDepot(::GetVehicle(vehicle_id)));
+	return AIObject::DoCommand(0, vehicle_id, DEPOT_DONT_CANCEL, GetCmdSendToDepot(::GetVehicle(vehicle_id)));
 }
 
 /* static */ bool AIVehicle::IsInDepot(VehicleID vehicle_id)
