Index: src/script/api/script_text.cpp
===================================================================
--- src/script/api/script_text.cpp	(revision 27101)
+++ src/script/api/script_text.cpp	(working copy)
@@ -201,7 +201,7 @@
 			p = this->paramt[i]->_GetEncodedText(p, lastofp, param_count);
 			continue;
 		}
-		p += seprintf(p, lastofp,":%X", (uint32)this->parami[i]);
+		p += seprintf(p, lastofp,":" OTTD_PRINTFHEX64, this->parami[i]);
 		param_count++;
 	}
 
Index: src/script/api/script_text.hpp
===================================================================
--- src/script/api/script_text.hpp	(revision 27101)
+++ src/script/api/script_text.hpp	(working copy)
@@ -129,7 +129,7 @@
 private:
 	StringID string;
 	char *params[SCRIPT_TEXT_MAX_PARAMETERS];
-	int parami[SCRIPT_TEXT_MAX_PARAMETERS];
+	int64 parami[SCRIPT_TEXT_MAX_PARAMETERS];
 	ScriptText *paramt[SCRIPT_TEXT_MAX_PARAMETERS];
 	int paramc;
 
Index: src/strings.cpp
===================================================================
--- src/strings.cpp	(revision 27101)
+++ src/strings.cpp	(working copy)
@@ -874,7 +874,7 @@
 						bool lookup = (l == SCC_ENCODED);
 						if (lookup) s += len;
 
-						param = (int32)strtoul(s, &p, 16);
+						param = strtoull(s, &p, 16);
 
 						if (lookup) {
 							if (param >= TAB_SIZE) {
