Index: src/misc_gui.cpp
===================================================================
--- src/misc_gui.cpp	(revisi¢n: 17491)
+++ src/misc_gui.cpp	(copia de trabajo)
@@ -1182,23 +1182,41 @@

 void QueryString::DrawEditBox(Window *w, int wid)
 {
-	const Widget *wi = &w->widget[wid];
+	uint left;
+	uint right;
+	uint top;
+	uint bottom;
+	if (w->widget == NULL) {
+		const NWidgetCore *wi = w->nested_array[wid];

-	assert((wi->type & WWT_MASK) == WWT_EDITBOX);
+		assert((wi->type & WWT_MASK) == WWT_EDITBOX);

-	GfxFillRect(wi->left + 1, wi->top + 1, wi->right - 1, wi->bottom - 1, 215);
+		left   = wi->pos_x;
+		right  = wi->pos_x + wi->current_x - 1;
+		top    = wi->pos_y;
+		bottom = wi->pos_y + wi->current_y - 1;
+	} else {
+		const Widget *wi = &w->widget[wid];

+		assert((wi->type & WWT_MASK) == WWT_EDITBOX);
+
+		left   = wi->left;
+		right  = wi->right;
+		top    = wi->top;
+		bottom = wi->bottom;
+	}
+
+	GfxFillRect(left + 1, top + 1, right - 1, bottom - 1, 215);
+
 	DrawPixelInfo dpi;
 	int delta;

 	/* Limit the drawing of the string inside the widget boundaries */
 	if (!FillDrawPixelInfo(&dpi,
-			wi->left + 4,
-			wi->top + 1,
-			wi->right - wi->left - 4,
-			wi->bottom - wi->top - 1)) {
-		return;
-	}
+			left + WD_FRAMERECT_LEFT + WD_FRAMERECT_RIGHT,
+			top + WD_FRAMERECT_TOP,
+			right - left - WD_FRAMERECT_LEFT - WD_FRAMERECT_RIGHT,
+			bottom - top - WD_FRAMERECT_BOTTOM)) return;

 	DrawPixelInfo *old_dpi = _cur_dpi;
 	_cur_dpi = &dpi;
@@ -1207,7 +1225,7 @@
 	 * space reserved at the end for the caret to show */
 	const Textbuf *tb = &this->text;

-	delta = (wi->right - wi->left) - tb->width - 10;
+	delta = (right - left) - tb->width - 10;
 	if (delta > 0) delta = 0;

 	if (tb->caretxoffs + delta < 0) delta = -tb->caretxoffs;
Index: src/querystring_gui.h
===================================================================
--- src/querystring_gui.h	(revisi¢n: 17491)
+++ src/querystring_gui.h	(copia de trabajo)
@@ -62,6 +62,12 @@
 	char *orig_str_buf;
 	const uint16 edit_str_size; ///< maximum length of string (in bytes), including terminating '\0'

+	QueryStringBaseWindow(uint16 size) : Window(), edit_str_size(size)
+	{
+		assert(size != 0);
+		this->edit_str_buf = CallocT<char>(size);
+	}
+
 	QueryStringBaseWindow(uint16 size, const WindowDesc *desc, WindowNumber window_number = 0) : Window(desc, window_number), edit_str_size(size)
 	{
 		assert(size != 0);
Index: src/window_gui.h
===================================================================
--- src/window_gui.h	(revisi¢n: 17491)
+++ src/window_gui.h	(copia de trabajo)
@@ -503,15 +503,29 @@
 	 */
 	inline bool SetFocusedWidget(byte widget_index)
 	{
-		if (widget_index >= this->widget_count || this->widget + widget_index == this->focused_widget) {
-			return false;
+		/* Do nothing if widget_index is already focused, or if it wasn't a valid widget */
+		if (this->widget != NULL && (widget_index >= this->widget_count || this->widget + widget_index == this->focused_widget)) return false;
+		if (this->nested_array != NULL) {
+			if (widget_index >= this->nested_array_size) return false;
+			if (this->nested_focus != NULL && this->nested_array[widget_index] == this->nested_focus) return false;
 		}

-		if (this->focused_widget != NULL) {
-			/* Repaint the widget that lost focus. A focused edit box may else leave the caret on the screen. */
-			this->InvalidateWidget(this->focused_widget - this->widget);
+		/* Repaint the widget that lost focus. A focused edit box may else leave the caret on the screen. */
+		if (this->nested_array != NULL) {
+			assert(widget_index < this->nested_array_size);
+			if (this->nested_focus != NULL) {
+				this->nested_focus->Invalidate(this);
+			}
+			assert(this->nested_array[widget_index] != NULL);
+			this->nested_focus = this->nested_array[widget_index];
 		}
-		this->focused_widget = &this->widget[widget_index];
+		if (this->widget != NULL) {
+			assert(widget_index < this->widget_count);
+			if (this->focused_widget != NULL) {
+				this->InvalidateWidget(this->focused_widget - this->widget);
+			}
+			this->focused_widget = &this->widget[widget_index];
+		}
 		return true;
 	}

Index: src/osk_gui.cpp
===================================================================
--- src/osk_gui.cpp	(revisi¢n: 17491)
+++ src/osk_gui.cpp	(copia de trabajo)
@@ -75,7 +75,13 @@
 		this->parent = parent;
 		assert(parent != NULL);

-		this->caption = (parent->widget[button].data != STR_NULL) ? parent->widget[button].data : parent->caption;
+		if (parent->widget != NULL) {
+			this->caption = (parent->widget[button].data != STR_NULL) ? parent->widget[button].data : parent->caption;
+		}
+		if (parent->nested_array != NULL) {
+			assert(parent->nested_array[button] != NULL);
+			this->caption = (parent->nested_array[button]->widget_data != STR_NULL) ? parent->nested_array[button]->widget_data : parent->caption;
+		}

 		this->qs         = parent;
 		this->text_btn   = button;
