Index: os/rpm/openttd.spec
===================================================================
--- os/rpm/openttd.spec	(Revision 14505)
+++ os/rpm/openttd.spec	(Arbeitskopie)
@@ -1,75 +1,162 @@
+# $id$
+#-------------------------------------------------------------------------------
+# spec file for the openttd rpm package
 #
-# spec file for package openttd (trunk)
+# Copyright (c) 2007-2008 The OpenTTD developers
 #
-# Copyright (c) 2007 The OpenTTD team.
 # This file and all modifications and additions to the pristine
 # package are under the same license as the package itself
 #
+#-------------------------------------------------------------------------------
+
 Name:          openttd
-Version:       svn
-Release:       head
-Group:         Applications/Games
-Source:        %{name}-%{version}-%{release}.tar.gz
-License:       GPL
+Version:       0.7.0
+Release:       1%{?dist}
+Group:         Amusements/Games
+Source:        %{name}-%{version}.tar.bz2
+License:       GPLv2
 URL:           http://www.openttd.org
-Packager:      Denis Burlaka <burlaka@yandex.ru>
 Summary:       OpenTTD is an Open Source clone of Chris Sawyer's Transport Tycoon Deluxe
-Requires:      SDL zlib libpng freetype2 fontconfig
-BuildRequires: gcc SDL-devel zlib-devel libpng-devel fontconfig-devel
-%if %{_vendor}=="suse"
-BuildRequires: freetype2-devel
-%endif
-%if %{_vendor}=="fedora"
+
+Requires:      fontconfig
+Requires:      libicu
+Requires:      libpng
+Requires:      SDL
+Requires:      zlib
+BuildRequires: fontconfig-devel
+BuildRequires: libpng-devel
+BuildRequires: libicu-devel
+BuildRequires: SDL-devel
+BuildRequires: zlib-devel
+# vendor specific dependencies
+%if %{_vendor}=="alt"
+# -> Alt Linux: Andrew Kornilov <hiddenman@altlinux.ru>
+Requires:      freetype
 BuildRequires: freetype-devel
 %endif
-%if %{_vendor}=="mandriva"
+%if %{_vendor}=="MandrakeSoft" || %{_vendor}=="mandriva"
+# -> Mandriva: Gustavo Pichorim Boiko <boiko@mandriva.com>
+Requires:      freetype2
 BuildRequires: libfreetype6-devel
 %endif
+%if %{_vendor}=="redhat" || %{_vendor}=="fedora"
+Requires:      freetype
+BuildRequires: freetype-devel
+BuildRequires: desktop-file-utils
+%endif
+%if %{_vendor}=="suse"
+# -> packman: Toni Graffy <toni@links2linux.de>
+Requires:      freetype2
+BuildRequires: freetype2-devel
+%endif
+
 BuildRoot:     %{_tmppath}/%{name}-%{version}-%{release}-buildroot
-Prefix:        /usr
 
 %description
-OpenTTD is a clone of the Microprose game "Transport Tycoon Deluxe", a popular game originally written by Chris Sawyer. It attempts to mimic the original game as closely as possible while extending it with new features.
+OpenTTD is a reimplementation of the Microprose game "Transport Tycoon Deluxe"
+with lots of new features and enhancements. To play the game you need to copy
+the following 6 data files from your Transport Tycoon Deluxe CD to the game
+data directory in %{_datadir}/games/%{name}/data:
 
-OpenTTD is licensed under the GNU General Public License version 2.0. For more information, see the file 'COPYING' included with every release and source download of the game.
+From the Windows version of TTD you need:
+sample.cat trg1r.grf trgcr.grf trghr.grf trgir.grf trgtr.grf
 
+Or if you have the DOS version you need:
+sample.cat TRG1.GRF TRGC.GRF TRGH.GRF TRGI.GRF TRGT.GRF
+
+OpenTTD is licensed under the GNU General Public License version 2.0. For more
+information, see the file 'COPYING' included with every release and source
+download of the game.
+
 %prep
-%setup
+%setup -q
 
 %build
-./configure --prefix-dir=%{prefix} --binary-dir=bin --install-dir="$RPM_BUILD_ROOT"
-make
+./configure \
+	--revision=%{version} \
+	--prefix-dir="%{_prefix}" \
+	--binary-dir="bin" \
+	--binary-name="%{name}" \
+#	--menu-name="OpenTTD" \
+#	--data-dir="share\games\%{name}" \
+#	--doc-dir="share\doc\%{name}" \
+#	--icon-dir="share/pixmaps" \
+#	--icon-theme-dir="share/icons/hicolor" \
+#	--man-dir="share/man/man6" \
+#	--menu-dir="share/applications" \
+	--enable-debug=0 \
+#	--menu_group="Game;" \
+	--with-sdl \
+	--with-zlib \
+	--with-png \
+	--with-freetype \
+	--with-fontconfig \
+	--with-icu \
+	--enable-strip
+make %{?_smp_mflags}
 
 %install
-make ROOT="$RPM_BUILD_ROOT" install
+rm -rf "%{buildroot}"
+make install INSTALL_DIR="%{buildroot}"
 
-mkdir -p $RPM_BUILD_ROOT/%{_datadir}/applications
-cat << EOF > $RPM_BUILD_ROOT/%{_datadir}/applications/%{name}.desktop
-[Desktop Entry]
-Categories=Games;
-Encoding=UTF-8
-Exec=/usr/bin/openttd
-Name=OpenTTD
-Icon=openttd.32
-Terminal=false
-Type=Application
-EOF
+%if %{_vendor} == "redhat" || %{_vendor}=="fedora"
+desktop-file-install \
+#	--delete-original \
+	--vendor="%{_vendor}" \
+	--remove-key Version \
+	--dir="%{buildroot}/%{_datadir}/applications/" \
+	"%{buildroot}/%{_datadir}/applications/%{name}.desktop"
+%endif
 
 %clean
-rm -Rf "$RPM_BUILD_ROOT"
+rm -rf "%{buildroot}"
 
+%post
+touch --no-create %{_datadir}/icons/hicolor
+if [ -x %{_bindir}/gtk-update-icon-cache ]; then
+	%{_bindir}/gtk-update-icon-cache --quiet %{_datadir}/icons/hicolor || :
+fi
+
+%postun
+touch --no-create %{_datadir}/icons/hicolor
+if [ -x %{_bindir}/gtk-update-icon-cache ]; then
+	%{_bindir}/gtk-update-icon-cache --quiet %{_datadir}/icons/hicolor || :
+fi
+
 %files
+%defattr(-, root, games, -)
+%dir %{_datadir}/doc/%{name}
 %dir %{_datadir}/games/%{name}
 %dir %{_datadir}/games/%{name}/lang
 %dir %{_datadir}/games/%{name}/data
 %dir %{_datadir}/games/%{name}/gm
-%dir %{_datadir}/games/%{name}/docs
-%dir %{_datadir}/pixmaps
-%defattr(644, root, games, 755)
 %attr(755, root, games) %{_bindir}/%{name}
+%{_datadir}/doc/%{name}/*
 %{_datadir}/games/%{name}/lang/*
 %{_datadir}/games/%{name}/data/*
-%{_datadir}/games/%{name}/docs/*
+%{_datadir}/applications/%{name}.desktop
 %{_datadir}/pixmaps/*
-%{_datadir}/applications/%{name}.desktop
+%{_datadir}/icons/hicolor/*/apps/openttd.png
+%doc %{_mandir}/man6/openttd.6.gz
 
+%changelog
+* Mon Oct 20 2008 Benedikt Brüggemeier <skidd13@openttd.org>
+
+- Added libicu dependency
+
+* Thu Sep 23 2008 Benedikt Brüggemeier <skidd13@openttd.org>
+
+- Merged both versions of the spec file
+
+* Fri Aug 29 2008 Jonathan Coome <maedhros@openttd.org>
+
+- Rewrite spec file from scratch.
+
+* Sat Aug 02 2008 Benedikt Brüggemeier <skidd13@openttd.org>
+
+- Updated spec file
+
+* Thu Mar 27 2008 Denis Burlaka <burlaka@yandex.ru>
+
+- Universal spec file
+
