Index: src/video/cocoa/wnd_quartz.mm
===================================================================
--- src/video/cocoa/wnd_quartz.mm	(revision 20612)
+++ src/video/cocoa/wnd_quartz.mm	(working copy)
@@ -456,6 +456,8 @@
 		[ this->window setMiniwindowTitle:nsscaption ];
 		[ nsscaption release ];
 
+		[ this->window setContentMinSize:NSMakeSize(64.0f,64.0f) ];
+
 		[ this->window setAcceptsMouseMovedEvents:YES ];
 		[ this->window setViewsNeedDisplay:NO ];
 
Index: src/video/cocoa/wnd_quickdraw.mm
===================================================================
--- src/video/cocoa/wnd_quickdraw.mm	(revision 20612)
+++ src/video/cocoa/wnd_quickdraw.mm	(working copy)
@@ -386,6 +386,8 @@
 		[ this->window setMiniwindowTitle:nsscaption ];
 		[ nsscaption release ];
 
+		[ this->window setContentMinSize:NSMakeSize(64.0f,64.0f) ];
+
 		[ this->window setAcceptsMouseMovedEvents:YES ];
 		[ this->window setViewsNeedDisplay:NO ];
 
