diff --git a/src/settings_gui.cpp b/src/settings_gui.cpp
index 7ddc725..598ff74 100644
--- a/src/settings_gui.cpp
+++ b/src/settings_gui.cpp
@@ -1753,6 +1753,7 @@ static SettingEntry _settings_ui_news[] = {
 	SettingEntry("news_display.acceptance"),
 	SettingEntry("news_display.subsidies"),
 	SettingEntry("news_display.general"),
+	SettingEntry("gui.coloured_news_year"),
 	SettingEntry("misc.news_ticker_sound"),
 };
 
@@ -1777,7 +1778,6 @@ static SettingEntry _settings_ui[] = {
 	SettingEntry("gui.default_rail_type"),
 	SettingEntry("gui.disable_unsuitable_building"),
 	SettingEntry("gui.persistent_buildingtools"),
-	SettingEntry("gui.coloured_news_year"),
 };
 /** Interface subpage */
 static SettingsPage _settings_ui_page = {_settings_ui, lengthof(_settings_ui)};
