Index: src/rail_cmd.cpp
===================================================================
--- src/rail_cmd.cpp	(revision 16121)
+++ src/rail_cmd.cpp	(working copy)
@@ -818,6 +818,7 @@
 	SignalType cycle_stop = (SignalType)GB(p1, 12, 3);
 	CommandCost cost;
 	uint num_dir_cycle = GB(p1, 15, 2);
+	if (IsTileType(tile, MP_TUNNELBRIDGE)) DEBUG(misc,0,"%u",GetTunnelBridgeLength(tile, GetOtherTunnelBridgeEnd(tile) + 2));
 
 	if (sigtype > SIGTYPE_LAST) return CMD_ERROR;
 
Index: src/tunnelbridge.h
===================================================================
--- src/tunnelbridge.h	(revision 16121)
+++ src/tunnelbridge.h	(working copy)
@@ -18,7 +18,7 @@
 	int x2 = TileX(end);
 	int y2 = TileY(end);
 
-	return abs(x2 + y2 - x1 - y1) - 1;
+	return abs((x2 + y2 - x1 - y1) - 2) - 1;
 }
 
 extern TileIndex _build_tunnel_endtile;
