diff --git a/src/table/animcursors.h b/src/table/animcursors.h
index 14d22b7..08a9075 100644
--- a/src/table/animcursors.h
+++ b/src/table/animcursors.h
@@ -30,7 +30,7 @@
 #define ANIM_CURSOR_END() ANIM_CURSOR_LINE(AnimCursor::LAST, 0)
 
 /**
- * Animated cursor elements for demolishion
+ * Animated cursor elements for demolition
  */
 static const AnimCursor _demolish_animcursor[] = {
 	ANIM_CURSOR_LINE(SPR_CURSOR_DEMOLISH_FIRST, 8)
diff --git a/src/table/bridge_land.h b/src/table/bridge_land.h
index f1ff8b4..4f6f94b 100644
--- a/src/table/bridge_land.h
+++ b/src/table/bridge_land.h
@@ -13,7 +13,7 @@
  * <ul><li>_bridge_sprite_table_n_m. Defines all the sprites of a bridge besides the pylons.
  * n defines the number of the bridge type, m the number of the section. the highest m for
  * each bridge set defines the heads.<br>
- * Sprites for middle secionts are arranged in groups of four, the elements are:
+ * Sprites for middle sections are arranged in groups of four, the elements are:
  * <ol><li>Element containing the track. This element is logically behind the vehicle.</li>
  * <li>Element containing the structure that is logically between the vehicle and the camera</li>
  * <li>Element containing the pylons.</li></ol>
@@ -729,7 +729,7 @@ static const PalSpriteID * const * const _bridge_sprite_table[MAX_BRIDGES] = {
 
 /**
  * Describes the data that defines each bridge in the game
- * @param y   year of availablity
+ * @param y   year of availability
  * @param mnl minimum length (not counting bridge heads)
  * @param mxl maximum length (not counting bridge heads)
  * @param p   price multiplier
@@ -745,7 +745,7 @@ static const PalSpriteID * const * const _bridge_sprite_table[MAX_BRIDGES] = {
 
 const BridgeSpec _orig_bridge[] = {
 /*
-	       year of availablity
+	       year of availability
 	       |  minimum length
 	       |  |   maximum length
 	       |  |   |        price multiplier
diff --git a/src/table/build_industry.h b/src/table/build_industry.h
index bcb6e28..9c9a28d 100644
--- a/src/table/build_industry.h
+++ b/src/table/build_industry.h
@@ -1171,7 +1171,7 @@ enum IndustryTypes {
  * @param a3   accepted cargo 3
  * @param im3  input multiplier for cargo 3
  * @param pr   industry life (actually, the same as extractive, organic, processing in ttdpatch's specs)
- * @param clim climate availaility
+ * @param clim climate availability
  * @param bev  industry behaviour
  * @param in   name
  * @param intx text while building
diff --git a/src/table/elrail_data.h b/src/table/elrail_data.h
index 9ded0c3..156f0e9 100644
--- a/src/table/elrail_data.h
+++ b/src/table/elrail_data.h
@@ -117,7 +117,7 @@ static const byte PreferredPPPofTrackAtPCP[TRACK_END][DIAGDIR_END] = {
 #define NUM_IGNORE_GROUPS 3
 #define IGNORE_NONE 0xFF
 /**
- * In case we have a staight line, we place pylon only every two tiles,
+ * In case we have a straight line, we place pylon only every two tiles,
  * so there are certain tiles which we ignore. A straight line is found if
  * we have exactly two PPPs.
  */
@@ -195,7 +195,7 @@ static const byte IgnoredPCP[NUM_IGNORE_GROUPS][TLG_END][DIAGDIR_END] = {
 
 #undef NO_IGNORE
 
-/** Which pylons can definately NOT be built */
+/** Which pylons can definitely NOT be built */
 static const byte DisallowedPPPofTrackAtPCP[TRACK_END][DIAGDIR_END] = {
 	{1 << DIR_SW | 1 << DIR_NE, 0,           1 << DIR_SW | 1 << DIR_NE, 0          }, // X
 	{0,           1 << DIR_NW | 1 << DIR_SE, 0,           1 << DIR_NW | 1 << DIR_SE}, // Y
diff --git a/src/table/sprites.h b/src/table/sprites.h
index 5388883..eefbc5e 100644
--- a/src/table/sprites.h
+++ b/src/table/sprites.h
@@ -9,7 +9,7 @@
 
 /**
  * @file sprites.h
- * This file contails all sprite-related enums and defines. These consist mainly of
+ * This file contains all sprite-related enums and defines. These consist mainly of
  * the sprite numbers and a bunch of masks and macros to handle sprites and to get
  * rid of all the magic numbers in the code.
  *
@@ -26,7 +26,7 @@
  *
  * All sprites which are described here are referenced only one to a handful of times
  * throughout the code. When introducing new sprite enums, use meaningful names.
- * Don't be lazy and typing, and only use abbrevations when their meaning is clear or
+ * Don't be lazy and typing, and only use abbreviations when their meaning is clear or
  * the length of the enum would get out of hand. In that case EXPLAIN THE ABBREVATION
  * IN THIS FILE, and perhaps add some comments in the code where it is used.
  * Now, don't whine about this being too much typing work if the enums are like
@@ -709,7 +709,7 @@ static const SpriteID SPR_BTSGA_MGLV_Y_REAR       = 4365;
  * TILE_* denotes the different tiles a suspension bridge
  * can have
  * TILE_A and TILE_B are the "beginnings" and "ends" of the
- *   suspension system. they have small rectangluar endcaps
+ *   suspension system. they have small rectangular endcaps
  * TILE_C and TILE_D look almost identical to TILE_A and
  *   TILE_B, but they do not have the "endcaps". They form the
  *   middle part
@@ -860,7 +860,7 @@ static const SpriteID SPR_BTGIR_MGLV_Y        = 4403;
  * tubular bridges have 3 kinds of tiles:
  *  a start tile (with only half a tube on the far side, marked _BEG
  *  a middle tile (full tunnel), marked _MID
- *  and an end tile (half a tube on the near side, maked _END
+ *  and an end tile (half a tube on the near side, marked _END
  */
 static const SpriteID SPR_BTTUB_X_FRONT_BEG       = 2559;
 static const SpriteID SPR_BTTUB_X_FRONT_MID       = 2560;
diff --git a/src/table/strgen_tables.h b/src/table/strgen_tables.h
index d71a46f..14de3a4 100644
--- a/src/table/strgen_tables.h
+++ b/src/table/strgen_tables.h
@@ -197,7 +197,7 @@ static const char * const _pragmas[][4] = {
 	{ "plural",      "tp",  "0",      "Plural form to use" },
 	{ "textdir",     "tl",  "ltr",    "Text direction. Either ltr (left-to-right) or rtl (right-to-left)" },
 	{ "digitsep",    "td",  ",",      "Digit grouping separator for non-currency numbers" },
-	{ "digitsepcur", "td",  ",",      "Digit grouping seprarator for currency numbers" },
+	{ "digitsepcur", "td",  ",",      "Digit grouping separator for currency numbers" },
 	{ "decimalsep",  "td",  ".",      "Decimal separator" },
 	{ "winlangid",   "x2",  "0x0000", "Language ID for Windows" },
 	{ "grflangid",   "x1",  "0x00",   "Language ID for NewGRFs" },
diff --git a/src/table/townname.h b/src/table/townname.h
index 4c81c48..6e572b5 100644
--- a/src/table/townname.h
+++ b/src/table/townname.h
@@ -1729,7 +1729,7 @@ static const char * const _name_czech_patmod[][3] = {
 	/* CZG_PNEUT */ { "\xC3\xAD", "\xC3\xA1", "a" }
 };
 
-/* This way the substantives can choose only some adjectives/endings:
+/* This way the substantive can choose only some adjectives/endings:
  * At least one of these flags must be satisfied: */
 enum CzechAllow {
 	CZA_SHORT = 1,
