Index: train_cmd.c
===================================================================
--- train_cmd.c	(revision 4467)
+++ train_cmd.c	(working copy)
@@ -1957,6 +1957,10 @@
 		if (IsTileDepotType(v->tile, TRANSPORT_RAIL) || IsTunnelTile(v->tile))
 			continue;
 
+		// No sparks for electric vehicles on nonelectrified tracks
+		if (GetEngine(engtype)->railtype == RAILTYPE_ELECTRIC && !HasPowerOnRail(v->u.rail.railtype, GetTileRailType(v->tile, GetVehicleTrackdir(v))))
+			continue;
+
 		if (effect_type == 0) {
 			// Use default effect type for engine class.
 			effect_type = RailVehInfo(engtype)->engclass;
