diff -r 0af36eb10c3c src/ai/api/ai_airport.cpp
--- a/src/ai/api/ai_airport.cpp	Sat Jul 30 12:03:25 2011 +0200
+++ b/src/ai/api/ai_airport.cpp	Sat Jul 30 22:10:09 2011 +0200
@@ -131,7 +131,7 @@
 	extern uint8 GetAirportNoiseLevelForTown(const AirportSpec *as, TileIndex town_tile, TileIndex tile);
 
 	if (!::IsValidTile(tile)) return -1;
-	if (!IsValidAirportType(type)) return -1;
+	if (!IsAirportInformationAvailable(type)) return -1;
 
 	if (_settings_game.economy.station_noise_level) {
 		const AirportSpec *as = ::AirportSpec::Get(type);
