diff -r 08633bf16195 src/town_cmd.cpp
--- a/src/town_cmd.cpp	Sun Mar 01 00:30:53 2015 +0300
+++ b/src/town_cmd.cpp	Sun Mar 01 04:19:31 2015 +0300
@@ -1300,7 +1300,7 @@
  * Returns "growth" if a house was built, or no if the build failed.
  * @param t town to inquiry
  * @param tile to inquiry
- * @return something other than zero(0)if town expansion was possible
+ * @return something other than zero(0) if town expansion was possible
  */
 static int GrowTownAtRoad(Town *t, TileIndex tile)
 {
@@ -1339,7 +1339,7 @@
 		 * and return if no more road blocks available */
 		if (IsValidDiagDirection(target_dir)) cur_rb &= ~DiagDirToRoadBits(ReverseDiagDir(target_dir));
 		if (cur_rb == ROAD_NONE) {
-			return _grow_town_result;
+			return (_grow_town_result == GROWTH_SUCCEED);
 		}
 
 		if (IsTileType(tile, MP_TUNNELBRIDGE)) {
@@ -1355,7 +1355,7 @@
 		if (IsTileType(tile, MP_ROAD) && !IsRoadDepot(tile) && HasTileRoadType(tile, ROADTYPE_ROAD)) {
 			/* Don't allow building over roads of other cities */
 			if (IsRoadOwner(tile, ROADTYPE_ROAD, OWNER_TOWN) && Town::GetByTile(tile) != t) {
-				_grow_town_result = GROWTH_SUCCEED;
+				_grow_town_result = GROWTH_SEARCH_STOPPED;
 			} else if (IsRoadOwner(tile, ROADTYPE_ROAD, OWNER_NONE) && _game_mode == GM_EDITOR) {
 				/* If we are in the SE, and this road-piece has no town owner yet, it just found an
 				 * owner :) (happy happy happy road now) */
@@ -1389,7 +1389,7 @@
 /**
  * Grow the town
  * @param t town to grow
- * @return true iff a house was built
+ * @return true iff something was built (house, road or bridge)
  */
 static bool GrowTown(Town *t)
 {
