diff -r 4c84891cc6b6 src/lang/english.txt
--- a/src/lang/english.txt	Sun Nov 23 15:35:10 2008 +0100
+++ b/src/lang/english.txt	Sun Nov 23 15:46:11 2008 +0100
@@ -1225,8 +1225,11 @@
 
 STR_CONFIG_PATCHES_TITLE_STATIONS                               :{ORANGE}Stations
 STR_CONFIG_PATCHES_TITLE_STATIONS_AIRPORT                       :{ORANGE}Airports
+STR_CONFIG_PATCHES_TITLE_STATIONS_CARGO                         :{ORANGE}Cargo handling
+
 STR_CONFIG_PATCHES_TITLE_ECONOMY                                :{ORANGE}Economy
 STR_CONFIG_PATCHES_TITLE_TOWNS                                  :{ORANGE}Towns
+STR_CONFIG_PATCHES_TITLE_INDUSTRIES                             :{ORANGE}Industries
 
 STR_CONFIG_PATCHES_TITLE_COMPETITORS                            :{ORANGE}Competitors
 STR_CONFIG_PATCHES_TITLE_COMPETITORS_AI                         :{ORANGE}Computer players
diff -r 4c84891cc6b6 src/settings_gui.cpp
--- a/src/settings_gui.cpp	Sun Nov 23 15:35:10 2008 +0100
+++ b/src/settings_gui.cpp	Sun Nov 23 15:46:11 2008 +0100
@@ -1024,7 +1024,6 @@
 	PatchEntry("gui.errmsg_duration"),
 	PatchEntry("gui.toolbar_pos"),
 	PatchEntry("gui.link_terraform_toolbar"),
-	PatchEntry("gui.liveries"),
 	PatchEntry("gui.pause_on_newgame"),
 	PatchEntry("gui.advanced_vehicle_list"),
 	PatchEntry("gui.timetable_in_ticks"),
@@ -1059,6 +1058,14 @@
 
 /* == STATIONS PAGE == */
 
+static PatchEntry _patches_stations_cargo[] = {
+	PatchEntry("gui.liveries"),
+	PatchEntry("order.improved_load"),
+	PatchEntry("order.selectgoods"),
+	PatchEntry("order.gradual_loading"),
+};
+static PatchPage _patches_stations_cargo_page = {_patches_stations_cargo, lengthof(_patches_stations_cargo)};
+
 static PatchEntry _patches_stations_airport[] = {
 	PatchEntry("station.always_small_airport"),
 	PatchEntry("order.serviceathelipad"),
@@ -1067,15 +1074,13 @@
 static PatchPage _patches_stations_airport_page = {_patches_stations_airport, lengthof(_patches_stations_airport)};
 
 static PatchEntry _patches_stations[] = {
+	PatchEntry(&_patches_stations_cargo_page, STR_CONFIG_PATCHES_TITLE_STATIONS_CARGO),
 	PatchEntry(&_patches_stations_airport_page, STR_CONFIG_PATCHES_TITLE_STATIONS_AIRPORT),
 	PatchEntry("station.join_stations"),
-	PatchEntry("order.improved_load"),
-	PatchEntry("order.selectgoods"),
 	PatchEntry("gui.new_nonstop"),
 	PatchEntry("station.nonuniform_stations"),
 	PatchEntry("station.station_spread"),
 	PatchEntry("station.modified_catchment"),
-	PatchEntry("order.gradual_loading"),
 	PatchEntry("construction.road_stop_on_town_road"),
 	PatchEntry("station.adjacent_stations"),
 };
@@ -1094,13 +1099,19 @@
 };
 static PatchPage _patches_town_page = {_patches_town, lengthof(_patches_town)};
 
+/* == INDUSTRIES PAGE == */
+
+static PatchEntry _patches_industries[] = {
+	PatchEntry("construction.raw_industry_construction"),
+	PatchEntry("economy.multiple_industry_per_town"),
+	PatchEntry("economy.same_industry_close"),
+};
+static PatchPage _patches_industries_page = {_patches_industries, lengthof(_patches_industries)};
+
 /* == ECONOMY PAGE == */
 
 static PatchEntry _patches_economy[] = {
 	PatchEntry("economy.inflation"),
-	PatchEntry("construction.raw_industry_construction"),
-	PatchEntry("economy.multiple_industry_per_town"),
-	PatchEntry("economy.same_industry_close"),
 	PatchEntry("gui.colored_news_year"),
 	PatchEntry("gui.ending_year"),
 	PatchEntry("economy.smooth_economy"),
@@ -1182,6 +1193,7 @@
 	PatchEntry(&_patches_vehicles_page,     STR_CONFIG_PATCHES_TITLE_VEHICLES),
 	PatchEntry(&_patches_stations_page,     STR_CONFIG_PATCHES_TITLE_STATIONS),
 	PatchEntry(&_patches_town_page,         STR_CONFIG_PATCHES_TITLE_TOWNS),
+	PatchEntry(&_patches_industries_page,   STR_CONFIG_PATCHES_TITLE_INDUSTRIES),
 	PatchEntry(&_patches_economy_page,      STR_CONFIG_PATCHES_TITLE_ECONOMY),
 	PatchEntry(&_patches_competitors_page,  STR_CONFIG_PATCHES_TITLE_COMPETITORS),
 };
