--- 0.4.7.orig/macros.h	2006-03-18 18:31:34.000000000 +0300
+++ 0.4.7/macros.h	2006-06-03 03:02:57.000000000 +0400
@@ -115,9 +115,9 @@ static inline int KillFirstBit2x64(int v
 #define IS_INT_INSIDE(a,min,max) ((uint)((a)-(min)) < (uint)((max)-(min)))
 
 
-#define CHANCE16(a,b) ((uint16)Random() <= (uint16)((65536 * a) / b))
-#define CHANCE16R(a,b,r) ((uint16)(r=Random()) <= (uint16)((65536 * a) / b))
-#define CHANCE16I(a,b,v) ((uint16)(v) <= (uint16)((65536 * a) / b))
+#define CHANCE16(a,b) ((uint16)Random() <= (uint16)((65536 * (a)) / (b)))
+#define CHANCE16R(a,b,r) ((uint16)((r)=Random()) <= (uint16)((65536 * (a)) / (b)))
+#define CHANCE16I(a,b,v) ((uint16)(v) <= (uint16)((65536 * (a)) / (b)))
 
 
 #define for_each_bit(_i,_b)										\
