Index: src/road.cpp
===================================================================
--- src/road.cpp	(revision 11939)
+++ src/road.cpp	(working copy)
@@ -96,11 +96,13 @@
 {
 	RoadTypes rt = ROADTYPES_NONE;
 
-	for (EngineID i = 0; i != TOTAL_NUM_ENGINES; i++) {
+	EngineID i;
+	FOR_ALL_ENGINEIDS_OF_TYPE(i, VEH_ROAD) {
 		const Engine* e = GetEngine(i);
 		const EngineInfo *ei = EngInfo(i);
 
-		if (e->type == VEH_ROAD && HasBit(ei->climates, _opt.landscape) &&
+		assert(e->type == VEH_ROAD);
+		if (HasBit(ei->climates, _opt.landscape) &&
 				(HasBit(e->player_avail, p) || _date >= e->intro_date + 365)) {
 			SetBit(rt, HasBit(ei->misc_flags, EF_ROAD_TRAM) ? ROADTYPE_TRAM : ROADTYPE_ROAD);
 		}
Index: src/rail.cpp
===================================================================
--- src/rail.cpp	(revision 11939)
+++ src/rail.cpp	(working copy)
@@ -167,11 +167,13 @@
 {
 	RailTypes rt = RAILTYPES_NONE;
 
-	for (EngineID i = 0; i != TOTAL_NUM_ENGINES; i++) {
+	EngineID i;
+	FOR_ALL_ENGINEIDS_OF_TYPE(i, VEH_TRAIN) {
 		const Engine* e = GetEngine(i);
 		const EngineInfo *ei = EngInfo(i);
 
-		if (e->type == VEH_TRAIN && HasBit(ei->climates, _opt.landscape) &&
+		assert(e->type == VEH_TRAIN);
+		if (HasBit(ei->climates, _opt.landscape) &&
 				(HasBit(e->player_avail, p) || _date >= e->intro_date + 365)) {
 			const RailVehicleInfo *rvi = RailVehInfo(i);
 
