Index: src/map.cpp
===================================================================
--- src/map.cpp	(revision 21000)
+++ src/map.cpp	(working copy)
@@ -224,6 +224,23 @@
 }
 
 /*!
+ * Gets the distance to the edge of the map in given direction.
+ * @param tile the tile to get the distance from
+ * @param diagdir the direction of interest
+ * @return the distance from the edge in tiles
+ */
+uint DistanceFromEdgeDir(TileIndex tile, DiagDirection dir)
+{
+	switch(dir) {
+		case DIAGDIR_NE: return             TileX(tile) - 1;
+		case DIAGDIR_NW: return             TileY(tile) - 1;
+		case DIAGDIR_SW: return MapMaxX() - TileX(tile) - 1;
+		case DIAGDIR_SE: return MapMaxY() - TileY(tile) - 1;
+		default: NOT_REACHED();
+	}
+}
+
+/*!
  * Function performing a search around a center tile and going outward, thus in circle.
  * Although it really is a square search...
  * Every tile will be tested by means of the callback function proc,
Index: src/map_func.h
===================================================================
--- src/map_func.h	(revision 21000)
+++ src/map_func.h	(working copy)
@@ -331,6 +331,7 @@
 uint DistanceMax(TileIndex, TileIndex); ///< also known as L-Infinity-Norm
 uint DistanceMaxPlusManhattan(TileIndex, TileIndex); ///< Max + Manhattan
 uint DistanceFromEdge(TileIndex); ///< shortest distance from any edge of the map
+uint DistanceFromEdgeDir(TileIndex, DiagDirection); ///< distance from the map edge in given direction
 
 /**
  * A loop which iterates to a square of tiles
