Index: src/tunnelbridge_cmd.cpp
===================================================================
--- src/tunnelbridge_cmd.cpp	(revision 21000)
+++ src/tunnelbridge_cmd.cpp	(working copy)
@@ -217,6 +217,21 @@
 }
 
 /**
+ * Quick check if given tile type allows building bridge ramp.
+ * @param tile  The tile to check
+ * @return      True if we can build the bridge ramp
+ */
+static inline bool AllowBridgeRamp(TileIndex tile)
+{
+	switch (GetTileType(tile)) {
+		case MP_CLEAR: return true;
+		case MP_TREES: return true;
+		case MP_WATER: return (IsCoast(tile));
+		default: return false;
+	};
+}
+
+/**
  * Build a Bridge
  * @param end_tile end tile
  * @param flags type of operation
