Index: src/ai/ai_instance.cpp
===================================================================
--- src/ai/ai_instance.cpp	(revision 16552)
+++ src/ai/ai_instance.cpp	(working copy)
@@ -272,7 +272,7 @@
 
 void AIInstance::GameLoop()
 {
-	if (this->is_dead) return;
+	if (this->IsDead()) return;
 	if (this->engine->HasScriptCrashed()) {
 		/* The script crashed during saving, kill it here. */
 		this->Died();
@@ -338,7 +338,7 @@
 
 void AIInstance::CollectGarbage()
 {
-	if (this->is_started && !this->is_dead) this->engine->CollectGarbage();
+	if (this->is_started && !this->IsDead()) this->engine->CollectGarbage();
 }
 
 /* static */ void AIInstance::DoCommandReturn(AIInstance *instance)
Index: src/ai/ai_instance.hpp
===================================================================
--- src/ai/ai_instance.hpp	(revision 16552)
+++ src/ai/ai_instance.hpp	(working copy)
@@ -81,6 +81,11 @@
 	class AIController *GetController() { return controller; }
 
 	/**
+	 * Return the "this AI died" value
+	 */
+	bool IsDead() { return this->is_dead; }
+
+	/**
 	 * Call the AI Save function and save all data in the savegame.
 	 */
 	void Save();
Index: src/ai/ai_gui.cpp
===================================================================
--- src/ai/ai_gui.cpp	(revision 16552)
+++ src/ai/ai_gui.cpp	(working copy)
@@ -22,6 +22,7 @@
 #include "ai.hpp"
 #include "api/ai_log.hpp"
 #include "ai_config.hpp"
+#include "ai_instance.hpp"
 
 #include "table/strings.h"
 
@@ -692,6 +693,8 @@
 
 		/* Paint the company icons */
 		for (CompanyID i = COMPANY_FIRST; i < MAX_COMPANIES; i++) {
+			this->widget[i + AID_WIDGET_COMPANY_BUTTON_START].colour = COLOUR_GREY;
+
 			Company *c = Company::GetIfValid(i);
 			if (c == NULL || !c->is_ai) {
 				/* Check if we have the company as an active company */
@@ -705,6 +708,10 @@
 				continue;
 			}
 
+			if (c->ai_instance->IsDead()) {
+				this->widget[i + AID_WIDGET_COMPANY_BUTTON_START].colour = COLOUR_RED;
+			}
+
 			/* Check if we have the company marked as inactive */
 			if (this->IsWidgetDisabled(i + AID_WIDGET_COMPANY_BUTTON_START)) {
 				/* New AI! Yippie :p */
