Index: industry_cmd.cpp
===================================================================
--- industry_cmd.cpp	(revision 18821)
+++ industry_cmd.cpp	(working copy)
@@ -2234,9 +2234,9 @@
 				}
 			}
 		} else {
-			if (only_decrease || Chance16(1, 3)) {
-				/* If more than 60% transported, 66% chance of increase, else 33% chance of increase */
-				if (!only_decrease && (i->last_month_pct_transported[0] > PERCENT_TRANSPORTED_60) != Chance16(1, 3)) {
+			if (only_decrease || Chance16(1, Clamp(RandomRange(4), 1, 3))) {
+				/* If more than 60% transported, 66% or more chance of increase, else 33% chance of increase */
+				if (!only_decrease && (i->last_month_pct_transported[0] > PERCENT_TRANSPORTED_60) != Chance16(1, (i->last_month_pct_transported[0] > PERCENT_TRANSPORTED_80 ? 6 : 3))) {
 					mul = 1; // Increase production
 				} else {
 					div = 1; // Decrease production
@@ -2283,8 +2283,8 @@
 	 * For non-smooth economy these should always be synchronized with prod_level */
 	if (recalculate_multipliers) {
 		/* Rates are rounded up, so e.g. oilrig always produces some passengers */
-		i->production_rate[0] = min((indspec->production_rate[0] * i->prod_level + PRODLEVEL_DEFAULT - 1) / PRODLEVEL_DEFAULT, 0xFF);
-		i->production_rate[1] = min((indspec->production_rate[1] * i->prod_level + PRODLEVEL_DEFAULT - 1) / PRODLEVEL_DEFAULT, 0xFF);
+		i->production_rate[0] = min(((indspec->production_rate[0] + RandomRange(indspec->production_rate[0] + 1)) * i->prod_level + PRODLEVEL_DEFAULT - 1) / PRODLEVEL_DEFAULT, 0xFF);
+		i->production_rate[1] = min(((indspec->production_rate[1] + RandomRange(indspec->production_rate[1] + 1)) * i->prod_level + PRODLEVEL_DEFAULT - 1) / PRODLEVEL_DEFAULT, 0xFF);
 	}
 
 	/* Close if needed and allowed */
