Index: not_close2/src/date_type.h
===================================================================
--- not_close2/src/date_type.h	(revision 13649)
+++ not_close2/src/date_type.h	(working copy)
@@ -44,10 +44,14 @@
 typedef uint8  Month;
 typedef uint8  Day;
 
+/**
+ * Data structure to convert between Date and triplet (year, month, and day).
+ * @see ConvertDateToYMD(), ConvertYMDToDate()
+ */
 struct YearMonthDay {
-	Year  year;
-	Month month; ///< 0 - 11
-	Day   day;   ///< 1 - 31
+	Year  year;   ///< Year (0...)
+	Month month;  ///< Month (0..11)
+	Day   day;    ///< Day (1..31)
 };
 
 static const Year INVALID_YEAR = -1;
Index: not_close2/src/date.cpp
===================================================================
--- not_close2/src/date.cpp	(revision 13649)
+++ not_close2/src/date.cpp	(working copy)
@@ -19,9 +19,9 @@
 #include "saveload.h"
 #endif
 
-Year      _cur_year;
-Month     _cur_month;
-Date      _date;
+Year      _cur_year;   ///< Current year, starting at 0
+Month     _cur_month;  ///< Current month (0..11)
+Date      _date;       ///< Current date in days (day counter)
 DateFract _date_fract;
 
 
Index: not_close2/src/industry_cmd.cpp
===================================================================
--- not_close2/src/industry_cmd.cpp	(revision 13649)
+++ not_close2/src/industry_cmd.cpp	(working copy)
@@ -52,8 +52,8 @@
 static byte _industry_sound_ctr;
 static TileIndex _industry_sound_tile;
 
-int _total_industries;                      //general counter
-uint16 _industry_counts[NUM_INDUSTRYTYPES]; // Number of industries per type ingame
+int _total_industries;                      ///< General counter
+uint16 _industry_counts[NUM_INDUSTRYTYPES]; ///< Number of industries per type ingame
 
 IndustrySpec _industry_specs[NUM_INDUSTRYTYPES];
 IndustryTileSpec _industry_tile_specs[NUM_INDUSTRYTILES];
