Index: src/airport_gui.cpp
===================================================================
--- src/airport_gui.cpp	(revision 13003)
+++ src/airport_gui.cpp	(working copy)
@@ -95,7 +95,7 @@
 
 		case WE_PLACE_MOUSEUP:
 			if (e->we.place.pt.x != -1 && e->we.place.select_proc == DDSP_DEMOLISH_AREA) {
-				DoCommandP(e->we.place.tile, e->we.place.starttile, 0, CcPlaySound10, CMD_CLEAR_AREA | CMD_MSG(STR_00B5_CAN_T_CLEAR_THIS_AREA));
+				GUIPlaceProcDragXY(e);
 			}
 			break;
 
Index: src/road_gui.cpp
===================================================================
--- src/road_gui.cpp	(revision 13003)
+++ src/road_gui.cpp	(working copy)
@@ -574,7 +574,7 @@
 						break;
 
 					case DDSP_DEMOLISH_AREA:
-						DoCommandP(end_tile, start_tile, 0, CcPlaySound10, CMD_CLEAR_AREA | CMD_MSG(STR_00B5_CAN_T_CLEAR_THIS_AREA));
+						GUIPlaceProcDragXY(e);
 						break;
 
 					case DDSP_PLACE_ROAD_X_DIR:
