From ccba7e7486953171a4823a6371a668c1228a64f7 Mon Sep 17 00:00:00 2001
From: Bruce Cowan <bcowan@fastmail.co.uk>
Date: Thu, 17 Jan 2008 02:04:25 +0000
Subject: [PATCH] Add .gitignore

---
 .gitignore |   11 +++++++++++
 1 files changed, 11 insertions(+), 0 deletions(-)
 create mode 100644 .gitignore

diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000..965b40a
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1,11 @@
+Makefile*
+bin/*
+!bin/data/chars.grf
+!bin/data/openttdd.grf
+!bin/data/openttdw.grf
+!bin/data/opntitle.grf
+!bin/scenario/README
+!bin/scripts*
+config.*
+objs/*
+src/rev.cpp
-- 
1.5.2.5

