Index: src/console_cmds.cpp
===================================================================
--- src/console_cmds.cpp	(revision 14030)
+++ src/console_cmds.cpp	(working copy)
@@ -636,6 +636,22 @@
 	return true;
 }
 
+DEF_CONSOLE_CMD(ConStartCompany)
+{
+	if (argc == 0) {
+		IConsoleHelp("Start a company as a spectator in a multiplayer game. Usage: 'start_company'");
+		return true;
+	}
+	if (_network_playas != PLAYER_SPECTATOR) {
+		IConsoleError("You need to be a spectator");
+		return false;
+	}
+	_network_playas = PLAYER_NEW_COMPANY;
+	SetLocalPlayer(PLAYER_SPECTATOR);
+	NetworkSend_Command(0, 0, 0, CMD_PLAYER_CTRL, NULL);
+	return true;
+}
+
 DEF_CONSOLE_CMD(ConNetworkClients)
 {
 	NetworkClientInfo *ci;
@@ -1406,7 +1422,10 @@
 
 	IConsoleCmdRegister("reset_company",   ConResetCompany);
 	IConsoleCmdHookAdd("reset_company",    ICONSOLE_HOOK_ACCESS, ConHookServerOnly);
+    IConsoleCmdRegister("start_company",   ConStartCompany);
+	IConsoleCmdHookAdd("start_company",    ICONSOLE_HOOK_ACCESS, ConHookClientOnly);
 	IConsoleAliasRegister("clean_company", "reset_company %A");
+	IConsoleAliasRegister("setup_company", "start_company");
 	IConsoleCmdRegister("kick",            ConKick);
 	IConsoleCmdHookAdd("kick",             ICONSOLE_HOOK_ACCESS, ConHookServerOnly);
 	IConsoleCmdRegister("ban",             ConBan);
Index: src/network/network_server.cpp
===================================================================
--- src/network/network_server.cpp	(revision 14030)
+++ src/network/network_server.cpp	(working copy)
@@ -897,7 +897,7 @@
 	 * to match the player in the packet. If it doesn't, the client has done
 	 * something pretty naughty (or a bug), and will be kicked
 	 */
-	if (!(cp->cmd == CMD_PLAYER_CTRL && cp->p1 == 0 && ci->client_playas == PLAYER_NEW_COMPANY) && ci->client_playas != cp->player) {
+	if (!(cp->cmd == CMD_PLAYER_CTRL && cp->p1 == 0 && (ci->client_playas == PLAYER_NEW_COMPANY || ci->client_playas == PLAYER_SPECTATOR)) && ci->client_playas != cp->player) {
 		IConsolePrintF(CC_ERROR, "WARNING: player %d (IP: %s) tried to execute a command as player %d, kicking...",
 		               ci->client_playas + 1, GetPlayerIP(ci), cp->player + 1);
 		SEND_COMMAND(PACKET_SERVER_ERROR)(cs, NETWORK_ERROR_PLAYER_MISMATCH);
