diff -r 188857a3b0db src/network/network_gui.cpp
--- a/src/network/network_gui.cpp	Mon Dec 22 12:59:31 2008 +0000
+++ b/src/network/network_gui.cpp	Mon Dec 22 14:19:13 2008 +0100
@@ -1478,7 +1478,7 @@
 
 	NetworkClientListPopupWindow(int x, int y, const Widget *widgets, int client_no) :
 			Window(x, y, 150, 100, WC_TOOLBAR_MENU, widgets),
-			sel_index(0), client_no(client_no)
+			sel_index(255), client_no(client_no)
 	{
 		/*
 		 * Fill the actions this client has.
@@ -1630,7 +1630,7 @@
 
 	NetworkClientListWindow(const WindowDesc *desc, WindowNumber window_number) :
 			Window(desc, window_number),
-			selected_item(0),
+			selected_item(255),
 			selected_y(255)
 	{
 		this->FindWindowPlacementAndResize(desc);
