Index: group_cmd.cpp
===================================================================
--- group_cmd.cpp	(revision 12672)
+++ group_cmd.cpp	(working copy)
@@ -39,7 +39,7 @@
 		/* Decrease the num engines of EngineID i of the old group if it's not the default one */
 		if (!IsDefaultGroupID(old_g) && IsValidGroupID(old_g)) GetGroup(old_g)->num_engines[i]--;
 
-		/* Increase the num engines of EngineID i of the new group if it's not the new one */
+		/* Increase the num engines of EngineID i of the new group if it's not the default one */
 		if (!IsDefaultGroupID(new_g) && IsValidGroupID(new_g)) GetGroup(new_g)->num_engines[i]++;
 	}
 }
