Index: src/road_cmd.cpp
===================================================================
--- src/road_cmd.cpp	(revision 12276)
+++ src/road_cmd.cpp	(working copy)
@@ -1066,7 +1066,7 @@
 	}
 
 	/* Draw extra details. */
-	for (drts = _road_display_table[roadside][road]; drts->image != 0; drts++) {
+	for (drts = _road_display_table[roadside][road|tram]; drts->image != 0; drts++) {
 		DrawRoadDetail(drts->image, ti, drts->subcoord_x, drts->subcoord_y, 0x10);
 	}
 }
