Index: src/cheat_gui.cpp
===================================================================
--- src/cheat_gui.cpp	(revision 25193)
+++ src/cheat_gui.cpp	(working copy)
@@ -118,6 +118,7 @@
 	CHT_NO_JETCRASH,     ///< Disable jet-airplane crashes.
 	CHT_SETUP_PROD,      ///< Allow manually editing of industry production.
 	CHT_CHANGE_DATE,     ///< Do time traveling.
+	CHT_NO_LENGHT_PEN,   ///< No penalization for trains bigger then station
 
 	CHT_NUM_CHEATS,      ///< Number of cheats.
 };
@@ -143,13 +144,14 @@
  * Order matches with the values of #CheatNumbers
  */
 static const CheatEntry _cheats_ui[] = {
-	{SLE_INT32, STR_CHEAT_MONEY,           &_money_cheat_amount,                    &_cheats.money.been_used,            &ClickMoneyCheat         },
-	{SLE_UINT8, STR_CHEAT_CHANGE_COMPANY,  &_local_company,                         &_cheats.switch_company.been_used,   &ClickChangeCompanyCheat },
-	{SLE_BOOL,  STR_CHEAT_EXTRA_DYNAMITE,  &_cheats.magic_bulldozer.value,          &_cheats.magic_bulldozer.been_used,  NULL                     },
-	{SLE_BOOL,  STR_CHEAT_CROSSINGTUNNELS, &_cheats.crossing_tunnels.value,         &_cheats.crossing_tunnels.been_used, NULL                     },
-	{SLE_BOOL,  STR_CHEAT_NO_JETCRASH,     &_cheats.no_jetcrash.value,              &_cheats.no_jetcrash.been_used,      NULL                     },
-	{SLE_BOOL,  STR_CHEAT_SETUP_PROD,      &_cheats.setup_prod.value,               &_cheats.setup_prod.been_used,       &ClickSetProdCheat       },
-	{SLE_INT32, STR_CHEAT_CHANGE_DATE,     &_cur_year,                              &_cheats.change_date.been_used,      &ClickChangeDateCheat    },
+	{SLE_INT32, STR_CHEAT_MONEY,                &_money_cheat_amount,                    &_cheats.money.been_used,            &ClickMoneyCheat         },
+	{SLE_UINT8, STR_CHEAT_CHANGE_COMPANY,       &_local_company,                         &_cheats.switch_company.been_used,   &ClickChangeCompanyCheat },
+	{SLE_BOOL,  STR_CHEAT_EXTRA_DYNAMITE,       &_cheats.magic_bulldozer.value,          &_cheats.magic_bulldozer.been_used,  NULL                     },
+	{SLE_BOOL,  STR_CHEAT_CROSSINGTUNNELS,      &_cheats.crossing_tunnels.value,         &_cheats.crossing_tunnels.been_used, NULL                     },
+	{SLE_BOOL,  STR_CHEAT_NO_JETCRASH,          &_cheats.no_jetcrash.value,              &_cheats.no_jetcrash.been_used,      NULL                     },
+	{SLE_BOOL,  STR_CHEAT_SETUP_PROD,           &_cheats.setup_prod.value,               &_cheats.setup_prod.been_used,       &ClickSetProdCheat       },
+	{SLE_INT32, STR_CHEAT_CHANGE_DATE,          &_cur_year,                              &_cheats.change_date.been_used,      &ClickChangeDateCheat    },
+	{SLE_BOOL,  STR_CHEAT_NO_TRAIN_LEN_PENALTY, &_cheats.train_penalty.value,            &_cheats.train_penalty.been_used,    NULL                     },
 };
 
 assert_compile(CHT_NUM_CHEATS == lengthof(_cheats_ui));
Index: src/cheat_type.h
===================================================================
--- src/cheat_type.h	(revision 25193)
+++ src/cheat_type.h	(working copy)
@@ -36,6 +36,7 @@
 	Cheat change_date;      ///< changes date ingame
 	Cheat setup_prod;       ///< setup raw-material production in game
 	Cheat dummy3;           ///< empty cheat (enable running el-engines on normal rail)
+	Cheat train_penalty;    ///< no penalization for trains bigger then station
 };
 
 extern Cheats _cheats;
Index: src/economy.cpp
===================================================================
--- src/economy.cpp	(revision 25193)
+++ src/economy.cpp	(working copy)
@@ -46,6 +46,7 @@
 #include "game/game.hpp"
 #include "cargomonitor.h"
 #include "goal_base.h"
+#include "cheat_type.h"
 
 #include "table/strings.h"
 #include "table/pricebase.h"
@@ -1621,7 +1622,8 @@
 		SB(front->vehicle_flags, VF_LOADING_FINISHED, 1, finished_loading);
 	}
 
-	if (front->type == VEH_TRAIN) {
+	/*Check if the vehicle is train and if the cheat is disabled*/
+	if (front->type == VEH_TRAIN && !_cheats.train_penalty.value) {
 		/* Each platform tile is worth 2 rail vehicles. */
 		int overhang = front->GetGroundVehicleCache()->cached_total_length - st->GetPlatformLength(front->tile) * TILE_SIZE;
 		if (overhang > 0) {
Index: src/lang/english.txt
===================================================================
--- src/lang/english.txt	(revision 25193)
+++ src/lang/english.txt	(working copy)
@@ -1687,6 +1687,7 @@
 STR_CHEAT_CHANGE_DATE                                           :{LTBLUE}Change date: {ORANGE}{DATE_SHORT}
 STR_CHEAT_CHANGE_DATE_QUERY_CAPT                                :{WHITE}Change current year
 STR_CHEAT_SETUP_PROD                                            :{LTBLUE}Enable modifying production values: {ORANGE}{STRING1}
+STR_CHEAT_NO_TRAIN_LEN_PENALTY                                  :{LTBLUE}Disable loading time penalty for trains longer than stations: {ORANGE}{STRING1}
 
 # Livery window
 STR_LIVERY_CAPTION                                              :{WHITE}New Colour Scheme
