Index: src/openttd.cpp
===================================================================
--- src/openttd.cpp	(revision 10735)
+++ src/openttd.cpp	(working copy)
@@ -624,16 +624,16 @@
 
 	if (_network_available) NetworkShutDown(); // Shut down the network and close any open connections
 
-	_video_driver->Stop();
-	_music_driver->Stop();
-	_sound_driver->Stop();
-
 	/* only save config if we have to */
 	if (save_config) {
 		SaveToConfig();
 		SaveToHighScore();
 	}
 
+	_video_driver->Stop();
+	_music_driver->Stop();
+	_sound_driver->Stop();
+
 	/* Reset windowing system and free config file */
 	UnInitializeGame();
 
