Index: src/dock_gui.cpp
===================================================================
--- src/dock_gui.cpp	(revision 21000)
+++ src/dock_gui.cpp	(working copy)
@@ -313,6 +313,10 @@
 				tile_to = TileAddByDiagDir(tile_from, ReverseDiagDir(dir));
 				tile_from = this->last_clicked_widget == DTW_LOCK ? TileAddByDiagDir(tile_from, dir) : tile_from;
 			}
+
+		} else if (this->last_clicked_widget == DTW_BUILD_AQUEDUCT) {
+			/* get the correct error for the aqueduct */
+			_build_tunnel_endtile = TILE_ADDXY(tile_from, 1, 0);
 		}
 
 		VpSetPresizeRange(tile_from, tile_to);
