diff -r b570bcf795b6 src/group_gui.cpp
--- a/src/group_gui.cpp	Sat Oct 02 19:55:13 2010 +0000
+++ b/src/group_gui.cpp	Sun Oct 03 11:31:53 2010 +0200
@@ -280,7 +280,7 @@
 				 * We list all vehicles or ungrouped vehicles */
 				if (IsDefaultGroupID(this->vli.index) || IsAllGroupID(this->vli.index)) {
 					SetDParam(0, STR_COMPANY_NAME);
-					SetDParam(1, GB(this->window_number, 0, 8));
+					SetDParam(1, this->vli.company);
 					SetDParam(2, this->vehicles.Length());
 					SetDParam(3, this->vehicles.Length());
 				} else {
