Index: station_gui.c
===================================================================
--- station_gui.c	(revision 7130)
+++ station_gui.c	(working copy)
@@ -28,7 +28,7 @@
 
 static void StationsWndShowStationRating(int x, int y, int type, uint acceptance, int rating)
 {
-	static const byte _rating_colors[NUM_CARGO] = {152, 32, 15, 174, 208, 194, 191, 55, 184, 10, 191, 48};
+	static const byte _rating_colors[NUM_CARGO] = {152, 32, 15, 174, 208, 194, 191, 84, 184, 10, 202, 215};
 	int color = _rating_colors[type];
 	uint w;
 
