Index: vehicle.c
===================================================================
--- vehicle.c	(revision 6294)
+++ vehicle.c	(working copy)
@@ -2528,6 +2528,28 @@
 			v->current_order.flags = (v->current_order.type & 0xF0) >> 4;
 			v->current_order.type  =  v->current_order.type & 0x0F;
 		}
+
+		if (v->type == VEH_Train && HASBIT(v->subtype, Train_Wagon)) {
+			if (v->cur_speed != 0) {
+				printf("****************************\n");
+				printf("buggy wagon cur_speed found!\n");
+				printf("cur_speed=%d\n", v->cur_speed);
+				// change loaded wagon cur_speed to 0, as it always should be
+				v->cur_speed = 0;
+				printf("cur_speed nonzero *causes depot bug*!  it has been reset to 0\n");
+				printf("savegame unchanged - please report this error!\n");
+				printf("****************************\n");
+			}
+			if (v->max_speed != 0) {
+				printf("****************************\n");
+				printf("buggy wagon max_speed found!\n");
+				printf("max_speed=%d\n", v->max_speed);
+				v->max_speed = 0;
+				printf("max_speed nonzero does not cause depot bug, but has been set to 0 anyway\n");
+				printf("savegame unchanged - please report this error!\n");
+				printf("****************************\n");
+			}
+		}
 	}
 
 	/* Check for shared order-lists (we now use pointers for that) */
