Index: src/video/win32_v.cpp
===================================================================
--- src/video/win32_v.cpp	(revision 13480)
+++ src/video/win32_v.cpp	(working copy)
@@ -280,8 +280,8 @@
 #endif
 		w = r.right - r.left;
 		h = r.bottom - r.top;
-		x = (GetSystemMetrics(SM_CXSCREEN) - w) / 2;
-		y = (GetSystemMetrics(SM_CYSCREEN) - h) / 2;
+		x = (GetSystemMetrics(SM_CXVIRTUALSCREEN) - w) / 2;
+		y = (GetSystemMetrics(SM_CYVIRTUALSCREEN) - h) / 2;
 
 		if (_wnd.main_wnd) {
 			ShowWindow(_wnd.main_wnd, SW_SHOWNORMAL); // remove maximize-flag
