--- ../original/openttd-1.2.3/src/video/cocoa/wnd_quartz.mm	2012-10-31 19:38:49.000000000 -0200
+++ src/video/cocoa/wnd_quartz.mm	2012-11-11 14:12:34.000000000 -0200
@@ -516,7 +516,8 @@
 	NSPoint pt = [ event locationInWindow ];
 	pt = [ this->cocoaview convertPoint:pt fromView:nil ];
 
-	pt.y = this->window_height - pt.y;
+	pt.y = this->window_height - pt.y + [this->window frame].origin.y;
+	pt.x -= [this->window frame].origin.x;
 
 	return pt;
 }
