Index: src/fontcache.cpp
===================================================================
--- src/fontcache.cpp	(revision 24979)
+++ src/fontcache.cpp	(working copy)
@@ -488,7 +488,11 @@
 	const char *str;
 	bool result = false;
 
-	callback->FindMissingGlyphs(&str);
+	if (_game_mode == GM_BOOTSTRAP) {
+		str = GetStringPtr(STR_MISSING_GRAPHICS_SET_MESSAGE);
+	} else {
+		callback->FindMissingGlyphs(&str);
+	}
 
 #if (MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_5)
 	if (MacOSVersionIsAtLeast(10, 5, 0)) {
