diff --git a/src/signs_gui.cpp b/src/signs_gui.cpp
index 71b452b4..e572f51 100644
--- a/src/signs_gui.cpp
+++ b/src/signs_gui.cpp
@@ -542,7 +542,7 @@ static const NWidgetPart _nested_query_sign_edit_widgets[] = {
 };

 static WindowDesc _query_sign_edit_desc(
-	WDP_AUTO, "query_sign", 0, 0,
+	WDP_CENTER, "query_sign", 0, 0,
 	WC_QUERY_STRING, WC_NONE,
 	WDF_CONSTRUCTION,
 	_nested_query_sign_edit_widgets, lengthof(_nested_query_sign_edit_widgets)
