Index: src/script/api/script_town.cpp
===================================================================
--- src/script/api/script_town.cpp	(Revision 26375)
+++ src/script/api/script_town.cpp	(Arbeitskopie)
@@ -185,6 +185,8 @@
 
 	if (t->growth_rate == TOWN_GROW_RATE_CUSTOM_NONE) return TOWN_GROWTH_NONE;
 
+	if (!HasBit(t->flags, TOWN_IS_GROWING)) return TOWN_GROWTH_NONE;
+
 	return ((t->growth_rate & ~TOWN_GROW_RATE_CUSTOM) * TOWN_GROWTH_TICKS + DAY_TICKS) / DAY_TICKS;
 }
 
