Index: economy.c
===================================================================
--- economy.c	(revision 5893)
+++ economy.c	(working copy)
@@ -1230,19 +1230,24 @@
 	// Give the goods to the industry.
 	DeliverGoodsToIndustry(s_to->xy, cargo_type, num_pieces);
 
-	// Determine profit
-	profit = GetTransportedGoodsIncome(num_pieces, DistanceManhattan(s_from->xy, s_to->xy), days_in_transit, cargo_type);
+	// Correct invalid station bug.
+	if(s_from->xy != 0)	{
+		// Determine profit
+		profit = GetTransportedGoodsIncome(num_pieces, DistanceManhattan(s_from->xy, s_to->xy), days_in_transit, cargo_type);
 
-	// Modify profit if a subsidy is in effect
-	if (subsidised) {
-		switch (_opt.diff.subsidy_multiplier) {
-			case 0:  profit += profit >> 1; break;
-			case 1:  profit *= 2; break;
-			case 2:  profit *= 3; break;
-			default: profit *= 4; break;
+		// Modify profit if a subsidy is in effect
+		if (subsidised) {
+			switch (_opt.diff.subsidy_multiplier) {
+				case 0:  profit += profit >> 1; break;
+				case 1:  profit *= 2; break;
+				case 2:  profit *= 3; break;
+				default: profit *= 4; break;
+			}
 		}
 	}
-
+	else {
+		profit = 0;
+	}
 	return profit;
 }
 
