Index: src/rail_gui.cpp
===================================================================
--- src/rail_gui.cpp	(revision 14367)
+++ src/rail_gui.cpp	(working copy)
@@ -1443,6 +1443,12 @@
 
 				_cur_signal_type = (SignalType)((uint)((widget - BSW_SEMAPHORE_NORM) % (SIGTYPE_LAST + 1)));
 				_cur_signal_variant = widget >= BSW_ELECTRIC_NORM ? SIG_ELECTRIC : SIG_SEMAPHORE;
+
+				/* If 'remove' button of rail build toolbar is active, disable it */
+				if (_remove_button_clicked) {
+					Window *w = FindWindowById(WC_BUILD_TOOLBAR, TRANSPORT_RAIL);
+					if (w != NULL) ToggleRailButton_Remove(w);
+				}
 				break;
 
 			case BSW_CONVERT:
