diff -r e5575895a889 bin/data/opntitle.dat
Binary file bin/data/opntitle.dat has changed
diff -r e5575895a889 config.lib
--- a/config.lib	Sat Apr 10 02:17:18 2010 +0000
+++ b/config.lib	Sat Apr 10 14:48:12 2010 +0200
@@ -1330,6 +1330,10 @@
 
 	if [ "$os" = "OSX" ]; then
 		LDFLAGS="$LDFLAGS -framework Cocoa"
+		
+		# add macports include dir which is not always set a default system dir. This avoids zillions of bogus warnings.
+		CFLAGS="$CFLAGS -isystem/opt/local/include"
+		
 		if [ "$enable_dedicated" = "0" ] && ([ "$cpu_type" = "32" ] || [ "$enable_universal" != "0" ]); then
 			LIBS="$LIBS -framework QuickTime"
 		else
