diff --git a/src/game/game_core.cpp b/src/game/game_core.cpp
index f32a99b..6f301d3 100644
--- a/src/game/game_core.cpp
+++ b/src/game/game_core.cpp
@@ -159,7 +159,7 @@
 
 /* static */ void Game::ResetConfig()
 {
-	/* Check for both newgame as current game if we can reload the GameInfo insde
+	/* Check for both newgame as current game if we can reload the GameInfo inside
 	 *  the GameConfig. If not, remove the Game from the list. */
 	if (_settings_game.game_config != NULL && _settings_game.game_config->HasScript()) {
 		if (!_settings_game.game_config->ResetInfo(true)) {
diff --git a/src/game/game_scanner.hpp b/src/game/game_scanner.hpp
index 8f43d14..f1d334f 100644
--- a/src/game/game_scanner.hpp
+++ b/src/game/game_scanner.hpp
@@ -21,7 +21,7 @@ public:
 	/**
 	 * Check if we have a game by name and version available in our list.
 	 * @param nameParam The name of the game script.
-	 * @param versionParam The versionof the game script, or -1 if you want the latest.
+	 * @param versionParam The version of the game script, or -1 if you want the latest.
 	 * @param force_exact_match Only match name+version, never latest.
 	 * @return NULL if no match found, otherwise the game script that matched.
 	 */
