Index: network_server.cpp
===================================================================
--- network_server.cpp	(révision 15765)
+++ network_server.cpp	(copie de travail)
@@ -14,6 +14,7 @@
 #include "network_udp.h"
 #include "../console_func.h"
 #include "../command_func.h"
+#include "../command_type.h"
 #include "../saveload/saveload.h"
 #include "../station_base.h"
 #include "../genworld.h"
@@ -485,6 +486,11 @@
 	p->Send_bool  (cp->my_cmd);
 
 	cs->Send_Packet(p);
+
+	if (cp->cmd==CMD_COMPANY_CTRL) {
+		DEBUG( net, 0, "New Company to client" );
+		CheckMinActiveClients();
+	}
 }
 
 DEF_SERVER_SEND_COMMAND_PARAM(PACKET_SERVER_CHAT)(NetworkClientSocket *cs, NetworkAction action, ClientID client_id, bool self_send, const char *msg, int64 data)
