Index: video/cocoa/wnd_quartz.mm
===================================================================
--- video/cocoa/wnd_quartz.mm	(revision 24976)
+++ video/cocoa/wnd_quartz.mm	(working copy)
@@ -513,9 +513,18 @@
 
 NSPoint WindowQuartzSubdriver::GetMouseLocation(NSEvent *event)
 {
-	NSPoint pt = [ event locationInWindow ];
-	pt = [ this->cocoaview convertPoint:pt fromView:nil ];
+	NSPoint pt;
 
+	if (event.window == nil)
+#if (MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_7)
+		if ([ this->cocoaview respondsToSelector:@selector(convertRectFromScreen::) ])
+			pt = [ this->cocoaview convertPoint: [ [ this->cocoaview window ] convertRectFromScreen: NSMakeRect([ event locationInWindow ].x, [ event locationInWindow ].y, 0, 0) ].origin fromView: nil ];
+		else
+#endif
+			pt = [ this->cocoaview convertPoint: [ [ this->cocoaview window ] convertScreenToBase: [ event locationInWindow ] ] fromView: nil ];
+	else
+		pt = [ event locationInWindow ];
+
 	pt.y = this->window_height - pt.y;
 
 	return pt;
