Index: src/group_gui.cpp
===================================================================
--- src/group_gui.cpp	(revision 12763)
+++ src/group_gui.cpp	(working copy)
@@ -523,7 +523,11 @@
 					break;
 
 				case GRP_WIDGET_SORT_BY_DROPDOWN: // Select sorting criteria dropdown menu
-					ShowDropDownMenu(w, _vehicle_sort_listing, gv->l.sort_type,  GRP_WIDGET_SORT_BY_DROPDOWN, 0, 0);
+					if (gv->vehicle_type == VEH_TRAIN || gv->vehicle_type == VEH_ROAD) {
+						ShowDropDownMenu(w, _vehicle_sort_listing, gv->l.sort_type,  GRP_WIDGET_SORT_BY_DROPDOWN, 0, 0);
+					} else {
+						ShowDropDownMenu(w, _vehicle_sort_listing, gv->l.sort_type,  GRP_WIDGET_SORT_BY_DROPDOWN, 0, 10000000000);
+					}
 					return;
 
 				case GRP_WIDGET_ALL_VEHICLES: // All vehicles button
Index: src/lang/english.txt
===================================================================
--- src/lang/english.txt	(revision 12763)
+++ src/lang/english.txt	(working copy)
@@ -378,6 +378,7 @@
 STR_SORT_BY_MAX_SPEED                                           :Maximum speed
 STR_SORT_BY_MODEL                                               :Model
 STR_SORT_BY_VALUE                                               :Value
+STR_SORT_BY_LENGTH												:Length
 STR_SORT_BY_FACILITY                                            :Station type
 STR_SORT_BY_WAITING                                             :Waiting cargo value
 STR_SORT_BY_RATING_MAX                                          :Cargo rating
Index: src/vehicle_gui.cpp
===================================================================
--- src/vehicle_gui.cpp	(revision 12763)
+++ src/vehicle_gui.cpp	(working copy)
@@ -72,6 +72,7 @@
 static VehicleSortListingTypeFunction VehicleMaxSpeedSorter;
 static VehicleSortListingTypeFunction VehicleModelSorter;
 static VehicleSortListingTypeFunction VehicleValueSorter;
+static VehicleSortListingTypeFunction VehicleLengthSorter;
 
 static VehicleSortListingTypeFunction* const _vehicle_sorter[] = {
 	&VehicleNumberSorter,
@@ -84,6 +85,7 @@
 	&VehicleMaxSpeedSorter,
 	&VehicleModelSorter,
 	&VehicleValueSorter,
+	&VehicleLengthSorter,
 };
 
 const StringID _vehicle_sort_listing[] = {
@@ -97,6 +99,7 @@
 	STR_SORT_BY_MAX_SPEED,
 	STR_SORT_BY_MODEL,
 	STR_SORT_BY_VALUE,
+	STR_SORT_BY_LENGTH,
 	INVALID_STRING_ID
 };
 
@@ -698,6 +701,28 @@
 	return (_internal_sort_order & 1) ? -r : r;
 }
 
+static int CDECL VehicleLengthSorter(const void *a, const void *b)
+{
+	const Vehicle* va = *(const Vehicle**)a;
+	const Vehicle* vb = *(const Vehicle**)b;
+	const Vehicle *u;
+	int r;
+	int vacounter = 0, vbcounter = 0;
+
+	if (va->type == VEH_TRAIN) {
+		r = va->u.rail.cached_total_length - vb->u.rail.cached_total_length;
+	} else if (va->type == VEH_ROAD) { 
+		for (u = va; u != NULL; u = u->Next()) vacounter += va->u.road.cached_veh_length;
+		for (u = vb; u != NULL; u = u->Next()) vbcounter += vb->u.road.cached_veh_length;
+	}
+
+	r = vacounter - vbcounter;
+
+	VEHICLEUNITNUMBERSORTER(r, va, vb);
+
+	return (_internal_sort_order & 1) ? -r : r;
+}
+
 void InitializeGUI()
 {
 	memset(&_sorting, 0, sizeof(_sorting));
@@ -1061,7 +1086,11 @@
 					SetWindowDirty(w);
 					break;
 				case VLW_WIDGET_SORT_BY_PULLDOWN:/* Select sorting criteria dropdown menu */
-					ShowDropDownMenu(w, _vehicle_sort_listing, vl->l.sort_type, VLW_WIDGET_SORT_BY_PULLDOWN, 0, 0);
+					if (vl->vehicle_type == VEH_TRAIN || vl->vehicle_type == VEH_ROAD) {
+						ShowDropDownMenu(w, _vehicle_sort_listing, vl->l.sort_type, VLW_WIDGET_SORT_BY_PULLDOWN, 0, 0);
+					} else {
+						ShowDropDownMenu(w, _vehicle_sort_listing, vl->l.sort_type, VLW_WIDGET_SORT_BY_PULLDOWN, 0, 10000000000);
+					}
 					return;
 				case VLW_WIDGET_LIST: { /* Matrix to show vehicles */
 					uint32 id_v = (e->we.click.pt.y - PLY_WND_PRC__OFFSET_TOP_WIDGET) / w->resize.step_height;
