Index: website/servers.php
===================================================================
--- website/servers.php	(revision 10486)
+++ website/servers.php	(working copy)
@@ -83,7 +83,7 @@
 			$info['online'] = $entry['online'];
 
 			$info['name'] = ($entry['name'] == "") ? "Unnamed Server" : htmlspecialchars($entry['name']);
-			$info['revision'] = $entry['revision'];
+			$info['revision'] = htmlspecialchars($entry['revision']);
 			$info['use_password'] = $entry['use_password'];
 			$info['clients_max'] = $entry['clients_max'];
 			$info['clients_on'] = $entry['clients_on'];
Index: website/server_detail.php
===================================================================
--- website/server_detail.php	(revision 10486)
+++ website/server_detail.php	(working copy)
@@ -40,7 +40,7 @@
 	$template->assign("last_online",     $row['last_online']);
 	$template->assign("info_version",    $row['info_version']);
 	$template->assign("name",            htmlspecialchars($row['name']));
-	$template->assign("revision",        $row['revision']);
+	$template->assign("revision",        htmlspecialchars($row['revision']));
 	$template->assign("server_langid",   $row['server_lang']);
 	$template->assign("use_password",    $row['use_password']);
 	$template->assign("clients_max",     $row['clients_max']);
