Index: src/3rdparty/squirrel/include/squirrel.h
===================================================================
--- src/3rdparty/squirrel/include/squirrel.h	(wersja 25619)
+++ src/3rdparty/squirrel/include/squirrel.h	(kopia robocza)
@@ -36,7 +36,12 @@
 #endif
 
 #if defined(_MSC_VER)
-# define inline __forceinline
+# if (_MSC_VER >= 1700) && !defined(_ALLOW_KEYWORD_MACROS)
+#  define _ALLOW_KEYWORD_MACROS // prevent MSVC from complaining abound macroized keywords
+# endif
+# ifndef inline
+#  define inline __forceinline
+# endif
 #endif /* _MSC_VER */
 
 #if defined(_MSC_VER) && _MSC_VER >= 1400 // MSVC 2005 safety check

