Index: table/unicode.h
===================================================================
--- table/unicode.h	(revision 7257)
+++ table/unicode.h	(working copy)
@@ -14,6 +14,12 @@
 
 static DefaultUnicodeMapping _default_unicode_map[] = {
 	{ 0x010D, 0x63 }, /* Small letter c with caron */
+	{ 0x011E, 0x47 }, /* Capital letter G with breve */
+	{ 0x011F, 0x67 }, /* Small letter g with breve */
+	{ 0x0130, 0x69 }, /* Capital letter I with dot above */
+	{ 0x0131, 0x69 }, /* Small letter dotless i */
+	{ 0x015E, 0x53 }, /* Capital letter S with cedilla */
+	{ 0x015F, 0x73 }, /* Small letter s with cedilla */
 	{ 0x0160, 0xA6 }, /* Capital letter s with caron */
 	{ 0x0161, 0xA8 }, /* Small letter s with caron */
 	{ 0x017E, 0xB8 }, /* Small letter z with caron */
