diff -r 3da4cb4a0fb9 src/order_gui.cpp
--- a/src/order_gui.cpp	Mon Apr 05 21:18:32 2010 +0000
+++ b/src/order_gui.cpp	Tue Apr 06 07:44:12 2010 +0200
@@ -1277,7 +1277,7 @@
 	virtual void OnTimeout()
 	{
 		static const int raise_widgets[] = {
-			ORDER_WIDGET_SKIP, ORDER_WIDGET_DELETE, ORDER_WIDGET_REFIT, ORDER_WIDGET_SHARED_ORDER_LIST, WIDGET_LIST_END,
+			ORDER_WIDGET_TIMETABLE_VIEW, ORDER_WIDGET_SKIP, ORDER_WIDGET_DELETE, ORDER_WIDGET_REFIT, ORDER_WIDGET_SHARED_ORDER_LIST, WIDGET_LIST_END,
 		};
 
 		/* Unclick all buttons in raise_widgets[]. */
