diff --git a/src/newgrf.cpp b/src/newgrf.cpp
--- a/src/newgrf.cpp
+++ b/src/newgrf.cpp
@@ -8821,7 +8821,7 @@
 static void ActivateOldShore()
 {
 	/* Use default graphics, if no shore sprites were loaded.
-	 * Should not happen, as openttd(w/d).grf includes some. */
+	 * Should not happen, as the base set extra grf should includes some. */
 	if (_loaded_newgrf_features.shore == SHORE_REPLACE_NONE) _loaded_newgrf_features.shore = SHORE_REPLACE_ACTION_A;
 
 	if (_loaded_newgrf_features.shore != SHORE_REPLACE_ACTION_5) {
