diff -Naur opengfx-0.4.6.1-source-orig/scripts/Makefile.nml opengfx-0.4.6.1-source/scripts/Makefile.nml
--- opengfx-0.4.6.1-source-orig/scripts/Makefile.nml	2012-12-27 11:57:35.925571502 +0100
+++ opengfx-0.4.6.1-source/scripts/Makefile.nml	2013-02-26 21:06:59.580312559 +0100
@@ -14,11 +14,11 @@
 
 %.src.dep: $(SRC_DIR)/%.pnml
 	$(_E) "[DEP] $@"
-	$(_V) $(CC) -C -M -MF $@ -E -MT $(patsubst %.src.dep,%.nml,$@) - < $<
+	$(_V) $(CC) -C -ffreestanding -M -MF $@ -E -MT $(patsubst %.src.dep,%.nml,$@) - < $<
 
 %.nml: $(SRC_DIR)/%.pnml $(TAG_FILENAME)
 	$(_E) "[CPP] $@"
-	$(_V) $(CC) -D REPO_REVISION=$(REPO_REVISION) $(CC_FLAGS) $< > $@
+	$(_V) $(CC) -ffreestanding -D REPO_REVISION=$(REPO_REVISION) $(CC_FLAGS) $< > $@
 
 %.gfx.dep: %.nml
 	$(_E) "[DEP] $@"
