diff -r 4738abd493ce src/newgrf_engine.cpp
--- a/src/newgrf_engine.cpp	Sun Jun 24 22:17:31 2012 +0200
+++ b/src/newgrf_engine.cpp	Thu Jun 28 16:10:59 2012 +0200
@@ -23,6 +23,7 @@
 #include "station_base.h"
 #include "company_base.h"
 #include "newgrf_railtype.h"
+#include "ship.h"
 
 struct WagonOverride {
 	EngineID *train_id;
@@ -834,6 +835,14 @@
 			break;
 		}
 
+		case VEH_SHIP: {
+			Ship *s = Ship::From(v);
+			switch (variable - 0x80) {
+				case 0x62: return s->state;
+			}
+			break;
+		}
+
 		case VEH_AIRCRAFT: {
 			Aircraft *a = Aircraft::From(v);
 			switch (variable - 0x80) {
