diff -r 527e232c2aae src/order_cmd.cpp
--- a/src/order_cmd.cpp	Sun Jan 04 13:00:17 2009 +0100
+++ b/src/order_cmd.cpp	Sun Jan 04 14:23:28 2009 +0100
@@ -1458,12 +1458,7 @@
 		case VEH_TRAIN:     return st->train_tile;
 		case VEH_AIRCRAFT:  return CanAircraftUseStation(v->engine_type, st) ? st->airport_tile : INVALID_TILE;
 		case VEH_SHIP:      return st->dock_tile;
-		case VEH_ROAD:
-			if (IsCargoInClass(v->cargo_type, CC_PASSENGERS)) {
-				return (st->bus_stops != NULL) ? st->bus_stops->xy : INVALID_TILE;
-			} else {
-				return (st->truck_stops != NULL) ? st->truck_stops->xy : INVALID_TILE;
-			}
+		case VEH_ROAD:      return st->GetPrimaryRoadStop(v) != NULL ? st->GetPrimaryRoadStop(v)->xy : INVALID_TILE;
 	}
 }
 
