# HG changeset patch
# Parent cebc759208863986d47cb9c81303bbdcfb22fb1a

diff -r cebc75920886 src/rail_cmd.cpp
--- a/src/rail_cmd.cpp	Sun Apr 05 17:45:14 2015 +0000
+++ b/src/rail_cmd.cpp	Thu Apr 16 17:14:34 2015 +0200
@@ -485,31 +485,31 @@
 				RoadTypes roadtypes = GetRoadTypes(tile);
 				RoadBits road = GetRoadBits(tile, ROADTYPE_ROAD);
 				RoadBits tram = GetRoadBits(tile, ROADTYPE_TRAM);
-				switch (roadtypes) {
-					default: break;
-					case ROADTYPES_TRAM:
-						/* Tram crossings must always have road. */
-						if (flags & DC_EXEC) {
-							SetRoadOwner(tile, ROADTYPE_ROAD, _current_company);
-							Company *c = Company::GetIfValid(_current_company);
-							if (c != NULL) {
-								/* A full diagonal tile has two road bits. */
-								c->infrastructure.road[ROADTYPE_ROAD] += 2;
-								DirtyCompanyInfrastructureWindows(c->index);
+				if ((track == TRACK_X && (road | tram) == ROAD_Y) ||
+						(track == TRACK_Y && (road | tram) == ROAD_X)) {
+					switch (roadtypes) {
+						default: break;
+						case ROADTYPES_TRAM:
+							/* Tram crossings must always have road. */
+							if (flags & DC_EXEC) {
+								SetRoadOwner(tile, ROADTYPE_ROAD, _current_company);
+								Company *c = Company::GetIfValid(_current_company);
+								if (c != NULL) {
+									/* A full diagonal tile has two road bits. */
+									c->infrastructure.road[ROADTYPE_ROAD] += 2;
+									DirtyCompanyInfrastructureWindows(c->index);
+								}
 							}
-						}
-						roadtypes |= ROADTYPES_ROAD;
-						break;
-
-					case ROADTYPES_ALL:
-						if (road != tram) return CMD_ERROR;
-						break;
-				}
-
-				road |= tram;
-
-				if ((track == TRACK_X && road == ROAD_Y) ||
-						(track == TRACK_Y && road == ROAD_X)) {
+							roadtypes |= ROADTYPES_ROAD;
+							break;
+
+						case ROADTYPES_ALL:
+							if (road != tram) return CMD_ERROR;
+							break;
+					}
+
+					road |= tram;
+
 					if (flags & DC_EXEC) {
 						MakeRoadCrossing(tile, GetRoadOwner(tile, ROADTYPE_ROAD), GetRoadOwner(tile, ROADTYPE_TRAM), _current_company, (track == TRACK_X ? AXIS_Y : AXIS_X), railtype, roadtypes, GetTownIndex(tile));
 						UpdateLevelCrossing(tile, false);
