Index: train_cmd.c
===================================================================
--- train_cmd.c	(revision 5062)
+++ train_cmd.c	(working copy)
@@ -2355,7 +2355,7 @@
 static bool ProcessTrainOrder(Vehicle *v)
 {
 	const Order *order;
-	bool result;
+	bool at_waypoint = false;
 
 	switch (v->current_order.type) {
 		case OT_GOTO_DEPOT:
@@ -2374,6 +2374,7 @@
 	// check if we've reached the waypoint?
 	if (v->current_order.type == OT_GOTO_WAYPOINT && v->tile == v->dest_tile) {
 		v->cur_order_index++;
+		at_waypoint = true;
 	}
 
 	// check if we've reached a non-stop station while TTDPatch nonstop is enabled..
@@ -2408,29 +2409,28 @@
 
 	v->dest_tile = 0;
 
-	result = false;
+	InvalidateVehicleOrder(v);
+
 	switch (order->type) {
 		case OT_GOTO_STATION:
 			if (order->station == v->last_station_visited)
 				v->last_station_visited = INVALID_STATION;
 			v->dest_tile = GetStation(order->station)->xy;
-			result = CheckReverseTrain(v);
 			break;
 
 		case OT_GOTO_DEPOT:
 			v->dest_tile = GetDepot(order->station)->xy;
-			result = CheckReverseTrain(v);
 			break;
 
 		case OT_GOTO_WAYPOINT:
 			v->dest_tile = GetWaypoint(order->station)->xy;
-			result = CheckReverseTrain(v);
 			break;
-	}
 
-	InvalidateVehicleOrder(v);
+		default:
+			return false;
+	}
 
-	return result;
+	return !at_waypoint && CheckReverseTrain(v);
 }
 
 static void MarkTrainDirty(Vehicle *v)
