Index: saveload.c
===================================================================
--- saveload.c	(Revision 4244)
+++ saveload.c	(Arbeitskopie)
@@ -688,7 +688,7 @@
 	}

 	for (; sld->cmd != SL_END; sld++) {
-		void *ptr = (byte*)object + (unsigned long)sld->address;
+		void *ptr = (byte*)object + (ptrdiff_t)sld->address;
 		SlObjectMember(ptr, sld);
 	}
 }
Index: settings.h
===================================================================
--- settings.h	(Revision 4244)
+++ settings.h	(Arbeitskopie)
@@ -71,7 +71,7 @@
  * to add this to the address of the object */
 static inline void *ini_get_variable(const SaveLoad *sld, const void *object)
 {
-	return (object == NULL) ? sld->address : (byte*)object + (unsigned long)sld->address;
+	return (object == NULL) ? sld->address : (byte*)object + (ptrdiff_t)sld->address;
 }

 void IConsoleSetPatchSetting(const char *name, const char *value);