Index: src/industry_cmd.cpp
===================================================================
--- src/industry_cmd.cpp	(revision 10298)
+++ src/industry_cmd.cpp	(working copy)
@@ -282,14 +282,12 @@
 	z = ti->z;
 	/* Add bricks below the industry? */
 	if (ti->tileh != SLOPE_FLAT) {
-		AddSortableSpriteToDraw(SPR_FOUNDATION_BASE + ti->tileh, PAL_NONE, ti->x, ti->y, 16, 16, 7, z);
-		AddChildSpriteScreen(image, pal, 31, 1);
+		DrawFoundation(ti, ti->tileh);
 		z += TILE_HEIGHT;
-	} else {
-		/* Else draw regular ground */
-		DrawGroundSprite(image, pal);
 	}
 
+	DrawGroundSprite(image, pal);
+
 	/* Add industry on top of the ground? */
 	image = dits->building.sprite;
 	if (image != 0) {
