diff -r 9a87f6fd49a0 src/settings_gui.cpp
--- a/src/settings_gui.cpp	Sun Nov 23 12:34:24 2008 +0100
+++ b/src/settings_gui.cpp	Sun Nov 23 13:00:37 2008 +0100
@@ -1032,7 +1032,7 @@
  */
 void DrawArrowButtons(int x, int y, Colours button_colour, byte state, bool clickable_left, bool clickable_right)
 {
-	int colour = _colour_gradient[COLOUR_YELLOW][2];
+	int colour = _colour_gradient[button_colour][2];
 
 	DrawFrameRect(x,      y + 1, x +  9, y + 9, button_colour, (state == 1) ? FR_LOWERED : FR_NONE);
 	DrawFrameRect(x + 10, y + 1, x + 19, y + 9, button_colour, (state == 2) ? FR_LOWERED : FR_NONE);
