# HG changeset patch
# Parent 2e126826e81063a82cc5015c9787b43681a7a613

diff --git a/src/textfile_gui.cpp b/src/textfile_gui.cpp
--- a/src/textfile_gui.cpp
+++ b/src/textfile_gui.cpp
@@ -226,5 +226,8 @@
 	if (FioCheckFileExists(file_path, dir)) return file_path;
 
 	seprintf(slash + 1, lastof(file_path), "%s.txt", prefix);
-	return FioCheckFileExists(file_path, dir) ? file_path : NULL;
+	if (FioCheckFileExists(file_path, dir)) return file_path;
+
+	seprintf(slash + 1, lastof(file_path), "COPYING");
+	return (FioCheckFileExists(file_path, dir) && prefix == prefixes[2]) ? file_path : NULL;
 }
