Index: src/video/sdl_v.cpp
===================================================================
--- src/video/sdl_v.cpp	(revision 18031)
+++ src/video/sdl_v.cpp	(working copy)
@@ -461,8 +461,10 @@
 			break;
 
 		case SDL_VIDEORESIZE: {
+			printf("resizeevent %d %d\n", ev.resize.w, ev.resize.h);
 			int w = max(ev.resize.w, 64);
 			int h = max(ev.resize.h, 64);
+			CSleep(200);
 			ChangeResInGame(w, h);
 			break;
 		}
