Index: src/base_media_func.h
===================================================================
--- src/base_media_func.h	(revision 24494)
+++ src/base_media_func.h	(working copy)
@@ -403,5 +403,6 @@
 	template const set_type *repl_type::GetSet(int index); \
 	template const set_type *repl_type::GetUsedSet(); \
 	template bool repl_type::DetermineBestSet(); \
-	template set_type *repl_type::GetAvailableSets();
+	template set_type *repl_type::GetAvailableSets(); \
+	template const char *TryGetBaseSetFile(const ContentInfo *ci, bool md5sum, const set_type *s);
 
