Index: src/aircraft_cmd.cpp
===================================================================
--- src/aircraft_cmd.cpp	(revision 14293)
+++ src/aircraft_cmd.cpp	(working copy)
@@ -1596,7 +1596,16 @@
 			return;
 		default:  // orders have been deleted (no orders), goto depot and don't bother us
 			v->current_order.Free();
-			v->u.air.state = HANGAR;
+
+			/* Do we have any hangars? */
+			if (GetStation(v->u.air.targetairport)->Airport()->nof_depots != 0)
+			{
+				v->u.air.state = HANGAR;
+			}
+			else /* ... no we haven't. Sorry, fly away. */
+			{
+				v->u.air.state = (v->subtype == AIR_HELICOPTER) ? HELITAKEOFF : TAKEOFF;
+			}
 	}
 	AirportMove(v, apc);
 }
