Index: src/road_cmd.cpp
===================================================================
--- src/road_cmd.cpp	(revision 24623)
+++ src/road_cmd.cpp	(working copy)
@@ -795,7 +795,7 @@
 static bool CanConnectToRoad(TileIndex tile, RoadType rt, DiagDirection dir)
 {
 	RoadBits bits = GetAnyRoadBits(tile + TileOffsByDiagDir(dir), rt, false);
-	return (bits & DiagDirToRoadBits(ReverseDiagDir(dir)));
+	return (bits & DiagDirToRoadBits(ReverseDiagDir(dir))) != 0;
 }
 
 /**
