Index: src/group_cmd.cpp
===================================================================
--- src/group_cmd.cpp	(revision 11015)
+++ src/group_cmd.cpp	(working copy)
@@ -216,12 +216,13 @@
 
 	if (!IsValidVehicleID(p2) || (!IsValidGroupID(new_g) && !IsDefaultGroupID(new_g))) return CMD_ERROR;
 
+	Vehicle *v = GetVehicle(p2);
+
 	if (IsValidGroupID(new_g)) {
 		Group *g = GetGroup(new_g);
-		if (g->owner != _current_player) return CMD_ERROR;
+		if (g->owner != _current_player || g->vehicle_type != v->type) return CMD_ERROR;
 	}
 
-	Vehicle *v = GetVehicle(p2);
 	if (v->owner != _current_player || !v->IsPrimaryVehicle()) return CMD_ERROR;
 
 	if (flags & DC_EXEC) {
