diff -r 907f1c3d3119 findversion.sh
--- a/findversion.sh	Sat May 22 19:44:21 2010 +0200
+++ b/findversion.sh	Sat May 22 21:32:55 2010 +0200
@@ -99,7 +99,7 @@
 	if [ -n "`hg status \"$SRC_DIR\" | grep -v '^?'`" ]; then
 		MODIFIED="2"
 	fi
-	HASH=`LC_ALL=C hg parents --template="{node}"`
+	HASH=`LC_ALL=C hg id -i | cut -c1-12`
 	REV="h`echo $HASH | cut -c1-8`"
 	BRANCH=`hg branch | sed 's@^default$@@'`
 	REV_NR=`LC_ALL=C hg log -r $HASH:0 -k "svn" -l 1 --template "{desc}\n" "$SRC_DIR" | grep "^(svn r[0-9]*)" | head -n 1 | sed "s@.*(svn r\([0-9]*\)).*@\1@"`
