diff --git a/src/video/win32_v.cpp b/src/video/win32_v.cpp
index beeef29..be5c859 100644
--- a/src/video/win32_v.cpp
+++ b/src/video/win32_v.cpp
@@ -282,6 +282,11 @@ bool VideoDriver_Win32::MakeWindow(bool full_screen)
 		settings.dmPelsHeight = _wnd.height_org;
 		settings.dmDisplayFrequency = _display_hz;
 
+		/* Check for 8 bpp support. */
+		if (settings.dmBitsPerPel != 32 && ChangeDisplaySettings(&settings, CDS_FULLSCREEN | CDS_TEST) != DISP_CHANGE_SUCCESSFUL) {
+			settings.dmBitsPerPel = 32;
+		}
+
 		/* Test fullscreen with current resolution, if it fails use desktop resolution. */
 		if (ChangeDisplaySettings(&settings, CDS_FULLSCREEN | CDS_TEST) != DISP_CHANGE_SUCCESSFUL) {
 			RECT r;
