Index: src/widget_type.h
===================================================================
--- src/widget_type.h	(revision 20510)
+++ src/widget_type.h	(working copy)
@@ -186,7 +186,7 @@
  */
 FORCEINLINE uint NWidgetBase::GetHorizontalStepSize(SizingType sizing) const
 {
-	return (sizing == ST_RESIZE) ? this->resize_x : this->fill_x;
+	return (sizing == ST_RESIZE || this->resize_x != 0) ? this->resize_x : this->fill_x;
 }
 
 /**
@@ -195,7 +195,7 @@
  */
 FORCEINLINE uint NWidgetBase::GetVerticalStepSize(SizingType sizing) const
 {
-	return (sizing == ST_RESIZE) ? this->resize_y : this->fill_y;
+	return (sizing == ST_RESIZE || this->resize_y != 0) ? this->resize_y : this->fill_y;
 }
 
 /**
