Index: src/main_gui.cpp
===================================================================
--- src/main_gui.cpp	(revision 10380)
+++ src/main_gui.cpp	(working copy)
@@ -159,7 +159,7 @@
 /** Toggle all transparency options, except for signs */
 static void ToggleTransparency()
 {
-	static byte trans_opt = ~0;
+	static uint trans_opt = ~0;
 
 	if (GB(_transparent_opt, 1, 7) == 0) {
 		SB(_transparent_opt, 1, 7, GB(trans_opt, 1, 7));
Index: src/transparency_gui.cpp
===================================================================
--- src/transparency_gui.cpp	(revision 10380)
+++ src/transparency_gui.cpp	(working copy)
@@ -31,7 +31,7 @@
  * when clicking on a widget, and play a sound
  * @param widget been clicked.
  */
-static void Transparent_Click(byte widget)
+static void Transparent_Click(uint widget)
 {
 	TOGGLEBIT(_transparent_opt, widget);
 	SndPlayFx(SND_15_BEEP);
Index: src/variables.h
===================================================================
--- src/variables.h	(revision 10380)
+++ src/variables.h	(working copy)
@@ -284,7 +284,7 @@
 VARDEF int _autosave_ctr;
 
 VARDEF byte _display_opt;
-VARDEF byte _transparent_opt;
+VARDEF uint _transparent_opt;
 VARDEF int _caret_timer;
 VARDEF uint32 _news_display_opt;
 VARDEF bool _news_ticker_sound;
