Index: os/macosx/plistgen.sh
===================================================================
--- os/macosx/plistgen.sh	(revision 25755)
+++ os/macosx/plistgen.sh	(working copy)
@@ -44,5 +44,81 @@
         <string>Copyright 2004-$date The OpenTTD team</string>
         <key>NSPrincipalClass</key>
         <string>NSApplication</string>
+        <key>UTExportedTypeDeclarations</key>
+        <array>
+                <dict>
+                        <key>UTTypeConformsTo</key>
+                        <array>
+                                <string>public.data</string>
+                        </array>
+                        <key>UTTypeIdentifier</key>
+                        <string>org.openttd.savegame</string>
+                        <key>UTTypeDescription</key>
+                        <string>OpenTTD Save</string>
+                        <key>UTTypeTagSpecification</key>
+                        <dict>
+                                <key>public.filename-extension</key>
+                                <array>
+                                        <string>sav</string>
+                                </array>
+                        </dict>
+                </dict>
+                <dict>
+                        <key>UTTypeConformsTo</key>
+                        <array>
+                                <string>public.data</string>
+                        </array>
+                        <key>UTTypeIdentifier</key>
+                        <string>org.openttd.scenario</string>
+                        <key>UTTypeDescription</key>
+                        <string>OpenTTD Scenario</string>
+                        <key>UTTypeTagSpecification</key>
+                        <dict>
+                                <key>public.filename-extension</key>
+                                <array>
+                                        <string>scn</string>
+                                </array>
+                        </dict>
+                </dict>
+        </array>
+        <key>CFBundleDocumentTypes</key>
+        <array>
+                <dict>
+                        <key>CFBundleTypeExtensions</key>
+                        <array>
+                                <string>sav</string>
+                        </array>
+                        <key>CFBundleTypeName</key>
+                        <string>OpenTTD Save</string>
+                        <key>LSHandlerRank</key>
+                        <string>Owner</string>
+                        <key>CFBundleTypeRole</key>
+                        <string>Editor</string>
+                        <key>LSTypeIsPackage</key>
+                        <false/>
+                        <key>LSItemContentTypes</key>
+                        <array>
+                                <string>org.openttd.savegame</string>
+                        </array>
+                </dict>
+                <dict>
+                        <key>CFBundleTypeExtensions</key>
+                        <array>
+                                <string>scn</string>
+                        </array>
+                        <key>CFBundleTypeName</key>
+                        <string>OpenTTD Scenario</string>
+                        <key>LSHandlerRank</key>
+                        <string>Owner</string>
+                        <key>CFBundleTypeRole</key>
+                        <string>Editor</string>
+                        <key>LSTypeIsPackage</key>
+                        <false/>
+                        <key>LSItemContentTypes</key>
+                        <array>
+                                <string>org.openttd.scenario</string>
+                        </array>
+                </dict>
+        </array>
 </dict>
 </plist>" > "$1"/Contents/Info.plist
Index: src/video/cocoa/cocoa_v.mm
===================================================================
--- src/video/cocoa/cocoa_v.mm	(revision 25755)
+++ src/video/cocoa/cocoa_v.mm	(working copy)
@@ -35,6 +35,8 @@
 #include "../../gfx_func.h"
 #include "../../window_func.h"
 #include "../../window_gui.h"
+#include "../../saveload/saveload.h"
+#include "../../fios.h"
 
 #import <sys/param.h> /* for MAXPATHLEN */
 
@@ -105,6 +107,19 @@
 }
 
 /**
+ * Called to open a document. Can be called before applicationDidFinishLaunching:
+ */
+- (BOOL) application: (NSApplication *)theApplication openFile:(NSString *)filename
+{
+	strecpy(_file_to_saveload.name, [ filename fileSystemRepresentation ], lastof(_file_to_saveload.name));
+	_switch_mode = SM_LOAD_GAME;
+	_file_to_saveload.mode = SL_LOAD;
+	
+	return YES;
+}
+ 
+
+/**
  * Display the in game quit confirmation dialog.
  */
 - (NSApplicationTerminateReply)applicationShouldTerminate:(NSApplication*) sender
