Index: src/engine_gui.cpp
===================================================================
--- src/engine_gui.cpp	(revision 14557)
+++ src/engine_gui.cpp	(working copy)
@@ -191,13 +191,13 @@
 	SetDParam(0, GetEngineCategoryName(engine));
 	DrawStringMultiCenter(w->width >> 1, 20, STR_NEW_VEHICLE_NOW_AVAILABLE, w->width - 2);
 
-	GfxFillRect(25, 56, w->width - 25, w->height - 2, 10);
+	GfxFillRect(25, 56, w->width - 25, w->height - 2, ni->flags & NF_INCOLOR ? 222 : 10); //Changes the background color to a light blue (222) or uses the dark grey (10)
 
 	SetDParam(0, engine);
 	DrawStringMultiCenter(w->width >> 1, 57, STR_NEW_VEHICLE_TYPE, w->width - 2);
 
 	dei->engine_proc(w->width >> 1, 88, engine, 0);
-	GfxFillRect(25, 56, w->width - 56, 112, PALETTE_TO_STRUCT_GREY, FILLRECT_RECOLOR);
+	GfxFillRect(25, 56, w->width - 56, 112, (ni->flags & NF_INCOLOR ? PALETTE_TO_STRUCT_YELLOW : PALETTE_TO_STRUCT_GREY), FILLRECT_RECOLOR); //I'm a bit unsure why, but everything before PALETTE_TO_TRANSPARENT works
 	dei->info_proc(engine, w->width >> 1, 129, w->width - 52);
 }
 
Index: src/news_gui.cpp
===================================================================
--- src/news_gui.cpp	(revision 14557)
+++ src/news_gui.cpp	(working copy)
@@ -235,7 +235,7 @@
 					ViewPort *vp = this->viewport;
 					GfxFillRect(vp->left - this->left, vp->top - this->top,
 						vp->left - this->left + vp->width - 1, vp->top - this->top + vp->height - 1,
-						(this->ni->flags & NF_INCOLOR ? PALETTE_TO_TRANSPARENT : PALETTE_TO_STRUCT_GREY), FILLRECT_RECOLOR
+						(this->ni->flags & NF_INCOLOR ? PALETTE_TO_STRUCT_YELLOW : PALETTE_TO_STRUCT_GREY), FILLRECT_RECOLOR //I'm a bit unsure why, but everything before PALETTE_TO_TRANSPARENT works
 					);
 
 					CopyInDParam(0, this->ni->params, lengthof(this->ni->params));
