# HG changeset patch
# Parent db87c54707c9debd58ae3a9b905414cfa28f9d18

diff -r db87c54707c9 src/rail_gui.cpp
--- a/src/rail_gui.cpp	Tue Sep 03 09:44:22 2013 +0200
+++ b/src/rail_gui.cpp	Tue Sep 03 09:44:23 2013 +0200
@@ -1506,7 +1506,10 @@
 
 	virtual void UpdateWidgetSize(int widget, Dimension *size, const Dimension &padding, Dimension *fill, Dimension *resize)
 	{
-		if (IsInsideMM(widget, WID_BS_SEMAPHORE_NORM, WID_BS_ELECTRIC_PBS_OWAY + 1)) {
+		if (widget == WID_BS_DRAG_SIGNALS_DENSITY_LABEL) {
+			/* Two digits for signals density. */
+			size->width = max(size->width, 2 * GetDigitWidth() + padding.width);
+		} else if (IsInsideMM(widget, WID_BS_SEMAPHORE_NORM, WID_BS_ELECTRIC_PBS_OWAY + 1)) {
 			/* Extract signal from widget number. */
 			int type = (widget - WID_BS_SEMAPHORE_NORM) % SIGTYPE_END;
 			int var = (widget < WID_BS_ELECTRIC_NORM) ? SIG_SEMAPHORE : SIG_ELECTRIC;
