Index: src/tunnelbridge_cmd.cpp
===================================================================
--- src/tunnelbridge_cmd.cpp	(revision 20956)
+++ src/tunnelbridge_cmd.cpp	(working copy)
@@ -881,8 +881,8 @@
 
 	static const int INF = 1000; ///< big number compared to sprite size
 	static const SubSprite half_pillar_sub_sprite[2][2] = {
-		{ {  -14, -INF, INF, INF }, { -INF, -INF, -15, INF } }, // X axis, north and south
-		{ { -INF, -INF,  15, INF }, {   16, -INF, INF, INF } }, // Y axis, north and south
+		{ {  -16, -INF, INF, INF }, { -INF, -INF, -17, INF } }, // X axis, north and south
+		{ { -INF, -INF,  17, INF }, {   18, -INF, INF, INF } }, // Y axis, north and south
 	};
 
 	if (psid->sprite == 0) return;
