Index: src/settings.cpp
===================================================================
--- src/settings.cpp	(Revision 24113)
+++ src/settings.cpp	(Arbeitskopie)
@@ -359,7 +359,7 @@
 		SetDParamStr(0, str);
 		SetDParamStr(1, desc->name);
 		ShowErrorMessage(STR_CONFIG_ERROR, STR_CONFIG_ERROR_INVALID_VALUE, WL_CRITICAL);
-		return 0;
+		return desc->def;
 	}
 	case SDT_MANYOFMANY: {
 		size_t r = LookupManyOfMany(desc->many, str);
@@ -367,7 +367,7 @@
 		SetDParamStr(0, str);
 		SetDParamStr(1, desc->name);
 		ShowErrorMessage(STR_CONFIG_ERROR, STR_CONFIG_ERROR_INVALID_VALUE, WL_CRITICAL);
-		return NULL;
+		return desc->def;
 	}
 	case SDT_BOOLX:
 		if (strcmp(str, "true")  == 0 || strcmp(str, "on")  == 0 || strcmp(str, "1") == 0) return (void*)true;
@@ -376,7 +376,7 @@
 		SetDParamStr(0, str);
 		SetDParamStr(1, desc->name);
 		ShowErrorMessage(STR_CONFIG_ERROR, STR_CONFIG_ERROR_INVALID_VALUE, WL_CRITICAL);
-		break;
+		return desc->def;
 
 	case SDT_STRING: return orig_str;
 	case SDT_INTLIST: return str;
