diff -ur openttd-0.6.0-beta1/src/lang/english.txt openttd-0.6.0-beta1-UI-growth-rate-patch/src/lang/english.txt
--- openttd-0.6.0-beta1/src/lang/english.txt	2007-11-18 14:55:23.000000000 +0100
+++ openttd-0.6.0-beta1-UI-growth-rate-patch/src/lang/english.txt	2007-12-04 22:47:18.000000000 +0100
@@ -370,6 +370,7 @@
 
 STR_SORT_BY_POPULATION                                          :{BLACK}Population
 STR_SORT_BY_PRODUCTION                                          :{BLACK}Production
+STR_SORT_BY_GROWTH                                          :{BLACK}Growth Rate
 STR_SORT_BY_TYPE                                                :{BLACK}Type
 STR_SORT_BY_TRANSPORTED                                         :{BLACK}Transported
 STR_SORT_BY_NAME                                                :{BLACK}Name
@@ -1788,12 +1789,13 @@
 STR_TOWN_BRIBE_THE_LOCAL_AUTHORITY_DESC                         :{WHITE}{STRING}{}{YELLOW} Bribe the local authority to increase your rating, at the risk of a severe penalty if caught.{}  Cost: {CURRENCY}
 STR_2055_TRAFFIC_CHAOS_IN_ROAD_REBUILDING                       :{BIGFONT}{BLACK}Traffic chaos in {TOWN}!{}{}Road rebuilding programme funded by {COMPANY} brings 6 months of misery to motorists!
 STR_2056                                                        :{TINYFONT}{WHITE}{TOWN}
-STR_2057                                                        :{ORANGE}{TOWN}{BLACK} ({COMMA})
+STR_2057                                                        :{ORANGE}{TOWN}{BLACK} ({COMMA}){YELLOW} (growth-rate: {COMMA})
 STR_2058_UNDER_CONSTRUCTION                                     :{STRING} (under construction)
 STR_2059_IGLOO                                                  :Igloo
 STR_205A_TEPEES                                                 :Tepees
 STR_205B_TEAPOT_HOUSE                                           :Teapot-House
 STR_205C_PIGGY_BANK                                             :Piggy-Bank
+STR_205D_GROWTH_RATE                            :{BLACK}Growth-rate: {ORANGE}{COMMA}
 
 STR_INDUSTRY                                                    :{INDUSTRY}
 STR_TOWN                                                        :{TOWN}
diff -ur openttd-0.6.0-beta1/src/town_gui.cpp openttd-0.6.0-beta1-UI-growth-rate-patch/src/town_gui.cpp
--- openttd-0.6.0-beta1/src/town_gui.cpp	2007-11-04 01:08:57.000000000 +0100
+++ openttd-0.6.0-beta1-UI-growth-rate-patch/src/town_gui.cpp	2007-12-04 22:50:26.000000000 +0100
@@ -273,14 +273,17 @@
 		SetDParam(0, t->population);
 		SetDParam(1, t->num_houses);
 		DrawString(2, 107, STR_2006_POPULATION, TC_FROMSTRING);
+		
+		SetDParam(0, t->growthRateAsHumanReadable());
+		DrawString(2, 117,STR_205D_GROWTH_RATE, TC_FROMSTRING);
 
 		SetDParam(0, t->act_pass);
 		SetDParam(1, t->max_pass);
-		DrawString(2, 117, STR_200D_PASSENGERS_LAST_MONTH_MAX, TC_FROMSTRING);
+		DrawString(2, 127, STR_200D_PASSENGERS_LAST_MONTH_MAX, TC_FROMSTRING);
 
 		SetDParam(0, t->act_mail);
 		SetDParam(1, t->max_mail);
-		DrawString(2, 127, STR_200E_MAIL_LAST_MONTH_MAX, TC_FROMSTRING);
+		DrawString(2, 137, STR_200E_MAIL_LAST_MONTH_MAX, TC_FROMSTRING);
 
 		DrawWindowViewport(w);
 		break;
@@ -325,17 +328,17 @@
 {   WWT_CLOSEBOX,   RESIZE_NONE,    13,     0,    10,     0,    13, STR_00C5,                 STR_018B_CLOSE_WINDOW},
 {    WWT_CAPTION,   RESIZE_NONE,    13,    11,   247,     0,    13, STR_2005,                 STR_018C_WINDOW_TITLE_DRAG_THIS},
 {  WWT_STICKYBOX,   RESIZE_NONE,    13,   248,   259,     0,    13, 0x0,                      STR_STICKY_BUTTON},
-{      WWT_PANEL,   RESIZE_NONE,    13,     0,   259,    14,   105, 0x0,                      STR_NULL},
-{      WWT_INSET,   RESIZE_NONE,    13,     2,   257,    16,   103, 0x0,                      STR_NULL},
-{      WWT_PANEL,   RESIZE_NONE,    13,     0,   259,   106,   137, 0x0,                      STR_NULL},
-{ WWT_PUSHTXTBTN,   RESIZE_NONE,    13,     0,    85,   138,   149, STR_00E4_LOCATION,        STR_200B_CENTER_THE_MAIN_VIEW_ON},
-{ WWT_PUSHTXTBTN,   RESIZE_NONE,    13,    86,   171,   138,   149, STR_2020_LOCAL_AUTHORITY, STR_2021_SHOW_INFORMATION_ON_LOCAL},
-{ WWT_PUSHTXTBTN,   RESIZE_NONE,    13,   172,   259,   138,   149, STR_0130_RENAME,          STR_200C_CHANGE_TOWN_NAME},
+{      WWT_PANEL,   RESIZE_NONE,    13,     0,   259,    14,   105, 0x0,                      STR_NULL}, // view's border
+{      WWT_INSET,   RESIZE_NONE,    13,     2,   257,    16,   103, 0x0,                      STR_NULL}, // view
+{      WWT_PANEL,   RESIZE_NONE,    13,     0,   259,   106,   147, 0x0,                      STR_NULL}, //text area
+{ WWT_PUSHTXTBTN,   RESIZE_NONE,    13,     0,    85,   148,   159, STR_00E4_LOCATION,        STR_200B_CENTER_THE_MAIN_VIEW_ON},
+{ WWT_PUSHTXTBTN,   RESIZE_NONE,    13,    86,   178,   148,   159, STR_2020_LOCAL_AUTHORITY, STR_2021_SHOW_INFORMATION_ON_LOCAL},
+{ WWT_PUSHTXTBTN,   RESIZE_NONE,    13,   179,   259,   148,   159, STR_0130_RENAME,          STR_200C_CHANGE_TOWN_NAME},
 {   WIDGETS_END},
 };
 
 static const WindowDesc _town_view_desc = {
-	WDP_AUTO, WDP_AUTO, 260, 150, 260, 150,
+	WDP_AUTO, WDP_AUTO, 260, 160, 260, 160,
 	WC_TOWN_VIEW, WC_NONE,
 	WDF_STD_TOOLTIPS | WDF_STD_BTN | WDF_DEF_WIDGET | WDF_UNCLICK_BUTTONS | WDF_STICKY_BUTTON,
 	_town_view_widgets,
@@ -348,17 +351,17 @@
 {  WWT_STICKYBOX,   RESIZE_NONE,    13,   248,   259,     0,    13, 0x0,               STR_STICKY_BUTTON},
 {      WWT_PANEL,   RESIZE_NONE,    13,     0,   259,    14,   105, 0x0,               STR_NULL},
 {      WWT_INSET,   RESIZE_NONE,    13,     2,   257,    16,   103, 0x0,               STR_NULL},
-{      WWT_PANEL,   RESIZE_NONE,    13,     0,   259,   106,   137, 0x0,               STR_NULL},
-{ WWT_PUSHTXTBTN,   RESIZE_NONE,    13,     0,    85,   138,   149, STR_00E4_LOCATION, STR_200B_CENTER_THE_MAIN_VIEW_ON},
+{      WWT_PANEL,   RESIZE_NONE,    13,     0,   259,   106,   147, 0x0,               STR_NULL},
+{ WWT_PUSHTXTBTN,   RESIZE_NONE,    13,     0,    85,   148,   159, STR_00E4_LOCATION, STR_200B_CENTER_THE_MAIN_VIEW_ON},
 {      WWT_EMPTY,   RESIZE_NONE,     0,     0,     0,     0,     0, 0x0,               STR_NULL},
 { WWT_PUSHTXTBTN,   RESIZE_NONE,    13,   173,   247,     0,    13, STR_0130_RENAME,   STR_200C_CHANGE_TOWN_NAME},
-{ WWT_PUSHTXTBTN,   RESIZE_NONE,    13,    86,   171,   138,   149, STR_023C_EXPAND,   STR_023B_INCREASE_SIZE_OF_TOWN},
-{ WWT_PUSHTXTBTN,   RESIZE_NONE,    13,   172,   259,   138,   149, STR_0290_DELETE,   STR_0291_DELETE_THIS_TOWN_COMPLETELY},
+{ WWT_PUSHTXTBTN,   RESIZE_NONE,    13,    86,   171,   148,   159, STR_023C_EXPAND,   STR_023B_INCREASE_SIZE_OF_TOWN},
+{ WWT_PUSHTXTBTN,   RESIZE_NONE,    13,   172,   259,   148,   159, STR_0290_DELETE,   STR_0291_DELETE_THIS_TOWN_COMPLETELY},
 {   WIDGETS_END},
 };
 
 static const WindowDesc _town_view_scen_desc = {
-	WDP_AUTO, WDP_AUTO, 260, 150, 260, 150,
+	WDP_AUTO, WDP_AUTO, 260, 160, 260, 160,
 	WC_TOWN_VIEW, WC_NONE,
 	WDF_STD_TOOLTIPS | WDF_STD_BTN | WDF_DEF_WIDGET | WDF_UNCLICK_BUTTONS | WDF_STICKY_BUTTON,
 	_town_view_scen_widgets,
@@ -383,14 +386,15 @@
 
 static const Widget _town_directory_widgets[] = {
 {   WWT_CLOSEBOX,   RESIZE_NONE,    13,     0,    10,     0,    13, STR_00C5,               STR_018B_CLOSE_WINDOW},
-{    WWT_CAPTION,   RESIZE_NONE,    13,    11,   195,     0,    13, STR_2000_TOWNS,         STR_018C_WINDOW_TITLE_DRAG_THIS},
-{  WWT_STICKYBOX,   RESIZE_NONE,    13,   196,   207,     0,    13, 0x0,                    STR_STICKY_BUTTON},
+{    WWT_CAPTION,   RESIZE_NONE,    13,    11,   295,     0,    13, STR_2000_TOWNS,         STR_018C_WINDOW_TITLE_DRAG_THIS},
+{  WWT_STICKYBOX,   RESIZE_NONE,    13,   296,   307,     0,    13, 0x0,                    STR_STICKY_BUTTON},
 { WWT_PUSHTXTBTN,   RESIZE_NONE,    13,     0,    98,    14,    25, STR_SORT_BY_NAME,       STR_SORT_ORDER_TIP},
 { WWT_PUSHTXTBTN,   RESIZE_NONE,    13,    99,   195,    14,    25, STR_SORT_BY_POPULATION, STR_SORT_ORDER_TIP},
-{      WWT_PANEL, RESIZE_BOTTOM,    13,     0,   195,    26,   189, 0x0,                    STR_200A_TOWN_NAMES_CLICK_ON_NAME},
-{  WWT_SCROLLBAR, RESIZE_BOTTOM,    13,   196,   207,    14,   189, 0x0,                    STR_0190_SCROLL_BAR_SCROLLS_LIST},
-{      WWT_PANEL,     RESIZE_TB,    13,     0,   195,   190,   201, 0x0,                    STR_NULL},
-{  WWT_RESIZEBOX,     RESIZE_TB,    13,   196,   207,   190,   201, 0x0,                    STR_RESIZE_BUTTON},
+{ WWT_PUSHTXTBTN,   RESIZE_NONE,    13,    196,  295,    14,    25, STR_SORT_BY_GROWTH, STR_SORT_ORDER_TIP},
+{      WWT_PANEL, RESIZE_BOTTOM,    13,     0,   295,    26,   189, 0x0,                    STR_200A_TOWN_NAMES_CLICK_ON_NAME},
+{  WWT_SCROLLBAR, RESIZE_BOTTOM,    13,   296,   307,    14,   189, 0x0,                    STR_0190_SCROLL_BAR_SCROLLS_LIST},
+{      WWT_PANEL,     RESIZE_TB,    13,     0,   295,   190,   201, 0x0,                    STR_NULL}, // world population
+{  WWT_RESIZEBOX,     RESIZE_TB,    13,   296,   307,   190,   201, 0x0,                    STR_RESIZE_BUTTON},
 {   WIDGETS_END},
 };
 
@@ -434,6 +438,15 @@
 	return r;
 }
 
+static int CDECL TownGrowthSorter(const void *a, const void *b)
+{
+	const Town* ta = *(const Town**)a;
+	const Town* tb = *(const Town**)b;
+	int r = tb->growth_rate - ta->growth_rate;
+	if (_town_sort_order & 1) r = -r;
+	return r;
+}
+
 static void MakeSortedTownList()
 {
 	const Town* t;
@@ -448,7 +461,7 @@
 	_num_town_sort = n;
 
 	_last_town = NULL; // used for "cache"
-	qsort((void*)_town_sort, n, sizeof(_town_sort[0]), _town_sort_order & 2 ? TownPopSorter : TownNameSorter);
+	qsort((void*)_town_sort, n, sizeof(_town_sort[0]), _town_sort_order & 2 ? TownPopSorter : (_town_sort_order & 4 ? TownGrowthSorter : TownNameSorter));
 
 	DEBUG(misc, 3, "Resorting towns list");
 }
@@ -466,7 +479,8 @@
 		SetVScrollCount(w, _num_town_sort);
 
 		DrawWindowWidgets(w);
-		DoDrawString(_town_sort_order & 1 ? DOWNARROW : UPARROW, (_town_sort_order <= 1) ? 88 : 187, 15, TC_BLACK);
+		
+		DoDrawString(_town_sort_order & 1 ? DOWNARROW : UPARROW, _town_sort_order & 2 ? 185 : ( _town_sort_order & 4 ? 285 : 88)   , 15, TC_BLACK);
 
 		{
 			int n = 0;
@@ -480,6 +494,7 @@
 
 				SetDParam(0, t->index);
 				SetDParam(1, t->population);
+				SetDParam(2, t->growthRateAsHumanReadable());
 				DrawString(2, y, STR_2057, TC_FROMSTRING);
 
 				y += 10;
@@ -504,8 +519,13 @@
 			_town_sort_dirty = true;
 			SetWindowDirty(w);
 		} break;
-
-		case 5: { /* Click on Town Matrix */
+		
+		case 5 : { 
+			_town_sort_order = (_town_sort_order == 4 ) ? 5 : 4;
+			_town_sort_dirty = true;
+			SetWindowDirty(w);
+		} break;
+		case 6: { /* Click on Town Matrix */
 			const Town* t;
 
 			uint16 id_v = (e->we.click.pt.y - 28) / 10;
@@ -535,7 +555,7 @@
 }
 
 static const WindowDesc _town_directory_desc = {
-	WDP_AUTO, WDP_AUTO, 208, 202, 208, 202,
+	WDP_AUTO, WDP_AUTO, 308, 202, 308, 202,
 	WC_TOWN_DIRECTORY, WC_NONE,
 	WDF_STD_TOOLTIPS | WDF_STD_BTN | WDF_DEF_WIDGET | WDF_UNCLICK_BUTTONS | WDF_STICKY_BUTTON | WDF_RESIZABLE,
 	_town_directory_widgets,
diff -ur openttd-0.6.0-beta1/src/town.h openttd-0.6.0-beta1-UI-growth-rate-patch/src/town.h
--- openttd-0.6.0-beta1/src/town.h	2007-08-30 22:40:33.000000000 +0200
+++ openttd-0.6.0-beta1-UI-growth-rate-patch/src/town.h	2007-12-04 20:31:09.000000000 +0100
@@ -159,6 +159,12 @@
 	/** Destroy the town */
 	~Town();
 
+	inline uint16 growthRateAsHumanReadable() const {
+		// growth_rate = how many ticks to next growth; 250 is default; *100 to get nice int values
+		// +0.5 - not accurate, but fast (we don't need absolute accurancy here)
+		return (uint16)((1/(double)growth_rate)*250*100+0.5);
+	}
+
 	inline bool IsValid() const { return this->xy != 0; }
 };
 
