Index: src/airport_gui.cpp
===================================================================
--- src/airport_gui.cpp	(revision 12968)
+++ src/airport_gui.cpp	(working copy)
@@ -40,12 +40,7 @@
 	DoCommandP(tile, _selected_airport_type, _ctrl_pressed, CcBuildAirport, CMD_BUILD_AIRPORT | CMD_NO_WATER | CMD_MSG(STR_A001_CAN_T_BUILD_AIRPORT_HERE));
 }
 
-static void PlaceAir_DemolishArea(TileIndex tile)
-{
-	VpStartPlaceSizing(tile, VPM_X_AND_Y, DDSP_DEMOLISH_AREA);
-}
 
-
 enum {
 	ATW_AIRPORT  = 3,
 	ATW_DEMOLISH = 4
@@ -59,7 +54,7 @@
 
 static void BuildAirClick_Demolish(Window *w)
 {
-	HandlePlacePushButton(w, ATW_DEMOLISH, ANIMCURSOR_DEMOLISH, VHM_RECT, PlaceAir_DemolishArea);
+	HandlePlacePushButton(w, ATW_DEMOLISH, ANIMCURSOR_DEMOLISH, VHM_RECT, PlaceProc_DemolishArea);
 }
 
 
Index: src/dock_gui.cpp
===================================================================
--- src/dock_gui.cpp	(revision 12968)
+++ src/dock_gui.cpp	(working copy)
@@ -58,11 +58,6 @@
 	DoCommandP(tile, 0, 0, CcBuildDocks, CMD_BUILD_BUOY | CMD_MSG(STR_9835_CAN_T_POSITION_BUOY_HERE));
 }
 
-static void PlaceDocks_DemolishArea(TileIndex tile)
-{
-	VpStartPlaceSizing(tile, VPM_X_AND_Y, DDSP_DEMOLISH_AREA);
-}
-
 static void PlaceDocks_BuildCanal(TileIndex tile)
 {
 	VpStartPlaceSizing(tile, VPM_X_OR_Y, DDSP_CREATE_WATER);
@@ -96,7 +91,7 @@
 
 static void BuildDocksClick_Demolish(Window *w)
 {
-	HandlePlacePushButton(w, DTW_DEMOLISH, ANIMCURSOR_DEMOLISH, VHM_RECT, PlaceDocks_DemolishArea);
+	HandlePlacePushButton(w, DTW_DEMOLISH, ANIMCURSOR_DEMOLISH, VHM_RECT, PlaceProc_DemolishArea);
 }
 
 static void BuildDocksClick_Depot(Window *w)
Index: src/road_gui.cpp
===================================================================
--- src/road_gui.cpp	(revision 12968)
+++ src/road_gui.cpp	(working copy)
@@ -230,11 +230,6 @@
 	}
 }
 
-static void PlaceRoad_DemolishArea(TileIndex tile)
-{
-	VpStartPlaceSizing(tile, VPM_X_AND_Y, DDSP_DEMOLISH_AREA);
-}
-
 /** Enum referring to the widgets of the build road toolbar */
 enum RoadToolbarWidgets {
 	RTW_CLOSEBOX = 0,
@@ -322,7 +317,7 @@
 
 static void BuildRoadClick_Demolish(Window *w)
 {
-	HandlePlacePushButton(w, RTW_DEMOLISH, ANIMCURSOR_DEMOLISH, VHM_RECT, PlaceRoad_DemolishArea);
+	HandlePlacePushButton(w, RTW_DEMOLISH, ANIMCURSOR_DEMOLISH, VHM_RECT, PlaceProc_DemolishArea);
 }
 
 static void BuildRoadClick_Depot(Window *w)
