diff -r e25a96e39d2d src/window.cpp
--- a/src/window.cpp	Sat May 22 12:25:47 2010 +0000
+++ b/src/window.cpp	Sat May 22 15:37:53 2010 +0200
@@ -1382,6 +1382,24 @@
 	return false;
 }
 
+static bool HandleMouseDrag()
+{
+	if (_special_mouse_mode != WSM_DRAGDROP) return true;
+	if (!_left_button_down || (_cursor.delta.x == 0 && _cursor.delta.y == 0)) return true;
+
+	Window *w = GetCallbackWnd();
+
+	if (w != NULL) {
+		/* Send an event in client coordinates. */
+		Point pt;
+		pt.x = _cursor.pos.x - w->left;
+		pt.y = _cursor.pos.y - w->top;
+		w->OnMouseDrag(pt, GetWidgetFromPos(w, pt.x, pt.y));
+	}
+
+	return false;
+}
+
 static bool HandleMouseOver()
 {
 	Window *w = FindWindowFromPt(_cursor.pos.x, _cursor.pos.y);
@@ -2057,6 +2075,7 @@
 	UpdateTileSelection();
 
 	if (!VpHandlePlaceSizingDrag())  return;
+	if (!HandleMouseDrag())          return;
 	if (!HandleDragDrop())           return;
 	if (!HandleWindowDragging())     return;
 	if (!HandleScrollbarScrolling()) return;
diff -r e25a96e39d2d src/window_gui.h
--- a/src/window_gui.h	Sat May 22 12:25:47 2010 +0000
+++ b/src/window_gui.h	Sat May 22 15:37:53 2010 +0200
@@ -657,6 +657,13 @@
 	virtual void OnRightClick(Point pt, int widget) {}
 
 	/**
+	 * An 'object' is dragged.
+	 * @param pt     the point inside the window that the mouse hovers over.
+	 * @param widget the widget the mouse hovers over.
+	 */
+	virtual void OnMouseDrag(Point pt, int widget) {}
+
+	/**
 	 * A dragged 'object' has been released.
 	 * @param pt     the point inside the window where the release took place.
 	 * @param widget the widget where the release took place.
