Index: src/ai/ai_gui.cpp
===================================================================
--- src/ai/ai_gui.cpp	(wersja 27565)
+++ src/ai/ai_gui.cpp	(kopia robocza)
@@ -28,6 +28,7 @@
 #include "../widgets/dropdown_type.h"
 #include "../widgets/dropdown_func.h"
 #include "../hotkeys.h"
+#include "../core/geometry_func.hpp"
 
 #include "ai.hpp"
 #include "ai_gui.hpp"
@@ -766,6 +767,18 @@
 				this->line_height = FONT_HEIGHT_NORMAL + WD_MATRIX_TOP + WD_MATRIX_BOTTOM;
 				size->height = 8 * this->line_height;
 				break;
+
+
+			case WID_AIC_CHANGE: {
+				Dimension dim;
+				SetDParam(0, STR_AI_CONFIG_CHANGE_GAMESCRIPT), dim = GetStringBoundingBox(STR_AI_CONFIG_CHANGE);
+				SetDParam(0, STR_AI_CONFIG_CHANGE_NONE),       dim = maxdim(dim, GetStringBoundingBox(STR_AI_CONFIG_CHANGE));
+				SetDParam(0, STR_AI_CONFIG_CHANGE_AI),         dim = maxdim(dim, GetStringBoundingBox(STR_AI_CONFIG_CHANGE));
+				dim.width += padding.width;
+				dim.height += padding.height;
+				*size = maxdim(*size, dim);
+				break;
+			}
 		}
 	}
 
