Index: src/toolbar_gui.cpp
===================================================================
--- src/toolbar_gui.cpp	(revision 13610)
+++ src/toolbar_gui.cpp	(working copy)
@@ -952,6 +952,7 @@
 			case WKC_CTRL  | 'S': MenuClickSmallScreenshot(); break;
 			case WKC_CTRL  | 'G': MenuClickWorldScreenshot(); break;
 			case WKC_CTRL | WKC_ALT | 'C': if (!_networking) ShowCheatWindow(); break;
+			case WKC_SHIFT | WKC_BACKSPACE: DeleteAllNonVitalWindows(); break;
 			case 'A': if (CanBuildVehicleInfrastructure(VEH_TRAIN)) ShowBuildRailToolbar(_last_built_railtype, 4); break; // Invoke Autorail
 			case 'L': ShowTerraformToolbar(); break;
 			case 'M': ShowSmallMap(); break;
