diff --git a/src/settings_gui.cpp b/src/settings_gui.cpp
index ca25ad2..f68e60f 100644
--- a/src/settings_gui.cpp
+++ b/src/settings_gui.cpp
@@ -1979,7 +1979,7 @@ struct GameSettingsWindow : QueryStringBaseWindow {
 
 	Scrollbar *vscroll;
 
-	GameSettingsWindow(const WindowDesc *desc) : QueryStringBaseWindow(50), cur_restriction_mode(RM_BASIC)
+	GameSettingsWindow(const WindowDesc *desc) : QueryStringBaseWindow(50)
 	{
 		static bool first_time = true;
 
@@ -1999,6 +1999,11 @@ struct GameSettingsWindow : QueryStringBaseWindow {
 		this->valuedropdown_entry = NULL;
 		this->closing_dropdown = false;
 		this->manually_changed_folding = false;
+		this->cur_restriction_mode = (RestrictionMode) _settings_client.gui.game_settings_window_restrict_ddb;
+		/* ensure consistency with the settings.ini definition */
+		uint restrict_ddb_descridx = 0;
+		const SettingDesc *restrict_ddb_descr = GetSettingFromName("gui.game_settings_window_restrict_ddb", &restrict_ddb_descridx);
+		assert ((uint)RM_END == restrict_ddb_descr->desc.max+1);
 
 		this->CreateNestedTree(desc);
 		this->vscroll = this->GetScrollbar(WID_GS_SCROLLBAR);
@@ -2010,6 +2015,11 @@ struct GameSettingsWindow : QueryStringBaseWindow {
 		this->InvalidateData();
 	}
 
+	~GameSettingsWindow()
+	{
+		_settings_client.gui.game_settings_window_restrict_ddb = (uint8) this->cur_restriction_mode;
+	}
+
 	virtual void UpdateWidgetSize(int widget, Dimension *size, const Dimension &padding, Dimension *fill, Dimension *resize)
 	{
 		switch (widget) {
diff --git a/src/settings_type.h b/src/settings_type.h
index c29f512..672c1dc 100644
--- a/src/settings_type.h
+++ b/src/settings_type.h
@@ -120,6 +120,7 @@ struct GUISettings {
 
 	uint16 console_backlog_timeout;          ///< the minimum amount of time items should be in the console backlog before they will be removed in ~3 seconds granularity.
 	uint16 console_backlog_length;           ///< the minimum amount of items in the console backlog before items will be removed.
+	uint8  game_settings_window_restrict_ddb;///< last state of the restrict drop down box of GameSettingsWindow
 #ifdef ENABLE_NETWORK
 	uint16 network_chat_box_width;           ///< width of the chat box in pixels
 	uint8  network_chat_box_height;          ///< height of the chat box in lines
diff --git a/src/table/settings.ini b/src/table/settings.ini
index 2024a36..a7de6b9 100644
--- a/src/table/settings.ini
+++ b/src/table/settings.ini
@@ -2949,6 +2949,15 @@ max      = 65500
 cat      = SC_ADVANCED
 
 [SDTC_VAR]
+var      = gui.game_settings_window_restrict_ddb
+type     = SLE_UINT8
+flags    = SLF_NOT_IN_SAVE | SLF_NO_NETWORK_SYNC
+def      = 0
+min      = 0
+max      = 5
+cat      = SC_EXPERT
+
+[SDTC_VAR]
 var      = music.playlist
 type     = SLE_UINT8
 flags    = SLF_NOT_IN_SAVE | SLF_NO_NETWORK_SYNC
