Index: src/network/network_server.cpp
===================================================================
--- src/network/network_server.cpp	(revision 26607)
+++ src/network/network_server.cpp	(working copy)
@@ -2193,11 +2193,15 @@
 		ci->client_playas = c->index;
 		NetworkUpdateClientInfo(ci->client_id);
 		NetworkSendCommand(0, 0, 0, CMD_RENAME_PRESIDENT, NULL, ci->client_name, c->index);
-		NetworkServerSendChat(NETWORK_ACTION_COMPANY_NEW, DESTTYPE_BROADCAST, 0, "", ci->client_id, c->index + 1);
 	}
 
 	/* Announce new company on network. */
 	NetworkAdminCompanyInfo(c, true);
+
+	if (ci != NULL) {
+		/* ci is NULL when replaying, or for AIs. In neither case there is a client. */
+		NetworkServerSendChat(NETWORK_ACTION_COMPANY_NEW, DESTTYPE_BROADCAST, 0, "", ci->client_id, c->index + 1);
+	}
 }
 
 #endif /* ENABLE_NETWORK */
