Index: src/settings.cpp
===================================================================
--- src/settings.cpp	(revision 14286)
+++ src/settings.cpp	(working copy)
@@ -1130,8 +1130,8 @@
 
 static const SettingDesc _music_settings[] = {
 	 SDT_VAR(MusicFileSettings, playlist,   SLE_UINT8, S, 0,   0, 0,   5, 1,  STR_NULL, NULL),
-	 SDT_VAR(MusicFileSettings, music_vol,  SLE_UINT8, S, 0, 128, 0, 100, 1,  STR_NULL, NULL),
-	 SDT_VAR(MusicFileSettings, effect_vol, SLE_UINT8, S, 0, 128, 0, 100, 1,  STR_NULL, NULL),
+	 SDT_VAR(MusicFileSettings, music_vol,  SLE_UINT8, S, 0, 128, 0, 128, 1,  STR_NULL, NULL),
+	 SDT_VAR(MusicFileSettings, effect_vol, SLE_UINT8, S, 0, 128, 0, 128, 1,  STR_NULL, NULL),
 	SDT_LIST(MusicFileSettings, custom_1,   SLE_UINT8, S, 0, NULL,            STR_NULL, NULL),
 	SDT_LIST(MusicFileSettings, custom_2,   SLE_UINT8, S, 0, NULL,            STR_NULL, NULL),
 	SDT_BOOL(MusicFileSettings, playing,               S, 0, true,            STR_NULL, NULL),
Index: src/music_gui.cpp
===================================================================
--- src/music_gui.cpp	(revision 14286)
+++ src/music_gui.cpp	(working copy)
@@ -486,7 +486,7 @@
 					x -= 106;
 				}
 
-				new_vol = min(max(x - 21, 0) * 2, 127);
+				new_vol = min(max(x - 21, 0) * 2, 128);
 				if (new_vol != *vol) {
 					*vol = new_vol;
 					if (vol == &msf.music_vol) MusicVolumeChanged(new_vol);
