diff --git a/src/network/network.cpp b/src/network/network.cpp
index 0d1e84d..7413276 100644
--- a/src/network/network.cpp
+++ b/src/network/network.cpp
@@ -1076,6 +1076,8 @@ void NetworkGameLoop()
 
 	if (!NetworkReceive()) return;
 
+	if (!_networking) return; /* may get turned off by NetworkReceive() */
+
 	if (_network_server) {
 		/* Log the sync state to check for in-syncedness of replays. */
 		if (_date_fract == 0) {
