Index: widget.c
===================================================================
--- widget.c	(revision 3270)
+++ widget.c	(working copy)
@@ -191,7 +191,7 @@
 			goto draw_default;
 		}
 
-		case WWT_CLOSEBOX: /* WWT_TEXTBTN */
+		case WWT_TEXTBTN: /* WWT_TEXTBTN */
 		case WWT_4: {
 			DrawFrameRect(r.left, r.top, r.right, r.bottom, wi->color, (clicked) ? FR_LOWERED : 0);
 			}
@@ -409,7 +409,15 @@
 			DrawSprite(SPR_WINDOW_RESIZE, r.left + 3 + clicked, r.top + 3 + clicked);
 			break;
 		}
+		
+		case WWT_CLOSEBOX: {
+			assert(r.right - r.left == 10); // ensure the same sizes are used everywhere
 
+			DrawFrameRect(r.left, r.top, r.right, r.bottom, wi->color, 0);
+			DrawString(r.left + 2, r.top + 2, STR_00C5, 0);
+			break;
+		}
+
 		case WWT_CAPTION: {
 			assert(r.bottom - r.top == 13); // XXX - to ensure the same sizes are used everywhere!
 			DrawFrameRect(r.left, r.top, r.right, r.bottom, wi->color, FR_BORDERONLY);
Index: window.h
===================================================================
--- window.h	(revision 3270)
+++ window.h	(working copy)
@@ -509,7 +509,6 @@
 	WWT_PANEL_2 = 2,					/* button with diff image when clicked */
 
 	WWT_TEXTBTN = 3,					/* button with text */
-	WWT_CLOSEBOX = WWT_TEXTBTN,
 	WWT_4 = 4,								/* button with diff text when clicked */
 	WWT_5 = 5,								/* label */
 	WWT_6 = 6,								/* combo box text area */
@@ -522,7 +521,8 @@
 	WWT_STICKYBOX = 12,
 	WWT_SCROLL2BAR = 13,				/* 2nd vertical scrollbar*/
 	WWT_RESIZEBOX = 14,
-	WWT_LAST = 15,						/* Last Item. use WIDGETS_END to fill up padding!! */
+	WWT_CLOSEBOX = 15,
+	WWT_LAST = 16,						/* Last Item. use WIDGETS_END to fill up padding!! */
 
 	WWT_MASK = 31,
 
