--- ../bin/ottd/trunk.hg/os/rpm/openttd.spec	2009-08-21 20:04:01.000000000 +0200
+++ openttd-trunk.spec	2009-09-10 21:24:42.000000000 +0200
@@ -24,10 +24,11 @@
 Source:        %{name}-%{version}-source.tar.bz2
 
 Requires:      fontconfig
-Requires:      libicu
-Requires:      libpng
+#Requires:      libicu
+#Requires:      libpng
 Requires:      SDL
 Requires:      zlib
+BuildRequires: gcc-c++
 BuildRequires: fontconfig-devel
 BuildRequires: libpng-devel
 BuildRequires: libicu-devel
@@ -50,6 +51,7 @@
 %if %{_vendor}=="suse"
 Requires:      freetype2
 BuildRequires: freetype2-devel
+BuildRequires: update-desktop-files
 %endif
 
 BuildRoot:     %{_tmppath}/%{name}-%{version}-%{release}-buildroot
@@ -99,7 +101,7 @@
 make %{?_smp_mflags}
 
 %install
-rm -rf "%{buildroot}"
+#rm -rf "%{buildroot}"
 make install INSTALL_DIR="%{buildroot}"
 
 # Validate menu entrys (vendor specific)
@@ -111,6 +113,21 @@
 	"%{buildroot}/%{_datadir}/applications/%{name}.desktop" \
 #	--delete-original
 %endif
+%if %{_vendor}=="suse"
+%__cat > %{name}.desktop << EOF
+[Desktop Entry]
+Encoding=UTF-8
+Name=OpenTTD
+Comment=OpenTTD - A clone of the Microprose game 'Transport Tycoon Deluxe'
+GenericName=OpenTTD
+Type=Application
+Terminal=false
+Exec=%{name}
+Icon=%{name}
+Categories=Game;StrategyGame;
+EOF
+%suse_update_desktop_file -i %{name} Game StrategyGame
+%endif
 
 %clean
 rm -rf "%{buildroot}"
@@ -148,6 +165,7 @@
 %dir %{_datadir}/games/%{name}/lang
 %dir %{_datadir}/games/%{name}/data
 %dir %{_datadir}/games/%{name}/gm
+%dir %{_datadir}/games/%{name}/scripts
 %attr(755, root, games) %{_bindir}/%{name}
 %{_datadir}/doc/%{name}/*
 %{_datadir}/games/%{name}/lang/*
@@ -155,10 +173,17 @@
 %{_datadir}/games/%{name}/scripts/*
 %{_datadir}/applications/*%{name}.desktop
 %{_datadir}/pixmaps/*
-%{_datadir}/icons/hicolor/*/apps/%{name}.png
+%{_datadir}/icons/*
+%{_datadir}/icons/*/*
+%{_datadir}/icons/*/*/apps
+%{_datadir}/icons/*/*/apps/%{name}*
 %doc %{_mandir}/man6/%{name}.6.gz
 
 %changelog
+* Thu Sep 10 2009 Marcel Gmür <ammler@openttdcoop.org> - 0.7.2
+
+- Add openSUSE support
+
 * Mon Oct 20 2008 Benedikt Brüggemeier <skidd13@openttd.org>
 
 - Added libicu dependency
