Index: src/newgrf_canal.cpp
===================================================================
--- src/newgrf_canal.cpp	(revision 22643)
+++ src/newgrf_canal.cpp	(working copy)
@@ -50,7 +50,7 @@
 		case 0x80: {
 			uint z = GetTileZ(tile) / TILE_HEIGHT;
 			/* Return consistent height within locks */
-			if (IsTileType(tile, MP_WATER) && IsLock(tile) && GetSection(tile) == LOCK_UPPER) z--;
+			if (IsLockTile(tile) && GetLockPart(tile) == LOCK_UPPER) z--;
 			return z;
 		}

