Index: config.lib
===================================================================
--- config.lib	(revision 19217)
+++ config.lib	(working copy)
@@ -2303,6 +2303,12 @@
 		eval "res=\$$2"
 		if [ -z "$res" ]; then
 			log 2 "  trying /usr/local/include/$4$5... no"
+			#macports default directory
+			eval "$2=`ls -1 /opt/local/include/$4*.h 2>/dev/null | egrep \"\/$5\$\"`"
+		fi
+		eval "res=\$$2"
+		if [ -z "$res" ]; then
+			log 2 "  trying /opt/local/include/$4$5... no"
 		fi
 
 		eval "res=\$$2"
