Index: src/ai/api/ai_bridge.cpp
===================================================================
--- src/ai/api/ai_bridge.cpp	(revision 21000)
+++ src/ai/api/ai_bridge.cpp	(working copy)
@@ -152,7 +152,7 @@
 	if (!IsValidBridge(bridge_id)) return -1;
 
 	uint max = ::GetBridgeSpec(bridge_id)->max_length;
-	if (max >= 16 && _settings_game.construction.longbridges) max = 100;
+	if (max >= MAX_BRIDGE_LENGTH && _settings_game.construction.longbridges) max = MAX_BRIDGE_LENGTH_LONGBRIDGES;
 	return max + 2;
 }
 
Index: src/bridge.h
===================================================================
--- src/bridge.h	(revision 21000)
+++ src/bridge.h	(working copy)
@@ -72,4 +72,7 @@
 
 void ResetBridges();
 
+static const uint MAX_BRIDGE_LENGTH             =  16; ///<Maximum length of the bridge
+static const uint MAX_BRIDGE_LENGTH_LONGBRIDGES = 100; ///<Maximum length of the bridge with enabled longbridges
+
 #endif /* BRIDGE_H */
Index: src/tunnelbridge_cmd.cpp
===================================================================
--- src/tunnelbridge_cmd.cpp	(revision 21000)
+++ src/tunnelbridge_cmd.cpp	(working copy)
@@ -173,7 +173,7 @@
 CommandCost CheckBridgeAvailability(BridgeType bridge_type, uint bridge_len, DoCommandFlag flags)
 {
 	if (flags & DC_QUERY_COST) {
-		if (bridge_len <= (_settings_game.construction.longbridges ? 100U : 16U)) return CommandCost();
+		if (bridge_len <= (_settings_game.construction.longbridges ? MAX_BRIDGE_LENGTH_LONGBRIDGES : MAX_BRIDGE_LENGTH)) return CommandCost();
 		return_cmd_error(STR_ERROR_BRIDGE_TOO_LONG);
 	}
 
@@ -183,7 +183,7 @@
 	if (b->avail_year > _cur_year) return CMD_ERROR;
 
 	uint max = b->max_length;
-	if (max >= 16 && _settings_game.construction.longbridges) max = 100;
+	if (max >= MAX_BRIDGE_LENGTH && _settings_game.construction.longbridges) max = MAX_BRIDGE_LENGTH_LONGBRIDGES;
 
 	if (b->min_length > bridge_len) return CMD_ERROR;
 	if (bridge_len <= max) return CommandCost();
@@ -257,7 +257,7 @@
 		CommandCost ret = CheckBridgeAvailability(bridge_type, bridge_len, flags);
 		if (ret.Failed()) return ret;
 	} else {
-		if (bridge_len > (_settings_game.construction.longbridges ? 100U : 16U)) return_cmd_error(STR_ERROR_BRIDGE_TOO_LONG);
+		if (bridge_len > (_settings_game.construction.longbridges ? MAX_BRIDGE_LENGTH_LONGBRIDGES : MAX_BRIDGE_LENGTH)) return_cmd_error(STR_ERROR_BRIDGE_TOO_LONG);
 	}
 
 	uint z_start;
