Index: src/station_cmd.cpp
===================================================================
--- src/station_cmd.cpp	(Revision 24187)
+++ src/station_cmd.cpp	(Arbeitskopie)
@@ -3207,8 +3207,27 @@
 			}
 
 			if (!skip) {
-				int b = ge->last_speed - 85;
-				if (b >= 0) rating += b >> 2;
+				int b = ge->last_speed;
+ 
+				if ((st->last_vehicle_type == VEH_TRAIN) || (st->last_vehicle_type == VEH_AIRCRAFT)) {
+					b -= 85;
+					if (b >= 0) rating += b>>2;
+				}
+				else	{
+					if (st->last_vehicle_type == VEH_ROAD)	{
+						b -= 60;
+						if (b >= 0) rating += b>>1;
+					}
+					else	{
+						//ships LSB is 0.5km/h not 1km/h
+						if (st->last_vehicle_type == VEH_SHIP)	{
+							b -= 40;
+							if (b >= 0) rating += b;
+						}
+					}
+					//looks that rating <= 42, cause trains/plains has max 42
+					if (rating > 42 ) rating = 42;
+				}
 
 				byte days = ge->days_since_pickup;
 				if (st->last_vehicle_type == VEH_SHIP) days >>= 2;
