Index: station_cmd.c
===================================================================
--- station_cmd.c	(Revision 4244)
+++ station_cmd.c	(Arbeitskopie)
@@ -1245,7 +1245,7 @@
 	return cost;
 }
 
-int32 DoConvertStationRail(TileIndex tile, uint totype, bool exec)
+int32 DoConvertStationRail(TileIndex tile, RailType totype, bool exec)
 {
 	const Station* st = GetStationByTile(tile);
 
Index: tunnelbridge_cmd.c
===================================================================
--- tunnelbridge_cmd.c	(Revision 4244)
+++ tunnelbridge_cmd.c	(Arbeitskopie)
@@ -716,7 +716,7 @@
 	return CMD_ERROR;
 }
 
-int32 DoConvertTunnelBridgeRail(TileIndex tile, uint totype, bool exec)
+int32 DoConvertTunnelBridgeRail(TileIndex tile, RailType totype, bool exec)
 {
 	TileIndex endtile;
 	uint length;
Index: road_cmd.c
===================================================================
--- road_cmd.c	(Revision 4244)
+++ road_cmd.c	(Arbeitskopie)
@@ -419,7 +419,7 @@
 	return cost;
 }
 
-int32 DoConvertStreetRail(TileIndex tile, uint totype, bool exec)
+int32 DoConvertStreetRail(TileIndex tile, RailType totype, bool exec)
 {
 	// not a railroad crossing?
 	if (!IsLevelCrossing(tile)) return CMD_ERROR;
Index: rail_cmd.c
===================================================================
--- rail_cmd.c	(Revision 4244)
+++ rail_cmd.c	(Arbeitskopie)
@@ -916,9 +916,9 @@
 	return _price.build_rail / 2;
 }
 
-extern int32 DoConvertStationRail(TileIndex tile, uint totype, bool exec);
-extern int32 DoConvertStreetRail(TileIndex tile, uint totype, bool exec);
-extern int32 DoConvertTunnelBridgeRail(TileIndex tile, uint totype, bool exec);
+extern int32 DoConvertStationRail(TileIndex tile, RailType totype, bool exec);
+extern int32 DoConvertStreetRail(TileIndex tile, RailType totype, bool exec);
+extern int32 DoConvertTunnelBridgeRail(TileIndex tile, RailType totype, bool exec);
 
 /** Convert one rail type to the other. You can convert normal rail to
  * monorail/maglev easily or vice-versa.
