diff --git a/config.lib b/config.lib
index a92990e..145d6c6 100644
--- a/config.lib
+++ b/config.lib
@@ -1558,7 +1558,7 @@ make_cflags_and_ldflags() {
 			LDFLAGS="$OSX_LD_SYSROOT $LDFLAGS"
 		fi
 
-		if [ "$enable_universal" = "0" ] && [ $cc_version -ge 40 ]; then
+		if [ "$enable_universal" = "0" ] && [ $cc_version -gt 40 ]; then
 			# Only set the min version when not doing an universal build.
 			# Universal builds set the version elsewhere.
 			if [ "$cpu_type" = "64" ]; then
