diff --git a/config.lib b/config.lib
index c41d505..95df961 100644
--- a/config.lib
+++ b/config.lib
@@ -1424,12 +1424,16 @@ make_cflags_and_ldflags() {
 	# Special CXXFlags for HOST
 	CXXFLAGS="$CXXFLAGS"
 	# Libs to compile. In fact this is just LDFLAGS
-	LIBS="-lstdc++"
+	LIBS=""
 	# LDFLAGS used for HOST
 	LDFLAGS="$LDFLAGS"
 	# FEATURES for HOST (lto)
 	FEATURES=""
 
+	if [ "$os" != "OSX" ]; then
+		LIBS="$LIBS -lstdc++"
+	fi
+
 	make_compiler_cflags "$cc_build" "CFLAGS_BUILD" "CXXFLAGS_BUILD" "LDFLAGS_BUILD" "FEATURES_BUILD"
 	make_compiler_cflags "$cc_host" "CFLAGS" "CXXFLAGS" "LDFLAGS" "FEATURES"
 
