Index: src/saveload/saveload.cpp
===================================================================
--- src/saveload/saveload.cpp	(revision 27967)
+++ src/saveload/saveload.cpp	(working copy)
@@ -2469,7 +2469,7 @@
 {
 	_sl.ff_state = _fast_forward;
 	_fast_forward = 0;
-	SetMouseCursorBusy(true);
+	if (!_do_autosave) SetMouseCursorBusy(true);
 
 	InvalidateWindowData(WC_STATUS_BAR, 0, SBI_SAVELOAD_START);
 	_sl.saveinprogress = true;
