Index: src/elrail.cpp
===================================================================
--- src/elrail.cpp	(revision 10266)
+++ src/elrail.cpp	(working copy)
@@ -395,6 +395,8 @@
 
 void DrawCatenary(const TileInfo *ti)
 {
+	if (_patches.disable_elrails) return;
+
 	if (MayHaveBridgeAbove(ti->tile) && IsBridgeAbove(ti->tile)) {
 		TileIndex head = GetNorthernBridgeEnd(ti->tile);
 
@@ -402,7 +404,6 @@
 			DrawCatenaryOnBridge(ti);
 		}
 	}
-	if (_patches.disable_elrails) return;
 
 	switch (GetTileType(ti->tile)) {
 		case MP_RAILWAY:
