Index: src/window.cpp
===================================================================
--- src/window.cpp	(revision 19658)
+++ src/window.cpp	(working copy)
@@ -2086,6 +2086,7 @@
 			case MC_DOUBLE_LEFT:
 			case MC_LEFT:
 				DEBUG(misc, 2, "Cursor: 0x%X (%d)", _cursor.sprite, _cursor.sprite);
+#if 0
 				if (_thd.place_mode != HT_NONE &&
 						/* query button and place sign button work in pause mode */
 						_cursor.sprite != SPR_CURSOR_QUERY &&
@@ -2094,7 +2095,7 @@
 						!_cheats.build_in_pause.value) {
 					return;
 				}
-
+#endif
 				if (_thd.place_mode == HT_NONE) {
 					if (!HandleViewportClicked(vp, x, y) &&
 							!(w->flags4 & WF_DISABLE_VP_SCROLL) &&
