Index: src/news_gui.cpp
===================================================================
--- src/news_gui.cpp	(revision 21229)
+++ src/news_gui.cpp	(working copy)
@@ -342,7 +342,9 @@
 
 	virtual Point OnInitialPosition(const WindowDesc *desc, int16 sm_width, int16 sm_height, int window_number)
 	{
-		Point pt = { (_screen.width - max(sm_width, desc->default_width)) / 2, _screen.height };
+		const Window *w = FindWindowById(WC_STATUS_BAR, 0);
+		assert(w != NULL);
+		Point pt = { w->left + w->width / 2 - max(sm_width, desc->default_width) / 2, _screen.height };
 		return pt;
 	}
 
