Index: src/dock_gui.cpp
===================================================================
--- src/dock_gui.cpp	(revision 21000)
+++ src/dock_gui.cpp	(working copy)
@@ -28,6 +28,7 @@
 #include "company_base.h"
 #include "station_type.h"
 #include "hotkeys.h"
+#include "tunnelbridge.h"
 
 #include "table/sprites.h"
 #include "table/strings.h"
@@ -92,7 +93,7 @@
 
 static void PlaceDocks_Aqueduct(TileIndex tile)
 {
-	VpStartPlaceSizing(tile, VPM_X_OR_Y, DDSP_BUILD_BRIDGE);
+	DoCommandP(tile, _build_tunnel_endtile, TRANSPORT_WATER << 15, CMD_BUILD_BRIDGE | CMD_MSG(STR_ERROR_CAN_T_BUILD_AQUEDUCT_HERE), CcBuildBridge);
 }
 
 
@@ -184,7 +185,7 @@
   */
 static void BuildDocksClick_Aqueduct(Window *w)
 {
-	HandlePlacePushButton(w, DTW_BUILD_AQUEDUCT, SPR_CURSOR_AQUEDUCT, HT_RECT, PlaceDocks_Aqueduct);
+	HandlePlacePushButton(w, DTW_BUILD_AQUEDUCT, SPR_CURSOR_AQUEDUCT, HT_SPECIAL, PlaceDocks_Aqueduct);
 }
 
 
@@ -289,12 +290,26 @@
 
 	virtual void OnPlacePresize(Point pt, TileIndex tile_from)
 	{
-		DiagDirection dir = GetInclinedSlopeDirection(GetTileSlope(tile_from, NULL));
+		uint z;
+		DiagDirection dir = GetInclinedSlopeDirection(GetTileSlope(tile_from, &z));
 		TileIndex tile_to = tile_from;
 
 		if (IsValidDiagDirection(dir)) {
-			tile_to = TileAddByDiagDir(tile_from, ReverseDiagDir(dir));
-			tile_from = this->last_clicked_widget == DTW_LOCK ? TileAddByDiagDir(tile_from, dir) : tile_from;
+			if (this->last_clicked_widget == DTW_BUILD_AQUEDUCT) {
+				TileIndexDiff step = TileOffsByDiagDir(ReverseDiagDir(dir));
+				uint map_edge = DistanceFromEdgeDir(tile_to, ReverseDiagDir(dir));
+
+				for (uint length = 0; length <= map_edge; length++) {
+					tile_to = TILE_ADD(tile_to, step);
+					/* if this tile max height is higher than our aqueduct height, return this tile */
+					if (GetTileMaxZ(tile_to) > z) break;
+				}
+				_build_tunnel_endtile = tile_to;
+
+			} else {
+				tile_to = TileAddByDiagDir(tile_from, ReverseDiagDir(dir));
+				tile_from = this->last_clicked_widget == DTW_LOCK ? TileAddByDiagDir(tile_from, dir) : tile_from;
+			}
 		}
 
 		VpSetPresizeRange(tile_from, tile_to);
