Index: src/viewport.cpp
===================================================================
--- src/viewport.cpp	(wersja 23123)
+++ src/viewport.cpp	(kopia robocza)
@@ -1887,8 +1887,8 @@
 	if (pt.x == -1) return;
 
 	if ((_thd.place_mode & HT_DRAG_MASK) == HT_POINT) {
-		pt.x += 8;
-		pt.y += 8;
+		pt.x += TILE_SIZE / 2;
+		pt.y += TILE_SIZE / 2;
 	}
 
 	_tile_fract_coords.x = pt.x & TILE_UNIT_MASK;
