Index: currency.c
===================================================================
--- currency.c	(revision 4472)
+++ currency.c	(working copy)
@@ -34,6 +34,7 @@
 	{    6, ' ', CF_NOEURO, "",     " zl"  }, // polish zloty
 	{    6, '.', CF_NOEURO, "",     " Lei" }, // romanian Lei
 	{    5, ' ', CF_NOEURO, "",     " p"   }, // russian rouble
+	{   350, '.', CF_NOEURO, "",     " SIT"  }, // slovenian tolar
 	{   13, '.', CF_NOEURO, "",     " Kr"  }, // swedish krona
 	{    1, ' ', CF_NOEURO, "",     ""     }, // custom currency
 };
@@ -61,6 +62,7 @@
 	STR_CURR_PLN,
 	STR_CURR_ROL,
 	STR_CURR_RUR,
+	STR_CURR_SIT,
 	STR_CURR_SEK,
 	STR_CURR_CUSTOM,
 	INVALID_STRING_ID
@@ -84,7 +86,7 @@
 		if (to_euro == CF_ISEURO && _cur_year < 2000 - MAX_YEAR_BEGIN_REAL) continue;
 		mask |= (1 << i);
 	}
-	mask |= (1 << 23); // always allow custom currency
+	mask |= (1 << 24); // always allow custom currency
 	return mask;
 }
 
Index: currency.h
===================================================================
--- currency.h	(revision 4472)
+++ currency.h	(working copy)
@@ -20,7 +20,7 @@
 extern const StringID _currency_string_list[];
 
 // XXX small hack, but makes the rest of the code a bit nicer to read
-#define _custom_currency (_currency_specs[23])
+#define _custom_currency (_currency_specs[24])
 #define _currency ((const CurrencySpec*)&_currency_specs[_opt_ptr->currency])
 
 uint GetMaskOfAllowedCurrencies(void);
Index: lang/english.txt
===================================================================
--- lang/english.txt	(revision 4472)
+++ lang/english.txt	(working copy)
@@ -894,6 +894,7 @@
 STR_CURR_PLN                                                    :Polish Zloty (PLN)
 STR_CURR_ROL                                                    :Romanian Leu (ROL)
 STR_CURR_RUR                                                    :Russian Rubles (RUR)
+STR_CURR_SIT                                                    :Slovenian Tolar (SIT)
 STR_CURR_SEK                                                    :Swedish Krona (SEK)
 
 STR_CURR_CUSTOM                                                 :Custom...
Index: settings_gui.c
===================================================================
--- settings_gui.c	(revision 4472)
+++ settings_gui.c	(working copy)
@@ -171,7 +171,7 @@
 			}
 			break;
 		case 5: /* Currency */
-			if (e->dropdown.index == 23)
+			if (e->dropdown.index == 24)
 				ShowCustCurrency();
 			_opt_ptr->currency = e->dropdown.index;
 			MarkWholeScreenDirty();
