Index: station_gui.c
===================================================================
--- station_gui.c	(revision 7201)
+++ station_gui.c	(working copy)
@@ -376,6 +376,7 @@
 			for (i = 0; i < NUM_CARGO; i++) {
 				LowerWindowWidget(w, i + 12);
 			}
+			SetWindowWidgetLoweredState(w, 24, false); // Toggle 'NO' to false, because ALL shouldn't include it
 			LowerWindowWidget(w, 27);
 
 			cargo_filter = 0x1FFF; /* select everything */
@@ -393,7 +394,7 @@
 		break;
 		default:
 			if (e->we.click.widget >= 12 && e->we.click.widget <= 24) { //change cargo_filter
-				if (_ctrl_pressed) {
+				if (_ctrl_pressed && e->we.click.widget != 24) { // NO and other options can't be selected at the same time
 					TOGGLEBIT(cargo_filter, e->we.click.widget - 12);
 					ToggleWidgetLoweredState(w, e->we.click.widget);
 				} else {
