Index: config.lib
===================================================================
--- config.lib	(revision 12309)
+++ config.lib	(working copy)
@@ -1146,10 +1146,10 @@
 	fi
 
 	if [ -n "$with_midi" ]; then
-		CFLAGS="$CFLAGS -DEXTERNAL_PLAYER=\\\\\"$with_midi\\\\\""
+		CFLAGS="$CFLAGS -DEXTERNAL_PLAYER=\"\\\\\"$with_midi\\\\\"\""
 	fi
 	if [ -n "$with_midi_arg" ]; then
-		CFLAGS="$CFLAGS -DMIDI_ARG=\\\\\"$with_midi_arg\\\\\""
+		CFLAGS="$CFLAGS -DMIDI_ARG=\"\\\\\"$with_midi_arg\\\\\"\""
 	fi
 
 	if [ "$enable_dedicated" != "0" ]; then
@@ -1200,14 +1200,14 @@
 	fi
 
 	if [ -n "$personal_dir" ]; then
-		CFLAGS="$CFLAGS -DWITH_PERSONAL_DIR -DPERSONAL_DIR=\\\\\"$personal_dir\\\\\""
+		CFLAGS="$CFLAGS -DWITH_PERSONAL_DIR -DPERSONAL_DIR=\"\\\\\"$personal_dir\\\\\"\""
 	fi
 
 	if [ -n "$shared_dir" ]; then
-		CFLAGS="$CFLAGS -DWITH_SHARED_DIR -DSHARED_DIR=\\\\\"$shared_dir\\\\\""
+		CFLAGS="$CFLAGS -DWITH_SHARED_DIR -DSHARED_DIR=\"\\\\\"$shared_dir\\\\\"\""
 	fi
 
-	CFLAGS="$CFLAGS -DGLOBAL_DATA_DIR=\\\\\"$prefix_dir/$data_dir\\\\\""
+	CFLAGS="$CFLAGS -DGLOBAL_DATA_DIR=\"\\\\\"$prefix_dir/$data_dir\\\\\"\""
 
 	log 1 "using CFLAGS... $CFLAGS $CC_CFLAGS"
 	log 1 "using LDFLAGS... $LIBS $LDFLAGS"
