Index: src/window.cpp
===================================================================
--- src/window.cpp	(revision 21434)
+++ src/window.cpp	(working copy)
@@ -2213,36 +2213,6 @@
 			case MC_DOUBLE_LEFT:
 			case MC_LEFT:
 				DEBUG(misc, 2, "Cursor: 0x%X (%d)", _cursor.sprite, _cursor.sprite);
-				if (_thd.place_mode != HT_NONE && _pause_mode != PM_UNPAUSED && _game_mode != GM_EDITOR) {
-					switch (_settings_game.construction.command_pause_level) {
-						case CMDPL_ALL_ACTIONS:
-							/* We allow all actions. */
-							break;
-
-						case CMDPL_NO_LANDSCAPING:
-							if (_cursor.sprite == SPR_CURSOR_CLONE_TRAIN ||
-									_cursor.sprite == SPR_CURSOR_CLONE_ROADVEH ||
-									_cursor.sprite == SPR_CURSOR_CLONE_SHIP ||
-									_cursor.sprite == SPR_CURSOR_CLONE_AIRPLANE) {
-								/* Cloning is allowed. */
-								break;
-							}
-							/* FALL THROUGH */
-						case CMDPL_NO_CONSTRUCTION:
-							if (_cursor.sprite == SPR_CURSOR_SIGN ||
-									(_cursor.sprite >= SPR_CURSOR_PICKSTATION_FIRST && _cursor.sprite <= SPR_CURSOR_PICKSTATION_LAST)) {
-								/* Building signs or making orders is allowed. */
-								break;
-							}
-							/* FALL THROUGH */
-						case CMDPL_NO_ACTIONS:
-							if (_cursor.sprite == SPR_CURSOR_QUERY) break;
-
-							/* All other ones are not allowed to build. */
-							return;
-					}
-				}
-
 				if (!HandleViewportClicked(vp, x, y) &&
 						!(w->flags4 & WF_DISABLE_VP_SCROLL) &&
 						_settings_client.gui.left_mouse_btn_scrolling) {
