Index: src/road_cmd.cpp
===================================================================
--- src/road_cmd.cpp	(revision 11199)
+++ src/road_cmd.cpp	(working copy)
@@ -835,11 +835,12 @@
 	switch (GetRoadTileType(tile)) {
 		case ROAD_TILE_NORMAL: {
 			RoadBits b = GetAllRoadBits(tile);
+			RoadBits tram_bits = GetRoadBits(tile, ROADTYPE_TRAM);
 
 #define M(x) (1 << (x))
-			/* Clear the road if only one piece is on the tile OR the AI tries
+			/* Clear the road if only one (non-tram-)piece is on the tile OR the AI tries
 			 * to clear town road OR we are not using the DC_AUTO flag */
-			if ((M(b) & (M(ROAD_NW) | M(ROAD_SW) | M(ROAD_SE) | M(ROAD_NE))) ||
+			if (((M(b) & (M(ROAD_NW) | M(ROAD_SW) | M(ROAD_SE) | M(ROAD_NE))) && (tram_bits == ROAD_NONE)) ||
 			    ((flags & DC_AI_BUILDING) && IsTileOwner(tile, OWNER_TOWN)) ||
 			    !(flags & DC_AUTO)
 				) {
