Index: saveload.h
===================================================================
--- saveload.h	(revision 3639)
+++ saveload.h	(working copy)
@@ -136,7 +136,11 @@
 };
 
 typedef byte SaveLoadType;
-typedef uint16 OffSetType;
+#ifdef __arch64__
+typedef uint64 OffSetType;
+#else
+typedef uint32 OffSetType;
+#endif
 
 /** SaveLoad type struct. Do NOT use this directly but use the SLE_ macros defined just below! */
 typedef struct SaveLoad {
