Index: src/rail_cmd.cpp
===================================================================
--- src/rail_cmd.cpp	(Revision 22728)
+++ src/rail_cmd.cpp	(Arbeitskopie)
@@ -1729,7 +1729,7 @@
 static void DrawTrackFence_NW(const TileInfo *ti, SpriteID base_image)
 {
 	RailFenceOffset rfo = RFO_FLAT_X;
-	if (ti->tileh != SLOPE_FLAT) rfo = (ti->tileh & SLOPE_S) ? RFO_SLOPE_SW : RFO_SLOPE_NE;
+	if ((ti->tileh != SLOPE_FLAT) && !IsHalftileSlope(ti->tileh)) rfo = (ti->tileh & SLOPE_N) ? RFO_SLOPE_NE : RFO_SLOPE_SW;
 	AddSortableSpriteToDraw(base_image + rfo, _drawtile_track_palette,
 		ti->x, ti->y + 1, 16, 1, 4, ti->z);
 }
@@ -1737,7 +1737,7 @@
 static void DrawTrackFence_SE(const TileInfo *ti, SpriteID base_image)
 {
 	RailFenceOffset rfo = RFO_FLAT_X;
-	if (ti->tileh != SLOPE_FLAT) rfo = (ti->tileh & SLOPE_S) ? RFO_SLOPE_SW : RFO_SLOPE_NE;
+	if ((ti->tileh != SLOPE_FLAT) && !IsHalftileSlope(ti->tileh)) rfo = (ti->tileh & SLOPE_S) ? RFO_SLOPE_SW : RFO_SLOPE_NE;
 	AddSortableSpriteToDraw(base_image + rfo, _drawtile_track_palette,
 		ti->x, ti->y + TILE_SIZE - 1, 16, 1, 4, ti->z);
 }
@@ -1751,7 +1751,7 @@
 static void DrawTrackFence_NE(const TileInfo *ti, SpriteID base_image)
 {
 	RailFenceOffset rfo = RFO_FLAT_Y;
-	if (ti->tileh != SLOPE_FLAT) rfo = (ti->tileh & SLOPE_S) ? RFO_SLOPE_SE : RFO_SLOPE_NW;
+	if ((ti->tileh != SLOPE_FLAT) && !IsHalftileSlope(ti->tileh)) rfo = (ti->tileh & SLOPE_N) ? RFO_SLOPE_NW : RFO_SLOPE_SE;
 	AddSortableSpriteToDraw(base_image + rfo, _drawtile_track_palette,
 		ti->x + 1, ti->y, 1, 16, 4, ti->z);
 }
@@ -1759,7 +1759,7 @@
 static void DrawTrackFence_SW(const TileInfo *ti, SpriteID base_image)
 {
 	RailFenceOffset rfo = RFO_FLAT_Y;
-	if (ti->tileh != SLOPE_FLAT) rfo = (ti->tileh & SLOPE_S) ? RFO_SLOPE_SE : RFO_SLOPE_NW;
+	if ((ti->tileh != SLOPE_FLAT) && !IsHalftileSlope(ti->tileh)) rfo = (ti->tileh & SLOPE_S) ? RFO_SLOPE_SE : RFO_SLOPE_NW;
 	AddSortableSpriteToDraw(base_image + rfo, _drawtile_track_palette,
 		ti->x + TILE_SIZE - 1, ti->y, 1, 16, 4, ti->z);
 }
