Index: src/town_cmd.cpp
===================================================================
--- src/town_cmd.cpp	(Revision 11256)
+++ src/town_cmd.cpp	(Arbeitskopie)
@@ -730,21 +730,22 @@
 		 * maybe terraform some. */
 		desired_slope = (dir == DIAGDIR_NW || dir == DIAGDIR_SE) ? SLOPE_NE : SLOPE_NW;
 		if (desired_slope != cur_slope && ComplementSlope(desired_slope) != cur_slope) {
-			uint32 r = Random();
 
-			if (CHANCE16I(1, 8, r) && !_generating_world) {
+			if (CHANCE16(1, 8)) {
 				CommandCost res;
 
-				if (CHANCE16I(1, 16, r)) {
+				if (CHANCE16(1, 16)) {
 					res = DoCommand(tile, cur_slope, 0, DC_EXEC | DC_AUTO | DC_NO_WATER, CMD_TERRAFORM_LAND);
 				} else {
 					/* Note: Do not replace " ^ 0xF" with ComplementSlope(). The slope might be steep. */
 					res = DoCommand(tile, cur_slope ^ 0xF, 1, DC_EXEC | DC_AUTO | DC_NO_WATER, CMD_TERRAFORM_LAND);
 				}
-				if (CmdFailed(res) && CHANCE16I(1, 3, r)) {
+				if (CmdFailed(res) && CHANCE16(1, 4)) {
 					/* We can consider building on the slope, though. */
 					goto no_slope;
 				}
+			} else if (CHANCE16(1, 8)) {
+				goto no_slope;
 			}
 			return false;
 		}
@@ -902,10 +903,10 @@
  *
  * @param t The current town
  * @param tile The current tile
- * @param rcmd The RoadBits which are possible on this tile
+ * @param target_dir The current town growth dir
  * @return true if a bridge has been build else false
  */
-static bool GrowTownWithBridge(const Town *t, TileIndex tile, RoadBits rcmd)
+static bool GrowTownWithBridge(const Town *t, TileIndex tile, DiagDirection target_dir)
 {
 	DiagDirection bridge_dir; // The direction of a bridge we maybe want to build
 
@@ -920,8 +921,8 @@
 		default: return false;
 	}
 
-	/* Check if the bridge will be compatible to the RoadBits */
-	if (!(rcmd & DiagDirToRoadBits(ReverseDiagDir(bridge_dir)))) return false;
+	/* Check if the bridge will be compatible to the town growth dir */
+	if (target_dir != bridge_dir) return false;
 
 	/* We are in the right direction */
 	uint32 bridge_length = 0;     // This value stores the length of the possible bridge
@@ -1127,7 +1128,7 @@
 	rcmd = CleanUpRoadBits(tile, rcmd);
 	if (rcmd == ROAD_NONE) return;
 
-	if (GrowTownWithBridge(t1, tile, rcmd)) return;
+	if (GrowTownWithBridge(t1, tile, target_dir)) return;
 
 	GrowTownWithRoad(t1, tile, rcmd);
 }
