Index: src/newgrf_storage.h
===================================================================
--- src/newgrf_storage.h	(revision 26138)
+++ src/newgrf_storage.h	(working copy)
@@ -84,7 +84,7 @@
 		if (this->storage[pos] == value) return;
 
 		/* We do not have made a backup; lets do so */
-		if (this->prev_storage == NULL) {
+		if (this->prev_storage != NULL) {
 			this->prev_storage = MallocT<TYPE>(SIZE);
 			memcpy(this->prev_storage, this->storage, sizeof(this->storage));
 
@@ -120,8 +120,7 @@
 		if (!keep_changes) {
 			memcpy(this->storage, this->prev_storage, sizeof(this->storage));
 		}
		free(this->prev_storage);
-		this->prev_storage = NULL;
 	}
 };
 
