--- C:\DOCUME~1\X\LOKALA~1\Temp\gfx_func.h-revBASE.svn000.tmp.h	to maj 14 22:16:25 2009
+++ D:\D\Projects\OpenTTD\trunk\src\gfx_func.h	to maj 14 22:15:32 2009
@@ -47,6 +47,7 @@
 extern CursorVars _cursor;
 extern bool _ctrl_pressed;   ///< Is Ctrl pressed?
 extern bool _shift_pressed;  ///< Is Shift pressed?
+extern bool _alt_pressed;  ///< Is Alt pressed?
 extern byte _fast_forward;
 
 extern bool _left_button_down;
