Index: config.lib
===================================================================
--- config.lib	(wersja 27523)
+++ config.lib	(kopia robocza)
@@ -1432,21 +1432,27 @@
 
 make_cflags_and_ldflags() {
 	# General CFlags for BUILD
-	CFLAGS_BUILD="$CFLAGS_BUILD"
+	CFLAGS_BUILD_ORIG="$CFLAGS_BUILD"
+	CFLAGS_BUILD=""
 	# Special CXXFlags for BUILD
-	CXXFLAGS_BUILD="$CXXFLAGS_BUILD"
+	CXXFLAGS_BUILD_ORIG="$CXXFLAGS_BUILD"
+	CXXFLAGS_BUILD=""
 	# LDFLAGS for BUILD
-	LDFLAGS_BUILD="$LDFLAGS_BUILD"
+	LDFLAGS_BUILD_ORIG="$LDFLAGS_BUILD"
+	LDFLAGS_BUILD=""
 	# FEATURES for BUILD (lto)
 	FEATURES_BUILD=""
 	# General CFlags for HOST
-	CFLAGS="$CFLAGS"
+	CFLAGS_ORIG="$CFLAGS"
+	CFLAGS=""
 	# Special CXXFlags for HOST
-	CXXFLAGS="$CXXFLAGS"
+	CXXFLAGS_ORIG="$CXXFLAGS"
+	CXXFLAGS=""
 	# Libs to compile. In fact this is just LDFLAGS
 	LIBS="-lstdc++"
 	# LDFLAGS used for HOST
-	LDFLAGS="$LDFLAGS"
+	LDFLAGS_ORIG="$LDFLAGS"
+	LDFLAGS=""
 	# FEATURES for HOST (lto)
 	FEATURES=""
 
@@ -1881,6 +1887,14 @@
 		fi
 	fi
 
+	# To allow overriding auto-generated flags, put user flags behind them.
+	CFLAGS_BUILD="$CFLAGS_BUILD $CFLAGS_BUILD_ORIG"
+	CXXFLAGS_BUILD="$CXXFLAGS_BUILD $CXXFLAGS_BUILD_ORIG"
+	LDFLAGS_BUILD="$LDFLAGS_BUILD $LDFLAGS_BUILD_ORIG"
+	CFLAGS="$CFLAGS $CFLAGS_ORIG"
+	CXXFLAGS="$CXXFLAGS $CXXFLAGS_ORIG"
+	LDFLAGS="$LDFLAGS $LDFLAGS_ORIG"
+
 	log 1 "using CFLAGS_BUILD... $CFLAGS_BUILD"
 	log 1 "using CXXFLAGS_BUILD... $CXXFLAGS_BUILD"
 	log 1 "using LDFLAGS_BUILD... $LDFLAGS_BUILD"
