diff -r 5dd72d594bf6 -r a1e148317984 src/settings_gui.cpp
--- a/src/settings_gui.cpp	Sun Sep 28 14:09:59 2008 +0200
+++ b/src/settings_gui.cpp	Sun Sep 28 14:09:59 2008 +0200
@@ -624,8 +624,14 @@
 	"construction.autoslope",
 	"construction.extra_dynamite",
 	"construction.longbridges",
+	"vehicle.disable_elrails",
+	"",
+	"vehicle.max_trains",
+	"vehicle.max_roadveh",
+	"vehicle.max_aircraft",
+	"vehicle.max_ships",
+	"",
 	"construction.signal_side",
-	"station.always_small_airport",
 	"gui.enable_signal_gui",
 	"gui.drag_signals_density",
 	"gui.semaphore_build_before",
@@ -646,6 +652,7 @@
 	"order.gradual_loading",
 	"construction.road_stop_on_town_road",
 	"station.adjacent_stations",
+	"station.always_small_airport",
 	"economy.station_noise_level",
 	NULL
 };
@@ -703,10 +710,6 @@
 	"gui.autorenew",
 	"gui.autorenew_months",
 	"gui.autorenew_money",
-	"vehicle.max_trains",
-	"vehicle.max_roadveh",
-	"vehicle.max_aircraft",
-	"vehicle.max_ships",
 	"vehicle.servint_ispercent",
 	"vehicle.servint_trains",
 	"vehicle.servint_roadveh",
@@ -714,7 +717,6 @@
 	"vehicle.servint_aircraft",
 	"order.no_servicing_if_no_breakdowns",
 	"vehicle.wagon_speed_limits",
-	"vehicle.disable_elrails",
 	"vehicle.freight_trains",
 	"vehicle.plane_speed",
 	"order.timetabling",
