Index: src/road_gui.cpp
===================================================================
--- src/road_gui.cpp	(revision 24940)
+++ src/road_gui.cpp	(working copy)
@@ -956,8 +956,7 @@
 	virtual void OnPaint()
 	{
 		this->DrawWidgets();
-
-		int rad = _settings_game.station.modified_catchment ? CA_TRUCK /* = CA_BUS */ : CA_UNMODIFIED;
+		int rad = _settings_game.station.modified_catchment ? ((this->window_class == WC_BUS_STATION) ? CA_BUS : CA_TRUCK) : CA_UNMODIFIED;
 		if (_settings_client.gui.station_show_coverage) {
 			SetTileSelectBigSize(-rad, -rad, 2 * rad, 2 * rad);
 		} else {
