Index: src/toolbar_gui.cpp
===================================================================
--- src/toolbar_gui.cpp	(revision 12736)
+++ src/toolbar_gui.cpp	(working copy)
@@ -427,6 +427,7 @@
 				case WKC_SHIFT | WKC_F4: ShowVehicleListWindow(_local_player, VEH_AIRCRAFT); break;
 				case WKC_NUM_PLUS: // Fall through
 				case WKC_EQUALS: // Fall through
+				case WKC_PLUS: // Fall through
 				case WKC_SHIFT | WKC_EQUALS: // Fall through
 				case WKC_SHIFT | WKC_F5: ToolbarZoomInClick(w); break;
 				case WKC_NUM_MINUS: // Fall through
Index: src/video/cocoa/cocoa_keys.h
===================================================================
--- src/video/cocoa/cocoa_keys.h	(revision 12736)
+++ src/video/cocoa/cocoa_keys.h	(working copy)
@@ -36,6 +36,7 @@
 #define QZ_9            0x19
 #define QZ_0            0x1D
 #define QZ_MINUS        0x1B
+#define QZ_PLUS         0x18  ///< same key code as equals
 #define QZ_EQUALS       0x18
 #define QZ_BACKSPACE    0x33
 #define QZ_INSERT       0x72
Index: src/video/cocoa/event.mm
===================================================================
--- src/video/cocoa/event.mm	(revision 12736)
+++ src/video/cocoa/event.mm	(working copy)
@@ -270,6 +270,7 @@
 	AS(QZ_QUOTE,   WKC_SINGLEQUOTE),
 	AS(QZ_COMMA,   WKC_COMMA),
 	AS(QZ_MINUS,   WKC_MINUS),
+	AS(QZ_PLUS,    WKC_PLUS),
 	AS(QZ_PERIOD,  WKC_PERIOD)
 };
 
Index: src/video/sdl_v.cpp
===================================================================
--- src/video/sdl_v.cpp	(revision 12736)
+++ src/video/sdl_v.cpp	(working copy)
@@ -287,6 +287,7 @@
 	AS(SDLK_QUOTE,   WKC_SINGLEQUOTE),
 	AS(SDLK_COMMA,   WKC_COMMA),
 	AS(SDLK_MINUS,   WKC_MINUS),
+	AS(SDLK_PLUS,    WKC_PLUS),
 	AS(SDLK_PERIOD,  WKC_PERIOD)
 };
 
Index: src/video/win32_v.cpp
===================================================================
--- src/video/win32_v.cpp	(revision 12736)
+++ src/video/win32_v.cpp	(working copy)
@@ -119,6 +119,7 @@
 	AS(0xDE,  WKC_SINGLEQUOTE),
 	AS(0xBC,  WKC_COMMA),
 	AS(0xBD,  WKC_MINUS),
+	AS(0xDB,  WKC_PLUS),
 	AS(0xBE,  WKC_PERIOD)
 };
 
Index: src/gfx_type.h
===================================================================
--- src/gfx_type.h	(revision 12736)
+++ src/gfx_type.h	(working copy)
@@ -100,6 +100,7 @@
 	WKC_COMMA       = 151, ///< , Comma
 	WKC_PERIOD      = 152, ///< . Period
 	WKC_MINUS       = 153, ///< - Minus
+	WKC_PLUS        = 154, ///< + Plus
 };
 
 /** A single sprite of a list of animated cursors */
