Index: src/autoreplace_gui.cpp
===================================================================
--- src/autoreplace_gui.cpp	(revision 13002)
+++ src/autoreplace_gui.cpp	(working copy)
@@ -85,6 +85,8 @@
 /* General Vehicle GUI based procedures that are independent of vehicle types */
 void InitializeVehiclesGuiList()
 {
+	error("sizeof(replaceveh_d) == %u, sizeof(EngineList) == %u", (uint)sizeof(replaceveh_d), (uint)sizeof(EngineList));
+
 	_railtype_selected_in_replace_gui = RAILTYPE_RAIL;
 }
 
Index: src/window_type.h
===================================================================
--- src/window_type.h	(revision 13002)
+++ src/window_type.h	(working copy)
@@ -105,6 +105,6 @@
  * the number of pointers in it and alignment will have a huge impact.
  * 96 is the largest window-size for 64-bit machines currently.
  */
-#define WINDOW_CUSTOM_SIZE 96
+#define WINDOW_CUSTOM_SIZE 256
 
 #endif /* WINDOW_TYPE_H */
