diff -r 21473a87781c src/lang/english.txt
--- a/src/lang/english.txt	Tue Apr 17 17:45:20 2012 +0000
+++ b/src/lang/english.txt	Wed Apr 25 20:21:58 2012 -0600
@@ -2723,11 +2723,12 @@
 
 STR_STATION_VIEW_ACCEPTS_BUTTON                                 :{BLACK}Accepts
 STR_STATION_VIEW_ACCEPTS_TOOLTIP                                :{BLACK}Show list of accepted cargo
-STR_STATION_VIEW_ACCEPTS_CARGO                                  :{BLACK}Accepts: {WHITE}{CARGO_LIST}
+STR_STATION_VIEW_ACCEPTS_CARGO_TITLE                            :{BLACK}Accepts:
+STR_STATION_VIEW_ACCEPTS_CARGO                                  :{WHITE}{CARGO_LIST}
 
 STR_STATION_VIEW_RATINGS_BUTTON                                 :{BLACK}Ratings
 STR_STATION_VIEW_RATINGS_TOOLTIP                                :{BLACK}Show station ratings
-STR_STATION_VIEW_CARGO_RATINGS_TITLE                            :{BLACK}Local rating of transport service:
+STR_STATION_VIEW_CARGO_RATINGS_TITLE                            :{BLACK}Supplies:
 STR_STATION_VIEW_CARGO_RATING                                   :{WHITE}{STRING}: {YELLOW}{STRING} ({COMMA}%)
 
 ############ range for rating starts
diff -r 21473a87781c src/script/api/game/game_window.hpp.sq
--- a/src/script/api/game/game_window.hpp.sq	Tue Apr 17 17:45:20 2012 +0000
+++ b/src/script/api/game/game_window.hpp.sq	Wed Apr 25 20:21:58 2012 -0600
@@ -1038,7 +1038,6 @@
 	SQGSWindow.DefSQConst(engine, ScriptWindow::WID_SV_SCROLLBAR,                          "WID_SV_SCROLLBAR");
 	SQGSWindow.DefSQConst(engine, ScriptWindow::WID_SV_ACCEPT_RATING_LIST,                 "WID_SV_ACCEPT_RATING_LIST");
 	SQGSWindow.DefSQConst(engine, ScriptWindow::WID_SV_LOCATION,                           "WID_SV_LOCATION");
-	SQGSWindow.DefSQConst(engine, ScriptWindow::WID_SV_ACCEPTS_RATINGS,                    "WID_SV_ACCEPTS_RATINGS");
 	SQGSWindow.DefSQConst(engine, ScriptWindow::WID_SV_RENAME,                             "WID_SV_RENAME");
 	SQGSWindow.DefSQConst(engine, ScriptWindow::WID_SV_TRAINS,                             "WID_SV_TRAINS");
 	SQGSWindow.DefSQConst(engine, ScriptWindow::WID_SV_ROADVEHS,                           "WID_SV_ROADVEHS");
diff -r 21473a87781c src/script/api/script_window.hpp
--- a/src/script/api/script_window.hpp	Tue Apr 17 17:45:20 2012 +0000
+++ b/src/script/api/script_window.hpp	Wed Apr 25 20:21:58 2012 -0600
@@ -2057,7 +2057,6 @@
 		WID_SV_SCROLLBAR                     = ::WID_SV_SCROLLBAR,                     ///< Scrollbar.
 		WID_SV_ACCEPT_RATING_LIST            = ::WID_SV_ACCEPT_RATING_LIST,            ///< List of accepted cargoes / rating of cargoes.
 		WID_SV_LOCATION                      = ::WID_SV_LOCATION,                      ///< 'Location' button.
-		WID_SV_ACCEPTS_RATINGS               = ::WID_SV_ACCEPTS_RATINGS,               ///< 'Accepts' / 'Ratings' button.
 		WID_SV_RENAME                        = ::WID_SV_RENAME,                        ///< 'Rename' button.
 		WID_SV_TRAINS                        = ::WID_SV_TRAINS,                        ///< List of scheduled trains button.
 		WID_SV_ROADVEHS                      = ::WID_SV_ROADVEHS,                      ///< List of scheduled road vehs button.
diff -r 21473a87781c src/station_gui.cpp
--- a/src/station_gui.cpp	Tue Apr 17 17:45:20 2012 +0000
+++ b/src/station_gui.cpp	Wed Apr 25 20:21:58 2012 -0600
@@ -765,8 +765,6 @@
 	NWidget(NWID_HORIZONTAL, NC_EQUALSIZE),
 		NWidget(WWT_PUSHTXTBTN, COLOUR_GREY, WID_SV_LOCATION), SetMinimalSize(60, 12), SetResize(1, 0), SetFill(1, 1),
 				SetDataTip(STR_BUTTON_LOCATION, STR_STATION_VIEW_CENTER_TOOLTIP),
-		NWidget(WWT_PUSHTXTBTN, COLOUR_GREY, WID_SV_ACCEPTS_RATINGS), SetMinimalSize(61, 12), SetResize(1, 0), SetFill(1, 1),
-				SetDataTip(STR_STATION_VIEW_RATINGS_BUTTON, STR_STATION_VIEW_RATINGS_TOOLTIP),
 		NWidget(WWT_PUSHTXTBTN, COLOUR_GREY, WID_SV_RENAME), SetMinimalSize(60, 12), SetResize(1, 0), SetFill(1, 1),
 				SetDataTip(STR_BUTTON_RENAME, STR_STATION_VIEW_RENAME_TOOLTIP),
 		NWidget(WWT_PUSHTXTBTN, COLOUR_GREY, WID_SV_TRAINS), SetMinimalSize(14, 12), SetFill(0, 1), SetDataTip(STR_TRAIN, STR_STATION_VIEW_SCHEDULED_TRAINS_TOOLTIP),
@@ -822,24 +820,17 @@
 	uint32 cargo;                 ///< Bitmask of cargo types to expand
 	uint16 cargo_rows[NUM_CARGO]; ///< Header row for each cargo type
 	uint expand_shrink_width;     ///< The width allocated to the expand/shrink 'button'
-	int rating_lines;             ///< Number of lines in the cargo ratings view.
-	int accepts_lines;            ///< Number of lines in the accepted cargo view.
+	int accept_rating_lines;      ///< Number of lines in the cargo ratings view.
 	Scrollbar *vscroll;
 
-	/** Height of the #WID_SV_ACCEPT_RATING_LIST widget for different views. */
-	enum AcceptListHeight {
-		ALH_RATING  = 13, ///< Height of the cargo ratings view.
-		ALH_ACCEPTS = 3,  ///< Height of the accepted cargo view.
-	};
-
 	StationViewWindow(const WindowDesc *desc, WindowNumber window_number) : Window()
 	{
-		this->rating_lines  = ALH_RATING;
-		this->accepts_lines = ALH_ACCEPTS;
+		this->accept_rating_lines  = 0;
 
 		this->CreateNestedTree(desc);
 		this->vscroll = this->GetScrollbar(WID_SV_SCROLLBAR);
 		/* Nested widget tree creation is done in two steps to ensure that this->GetWidget<NWidgetCore>(WID_SV_ACCEPTS_RATINGS) exists in UpdateWidgetSize(). */
+		/* XXX That's gone so what is this doing now? */
 		this->FinishInitNested(desc, window_number);
 
 		Owner owner = Station::Get(window_number)->owner;
@@ -867,7 +858,7 @@
 				break;
 
 			case WID_SV_ACCEPT_RATING_LIST:
-				size->height = WD_FRAMERECT_TOP + ((this->GetWidget<NWidgetCore>(WID_SV_ACCEPTS_RATINGS)->widget_data == STR_STATION_VIEW_RATINGS_BUTTON) ? this->accepts_lines : this->rating_lines) * FONT_HEIGHT_NORMAL + WD_FRAMERECT_BOTTOM;
+				size->height = WD_FRAMERECT_TOP + this->accept_rating_lines * FONT_HEIGHT_NORMAL + WD_FRAMERECT_BOTTOM;
 				break;
 		}
 	}
@@ -894,20 +885,11 @@
 			/* Draw 'accepted cargo' or 'cargo ratings'. */
 			const NWidgetBase *wid = this->GetWidget<NWidgetBase>(WID_SV_ACCEPT_RATING_LIST);
 			const Rect r = {wid->pos_x, wid->pos_y, wid->pos_x + wid->current_x - 1, wid->pos_y + wid->current_y - 1};
-			if (this->GetWidget<NWidgetCore>(WID_SV_ACCEPTS_RATINGS)->widget_data == STR_STATION_VIEW_RATINGS_BUTTON) {
-				int lines = this->DrawAcceptedCargo(r);
-				if (lines > this->accepts_lines) { // Resize the widget, and perform re-initialization of the window.
-					this->accepts_lines = lines;
-					this->ReInit();
-					return;
-				}
-			} else {
-				int lines = this->DrawCargoRatings(r);
-				if (lines > this->rating_lines) { // Resize the widget, and perform re-initialization of the window.
-					this->rating_lines = lines;
-					this->ReInit();
-					return;
-				}
+			int lines = this->DrawAcceptRating(r);
+			if (lines > this->accept_rating_lines) { // Resize the widget, and perform re-initialization of the window.
+				this->accept_rating_lines = lines;
+				this->ReInit();
+				return;
 			}
 
 			/* Draw waiting cargo. */
@@ -1040,33 +1022,28 @@
 	}
 
 	/**
-	 * Draw accepted cargo in the #WID_SV_ACCEPT_RATING_LIST widget.
+	 * Draw accepted and ratings in the #WID_SV_ACCEPT_RATING_LIST widget.
 	 * @param r Rectangle of the widget.
-	 * @return Number of lines needed for drawing the accepted cargo.
+	 * @return Number of lines needed for drawing the accepted and ratings text.
 	 */
-	int DrawAcceptedCargo(const Rect &r) const
+	int DrawAcceptRating(const Rect &r) const
 	{
 		const Station *st = Station::Get(this->window_number);
+		int y = r.top + WD_FRAMERECT_TOP;
+
+		/* Accepted cargo */
+		DrawString(r.left + WD_FRAMERECT_LEFT, r.right - WD_FRAMERECT_RIGHT, y, STR_STATION_VIEW_ACCEPTS_CARGO_TITLE);
+		y += FONT_HEIGHT_NORMAL;
 
 		uint32 cargo_mask = 0;
 		for (CargoID i = 0; i < NUM_CARGO; i++) {
 			if (HasBit(st->goods[i].acceptance_pickup, GoodsEntry::GES_ACCEPTANCE)) SetBit(cargo_mask, i);
 		}
 		SetDParam(0, cargo_mask);
-		int bottom = DrawStringMultiLine(r.left + WD_FRAMERECT_LEFT, r.right - WD_FRAMERECT_RIGHT, r.top + WD_FRAMERECT_TOP, INT32_MAX, STR_STATION_VIEW_ACCEPTS_CARGO);
-		return CeilDiv(bottom - r.top - WD_FRAMERECT_TOP, FONT_HEIGHT_NORMAL);
-	}
+		y = DrawStringMultiLine(r.left + WD_FRAMERECT_LEFT + 6, r.right - WD_FRAMERECT_RIGHT, y, INT32_MAX, STR_STATION_VIEW_ACCEPTS_CARGO);
+		y += FONT_HEIGHT_NORMAL;
 
-	/**
-	 * Draw cargo ratings in the #WID_SV_ACCEPT_RATING_LIST widget.
-	 * @param r Rectangle of the widget.
-	 * @return Number of lines needed for drawing the cargo ratings.
-	 */
-	int DrawCargoRatings(const Rect &r) const
-	{
-		const Station *st = Station::Get(this->window_number);
-		int y = r.top + WD_FRAMERECT_TOP;
-
+		/* Ratings of supplied cargo */
 		DrawString(r.left + WD_FRAMERECT_LEFT, r.right - WD_FRAMERECT_RIGHT, y, STR_STATION_VIEW_CARGO_RATINGS_TITLE);
 		y += FONT_HEIGHT_NORMAL;
 
@@ -1081,6 +1058,7 @@
 			DrawString(r.left + WD_FRAMERECT_LEFT + 6, r.right - WD_FRAMERECT_RIGHT - 6, y, STR_STATION_VIEW_CARGO_RATING);
 			y += FONT_HEIGHT_NORMAL;
 		}
+
 		return CeilDiv(y - r.top - WD_FRAMERECT_TOP, FONT_HEIGHT_NORMAL);
 	}
 
@@ -1112,21 +1090,6 @@
 				}
 				break;
 
-			case WID_SV_ACCEPTS_RATINGS: {
-				/* Swap between 'accepts' and 'ratings' view. */
-				int height_change;
-				NWidgetCore *nwi = this->GetWidget<NWidgetCore>(WID_SV_ACCEPTS_RATINGS);
-				if (this->GetWidget<NWidgetCore>(WID_SV_ACCEPTS_RATINGS)->widget_data == STR_STATION_VIEW_RATINGS_BUTTON) {
-					nwi->SetDataTip(STR_STATION_VIEW_ACCEPTS_BUTTON, STR_STATION_VIEW_ACCEPTS_TOOLTIP); // Switch to accepts view.
-					height_change = this->rating_lines - this->accepts_lines;
-				} else {
-					nwi->SetDataTip(STR_STATION_VIEW_RATINGS_BUTTON, STR_STATION_VIEW_RATINGS_TOOLTIP); // Switch to ratings view.
-					height_change = this->accepts_lines - this->rating_lines;
-				}
-				this->ReInit(0, height_change * FONT_HEIGHT_NORMAL);
-				break;
-			}
-
 			case WID_SV_RENAME:
 				SetDParam(0, this->window_number);
 				ShowQueryString(STR_STATION_NAME, STR_STATION_VIEW_RENAME_STATION_CAPTION, MAX_LENGTH_STATION_NAME_CHARS,
diff -r 21473a87781c src/widgets/station_widget.h
--- a/src/widgets/station_widget.h	Tue Apr 17 17:45:20 2012 +0000
+++ b/src/widgets/station_widget.h	Wed Apr 25 20:21:58 2012 -0600
@@ -19,7 +19,6 @@
 	WID_SV_SCROLLBAR,          ///< Scrollbar.
 	WID_SV_ACCEPT_RATING_LIST, ///< List of accepted cargoes / rating of cargoes.
 	WID_SV_LOCATION,           ///< 'Location' button.
-	WID_SV_ACCEPTS_RATINGS,    ///< 'Accepts' / 'Ratings' button.
 	WID_SV_RENAME,             ///< 'Rename' button.
 	WID_SV_TRAINS,             ///< List of scheduled trains button.
 	WID_SV_ROADVEHS,           ///< List of scheduled road vehs button.
