diff -r 4bde864867e2 src/viewport.cpp
--- a/src/viewport.cpp	Thu Aug 18 23:32:47 2011 +0200
+++ b/src/viewport.cpp	Thu Aug 18 23:49:38 2011 +0200
@@ -1833,6 +1833,9 @@
 		/* Don't check if the display options are disabled */
 		if (!HasBit(_display_opt, is_station ? DO_SHOW_STATION_NAMES : DO_SHOW_WAYPOINT_NAMES)) continue;
 
+		/* Don't check if competitor signs are not shown and the sign isn't owned by the local company */
+		if (!HasBit(_display_opt, DO_SHOW_COMPETITOR_SIGNS) && _local_company != st->owner && st->owner != OWNER_NONE) continue; 
+
 		if (CheckClickOnViewportSign(vp, x, y, &st->sign)) {
 			if (is_station) {
 				ShowStationViewWindow(st->index);
@@ -1854,6 +1857,9 @@
 
 	const Sign *si;
 	FOR_ALL_SIGNS(si) {
+		/* If competitor signs are hidden, don't check signs that aren't owned by local company */
+		if (!HasBit(_display_opt, DO_SHOW_COMPETITOR_SIGNS) && _local_company != si->owner) continue; 
+
 		if (CheckClickOnViewportSign(vp, x, y, &si->sign)) {
 			HandleClickOnSign(si);
 			return true;
