diff -r 56669363ce41 src/settings_gui.cpp
--- a/src/settings_gui.cpp	Sun Nov 02 11:25:57 2008 +0100
+++ b/src/settings_gui.cpp	Sun Nov 02 12:53:51 2008 +0100
@@ -583,6 +583,8 @@
 	new GameDifficultyWindow();
 }
 
+static const int SETTING_HEIGHT = 11;         ///< Height of a single patch setting in the tree view
+
 static const char *_patches_ui[] = {
 	"gui.vehicle_speed",
 	"gui.status_long_date",
@@ -784,10 +786,10 @@
 		}
 
 		/* Resize the window to fit the largest patch tab */
-		ResizeWindowForWidget(this, PATCHSEL_OPTIONSPANEL, 0, patches_max * 11);
+		ResizeWindowForWidget(this, PATCHSEL_OPTIONSPANEL, 0, patches_max * SETTING_HEIGHT);
 
 		/* Recentre the window for the new size */
-		this->top = this->top - (patches_max * 11) / 2;
+		this->top = this->top - (patches_max * SETTING_HEIGHT) / 2;
 
 		this->LowerWidget(PATCHSEL_INTERFACE);
 
@@ -847,7 +849,7 @@
 				}
 			}
 			DrawString(30, y, (sdb->str) + disabled, TC_FROMSTRING);
-			y += 11;
+			y += SETTING_HEIGHT;
 		}
 	}
 
@@ -868,8 +870,8 @@
 				x = pt.x - 5;  // Shift x coordinate
 				if (x < 0) return;  // Clicked left of the entry
 
-				btn = y / 11;  // Compute which setting is selected
-				if (y % 11 > 9) return;  // Clicked too low at the setting
+				btn = y / SETTING_HEIGHT;  // Compute which setting is selected
+				if (y % SETTING_HEIGHT > SETTING_HEIGHT - 2) return;  // Clicked too low at the setting
 				if (btn >= page->num) return;  // Clicked below the last setting of the page
 
 				sd = page->entries[btn].setting;
