Index: src/settings_gui.cpp
===================================================================
--- src/settings_gui.cpp	(revision 14937)
+++ src/settings_gui.cpp	(working copy)
@@ -870,7 +870,7 @@
 				SetDParam(1, value);
 			}
 		}
-		DrawString(x + 25, y, (sdb->str) + disabled, TC_FROMSTRING);
+		DrawStringTruncated(x + 25, y, (sdb->str) + disabled, TC_FROMSTRING, this->width - x - 38);
 	}
 
 	virtual void OnClick(Point pt, int widget)
