Index: src/tunnelbridge_cmd.cpp
===================================================================
--- src/tunnelbridge_cmd.cpp	(revision 21000)
+++ src/tunnelbridge_cmd.cpp	(working copy)
@@ -231,6 +231,31 @@
 	};
 }
 
+/** 
+ * Can we can build bridge middle on given tile in given direction?
+ *
+ * @param tile   The tile to check
+ * @param dir    The direction of the bridge
+ * @param z      The height of the bridge
+ * @return       True if we can build the bridge middle above
+ */
+static inline bool AllowBridgeAbove(TileIndex tile, DiagDirection dir, uint z) {
+	switch (GetTileType(tile)) {
+		case MP_CLEAR:        return true;
+		case MP_WATER:        return (IsWater(tile) || IsCoast(tile));
+		case MP_RAILWAY:      return (IsPlainRailTile(tile));
+		case MP_ROAD:         return (!IsRoadDepot(tile));
+		case MP_TUNNELBRIDGE: return (IsTunnel(tile) ||
+																	(z >= GetBridgeHeight(tile) && dir != GetTunnelBridgeDirection(tile)));
+		case MP_OBJECT: {
+			const ObjectSpec *spec = ObjectSpec::GetByTile(tile);
+			                    return ((spec->flags & OBJECT_FLAG_ALLOW_UNDER_BRIDGE) &&
+																	(GetTileMaxZ(tile) + spec->height * TILE_HEIGHT <= z));
+		}
+		default: return false;
+	}
+}
+
 /**
  * Build a Bridge
  * @param end_tile end tile
