Index: C:/Users/Alan/Desktop/openttd-dev/src/rail_gui.cpp
===================================================================
--- C:/Users/Alan/Desktop/openttd-dev/src/rail_gui.cpp	(revision 22844)
+++ C:/Users/Alan/Desktop/openttd-dev/src/rail_gui.cpp	(working copy)
@@ -386,19 +386,22 @@
 		return;
 	}
 
+	/* Ensure that the signal density is always effectively 1 when removing signals.
+	   This is achieved by simulating CTRL being pressed. */
+
 	const Window *w = FindWindowById(WC_BUILD_SIGNAL, 0);
 
 	if (w != NULL) {
 		/* signal GUI is used */
 		SB(p2,  3, 1, 0);
 		SB(p2,  4, 1, _cur_signal_variant);
-		SB(p2,  6, 1, _ctrl_pressed);
+		SB(p2,  6, 1, _remove_button_clicked ? 1 : _ctrl_pressed);
 		SB(p2,  7, 3, _cur_signal_type);
 		SB(p2, 24, 8, _settings_client.gui.drag_signals_density);
 	} else {
 		SB(p2,  3, 1, 0);
 		SB(p2,  4, 1, (_cur_year < _settings_client.gui.semaphore_build_before ? SIG_SEMAPHORE : SIG_ELECTRIC));
-		SB(p2,  6, 1, _ctrl_pressed);
+		SB(p2,  6, 1, _remove_button_clicked ? 1 : _ctrl_pressed);
 		SB(p2,  7, 3, _default_signal_type[_settings_client.gui.default_signal_type]);
 		SB(p2, 24, 8, _settings_client.gui.drag_signals_density);
 	}