Index: src/gfx_func.h
===================================================================
--- src/gfx_func.h	(revision 14151)
+++ src/gfx_func.h	(working copy)
@@ -179,6 +179,6 @@
  */
 extern byte _colour_gradient[COLOUR_END][8];
 
-extern Palette _use_palette;
+extern Palettes _use_palette;
 
 #endif /* GFX_FUNC_H */
Index: src/gfx_type.h
===================================================================
--- src/gfx_type.h	(revision 14151)
+++ src/gfx_type.h	(working copy)
@@ -236,7 +236,7 @@
 };
 
 /** Palettes OpenTTD supports. */
-enum Palette {
+enum Palettes {
 	PAL_DOS,        ///< Use the DOS palette.
 	PAL_WINDOWS,    ///< Use the Windows palette.
 	PAL_AUTODETECT, ///< Automatically detect the palette based on the graphics pack.
Index: src/gfxinit.cpp
===================================================================
--- src/gfxinit.cpp	(revision 14151)
+++ src/gfxinit.cpp	(working copy)
@@ -22,7 +22,7 @@
 
 #include "table/sprites.h"
 
-Palette _use_palette = PAL_AUTODETECT;
+Palettes _use_palette = PAL_AUTODETECT;
 
 struct MD5File {
 	const char * filename;     ///< filename
@@ -35,7 +35,7 @@
 struct GraphicsSet {
 	const char *name;          ///< The name of the graphics set
 	const char *description;   ///< Description of the graphics set
-	Palette palette;           ///< Palette of this graphics set
+	Palettes palette;           ///< Palette of this graphics set
 	MD5File basic[2];          ///< GRF files that always have to be loaded
 	MD5File landscape[3];      ///< Landscape specific grf files
 	const char *base_missing;  ///< Warning when one of the base GRF files is missing
