Index: src/fileio.cpp
===================================================================
--- src/fileio.cpp	(revision 19135)
+++ src/fileio.cpp	(working copy)
@@ -18,6 +18,9 @@
 #include "tar_type.h"
 #ifdef WIN32
 #include <windows.h>
+#elif defined(__HAIKU__)
+#include <Path.h>
+#include <storage/FindDirectory.h>
 #else
 #ifdef OPENBSD
 #include <unistd.h>
@@ -849,12 +852,18 @@
 #if defined(__MORPHOS__) || defined(__AMIGA__) || defined(DOS) || defined(OS2) || !defined(WITH_PERSONAL_DIR)
 	_searchpaths[SP_PERSONAL_DIR] = NULL;
 #else
+#ifdef __HAIKU__
+	BPath path;
+	find_directory(B_USER_SETTINGS_DIRECTORY, &path);
+	const char *homedir = path.Path();
+#else
 	const char *homedir = getenv("HOME");
 
 	if (homedir == NULL) {
 		const struct passwd *pw = getpwuid(getuid());
 		homedir = (pw == NULL) ? "" : pw->pw_dir;
 	}
+#endif
 
 	snprintf(tmp, MAX_PATH, "%s" PATHSEP "%s", homedir, PERSONAL_DIR);
 	AppendPathSeparator(tmp, MAX_PATH);
Index: config.lib
===================================================================
--- config.lib	(revision 19135)
+++ config.lib	(working copy)
@@ -1018,7 +1018,7 @@
 	fi
 
 	if [ "$personal_dir" = "1" ]; then
-		if [ "$os" = "MINGW" ] || [ "$os" = "CYGWIN" ] || [ "$os" = "WINCE" ] || [ "$os" = "DOS" ]; then
+		if [ "$os" = "MINGW" ] || [ "$os" = "CYGWIN" ] || [ "$os" = "WINCE" ] || [ "$os" = "DOS" ] || [ "$os" == "HAIKU" ]; then
 			personal_dir="OpenTTD"
 		elif [ "$os" = "OSX" ]; then
 			personal_dir="Documents/OpenTTD"
