diff --git a/src/group_gui.cpp b/src/group_gui.cpp
--- a/src/group_gui.cpp
+++ b/src/group_gui.cpp
@@ -630,6 +630,8 @@
 	{
 		switch (widget) {
 			case WID_GL_ALL_VEHICLES: // All vehicles
+				break;
+
 			case WID_GL_DEFAULT_VEHICLES: // Ungrouped vehicles
 				DoCommandP(0, DEFAULT_GROUP, this->vehicle_sel | (_ctrl_pressed ? 1 << 31 : 0), CMD_ADD_VEHICLE_GROUP | CMD_MSG(STR_ERROR_GROUP_CAN_T_ADD_VEHICLE));
 
@@ -753,13 +755,13 @@
 		GroupID new_group_over = INVALID_GROUP;
 		switch (widget) {
 			case WID_GL_ALL_VEHICLES: // ... the 'all' group.
-				/* Moving a vehicle from the all group to the default group
-				 * is a no-op, so do not highlight then. */
-				if (!IsDefaultGroupID(Vehicle::Get(vehicle_sel)->group_id)) new_group_over = ALL_GROUP;
+				/* Moving a vehicle from any group to the all group
+				 * is always a no-op, so do not highlight then. */
+				/* this case can be omitted to simplify code */
 				break;
 
 			case WID_GL_DEFAULT_VEHICLES: // ... the 'default' group.
-				/* Moving a vehicle from the default group to the all group
+				/* Moving a vehicle from default group to default group
 				 * is a no-op, so do not highlight then. */
 				if (!IsDefaultGroupID(Vehicle::Get(vehicle_sel)->group_id)) new_group_over = DEFAULT_GROUP;
 				break;
