Index: src/openttd.h
===================================================================
--- src/openttd.h	(revision 11760)
+++ src/openttd.h	(working copy)
@@ -45,9 +45,9 @@
 /* DestinationID must be at least as large as every these below, because it can
  * be any of them
  */
-assert_compile(sizeof(DestinationID) == sizeof(DepotID));
-assert_compile(sizeof(DestinationID) == sizeof(WaypointID));
-assert_compile(sizeof(DestinationID) == sizeof(StationID));
+assert_compile(sizeof(DestinationID) >= sizeof(DepotID));
+assert_compile(sizeof(DestinationID) >= sizeof(WaypointID));
+assert_compile(sizeof(DestinationID) >= sizeof(StationID));
 
 typedef uint32 PlayerFace; ///< player face bits, info see in player_face.h
 
