Index: src/console_cmds.cpp
===================================================================
--- src/console_cmds.cpp	(revision 12344)
+++ src/console_cmds.cpp	(working copy)
@@ -1229,6 +1229,21 @@
 	return true;
 }
 
+DEF_CONSOLE_CMD(ConDo)
+{
+	if (argc == 0) {
+		IConsoleHelp("Display an action for all clients in a multiplayer game. Usage: 'do \"<msg>\"'");
+		return true;
+	}
+
+	if (argc != 2) return false;
+
+	NetworkServer_HandleChat(NETWORK_ACTION_SERVER_MESSAGE, DESTTYPE_BROADCAST, 0, argv[1], NETWORK_SERVER_INDEX);
+
+
+	return true;
+}
+
 DEF_CONSOLE_CMD(ConPlayers)
 {
 	Player *p;
@@ -1559,6 +1574,8 @@
 	IConsoleCmdHookAdd("say_player",       ICONSOLE_HOOK_ACCESS, ConHookNeedNetwork);
 	IConsoleCmdRegister("say_client",      ConSayClient);
 	IConsoleCmdHookAdd("say_client",       ICONSOLE_HOOK_ACCESS, ConHookNeedNetwork);
+	IConsoleCmdRegister("do",                   ConDo);
+	IConsoleCmdHookAdd("do",                  ICONSOLE_HOOK_ACCESS, ConHookServerOnly);
 
 	IConsoleCmdRegister("connect",         ConNetworkConnect);
 	IConsoleCmdHookAdd("connect",          ICONSOLE_HOOK_ACCESS, ConHookClientOnly);
