Index: src/ai/ai_instance.cpp
===================================================================
--- src/ai/ai_instance.cpp	(revision 19386)
+++ src/ai/ai_instance.cpp	(working copy)
@@ -121,10 +121,10 @@
 	/* Register the API functions and classes */
 	this->RegisterAPI();
 
-	if (!this->LoadCompatibilityScripts(info->GetAPIVersion())) {
+//	if (!this->LoadCompatibilityScripts(info->GetAPIVersion())) {
 		this->Died();
 		return;
-	}
+//	}
 
 	try {
 		AIObject::SetAllowDoCommand(false);
