Index: src/oldloader.cpp
===================================================================
--- src/oldloader.cpp	(revision 11276)
+++ src/oldloader.cpp	(working copy)
@@ -224,7 +224,7 @@
 					default: break;
 				}
 			} else {
-				uint32 res = 0;
+				uint64 res = 0;
 
 				/* Reading from the file: bits 16 to 23 have the FILE type */
 				switch (GetOldChunkFileType(chunk->type)) {
@@ -1184,8 +1184,8 @@
 	OCL_SVAR( OC_UINT16, Vehicle, reliability ),
 	OCL_SVAR( OC_UINT16, Vehicle, reliability_spd_dec ),
 
-	OCL_SVAR(  OC_INT32, Vehicle, profit_this_year ),
-	OCL_SVAR(  OC_INT32, Vehicle, profit_last_year ),
+	OCL_SVAR( OC_FILE_I32 | OC_VAR_I64, Vehicle, profit_this_year ),
+	OCL_SVAR( OC_FILE_I32 | OC_VAR_I64, Vehicle, profit_last_year ),
 
 	OCL_VAR ( OC_UINT16,   1, &_old_next_ptr ),
 
