From c51d672cf68e4ec845b0f71364e56571e8d709f4 Mon Sep 17 00:00:00 2001
From: Chris Angelico <rosuav@gmail.com>
Date: Sat, 6 Oct 2012 00:18:38 +1000
Subject: [PATCH 3/5] Permit interest rate to go much higher

---
 src/table/settings.ini |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/table/settings.ini b/src/table/settings.ini
index 933e792..a898ba1 100644
--- a/src/table/settings.ini
+++ b/src/table/settings.ini
@@ -156,7 +156,7 @@ from     = 97
 guiflags = SGF_NEWGAME_ONLY | SGF_SCENEDIT_TOO
 def      = 2
 min      = 2
-max      = 4
+max      = 40
 interval = 1
 proc     = DifficultyChange
 
-- 
1.7.9.GIT

