diff --git a/src/gfx.cpp b/src/gfx.cpp
index 26d8b66..bc7f932 100644
--- a/src/gfx.cpp
+++ b/src/gfx.cpp
@@ -23,6 +23,7 @@
 #include "newgrf_debug.h"
 
 #include "table/palettes.h"
+#include "table/string_colours.h"
 #include "table/sprites.h"
 #include "table/control_codes.h"
 
diff --git a/src/gfx_type.h b/src/gfx_type.h
index 57a7ca3..d25abff 100644
--- a/src/gfx_type.h
+++ b/src/gfx_type.h
@@ -237,7 +237,7 @@ enum Colours {
 };
 template <> struct EnumPropsT<Colours> : MakeEnumPropsT<Colours, byte, COLOUR_BEGIN, COLOUR_END, INVALID_COLOUR, 4> {};
 
-/** Colour of the strings, see _string_colourmap in table/palettes.h or docs/ottd-colourtext-palette.png */
+/** Colour of the strings, see _string_colourmap in table/string_colours.h or docs/ottd-colourtext-palette.png */
 enum TextColour {
 	TC_BEGIN       = 0x00,
 	TC_FROMSTRING  = 0x00,
diff --git a/src/table/palettes.h b/src/table/palettes.h
index e190134..93cb738 100644
--- a/src/table/palettes.h
+++ b/src/table/palettes.h
@@ -147,24 +147,3 @@ static const ExtraPaletteValues _extra_palette_values = {
 	  M(116, 180, 196), M(148, 200, 216), M(180, 220, 232) }
 };
 #undef M
-
-/** Colour mapping for the TextColours. */
-static const byte _string_colourmap[17] = {
-		150, // TC_BLUE
-		 12, // TC_SILVER
-		189, // TC_GOLD
-		184, // TC_RED
-		174, // TC_PURPLE
-		 30, // TC_LIGHT_BROWN
-		195, // TC_ORANGE
-		209, // TC_GREEN
-		 68, // TC_YELLOW
-		 95, // TC_DARK_GREEN
-		 79, // TC_CREAM
-		116, // TC_BROWN
-		 15, // TC_WHITE
-		152, // TC_LIGHT_BLUE
-		  6, // TC_GREY
-		133, // TC_DARK_BLUE
-		  1, // TC_BLACK
-};
diff --git a/src/table/string_colours.h b/src/table/string_colours.h
new file mode 100644
index 0000000..1e3b02b
--- /dev/null
+++ b/src/table/string_colours.h
@@ -0,0 +1,31 @@
+/* $Id$ */
+
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file string_colours.h The colour translation of GRF's strings. */
+
+/** Colour mapping for the TextColours. */
+static const byte _string_colourmap[17] = {
+		150, // TC_BLUE
+		 12, // TC_SILVER
+		189, // TC_GOLD
+		184, // TC_RED
+		174, // TC_PURPLE
+		 30, // TC_LIGHT_BROWN
+		195, // TC_ORANGE
+		209, // TC_GREEN
+		 68, // TC_YELLOW
+		 95, // TC_DARK_GREEN
+		 79, // TC_CREAM
+		116, // TC_BROWN
+		 15, // TC_WHITE
+		152, // TC_LIGHT_BLUE
+		  6, // TC_GREY
+		133, // TC_DARK_BLUE
+		  1, // TC_BLACK
+};
diff --git a/src/viewport.cpp b/src/viewport.cpp
index c1d4d61..0dfc894 100644
--- a/src/viewport.cpp
+++ b/src/viewport.cpp
@@ -88,7 +88,7 @@
 #include <map>
 
 #include "table/strings.h"
-#include "table/palettes.h"
+#include "table/string_colours.h"
 
 #include "safeguards.h"
 
diff --git a/src/widget.cpp b/src/widget.cpp
index 350db9d..3b50f71 100644
--- a/src/widget.cpp
+++ b/src/widget.cpp
@@ -22,7 +22,7 @@
 
 #include "table/sprites.h"
 #include "table/strings.h"
-#include "table/palettes.h"
+#include "table/string_colours.h"
 
 #include "safeguards.h"
 
