Index: src/station_cmd.cpp
===================================================================
--- src/station_cmd.cpp	(revision 17152)
+++ src/station_cmd.cpp	(working copy)
@@ -711,6 +711,19 @@
 				}
 			}
 		} else if (check_clear) {
+			/* station is non-NULL only when building a railway station. */
+			if (GetTileType(tile_cur) == MP_RAILWAY && station != NULL) {
+				TrackBits tracks = GetTrackBits(tile_cur);
+				Track track = RemoveFirstTrack(&tracks);
+				/* invalid_dirs is either 5 or 10 and valid track values are 0 or 1 */
+				if (RemoveFirstTrack(&tracks) == INVALID_TRACK && track == (Track)invalid_dirs >> 3) {
+					CommandCost ret = DoCommand(tile_cur, 0, track, flags, CMD_REMOVE_SINGLE_RAIL);
+					if (CmdFailed(ret)) return ret;
+					cost.AddCost(ret);
+					/* With flags & ~DC_EXEC CmdLandscapeClear would fail since the rail still exists */
+					continue;
+				}
+			}
 			CommandCost ret = DoCommand(tile_cur, 0, 0, flags, CMD_LANDSCAPE_CLEAR);
 			if (CmdFailed(ret)) return ret;
 			cost.AddCost(ret);
