Index: src/window.cpp
===================================================================
--- src/window.cpp	(revision 15895)
+++ src/window.cpp	(working copy)
@@ -1999,7 +1999,8 @@
 						_cursor.sprite != SPR_CURSOR_QUERY &&
 						_cursor.sprite != SPR_CURSOR_SIGN &&
 						_pause_game != 0 &&
-						!_cheats.build_in_pause.value) {
+						!_cheats.build_in_pause.value &&
+						_game_mode != GM_EDITOR) {
 					return;
 				}
 
Index: src/openttd.cpp
===================================================================
--- src/openttd.cpp	(revision 15895)
+++ src/openttd.cpp	(working copy)
@@ -1187,7 +1187,7 @@
 
 	if (!_pause_game && HasBit(_display_opt, DO_FULL_ANIMATION)) DoPaletteAnimations();
 
-	if (!_pause_game || _cheats.build_in_pause.value) MoveAllTextEffects();
+	if (!_pause_game || _cheats.build_in_pause.value || _game_mode == GM_EDITOR) MoveAllTextEffects();
 
 	InputLoop();
 
