Index: src/terraform_cmd.cpp
===================================================================
--- src/terraform_cmd.cpp	(revision 15100)
+++ src/terraform_cmd.cpp	(working copy)
@@ -359,7 +359,10 @@
 
 	/* Check range of destination height */
 	if (h > MAX_TILE_HEIGHT) return_cmd_error((oldh == 0) ? STR_1003_ALREADY_AT_SEA_LEVEL : STR_1004_TOO_HIGH);
-	if (p2 == 0) _error_message = STR_ALREADY_LEVELLED;
+	if (p2 == 0) {
+		_terraform_err_tile = INVALID_TILE;
+		_error_message = STR_ALREADY_LEVELLED;
+	}
 
 	/* make sure sx,sy are smaller than ex,ey */
 	int ex = TileX(tile);
