Index: Makefile.in
===================================================================
--- Makefile.in	(Revision 10312)
+++ Makefile.in	(Arbeitskopie)
@@ -35,6 +35,30 @@
 
 RES := $(shell if ! [ -f $(CONFIG_CACHE_SOURCE_LIST) ] || [ -n "`cmp $(CONFIG_CACHE_SOURCE_LIST) $(SOURCE_LIST)`" ]; then cp $(SOURCE_LIST) $(CONFIG_CACHE_SOURCE_LIST); fi )
 
+help:
+	@echo "----------------"
+	@echo "Command list:"
+	@echo "----------------"
+	@echo "==Compilation:"
+	@echo " make all      => compile the executable and the lang files"
+	@echo " make lang     => compile the lang files only"
+	@echo "==Clean up:"
+	@echo " make clean    => remove the files generated during compillation"
+	@echo " make mrproper => make clean + remove the files gernerated during configuration"
+	@echo "==Run after compilation:"
+	@echo " make run      => execute openttd after the compilation"
+	@echo " make run-gdb  => execute openttd in debug mode after the compilation"
+	@echo " make run-prof => execute openttd in profiling mode after the compilation"
+	@echo "==Installation:"
+	@echo " install       => install the compiled files and the data-files after the compilation"
+	@echo " bundle        => create all installation bundles"
+	@echo " bundle_zip    => create the zip installation bundle"
+	@echo " bundle_gzip   => create the gzip installation bundle"
+	@echo " bundle_bzip2  => create the bzip2 installation bundle"
+	@echo " bundle_lha    => create the lha installation bundle"
+	@echo " bundle_dmg    => create the dmg installation bundle"
+	@echo "----------------"
+
 all: config.cache
 ifdef DISTCC
 	@if [ -z "`echo '$(MFLAGS)' | grep '\-j'`" ]; then echo; echo "WARNING: you enabled distcc support, but you don't seem to be using the -jN paramter"; echo; fi
