Index: C:/Documents and Settings/ac4/Desktop/openttd/src/rail_gui.cpp
===================================================================
--- C:/Documents and Settings/ac4/Desktop/openttd/src/rail_gui.cpp	(revision 22817)
+++ C:/Documents and Settings/ac4/Desktop/openttd/src/rail_gui.cpp	(working copy)
@@ -385,6 +385,9 @@
 		GenericPlaceSignals(TileVirtXY(_thd.selend.x, _thd.selend.y));
 		return;
 	}
+	
+	/* Ensure that the spacing for removing signals is always 1. */
+	byte drag_signals_density = _remove_button_clicked ? 1 : _settings_client.gui.drag_signals_density;
 
 	const Window *w = FindWindowById(WC_BUILD_SIGNAL, 0);
 
@@ -394,13 +397,13 @@
 		SB(p2,  4, 1, _cur_signal_variant);
 		SB(p2,  6, 1, _ctrl_pressed);
 		SB(p2,  7, 3, _cur_signal_type);
-		SB(p2, 24, 8, _settings_client.gui.drag_signals_density);
+		SB(p2, 24, 8, drag_signals_density);
 	} else {
 		SB(p2,  3, 1, 0);
 		SB(p2,  4, 1, (_cur_year < _settings_client.gui.semaphore_build_before ? SIG_SEMAPHORE : SIG_ELECTRIC));
 		SB(p2,  6, 1, _ctrl_pressed);
 		SB(p2,  7, 3, _default_signal_type[_settings_client.gui.default_signal_type]);
-		SB(p2, 24, 8, _settings_client.gui.drag_signals_density);
+		SB(p2, 24, 8, drag_signals_density);
 	}
 
 	/* _settings_client.gui.drag_signals_density is given as a parameter such that each user
