Index: projects/generate_vs80.vcproj
===================================================================
--- projects/generate_vs80.vcproj	(revision 0)
+++ projects/generate_vs80.vcproj	(revision 0)
@@ -0,0 +1,40 @@
+<?xml version="1.0" encoding="Windows-1252"?>
+<VisualStudioProject
+	ProjectType="Visual C++"
+	Version="8.00"
+	Name="generate"
+	ProjectGUID="{2F31FD79-D1AC-43C4-89F3-B0D5E4E53E34}"
+	RootNamespace="generate"
+	>
+	<Platforms>
+		<Platform
+			Name="Win32"
+		/>
+	</Platforms>
+	<Configurations>
+		<Configuration
+			Name="Debug|Win32"
+			OutputDirectory="$(SolutionDir)"
+			IntermediateDirectory="..\objs\generate"
+			ConfigurationType="10"
+			>
+		</Configuration>
+	</Configurations>
+	<Files>
+		<File
+			RelativePath="..\source.list"
+			>
+			<FileConfiguration
+				Name="Debug|Win32"
+				>
+				<Tool
+					Name="VCCustomBuildTool"
+					Description="Running 'generate.vbs'..."
+					CommandLine="cscript &quot;$(ProjectDir)generate.vbs&quot;"
+					AdditionalDependencies="$(InputPath)"
+					Outputs="$(SolutionDir)openttd_vs80.vcproj;$(SolutionDir)openttd_vs90.vcproj"
+				/>
+			</FileConfiguration>
+		</File>
+	</Files>
+</VisualStudioProject>
Index: projects/generate_vs90.vcproj
===================================================================
--- projects/generate_vs90.vcproj	(revision 0)
+++ projects/generate_vs90.vcproj	(revision 0)
@@ -0,0 +1,41 @@
+<?xml version="1.0" encoding="Windows-1252"?>
+<VisualStudioProject
+	ProjectType="Visual C++"
+	Version="9.00"
+	Name="generate"
+	ProjectGUID="{2F31FD79-D1AC-43C4-89F3-B0D5E4E53E34}"
+	RootNamespace="generate"
+	TargetFrameworkVersion="131072"
+	>
+	<Platforms>
+		<Platform
+			Name="Win32"
+		/>
+	</Platforms>
+	<Configurations>
+		<Configuration
+			Name="Debug|Win32"
+			OutputDirectory="$(SolutionDir)"
+			IntermediateDirectory="..\objs\generate"
+			ConfigurationType="10"
+			>
+		</Configuration>
+	</Configurations>
+	<Files>
+		<File
+			RelativePath="..\source.list"
+			>
+			<FileConfiguration
+				Name="Debug|Win32"
+				>
+				<Tool
+					Name="VCCustomBuildTool"
+					Description="Running 'generate.vbs' ..."
+					CommandLine="cscript &quot;$(ProjectDir)generate.vbs&quot;"
+					AdditionalDependencies="$(InputPath)"
+					Outputs="$(SolutionDir)openttd_vs80.vcproj;$(SolutionDir)openttd_vs90.vcproj"
+				/>
+			</FileConfiguration>
+		</File>
+	</Files>
+</VisualStudioProject>
Index: projects/openttd_vs80.sln
===================================================================
--- projects/openttd_vs80.sln	(revision 19304)
+++ projects/openttd_vs80.sln	(working copy)
@@ -16,6 +16,8 @@
 EndProject
 Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "version", "version_vs80.vcproj", "{1A2B3C5E-1C23-41A5-9C9B-ACBA2AA75FEC}"
 EndProject
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "generate", "generate_vs80.vcproj", "{2F31FD79-D1AC-43C4-89F3-B0D5E4E53E34}"
+EndProject
 Global
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
 		Debug|Win32 = Debug|Win32
@@ -56,6 +58,10 @@
 		{1A2B3C5E-1C23-41A5-9C9B-ACBA2AA75FEC}.Release|Win32.Build.0 = Debug|Win32
 		{1A2B3C5E-1C23-41A5-9C9B-ACBA2AA75FEC}.Release|x64.ActiveCfg = Debug|Win32
 		{1A2B3C5E-1C23-41A5-9C9B-ACBA2AA75FEC}.Release|x64.Build.0 = Debug|Win32
+		{2F31FD79-D1AC-43C4-89F3-B0D5E4E53E34}.Debug|Win32.ActiveCfg = Debug|Win32
+		{2F31FD79-D1AC-43C4-89F3-B0D5E4E53E34}.Debug|x64.ActiveCfg = Debug|Win32
+		{2F31FD79-D1AC-43C4-89F3-B0D5E4E53E34}.Release|Win32.ActiveCfg = Debug|Win32
+		{2F31FD79-D1AC-43C4-89F3-B0D5E4E53E34}.Release|x64.ActiveCfg = Debug|Win32
 	EndGlobalSection
 	GlobalSection(SolutionProperties) = preSolution
 		HideSolutionNode = FALSE
Index: projects/openttd_vs90.sln
===================================================================
--- projects/openttd_vs90.sln	(revision 19304)
+++ projects/openttd_vs90.sln	(working copy)
@@ -16,6 +16,8 @@
 EndProject
 Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "version", "version_vs90.vcproj", "{1A2B3C5E-1C23-41A5-9C9B-ACBA2AA75FEC}"
 EndProject
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "generate", "generate_vs90.vcproj", "{2F31FD79-D1AC-43C4-89F3-B0D5E4E53E34}"
+EndProject
 Global
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
 		Debug|Win32 = Debug|Win32
@@ -56,6 +58,10 @@
 		{1A2B3C5E-1C23-41A5-9C9B-ACBA2AA75FEC}.Release|Win32.Build.0 = Debug|Win32
 		{1A2B3C5E-1C23-41A5-9C9B-ACBA2AA75FEC}.Release|x64.ActiveCfg = Debug|Win32
 		{1A2B3C5E-1C23-41A5-9C9B-ACBA2AA75FEC}.Release|x64.Build.0 = Debug|Win32
+		{2F31FD79-D1AC-43C4-89F3-B0D5E4E53E34}.Debug|Win32.ActiveCfg = Debug|Win32
+		{2F31FD79-D1AC-43C4-89F3-B0D5E4E53E34}.Debug|x64.ActiveCfg = Debug|Win32
+		{2F31FD79-D1AC-43C4-89F3-B0D5E4E53E34}.Release|Win32.ActiveCfg = Debug|Win32
+		{2F31FD79-D1AC-43C4-89F3-B0D5E4E53E34}.Release|x64.ActiveCfg = Debug|Win32
 	EndGlobalSection
 	GlobalSection(SolutionProperties) = preSolution
 		HideSolutionNode = FALSE
