Index: src/airport_gui.cpp
===================================================================
--- src/airport_gui.cpp	(wersja 23123)
+++ src/airport_gui.cpp	(kopia robocza)
@@ -118,7 +118,7 @@
 	{
 		switch (this->last_user_action) {
 			case ATW_AIRPORT:
-				PlaceAirport(tile);
+				PlaceAirport(TileVirtXY(_thd.pos.x, _thd.pos.y));
 				break;
 
 			case ATW_DEMOLISH:
Index: src/terraform_gui.cpp
===================================================================
--- src/terraform_gui.cpp	(wersja 23123)
+++ src/terraform_gui.cpp	(kopia robocza)
@@ -274,7 +274,7 @@
 				break;
 
 			case TTW_PLACE_OBJECT: // Place object button
-				PlaceProc_Object(tile);
+				PlaceProc_Object(TileVirtXY(_thd.pos.x, _thd.pos.y));
 				break;
 
 			default: NOT_REACHED();
@@ -707,11 +707,11 @@
 				break;
 
 			case ETTW_LOWER_LAND: // Lower land button
-				CommonRaiseLowerBigLand(tile, 0);
+				CommonRaiseLowerBigLand(TileVirtXY(_thd.pos.x, _thd.pos.y), 0);
 				break;
 
 			case ETTW_RAISE_LAND: // Raise land button
-				CommonRaiseLowerBigLand(tile, 1);
+				CommonRaiseLowerBigLand(TileVirtXY(_thd.pos.x, _thd.pos.y), 1);
 				break;
 
 			case ETTW_LEVEL_LAND: // Level land button
@@ -727,7 +727,7 @@
 				break;
 
 			case ETTW_PLACE_OBJECT: // Place transmitter button
-				PlaceProc_Object(tile);
+				PlaceProc_Object(TileVirtXY(_thd.pos.x, _thd.pos.y));
 				break;
 
 			default: NOT_REACHED();
Index: src/rail_gui.cpp
===================================================================
--- src/rail_gui.cpp	(wersja 23123)
+++ src/rail_gui.cpp	(kopia robocza)
@@ -646,7 +646,7 @@
 				break;
 
 			case RTW_BUILD_STATION:
-				PlaceRail_Station(tile);
+				PlaceRail_Station(TileVirtXY(_thd.pos.x, _thd.pos.y));
 				break;
 
 			case RTW_BUILD_SIGNALS:
Index: src/company_gui.cpp
===================================================================
--- src/company_gui.cpp	(wersja 23123)
+++ src/company_gui.cpp	(kopia robocza)
@@ -2054,7 +2054,7 @@
 
 	virtual void OnPlaceObject(Point pt, TileIndex tile)
 	{
-		if (DoCommandP(tile, OBJECT_HQ, 0, CMD_BUILD_OBJECT | CMD_MSG(STR_ERROR_CAN_T_BUILD_COMPANY_HEADQUARTERS))) {
+		if (DoCommandP(TileVirtXY(_thd.pos.x, _thd.pos.y), OBJECT_HQ, 0, CMD_BUILD_OBJECT | CMD_MSG(STR_ERROR_CAN_T_BUILD_COMPANY_HEADQUARTERS))) {
 			ResetObjectToPlace();
 			this->RaiseButtons();
 		}
Index: src/dock_gui.cpp
===================================================================
--- src/dock_gui.cpp	(wersja 23123)
+++ src/dock_gui.cpp	(kopia robocza)
@@ -202,7 +202,7 @@
 				break;
 
 			case DTW_DEPOT: // Build depot button
-				DoCommandP(tile, _ship_depot_direction, 0, CMD_BUILD_SHIP_DEPOT | CMD_MSG(STR_ERROR_CAN_T_BUILD_SHIP_DEPOT), CcBuildDocks);
+				DoCommandP(TileVirtXY(_thd.pos.x, _thd.pos.y), _ship_depot_direction, 0, CMD_BUILD_SHIP_DEPOT | CMD_MSG(STR_ERROR_CAN_T_BUILD_SHIP_DEPOT), CcBuildDocks);
 				break;
 
 			case DTW_STATION: { // Build station button

