Index: src/core/smallvec_type.hpp
===================================================================
--- src/core/smallvec_type.hpp	(revision 15067)
+++ src/core/smallvec_type.hpp	(working copy)
@@ -28,6 +28,14 @@
 public:
 	SmallVector() : data(NULL), items(0), capacity(0) { }
 
+	SmallVector(SmallVector &copy) :
+		items(copy.items),
+		capacity(copy.capacity)
+	{
+		this->data = MallocT<T>(this->capacity);
+		memcpy(this->data, copy.data, sizeof(T) * this->items);
+	}
+
 	~SmallVector()
 	{
 		free(this->data);
