diff -r ec8bc8f89539 src/3rdparty/squirrel/include/squirrel.h
--- a/src/3rdparty/squirrel/include/squirrel.h	Sat Apr 13 17:45:14 2013 +0000
+++ b/src/3rdparty/squirrel/include/squirrel.h	Sun Apr 14 19:25:55 2013 +0200
@@ -35,9 +35,9 @@
 extern "C" {
 #endif
 
-#if defined(_MSC_VER)
+#if defined(_MSC_VER) && _MSC_VER < 1700 // Visual Studio 2012 is stricter with keyword redefinition; it also deprecates __forceinline
 # define inline __forceinline
-#endif /* _MSC_VER */
+#endif /* _MSC_VER < 1700 */
 
 #if defined(_MSC_VER) && _MSC_VER >= 1400 // MSVC 2005 safety checks
 # pragma warning(disable: 4996)   // '_wfopen' was declared deprecated
diff -r ec8bc8f89539 src/stdafx.h
--- a/src/stdafx.h	Sat Apr 13 17:45:14 2013 +0000
+++ b/src/stdafx.h	Sun Apr 14 19:25:55 2013 +0200
@@ -208,7 +208,9 @@
 
 	#include <malloc.h> // alloca()
 	#define NORETURN __declspec(noreturn)
-	#define inline __forceinline
+	#if (_MSC_VER < 1700) // Visual Studio 2012 is stricter with keyword redefinition; it also deprecates __forceinline
+		#define inline __forceinline
+	#endif  /* (_MSC_VER < 1700) */
 
 	#if !defined(WINCE)
 		#define CDECL _cdecl
