Index: industry_cmd.cpp
===================================================================
--- industry_cmd.cpp	(revision 9817)
+++ industry_cmd.cpp	(working copy)
@@ -1642,13 +1642,18 @@
 				uint32 r = Random();
 				int old_prod, new_prod, percent;
 				int mag;
+				int influence = (i->pct_transported[j] > 153) ? 1 : -1;
 
 				new_prod = old_prod = i->production_rate[j];
-				if (CHANCE16I(20, 1024, r))
-					new_prod -= ((RandomRange(50) + 10) * old_prod) >> 8;
-				if (CHANCE16I(20 + (i->pct_transported[j] * 20 >> 8), 1024, r >> 16))
-					new_prod += ((RandomRange(50) + 10) * old_prod) >> 8;
 
+				/* somewhat exponential increases - chance = up to 3% if more than 60% transported */
+				if (CHANCE16I((2 + influence), 100, r))
+					new_prod += ((old_prod / 3) - (old_prod / 6) - RandomRange((old_prod / 12) + 1) + (old_prod >> 4) - (old_prod >> 5) - (old_prod >> 6) - (old_prod >> 7));
+				
+				/* somewhat logarithmical decreases - chance = up to 3% if less than 60% transported */
+				if (CHANCE16I((2 - influence), 100, r >> 16))
+					new_prod -= (RandomRange((old_prod >> 2) + 1) - (old_prod >> 3) + (old_prod >> 4) + (old_prod >> 5) + (old_prod >> 6) + (old_prod >> 7));
+				
 				new_prod = clamp(new_prod, 0, 255);
 				if (new_prod == old_prod) {
 					closeit = false;
