Index: bin/ai/regression/regression.nut
===================================================================
--- bin/ai/regression/regression.nut	(revision 18973)
+++ bin/ai/regression/regression.nut	(working copy)
@@ -1494,6 +1494,7 @@
 {
 	print("");
 	print("--Tunnel--");
+	AIBridge.BuildBridge(AIVehicle.VT_ROAD, 1, 0x727A, 0x7278);
 	print("  IsTunnelTile():       " + AITunnel.IsTunnelTile(29050));
 	print("  RemoveTunnel():       " + AITunnel.RemoveTunnel(29050));
 	print("  GetOtherTunnelEnd():  " + AITunnel.GetOtherTunnelEnd(29050));
