Index: src/station_cmd.cpp
===================================================================
--- src/station_cmd.cpp	(revision 12875)
+++ src/station_cmd.cpp	(working copy)
@@ -2764,7 +2764,7 @@
 		Station *st = *st_iter;
 
 		/* Is the station reserved exclusively for somebody else? */
-		if (st->town->exclusive_counter > 0 && st->town->exclusivity != st->owner) continue;
+		if (st->town->exclusive_counter > 0 && st->town->exclusivity != st->owner && st->owner != OWNER_NONE) continue;
 
 		if (st->goods[type].rating == 0) continue; // Lowest possible rating, better not to give cargo anymore
 
