Index: src/settings_gui.cpp
===================================================================
--- src/settings_gui.cpp	(revision 12176)
+++ src/settings_gui.cpp	(working copy)
@@ -949,7 +949,7 @@
 						SetDParam(1, value);
 					}
 				}
-				DrawString(30, y, (sdb->str) + disabled, TC_FROMSTRING);
+				DrawStringTruncated(30, y, (sdb->str) + disabled, TC_FROMSTRING, w->width - 30);
 				y += 11;
 			}
 		} break;
