Index: src/rail_cmd.cpp
===================================================================
--- src/rail_cmd.cpp	(revision 27653)
+++ src/rail_cmd.cpp	(working copy)
@@ -1595,7 +1595,7 @@
 		 * Tunnels and bridges have special check later */
 		if (tt != MP_TUNNELBRIDGE) {
 			if (!IsCompatibleRail(type, totype)) {
-				CommandCost ret = EnsureNoVehicleOnGround(tile);
+				CommandCost ret = EnsureNoTrainOnTrackBits(tile, GetTrackBits(tile));
 				if (ret.Failed()) {
 					error = ret;
 					continue;
