diff -r c7487e103f60 src/os/macosx/osx_stdafx.h
--- a/src/os/macosx/osx_stdafx.h	Fri Sep 18 17:45:12 2009 +0000
+++ b/src/os/macosx/osx_stdafx.h	Thu Sep 24 08:24:34 2009 +0200
@@ -26,7 +26,7 @@
 #endif
 
 /* Check for mismatching 'architectures' */
-#if (__LP64__ && !defined(_SQ64)) || (!__LP64__ && defined(_SQ64))
+#if ((__LP64__ && !defined(_SQ64)) || (!__LP64__ && defined(_SQ64))) && (!STRGEN)
 #	error "Compiling 64 bits without _SQ64 set! (or vice versa)"
 #endif
 
