Index: src/waypoint_cmd.cpp
===================================================================
--- src/waypoint_cmd.cpp	(revision 27653)
+++ src/waypoint_cmd.cpp	(working copy)
@@ -297,12 +297,15 @@
 	Waypoint *wp = FindDeletedWaypointCloseTo(tile, STR_SV_STNAME_BUOY, OWNER_NONE);
 	if (wp == NULL && !Waypoint::CanAllocateItem()) return_cmd_error(STR_ERROR_TOO_MANY_STATIONS_LOADING);
 
+	CommandCost ret;
 	CommandCost cost(EXPENSES_CONSTRUCTION, _price[PR_BUILD_WAYPOINT_BUOY]);
 	if (!IsWaterTile(tile)) {
-		CommandCost ret = DoCommand(tile, 0, 0, flags | DC_AUTO, CMD_LANDSCAPE_CLEAR);
+		ret = DoCommand(tile, 0, 0, flags | DC_AUTO, CMD_LANDSCAPE_CLEAR);
 		if (ret.Failed()) return ret;
 		cost.AddCost(ret);
 	}
+	ret = EnsureNoVehicleOnGround(tile);
+	if (ret.Failed()) return ret;
 
 	if (flags & DC_EXEC) {
 		if (wp == NULL) {
