Index: src/console_gui.cpp
===================================================================
--- src/console_gui.cpp	(revision 16026)
+++ src/console_gui.cpp	(working copy)
@@ -332,7 +332,6 @@
 	IConsolePrint(CC_WHITE,  "use \"help\" for more information");
 	IConsolePrint(CC_WHITE,  "");
 	IConsoleClearCommand();
-	IConsoleHistoryAdd("");
 }
 
 void IConsoleClearBuffer()
@@ -388,6 +387,11 @@
  */
 static void IConsoleHistoryAdd(const char *cmd)
 {
+	//Do not put empty command in history
+	if (StrEmpty(cmd)) return;
+	//Do not put in history if command is same as previous
+	if (_iconsole_history[0] && strcmp(_iconsole_history[0],cmd)==0) return;
+
 	free(_iconsole_history[ICON_HISTORY_SIZE - 1]);
 
 	memmove(&_iconsole_history[1], &_iconsole_history[0], sizeof(_iconsole_history[0]) * (ICON_HISTORY_SIZE - 1));
@@ -401,6 +405,7 @@
  */
 static void IConsoleHistoryNavigate(int direction)
 {
+	if (!_iconsole_history[0]) return; //Empty history
 	int i = _iconsole_historypos + direction;
 
 	/* watch out for overflows, just wrap around */
