Index: saveload.c
===================================================================
--- saveload.c	(revision 3308)
+++ saveload.c	(working copy)
@@ -1232,7 +1232,7 @@
 }
 
 // actual loader/saver function
-void InitializeGame(uint size_x, uint size_y);
+void InitializeGame(int mode, uint size_x, uint size_y);
 extern bool AfterLoadGame(uint version);
 extern void BeforeSaveGame(void);
 extern bool LoadOldSaveGame(const char *file);
@@ -1363,7 +1363,7 @@
 
 	/* Load a TTDLX or TTDPatch game */
 	if (mode == SL_OLD_LOAD) {
-		InitializeGame(256, 256); // set a mapsize of 256x256 for TTDPatch games or it might get confused
+		InitializeGame(0, 256, 256); // set a mapsize of 256x256 for TTDPatch games or it might get confused
 		if (!LoadOldSaveGame(filename)) return SL_REINIT;
 		AfterLoadGame(0);
 		return SL_OK;
@@ -1486,7 +1486,7 @@
 		/* Old maps were hardcoded to 256x256 and thus did not contain
 		 * any mapsize information. Pre-initialize to 256x256 to not to
 		 * confuse old games */
-		InitializeGame(256, 256);
+		InitializeGame(0, 256, 256);
 
 		SlLoadChunks();
 		fmt->uninit_read();
Index: misc.c
===================================================================
--- misc.c	(revision 3308)
+++ misc.c	(working copy)
@@ -122,7 +122,7 @@
 
 void ConvertGroundTilesIntoWaterTiles(void);
 
-void InitializeGame(uint size_x, uint size_y)
+void InitializeGame(int mode, uint size_x, uint size_y)
 {
 	AllocateMap(size_x, size_y);
 
@@ -136,7 +136,7 @@
 	_date_fract = 0;
 	_cur_tileloop_tile = 0;
 
-	{
+	if (mode != 2) { //if this is not creating random land in the editor...
 		uint starting = ConvertIntDate(_patches.starting_date);
 		if ( starting == (uint)-1) starting = 10958;
 		SetDate(starting);
@@ -189,7 +189,7 @@
 	_current_player = OWNER_NONE;
 
 	_generating_world = true;
-	InitializeGame(size_x, size_y);
+	InitializeGame(mode, size_x, size_y);
 	SetObjectToPlace(SPR_CURSOR_ZZZ, 0, 0, 0);
 
 	// Must start economy early because of the costs.
