Index: Makefile
===================================================================
--- Makefile	(revision 3285)
+++ Makefile	(working copy)
@@ -538,9 +538,9 @@
 		SECOND_DATA_PATH:="$(OSXAPP)/Contents/Data/"
 	endif
 
-	ifndef CUSTOM_LANG_DIR
+	ifndef CUSTOM_LANG_PATH
 		ifndef DEDICATED
-		CUSTOM_LANG_DIR:="$(OSXAPP)/Contents/Lang/"
+		CUSTOM_LANG_PATH:="$(OSXAPP)/Contents/Lang/"
 		endif
 	endif
 endif
@@ -568,8 +568,8 @@
 CDEFS += -DSECOND_DATA_DIR=\"$(SECOND_DATA_PATH)/\"
 endif
 
-ifdef CUSTOM_LANG_DIR
-CDEFS += -DCUSTOM_LANG_DIR=\"$(CUSTOM_LANG_DIR)/\"
+ifdef CUSTOM_LANG_PATH
+CDEFS += -DCUSTOM_LANG_DIR=\"$(CUSTOM_LANG_PATH)/\"
 endif
 
 ifdef WITH_DIRECTMUSIC
