Index: src/console.cpp
===================================================================
--- src/console.cpp	(révision 14450)
+++ src/console.cpp	(copie de travail)
@@ -13,6 +13,7 @@
 #include "rev.h"
 
 #include <stdarg.h>
+#include <time.h>
 
 #include "table/strings.h"
 
@@ -104,7 +105,16 @@
 	str_validate(str);
 
 	if (_network_dedicated) {
-		fprintf(stdout, "%s\n", str);
+		// Quelle heure qu'il est il maintenant ?
+		char l_out_time[32];
+		time_t l_t;
+		struct tm *l_tmp;
+
+		l_t = time(NULL);
+		l_tmp = localtime(&l_t);
+		strftime( l_out_time, sizeof(l_out_time), "%Y%m%d %H:%M:%S", l_tmp );
+	
+		fprintf(stdout, "%s : %s\n", l_out_time, str);
 		fflush(stdout);
 		IConsoleWriteToLogFile(str);
 		free(str); // free duplicated string since it's not used anymore
Index: src/debug.cpp
===================================================================
--- src/debug.cpp	(révision 14450)
+++ src/debug.cpp	(copie de travail)
@@ -3,6 +3,7 @@
 /** @file debug.cpp Handling of printing debug messages. */
 
 #include "stdafx.h"
+#include <time.h>
 #include <stdio.h>
 #include <stdarg.h>
 #include "openttd.h"
@@ -63,6 +64,15 @@
 
 static void debug_print(const char *dbg, const char *buf)
 {
+        char out_time[32];
+        time_t l_time;
+        struct tm *l_tmp;
+
+        // What time is it ?
+        l_time = time(NULL);
+        l_tmp = localtime(&l_time);
+        strftime( out_time, sizeof(out_time), "%Y%m%d %H:%M:%S", l_tmp );
+
 #if defined(ENABLE_NETWORK)
 	if (_debug_socket != INVALID_SOCKET) {
 		char buf2[1024 + 32];
@@ -78,7 +88,7 @@
 		_sntprintf(tbuf, sizeof(tbuf), _T("%s"), OTTD2FS(dbg));
 		NKDbgPrintfW(_T("dbg: [%s] %s\n"), tbuf, OTTD2FS(buf));
 #else
-		fprintf(stderr, "dbg: [%s] %s\n", dbg, buf);
+		fprintf(stderr, "%s dbg: [%s] %s\n", out_time, dbg, buf);
 #endif
 		IConsoleDebug(dbg, buf);
 	}
@@ -86,6 +96,7 @@
 
 void CDECL debug(const char *dbg, ...)
 {
+
 	va_list va;
 	va_start(va, dbg);
 	const char *s;
