Index: src/station_gui.cpp
===================================================================
--- src/station_gui.cpp	(Revision 10606)
+++ src/station_gui.cpp	(Arbeitskopie)
@@ -779,7 +779,7 @@
 					*b++ = ',';
 					*b++ = ' ';
 				}
-				b = InlineString(b, GetCargo(i)->name);
+				b = InlineString(b, GetCargo(i)->name_plural);
 			}
 		}
 
Index: src/misc_gui.cpp
===================================================================
--- src/misc_gui.cpp	(Revision 10606)
+++ src/misc_gui.cpp	(Arbeitskopie)
@@ -775,7 +775,7 @@
 				*b++ = ',';
 				*b++ = ' ';
 			}
-			b = InlineString(b, GetCargo(i)->name);
+			b = InlineString(b, GetCargo(i)->name_plural);
 		}
 	}
 
