Index: src/openttd.cpp
===================================================================
--- src/openttd.cpp	(revision 9518)
+++ src/openttd.cpp	(working copy)
@@ -684,6 +684,9 @@
 
 	GenerateWorldSetCallback(&MakeNewGameDone);
 	GenerateWorld(from_heightmap ? GW_HEIGHTMAP : GW_NEWGAME, 1 << _patches.map_x, 1 << _patches.map_y);
+	if(_patches.start_paused && !_network_dedicated && !_network_server && !_networking) {
+		_pause_game = 1;
+	}
 }
 
 static void MakeNewEditorWorldDone()
@@ -752,6 +755,7 @@
 	DoCommandP(0, (_patches.autorenew << 15 ) | (_patches.autorenew_months << 16) | 4, _patches.autorenew_money, NULL, CMD_SET_AUTOREPLACE);
 
 	MarkWholeScreenDirty();
+
 }
 
 bool SafeSaveOrLoad(const char *filename, int mode, int newgm)
Index: src/settings.cpp
===================================================================
--- src/settings.cpp	(revision 9518)
+++ src/settings.cpp	(working copy)
@@ -1400,7 +1400,9 @@
 	 SDT_VAR(Patches, dist_local_authority,SLE_UINT8, 0, 0, 20, 5,  60, 0, STR_NULL, NULL),
 	 SDT_VAR(Patches, wait_oneway_signal,  SLE_UINT8, 0, 0, 15, 2, 100, 0, STR_NULL, NULL),
 	 SDT_VAR(Patches, wait_twoway_signal,  SLE_UINT8, 0, 0, 41, 2, 100, 0, STR_NULL, NULL),
+	SDT_BOOL(Patches, start_paused,					  S, 0, false,		   STR_NULL, NULL),
 
+
 	/***************************************************************************/
 	/* New Pathfinding patch settings */
 	SDT_VAR(Patches, pf_maxlength,      SLE_UINT16, 0, 0,  4096,  64,  65535, 0, STR_NULL, NULL),
Index: src/variables.h
===================================================================
--- src/variables.h	(revision 9518)
+++ src/variables.h	(working copy)
@@ -218,6 +218,8 @@
 
 	uint8 freight_trains; ///< Value to multiply the weight of cargo by
 
+	bool start_paused;					   // Start the game paused or not 
+
 	/** YAPF settings */
 	YapfSettings  yapf;
 
@@ -290,6 +292,7 @@
 VARDEF uint32 _decode_parameters[20];
 
 VARDEF bool _rightclick_emulate;
+VARDEF bool _start_paused;
 
 // IN/OUT parameters to commands
 VARDEF byte _yearly_expenses_type;
