diff -r 9c799a75b273 src/vehicle_gui.cpp
--- a/src/vehicle_gui.cpp
+++ b/src/vehicle_gui.cpp
@@ -776,7 +776,7 @@
 	EndContainer(),
 
 	NWidget(NWID_HORIZONTAL),
-		NWidget(WWT_MATRIX, COLOUR_GREY, VLW_WIDGET_LIST), SetMinimalSize(248, 0), SetFill(1, 0),
+		NWidget(WWT_MATRIX, COLOUR_GREY, VLW_WIDGET_LIST), SetMinimalSize(248, 0), SetFill(1, 0), SetResize(1, 0),
 		NWidget(WWT_SCROLLBAR, COLOUR_GREY, VLW_WIDGET_SCROLLBAR),
 	EndContainer(),
 
@@ -1001,13 +1001,10 @@
 	{
 		if (widget != VLW_WIDGET_LIST) return;
 
-		resize->width = 0;
 		resize->height = GetVehicleListHeight(this->vehicle_type, 1);
 
 		switch (this->vehicle_type) {
 			case VEH_TRAIN:
-				resize->width = 1;
-				/* Fallthrough */
 			case VEH_ROAD:
 				size->height = 6 * resize->height;
 				break;
