diff -r 15d1b5f7c9e3 src/lang/english.txt
--- a/src/lang/english.txt	Mon Nov 17 21:57:17 2008 +0100
+++ b/src/lang/english.txt	Sun Nov 23 12:08:39 2008 +0100
@@ -1212,12 +1212,6 @@
 STR_CONFIG_PATCHES_CITY_SIZE_MULTIPLIER                         :{LTBLUE}Initial city size multiplier: {ORANGE}{STRING1}
 STR_CONFIG_MODIFIED_ROAD_REBUILD                                :{LTBLUE}Remove absurd road-elements during the road construction
 
-STR_CONFIG_PATCHES_GUI                                          :{BLACK}Interface
-STR_CONFIG_PATCHES_CONSTRUCTION                                 :{BLACK}Construction
-STR_CONFIG_PATCHES_VEHICLES                                     :{BLACK}Vehicles
-STR_CONFIG_PATCHES_STATIONS                                     :{BLACK}Stations
-STR_CONFIG_PATCHES_ECONOMY                                      :{BLACK}Economy
-STR_CONFIG_PATCHES_AI                                           :{BLACK}Competitors
 STR_CONFIG_PATCHES_TITLE_SIGNALS                                :{ORANGE}Signals
 
 STR_CONFIG_PATCHES_DISABLED                                     :disabled
diff -r 15d1b5f7c9e3 src/settings_gui.cpp
--- a/src/settings_gui.cpp	Mon Nov 17 21:57:17 2008 +0100
+++ b/src/settings_gui.cpp	Sun Nov 23 12:08:39 2008 +0100
@@ -1126,15 +1126,9 @@
 
 /** Widget numbers of config patches window */
 enum PatchesSelectionWidgets {
-	PATCHSEL_OPTIONSPANEL = 3, ///< Panel widget containing the option lists
+	PATCHSEL_OPTIONSPANEL = 2, ///< Panel widget containing the option lists
 	PATCHSEL_SCROLLBAR,        ///< Scrollbar
 	PATCHSEL_RESIZE,           ///< Resize button
-	PATCHSEL_INTERFACE,        ///< Button 'Interface'
-	PATCHSEL_CONSTRUCTION,     ///< Button 'Construction'
-	PATCHSEL_VEHICLES,         ///< Button 'Vehicles'
-	PATCHSEL_STATIONS,         ///< Button 'Stations'
-	PATCHSEL_ECONOMY,          ///< Button 'Economy'
-	PATCHSEL_COMPETITORS       ///< Button 'Competitors'
 };
 
 struct PatchesSelectionWindow : Window {
@@ -1149,7 +1143,7 @@
 
 	PatchesSelectionWindow(const WindowDesc *desc) : Window(desc)
 	{
-		/* Check that the widget doesn't get moved without adapting the constant as well.
+		/* Check that the widget doesn't get moved without adapting the constants as well.
 		 *  - SETTINGTREE_LEFT_OFFSET should be 5 pixels to the right of the left edge of the panel
 		 *  - SETTINGTREE_TOP_OFFSET should be 5 pixels below the top edge of the panel
 		 */
@@ -1179,8 +1173,6 @@
 		this->resize.height = this->height;
 		this->resize.step_width = 1;
 		this->resize.width = this->width;
-
-		this->LowerWidget(page + PATCHSEL_INTERFACE); // Depress button of currently selected page
 
 		this->FindWindowPlacementAndResize(desc);
 	}
@@ -1293,20 +1285,6 @@
 					}
 				}
 			} break;
-
-			case PATCHSEL_INTERFACE: case PATCHSEL_CONSTRUCTION: case PATCHSEL_VEHICLES:
-			case PATCHSEL_STATIONS:  case PATCHSEL_ECONOMY:      case PATCHSEL_COMPETITORS:
-				if (depressed_entry != NULL) { // If at the current page an entry has depressed buttons, release them
-					depressed_entry->SetButtons(0);
-					depressed_entry = NULL;
-				}
-				this->RaiseWidget(this->page + PATCHSEL_INTERFACE);
-				this->page = widget - PATCHSEL_INTERFACE;
-				this->LowerWidget(this->page + PATCHSEL_INTERFACE);
-				SetVScrollCount(this, _patches_page[page].Length());
-				DeleteWindowById(WC_QUERY_STRING, 0);
-				this->SetDirty();
-				break;
 		}
 	}
 
@@ -1344,27 +1322,19 @@
 
 GameSettings *PatchesSelectionWindow::patches_ptr = NULL;
 const int PatchesSelectionWindow::SETTINGTREE_LEFT_OFFSET = 5;
-const int PatchesSelectionWindow::SETTINGTREE_TOP_OFFSET = 47;
+const int PatchesSelectionWindow::SETTINGTREE_TOP_OFFSET = 19;
 
 static const Widget _patches_selection_widgets[] = {
 {   WWT_CLOSEBOX,   RESIZE_NONE,  COLOUR_MAUVE,     0,    10,     0,    13, STR_00C5,                        STR_018B_CLOSE_WINDOW},
 {    WWT_CAPTION,   RESIZE_NONE,  COLOUR_MAUVE,    11,   411,     0,    13, STR_CONFIG_PATCHES_CAPTION,      STR_018C_WINDOW_TITLE_DRAG_THIS},
-{      WWT_PANEL,   RESIZE_NONE,  COLOUR_MAUVE,     0,   411,    14,    41, 0x0,                             STR_NULL},
-{      WWT_PANEL, RESIZE_BOTTOM,  COLOUR_MAUVE,     0,   399,    42,   215, 0x0,                             STR_NULL}, // PATCHSEL_OPTIONSPANEL
-{  WWT_SCROLLBAR, RESIZE_BOTTOM,  COLOUR_MAUVE,   400,   411,    42,   203, 0x0,                             STR_0190_SCROLL_BAR_SCROLLS_LIST}, // PATCHSEL_SCROLLBAR
-{  WWT_RESIZEBOX,     RESIZE_TB,  COLOUR_MAUVE,   400,   411,   204,   215, 0x0,                             STR_RESIZE_BUTTON}, // PATCHSEL_RESIZE
-
-{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,   10,   107,    16,    27, STR_CONFIG_PATCHES_GUI,          STR_NULL}, // PATCHSEL_INTERFACE
-{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,  108,   205,    16,    27, STR_CONFIG_PATCHES_CONSTRUCTION, STR_NULL}, // PATCHSEL_CONSTRUCTION
-{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,  206,   303,    16,    27, STR_CONFIG_PATCHES_VEHICLES,     STR_NULL}, // PATCHSEL_VEHICLES
-{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,  304,   401,    16,    27, STR_CONFIG_PATCHES_STATIONS,     STR_NULL}, // PATCHSEL_STATIONS
-{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,   10,   107,    28,    39, STR_CONFIG_PATCHES_ECONOMY,      STR_NULL}, // PATCHSEL_ECONOMY
-{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,  108,   205,    28,    39, STR_CONFIG_PATCHES_AI,           STR_NULL}, // PATCHSEL_COMPETITORS
+{      WWT_PANEL, RESIZE_BOTTOM,  COLOUR_MAUVE,     0,   399,    14,   187, 0x0,                             STR_NULL}, // PATCHSEL_OPTIONSPANEL
+{  WWT_SCROLLBAR, RESIZE_BOTTOM,  COLOUR_MAUVE,   400,   411,    14,   175, 0x0,                             STR_0190_SCROLL_BAR_SCROLLS_LIST}, // PATCHSEL_SCROLLBAR
+{  WWT_RESIZEBOX,     RESIZE_TB,  COLOUR_MAUVE,   400,   411,   176,   187, 0x0,                             STR_RESIZE_BUTTON}, // PATCHSEL_RESIZE
 {   WIDGETS_END},
 };
 
 static const WindowDesc _patches_selection_desc = {
-	WDP_CENTER, WDP_CENTER, 412, 216, 412, 216,
+	WDP_CENTER, WDP_CENTER, 412, 188, 412, 188,
 	WC_GAME_OPTIONS, WC_NONE,
 	WDF_STD_TOOLTIPS | WDF_STD_BTN | WDF_DEF_WIDGET | WDF_RESIZABLE,
 	_patches_selection_widgets,
