Index: src/saveload/town_sl.cpp
===================================================================
--- src/saveload/town_sl.cpp	(revision 24180)
+++ src/saveload/town_sl.cpp	(working copy)
@@ -35,7 +35,7 @@
 	for (TileIndex t = 0; t < MapSize(); t++) {
 		if (!IsTileType(t, MP_HOUSE)) continue;
 
-		HouseID house_id = GetCleanHouseType(t);
+		HouseID house_id = GetHouseType(t);
 		town = Town::GetByTile(t);
 		IncreaseBuildingCount(town, house_id);
 		if (IsHouseCompleted(t)) town->cache.population += HouseSpec::Get(house_id)->population;
Index: src/town_cmd.cpp
===================================================================
--- src/town_cmd.cpp	(revision 24180)
+++ src/town_cmd.cpp	(working copy)
@@ -2304,6 +2304,7 @@
 		}
 
 		MakeTownHouse(tile, t, construction_counter, construction_stage, house, random_bits);
+		UpdateTownRadius(t);
 		UpdateTownCargoes(t, tile);
 
 		return true;
@@ -2387,6 +2388,8 @@
 	if (eflags & BUILDING_2_TILES_X)   DoClearTownHouseHelper(tile + TileDiffXY(1, 0), t, ++house);
 	if (eflags & BUILDING_HAS_4_TILES) DoClearTownHouseHelper(tile + TileDiffXY(1, 1), t, ++house);
 
+	UpdateTownRadius(t);
+
 	/* Update cargo acceptance. */
 	UpdateTownCargoes(t, tile);
 }
