Index: src/driver.h
===================================================================
--- src/driver.h	(r‚vision 13071)
+++ src/driver.h	(copie de travail)
@@ -71,7 +71,16 @@
 	 */
 	virtual ~DriverFactoryBase() {
 		if (this->name == NULL) return;
-		GetDrivers().erase(this->name);
+		/* Prefix the name with driver type to make it unique */
+		char buf[32];
+		strecpy(buf, GetDriverTypeName(type), lastof(buf));
+		strecpy(buf + 5, this->name, lastof(buf));
+		for (Drivers::iterator it = GetDrivers().begin(); it != GetDrivers().end(); it++) {
+			if (it->first == Drivers::key_type(buf)) {
+				GetDrivers().erase(it);
+				break;
+			}
+		}
 		free(this->name);
 	}
 
