Index: src/video/win32_v.cpp
===================================================================
--- src/video/win32_v.cpp	(revision 13571)
+++ src/video/win32_v.cpp	(working copy)
@@ -18,6 +18,9 @@
 #include <windows.h>
 #include <tchar.h>
 
+#define COMPILE_MULTIMON_STUBS
+#include <multimon.h>
+
 static struct {
 	HWND main_wnd;
 	HBITMAP dib_sect;
@@ -280,8 +283,8 @@
 #endif
 		w = r.right - r.left;
 		h = r.bottom - r.top;
-		x = (GetSystemMetrics(SM_CXSCREEN) - w) / 2;
-		y = (GetSystemMetrics(SM_CYSCREEN) - h) / 2;
+		x = ((GetSystemMetrics(SM_CXVIRTUALSCREEN) - w) / 2) - GetSystemMetrics(SM_XVIRTUALSCREEN);
+		y = ((GetSystemMetrics(SM_CYVIRTUALSCREEN) - h) / 2) - GetSystemMetrics(SM_YVIRTUALSCREEN);
 
 		if (_wnd.main_wnd) {
 			ShowWindow(_wnd.main_wnd, SW_SHOWNORMAL); // remove maximize-flag
