Index: src/road_cmd.cpp
===================================================================
--- src/road_cmd.cpp	(revision 11051)
+++ src/road_cmd.cpp	(working copy)
@@ -198,10 +198,12 @@
 				if (present == ROAD_NONE) {
 					RoadTypes rts = GetRoadTypes(tile) & ComplementRoadTypes(RoadTypeToRoadTypes(rt));
 					if (rts == ROADTYPES_NONE) {
+						/* Includes MarkTileDirtyByTile() */
 						DoClearSquare(tile);
 					} else {
 						SetRoadBits(tile, ROAD_NONE, rt);
 						SetRoadTypes(tile, rts);
+						MarkTileDirtyByTile(tile);
 					}
 				} else {
 					/* When bits are removed, you *always* end up with something that
