Index: src/network/network_admin.cpp
===================================================================
--- src/network/network_admin.cpp	(revision 25785)
+++ src/network/network_admin.cpp	(working copy)
@@ -336,7 +336,12 @@
 	p->Send_bool  (NetworkCompanyIsPassworded(c->index));
 	p->Send_uint32(c->inaugurated_year);
 	p->Send_bool  (c->is_ai);
+	p->Send_uint8 (CeilDiv(c->months_of_bankruptcy, 3)); // send as quarters_of_bankruptcy
 
+	for (size_t i = 0; i < lengthof(c->share_owners); i++) {
+		p->Send_uint8(c->share_owners[i]);
+	}
+	
 	this->SendPacket(p);
 
 	return NETWORK_RECV_STATUS_OKAY;
