From 39ac5365628e6e4fd667be83fef550fc05da0a6f Mon Sep 17 00:00:00 2001
From: Chris Angelico <rosuav@gmail.com>
Date: Sun, 16 Sep 2012 00:40:08 +1000
Subject: [PATCH] Increase max loan to company size

---
 src/economy.cpp |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/economy.cpp b/src/economy.cpp
index 66972b8..04def71 100644
--- a/src/economy.cpp
+++ b/src/economy.cpp
@@ -685,7 +685,10 @@ void AddInflation(bool check_year)
 void RecomputePrices()
 {
 	/* Setup maximum loan */
-	_economy.max_loan = (_settings_game.difficulty.max_loan * _economy.inflation_prices >> 16) / 50000 * 50000;
+	_economy.max_loan = _settings_game.difficulty.max_loan * _economy.inflation_prices >> 16;
+	Company *c = Company::GetIfValid(_local_company);
+	if (c) _economy.max_loan = max(_economy.max_loan, CalculateCompanyValue(c, true));
+	_economy.max_loan = _economy.max_loan / 50000 * 50000;
 
 	/* Setup price bases */
 	for (Price i = PR_BEGIN; i < PR_END; i++) {
-- 
1.7.9.GIT

