Index: src/window.cpp
===================================================================
--- src/window.cpp	(revision 21421)
+++ src/window.cpp	(working copy)
@@ -2233,7 +2233,11 @@
 			case MC_RIGHT:
 				if (!(w->flags4 & WF_DISABLE_VP_SCROLL)) {
 					_scrolling_viewport = true;
-					_cursor.fix_at = true;
+					if (_settings_client.gui.reverse_scroll) {
+						_cursor.fix_at = false;
+					} else {
+						_cursor.fix_at = true;
+					}
 
 					/* clear 2D scrolling caches before we start a 2D scroll */
 					_cursor.h_wheel = 0;
