Index: src/train_cmd.cpp
===================================================================
--- src/train_cmd.cpp	(revision 20240)
+++ src/train_cmd.cpp	(working copy)
@@ -2977,7 +2977,7 @@
 {
 	uint pass = 0;
 	if (this->IsFrontEngine()) {
-		pass += 4; // driver
+		pass += 2; // driver
 
 		/* Remove the reserved path in front of the train if it is not stuck.
 		 * Also clear all reserved tracks the train is currently on. */
@@ -3023,7 +3023,7 @@
 	}
 
 	/* Try to re-reserve track under already crashed train too.
-	 * SetVehicleCrashed() clears the reservation! */
+	 * Crash() clears the reservation! */
 	v->ReserveTrackUnderConsist();
 
 	return num;
