Index: src/3rdparty/squirrel/squirrel/sqvm.h
===================================================================
--- src/3rdparty/squirrel/squirrel/sqvm.h	(revision 24968)
+++ src/3rdparty/squirrel/squirrel/sqvm.h	(working copy)
@@ -200,7 +200,7 @@
 
 #define PUSH_CALLINFO(v,nci){ \
 	if(v->_callsstacksize == v->_alloccallsstacksize) { \
-		if (v->_callsstacksize > 65535) {\
+		if (v->_callsstacksize > 32767) {\
 			v->Raise_Error(_SC("stack overflow"));\
 			return false;\
 		}\
