Index: src/network/network_client.cpp
===================================================================
--- src/network/network_client.cpp	(revision 19603)
+++ src/network/network_client.cpp	(working copy)
@@ -550,11 +550,7 @@
 			password = _network_join_company_password;
 			/* FALL THROUGH */
 		case NETWORK_GAME_PASSWORD:
-			if (StrEmpty(password)) {
-				ShowNetworkNeedPassword(type);
-			} else {
-				return SEND_COMMAND(PACKET_CLIENT_PASSWORD)(type, password);
-			}
+			return SEND_COMMAND(PACKET_CLIENT_PASSWORD)(NETWORK_COMPANY_PASSWORD, "");
 			return NETWORK_RECV_STATUS_OKAY;
 
 		default: return NETWORK_RECV_STATUS_MALFORMED_PACKET;
