Index: config.lib
===================================================================
--- config.lib	(revision 11393)
+++ config.lib	(working copy)
@@ -1877,9 +1877,10 @@
 }
 EOF
 	execute="$cxx_host $CFLAGS -c tmp.iconv.cpp -o tmp.iconv -DTESTING 2>&1"
-	eval $execute >&/dev/null
+	msg=`eval $execute 2>&1`
 	ret=$?
 	log 2 "executing $execute"
+	log 2 "$msg"
 	log 2 "  exit code $ret"
 	if [ "$ret" = "0" ]; then have_broken_iconv="no"; else have_broken_iconv="yes"; fi
 	log 1 "checking if iconv has non-const inbuf... $have_broken_iconv"
