diff --git a/src/blitter/32bpp_base.cpp b/src/blitter/32bpp_base.cpp
index cf2a78f..3ea0332 100644
--- a/src/blitter/32bpp_base.cpp
+++ b/src/blitter/32bpp_base.cpp
@@ -107,7 +107,7 @@ void Blitter_32bppBase::ScrollBuffer(void *video, int &left, int &top, int &widt
 		dst = (uint32 *)video + left + top * _screen.pitch;
 		src = dst - scroll_y * _screen.pitch;
 
-		/* Decrese height. (scroll_y is <=0). */
+		/* Decrease height. (scroll_y is <=0). */
 		height += scroll_y;
 		assert(height > 0);
 
diff --git a/src/blitter/8bpp_base.cpp b/src/blitter/8bpp_base.cpp
index 39960f8..f2ba704 100644
--- a/src/blitter/8bpp_base.cpp
+++ b/src/blitter/8bpp_base.cpp
@@ -112,7 +112,7 @@ void Blitter_8bppBase::ScrollBuffer(void *video, int &left, int &top, int &width
 		dst = (uint8 *)video + left + top * _screen.pitch;
 		src = dst - scroll_y * _screen.pitch;
 
-		/* Decrese height. (scroll_y is <=0). */
+		/* Decrease height. (scroll_y is <=0). */
 		height += scroll_y;
 		assert(height > 0);
 
diff --git a/src/blitter/base.hpp b/src/blitter/base.hpp
index 0a761a1..45aa19b 100644
--- a/src/blitter/base.hpp
+++ b/src/blitter/base.hpp
@@ -123,7 +123,7 @@ public:
 
 	/**
 	 * Copy from a buffer to the screen.
-	 * @param video The destionation pointer (video-buffer).
+	 * @param video The destination pointer (video-buffer).
 	 * @param src The buffer from which the data will be read.
 	 * @param width The width of the buffer.
 	 * @param height The height of the buffer.
diff --git a/src/blitter/null.hpp b/src/blitter/null.hpp
index 584afd0..4ec5ff4 100644
--- a/src/blitter/null.hpp
+++ b/src/blitter/null.hpp
@@ -37,7 +37,7 @@ public:
 	/* virtual */ int GetBytesPerPixel() { return 0; }
 };
 
-/** Factory for the blitter that doesn nothing. */
+/** Factory for the blitter that does nothing. */
 class FBlitter_Null: public BlitterFactory<FBlitter_Null> {
 public:
 	/* virtual */ const char *GetName() { return "null"; }
