Index: src/fileio.cpp
===================================================================
--- src/fileio.cpp	(revision 18831)
+++ src/fileio.cpp	(working copy)
@@ -19,6 +19,9 @@
 #ifdef WIN32
 #include <windows.h>
 #else
+#ifdef OPENBSD
+#include <unistd.h>
+#endif
 #include <pwd.h>
 #endif
 #include <sys/stat.h>
Index: src/network/core/os_abstraction.h
===================================================================
--- src/network/core/os_abstraction.h	(revision 18831)
+++ src/network/core/os_abstraction.h	(working copy)
@@ -127,6 +127,9 @@
 
 /* UNIX stuff */
 #if defined(UNIX) && !defined(__OS2__)
+#	if defined(OPENBSD)
+#		define AI_ADDRCONFIG 0
+#	endif
 #	define SOCKET int
 #	define INVALID_SOCKET -1
 #	if !defined(__MORPHOS__) && !defined(__AMIGA__)
