Index: config.lib
===================================================================
--- config.lib	(revision 12309)
+++ config.lib	(working copy)
@@ -823,7 +823,7 @@
 	if [ "$shared_dir" = "1" ]; then
 		# we are using default values
 		if [ "$os" = "OSX" ]; then
-			shared_dir="/Library/Application\\\\ Support/OpenTTD"
+			shared_dir="/Library/Application Support/OpenTTD"
 		else
 			shared_dir=""
 		fi
@@ -831,6 +831,7 @@
 
 	if [ -n "$personal_dir" ]
 	then
+		personal_dir=\"\"$personal_dir\"\"
 		log 1 "personal home directory... $personal_dir"
 	else
 		log 1 "personal home directory... none"
@@ -838,6 +839,7 @@
 
 	if [ -n "$shared_dir" ]
 	then
+		shared_dir=\"\"$shared_dir\"\"
 		log 1 "shared data directory... $shared_dir"
 	else
 		log 1 "shared data directory... none"
@@ -845,6 +847,7 @@
 
 	if [ -n "$install_dir" ]
 	then
+		install_dir=\"\"$install_dir\"\"
 		log 1 "installation directory... $install_dir"
 	else
 		log 1 "installation directory... none"
@@ -1146,10 +1149,10 @@
 	fi
 
 	if [ -n "$with_midi" ]; then
-		CFLAGS="$CFLAGS -DEXTERNAL_PLAYER=\\\\\"$with_midi\\\\\""
+		CFLAGS="$CFLAGS -DEXTERNAL_PLAYER='\\\\\"$with_midi\\\\\"'"
 	fi
 	if [ -n "$with_midi_arg" ]; then
-		CFLAGS="$CFLAGS -DMIDI_ARG=\\\\\"$with_midi_arg\\\\\""
+		CFLAGS="$CFLAGS -DMIDI_ARG='\\\\\"$with_midi_arg\\\\\"'"
 	fi
 
 	if [ "$enable_dedicated" != "0" ]; then
@@ -1200,14 +1203,14 @@
 	fi
 
 	if [ -n "$personal_dir" ]; then
-		CFLAGS="$CFLAGS -DWITH_PERSONAL_DIR -DPERSONAL_DIR=\\\\\"$personal_dir\\\\\""
+		CFLAGS="$CFLAGS -DWITH_PERSONAL_DIR -DPERSONAL_DIR='\\\\\"$personal_dir\\\\\"'"
 	fi
 
 	if [ -n "$shared_dir" ]; then
-		CFLAGS="$CFLAGS -DWITH_SHARED_DIR -DSHARED_DIR=\\\\\"$shared_dir\\\\\""
+		CFLAGS="$CFLAGS -DWITH_SHARED_DIR -DSHARED_DIR='\\\\\"$shared_dir\\\\\"'"
 	fi
 
-	CFLAGS="$CFLAGS -DGLOBAL_DATA_DIR=\\\\\"$prefix_dir/$data_dir\\\\\""
+	CFLAGS="$CFLAGS -DGLOBAL_DATA_DIR='\\\\\"$prefix_dir/$data_dir\\\\\"'"
 
 	log 1 "using CFLAGS... $CFLAGS $CC_CFLAGS"
 	log 1 "using LDFLAGS... $LIBS $LDFLAGS"
