Index: src/rail_cmd.cpp
===================================================================
--- src/rail_cmd.cpp	(revision 18821)
+++ src/rail_cmd.cpp	(working copy)
@@ -1279,7 +1279,7 @@
 {
 	if (v->type != VEH_TRAIN) return NULL;

-	/* Similiar checks as in TrainPowerChanged() */
+	/* Similar checks as in Train::PowerChanged() */

 	Train *t = Train::From(v);
 	if (t->IsArticulatedPart()) return NULL;
Index: src/train_cmd.cpp
===================================================================
--- src/train_cmd.cpp	(revision 18821)
+++ src/train_cmd.cpp	(working copy)
@@ -141,7 +141,6 @@
 /**
  * Recalculates the cached weight of a train and its vehicles. Should be called each time the cargo on
  * the consist changes.
- * @param v First vehicle of the consist.
  */
 void Train::CargoChanged()
 {
