Index: src/tunnelbridge_cmd.cpp
===================================================================
--- src/tunnelbridge_cmd.cpp	(revision 21000)
+++ src/tunnelbridge_cmd.cpp	(working copy)
@@ -190,6 +190,32 @@
 	return_cmd_error(STR_ERROR_BRIDGE_TOO_LONG);
 }
 
+/** 
+ * Can we build bridge ramp on tile with given slope in given direction?
+ * modifies height, so that wiil be the height of the bridge we try to build
+ *
+ * @param dir    The direction of the bridge
+ * @param tileh  The slope of the tile
+ * @param z      TileZ corresponding to tileh, it gets modified (height of the ramp + foundation, if they will be built)
+ * @return       True if we can build the bridge ramp
+ */
+static inline bool CanBuildBridgeRamp(DiagDirection dir, Slope tileh, uint *z) {
+	Axis axis = DiagDirToAxis(dir);
+	CommandCost ret;
+	switch(dir) {
+		case DIAGDIR_NE:
+		case DIAGDIR_NW:
+			ret = CheckBridgeSlopeSouth(axis, &tileh, z);
+			break;
+		case DIAGDIR_SW:
+		case DIAGDIR_SE:
+			ret = CheckBridgeSlopeNorth(axis, &tileh, z);
+			break;
+		default: NOT_REACHED();
+	}
+	return (ret.Succeeded() && (_settings_game.construction.build_on_slopes || !ret.GetCost()));
+}
+
 /**
  * Build a Bridge
  * @param end_tile end tile
