diff -r bdb22d62f768 src/industry_gui.cpp
--- a/src/industry_gui.cpp	Sat Dec 06 09:27:49 2008 +0100
+++ b/src/industry_gui.cpp	Sat Dec 06 11:14:23 2008 +0100
@@ -35,10 +35,11 @@
 
 bool _ignore_restrictions;
 
+/** Cargo suffix type (for which window is it requested) */
 enum CargoSuffixType {
-	CST_FUND,
-	CST_VIEW,
-	CST_DIR,
+	CST_FUND,  ///< Fund-industry window
+	CST_VIEW,  ///< View-industry window
+	CST_DIR,   ///< Industry-directory window
 };
 
 /**
@@ -49,7 +50,7 @@
  * - 02 - third accepted cargo type
  * - 03 - first produced cargo type
  * - 04 - second produced cargo type
- * @param cst the cargo suffix type (for which window is it requested)
+ * @param cst the cargo suffix type (for which window is it requested). @see CargoSuffixType
  * @param ind the industry (NULL if in fund window)
  * @param ind_type the industry type
  * @param indspec the industry spec
