Index: src/viewport.cpp
===================================================================
--- src/viewport.cpp	(revision 11101)
+++ src/viewport.cpp	(working copy)
@@ -1147,9 +1147,9 @@
 				/* Decide which comparator to use, based on whether the bounding
 				 * boxes overlap
 				 */
-				if (ps->xmax > ps2->xmin && ps->xmin < ps2->xmax && // overlap in X?
-						ps->ymax > ps2->ymin && ps->ymin < ps2->ymax && // overlap in Y?
-						ps->zmax > ps2->zmin && ps->zmin < ps2->zmax) { // overlap in Z?
+				if (ps->xmax >= ps2->xmin && ps->xmin <= ps2->xmax && // overlap in X?
+						ps->ymax >= ps2->ymin && ps->ymin <= ps2->ymax && // overlap in Y?
+						ps->zmax >= ps2->zmin && ps->zmin <= ps2->zmax) { // overlap in Z?
 					/* Use X+Y+Z as the sorting order, so sprites closer to the bottom of
 					 * the screen and with higher Z elevation, are drawn in front.
 					 * Here X,Y,Z are the coordinates of the "center of mass" of the sprite,
@@ -1161,13 +1161,13 @@
 						continue;
 					}
 				} else {
+					/* We only change the order, if it is definite.
+					 * I.e. every single order of X, Y, Z says ps2 is behind ps or they overlap.
+					 * That is: If one partial order says ps behind ps2, do not change the order.
+					 */
 					if (ps->xmax < ps2->xmin ||
 							ps->ymax < ps2->ymin ||
-							ps->zmax < ps2->zmin || (
-								ps->xmin < ps2->xmax &&
-								ps->ymin < ps2->ymax &&
-								ps->zmin < ps2->zmax
-							)) {
+							ps->zmax < ps2->zmin) {
 						continue;
 					}
 				}
