Index: src/music/cocoa_m.cpp
===================================================================
diff --git a/tags/1.6.1/src/music/cocoa_m.cpp b/tags/1.6.1/src/music/cocoa_m.cpp
--- a/tags/1.6.1/src/music/cocoa_m.cpp	(revision 27632)
+++ b/tags/1.6.1/src/music/cocoa_m.cpp	(working copy)
@@ -68,11 +68,7 @@
 			 * risk compilation errors. The header AudioComponent.h
 			 * was introduced in 10.6 so use it to decide which
 			 * type definition to use. */
-#ifdef __AUDIOCOMPONENT_H__
 			AudioComponentDescription desc;
-#else
-			ComponentDescription desc;
-#endif
 			AUGraphNodeInfo(graph, node, &desc, &unit);
 			comp_type = desc.componentType;
 		} else
Index: src/video/cocoa/wnd_quartz.mm
===================================================================
diff --git a/tags/1.6.1/src/video/cocoa/wnd_quartz.mm b/tags/1.6.1/src/video/cocoa/wnd_quartz.mm
--- a/tags/1.6.1/src/video/cocoa/wnd_quartz.mm	(revision 27632)
+++ b/tags/1.6.1/src/video/cocoa/wnd_quartz.mm	(working copy)
@@ -110,12 +110,10 @@
 	static CGColorSpaceRef colorSpace = NULL;
 
 	if (colorSpace == NULL) {
-		CMProfileRef sysProfile;
+		/* http://stackoverflow.com/a/33076871 */
+		colorSpace = CGDisplayCopyColorSpace(CGMainDisplayID());
 
-		if (CMGetSystemProfile(&sysProfile) == noErr) {
-			colorSpace = CGColorSpaceCreateWithPlatformColorSpace(sysProfile);
-			CMCloseProfile(sysProfile);
-		} else {
+		if (colorSpace == NULL) {
 			colorSpace = CGColorSpaceCreateDeviceRGB();
 		}
 
