Index: src/saveload/signs_sl.cpp
===================================================================
--- src/saveload/signs_sl.cpp	(Revision 20914)
+++ src/saveload/signs_sl.cpp	(Arbeitskopie)
@@ -22,8 +22,9 @@
   SLE_CONDVAR(Sign, x,     SLE_INT32,                  5, SL_MAX_VERSION),
   SLE_CONDVAR(Sign, y,     SLE_INT32,                  5, SL_MAX_VERSION),
   SLE_CONDVAR(Sign, owner, SLE_UINT8,                  6, SL_MAX_VERSION),
-      SLE_VAR(Sign, z,     SLE_UINT8),
-	SLE_END()
+  SLE_CONDVAR(Sign, z,     SLE_FILE_U8 | SLE_VAR_U32,  0, MORE_HEIGHTLEVEL_SAVEGAME_VERSION - 1),
+  SLE_CONDVAR(Sign, z,     SLE_UINT32,                 MORE_HEIGHTLEVEL_SAVEGAME_VERSION, SL_MAX_VERSION),
+  SLE_END()
 };
 
 /** Save all signs */
Index: src/signs_base.h
===================================================================
--- src/signs_base.h	(Revision 20914)
+++ src/signs_base.h	(Arbeitskopie)
@@ -25,7 +25,7 @@
 	ViewportSign sign;
 	int32        x;
 	int32        y;
-	byte         z;
+	int32        z;
 	OwnerByte    owner; // placed by this company. Anyone can delete them though. OWNER_NONE for gray signs from old games.
 
 	/**
