Index: src/video/cocoa/cocoa_v.mm
===================================================================
--- src/video/cocoa/cocoa_v.mm	(revision 21674)
+++ src/video/cocoa/cocoa_v.mm	(working copy)
@@ -511,7 +511,7 @@
 /**
  * Initialize event system for the application rectangle
  */
-- (id)initWithContentRect:(NSRect)contentRect styleMask:(unsigned int)styleMask backing:(NSBackingStoreType)backingType defer:(BOOL)flag
+- (id)initWithContentRect:(NSRect)contentRect styleMask:(NSUInteger)styleMask backing:(NSBackingStoreType)backingType defer:(BOOL)flag
 {
 	/* Make our window subclass receive these application notifications */
 	[ [ NSNotificationCenter defaultCenter ] addObserver:self
Index: src/video/cocoa/wnd_quartz.mm
===================================================================
--- src/video/cocoa/wnd_quartz.mm	(revision 21674)
+++ src/video/cocoa/wnd_quartz.mm	(working copy)
@@ -301,7 +301,7 @@
 
 		[ this->cocoaview setDriver:this ];
 
-		[ this->cocoaview setAutoresizingMask:NSViewWidthSizable | NSViewHeightSizable ];
+		[ (NSView*)(this->cocoaview) setAutoresizingMask:NSViewWidthSizable | NSViewHeightSizable ];
 		[ this->window setContentView:cocoaview ];
 		[ this->cocoaview release ];
 		[ this->window makeKeyAndOrderFront:nil ];
