diff --git a/src/video/cocoa/cocoa_v.mm b/src/video/cocoa/cocoa_v.mm
index a9f3981..df2e780 100644
--- a/src/video/cocoa/cocoa_v.mm
+++ b/src/video/cocoa/cocoa_v.mm
@@ -573,7 +573,7 @@ bool VideoDriver_Cocoa::AfterBlitterChange()
 void VideoDriver_Cocoa::EditBoxLostFocus()
 {
 	if (_cocoa_subdriver != NULL) {
-		if ([ _cocoa_subdriver->cocoaview respondsToSelector:@selector(inputContext) ]) {
+		if ([ _cocoa_subdriver->cocoaview respondsToSelector:@selector(inputContext) ] && [ [ _cocoa_subdriver->cocoaview performSelector:@selector(inputContext) ] respondsToSelector:@selector(discardMarkedText) ]) {
 			[ [ _cocoa_subdriver->cocoaview performSelector:@selector(inputContext) ] performSelector:@selector(discardMarkedText) ];
 		} else {
 			[ [ NSInputManager currentInputManager ] markedTextAbandoned:_cocoa_subdriver->cocoaview ];
