Index: config.lib
===================================================================
--- config.lib	(Revision 14505)
+++ config.lib	(Arbeitskopie)
@@ -55,7 +55,6 @@
 	enable_cocoa_quickdraw="1"
 	with_osx_sysroot="1"
 	with_application_bundle="1"
-	with_menu_entry="1"
 	with_sdl="1"
 	with_cocoa="1"
 	with_zlib="1"
@@ -2428,9 +2427,7 @@
 		s~!!BINARY_DIR!!~$prefix_dir/$binary_dir~g;
 		s~!!DATA_DIR!!~$prefix_dir/$data_dir~g;
 		s~!!DOC_DIR!!~$prefix_dir/$doc_dir~g;
-		s~!!MAN_DIR!!~$prefix_dir/$man_dir~g;
 		s~!!ICON_DIR!!~$prefix_dir/$icon_dir~g;
-		s~!!ICON_THEME_DIR!!~$prefix_dir/$icon_theme_dir~g;
 		s~!!PERSONAL_DIR!!~$personal_dir~g;
 		s~!!SHARED_DIR!!~$shared_dir~g;
 		s~!!INSTALL_DIR!!~$install_dir~g;
@@ -2628,15 +2625,15 @@
 	echo "                                 [share/games/openttd]"
 	echo "  --doc-dir=dir                  location of the doc files"
 	echo "                                 Will be prefixed with the prefix-dir"
-	echo "                                 [$doc_dir]"
+	echo "                                 [share/doc/openttd]"
 	echo "  --icon-dir=dir                 location of icons. Will be prefixed"
 	echo "                                 with the prefix-dir [share/pixmaps]"
 	echo "  --icon-theme-dir=dir           location of icon theme."
 	echo "                                 Will be prefixed with the prefix-dir"
-	echo "                                 and postfixed with size-dirs [$icon_theme_dir]"
+	echo "                                 and postfixed with size-dirs [share/icons/hicolor]"
 	echo "  --man-dir=dir                  location of the manual page (UNIX only)"
 	echo "                                 Will be prefixed with the prefix-dir"
-	echo "                                 [$man_dir]"
+	echo "                                 [share/man/man6]"
 	echo "  --menu-dir=dir                 location of the menu item. (UNIX only, except OSX)"
 	echo "                                 Will be prefixed with the prefix-dir"
 	echo "                                 [share/applications]"
