Index: src/industry_cmd.cpp
===================================================================
--- src/industry_cmd.cpp	(revision 11704)
+++ src/industry_cmd.cpp	(working copy)
@@ -1612,7 +1612,11 @@
 			 * is nothing we can really do about that. */
 			if (Random() <= indspec->prospecting_chance) {
 				for (int i = 0; i < 5000; i++) {
-					const Industry *ind = CreateNewIndustryHelper(RandomTile(), p1, flags, indspec, RandomRange(indspec->num_table));
+					/* Only one Random() call inside a function call, else we get
+					 * desyncs between clients on different compilers/optimization
+					 * settings. */
+					TileIndex location = RandomTile();
+					const Industry *ind = CreateNewIndustryHelper(location, p1, flags, indspec, RandomRange(indspec->num_table));
 					if (ind != NULL) {
 						SetDParam(0, indspec->name);
 						if (indspec->new_industry_text > STR_LAST_STRINGID) {
