diff -r 78589fa46b4e bin/data/opntitle.dat
Binary file bin/data/opntitle.dat has changed
diff -r 78589fa46b4e src/core/random_func.hpp
--- a/src/core/random_func.hpp	Sat Feb 13 20:13:34 2010 +0000
+++ b/src/core/random_func.hpp	Sun Feb 14 11:54:29 2010 +0100
@@ -85,7 +85,11 @@
 
 void SetRandomSeed(uint32 seed);
 #ifdef RANDOM_DEBUG
-	#define Random() DoRandom(__LINE__, __FILE__)
+	#ifdef __APPLE__
+		#define OTTD_Random() DoRandom(__LINE__, __FILE__)
+	#else
+		#define Random() DoRandom(__LINE__, __FILE__)
+	#endif
 	uint32 DoRandom(int line, const char *file);
 	#define RandomRange(max) DoRandomRange(max, __LINE__, __FILE__)
 	uint DoRandomRange(uint max, int line, const char *file);
