Index: src/network/core/os_abstraction.h
===================================================================
--- src/network/core/os_abstraction.h	(revision 19751)
+++ src/network/core/os_abstraction.h	(working copy)
@@ -127,7 +127,7 @@
 
 /* UNIX stuff */
 #if defined(UNIX) && !defined(__OS2__)
-#	if defined(OPENBSD)
+#	if defined(OPENBSD) || defined(__NetBSD__)
 #		define AI_ADDRCONFIG 0
 #	endif
 #	define SOCKET int
Index: src/stdafx.h
===================================================================
--- src/stdafx.h	(revision 19751)
+++ src/stdafx.h	(working copy)
@@ -24,6 +24,9 @@
 #elif defined(__NDS__)
 	#include <nds/jtypes.h>
 	#define TROUBLED_INTS
+#elif defined(__NetBSD__)
+	#include <unistd.h>
+	#define _GNU_SOURCE
 #endif
 
 /* It seems that we need to include stdint.h before anything else
Index: config.lib
===================================================================
--- config.lib	(revision 19751)
+++ config.lib	(working copy)
@@ -2304,6 +2304,13 @@
 		if [ -z "$res" ]; then
 			log 2 "  trying /usr/local/include/$4$5... no"
 		fi
+		if [ -z "$res" ] && [ "$os" = "NETBSD" ]; then
+			eval "$2=`ls -1 /usr/pkg/include/$4*.h 2>/dev/null | egrep \"\/$5\$\"`"
+			eval "res=\$$2"
+			if [ -z "$res" ]; then
+				log 2 "  trying /usr/pkg/include/$4$5... no"
+			fi
+		fi
 
 		eval "res=\$$2"
 		if [ -n "$res" ] && ( [ -n "$force_static" ] || ( [ "$enable_static" != "0" ] && [ "$os" != "OSX" ] ) ); then
