From fe4f2aa904ad4f0e4e2cb527edcb2fad0e2e7494 Mon Sep 17 00:00:00 2001
From: Michael Lutz <michi@icosahedron.de>
Date: Mon, 5 Jan 2009 17:28:31 +0100
Subject: -Fix [FS#2503]: Reversing a stuck train that is then not stuck anymore did not always reset the waiting timer.

---
 src/train_cmd.cpp |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/src/train_cmd.cpp b/src/train_cmd.cpp
index 5521b0b..e71ef9e 100644
--- a/src/train_cmd.cpp
+++ b/src/train_cmd.cpp
@@ -3136,7 +3136,10 @@ bool TryPathReserve(Vehicle *v, bool mark_as_stuck, bool first_tile_okay)
 		return false;
 	}
 
-	if (HasBit(v->u.rail.flags, VRF_TRAIN_STUCK)) InvalidateWindowWidget(WC_VEHICLE_VIEW, v->index, VVW_WIDGET_START_STOP_VEH);
+	if (HasBit(v->u.rail.flags, VRF_TRAIN_STUCK)) {
+		v->load_unload_time_rem = 0;
+		InvalidateWindowWidget(WC_VEHICLE_VIEW, v->index, VVW_WIDGET_START_STOP_VEH);
+	}
 	ClrBit(v->u.rail.flags, VRF_TRAIN_STUCK);
 	return true;
 }
-- 
1.6.0.4

