diff -r 3514f1c5e182 src/video/cocoa/cocoa_v.mm
--- a/src/video/cocoa/cocoa_v.mm	Tue Dec 29 04:27:39 2009 +0000
+++ b/src/video/cocoa/cocoa_v.mm	Tue Dec 29 21:58:35 2009 +0100
@@ -203,8 +203,10 @@
 
 static CocoaSubdriver *QZ_CreateWindowSubdriver(int width, int height, int bpp)
 {
+#if defined ENABLE_COCOA_QUARTZ || defined ENABLE_COCOA_QUICKDRAW
 	CocoaSubdriver *ret;
-
+#endif
+	
 #ifdef ENABLE_COCOA_QUARTZ
 #if MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_4
 	/* The reason for the version mismatch is due to the fact that the 10.4 binary needs to work on 10.5 as well. */
