Index: src/terraform_gui.cpp
===================================================================
--- src/terraform_gui.cpp	(revision 13265)
+++ src/terraform_gui.cpp	(working copy)
@@ -465,7 +465,8 @@
 {    WWT_IMGBTN,   RESIZE_NONE,    14,    90,   111,    16,    37, SPR_IMG_BUILD_CANAL,       STR_CREATE_LAKE},                         // ETTW_BUILD_CANAL
 {    WWT_IMGBTN,   RESIZE_NONE,    14,   112,   133,    16,    37, SPR_IMG_BUILD_RIVER,       STR_CREATE_RIVER},                        // ETTW_BUILD_RIVER
 {    WWT_IMGBTN,   RESIZE_NONE,    14,   134,   156,    16,    37, SPR_IMG_ROCKS,             STR_028C_PLACE_ROCKY_AREAS_ON_LANDSCAPE}, // ETTW_PLACE_ROCKS
-{    WWT_IMGBTN,   RESIZE_NONE,    14,   157,   179,    16,    37, SPR_IMG_LIGHTHOUSE_DESERT, STR_NULL},                                // ETTW_PLACE_DESERT_LIGHTHOUSE XXX - dynamic
+{    WWT_IMGBTN,   RESIZE_NONE,    14,   157,   179,    16,    37, SPR_IMG_LIGHTHOUSE_DESERT, STR_028D_PLACE_LIGHTHOUSE},               // ETTW_PLACE_LIGHTHOUSE
+{    WWT_IMGBTN,   RESIZE_NONE,    14,   157,   179,    16,    37, SPR_IMG_LIGHTHOUSE_DESERT, STR_028F_DEFINE_DESERT_AREA},             // ETTW_PLACE_DESERT
 {    WWT_IMGBTN,   RESIZE_NONE,    14,   180,   201,    16,    37, SPR_IMG_TRANSMITTER,       STR_028E_PLACE_TRANSMITTER},              // ETTW_PLACE_TRANSMITTER
 {    WWT_IMGBTN,   RESIZE_NONE,    14,   150,   161,    45,    56, SPR_ARROW_UP,              STR_0228_INCREASE_SIZE_OF_LAND_AREA},     // ETTW_INCREASE_SIZE
 {    WWT_IMGBTN,   RESIZE_NONE,    14,   150,   161,    58,    69, SPR_ARROW_DOWN,            STR_0229_DECREASE_SIZE_OF_LAND_AREA},     // ETTW_DECREASE_SIZE
@@ -500,7 +501,8 @@
 	ETTW_BUILD_CANAL,                      ///< Build canal button
 	ETTW_BUILD_RIVER,                      ///< Build river button
 	ETTW_PLACE_ROCKS,                      ///< Place rocks button
-	ETTW_PLACE_DESERT_LIGHTHOUSE,          ///< Place desert button (in tropical climate) / place lighthouse button (else)
+	ETTW_PLACE_LIGHTHOUSE,                 ///< Place lighthouse button (not available in tropical climate)
+	ETTW_PLACE_DESERT,                     ///< Place desert button (only in tropical climate)
 	ETTW_PLACE_TRANSMITTER,                ///< Place transmitter button
 	ETTW_BUTTONS_END,                      ///< End of pushable buttons
 	ETTW_INCREASE_SIZE = ETTW_BUTTONS_END, ///< Upwards arrow button to increase terraforming size
@@ -548,11 +550,16 @@
 	HandlePlacePushButton(w, ETTW_PLACE_ROCKS, SPR_CURSOR_ROCKY_AREA, VHM_RECT, PlaceProc_RockyArea);
 }
 
-static void EditorTerraformClick_DesertLightHouse(Window *w)
+static void EditorTerraformClick_LightHouse(Window *w)
 {
-	HandlePlacePushButton(w, ETTW_PLACE_DESERT_LIGHTHOUSE, SPR_CURSOR_LIGHTHOUSE, VHM_RECT, (_settings.game_creation.landscape == LT_TROPIC) ? PlaceProc_DesertArea : PlaceProc_LightHouse);
+	HandlePlacePushButton(w, ETTW_PLACE_LIGHTHOUSE, SPR_CURSOR_LIGHTHOUSE, VHM_RECT, PlaceProc_LightHouse);
 }
 
+static void EditorTerraformClick_Desert(Window *w)
+{
+	HandlePlacePushButton(w, ETTW_PLACE_DESERT, SPR_CURSOR_LIGHTHOUSE, VHM_RECT, PlaceProc_DesertArea);
+}
+
 static void EditorTerraformClick_Transmitter(Window *w)
 {
 	HandlePlacePushButton(w, ETTW_PLACE_TRANSMITTER, SPR_CURSOR_TRANSMITTER, VHM_RECT, PlaceProc_Transmitter);
@@ -567,6 +574,7 @@
 	'T',
 	'Y',
 	'U',
+	'U',
 	'I'
 };
 
@@ -579,7 +587,8 @@
 	EditorTerraformClick_WaterArea,
 	EditorTerraformClick_RiverArea,
 	EditorTerraformClick_RockyArea,
-	EditorTerraformClick_DesertLightHouse,
+	EditorTerraformClick_LightHouse,
+	EditorTerraformClick_Desert,
 	EditorTerraformClick_Transmitter
 };
 
@@ -615,7 +624,10 @@
 struct ScenarioEditorLandscapeGenerationWindow : Window {
 	ScenarioEditorLandscapeGenerationWindow(const WindowDesc *desc, WindowNumber window_number) : Window(desc, window_number)
 	{
-		this->widget[ETTW_PLACE_DESERT_LIGHTHOUSE].tooltips = (_settings.game_creation.landscape == LT_TROPIC) ? STR_028F_DEFINE_DESERT_AREA : STR_028D_PLACE_LIGHTHOUSE;
+		/* Hide Lighthouse on tropic climate, hide desert on non-tropic climate, and vice versa */
+		bool desert = _settings.game_creation.landscape == LT_TROPIC;
+		this->SetWidgetHiddenState(ETTW_PLACE_LIGHTHOUSE, desert);
+		this->SetWidgetHiddenState(ETTW_PLACE_DESERT, !desert);
 		this->FindWindowPlacementAndResize(desc);
 	}
 
@@ -638,9 +650,11 @@
 
 	virtual EventState OnKeyPress(uint16 key, uint16 keycode)
 	{
-		for (uint i = 0; i != lengthof(_editor_terraform_keycodes); i++) {
-			if (keycode == _editor_terraform_keycodes[i]) {
-				_editor_terraform_button_proc[i](this);
+		for (uint i = ETTW_BUTTONS_START; i < ETTW_BUTTONS_END; i++) {
+			if (keycode == _editor_terraform_keycodes[i - ETTW_BUTTONS_START]) {
+				/* If lighthouse is not available, choose desert instead (they use the same hotkey) */
+				if (i == ETTW_PLACE_LIGHTHOUSE && this->IsWidgetHidden(ETTW_PLACE_LIGHTHOUSE)) i = ETTW_PLACE_DESERT;
+				_editor_terraform_button_proc[i - ETTW_BUTTONS_START](this);
 				return ES_HANDLED;
 			}
 		}
