Index: src/autoreplace_gui.cpp
===================================================================
--- src/autoreplace_gui.cpp	(revision 13003)
+++ src/autoreplace_gui.cpp	(working copy)
@@ -26,15 +26,15 @@
 #include "table/strings.h"
 
 struct replaceveh_d {
-	byte sel_index[2];
+	EngineList list[2];
 	EngineID sel_engine[2];
 	uint16 count[2];
+	GroupID sel_group;
+	byte sel_index[2];
 	bool wagon_btnstate; ///< true means engine is selected
-	EngineList list[2];
 	bool update_left;
 	bool update_right;
 	bool init_lists;
-	GroupID sel_group;
 };
 assert_compile(WINDOW_CUSTOM_SIZE >= sizeof(replaceveh_d));
 
