Index: src/news_gui.cpp
===================================================================
--- src/news_gui.cpp	2008-05-30 22:57:19.000000000 +0200
+++ src/news_gui.cpp	2008-06-01 11:19:17.000000000 +0200
@@ -364,8 +364,6 @@
 /** Open up an own newspaper window for the news item */
 static void ShowNewspaper(NewsItem *ni)
 {
-	ni->flags &= ~NF_FORCE_BIG;
-
 	SoundFx sound = _news_type_data[_news_subtype_data[ni->subtype].type].sound;
 	if (sound != 0) SndPlayFx(sound);
 
@@ -468,12 +466,9 @@
 				InvalidateWindowData(WC_STATUS_BAR, 0, SBI_SHOW_REMINDER);
 				break;
 
-			case ND_SUMMARY: // Summary - show ticker, but if forced big, cascade to full
-				if (!(ni->flags & NF_FORCE_BIG)) {
-					ShowTicker(ni);
-					break;
-				}
-				/* Fallthrough */
+			case ND_SUMMARY: // Summary - show ticker
+				ShowTicker(ni);
+				break;
 
 			case ND_FULL: // Full - show newspaper
 				ShowNewspaper(ni);
@@ -616,8 +611,6 @@
 	_forced_news = ni;
 
 	if (_forced_news != NULL) {
-		NewsWindow::duration = 555;
-		ni->flags |= NF_FORCE_BIG;
 		DeleteWindowById(WC_NEWS_WINDOW, 0);
 		ShowNewspaper(ni);
 	}
Index: src/news_type.h
===================================================================
--- src/news_type.h	2008-05-30 22:57:19.000000000 +0200
+++ src/news_type.h	2008-06-01 11:21:09.000000000 +0200
@@ -74,7 +74,6 @@
 	NF_VIEWPORT  = (1 << 1), ///< Does the news message have a viewport? (ingame picture of happening)
 	NF_TILE      = (1 << 2), ///< When clicked on the news message scroll to a given tile? Tile is in data_a
 	NF_VEHICLE   = (1 << 3), ///< When clicked on the message scroll to the vehicle? VehicleID is in data_a
-	NF_FORCE_BIG = (1 << 4), ///< Force the appearance of a news message if it has already been shown (internal)
 	NF_INCOLOR   = (1 << 5), ///< Show the newsmessage in colour, otherwise it defaults to black & white
 	NF_TILE2     = (1 << 6), ///< There is a second tile to scroll to; tile is in data_b
 };
