Index: industry_cmd.cpp
===================================================================
--- industry_cmd.cpp	(revision 17352)
+++ industry_cmd.cpp	(working copy)
@@ -1507,8 +1507,8 @@
 	i->accepts_cargo[0] = indspec->accepts_cargo[0];
 	i->accepts_cargo[1] = indspec->accepts_cargo[1];
 	i->accepts_cargo[2] = indspec->accepts_cargo[2];
-	i->production_rate[0] = indspec->production_rate[0];
-	i->production_rate[1] = indspec->production_rate[1];
+	i->production_rate[0] = ((indspec->production_rate[0] * 23) / 15);
+	i->production_rate[1] = ((indspec->production_rate[1] * 23) / 15);
 
 	/* don't use smooth economy for industries using production related callbacks */
 	if (_settings_game.economy.smooth_economy &&
@@ -2207,8 +2207,8 @@
 			}
 		} else {
 			if (only_decrease || Chance16(1, 3)) {
-				/* If more than 60% transported, 66% chance of increase, else 33% chance of increase */
-				if (!only_decrease && (i->last_month_pct_transported[0] > PERCENT_TRANSPORTED_60) != Chance16(1, 3)) {
+				/* If more than 60% transported, 66% or more chance of increase, else 33% chance of increase */
+				if (!only_decrease && (i->last_month_pct_transported[0] > PERCENT_TRANSPORTED_60) != Chance16(1, (3 + ((i->last_month_pct_transported[0] / 169) * (i->last_month_pct_transported[0] >> 5))))) {
 					mul = 1; // Increase production
 				} else {
 					div = 1; // Decrease production
@@ -2255,8 +2255,8 @@
 	 * For non-smooth economy these should always be synchronized with prod_level */
 	if (recalculate_multipliers) {
 		/* Rates are rounded up, so e.g. oilrig always produces some passengers */
-		i->production_rate[0] = min((indspec->production_rate[0] * i->prod_level + PRODLEVEL_DEFAULT - 1) / PRODLEVEL_DEFAULT, 0xFF);
-		i->production_rate[1] = min((indspec->production_rate[1] * i->prod_level + PRODLEVEL_DEFAULT - 1) / PRODLEVEL_DEFAULT, 0xFF);
+		i->production_rate[0] = min((((indspec->production_rate[0] * 23) / 15) * i->prod_level + PRODLEVEL_DEFAULT - 1) / PRODLEVEL_DEFAULT, 0xFF);
+		i->production_rate[1] = min((((indspec->production_rate[1] * 23) / 15) * i->prod_level + PRODLEVEL_DEFAULT - 1) / PRODLEVEL_DEFAULT, 0xFF);
 	}
 
 	/* Close if needed and allowed */
