diff -r 3ddd1a3a1c67 src/settings_gui.cpp
--- a/src/settings_gui.cpp	Sun Nov 02 13:24:19 2008 +0100
+++ b/src/settings_gui.cpp	Sun Nov 02 13:26:17 2008 +0100
@@ -1049,7 +1049,7 @@
 			SetDParam(1, value);
 		}
 	}
-	DrawString(x + 25, y, (sdb->str) + disabled, TC_FROMSTRING);
+	DrawString(x + TREE_PATCH_TEXT_OFFSET, y, (sdb->str) + disabled, TC_FROMSTRING);
 }
 
 static const char *_patches_ui[] = {
