Index: src/ai/api/ai_tile.hpp
===================================================================
--- src/ai/api/ai_tile.hpp	(revision 12185)
+++ src/ai/api/ai_tile.hpp	(working copy)
@@ -118,7 +118,7 @@
 	 * @param slope corners to raise (SLOPE_xxx).
 	 * @return 0 means failed, 1 means success.
 	 */
-	bool RaiseTile(TileIndex tile, int32 slope);
+	static bool RaiseTile(TileIndex tile, int32 slope);
 
 	/**
 	 * Lower the given corners of the tile. The corners can be combined,
@@ -128,7 +128,7 @@
 	 * @param slope corners to lower (SLOPE_xxx).
 	 * @return 0 means failed, 1 means success.
 	 */
-	bool LowerTile(TileIndex tile, int32 slope);
+	static bool LowerTile(TileIndex tile, int32 slope);
 
 };

Index: src/ai/api/ai_tile.cpp
===================================================================
--- src/ai/api/ai_tile.cpp	(revision 12185)
+++ src/ai/api/ai_tile.cpp	(working copy)
@@ -75,7 +75,7 @@
 	/* Outside of the map */
 	if (tile >= ::MapSize()) return false;
 
-	return this->DoCommand(tile, slope, 1, CMD_TERRAFORM_LAND);
+	return DoCommand(tile, slope, 1, CMD_TERRAFORM_LAND);
 }
 
 bool AITile::LowerTile(TileIndex tile, int32 slope)
@@ -83,5 +83,5 @@
 	/* Outside of the map */
 	if (tile >= ::MapSize()) return false;
 
-	return this->DoCommand(tile, slope, 0, CMD_TERRAFORM_LAND);
+	return DoCommand(tile, slope, 0, CMD_TERRAFORM_LAND);
 }
Index: src/ai/api/ai_tile.hpp.sq
===================================================================
--- src/ai/api/ai_tile.hpp.sq	(revision 12185)
+++ src/ai/api/ai_tile.hpp.sq	(working copy)
@@ -47,9 +47,8 @@
 	SQAITile.DefSQStaticMethod(engine, &AITile::GetHeight,          "GetHeight",          2, "xi");
 	SQAITile.DefSQStaticMethod(engine, &AITile::GetCargoAcceptance, "GetCargoAcceptance", 6, "xiiiii");
 	SQAITile.DefSQStaticMethod(engine, &AITile::GetCargoProduction, "GetCargoProduction", 6, "xiiiii");
+	SQAITile.DefSQStaticMethod(engine, &AITile::RaiseTile,          "RaiseTile",          3, "xii");
+	SQAITile.DefSQStaticMethod(engine, &AITile::LowerTile,          "LowerTile",          3, "xii");
 
-	SQAITile.DefSQMethod(engine, &AITile::RaiseTile, "RaiseTile", 3, "xii");
-	SQAITile.DefSQMethod(engine, &AITile::LowerTile, "LowerTile", 3, "xii");
-
 	SQAITile.PostRegister(engine);
 }
