Index: src/train_cmd.cpp
===================================================================
--- src/train_cmd.cpp	(revision 10369)
+++ src/train_cmd.cpp	(working copy)
@@ -2400,6 +2400,23 @@
 		at_waypoint = true;
 	}
 
+	bool may_reverse = false;
+
+	/* Do not reverse unless at a station (and reversing there is allowed) ... */
+	if (v->current_order.type == OT_GOTO_STATION &&
+			_opt.diff.line_reverse_mode == 0 &&
+			IsTileType(v->tile, MP_STATION) &&
+			GetStationByTile(v->tile)->index == v->current_order.dest) {
+		may_reverse = true;
+	}
+	/* ... or at a depot */
+	if (v->current_order.type == OT_GOTO_DEPOT &&
+			IsTileType(v->tile, MP_RAILWAY) &&
+			GetRailTileType(v->tile) == RAIL_TILE_DEPOT &&
+			GetDepotByTile(v->tile)->index == v->current_order.dest) {
+		may_reverse = true;
+	}
+
 	/* check if we've reached a non-stop station while TTDPatch nonstop is enabled.. */
 	if (_patches.new_nonstop &&
 			v->current_order.flags & OF_NON_STOP &&
@@ -2453,7 +2470,7 @@
 			return false;
 	}
 
-	return !at_waypoint && CheckReverseTrain(v);
+	return may_reverse && !at_waypoint && CheckReverseTrain(v);
 }
 
 void Train::MarkDirty()
