Index: src/settings.cpp
===================================================================
--- src/settings.cpp	(Revision 12658)
+++ src/settings.cpp	(Arbeitskopie)
@@ -1514,6 +1514,7 @@
 	 SDT_VAR(Patches, dist_local_authority,SLE_UINT8, 0, 0, 20, 5,  60, 0, STR_NULL, NULL),
 	 SDT_VAR(Patches, wait_oneway_signal,  SLE_UINT8, 0, 0, 15, 2, 100, 0, STR_NULL, NULL),
 	 SDT_VAR(Patches, wait_twoway_signal,  SLE_UINT8, 0, 0, 41, 2, 100, 0, STR_NULL, NULL),
+	 SDT_VAR(Patches, max_year, SLE_UINT32, 0, 0, MAX_YEAR, MIN_YEAR, MAX_YEAR, 0, STR_NULL, NULL),
 
 	/***************************************************************************/
 	/* New Pathfinding patch settings */
Index: src/date.cpp
===================================================================
--- src/date.cpp	(Revision 12658)
+++ src/date.cpp	(Arbeitskopie)
@@ -289,7 +289,7 @@
 	if (_cur_year == _patches.ending_year) {
 			ShowEndGameChart();
 	/* check if we reached the maximum year, decrement dates by a year */
-	} else if (_cur_year == MAX_YEAR + 1) {
+	} else if (_cur_year == _patches.max_year + 1) {
 		Vehicle *v;
 		uint days_this_year;
 
Index: src/settings_type.h
===================================================================
--- src/settings_type.h	(Revision 12658)
+++ src/settings_type.h	(Arbeitskopie)
@@ -148,6 +148,7 @@
 	Year starting_year;                    ///< starting date
 	Year ending_year;                      ///< end of the game (just show highscore)
 	Year colored_news_year;                ///< when does newspaper become colored?
+	Year max_year;                         ///< if reached, it will be looped endless.
 
 	bool keep_all_autosave;                ///< name the autosave in a different way.
 	bool autosave_on_exit;                 ///< save an autosave when you quit the game, but do not ask "Do you really want to quit?"
