Index: table/namegen.h
===================================================================
--- table/namegen.h	(revision 9561)
+++ table/namegen.h	(working copy)
@@ -455,7 +455,6 @@
 	"Chemnitz",
 	"Dortmund",
 	"Dresden",
-	"Erfurt",
 	"Erlangen",
 	"Essen",
 	"Fulda",
@@ -495,6 +494,7 @@
 	"Elster",
 	"Eichen",
 	"Ems",
+	"Er",
 	"Fahr",
 	"Falken",
 	"Flens",
@@ -574,21 +574,36 @@
 };
 
 static const char *name_german_2[] = {
+	"au",
 	"bach",
+	"beck",
 	"berg",
+	"born",
 	"brück",
 	"brücken",
+	"büren",
 	"burg",
 	"dorf",
 	"feld",
 	"furt",
+	"hagen",
 	"hausen",
 	"haven",
 	"heim",
+	"hofen",
 	"horst",
+	"hütte",
+	"kamp",
+	"kirchen",
+	"leben",
+	"lingen",
+	"loh",
 	"mund",
 	"münster",
 	"stadt",
+	"stedt",
+	"thal",
+	"wig"
 	"wald"
 };
 
