Index: src/order_gui.cpp
===================================================================
--- src/order_gui.cpp	(revision 12871)
+++ src/order_gui.cpp	(working copy)
@@ -498,6 +498,7 @@
 			(facil = FACIL_TRUCK_STOP, 1);
 			if (st->facilities & facil) {
 				order.MakeGoToStation(st_index);
+				if (_ctrl_pressed) order.SetLoadType(OLFB_FULL_LOAD);
 				if (_patches.new_nonstop && (v->type == VEH_TRAIN || v->type == VEH_ROAD)) order.SetNonStopType(ONSF_NO_STOP_AT_INTERMEDIATE_STATIONS);
 				return order;
 			}
@@ -542,7 +543,7 @@
 
 	if (DoCommandP(v->tile, v->index + (OrderGetSel(w) << 16), cmd.Pack(), NULL, CMD_INSERT_ORDER | CMD_MSG(STR_8833_CAN_T_INSERT_NEW_ORDER))) {
 		if (WP(w, order_d).sel != -1) WP(w, order_d).sel++;
-		ResetObjectToPlace();
+		if (false) ResetObjectToPlace();	// todo: patch option
 	}
 }
