Index: src/industry_gui.cpp
===================================================================
--- src/industry_gui.cpp	(revision 17411)
+++ src/industry_gui.cpp	(working copy)
@@ -994,7 +994,7 @@
 			}
 
 			case IDW_INDUSTRY_LIST: {
-				Dimension d = {0, 0};
+				Dimension d = {0, FONT_HEIGHT_NORMAL}; // Give an height estimation, otherwise stuff goes haywire
 				for (uint i = 0; i < this->industries.Length(); i++) {
 					d = maxdim(d, GetStringBoundingBox(this->GetIndustryString(this->industries[i])));
 				}
Index: src/town_gui.cpp
===================================================================
--- src/town_gui.cpp	(revision 17411)
+++ src/town_gui.cpp	(working copy)
@@ -663,7 +663,7 @@
 				break;
 			}
 			case TDW_CENTERTOWN: {
-				Dimension d = {0, 0};
+				Dimension d = {0, FONT_HEIGHT_NORMAL}; // Give an height estimation, otherwise stuff goes haywire
 				for (uint i = 0; i < this->towns.Length(); i++) {
 					const Town *t = this->towns[i];
 
