Index: src/industry_map.h
===================================================================
--- src/industry_map.h	(revision 27170)
+++ src/industry_map.h	(working copy)
@@ -280,7 +280,7 @@
 static inline void MakeIndustry(TileIndex t, IndustryID index, IndustryGfx gfx, uint8 random, WaterClass wc)
 {
 	SetTileType(t, MP_INDUSTRY);
-	_m[t].m1 = 0;
+	_m[t].m1 &= 1<<4; // Make sure bit 4 is unaffected.
 	_m[t].m2 = index;
 	SetIndustryRandomBits(t, random); // m3
 	_m[t].m4 = 0;
