diff --git a/src/newgrf_engine.cpp b/src/newgrf_engine.cpp
--- a/src/newgrf_engine.cpp
+++ b/src/newgrf_engine.cpp
@@ -648,7 +648,8 @@
 		case 0x4A: {
 			if (v->type != VEH_TRAIN) return 0;
 			RailType rt = GetTileRailType(v->tile);
-			return (HasPowerOnRail(Train::From(v)->railtype, rt) ? 0x100 : 0) | GetReverseRailTypeTranslation(rt, object->grffile);
+			const Train *t = Train::From(v);
+			return (t->IsEngine() || (t->IsWagon() && HasBit(t->vehicle_flags, VRF_POWEREDWAGON) && HasPowerOnRail(t->railtype, rt)) ? 0x100 : 0) | GetReverseRailTypeTranslation(rt, object->grffile);
 		}
 
 		/* Variables which use the parameter */
