Index: src/industry_cmd.cpp
===================================================================
--- src/industry_cmd.cpp	(revision 11266)
+++ src/industry_cmd.cpp	(working copy)
@@ -1185,6 +1185,7 @@
 
 bool IsSlopeRefused(Slope current, Slope refused)
 {
+	if (IsSteepSlope(current)) return true;
 	if (current != SLOPE_FLAT) {
 		if (refused & SLOPE_STEEP) return true;
 
@@ -1242,7 +1243,6 @@
 					if (IsClearWaterTile(cur_tile)) return false;
 
 					tileh = GetTileSlope(cur_tile, NULL);
-					if (IsSteepSlope(tileh)) return false;
 
 					refused_slope |= IsSlopeRefused(tileh, its->slopes_refused);
 				}
