Index: media/openttd.desktop.in
===================================================================
--- media/openttd.desktop.in	(revision 18495)
+++ media/openttd.desktop.in	(working copy)
@@ -2,10 +2,9 @@
 # http://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-1.1.html
 [Desktop Entry]
 Type=Application
-Version=1.1
 Name=!!MENU_NAME!!
 Comment=A clone of Transport Tycoon Deluxe
-Icon=openttd
+Icon=!!TTD!!
 Exec=!!TTD!!
 Terminal=false
 Categories=!!MENU_GROUP!!
