Index: split/flags.pcx
===================================================================
Cannot display: file marked as a binary type.
svn:mime-type = application/octet-stream
Index: split/flags.nfo
===================================================================
--- split/flags.nfo	(revision 12364)
+++ split/flags.nfo	(working copy)
@@ -2,7 +2,7 @@
 // $Id$
 //
    -1 * 0	 0C "Flag graphics"
-   -1 * 0	 05 14 1D
+   -1 * 0	 05 14 24
    -1 sprites/flags.pcx 18 8 09 8 11 0 0
    -1 sprites/flags.pcx 34 8 09 8 11 0 0
    -1 sprites/flags.pcx 50 8 09 8 11 0 0
@@ -32,3 +32,10 @@
    -1 sprites/flags.pcx 434 8 09 8 11 0 0
    -1 sprites/flags.pcx 450 8 09 8 11 0 0
    -1 sprites/flags.pcx 466 8 09 8 11 0 0
+   -1 sprites/flags.pcx 482 8 09 8 11 0 0
+   -1 sprites/flags.pcx 498 8 09 8 11 0 0
+   -1 sprites/flags.pcx 514 8 09 8 11 0 0
+   -1 sprites/flags.pcx 530 8 09 8 11 0 0
+   -1 sprites/flags.pcx 546 8 09 8 11 0 0
+   -1 sprites/flags.pcx 562 8 09 8 11 0 0
+   -1 sprites/flags.pcx 578 8 09 8 11 0 0
