diff --git a/src/vehicle.cpp b/src/vehicle.cpp
index d567595..33d581c 100644
--- a/src/vehicle.cpp
+++ b/src/vehicle.cpp
@@ -927,6 +927,9 @@ void AgeVehicle(Vehicle *v)
        /* Don't warn if a renew is active */
        if (GetCompany(v->owner)->engine_renew && GetEngine(v->engine_type)->company_avail != 0) return;
 
+       /* Don't warn if vehicle is destroyed */
+       if (v->vehstatus & VS_CRASHED) return;
+
        StringID str;
        if (age == -DAYS_IN_LEAP_YEAR) {
                str = STR_01A0_IS_GETTING_OLD;
