Index: src/spriteloader/png.cpp
===================================================================
--- src/spriteloader/png.cpp	(révision 19500)
+++ src/spriteloader/png.cpp	(copie de travail)
@@ -132,7 +132,10 @@
 		}
 
 		if (colour_type == PNG_COLOR_TYPE_RGB) {
-			png_set_filler(png_ptr, 0xff, PNG_FILLER_AFTER);
+			if (png_get_valid(png_ptr, info_ptr, PNG_INFO_tRNS))
+				png_set_tRNS_to_alpha(png_ptr);
+			else
+				png_set_filler(png_ptr, 0xff, PNG_FILLER_AFTER);
 		}
 
 		pixelsize = sizeof(uint32);
