Subject: Add game event about ships arriving at a station.


diff --git a/src/ship_cmd.cpp b/src/ship_cmd.cpp
index de1bf4f..771863a 100644
--- a/src/ship_cmd.cpp
+++ b/src/ship_cmd.cpp
@@ -28,6 +28,7 @@
 #include "vehicle_func.h"
 #include "sound_func.h"
 #include "ai/ai.hpp"
+#include "game/game.hpp"
 #include "pathfinder/opf/opf_ship.h"
 #include "engine_base.h"
 #include "company_base.h"
@@ -421,6 +422,7 @@ static void ShipArrivesAt(const Vehicle *v, Station *st)
 			st->index
 		);
 		AI::NewEvent(v->owner, new ScriptEventStationFirstVehicle(st->index, v->index));
+		Game::NewEvent(new ScriptEventStationFirstVehicle(st->index, v->index));
 	}
 }
 
