Index: src/tunnelbridge_cmd.cpp
===================================================================
--- src/tunnelbridge_cmd.cpp	(revision 21141)
+++ src/tunnelbridge_cmd.cpp	(working copy)
@@ -1496,7 +1496,7 @@
  * When leaving a tunnel, show the train when it is one frame further
  * to the 'outside', i.e. at (TILE_SIZE-1) - (frame) + 1
  */
-static const byte _train_tunnel_frame[DIAGDIR_END] = {14, 9, 7, 12};
+static const byte _train_tunnel_frame[DIAGDIR_END] = {12, 12, 12, 12};
 
 /**
  * Frame when a road vehicle enters a tunnel with a certain direction.
@@ -1504,7 +1504,7 @@
  * the entry and exit frames are not consistent. This is the entry frame,
  * the road vehicle should be hidden when it reaches this frame.
  */
-static const byte _road_enter_tunnel_frame[DIAGDIR_END] = {13, 8, 8, 13};
+static const byte _road_enter_tunnel_frame[DIAGDIR_END] = {12, 12, 12, 12};
 
 /**
  * Frame when a road vehicle exits a tunnel with a certain direction.
@@ -1512,7 +1512,7 @@
  * vehicle direction. As stated above, this frame is not the same as the
  * entry frame, for unclear (historical?) reasons.
  */
-static const byte _road_exit_tunnel_frame[DIAGDIR_END] = {2, 7, 9, 4};
+static const byte _road_exit_tunnel_frame[DIAGDIR_END] = {3, 3, 3, 3};
 
 static VehicleEnterTileStatus VehicleEnter_TunnelBridge(Vehicle *v, TileIndex tile, int x, int y)
 {
