Index: src/tunnelbridge_cmd.cpp
===================================================================
--- src/tunnelbridge_cmd.cpp	(revision 20896)
+++ src/tunnelbridge_cmd.cpp	(working copy)
@@ -877,8 +877,8 @@
 	const int h = bounding_box_size[axis ^ 1];
 	int x_back = x - back_pillar_offset[axis];
 	int y_back = y - back_pillar_offset[axis ^ 1];
-	/* skip 2 top rows for back facing pillars hidden by bridge, they are unseen anyway */
-	int z_back = z_bridge - TILE_HEIGHT * 2;
+	/* skip 2 top rows for back facing pillars hidden by bridge, they are unseen anyway (flags bit 1 not set) */
+	int z_back = (HasBit(flags, 1)) ? z_bridge : z_bridge - TILE_HEIGHT * 2;
 	/* back_height higher than z_bridge prevents drawing far pillar (flags bit 0) */
 	int back_height = (HasBit(flags, 0)) ? z_bridge + 1 : ti->z;
 	int front_height = ti->z;
