diff -r ba8c63598f72 bananas/views.py
--- a/bananas/views.py	Sun Oct 28 12:13:06 2012 +0000
+++ b/bananas/views.py	Fri Nov 02 21:23:06 2012 +0100
@@ -82,10 +82,18 @@
 		if scanning == 2 and line.find("return") != -1:
 			line = line[line.find("return"):]
 			scanning = 3
+		if scanning == 3 and line.find('/*') != -1: # watch out for // ".."
+			continue
+		if scanning == 3 and line.find('/*') != -1: # watch out for /*".."*/
+			line = line[line.find('/*') + 1:]
+			scanning = 4
+		if scanning == 4 and line.find('*/') != -1:
+			line = line[line.find('*/') + 1:]
+			scanning = 3 # return to looking for "
 		if scanning == 3 and line.find('"') != -1:
 			line = line[line.find('"') + 1:]
-			scanning = 4
-		if scanning == 4 and line.find('"') != -1:
+			scanning = 5
+		if scanning == 5 and line.find('"') != -1:
 			shortName = line[:line.find('"')]
 			break
 
