Index: config.lib
===================================================================
--- config.lib	(revision 8903)
+++ config.lib	(working copy)
@@ -779,7 +779,7 @@
 
 		if [ "$enable_static" != "0" ]; then
 			if [ "$os" = "OSX" ]; then
-				LIBS="$LIBS `$fontconfig_config --prefix`/lib/libfontconfig.a"
+				LIBS="$LIBS `$fontconfig_config --variable=prefix`/lib/libfontconfig.a"
 			else
 				LIBS="$LIBS `$fontconfig_config --libs --static | tr '\n\r' '  '`"
 			fi
