Index: src/currency.cpp
===================================================================
--- src/currency.cpp	(revision 9815)
+++ src/currency.cpp	(working copy)
@@ -35,7 +35,7 @@
 	{    3, ',', 2002,      "NLG ", "",      0,  STR_CURR_NLG    }, ///< dutch gulden
 	{   12, '.', CF_NOEURO, "",     " Kr",   1,  STR_CURR_NOK    }, ///< norwegian krone
 	{    6, ' ', CF_NOEURO, "",     " zl",   1,  STR_CURR_PLN    }, ///< polish zloty
-	{    5, '.', CF_NOEURO, "",     " Lei",  1,  STR_CURR_ROL    }, ///< romanian Lei
+	{    5, '.', 2010,      "",     " Lei",  1,  STR_CURR_RON    }, ///< romanian Lei
 	{   50, ' ', CF_NOEURO, "",     " p",    1,  STR_CURR_RUR    }, ///< russian rouble
 	{  352, '.', CF_NOEURO, "",     " SIT",  1,  STR_CURR_SIT    }, ///< slovenian tolar
 	{   13, '.', CF_NOEURO, "",     " Kr",   1,  STR_CURR_SEK    }, ///< swedish krona
@@ -75,7 +75,7 @@
 	CURR_NLG,
 	CURR_NOK,
 	CURR_PLN,
-	CURR_ROL,
+	CURR_RON,
 	CURR_RUR,
 	CURR_SIT,
 	CURR_SEK,
