Index: src/train_cmd.cpp
===================================================================
--- src/train_cmd.cpp	(revision 16069)
+++ src/train_cmd.cpp	(working copy)
@@ -3775,6 +3775,7 @@
 							{TRACK_BIT_NONE,  TRACK_BIT_RIGHT, TRACK_BIT_Y,     TRACK_BIT_UPPER},
 							{TRACK_BIT_RIGHT, TRACK_BIT_NONE,  TRACK_BIT_LOWER, TRACK_BIT_X    }
 						};
+						DEBUG(misc, 0, "%u %u",gp.new_tile,TileVirtXY(prev->x_pos, prev->y_pos));
 						DiagDirection exitdir = DiagdirBetweenTiles(gp.new_tile, prev->tile);
 						assert(IsValidDiagDirection(exitdir));
 						chosen_track = _connecting_track[enterdir][exitdir];
