Index: src/order_gui.cpp
===================================================================
--- src/order_gui.cpp	(revision 13061)
+++ src/order_gui.cpp	(working copy)
@@ -1183,6 +1183,7 @@
 
 	DeleteWindowById(WC_VEHICLE_ORDERS, veh);
 	DeleteWindowById(WC_VEHICLE_DETAILS, veh);
+	DeleteWindowById(WC_VEHICLE_TIMETABLE, veh);
 
 	if (v->owner != _local_player) {
 		w = AllocateWindowDescFront<Window>(&_other_orders_desc, veh);
Index: src/timetable_gui.cpp
===================================================================
--- src/timetable_gui.cpp	(revision 13061)
+++ src/timetable_gui.cpp	(working copy)
@@ -8,6 +8,7 @@
 #include "command_func.h"
 #include "gui.h"
 #include "window_gui.h"
+#include "window_func.h"
 #include "textbuf_gui.h"
 #include "cargotype.h"
 #include "strings_func.h"
@@ -24,7 +25,6 @@
 	TTV_WIDGET_CLOSEBOX = 0,
 	TTV_CAPTION,
 	TTV_ORDER_VIEW,
-	TTV_STICKY,
 	TTV_TIMETABLE_PANEL,
 	TTV_SCROLLBAR,
 	TTV_SUMMARY_PANEL,
@@ -274,8 +274,7 @@
 static const Widget _timetable_widgets[] = {
 	{   WWT_CLOSEBOX,   RESIZE_NONE,    14,     0,    10,     0,    13, STR_00C5,                   STR_018B_CLOSE_WINDOW},                // TTV_WIDGET_CLOSEBOX
 	{    WWT_CAPTION,   RESIZE_RIGHT,   14,    11,   326,     0,    13, STR_TIMETABLE_TITLE,        STR_018C_WINDOW_TITLE_DRAG_THIS},      // TTV_CAPTION
-	{ WWT_PUSHTXTBTN,   RESIZE_LR,      14,   327,   387,     0,    13, STR_ORDER_VIEW,             STR_ORDER_VIEW_TOOLTIP},               // TTV_ORDER_VIEW
-	{  WWT_STICKYBOX,   RESIZE_LR,      14,   388,   399,     0,    13, STR_NULL,                   STR_STICKY_BUTTON},                    // TTV_STICKY
+	{ WWT_PUSHTXTBTN,   RESIZE_LR,      14,   327,   399,     0,    13, STR_ORDER_VIEW,             STR_ORDER_VIEW_TOOLTIP},               // TTV_ORDER_VIEW
 
 	{      WWT_PANEL,   RESIZE_RB,      14,     0,   387,    14,    95, STR_NULL,                   STR_TIMETABLE_TOOLTIP},                // TTV_TIMETABLE_PANEL
 	{  WWT_SCROLLBAR,   RESIZE_LRB,     14,   388,   399,    14,    95, STR_NULL,                   STR_0190_SCROLL_BAR_SCROLLS_LIST},     // TTV_SCROLLBAR
@@ -296,13 +295,16 @@
 static const WindowDesc _timetable_desc = {
 	WDP_AUTO, WDP_AUTO, 400, 130, 400, 130,
 	WC_VEHICLE_TIMETABLE, WC_VEHICLE_VIEW,
-	WDF_STD_TOOLTIPS | WDF_STD_BTN | WDF_DEF_WIDGET | WDF_UNCLICK_BUTTONS | WDF_STICKY_BUTTON | WDF_RESIZABLE,
+	WDF_STD_TOOLTIPS | WDF_STD_BTN | WDF_DEF_WIDGET | WDF_UNCLICK_BUTTONS | WDF_RESIZABLE,
 	_timetable_widgets,
 	TimetableWndProc
 };
 
 void ShowTimetableWindow(const Vehicle *v)
 {
+	DeleteWindowById(WC_VEHICLE_ORDERS, v->index);
+	DeleteWindowById(WC_VEHICLE_DETAILS, v->index);
+	DeleteWindowById(WC_VEHICLE_TIMETABLE, v->index);
 	Window *w = AllocateWindowDescFront<Window>(&_timetable_desc, v->index);
 
 	if (w != NULL) {
Index: src/vehicle_gui.cpp
===================================================================
--- src/vehicle_gui.cpp	(revision 13061)
+++ src/vehicle_gui.cpp	(working copy)
@@ -1663,6 +1663,7 @@
 {
 	DeleteWindowById(WC_VEHICLE_ORDERS, v->index);
 	DeleteWindowById(WC_VEHICLE_DETAILS, v->index);
+	DeleteWindowById(WC_VEHICLE_TIMETABLE, v->index);
 	AllocateWindowDescFront<Window>(&_vehicle_details_desc, v->index);
 }
 
