Index: src/settings.cpp
===================================================================
--- src/settings.cpp	(revision 13435)
+++ src/settings.cpp	(working copy)
@@ -1744,6 +1744,9 @@
 	     SDT_VAR(GameSettings, game_creation.map_x,                           SLE_UINT8,                     S, 0,     8,                     6,      11, 0, STR_CONFIG_PATCHES_MAP_X,                 NULL),
 	     SDT_VAR(GameSettings, game_creation.map_y,                           SLE_UINT8,                     S, 0,     8,                     6,      11, 0, STR_CONFIG_PATCHES_MAP_Y,                 NULL),
 
+	SDT_CONDBOOL(GameSettings, construction.disable_tree_growth,                         98, SL_MAX_VERSION, 0, 0, false,                                   STR_CONFIG_PATCHES_TREE_GROWTH,            NULL),
+
+
  SDT_CONDOMANY(GameSettings, locale.currency,                               SLE_UINT8, 97, SL_MAX_VERSION, N, 0, 0, CUSTOM_CURRENCY_ID, "GBP|USD|EUR|YEN|ATS|BEF|CHF|CZK|DEM|DKK|ESP|FIM|FRF|GRD|HUF|ISK|ITL|NLG|NOK|PLN|ROL|RUR|SIT|SEK|YTL|SKK|BRR|custom", STR_NULL, NULL, NULL),
  SDT_CONDOMANY(GameSettings, locale.units,                                  SLE_UINT8, 97, SL_MAX_VERSION, N, 0, 1, 2, "imperial|metric|si", STR_NULL, NULL, NULL),
 
Index: src/lang/english.txt
===================================================================
--- src/lang/english.txt	(revision 13435)
+++ src/lang/english.txt	(working copy)
@@ -1179,6 +1179,9 @@
 STR_CONFIG_PATCHES_TOWN_LAYOUT_3X3_GRID                         :3x3 grid
 STR_CONFIG_PATCHES_TOWN_LAYOUT_RANDOM                           :random
 
+STR_CONFIG_PATCHES_TREE_GROWTH                                  :{LTBLUE}Disable Tree Growth: {ORANGE}{STRING1}
+
+
 STR_CONFIG_PATCHES_TOOLBAR_POS                                  :{LTBLUE}Position of main toolbar: {ORANGE}{STRING1}
 STR_CONFIG_PATCHES_TOOLBAR_POS_LEFT                             :Left
 STR_CONFIG_PATCHES_TOOLBAR_POS_CENTER                           :Centre
Index: src/settings_gui.cpp
===================================================================
--- src/settings_gui.cpp	(revision 13435)
+++ src/settings_gui.cpp	(working copy)
@@ -625,6 +625,7 @@
 	"gui.drag_signals_density",
 	"game_creation.oil_refinery_limit",
 	"gui.semaphore_build_before",
+	"construction.disable_tree_growth",
 };
 
 static const char *_patches_stations[] = {
Index: src/tree_cmd.cpp
===================================================================
--- src/tree_cmd.cpp	(revision 13435)
+++ src/tree_cmd.cpp	(working copy)
@@ -623,6 +623,8 @@
 
 static void TileLoop_Trees(TileIndex tile)
 {
+	if (_settings_game.construction.disable_tree_growth) return;
+
 	if (GetTreeGround(tile) == TREE_GROUND_SHORE) {
 		TileLoop_Water(tile);
 	} else {
@@ -724,6 +726,8 @@
 	TileIndex tile;
 	TreeType tree;
 
+	if (_settings_game.construction.disable_tree_growth) return;
+
 	/* place a tree at a random rainforest spot */
 	if (_settings_game.game_creation.landscape == LT_TROPIC &&
 			(r = Random(), tile = RandomTileSeed(r), GetTropicZone(tile) == TROPICZONE_RAINFOREST) &&
Index: src/settings_type.h
===================================================================
--- src/settings_type.h	(revision 13435)
+++ src/settings_type.h	(working copy)
@@ -148,6 +148,7 @@
 	bool   extra_dynamite;                   ///< extra dynamite
 	bool   road_stop_on_town_road;           ///< allow building of drive-through road stops on town owned roads
 	uint8  raw_industry_construction;        ///< type of (raw) industry construction (none, "normal", prospecting)
+	bool   disable_tree_growth;              ///< If trees are allowed to grow
 };
 
 /** Settings related to the AI. */
