Index: media/openttd.menu.in
===================================================================
--- media/openttd.menu.in	(Revision 0)
+++ media/openttd.menu.in	(Revision 0)
@@ -0,0 +1,6 @@
+?package(openttd):\
+needs="X11"\
+section="Games/Simulation"\
+title="!!MENU_NAME!!"\
+command="!!BINARY_DIR!!/!!BINARY_NAME!!"\
+icon="!!ICON_DIR!!/!!BINARY_NAME!!.32.xpm"
Index: Makefile.in
===================================================================
--- Makefile.in	(Revision 14505)
+++ Makefile.in	(Arbeitskopie)
@@ -121,7 +121,7 @@
 	done
 	$(Q)rm -rf objs
 	$(Q)rm -f Makefile Makefile.am Makefile.bundle
-	$(Q)rm -f media/openttd.desktop
+	$(Q)rm -f media/openttd.desktop media/openttd.menu
 	$(Q)rm -f $(CONFIG_CACHE_SOURCE_LIST) config.cache config.pwd config.log $(CONFIG_CACHE_PWD)
 	$(Q)rm -rf $(BUNDLE_DIR)
 	$(Q)rm -rf $(BUNDLES_DIR)
Index: Makefile.bundle.in
===================================================================
--- Makefile.bundle.in	(Revision 14505)
+++ Makefile.bundle.in	(Arbeitskopie)
@@ -70,6 +70,7 @@
 	$(Q)cp "$(ROOT_DIR)/media/openttd.32.xpm" "$(BUNDLE_DIR)/media/"
 	$(Q)cp "$(ROOT_DIR)/media/openttd."*.png  "$(BUNDLE_DIR)/media/"
 ifdef MENU_DIR
+	$(Q)cp "$(ROOT_DIR)/media/openttd.menu" "$(BUNDLE_DIR)/media/"
 	$(Q)cp "$(ROOT_DIR)/media/openttd.desktop" "$(BUNDLE_DIR)/media/"
 	$(Q)cat "$(ROOT_DIR)/media/openttd.desktop" | sed s/=openttd/=$(BINARY_NAME)/g > "$(ROOT_DIR)/media/openttd.desktop.install"
 endif
Index: config.lib
===================================================================
--- config.lib	(Revision 14505)
+++ config.lib	(Arbeitskopie)
@@ -2505,6 +2505,18 @@
 	< $ROOT_DIR/media/openttd.desktop.in sed "$MENU_REPLACE" > media/openttd.desktop
 }
 
+generate_menu_info() {
+	INFO_REPLACE="
+		s~!!BINARY_NAME!!~$binary_name~g;
+		s~!!BINARY_DIR!!~$prefix_dir/$binary_dir~g;
+		s~!!ICON_DIR!!~$prefix_dir/$icon_dir~g;
+		s~!!MENU_NAME!!~$menu_name~g
+	"
+	echo "Generating menu info.."
+	mkdir -p media
+	< $ROOT_DIR/media/openttd.menu.in sed "$INFO_REPLACE" > media/openttd.menu
+}
+
 generate_main() {
 	STAGE="[MAIN]"
 
@@ -2527,6 +2539,7 @@
 
 	if [ "$menu_dir" != "" ]; then
 		generate_menu_item
+		generate_menu_info
 	fi
 }
 
