Index: src/pathfinder/yapf/yapf_costrail.hpp
===================================================================
--- src/pathfinder/yapf/yapf_costrail.hpp	(revision 24978)
+++ src/pathfinder/yapf/yapf_costrail.hpp	(working copy)
@@ -524,13 +524,6 @@
 				}
 			}
 
-			/* Check the next tile for the rail type. */
-			if (next.rail_type != cur.rail_type) {
-				/* Segment must consist from the same rail_type tiles. */
-				end_segment_reason |= ESRB_RAIL_TYPE;
-				break;
-			}
-
 			/* Avoid infinite looping. */
 			if (next.tile == n.m_key.m_tile && next.td == n.m_key.m_td) {
 				end_segment_reason |= ESRB_INFINITE_LOOP;
