diff --git a/src/script/api/ai_changelog.hpp b/src/script/api/ai_changelog.hpp
index 91429fc..aabe962 100644
--- a/src/script/api/ai_changelog.hpp
+++ b/src/script/api/ai_changelog.hpp
@@ -215,7 +215,7 @@
  *     AIBaseStation, but can still be used as AIStation.GetConstructionDate
  * \li WaypointID was replaced by StationID. All WaypointIDs from previous
  *     savegames are invalid. Use STATION_INVALID instead of WAYPOINT_INVALID
- * \li AIWaypointList constructor now needs a WaypointType similiar to AIStationList,
+ * \li AIWaypointList constructor now needs a WaypointType similar to AIStationList,
  *     it can also handle buoys.
  * \li AIVehicleList_Station now also works for waypoints
  * \li Stations can be build over rail without signals that is in the right
diff --git a/src/script/api/script_company.hpp b/src/script/api/script_company.hpp
index 6fcf01d..e6c3dcb 100644
--- a/src/script/api/script_company.hpp
+++ b/src/script/api/script_company.hpp
@@ -172,13 +172,13 @@ public:
 	 * @param quarter The quarter to get the income of.
 	 * @pre ResolveCompanyID(company) != COMPANY_INVALID.
 	 * @pre quarter <= EARLIEST_QUARTER.
-	 * @return The bruto income of the company in the given quarter.
+	 * @return The gross income of the company in the given quarter.
 	 */
 	static Money GetQuarterlyIncome(CompanyID company, uint32 quarter);
 
 	/**
 	 * Get the expenses of the company in the given quarter.
-	 * Note that this function only considers recurring expensens from vehicle
+	 * Note that this function only considers recurring expenses from vehicle
 	 * running cost, maintenance and interests; it does not include one-time
 	 * expenses from construction and buying stuff.
 	 * @param company The company to get the quarterly expenses of.
@@ -223,7 +223,7 @@ public:
 
 	/**
 	 * Build your company's HQ on the given tile.
-	 * @param tile The tile to build your HQ on, this tile is the most nothern tile of your HQ.
+	 * @param tile The tile to build your HQ on, this tile is the most northern tile of your HQ.
 	 * @pre ScriptMap::IsValidTile(tile).
 	 * @game @pre Valid ScriptCompanyMode active in scope.
 	 * @exception ScriptError::ERR_AREA_NOT_CLEAR
@@ -238,7 +238,7 @@ public:
 	 * Return the location of a company's HQ.
 	 * @param company The company the get the HQ of.
 	 * @pre ResolveCompanyID(company) != COMPANY_INVALID.
-	 * @return The tile of the company's HQ, this tile is the most nothern tile
+	 * @return The tile of the company's HQ, this tile is the most northern tile
 	 *  of that HQ, or ScriptMap::TILE_INVALID if there is no HQ yet.
 	 */
 	static TileIndex GetCompanyHQ(CompanyID company);
diff --git a/src/script/api/script_controller.hpp b/src/script/api/script_controller.hpp
index f9eeed7..20a849f 100644
--- a/src/script/api/script_controller.hpp
+++ b/src/script/api/script_controller.hpp
@@ -129,7 +129,7 @@ public:
 	 * Import a library.
 	 * @param library The name of the library to import.
 	 * @param class_name Under which name you want it to be available (or "" if you just want the returning object).
-	 * @param version Which version you want specificly.
+	 * @param version Which version you want specifically.
 	 * @return The loaded library object. If class_name is set, it is also available (under the scope of the import) under that name.
 	 * @note This command can be called from the global space, and does not need an instance.
 	 */
diff --git a/src/script/api/script_engine.hpp b/src/script/api/script_engine.hpp
index c5525ba..058d230 100644
--- a/src/script/api/script_engine.hpp
+++ b/src/script/api/script_engine.hpp
@@ -255,7 +255,7 @@ public:
 
 	/**
 	 * Get the maximum allowed distance between two orders for an engine.
-	 * The distance returned is a vehicle-type specific distance indepenent from other
+	 * The distance returned is a vehicle-type specific distance independent from other
 	 * map distances, you may use the result of this function to compare it
 	 * with the result of ScriptOrder::GetOrderDistance.
 	 * @param engine_id The engine to get the max distance for.
diff --git a/src/script/api/script_error.hpp b/src/script/api/script_error.hpp
index ccbd8b1..33a69e5 100644
--- a/src/script/api/script_error.hpp
+++ b/src/script/api/script_error.hpp
@@ -129,7 +129,7 @@ public:
 	/**
 	 * Check the membership of the last thrown error.
 	 * @return The category the error belongs to.
-	 * @note The last throw error can be aquired by calling GetLastError().
+	 * @note The last throw error can be acquired by calling GetLastError().
 	 */
 	static ErrorCategories GetErrorCategory();
 
diff --git a/src/script/api/script_event_types.hpp b/src/script/api/script_event_types.hpp
index f7b58c5..6283411 100644
--- a/src/script/api/script_event_types.hpp
+++ b/src/script/api/script_event_types.hpp
@@ -141,7 +141,7 @@ private:
 };
 
 /**
- * Event Subidy Awarded, indicating a subsidy is awarded to some company.
+ * Event Subsidy Awarded, indicating a subsidy is awarded to some company.
  * @api ai game
  */
 class ScriptEventSubsidyAwarded : public ScriptEvent {
@@ -168,7 +168,7 @@ public:
 	SubsidyID GetSubsidyID() { return this->subsidy_id; }
 
 private:
-	SubsidyID subsidy_id; ///< The subsidy that was awared.
+	SubsidyID subsidy_id; ///< The subsidy that was awarded.
 };
 
 /**
@@ -412,7 +412,7 @@ class ScriptEventCompanyMerger : public ScriptEvent {
 public:
 	/**
 	 * @param old_owner The company bought off.
-	 * @param new_owner The company that bougth owner.
+	 * @param new_owner The company that bought owner.
 	 */
 	ScriptEventCompanyMerger(Owner old_owner, Owner new_owner) :
 		ScriptEvent(ET_COMPANY_MERGER),
@@ -432,7 +432,7 @@ public:
 	 * @return The CompanyID of the company that has been bought.
 	 * @note: The value below is not valid anymore as CompanyID, and
 	 *  ScriptCompany::ResolveCompanyID will return COMPANY_COMPANY. It's
-	 *  only usefull if you're keeping track of company's yourself.
+	 *  only useful if you're keeping track of company's yourself.
 	 */
 	ScriptCompany::CompanyID GetOldCompanyID() { return this->old_owner; }
 
@@ -700,7 +700,7 @@ public:
 	VehicleID GetVehicleID() { return this->vehicle; }
 
 private:
-	StationID station; ///< The station the vehicle arived at.
+	StationID station; ///< The station the vehicle arrived at.
 	VehicleID vehicle; ///< The vehicle that arrived at the station.
 };
 
diff --git a/src/script/api/script_execmode.hpp b/src/script/api/script_execmode.hpp
index 1ac7565..6591399 100644
--- a/src/script/api/script_execmode.hpp
+++ b/src/script/api/script_execmode.hpp
@@ -25,7 +25,7 @@
 class ScriptExecMode : public ScriptObject {
 private:
 	ScriptModeProc *last_mode;   ///< The previous mode we were in.
-	ScriptObject *last_instance; ///< The previous instace of the mode.
+	ScriptObject *last_instance; ///< The previous instance of the mode.
 
 protected:
 	/**
diff --git a/src/script/api/script_game.hpp b/src/script/api/script_game.hpp
index 3b66ba1..3531c9d 100644
--- a/src/script/api/script_game.hpp
+++ b/src/script/api/script_game.hpp
@@ -26,7 +26,7 @@ public:
 	 */
 	enum LandscapeType {
 		/* Note: these values represent part of the in-game LandscapeType enum */
-		LT_TEMPERATE  = ::LT_TEMPERATE, ///< Temperate climiate.
+		LT_TEMPERATE  = ::LT_TEMPERATE, ///< Temperate climate.
 		LT_ARCTIC     = ::LT_ARCTIC,    ///< Arctic climate.
 		LT_TROPIC     = ::LT_TROPIC,    ///< Tropic climate.
 		LT_TOYLAND    = ::LT_TOYLAND,   ///< Toyland climate.
diff --git a/src/script/api/script_gamesettings.hpp b/src/script/api/script_gamesettings.hpp
index ad718f3..6f92ca7 100644
--- a/src/script/api/script_gamesettings.hpp
+++ b/src/script/api/script_gamesettings.hpp
@@ -43,7 +43,7 @@ public:
 	 * @warning Results of this function are not governed by the API. This means
 	 *          that a setting that previously existed can be gone or has
 	 *          changed its name.
-	 * @note Results achieved in the past offer no gurantee for the future.
+	 * @note Results achieved in the past offer no guarantee for the future.
 	 * @return True if and only if the setting is valid.
 	 */
 	static bool IsValid(const char *setting);
@@ -56,7 +56,7 @@ public:
 	 *          that the value of settings may be out of the expected range. It
 	 *          also means that a setting that previously existed can be gone or
 	 *          has changed its name/characteristics.
-	 * @note Results achieved in the past offer no gurantee for the future.
+	 * @note Results achieved in the past offer no guarantee for the future.
 	 * @return The value for the setting.
 	 */
 	static int32 GetValue(const char *setting);
@@ -67,7 +67,7 @@ public:
 	 * @param value The value to set the setting to.
 	 * @pre IsValid(setting).
 	 * @return True if the action succeeded.
-	 * @note Results achieved in the past offer no gurantee for the future.
+	 * @note Results achieved in the past offer no guarantee for the future.
 	 * @api -ai
 	 */
 	static bool SetValue(const char *setting, int value);
diff --git a/src/script/api/script_list.cpp b/src/script/api/script_list.cpp
index e2afbff..353d25d 100644
--- a/src/script/api/script_list.cpp
+++ b/src/script/api/script_list.cpp
@@ -835,7 +835,7 @@ SQInteger ScriptList::Valuate(HSQUIRRELVM vm)
 			return SQ_ERROR;
 		}
 
-		/* Retreive the return value */
+		/* Retrieve the return value */
 		SQInteger value;
 		switch (sq_gettype(vm, -1)) {
 			case OT_INTEGER: {
diff --git a/src/script/api/script_list.hpp b/src/script/api/script_list.hpp
index 8799282..88b0d6f 100644
--- a/src/script/api/script_list.hpp
+++ b/src/script/api/script_list.hpp
@@ -164,7 +164,7 @@ public:
 	/**
 	 * Removes all items with a value above start and below end.
 	 * @param start the lower bound of the to be removed values (exclusive).
-	 * @param end   the upper bound of the to be removed valuens (exclusive).
+	 * @param end   the upper bound of the to be removed values (exclusive).
 	 */
 	void RemoveBetweenValue(int32 start, int32 end);
 
diff --git a/src/script/api/script_order.hpp b/src/script/api/script_order.hpp
index 4df6f7e..68702af 100644
--- a/src/script/api/script_order.hpp
+++ b/src/script/api/script_order.hpp
@@ -47,7 +47,7 @@ public:
 
 		/** Do not stop at the stations that are passed when going to the destination. Only for trains and road vehicles. */
 		OF_NON_STOP_INTERMEDIATE = 1 << 0,
-		/** Do not stop at the destionation station. Only for trains and road vehicles. */
+		/** Do not stop at the destination station. Only for trains and road vehicles. */
 		OF_NON_STOP_DESTINATION  = 1 << 1,
 
 		/** Always unload the vehicle; only for stations. Cannot be set when OF_TRANSFER or OF_NO_UNLOAD is set. */
@@ -586,7 +586,7 @@ public:
 	/**
 	 * Get the distance between two points for a vehicle type.
 	 * Use this function to compute the distance between two tiles wrt. a vehicle type.
-	 * These vehicle-type specific distances are indepenent from other map distances, you may
+	 * These vehicle-type specific distances are independent from other map distances, you may
 	 * use the result of this function to compare it with the result of
 	 * ScriptEngine::GetMaximumOrderDistance or ScriptVehicle::GetMaximumOrderDistance.
 	 * @param vehicle_type The vehicle type to get the distance for.
diff --git a/src/script/api/script_rail.hpp b/src/script/api/script_rail.hpp
index 21dfab3..787b5b5 100644
--- a/src/script/api/script_rail.hpp
+++ b/src/script/api/script_rail.hpp
@@ -314,7 +314,7 @@ public:
 	/**
 	 * Remove all rail waypoint pieces within a rectangle on the map.
 	 * @param tile One corner of the rectangle to clear.
-	 * @param tile2 The oppposite corner.
+	 * @param tile2 The opposite corner.
 	 * @param keep_rail Whether to keep the rail after removal.
 	 * @pre IsValidTile(tile).
 	 * @pre IsValidTile(tile2).
@@ -326,7 +326,7 @@ public:
 	/**
 	 * Remove all rail station platform pieces within a rectangle on the map.
 	 * @param tile One corner of the rectangle to clear.
-	 * @param tile2 The oppposite corner.
+	 * @param tile2 The opposite corner.
 	 * @param keep_rail Whether to keep the rail after removal.
 	 * @pre IsValidTile(tile).
 	 * @pre IsValidTile(tile2).
diff --git a/src/script/api/script_road.hpp b/src/script/api/script_road.hpp
index 6f928e2..36290e2 100644
--- a/src/script/api/script_road.hpp
+++ b/src/script/api/script_road.hpp
@@ -147,7 +147,7 @@ public:
 	/**
 	 * Checks whether the given tiles are directly connected, i.e. whether
 	 *  a road vehicle can travel from the center of the first tile to the
-	  * center of the second tile.
+	 * center of the second tile.
 	 * @param tile_from The source tile.
 	 * @param tile_to The destination tile.
 	 * @pre IsRoadTypeAvailable(GetCurrentRoadType()).
diff --git a/src/script/api/script_subsidy.hpp b/src/script/api/script_subsidy.hpp
index bf56505..3069ad9 100644
--- a/src/script/api/script_subsidy.hpp
+++ b/src/script/api/script_subsidy.hpp
@@ -26,7 +26,7 @@ public:
 	 */
 	enum SubsidyParticipantType {
 		/* Values are important, as they represent the internal state of the game.
-		 *  It is orignally named SourceType. ST_HEADQUARTERS is intentionally
+		 *  It is originally named SourceType. ST_HEADQUARTERS is intentionally
 		 *  left out, as it cannot be used for Subsidies. */
 		SPT_INDUSTRY =    0, ///< Subsidy participant is an industry
 		SPT_TOWN     =    1, ///< Subsidy participant is a town
diff --git a/src/script/api/script_testmode.hpp b/src/script/api/script_testmode.hpp
index 5ef3382..4ca29d5 100644
--- a/src/script/api/script_testmode.hpp
+++ b/src/script/api/script_testmode.hpp
@@ -27,7 +27,7 @@
 class ScriptTestMode : public ScriptObject {
 private:
 	ScriptModeProc *last_mode;   ///< The previous mode we were in.
-	ScriptObject *last_instance; ///< The previous instace of the mode.
+	ScriptObject *last_instance; ///< The previous instance of the mode.
 
 protected:
 	/**
diff --git a/src/script/api/script_town.hpp b/src/script/api/script_town.hpp
index c10e1f6..1c544ff 100644
--- a/src/script/api/script_town.hpp
+++ b/src/script/api/script_town.hpp
@@ -228,7 +228,7 @@ public:
 	/**
 	 * Set the amount of days between town growth.
 	 * @param town_id The index of the town.
-	 * @param days_between_town_growth The amont of days between town growth.
+	 * @param days_between_town_growth The amount of days between town growth.
 	 * @pre IsValidTown(town_id).
 	 * @return True if the action succeeded.
 	 * @note If 'Fund Building' and 'economy.town_growth_rate' is active, the game will often set a new GrowthRate.
diff --git a/src/script/api/script_vehicle.hpp b/src/script/api/script_vehicle.hpp
index 9af3aca..f6b22c2 100644
--- a/src/script/api/script_vehicle.hpp
+++ b/src/script/api/script_vehicle.hpp
@@ -370,7 +370,7 @@ public:
 	static bool MoveWagonChain(VehicleID source_vehicle_id, int source_wagon, int dest_vehicle_id, int dest_wagon);
 
 	/**
-	 * Gets the capacity of the given vehicle when refited to the given cargo type.
+	 * Gets the capacity of the given vehicle when refitted to the given cargo type.
 	 * @param vehicle_id The vehicle to refit.
 	 * @param cargo The cargo to refit to.
 	 * @pre IsValidVehicle(vehicle_id).
@@ -549,7 +549,7 @@ public:
 
 	/**
 	 * Get the maximum allowed distance between two orders for a vehicle.
-	 * The distance returned is a vehicle-type specific distance indepenent from other
+	 * The distance returned is a vehicle-type specific distance independent from other
 	 * map distances, you may use the result of this function to compare it
 	 * with the result of ScriptOrder::GetOrderDistance.
 	 * @param vehicle_id The vehicle to get the distance for.
diff --git a/src/script/api/script_viewport.hpp b/src/script/api/script_viewport.hpp
index a3b82a7..3340dc4 100644
--- a/src/script/api/script_viewport.hpp
+++ b/src/script/api/script_viewport.hpp
@@ -15,7 +15,7 @@
 #include "script_object.hpp"
 
 /**
- * Class that manipultes the users viewport.
+ * Class that manipulates the users viewport.
  * @api game
  */
 class ScriptViewport : public ScriptObject {
diff --git a/src/script/api/script_window.hpp b/src/script/api/script_window.hpp
index 3bdc0c1..b4df7a7 100644
--- a/src/script/api/script_window.hpp
+++ b/src/script/api/script_window.hpp
@@ -910,7 +910,7 @@ public:
 		WID_RV_LEFT_MATRIX                           = ::WID_RV_LEFT_MATRIX,                           ///< The matrix on the left.
 		WID_RV_LEFT_SCROLLBAR                        = ::WID_RV_LEFT_SCROLLBAR,                        ///< The scrollbar for the matrix on the left.
 		WID_RV_RIGHT_MATRIX                          = ::WID_RV_RIGHT_MATRIX,                          ///< The matrix on the right.
-		WID_RV_RIGHT_SCROLLBAR                       = ::WID_RV_RIGHT_SCROLLBAR,                       ///< The scrollbar for the matrxi on the right.
+		WID_RV_RIGHT_SCROLLBAR                       = ::WID_RV_RIGHT_SCROLLBAR,                       ///< The scrollbar for the matrix on the right.
 		WID_RV_LEFT_DETAILS                          = ::WID_RV_LEFT_DETAILS,                          ///< Details of the entry on the left.
 		WID_RV_RIGHT_DETAILS                         = ::WID_RV_RIGHT_DETAILS,                         ///< Details of the entry on the right.
 
@@ -2012,10 +2012,10 @@ public:
 	/** Widgets of the #BuildRailDepotWindow class. */
 	enum BuildRailDepotWidgets {
 		/* Name starts with BRA instead of BR, because of collision with BuildRoadDepotWidgets */
-		WID_BRAD_DEPOT_NE                            = ::WID_BRAD_DEPOT_NE,                            ///< Build a depot with the entrace in the north east.
-		WID_BRAD_DEPOT_SE                            = ::WID_BRAD_DEPOT_SE,                            ///< Build a depot with the entrace in the south east.
-		WID_BRAD_DEPOT_SW                            = ::WID_BRAD_DEPOT_SW,                            ///< Build a depot with the entrace in the south west.
-		WID_BRAD_DEPOT_NW                            = ::WID_BRAD_DEPOT_NW,                            ///< Build a depot with the entrace in the north west.
+		WID_BRAD_DEPOT_NE                            = ::WID_BRAD_DEPOT_NE,                            ///< Build a depot with the entrance in the north east.
+		WID_BRAD_DEPOT_SE                            = ::WID_BRAD_DEPOT_SE,                            ///< Build a depot with the entrance in the south east.
+		WID_BRAD_DEPOT_SW                            = ::WID_BRAD_DEPOT_SW,                            ///< Build a depot with the entrance in the south west.
+		WID_BRAD_DEPOT_NW                            = ::WID_BRAD_DEPOT_NW,                            ///< Build a depot with the entrance in the north west.
 	};
 
 	/** Widgets of the #BuildRailWaypointWindow class. */
@@ -2111,8 +2111,8 @@ public:
 		WID_CC_RATE_DOWN                             = ::WID_CC_RATE_DOWN,                             ///< Down button.
 		WID_CC_RATE_UP                               = ::WID_CC_RATE_UP,                               ///< Up button.
 		WID_CC_RATE                                  = ::WID_CC_RATE,                                  ///< Rate of currency.
-		WID_CC_SEPARATOR_EDIT                        = ::WID_CC_SEPARATOR_EDIT,                        ///< Seperator edit button.
-		WID_CC_SEPARATOR                             = ::WID_CC_SEPARATOR,                             ///< Current seperator.
+		WID_CC_SEPARATOR_EDIT                        = ::WID_CC_SEPARATOR_EDIT,                        ///< Separator edit button.
+		WID_CC_SEPARATOR                             = ::WID_CC_SEPARATOR,                             ///< Current separator.
 		WID_CC_PREFIX_EDIT                           = ::WID_CC_PREFIX_EDIT,                           ///< Prefix edit button.
 		WID_CC_PREFIX                                = ::WID_CC_PREFIX,                                ///< Current prefix.
 		WID_CC_SUFFIX_EDIT                           = ::WID_CC_SUFFIX_EDIT,                           ///< Suffix edit button.
@@ -2243,7 +2243,7 @@ public:
 		WID_TT_LEVEL_LAND                            = ::WID_TT_LEVEL_LAND,                            ///< Level land button.
 		WID_TT_DEMOLISH                              = ::WID_TT_DEMOLISH,                              ///< Demolish aka dynamite button.
 		WID_TT_BUY_LAND                              = ::WID_TT_BUY_LAND,                              ///< Buy land button.
-		WID_TT_PLANT_TREES                           = ::WID_TT_PLANT_TREES,                           ///< Plant trees button (note: opens seperate window, no place-push-button).
+		WID_TT_PLANT_TREES                           = ::WID_TT_PLANT_TREES,                           ///< Plant trees button (note: opens separate window, no place-push-button).
 		WID_TT_PLACE_SIGN                            = ::WID_TT_PLACE_SIGN,                            ///< Place sign button.
 		WID_TT_PLACE_OBJECT                          = ::WID_TT_PLACE_OBJECT,                          ///< Place object button.
 	};
@@ -2487,7 +2487,7 @@ public:
 		WID_VD_DETAILS_CARGO_CARRIED                 = ::WID_VD_DETAILS_CARGO_CARRIED,                 ///< Show carried cargo per part of the train.
 		WID_VD_DETAILS_TRAIN_VEHICLES                = ::WID_VD_DETAILS_TRAIN_VEHICLES,                ///< Show all parts of the train with their description.
 		WID_VD_DETAILS_CAPACITY_OF_EACH              = ::WID_VD_DETAILS_CAPACITY_OF_EACH,              ///< Show the capacity of all train parts.
-		WID_VD_DETAILS_TOTAL_CARGO                   = ::WID_VD_DETAILS_TOTAL_CARGO,                   ///< Show the capacity and carried cargo amounts aggregrated per cargo of the train.
+		WID_VD_DETAILS_TOTAL_CARGO                   = ::WID_VD_DETAILS_TOTAL_CARGO,                   ///< Show the capacity and carried cargo amounts aggregated per cargo of the train.
 	};
 
 	/** Widgets of the #VehicleListWindow class. */
diff --git a/src/script/squirrel.cpp b/src/script/squirrel.cpp
index 1f4ce7a..4584b9d 100644
--- a/src/script/squirrel.cpp
+++ b/src/script/squirrel.cpp
@@ -345,7 +345,7 @@ Squirrel::Squirrel(const char *APIName) :
 	sq_newclosure(this->vm, &Squirrel::_RunError, 0);
 	sq_seterrorhandler(this->vm);
 
-	/* Set the foreigh pointer, so we can always find this instance from within the VM */
+	/* Set the foreign pointer, so we can always find this instance from within the VM */
 	sq_setforeignptr(this->vm, this);
 
 	sq_pushroottable(this->vm);
diff --git a/src/script/squirrel_std.hpp b/src/script/squirrel_std.hpp
index 248b977..b1c7b30 100644
--- a/src/script/squirrel_std.hpp
+++ b/src/script/squirrel_std.hpp
@@ -58,7 +58,7 @@ void squirrel_register_std(Squirrel *engine);
 
 /**
  * Register all standard functions that are available on first startup.
- * @note this set is very limited, and is only ment to load other scripts and things like that.
+ * @note this set is very limited, and is only meant to load other scripts and things like that.
  */
 void squirrel_register_global_std(Squirrel *engine);
 
