diff -ur ./openttd-r14686/src/town_cmd.cpp ./openttd-r14686-town-placement/src/town_cmd.cpp
--- src/town_cmd.cpp	2008-11-23 15:17:41.000000000 +0100
+++ src/town_cmd.cpp	2008-12-17 22:39:35.000000000 +0100
@@ -1579,6 +1579,17 @@
 	do {
 		/* Generate a tile index not too close from the edge */
 		TileIndex tile = RandomTile();
+		switch (_settings_game.economy.town_layout) {
+			case TL_2X2_GRID:
+				tile = TileXY(TileX(tile) - TileX(tile) % 3,
+				              TileY(tile) - TileY(tile) % 3);
+				break;
+			case TL_3X3_GRID:
+				tile = TileXY(TileX(tile) & ~3,
+				              TileY(tile) & ~3);
+				break;
+			default: break;
+		}
 		if (DistanceFromEdge(tile) < 20) continue;
 
 		/* Make sure the tile is plain */
