Index: src/station_cmd.c
===================================================================
--- src/station_cmd.c	(revision 7794)
+++ src/station_cmd.c	(working copy)
@@ -1746,7 +1746,14 @@
 
 		st->build_date = _date;
 
-		StationRect_BeforeAddRect(st, tile, w, h, RECT_MODE_TRY);
+		/* New airports bigger than station_spread needs to be forced or
+		 * rejected. Sending RECT_MODE_TRY as usual gives an assertion. 
+		 */
+		if (airport_upgrade) {
+			StationRect_BeforeAddRect(st, tile, w, h, RECT_MODE_TRY);
+		} else {
+			StationRect_BeforeAddRect(st, tile, w, h, RECT_MODE_FORCE);
+		}
 
 		/* if airport was demolished while planes were en-route to it, the
 		 * positions can no longer be the same (v->u.air.pos), since different
