Index: src/autoreplace_gui.cpp
===================================================================
--- src/autoreplace_gui.cpp	(revision 16598)
+++ src/autoreplace_gui.cpp	(working copy)
@@ -108,7 +108,7 @@
 	/** Figure out if an engine should be added to a list
 	 * @param e The EngineID
 	 * @param draw_left If true, then the left list is drawn (the engines specific to the railtype you selected)
-	 * @param show_engines if truem then locomotives are drawn, else wagons (never both)
+	 * @param show_engines if true, then locomotives are drawn, else wagons (never both)
 	 * @return true if the engine should be in the list (based on this check)
 	 */
 	bool GenerateReplaceRailList(EngineID e, bool draw_left, bool show_engines)
