Index: src/widgets/dropdown.cpp
===================================================================
--- src/widgets/dropdown.cpp	(revision 12650)
+++ src/widgets/dropdown.cpp	(working copy)
@@ -258,7 +258,7 @@
 		_dropdown_menu_widgets);
 
 	dw->widget[0].color = wi->color;
-	dw->widget[0].right = width + 1;
+	dw->widget[0].right = width - 1;
 	dw->widget[0].bottom = height - 1;
 
 	dw->SetWidgetHiddenState(1, !scroll);
