diff -r bf4d7df60a83 src/settings_gui.cpp
--- a/src/settings_gui.cpp	Fri Nov 07 11:45:44 2008 +0100
+++ b/src/settings_gui.cpp	Fri Nov 07 11:48:06 2008 +0100
@@ -777,11 +777,9 @@
 
 		/* Build up the dynamic settings-array only once per OpenTTD session */
 		if (first_time) {
-			PatchPage *page;
-			for (page = &_patches_page[0]; page != endof(_patches_page); page++) {
-				uint i;
+			for (PatchPage *page = &_patches_page[0]; page != endof(_patches_page); page++) {
 				page->entries = MallocT<PatchEntry>(page->num);
-				for (i = 0; i != page->num; i++) {
+				for (uint i = 0; i != page->num; i++) {
 					uint index;
 					const SettingDesc *sd = GetPatchFromName(page->names[i], &index);
 					assert(sd != NULL);
@@ -811,14 +809,13 @@
 	virtual void OnPaint()
 	{
 		const PatchPage *page = &_patches_page[this->page];
-		uint i;
 
 		/* Set up selected category */
 		this->DrawWidgets();
 
 		int x = SETTINGTREE_LEFT_OFFSET;
 		int y = SETTINGTREE_TOP_OFFSET;
-		for (i = vscroll.pos; i != page->num && i < vscroll.pos + vscroll.cap; i++) {
+		for (uint i = vscroll.pos; i != page->num && i < vscroll.pos + vscroll.cap; i++) {
 			const SettingDesc *sd = page->entries[i].setting;
 			DrawPatch(patches_ptr, sd, x, y, this->click - (i * 2));
 			y += SETTING_HEIGHT;
