Index: src/video/sdl_v.cpp
===================================================================
--- src/video/sdl_v.cpp	(revision 18031)
+++ src/video/sdl_v.cpp	(working copy)
@@ -379,6 +379,9 @@
 
 	if (!SDL_CALL SDL_PollEvent(&ev)) return -2;
 
+	SDL_Surface *surf = SDL_CALL SDL_GetVideoSurface();
+	printf("size %d %d %d %d %d %d\n", surf->w, surf->h, surf->clip_rect.x, surf->clip_rect.y, surf->clip_rect.w, surf->clip_rect.h);
+
 	switch (ev.type) {
 		case SDL_MOUSEMOTION:
 			if (_cursor.fix_at) {
@@ -461,6 +464,7 @@
 			break;
 
 		case SDL_VIDEORESIZE: {
+			printf("resizeevent %d %d\n", ev.resize.w, ev.resize.h);
 			int w = max(ev.resize.w, 64);
 			int h = max(ev.resize.h, 64);
 			ChangeResInGame(w, h);
