diff --git a/src/station_cmd.cpp b/src/station_cmd.cpp
--- a/src/station_cmd.cpp
+++ b/src/station_cmd.cpp
@@ -1796,7 +1796,7 @@
 	/* Safeguard the parameters. */
 	if (!IsValidDiagDirection(ddir)) return CMD_ERROR;
 	/* If it is a drive-through stop, check for valid axis. */
-	if (is_drive_through && !IsValidAxis((Axis)ddir)) return CMD_ERROR;
+	if (is_drive_through && !IsValidAxis(DiagDirToAxis(ddir))) return CMD_ERROR;
 
 	CommandCost ret = CheckIfAuthorityAllowsNewStation(tile, flags);
 	if (ret.Failed()) return ret;
