From 9d88bb10f16150f71809228236f4d202d3a4315e Mon Sep 17 00:00:00 2001
From: Michael Lutz <michi@icosahedron.de>
Date: Fri, 22 Aug 2008 19:31:47 +0200
Subject: -Fix (r14039): Only reserve a path on leaving a station if the current block really is a PBS block.

---
 src/vehicle.cpp |    7 ++++---
 1 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/vehicle.cpp b/src/vehicle.cpp
index 4d4b774..870c135 100644
--- a/src/vehicle.cpp
+++ b/src/vehicle.cpp
@@ -2467,9 +2467,10 @@ void Vehicle::LeaveStation()
 
 		/* Try to reserve a path when leaving the station as we
 		 * might not be marked as wanting a reservation, e.g.
-		 * when an overlenght train gets turned around in a station.
-		 */
-		TryPathReserve(this, true, true);
+		 * when an overlength train gets turned around in a station. */
+		if (UpdateSignalsOnSegment(this->tile, TrackdirToExitdir(GetVehicleTrackdir(this)), this->owner) == SIGSEG_PBS || _settings_game.pf.reserve_paths) {
+			TryPathReserve(this, true, true);
+		}
 	}
 }
 
-- 
1.5.6.4

