diff -r 7c4cb55a8ba1 -r 759b36a5469d src/settings_gui.cpp
--- a/src/settings_gui.cpp	Sun Sep 28 14:09:59 2008 +0200
+++ b/src/settings_gui.cpp	Sun Sep 28 14:09:59 2008 +0200
@@ -695,34 +695,39 @@
 };
 
 static const char *_patches_vehicles[] = {
+	"order.gotodepot",
 	"gui.new_nonstop",
+	"vehicle.dynamic_engines",
+	"vehicle.never_expire_vehicles",
+	"order.timetabling",
 	"gui.timetable_in_ticks",
+	"",
 	"vehicle.realistic_acceleration",
+	"vehicle.wagon_speed_limits",
+	"vehicle.mammoth_trains",
+	"vehicle.freight_trains",
+	"vehicle.plane_speed",
+	"",
+	"pf.pathfinder_for_trains",
 	"pf.forbid_90_deg",
-	"vehicle.mammoth_trains",
-	"order.gotodepot",
+	"pf.pathfinder_for_roadvehs",
 	"pf.roadveh_queue",
-	"pf.pathfinder_for_trains",
-	"pf.pathfinder_for_roadvehs",
 	"pf.pathfinder_for_ships",
+	"",
+	"gui.lost_train_warn",
 	"gui.train_income_warn",
 	"gui.order_review_system",
-	"vehicle.never_expire_vehicles",
-	"gui.lost_train_warn",
+	"",
 	"gui.autorenew",
 	"gui.autorenew_months",
 	"gui.autorenew_money",
+	"",
 	"vehicle.servint_ispercent",
 	"vehicle.servint_trains",
 	"vehicle.servint_roadveh",
 	"vehicle.servint_ships",
 	"vehicle.servint_aircraft",
 	"order.no_servicing_if_no_breakdowns",
-	"vehicle.wagon_speed_limits",
-	"vehicle.freight_trains",
-	"vehicle.plane_speed",
-	"order.timetabling",
-	"vehicle.dynamic_engines",
 	NULL
 };
 
