diff --git a/src/3rdparty/squirrel/include/squirrel.h b/src/3rdparty/squirrel/include/squirrel.h
--- a/src/3rdparty/squirrel/include/squirrel.h
+++ b/src/3rdparty/squirrel/include/squirrel.h
@@ -35,10 +35,6 @@
 extern "C" {
 #endif
 
-#if defined(_MSC_VER)
-# define inline __forceinline
-#endif /* _MSC_VER */
-
 #if defined(_MSC_VER) && _MSC_VER >= 1400 // MSVC 2005 safety checks
 # pragma warning(disable: 4996)   // '_wfopen' was declared deprecated
 # define _CRT_SECURE_NO_DEPRECATE // all deprecated 'unsafe string functions
diff --git a/src/stdafx.h b/src/stdafx.h
--- a/src/stdafx.h
+++ b/src/stdafx.h
@@ -185,8 +185,6 @@
 
 	#include <malloc.h> // alloca()
 	#define NORETURN __declspec(noreturn)
-	#define inline __forceinline
-
 	#if !defined(WINCE)
 		#define CDECL _cdecl
 	#endif
