Index: src/players.cpp
===================================================================
--- src/players.cpp	(revision 11876)
+++ src/players.cpp	(working copy)
@@ -13,6 +13,7 @@
 #include "saveload.h"
 #include "command_func.h"
 #include "network/network.h"
+#include "network/network_gui.h"
 #include "network/network_internal.h"
 #include "variables.h"
 #include "engine.h"
@@ -833,6 +834,9 @@
 			if (!StrEmpty(_network_default_company_pass)) {
 				char *password = _network_default_company_pass;
 				NetworkChangeCompanyPassword(1, &password);
+			} else {
+				/* When joining a new company, ask the player if he wants to set a password. */
+				ShowNetworkCompanyPasswordWindow();
 			}
 #endif /* ENABLE_NETWORK */
 			MarkWholeScreenDirty();
