Index: src/road_cmd.cpp
===================================================================
--- src/road_cmd.cpp	(revision 12286)
+++ src/road_cmd.cpp	(working copy)
@@ -1286,6 +1286,12 @@
 			} else {
 				cur_rs = ROADSIDE_BARREN;
 			}
+
+			if (HasTileRoadType(tile, ROADTYPE_TRAM) &&
+					(cur_rs == ROADSIDE_STREET_LIGHTS || cur_rs == ROADSIDE_TREES)){
+				cur_rs = ROADSIDE_PAVED;
+			}
+
 			SetRoadside(tile, cur_rs);
 			MarkTileDirtyByTile(tile);
 		}
