Index: src/console_cmds.cpp
===================================================================
--- src/console_cmds.cpp	(revision 16003)
+++ src/console_cmds.cpp	(working copy)
@@ -1401,7 +1401,7 @@
 	Company *c;
 
 	if (argc == 0) {
-		IConsoleHelp("List the in-game details of all clients connected to the server. Usage 'companies'");
+		IConsoleHelp("List the in-game details of all companies in the game. Usage 'companies'");
 		return true;
 	}
 	NetworkCompanyStats company_stats[MAX_COMPANIES];
@@ -1417,13 +1417,13 @@
 		const NetworkCompanyStats *stats = &company_stats[c->index];
 
 		GetString(buffer, STR_00D1_DARK_BLUE + _company_colours[c->index], lastof(buffer));
-		IConsolePrintF(CC_INFO, "#:%d(%s) Company Name: '%s'  Year Founded: %d  Money: %" OTTD_PRINTF64 "d  Loan: %" OTTD_PRINTF64 "d  Value: %" OTTD_PRINTF64 "d  (T:%d, R:%d, P:%d, S:%d) %sprotected",
+		IConsolePrintF(CC_INFO, "#:%d(%s) Company Name: '%s'  Year Founded: %d  Money: %" OTTD_PRINTF64 "d  Loan: %" OTTD_PRINTF64 "d  Value: %" OTTD_PRINTF64 "d  (T:%d, R:%d, P:%d, S:%d) %s",
 			c->index + 1, buffer, company_name, c->inaugurated_year, (int64)c->money, (int64)c->current_loan, (int64)CalculateCompanyValue(c),
 			/* trains      */ stats->num_vehicle[0],
 			/* lorry + bus */ stats->num_vehicle[1] + stats->num_vehicle[2],
 			/* planes      */ stats->num_vehicle[3],
 			/* ships       */ stats->num_vehicle[4],
-			/* protected   */ StrEmpty(_network_company_states[c->index].password) ? "un" : "");
+			/* protected   */ _networking?(StrEmpty(_network_company_states[c->index].password) ? "unprotected" : "protected"):(IsHumanCompany(c->index)?"":"AI"));
 	}
 
 	return true;
@@ -1765,7 +1765,6 @@
 	IConsoleCmdRegister("say",             ConSay);
 	IConsoleCmdHookAdd("say",              ICONSOLE_HOOK_ACCESS, ConHookNeedNetwork);
 	IConsoleCmdRegister("companies",       ConCompanies);
-	IConsoleCmdHookAdd("companies",        ICONSOLE_HOOK_ACCESS, ConHookServerOnly);
 	IConsoleAliasRegister("players",       "companies");
 	IConsoleCmdRegister("say_company",     ConSayCompany);
 	IConsoleCmdHookAdd("say_company",      ICONSOLE_HOOK_ACCESS, ConHookNeedNetwork);
