Index: src/train_cmd.c
===================================================================
--- src/train_cmd.c	(revision 7830)
+++ src/train_cmd.c	(working copy)
@@ -2973,10 +2973,16 @@
 	coll = GetFirstVehicleInChain(realcoll);
 
 	/* it can't collide with its own wagons */
-	if (v == coll ||
-			(v->u.rail.track & 0x40 && (v->direction & 2) != (realcoll->direction & 2)))
+	// yes it is !
+	//if (v == coll ||
+	if 
+			(v->u.rail.track & 0x40 && (v->direction & 2) != (realcoll->direction & 2))
 		return;
 
+	// But it can't collide with neightbor wagons
+	if (v->next == coll) return;
+	if (coll->next == coll) return;
+
 	//two drivers + passangers killed in train v
 	num = 2 + CountPassengersInTrain(v);
 	if (!(coll->vehstatus & VS_CRASHED))
