Index: src/table/water_land.h
===================================================================
--- src/table/water_land.h	(revision 22643)
+++ src/table/water_land.h	(working copy)
@@ -54,10 +54,12 @@
 };

 static const DrawTileSprites _shipdepot_display_data[] = {
-	TILE_SPRITE_LINE(0xFDD, _shipdepot_display_seq_1)
-	TILE_SPRITE_LINE(0xFDD, _shipdepot_display_seq_2)
-	TILE_SPRITE_LINE(0xFDD, _shipdepot_display_seq_3)
-	TILE_SPRITE_LINE(0xFDD, _shipdepot_display_seq_4)
+	/* AXIS_X */
+	TILE_SPRITE_LINE(0xFDD, _shipdepot_display_seq_1) // DEPOT_NORTH
+	TILE_SPRITE_LINE(0xFDD, _shipdepot_display_seq_2) // DEPOT_SOUTH
+	/* AXIS_Y */
+	TILE_SPRITE_LINE(0xFDD, _shipdepot_display_seq_3) // DEPOT_NORTH
+	TILE_SPRITE_LINE(0xFDD, _shipdepot_display_seq_4) // DEPOT_SOUTH
 };

 static const DrawTileSeqStruct _lock_display_seq_0[] = {
@@ -133,20 +135,23 @@
 };

 static const DrawTileSprites _lock_display_data[] = {
-	TILE_SPRITE_LINE(1, _lock_display_seq_0)
-	TILE_SPRITE_LINE(0, _lock_display_seq_1)
-	TILE_SPRITE_LINE(2, _lock_display_seq_2)
-	TILE_SPRITE_LINE(3, _lock_display_seq_3)
+	/* LOCK_MIDDLE */
+	TILE_SPRITE_LINE(1, _lock_display_seq_0) // NE
+	TILE_SPRITE_LINE(0, _lock_display_seq_1) // SE
+	TILE_SPRITE_LINE(2, _lock_display_seq_2) // SW
+	TILE_SPRITE_LINE(3, _lock_display_seq_3) // NW

-	TILE_SPRITE_LINE(0xFDD, _lock_display_seq_0b)
-	TILE_SPRITE_LINE(0xFDD, _lock_display_seq_1b)
-	TILE_SPRITE_LINE(0xFDD, _lock_display_seq_2b)
-	TILE_SPRITE_LINE(0xFDD, _lock_display_seq_3b)
+	/* LOCK_LOWER */
+	TILE_SPRITE_LINE(0xFDD, _lock_display_seq_0b) // NE
+	TILE_SPRITE_LINE(0xFDD, _lock_display_seq_1b) // SE
+	TILE_SPRITE_LINE(0xFDD, _lock_display_seq_2b) // SW
+	TILE_SPRITE_LINE(0xFDD, _lock_display_seq_3b) // NW

-	TILE_SPRITE_LINE(0xFDD, _lock_display_seq_0t)
-	TILE_SPRITE_LINE(0xFDD, _lock_display_seq_1t)
-	TILE_SPRITE_LINE(0xFDD, _lock_display_seq_2t)
-	TILE_SPRITE_LINE(0xFDD, _lock_display_seq_3t)
+	/* LOCK_UPPER */
+	TILE_SPRITE_LINE(0xFDD, _lock_display_seq_0t) // NE
+	TILE_SPRITE_LINE(0xFDD, _lock_display_seq_1t) // SE
+	TILE_SPRITE_LINE(0xFDD, _lock_display_seq_2t) // SW
+	TILE_SPRITE_LINE(0xFDD, _lock_display_seq_3t) // NW
 };

 #undef TILE_SEQ_LINE
Index: src/water_cmd.cpp
===================================================================
--- src/water_cmd.cpp	(revision 22643)
+++ src/water_cmd.cpp	(working copy)
@@ -427,9 +427,10 @@

 		case WATER_TILE_LOCK: {
 			static const TileIndexDiffC _lock_tomiddle_offs[] = {
-				{ 0,  0}, {0,  0}, { 0, 0}, {0,  0}, // middle
-				{-1,  0}, {0,  1}, { 1, 0}, {0, -1}, // lower
-				{ 1,  0}, {0, -1}, {-1, 0}, {0,  1}, // upper
+				/* NE       SE        SW      NW */
+				{ 0,  0}, {0,  0}, { 0, 0}, {0,  0}, // LOCK_MIDDLE
+				{-1,  0}, {0,  1}, { 1, 0}, {0, -1}, // LOCK_LOWER
+				{ 1,  0}, {0, -1}, {-1, 0}, {0,  1}, // LOCK_UPPER
 			};

 			if (flags & DC_AUTO) return_cmd_error(STR_ERROR_BUILDING_MUST_BE_DEMOLISHED);
