Index: src/table/newgrf_debug_data.h
===================================================================
--- src/table/newgrf_debug_data.h	(wersja 26371)
+++ src/table/newgrf_debug_data.h	(kopia robocza)
@@ -97,8 +97,8 @@
 #define NICS(cb_id, bit) NIC(cb_id, StationSpec, callback_mask, bit)
 static const NICallback _nic_stations[] = {
 	NICS(CBID_STATION_AVAILABILITY,     CBM_STATION_AVAIL),
-	NICS(CBID_STATION_SPRITE_LAYOUT,    CBM_NO_BIT),
-	NICS(CBID_STATION_TILE_LAYOUT,      CBM_STATION_SPRITE_LAYOUT),
+	NICS(CBID_STATION_SPRITE_LAYOUT,    CBM_STATION_SPRITE_LAYOUT),
+	NICS(CBID_STATION_TILE_LAYOUT,      CBM_NO_BIT),
 	NICS(CBID_STATION_ANIM_START_STOP,  CBM_NO_BIT),
 	NICS(CBID_STATION_ANIM_NEXT_FRAME,  CBM_STATION_ANIMATION_NEXT_FRAME),
 	NICS(CBID_STATION_ANIMATION_SPEED,  CBM_STATION_ANIMATION_SPEED),
