Index: src/console_cmds.cpp
===================================================================
--- src/console_cmds.cpp        (révision 19757)
+++ src/console_cmds.cpp        (copie de travail)
@@ -964,12 +964,18 @@
 DEF_CONSOLE_CMD(ConNewGame)
 {
        if (argc == 0) {
-               IConsoleHelp("Start a new game. Usage: 'newgame [seed]'");
+               IConsoleHelp("Start a new game. Usage: 'newgame [landscape] [map_x] [map_y] [seed]'");
+               IConsoleHelp("landcape : 0-3, temperate, arctic, desert, candy");
+               IConsoleHelp("map_x, map_y : 6-11, 6:64, 7:128, 8:256, 9:512, 10:1024, 11:2048");
                IConsoleHelp("The server can force a new game using 'newgame'; any client joined will rejoin after the server is done generating the new game.");
                return true;
        }
-
-       StartNewGameWithoutGUI((argc == 2) ? (uint)atoi(argv[1]) : GENERATE_NEW_SEED);
+       _settings_newgame.game_creation.landscape = (argc>=2?(uint)atoi(argv[1]) & 0x03:0) ;
+       _settings_newgame.game_creation.map_x = (argc>=3&&(uint)atoi(argv[2])>5&&(uint)atoi(argv[2])<12?(uint)atoi(argv[2]):9);
+       _settings_newgame.game_creation.map_y = (argc>=4&&(uint)atoi(argv[3])>5&&(uint)atoi(argv[3])<12?(uint)atoi(argv[3]):9);
+       IConsolePrintF( CC_DEFAULT, "new game with landscape %d map size %dx%d", _settings_newgame.game_creation.landscape, _settings_newgame.game_creation.map_x, _settings_newgame.game_creation.map_y );
+       // StartNewGameWithoutGUI((argc == 2) ? (uint)atoi(argv[1]) : GENERATE_NEW_SEED);
+       StartNewGameWithoutGUI((argc >= 5) ? (uint)atoi(argv[4]) : GENERATE_NEW_SEED);
        return true;
 }