Index: src/main_gui.cpp
===================================================================
--- src/main_gui.cpp	(revision 10292)
+++ src/main_gui.cpp	(working copy)
@@ -1903,6 +1903,7 @@
 		case WKC_SHIFT | WKC_F12: ShowMusicWindow(); break;
 		case WKC_CTRL  | 'S': MenuClickSmallScreenshot(); break;
 		case WKC_CTRL  | 'G': MenuClickWorldScreenshot(); break;
+		case WKC_CTRL | WKC_ALT | WKC_SHIFT | 'C': /* Fall thru */
 		case WKC_CTRL | WKC_ALT | 'C': if (!_networking) ShowCheatWindow(); break;
 		case 'A': ShowBuildRailToolbar(_last_built_railtype, 4); break; /* Invoke Autorail */
 		case 'L': ShowTerraformToolbar(); break;
