Index: src/video/cocoa/wnd_quartz.mm
===================================================================
--- src/video/cocoa/wnd_quartz.mm	(revision 18066)
+++ src/video/cocoa/wnd_quartz.mm	(working copy)
@@ -167,6 +167,8 @@
 		if (CMGetSystemProfile(&sysProfile) == noErr) {
 			colorSpace = CGColorSpaceCreateWithPlatformColorSpace(sysProfile);
 			CMCloseProfile(sysProfile);
+		} else {
+			colorSpace = CGColorSpaceCreateDeviceRGB();
 		}
 
 		if (colorSpace == NULL) error("Could not get system colour space. You might need to recalibrate your monitor.");
