diff -r 94e78312260d src/os/windows/win32.cpp
--- a/src/os/windows/win32.cpp	Sun May 08 18:20:21 2011 +0000
+++ b/src/os/windows/win32.cpp	Mon May 09 22:21:47 2011 +0200
@@ -297,7 +297,9 @@
 
 	/* redirect unbuffered STDIN, STDOUT, STDERR to the console */
 #if !defined(__CYGWIN__)
-	*stdout = *_fdopen( _open_osfhandle((intptr_t)hand, _O_TEXT), "w" );
+	int fd = _open_osfhandle((intptr_t)hand, _O_TEXT);
+	assert(fd != -1);
+	*stdout = *_fdopen(fd , "w");
 	*stdin = *_fdopen(_open_osfhandle((intptr_t)GetStdHandle(STD_INPUT_HANDLE), _O_TEXT), "r" );
 	*stderr = *_fdopen(_open_osfhandle((intptr_t)GetStdHandle(STD_ERROR_HANDLE), _O_TEXT), "w" );
 #else
