Index: src/direction_func.h
===================================================================
--- src/direction_func.h	(wersja 22276)
+++ src/direction_func.h	(kopia robocza)
@@ -35,7 +35,7 @@
  */
 static inline DirDiff DirDifference(Direction d0, Direction d1)
 {
-	return (DirDiff)((d0 + 8 - d1) % 8);
+	return (DirDiff)((d0 - d1) & 0x7);
 }
 
 /**
@@ -51,7 +51,7 @@
  */
 static inline DirDiff ChangeDirDiff(DirDiff d, DirDiff delta)
 {
-	return (DirDiff)((d + delta) % 8);
+	return (DirDiff)((d + delta) & 0x7);
 }
 
 /**
@@ -66,7 +66,7 @@
  */
 static inline Direction ChangeDir(Direction d, DirDiff delta)
 {
-	return (Direction)((d + delta) % 8);
+	return (Direction)((d + delta) & 0x7);
 }
 
 
@@ -94,7 +94,7 @@
  */
 static inline DiagDirection ChangeDiagDir(DiagDirection d, DiagDirDiff delta)
 {
-	return (DiagDirection)((d + delta) % 4);
+	return (DiagDirection)((d + delta) & 0x3);
 }
 
 /**
@@ -124,7 +124,7 @@
  */
 static inline Direction DiagDirToDir(DiagDirection dir)
 {
-	return (Direction)(dir * 2 + 1);
+	return (Direction)((dir << 1) + 1);
 }
 
 
@@ -187,7 +187,7 @@
  */
 static inline Direction AxisToDirection(Axis a)
 {
-	return (Direction)(5 - 2 * a);
+	return (Direction)(5 - (a << 1));
 }
 
 /**
@@ -198,7 +198,7 @@
  */
 static inline DiagDirection XYNSToDiagDir(Axis xy, uint ns)
 {
-	return (DiagDirection)(xy * 3 ^ ns * 2);
+	return (DiagDirection)((4 - (ns << 1) - xy) & 0x3);
 }
 
 /**
