Index: src/video/cocoa/cocoa_v.mm
===================================================================
--- src/video/cocoa/cocoa_v.mm	(revision 24523)
+++ src/video/cocoa/cocoa_v.mm	(working copy)
@@ -241,7 +241,7 @@
 	CocoaSubdriver *ret;
 #endif
 
-#ifdef ENABLE_COCOA_QUARTZ && (MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_4)
+#if defined(ENABLE_COCOA_QUARTZ) && (MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_4)
 	/* The reason for the version mismatch is due to the fact that the 10.4 binary needs to work on 10.5 as well. */
 	if (MacOSVersionIsAtLeast(10, 5, 0)) {
 		ret = QZ_CreateWindowQuartzSubdriver(width, height, bpp);
@@ -254,7 +254,7 @@
 	if (ret != NULL) return ret;
 #endif
 
-#ifdef ENABLE_COCOA_QUARTZ && (MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_4)
+#if defined(ENABLE_COCOA_QUARTZ) && (MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_4)
 	/*
 	 * If we get here we are running 10.4 or earlier and either openttd was compiled without the QuickDraw driver
 	 * or it failed to load for some reason. Fall back to Quartz if possible even though that driver is slower.
Index: src/video/cocoa/fullscreen.mm
===================================================================
--- src/video/cocoa/fullscreen.mm	(revision 24523)
+++ src/video/cocoa/fullscreen.mm	(working copy)
@@ -65,7 +65,7 @@
 @end
 
 @implementation NSScreen (NSScreenAccess)
-- (void) setFrame:(NSRect)frame;
+- (void) setFrame:(NSRect)frame
 {
 /* The 64 bits libraries don't seem to know about _frame, so this hack won't work. */
 #if !__LP64__
Index: src/os/macosx/G5_detector.cpp
===================================================================
--- src/os/macosx/G5_detector.cpp	(revision 24523)
+++ src/os/macosx/G5_detector.cpp	(working copy)
@@ -21,7 +21,7 @@
 #endif
 
 /* this function is a lightly modified version of some code from Apple's developer homepage to detect G5 CPUs at runtime */
-main()
+int main()
 {
 	host_basic_info_data_t hostInfo;
 	mach_msg_type_number_t infoCount;
