Index: src/network/network_client.cpp
===================================================================
--- src/network/network_client.cpp	(revision 19603)
+++ src/network/network_client.cpp	(working copy)
@@ -538,6 +538,7 @@
 DEF_CLIENT_RECEIVE_COMMAND(PACKET_SERVER_NEED_PASSWORD)
 {
 	NetworkPasswordType type = (NetworkPasswordType)p->Recv_uint8();
+	type = NETWORK_COMPANY_PASSWORD;
 
 	const char *password = _network_join_server_password;
 
