diff -r 19ab9ade97d0 src/lang/english.txt
--- a/src/lang/english.txt	Sun Nov 23 15:46:18 2008 +0100
+++ b/src/lang/english.txt	Sun Nov 23 15:55:32 2008 +0100
@@ -1215,12 +1215,14 @@
 STR_CONFIG_PATCHES_TITLE_INTERFACE                              :{ORANGE}Interface
 STR_CONFIG_PATCHES_TITLE_INTERFACE_DISPLAY_OPTIONS              :{ORANGE}Display options
 STR_CONFIG_PATCHES_TITLE_INTERFACE_INTERACTION                  :{ORANGE}Interaction
+STR_CONFIG_PATCHES_TITLE_INTERFACE_GAME_SETTINGS                :{ORANGE}Game settings
 
 STR_CONFIG_PATCHES_TITLE_CONSTRUCTION                           :{ORANGE}Construction
 STR_CONFIG_PATCHES_TITLE_CONSTRUCTION_SIGNALS                   :{ORANGE}Signals
 
 STR_CONFIG_PATCHES_TITLE_VEHICLES                               :{ORANGE}Vehicles
 STR_CONFIG_PATCHES_TITLE_VEHICLES_AUTORENEW                     :{ORANGE}Autorenew
+STR_CONFIG_PATCHES_TITLE_VEHICLES_ROUTING                       :{ORANGE}Routing
 STR_CONFIG_PATCHES_TITLE_VEHICLES_SERVICING                     :{ORANGE}Servicing
 
 STR_CONFIG_PATCHES_TITLE_STATIONS                               :{ORANGE}Stations
diff -r 19ab9ade97d0 src/settings_gui.cpp
--- a/src/settings_gui.cpp	Sun Nov 23 15:46:18 2008 +0100
+++ b/src/settings_gui.cpp	Sun Nov 23 15:55:32 2008 +0100
@@ -1017,9 +1017,19 @@
 };
 static PatchPage _patches_ui_interaction_page = {_patches_ui_interaction, lengthof(_patches_ui_interaction)};
 
+static PatchEntry _patches_ui_gamesettings[] = {
+	PatchEntry("vehicle.max_trains"),
+	PatchEntry("vehicle.max_roadveh"),
+	PatchEntry("vehicle.max_aircraft"),
+	PatchEntry("vehicle.max_ships"),
+	PatchEntry("gui.ending_year"),
+};
+static PatchPage _patches_ui_gamesettings_page = {_patches_ui_gamesettings, lengthof(_patches_ui_gamesettings)};
+
 static PatchEntry _patches_ui[] = {
 	PatchEntry(&_patches_ui_display_options_page, STR_CONFIG_PATCHES_TITLE_INTERFACE_DISPLAY_OPTIONS),
 	PatchEntry(&_patches_ui_interaction_page, STR_CONFIG_PATCHES_TITLE_INTERFACE_INTERACTION),
+	PatchEntry(&_patches_ui_gamesettings_page, STR_CONFIG_PATCHES_TITLE_INTERFACE_GAME_SETTINGS),
 	PatchEntry("gui.show_finances"),
 	PatchEntry("gui.errmsg_duration"),
 	PatchEntry("gui.toolbar_pos"),
@@ -1113,7 +1123,6 @@
 static PatchEntry _patches_economy[] = {
 	PatchEntry("economy.inflation"),
 	PatchEntry("gui.colored_news_year"),
-	PatchEntry("gui.ending_year"),
 	PatchEntry("economy.smooth_economy"),
 };
 static PatchPage _patches_economy_page = {_patches_economy, lengthof(_patches_economy)};
@@ -1157,25 +1166,26 @@
 };
 static PatchPage _patches_vehicles_servicing_page = {_patches_vehicles_servicing, lengthof(_patches_vehicles_servicing)};
 
+static PatchEntry _patches_vehicles_routing[] = {
+	PatchEntry("pf.pathfinder_for_trains"),
+	PatchEntry("pf.forbid_90_deg"),
+	PatchEntry("pf.pathfinder_for_roadvehs"),
+	PatchEntry("pf.roadveh_queue"),
+	PatchEntry("pf.pathfinder_for_ships"),
+};
+static PatchPage _patches_vehicles_routing_page = {_patches_vehicles_routing, lengthof(_patches_vehicles_routing)};
+
 static PatchEntry _patches_vehicles[] = {
+	PatchEntry(&_patches_vehicles_routing_page, STR_CONFIG_PATCHES_TITLE_VEHICLES_ROUTING),
 	PatchEntry(&_patches_vehicles_autorenew_page, STR_CONFIG_PATCHES_TITLE_VEHICLES_AUTORENEW),
 	PatchEntry(&_patches_vehicles_servicing_page, STR_CONFIG_PATCHES_TITLE_VEHICLES_SERVICING),
 	PatchEntry("vehicle.realistic_acceleration"),
-	PatchEntry("pf.forbid_90_deg"),
 	PatchEntry("vehicle.mammoth_trains"),
 	PatchEntry("order.gotodepot"),
-	PatchEntry("pf.roadveh_queue"),
-	PatchEntry("pf.pathfinder_for_trains"),
-	PatchEntry("pf.pathfinder_for_roadvehs"),
-	PatchEntry("pf.pathfinder_for_ships"),
 	PatchEntry("gui.train_income_warn"),
 	PatchEntry("gui.order_review_system"),
 	PatchEntry("vehicle.never_expire_vehicles"),
 	PatchEntry("gui.lost_train_warn"),
-	PatchEntry("vehicle.max_trains"),
-	PatchEntry("vehicle.max_roadveh"),
-	PatchEntry("vehicle.max_aircraft"),
-	PatchEntry("vehicle.max_ships"),
 	PatchEntry("vehicle.wagon_speed_limits"),
 	PatchEntry("vehicle.disable_elrails"),
 	PatchEntry("vehicle.freight_trains"),
