commit 50a8a40de47a12409dba5658c00735122c0fd77b
Author: Ulf Hermann <ulf_hermann@gmx.net>
Date:   Tue May 7 23:24:08 2013 -0500

    -Fix: keep all cargo whenever cancelling reservations

diff --git a/src/vehicle.cpp b/src/vehicle.cpp
index b6545e2..b209092 100644
--- a/src/vehicle.cpp
+++ b/src/vehicle.cpp
@@ -2025,8 +2025,8 @@ void Vehicle::CancelReservation(Station *st)
 		if (cargo.ActionCount(VehicleCargoList::MTA_LOAD) > 0) {
 			DEBUG(misc, 1, "cancelling cargo reservation");
 			cargo.Return(UINT_MAX, &st->goods[v->cargo_type].cargo);
-			cargo.KeepAll();
 		}
+		cargo.KeepAll();
 	}
 }
 
