Index: economy.c
===================================================================
--- economy.c	(revision 5348)
+++ economy.c	(working copy)
@@ -1341,7 +1341,11 @@
 
 		/* unload? */
 		if (v->cargo_count != 0) {
-			if (v->cargo_source != last_visited && ge->waiting_acceptance & 0x8000 && !(u->current_order.flags & OF_TRANSFER)) {
+			if (	v->cargo_source != last_visited &&
+					ge->waiting_acceptance & 0x8000 &&
+					!(u->current_order.flags & OF_FULL_LOAD) &&  
+					!(u->current_order.flags & OF_TRANSFER)) {
+
 				// deliver goods to the station
 				st->time_since_unload = 0;
 
