diff -r cb7c10816048 src/order_gui.cpp
--- a/src/order_gui.cpp	Fri Jun 25 17:45:06 2010 +0000
+++ b/src/order_gui.cpp	Sat Jun 26 14:08:33 2010 +0200
@@ -608,6 +608,12 @@
 		}
 
 		DoCommandP(this->vehicle->tile, this->vehicle->index + (sel_ord << 16), MOF_UNLOAD | (unload_type << 4), CMD_MODIFY_ORDER | CMD_MSG(STR_ERROR_CAN_T_MODIFY_THIS_ORDER));
+
+		/* Transfer orders with leave emtpy as default */
+		if (unload_type == OUFB_TRANSFER) {
+			DoCommandP(this->vehicle->tile, this->vehicle->index + (sel_ord << 16), MOF_LOAD | (OLFB_NO_LOAD << 4), CMD_MODIFY_ORDER);
+			this->SetWidgetDirty(ORDER_WIDGET_FULL_LOAD);
+		}
 	}
 
 	/**
