Index: src/order_cmd.cpp
===================================================================
--- src/order_cmd.cpp	(revision 24917)
+++ src/order_cmd.cpp	(working copy)
@@ -1613,6 +1613,8 @@
 	/* Automatic refit cargo is only supported for goto station orders. */
 	if (cargo == CT_AUTO_REFIT && !order->IsType(OT_GOTO_STATION)) return CMD_ERROR;
 
+	if (order->GetLoadType() & OLFB_NO_LOAD) return CMD_ERROR;
+
 	if (flags & DC_EXEC) {
 		order->SetRefit(cargo, subtype);
 
