Index: src/date_type.h
===================================================================
--- src/date_type.h	(revision 22544)
+++ src/date_type.h	(working copy)
@@ -35,6 +35,7 @@
 static const int STATION_ACCEPTANCE_TICKS = 250; ///< cycle duration for updating station acceptance
 static const int CARGO_AGING_TICKS        = 185; ///< cycle duration for aging cargo
 static const int INDUSTRY_PRODUCE_TICKS   = 256; ///< cycle duration for industry production
+static const int INDUSTRY_CUT_TREE_TICKS  = 512; ///< cycle duration for lumber mill's extra action
 static const int TOWN_GROWTH_TICKS        = 70;  ///< cycle duration for towns trying to grow. (this originates from the size of the town array in TTD
 
 
Index: src/industry_cmd.cpp
===================================================================
--- src/industry_cmd.cpp	(revision 22544)
+++ src/industry_cmd.cpp	(working copy)
@@ -1086,7 +1086,7 @@
 			if (plant) PlantRandomFarmField(i);
 		}
 		if ((indbehav & INDUSTRYBEH_CUT_TREES) != 0) {
-			bool cut = ((i->counter & 0x1FF) == 0);
+			bool cut = ((i->counter % INDUSTRY_CUT_TREE_TICKS) == 0);
 			if (HasBit(indsp->callback_mask, CBM_IND_SPECIAL_EFFECT)) {
 				cut = (GetIndustryCallback(CBID_INDUSTRY_SPECIAL_EFFECT, 0, 1, i, i->type, i->location.tile) != 0);
 			}
