Index: src/road_cmd.cpp
===================================================================
--- src/road_cmd.cpp	(revision 10029)
+++ src/road_cmd.cpp	(working copy)
@@ -935,8 +935,13 @@
 	int x = ti->x | dx;
 	int y = ti->y | dy;
 	byte z = ti->z;
+	SpriteID pal = PAL_NONE;
 	if (ti->tileh != SLOPE_FLAT) z = GetSlopeZ(x, y);
-	AddSortableSpriteToDraw(img, PAL_NONE, x, y, 2, 2, h, z);
+	if (HASBIT(_transparent_opt, TO_HOUSES)) {
+		pal = PALETTE_TO_TRANSPARENT;
+		SETBIT(img, PALETTE_MODIFIER_TRANSPARENT);
+	}
+	AddSortableSpriteToDraw(img, pal, x, y, 2, 2, h, z);
 }
 
 /**
