diff -Nur bin.old/apple-darwin9-g++ bin/apple-darwin9-g++
--- bin.old/apple-darwin9-g++	2013-02-25 16:50:19.000000000 +0100
+++ bin/apple-darwin9-g++	2013-02-25 16:58:05.000000000 +0100
@@ -24,7 +24,12 @@
 	version="9"
 fi
 if [ -n "`echo $@ | grep '\mmacosx-version-min=10.4'`" ]; then
-	version="8"
+	case $platform in
+		powerpc*)
+			version="8";;
+		*)
+			version="9";;
+	esac
 fi
 if [ -n "`echo $@ | grep '\mmacosx-version-min=10.3'`" ]; then
 	version="8"
diff -Nur bin.old/apple-darwin9-gcc bin/apple-darwin9-gcc
--- bin.old/apple-darwin9-gcc	2013-02-25 16:50:19.000000000 +0100
+++ bin/apple-darwin9-gcc	2013-02-25 16:56:08.000000000 +0100
@@ -24,7 +24,12 @@
 	version="9"
 fi
 if [ -n "`echo $@ | grep '\mmacosx-version-min=10.4'`" ]; then
-	version="8"
+	case $platform in
+		powerpc*)
+			version="8";;
+		*)
+			version="9";;
+	esac
 fi
 if [ -n "`echo $@ | grep '\mmacosx-version-min=10.3'`" ]; then
 	version="8"
diff -Nur bin.old/apple-darwin9-strip bin/apple-darwin9-strip
--- bin.old/apple-darwin9-strip	2013-02-25 16:50:19.000000000 +0100
+++ bin/apple-darwin9-strip	2013-02-25 16:57:58.000000000 +0100
@@ -22,7 +22,12 @@
 	version="9"
 fi
 if [ -n "`echo $@ | grep '\mmacosx-version-min=10.4'`" ]; then
-	version="8"
+	case $platform in
+		powerpc*)
+			version="8";;
+		*)
+			version="9";;
+	esac
 fi
 if [ -n "`echo $@ | grep '\mmacosx-version-min=10.3'`" ]; then
 	version="8"
