Index: split/openttdgui.nfo
===================================================================
--- split/openttdgui.nfo	(revision 12405)
+++ split/openttdgui.nfo	(working copy)
@@ -2,7 +2,7 @@
 // $Id$
 //
    -1 * 0	 0C "OpenTTD GUI graphics"
-   -1 * 0	 05 15 8A
+   -1 * 0	 05 15 90
    -1 sprites/openttdgui.pcx 18 8 09 31 64 -31 7
    -1 sprites/openttdgui.pcx 98 8 09 31 64 -31 7
    -1 sprites/openttdgui.pcx 178 8 09 31 64 -31 7
@@ -141,3 +141,9 @@
    -1 sprites/openttdgui.pcx 447 376 01 22 19 0 0
    -1 sprites/openttdgui.pcx 482 376 01 20 20 0 0
    -1 sprites/openttdgui.pcx 514 376 01 44 54 -16 -35
+   -1 sprites/openttdgui.pcx 578 376 01 7 7 3 4
+   -1 sprites/openttdgui.pcx 594 376 01 7 7 3 4
+   -1 sprites/openttdgui.pcx 610 376 01 9 10 2 3
+   -1 sprites/openttdgui.pcx 626 376 01 9 10 2 3
+   -1 sprites/openttdgui.pcx 642 376 01 7 11 2 1
+   -1 sprites/openttdgui.pcx 658 376 01 9 9 3 3
Index: split/openttdgui.pcx
===================================================================
Cannot display: file marked as a binary type.
svn:mime-type = application/octet-stream
