allow relocation of station building spot

diff -r ace5839ac9be src/station_gui.cpp
--- a/src/station_gui.cpp	Tue Jan 13 11:53:33 2009 +0000
+++ b/src/station_gui.cpp	Tue Jan 13 13:11:06 2009 +0100
@@ -1196,6 +1196,7 @@ static const WindowDesc _select_station_
 	_select_station_widgets,
 };
 
+void UpdateTileSelection();
 
 /**
  * Check whether we need to show the station selection window.
@@ -1206,16 +1207,24 @@ static const WindowDesc _select_station_
  */
 static bool StationJoinerNeeded(CommandContainer cmd, int w, int h)
 {
-	if (CmdFailed(DoCommand(&cmd, CommandFlagsToDCFlags(GetCommandFlags(cmd.cmd))))) return false;
-
 	/* Only show selection if distant join is enabled in the settings */
 	if (!_settings_game.station.distant_join_stations) return false;
 
-	/* If a window is already opened, we always return true */
-	if (FindWindowById(WC_SELECT_STATION, 0) != NULL) return true;
+	/* If a window is already opened and we didn't ctrl-click on another tile,
+	 * we return true without checking for valid place again */
+	Window *win = NULL;
+	if ((win = FindWindowById(WC_SELECT_STATION, 0)) != NULL) {
+		if (!_ctrl_pressed || cmd.tile == TileVirtXY(_thd.pos.x, _thd.pos.y)) return true;
+
+		delete win;
+		UpdateTileSelection();
+	}
 
 	/* only show the popup, if we press ctrl */
 	if (!_ctrl_pressed) return false;
+
+	/* Now check if we could build there */
+	if (CmdFailed(DoCommand(&cmd, CommandFlagsToDCFlags(GetCommandFlags(cmd.cmd))))) return false;
 
 	/* First test for adjacent station */
 	FindStationsNearby(cmd.tile, w, h, false);
diff -r ace5839ac9be src/viewport.cpp
--- a/src/viewport.cpp	Tue Jan 13 11:53:33 2009 +0000
+++ b/src/viewport.cpp	Tue Jan 13 13:11:06 2009 +0100
@@ -2116,7 +2116,6 @@ static HighLightStyle GetAutorailHT(int 
  * Also drawstyle is determined. Uses _thd.new.* as a buffer and calls SetSelectionTilesDirty() twice,
  * Once for the old and once for the new selection.
  * _thd is TileHighlightData, found in viewport.h
- * Called by MouseLoop() in windows.cpp
  */
 void UpdateTileSelection()
 {
