Packing a vli is const.

diff --git a/src/vehiclelist.cpp b/src/vehiclelist.cpp
index 93dec48..7e42b25 100644
--- a/src/vehiclelist.cpp
+++ b/src/vehiclelist.cpp
@@ -20,7 +20,7 @@
  * Pack a VehicleListIdentifier in a single uint32.
  * @return The packed identifier.
  */
-uint32 VehicleListIdentifier::Pack()
+uint32 VehicleListIdentifier::Pack() const
 {
 	byte c = this->company == OWNER_NONE ? 0xF : (byte)this->company;
 	assert(c             < (1 <<  4));
diff --git a/src/vehiclelist.h b/src/vehiclelist.h
index d09f196..03bfb07 100644
--- a/src/vehiclelist.h
+++ b/src/vehiclelist.h
@@ -34,7 +34,7 @@ struct VehicleListIdentifier {
 	CompanyID company;    ///< The company associated with this list.
 	uint32 index;         ///< A vehicle list type specific index.
 
-	uint32 Pack();
+	uint32 Pack() const;
 	bool Unpack(uint32 data);
 
 	/**
