Index: src/window.cpp
===================================================================
--- src/window.cpp	(revision 19627)
+++ src/window.cpp	(working copy)
@@ -2091,7 +2091,8 @@
 						_cursor.sprite != SPR_CURSOR_QUERY &&
 						_cursor.sprite != SPR_CURSOR_SIGN &&
 						_pause_mode != PM_UNPAUSED &&
-						!_cheats.build_in_pause.value) {
+						!_cheats.build_in_pause.value &&
+						_game_mode != GM_EDITOR) {
 					return;
 				}
 
Index: src/openttd.cpp
===================================================================
--- src/openttd.cpp	(revision 19627)
+++ src/openttd.cpp	(working copy)
@@ -1309,7 +1309,7 @@
 
 	if (!_pause_mode && HasBit(_display_opt, DO_FULL_ANIMATION)) DoPaletteAnimations();
 
-	if (!_pause_mode || _cheats.build_in_pause.value) MoveAllTextEffects();
+	if (!_pause_mode || _cheats.build_in_pause.value || _game_mode == GM_EDITOR) MoveAllTextEffects();
 
 	InputLoop();
 
