Index: src/ai/api/ai_map.hpp.sq
===================================================================
--- src/ai/api/ai_map.hpp.sq	(revision 12185)
+++ src/ai/api/ai_map.hpp.sq	(working copy)
@@ -26,8 +26,7 @@
 	SQAIMap.DefSQStaticMethod(engine, &AIMap::DistanceMax,       "DistanceMax",       3, "xii");
 	SQAIMap.DefSQStaticMethod(engine, &AIMap::DistanceSquare,    "DistanceSquare",    3, "xii");
 	SQAIMap.DefSQStaticMethod(engine, &AIMap::DistanceFromEdge,  "DistanceFromEdge",  2, "xi");
+	SQAIMap.DefSQStaticMethod(engine, &AIMap::DemolishTile,      "DemolishTile",      2, "xi");
 
-	SQAIMap.DefSQMethod(engine, &AIMap::DemolishTile, "DemolishTile", 2, "xi");
-
 	SQAIMap.PostRegister(engine);
 }
Index: src/ai/api/ai_map.hpp
===================================================================
--- src/ai/api/ai_map.hpp	(revision 12185)
+++ src/ai/api/ai_map.hpp	(working copy)
@@ -122,7 +122,7 @@
 	 * @pre t has to be valid (use IsValidTile(t)).
 	 * @return true if and only if the destruction succeeded
 	 */
-	bool DemolishTile(TileIndex t);
+	static bool DemolishTile(TileIndex t);
 };
 
 #endif /* AI_MAP_HPP */
Index: src/ai/api/ai_map.cpp
===================================================================
--- src/ai/api/ai_map.cpp	(revision 12185)
+++ src/ai/api/ai_map.cpp	(working copy)
@@ -63,5 +63,5 @@
 
 bool AIMap::DemolishTile(TileIndex t)
 {
-	return this->DoCommand(t, 0, 0, CMD_LANDSCAPE_CLEAR);
+	return DoCommand(t, 0, 0, CMD_LANDSCAPE_CLEAR);
 }
