--- industry_cmd.cpp	2007-12-09 17:20:18.000000000 +0100
+++ industry_cmd.cpp-new	2007-12-09 17:20:30.000000000 +0100
@@ -1236,9 +1236,9 @@
 				refused_slope |= IsSlopeRefused(tileh, its->slopes_refused);
 			}
 
-			if (ind_behav & (INDUSTRYBEH_ONLY_INTOWN | INDUSTRYBEH_TOWN1200_MORE)) {
+			if (ind_behav & (INDUSTRYBEH_ONLY_INTOWN )) {
 				if (!IsTileType(cur_tile, MP_HOUSE)) {
-					_error_message = STR_029D_CAN_ONLY_BE_BUILT_IN_TOWNS;
+					_error_message = STR_030D_CAN_ONLY_BE_BUILT_IN_TOWNS;
 					return false;
 				}
 				if (CmdFailed(DoCommand(cur_tile, 0, 0, 0, CMD_LANDSCAPE_CLEAR))) return false;
