Index: src/core/alloc_type.hpp
===================================================================
--- src/core/alloc_type.hpp	(revision 12737)
+++ src/core/alloc_type.hpp	(working copy)
@@ -88,6 +88,9 @@
 	 * Memory release for a single class instance.
 	 * @param ptr  the memory to free.
 	 * @param size the amount of allocated memory (unused).
+	 *
+	 * @warning The value of the \a size parameter can only be trusted for
+	 *          classes that have their own (virtual) destructor method.
 	 */
 	void operator delete(void *ptr, size_t size) { free(ptr); }
 
@@ -95,6 +98,9 @@
 	 * Memory release for an array of class instances.
 	 * @param ptr  the memory to free.
 	 * @param size the amount of allocated memory (unused).
+	 *
+	 * @warning The value of the \a size parameter can only be trusted for
+	 *          classes that have their own (virtual) destructor method.
 	 */
 	void operator delete[](void *ptr, size_t size) { free(ptr); }
 };
