Index: src/console_gui.cpp
===================================================================
--- src/console_gui.cpp	(revision 16003)
+++ src/console_gui.cpp	(working copy)
@@ -383,6 +383,11 @@
  */
 static void IConsoleHistoryAdd(const char *cmd)
 {
+	//Do not put empty command in history
+	if (StrEmpty(cmd)) return;
+	//Do not put in history if command is same as previous
+	if (_iconsole_history[0] && strcmp(_iconsole_history[0],cmd)==0) return;
+
 	free(_iconsole_history[ICON_HISTORY_SIZE - 1]);
 
 	memmove(&_iconsole_history[1], &_iconsole_history[0], sizeof(_iconsole_history[0]) * (ICON_HISTORY_SIZE - 1));
