--- script_engine.hpp.old	2015-01-20 14:44:00.000000000 +0100
+++ script_engine.hpp	2015-01-20 14:44:08.000000000 +0100
@@ -108,7 +107,6 @@
 	 * Get the maximum speed of an engine.
 	 * @param engine_id The engine to get the maximum speed of.
 	 * @pre IsValidEngine(engine_id).
-	 * @pre GetVehicleType(engine_id) != ScriptVehicle::VT_TRAIN || !IsWagon(engine_id).
 	 * @return The maximum speed the engine has.
 	 * @note The speed is in OpenTTD's internal speed unit.
 	 *       This is mph / 1.6, which is roughly km/h.
