Index: src/saveload/saveload.cpp
===================================================================
--- src/saveload/saveload.cpp	(revision 27603)
+++ src/saveload/saveload.cpp	(working copy)
@@ -2008,7 +2008,7 @@
 		if (tmp[0] != lzo_adler32(0, out, size + sizeof(uint32))) SlErrorCorrupt("Bad checksum");
 
 		/* Decompress */
-		lzo1x_decompress_safe(out + sizeof(uint32) * 1, size, buf, &len, NULL);
+		lzo1x_decompress(out + sizeof(uint32) * 1, size, buf, &len, NULL);
 		return len;
 	}
 };
