Index: economy.c
===================================================================
--- economy.c	(revision 3285)
+++ economy.c	(working copy)
@@ -124,7 +124,7 @@
 /* Count vehicles */
 	{
 		Vehicle *v;
-		int32 min_profit = _score_info[SCORE_MIN_PROFIT].needed;
+		int32 min_profit = 0x7fffffff;
 		uint num = 0;
 
 		FOR_ALL_VEHICLES(v) {
@@ -143,7 +143,9 @@
 		}
 
 		_score_part[owner][SCORE_VEHICLES] = num;
-		if (min_profit > 0)
+		// set the minimum profit to 0, unless the player has vehicles that affected the min. profit
+		_score_part[owner][SCORE_MIN_PROFIT] = 0;
+		if ((min_profit > 0) && (num != 0))
 			_score_part[owner][SCORE_MIN_PROFIT] = min_profit;
 	}
 
