=== src/road.h
==================================================================
--- src/road.h	(/trunk)	(revision 295)
+++ src/road.h	(/branch/new_town)	(revision 295)
@@ -6,6 +6,7 @@
 #define ROAD_H
 
 #include "helpers.hpp"
+#include "rail_map.h"
 
 /**
  * The different roadtypes we support
@@ -102,6 +103,16 @@
 DECLARE_ENUM_AS_BIT_SET(RoadBits);
 
 /**
+ * Count the number of ON RoadBits
+ * @param The given RoadBits value
+ * @return The number of ON RoadBits
+ */
+static inline uint CountRoadBits(RoadBits r)
+{
+	return COUNTBITS(r, 4, 0);
+}
+
+/**
  * Calculate the complement of a RoadBits value
  *
  * Simply flips all bits in the RoadBits value to get the complement
@@ -116,6 +127,52 @@
 }
 
 /**
+ * Calculate the mirrored RoadBits
+ *
+ * Simply move the bits to their new position.
+ *
+ * @param r The given RoadBits value
+ * @return the mirrored
+ */
+static inline RoadBits MirrorRoadBits(RoadBits r)
+{
+	return ((RoadBits)GB(r, 2, 2) | (RoadBits)(GB(r, 0, 2) << 2)); 
+}
+
+/**
+ *
+ * Calculate rotated RoadBits
+ *
+ * Move the Roadbits til they are in their final position.
+ *
+ * @param r The givne RoadBits value
+ * @param times how many times we want to rotate the Roatbits
+ * @param clockwise the direction of the rotatiion
+ * @return the rotated
+ */
+static inline RoadBits RotateRoadBits(RoadBits r, int times, bool clockwise)
+{
+	/* Make negative rotation appear sensefull */
+	if (times < 0) {
+		clockwise ^= clockwise;
+		times = -times;
+	}
+
+	/* Do no rotations over 360° */
+	times = times % 4;
+
+	if (times == 0) return r;
+
+	if (clockwise) times = 4 - times;
+
+	/* rotate the bits counter clockwise */
+	for ( ; times > 0; times--) {
+		r = (RoadBits)GB(r, 3, 1) | (RoadBits)(GB(r, 0, 3) << 1);
+	}
+	return r;
+}
+
+/**
  * Create the road-part which belongs to the given DiagDirection
  *
  * This function returns a RoadBits value which belongs to
@@ -160,7 +217,56 @@
  */
 bool CheckAllowRemoveRoad(TileIndex tile, RoadBits remove, Owner owner, bool *edge_road, RoadType rt);
 
+const TileIndexDiffC _roadblock_tileadd[] = {
+	{ 0, -1}, ///< Direction NW
+	{ 1,  0}, ///< Direction SW
+	{ 0,  1}, ///< Direction SE
+	{-1,  0}, ///< Direction NE
+
+	/* Store the first 3 elements again.
+	 * Lets us rotate without using &3. */
+	{ 0, -1}, ///< Direction NW
+	{ 1,  0}, ///< Direction SW
+	{ 0,  1}  ///< Direction SE
+};
+
 /**
+ * Clean up unneccesary RoadBits of a planed tile.
+ * @param tile current tile
+ * @param org_rb planed RoadBits
+ * @return optimised RoadBits
+ */
+RoadBits CleanUpRoadBits(const TileIndex tile, RoadBits rcmd);
+
+/**
+ * Return if the tile is a valid tile for a crossing.
+ *
+ * @note function is overloaded
+ * @param tile the curent tile
+ * @param ax the axis of the road over the rail
+ * @return true if it is a valid tile
+ */
+static inline bool IsPossibleCrossing(const TileIndex tile, Axis ax)
+{
+	return (IsTileType(tile, MP_RAILWAY) &&
+		!HasSignals(tile) &&
+		GetTrackBits(tile) == (ax == AXIS_X ?  TRACK_BIT_Y : TRACK_BIT_X) &&
+		GetTileSlope(tile, NULL) == SLOPE_FLAT);
+}
+
+/**
+ * Return if the tile is a valid tile for a crossing.
+ *
+ * @note function is overloaded
+ * @param tile the curent tile
+ * @param rb the roadbits of the road over the rail
+ * @return true if it is a valid tile
+ */
+static inline bool IsPossibleCrossing(const TileIndex tile, RoadBits rb) {
+	return IsPossibleCrossing(tile, rb == ROAD_X ? AXIS_X : AXIS_Y);
+}
+
+/**
  * Draw the catenary for tram road bits
  * @param ti   information about the tile (position, slope)
  * @param tram the roadbits to draw the catenary for
=== src/macros.h
==================================================================
--- src/macros.h	(/trunk)	(revision 295)
+++ src/macros.h	(/branch/new_town)	(revision 295)
@@ -300,7 +300,25 @@
 	return x ^= (T)1 << y;
 }
 
+/**
+ * Counts the on bits.
+ * 
+ * @param x The variable form which the bits are counted.
+ * @param size The range how far the count should go.
+ * @param pos The bit where we want to start
+ * @return The number of on bits.
+ */
+template<typename T> static inline uint COUNTBITS(T& x, int size, int pos)
+{
+	uint count = 0;
+	size += pos;
+	for (; pos < size; pos++) {
+		if (x & (T)(1 << pos)) count++;
+	}
+	return count;
+}
 
+
 /* checking more bits. Maybe unneccessary, but easy to use */
 /**
  * Check several bits in a value.
=== src/town_cmd.cpp
==================================================================
--- src/town_cmd.cpp	(/trunk)	(revision 295)
+++ src/town_cmd.cpp	(/branch/new_town)	(revision 295)
@@ -563,20 +563,6 @@
 	/* not used */
 }
 
-
-static const TileIndexDiffC _roadblock_tileadd[] = {
-	{ 0, -1},
-	{ 1,  0},
-	{ 0,  1},
-	{-1,  0},
-
-	/* Store the first 3 elements again.
-	 * Lets us rotate without using &3. */
-	{ 0, -1},
-	{ 1,  0},
-	{ 0,  1}
-};
-
 /**
  * Distance multiplyer
  * Defines the possible distances between 2 road tiles
@@ -791,34 +777,7 @@
 	} else if (!lx && ly) {   ///< It is a X-dir road tile
 		return ROAD_X;
 	} else {                  ///< It is a crossing tile
-		/* Presets for junctions on slopes
-		 * not nice :( */
-		switch (GetTileSlope(tile, NULL)) {
-			case SLOPE_W:
-				return ROAD_NW | ROAD_SW;
-			case SLOPE_S:
-				return ROAD_SE | ROAD_SW;
-			case SLOPE_SW:
-				return ROAD_Y | ROAD_SW;
-			case SLOPE_E:
-				return ROAD_NE | ROAD_SE;
-			case SLOPE_SE:
-				return ROAD_X | ROAD_SE;
-			case SLOPE_N:
-				return ROAD_NW | ROAD_NE;
-			case SLOPE_NW:
-				return ROAD_X | ROAD_NW;
-			case SLOPE_NE:
-				return ROAD_Y | ROAD_NE;
-			case SLOPE_STEEP_W:
-			case SLOPE_STEEP_N:
-				return ROAD_X;
-			case SLOPE_STEEP_S:
-			case SLOPE_STEEP_E:
-				return ROAD_Y;
-			default:
-				return ROAD_ALL;
-		}
+		return ROAD_ALL;
 	}
 }
 
@@ -874,8 +833,6 @@
 {
 	RoadBits rcmd;
 	TileIndex tmptile;
-	DiagDirection i;
-	int j;
 	TileIndex tile = *tile_ptr;
 
 	TILE_ASSERT(tile);
@@ -947,22 +904,10 @@
 		 * Always OK. */
 		_grow_town_result = 0;
 
-		switch (_patches.town_layout) {
-			default: NOT_REACHED();
+		if (_patches.town_layout == TL_NO_ROADS) return; ///< Disallow roads
 
-			case TL_NO_ROADS: /* Disallow Roads */
-				return;
+		rcmd = ROAD_ALL;
 
-			case TL_3X3_GRID:
-			case TL_2X2_GRID:
-			 	rcmd = GetTownRoadGridElement(t1, tile);
-				break;
-
-			case TL_BETTER_ROADS:
-			case TL_ORIGINAL:
-				rcmd = (RoadBits)(1 << (block ^ 2));
-				break;
-		}
 	} else {
 		int i;
 		bool allow_house = false;
@@ -1045,44 +990,55 @@
 	/* Return if a water tile */
 	if (IsClearWaterTile(tile)) return;
 
-	/* Determine direction of slope,
-	 *  and build a road if not a special slope. */
-	switch (GetTileSlope(tile, NULL)) {
-		case SLOPE_SW: i = DIAGDIR_NE; break;
-		case SLOPE_SE: i = DIAGDIR_NW; break;
-		case SLOPE_NW: i = DIAGDIR_SE; break;
-		case SLOPE_NE: i = DIAGDIR_SW; break;
+	DiagDirection bridge_dir;
 
+	rcmd = CleanUpRoadBits(tile, rcmd);
+	switch (rcmd) {
+		case ROAD_NONE: return; break; ///< If there are no more RoadBits further actions are needless.
+
+		/* Determine direction of slope,
+		 *  and build a road if not a special slope. */
+		case ROAD_NW: bridge_dir = DIAGDIR_SE; break;
+		case ROAD_SW: bridge_dir = DIAGDIR_NE; break;
+		case ROAD_SE: bridge_dir = DIAGDIR_NW; break;
+		case ROAD_NE: bridge_dir = DIAGDIR_SW; break;
+
 		default:
 build_road_and_exit:
 			if (CmdSucceeded(DoCommand(tile, rcmd, t1->index, DC_EXEC | DC_AUTO | DC_NO_WATER, CMD_BUILD_ROAD))) {
 				_grow_town_result = -1;
 			}
 			return;
+build_road_and_exit_b:
+			/* XXX not really nice but it works this way
+			 * Build at first a L junction then remove the dummy RoadBit.
+			 * This improves the roads building near water. */
+			if (CmdSucceeded(DoCommand(tile, rcmd | RotateRoadBits(rcmd, 1, true), t1->index, DC_EXEC | DC_AUTO | DC_NO_WATER, CMD_BUILD_ROAD))) {
+				DoCommand(tile, RotateRoadBits(rcmd, 1, true), t1->index, DC_EXEC | DC_AUTO | DC_NO_WATER, CMD_REMOVE_ROAD);
+				_grow_town_result = -1;
+			} else if (CmdSucceeded(DoCommand(tile, rcmd | RotateRoadBits(rcmd, 1, false), t1->index, DC_EXEC | DC_AUTO | DC_NO_WATER, CMD_BUILD_ROAD))) {
+				DoCommand(tile, RotateRoadBits(rcmd, 1, false), t1->index, DC_EXEC | DC_AUTO | DC_NO_WATER, CMD_REMOVE_ROAD);
+				_grow_town_result = -1;
+			}
+			return;
 	}
 
-	/* Check if the bridge is in the right direction */
-	if ((rcmd == ROAD_X && (i == DIAGDIR_NW || i == DIAGDIR_SE)) ||
-			(rcmd == ROAD_Y && (i == DIAGDIR_NE || i == DIAGDIR_SW))) {
-		goto build_road_and_exit;
-	}
+	/* Don't build bridges on flat slopes */
+	if (GetTileSlope(tile, NULL) == SLOPE_FLAT) goto build_road_and_exit;
 
+	/* Get the end of the bridge. */
+	int dist = 0;
 	tmptile = tile;
-
-	/* Now it contains the direction of the slope */
-	j = -11; // max 11 tile long bridges
 	do {
-		if (++j == 0)
-			goto build_road_and_exit;
-		tmptile = TILE_MASK(tmptile + TileOffsByDiagDir(i));
+		if (++dist >= 11) goto build_road_and_exit_b; ///< max 11 tile long bridges
+		tmptile = TILE_MASK(tmptile + TileOffsByDiagDir(bridge_dir));
 	} while (IsClearWaterTile(tmptile));
 
-	/* no water tiles in between? */
-	if (j == -10)
-		goto build_road_and_exit;
+	/* No water tiles in between? */
+	if (dist == 1) goto build_road_and_exit;
 
 	/* Quit if it selecting an appropiate bridge type fails a large number of times. */
-	j = 22;
+	dist = 22;
 	{
 		int32 bridge_len = GetBridgeLength(tile, tmptile);
 		do {
@@ -1094,7 +1050,8 @@
 				/* obviously, if building any bridge would fail, there is no need to try other bridge-types */
 				return;
 			}
-		} while (--j != 0);
+		} while (--dist != 0);
+		goto build_road_and_exit_b;
 	}
 }
 
=== src/road.cpp
==================================================================
--- src/road.cpp	(/trunk)	(revision 295)
+++ src/road.cpp	(/branch/new_town)	(revision 295)
@@ -0,0 +1,136 @@
+#include "stdafx.h"
+#include "openttd.h"
+#include "functions.h"
+#include "road.h"
+#include "road_map.h"
+#include "water_map.h"
+#include "macros.h"
+
+RoadBits CleanUpRoadBits(const TileIndex tile, RoadBits org_rb)
+{
+	RoadBits prefered_rb = ROAD_NONE;
+
+	for (uint8 i = 0 ; i <= 3 ; i++) {
+		/* Cycle counter around the tile NW->SW->SE->NE */
+		const TileIndex neighbor_tile = AddTileIndexDiffCWrap(tile, _roadblock_tileadd[i]);
+
+		/* Get the Roadbit pointing to the neighbor_tile */
+		const RoadBits current_rb = RotateRoadBits(ROAD_NW, i, false);
+
+		/* If the roadbit is in the current plan */
+		if (HASBITS(org_rb, current_rb)) {
+			bool connective = false;
+			const RoadBits mirrored_rb = MirrorRoadBits(current_rb);
+
+			switch (GetTileType(neighbor_tile)) {
+				/* Allways connective ones */
+				case MP_CLEAR: case MP_TREES:
+					connective = true;
+					break;
+
+				/* The conditionaly connective ones */
+				case MP_TUNNELBRIDGE:
+				case MP_STATION:
+				case MP_ROAD: {
+					const RoadBits neighbor_rb_road = GetAnyRoadBits(neighbor_tile, ROADTYPE_ROAD);
+					const RoadBits neighbor_rb_tram = GetAnyRoadBits(neighbor_tile, ROADTYPE_TRAM);
+
+					/* Accept only connective tiles */
+					connective = HASBITS(neighbor_rb_road | neighbor_rb_tram, mirrored_rb) || CountRoadBits(neighbor_rb_road | neighbor_rb_tram) == 1;
+					/* Prefer road */
+					if (HASBITS(neighbor_rb_road, mirrored_rb)) prefered_rb |= current_rb;
+
+				} break;
+
+				case MP_RAILWAY:
+					connective = IsPossibleCrossing(neighbor_tile, current_rb | mirrored_rb);
+					break;
+
+				case MP_WATER:
+					/* Check for real water tile */
+					connective = !IsWater(neighbor_tile);
+					break;
+
+				/* The defentetly not connective ones */
+				default: break;
+			}
+
+			/* If the neighbor tile is inconnective remove the planed road connection to it */
+			if (!connective) org_rb ^= current_rb;
+
+		}
+	}
+
+	/* Optimise the Roadbits for slopes */
+	RoadBits slope_rb[3]; ///< recommended RoadBits
+	switch (GetTileSlope(tile, NULL)) {
+		case SLOPE_N:
+			slope_rb[0] = ROAD_NW | ROAD_NE;
+			slope_rb[1] = ROAD_X;
+			slope_rb[2] = ROAD_Y;
+			break;
+
+		case SLOPE_W:
+			slope_rb[0] = ROAD_NW | ROAD_SW;
+			slope_rb[1] = ROAD_X;
+			slope_rb[2] = ROAD_Y;
+			break;
+
+		case SLOPE_S:
+			slope_rb[0] = ROAD_SW | ROAD_SE;
+			slope_rb[1] = ROAD_X;
+			slope_rb[2] = ROAD_Y;
+			break;
+
+		case SLOPE_E:
+			slope_rb[0] = ROAD_SE | ROAD_NE;
+			slope_rb[1] = ROAD_X;
+			slope_rb[2] = ROAD_Y;
+			break;
+
+		case SLOPE_NW:
+			slope_rb[0] = ROAD_X | ROAD_NW;
+			slope_rb[1] = ROAD_Y;
+			slope_rb[2] = ROAD_NONE;
+			break;
+
+		case SLOPE_SW:
+			slope_rb[0] = ROAD_Y | ROAD_SW;
+			slope_rb[1] = ROAD_X;
+			slope_rb[2] = ROAD_NONE;
+			break;
+
+		case SLOPE_SE:
+			slope_rb[0] = ROAD_X | ROAD_SE;
+			slope_rb[1] = ROAD_Y;
+			slope_rb[2] = ROAD_NONE;
+			break;
+
+		case SLOPE_NE:
+			slope_rb[0] = ROAD_Y | ROAD_NE;
+			slope_rb[1] = ROAD_X;
+			slope_rb[2] = ROAD_NONE;
+			break;
+
+		default:
+			/* No further calculation needed cause all
+			 *  directions are nice at such slopes. */
+			return org_rb;
+			break;
+	}
+
+	for (uint cur = 0; cur < 3; cur++) {
+
+		if (slope_rb[cur] == ROAD_ALL)
+			return org_rb;
+
+		if (slope_rb[cur] == ROAD_NONE && cur > 0)
+			return org_rb & slope_rb[cur - 1];
+
+		if (HASBITS(prefered_rb, slope_rb[cur]))
+			return prefered_rb & slope_rb[cur];
+	}
+
+	return ROAD_NONE;
+}
+
=== src/road_cmd.cpp
==================================================================
--- src/road_cmd.cpp	(/trunk)	(revision 295)
+++ src/road_cmd.cpp	(/branch/new_town)	(revision 295)
@@ -33,19 +33,6 @@
 #include "tunnel_map.h"
 #include "misc/autoptr.hpp"
 
-
-static uint CountRoadBits(RoadBits r)
-{
-	uint count = 0;
-
-	if (r & ROAD_NW) ++count;
-	if (r & ROAD_SW) ++count;
-	if (r & ROAD_SE) ++count;
-	if (r & ROAD_NE) ++count;
-	return count;
-}
-
-
 bool CheckAllowRemoveRoad(TileIndex tile, RoadBits remove, Owner owner, bool *edge_road, RoadType rt)
 {
 	RoadBits present;
@@ -1206,8 +1193,8 @@
 			/* Show an animation to indicate road work */
 			if (t->road_build_months != 0 &&
 					(DistanceManhattan(t->xy, tile) < 8 || grp != 0) &&
-					GetRoadTileType(tile) == ROAD_TILE_NORMAL && (GetAllRoadBits(tile) == ROAD_X || GetAllRoadBits(tile) == ROAD_Y)) {
-				if (GetTileSlope(tile, NULL) == SLOPE_FLAT && EnsureNoVehicleOnGround(tile) && CHANCE16(1, 20)) {
+					GetRoadTileType(tile) == ROAD_TILE_NORMAL && CountRoadBits(GetAllRoadBits(tile)) > 1 ) {
+				if (GetTileSlope(tile, NULL) == SLOPE_FLAT && EnsureNoVehicleOnGround(tile) && CHANCE16(1, 40)) {
 					StartRoadWorks(tile);
 
 					SndPlayTileFx(SND_21_JACKHAMMER, tile);
@@ -1245,6 +1232,27 @@
 		}
 	} else if (IncreaseRoadWorksCounter(tile)) {
 		TerminateRoadWorks(tile);
+
+		/* Generate a nicer town surface */
+		const RoadBits old_rb = GetAnyRoadBits(tile, ROADTYPE_ROAD);
+		const RoadBits new_rb = CleanUpRoadBits(tile, old_rb);
+		if (old_rb != new_rb) {
+
+			/* Protect against roads with no roadbits */
+			if (new_rb == ROAD_NONE) {
+				const RoadTypes rts = GetRoadTypes(tile) & ComplementRoadTypes(RoadTypeToRoadTypes(ROADTYPE_ROAD));
+
+				if (rts == ROADTYPES_NONE) {
+					DoClearSquare(tile);
+				} else {
+					SetRoadBits(tile, ROAD_NONE, ROADTYPE_ROAD);
+					SetRoadTypes(tile, rts);
+				}
+			} else {
+				SetRoadBits(tile, new_rb, ROADTYPE_ROAD);
+			}
+		}
+
 		MarkTileDirtyByTile(tile);
 	}
 }
=== projects/openttd.vcproj
==================================================================
--- projects/openttd.vcproj	(/trunk)	(revision 295)
+++ projects/openttd.vcproj	(/branch/new_town)	(revision 295)
@@ -320,6 +320,9 @@
 				RelativePath=".\..\src\rail.cpp">
 			</File>
 			<File
+				RelativePath=".\..\src\road.cpp">
+			</File>
+			<File
 				RelativePath=".\..\src\saveload.cpp">
 			</File>
 			<File
=== projects/openttd_vs80.vcproj
==================================================================
--- projects/openttd_vs80.vcproj	(/trunk)	(revision 295)
+++ projects/openttd_vs80.vcproj	(/branch/new_town)	(revision 295)
@@ -656,6 +656,10 @@
 				>
 			</File>
 			<File
+				RelativePath=".\..\src\road.cpp"
+				>
+			</File>
+			<File
 				RelativePath=".\..\src\saveload.cpp"
 				>
 			</File>
=== source.list
==================================================================
--- source.list	(/trunk)	(revision 295)
+++ source.list	(/branch/new_town)	(revision 295)
@@ -57,6 +57,7 @@
 players.cpp
 queue.cpp
 rail.cpp
+road.cpp
 saveload.cpp
 screenshot.cpp
 #if SDL

Property changes on: 
___________________________________________________________________
Name: svk:merge
 +6aa0318a-3be1-0310-93fa-89fd2396df07:/trunk:10926

