Index: src/road_gui.cpp
===================================================================
--- src/road_gui.cpp	(revision 19439)
+++ src/road_gui.cpp	(working copy)
@@ -27,6 +27,7 @@
 #include "tunnelbridge.h"
 #include "tilehighlight_func.h"
 #include "company_base.h"
+#include "tunnelbridge_map.h"
 
 #include "table/sprites.h"
 #include "table/strings.h"
@@ -174,6 +175,10 @@
 
 static void PlaceRoad_Tunnel(TileIndex tile)
 {
+	if (IsTunnelTile(tile)) {
+		DoCommandP(tile, DiagDirToRoadBits(ReverseDiagDir(GetTunnelBridgeDirection(tile))) | _cur_roadtype << 4, 0, CMD_BUILD_ROAD | CMD_MSG(STR_ERROR_CAN_T_BUILD_TUNNEL_HERE), CcBuildRoadTunnel);
+		return;
+	}
 	DoCommandP(tile, 0x200 | RoadTypeToRoadTypes(_cur_roadtype), 0, CMD_BUILD_TUNNEL | CMD_MSG(STR_ERROR_CAN_T_BUILD_TUNNEL_HERE), CcBuildRoadTunnel);
 }
 
