Index: src/lang/english.txt
===================================================================
--- src/lang/english.txt	(revision 17224)
+++ src/lang/english.txt	(working copy)
@@ -712,6 +712,7 @@
 STR_NEWS_MESSAGE_TYPE_CHANGES_OF_CARGO_ACCEPTANCE               :{YELLOW}Changes to cargo acceptance
 STR_NEWS_MESSAGE_TYPE_SUBSIDIES                                 :{YELLOW}Subsidies
 STR_NEWS_MESSAGE_TYPE_GENERAL_INFORMATION                       :{YELLOW}General information
+STR_NEWS_MESSAGE_TYPE_ADVICE_VEH_OLD                            :{YELLOW}Advice on vehicle age
 
 STR_NEWS_FIRST_TRAIN_ARRIVAL                                    :{BIGFONT}{BLACK}Citizens celebrate . . .{}First train arrives at {STATION}!
 STR_NEWS_FIRST_BUS_ARRIVAL                                      :{BIGFONT}{BLACK}Citizens celebrate . . .{}First bus arrives at {STATION}!
Index: src/news_type.h
===================================================================
--- src/news_type.h	(revision 17224)
+++ src/news_type.h	(working copy)
@@ -25,6 +25,7 @@
 	NT_INDUSTRY_OTHER,  ///< Production changes of industry serviced by competitor(s)
 	NT_INDUSTRY_NOBODY, ///< Other industry production changes
 	NT_ADVICE,          ///< Bits of news about vehicles of the company
+	NT_AGE,             ///< Bits of news about old vehicles of the company
 	NT_NEW_VEHICLES,    ///< New vehicle has become available
 	NT_ACCEPTANCE,      ///< A type of cargo is (no longer) accepted
 	NT_SUBSIDIES,       ///< News about subsidies (announcements, expirations, acceptance)
@@ -50,6 +51,7 @@
 	NS_INDUSTRY_OTHER,   ///< NT_INDUSTRY_OTHER
 	NS_INDUSTRY_NOBODY,  ///< NT_INDUSTRY_NOBODY
 	NS_ADVICE,           ///< NT_ADVICE
+	NS_AGE,              ///< NT_AGE
 	NS_NEW_VEHICLES,     ///< NT_NEW_VEHICLES
 	NS_ACCEPTANCE,       ///< NT_ACCEPTANCE
 	NS_SUBSIDIES,        ///< NT_SUBSIDIES
Index: src/news_gui.cpp
===================================================================
--- src/news_gui.cpp	(revision 17224)
+++ src/news_gui.cpp	(working copy)
@@ -138,6 +138,7 @@
 	{ NT_INDUSTRY_OTHER,   NM_THIN,     NF_NONE, NULL                    }, ///< NS_INDUSTRY_OTHER
 	{ NT_INDUSTRY_NOBODY,  NM_THIN,     NF_NONE, NULL                    }, ///< NS_INDUSTRY_NOBODY
 	{ NT_ADVICE,           NM_SMALL,    NF_NONE, NULL                    }, ///< NS_ADVICE
+	{ NT_AGE,              NM_SMALL,    NF_NONE, NULL                    }, ///< NS_AGE
 	{ NT_NEW_VEHICLES,     NM_NORMAL,   NF_NONE, DrawNewsNewVehicleAvail }, ///< NS_NEW_VEHICLES
 	{ NT_ACCEPTANCE,       NM_SMALL,    NF_NONE, NULL                    }, ///< NS_ACCEPTANCE
 	{ NT_SUBSIDIES,        NM_NORMAL,   NF_NONE, NULL                    }, ///< NS_SUBSIDIES
@@ -162,6 +163,7 @@
 	NewsTypeData("production_other",  30, SND_BEGIN,       STR_NEWS_MESSAGE_TYPE_INDUSTRY_CHANGES_SERVED_BY_OTHER   ),  ///< NT_INDUSTRY_OTHER
 	NewsTypeData("production_nobody", 30, SND_BEGIN,       STR_NEWS_MESSAGE_TYPE_INDUSTRY_CHANGES_UNSERVED          ),  ///< NT_INDUSTRY_NOBODY
 	NewsTypeData("advice",           150, SND_BEGIN,       STR_NEWS_MESSAGE_TYPE_ADVICE_INFORMATION_ON_COMPANY      ),  ///< NT_ADVICE
+	NewsTypeData("age",              150, SND_BEGIN,       STR_NEWS_MESSAGE_TYPE_ADVICE_VEH_OLD                     ),  ///< NT_AGE
 	NewsTypeData("new_vehicles",      30, SND_1E_OOOOH,    STR_NEWS_MESSAGE_TYPE_NEW_VEHICLES                       ),  ///< NT_NEW_VEHICLES
 	NewsTypeData("acceptance",        90, SND_BEGIN,       STR_NEWS_MESSAGE_TYPE_CHANGES_OF_CARGO_ACCEPTANCE        ),  ///< NT_ACCEPTANCE
 	NewsTypeData("subsidies",        180, SND_BEGIN,       STR_NEWS_MESSAGE_TYPE_SUBSIDIES                          ),  ///< NT_SUBSIDIES
Index: src/vehicle.cpp
===================================================================
--- src/vehicle.cpp	(revision 17224)
+++ src/vehicle.cpp	(working copy)
@@ -897,7 +897,7 @@
 	}
 
 	SetDParam(0, v->index);
-	AddVehicleNewsItem(str, NS_ADVICE, v->index);
+	AddVehicleNewsItem(str, NS_AGE, v->index);
 }
 
 /**
