Index: src/genworld_gui.cpp
===================================================================
--- src/genworld_gui.cpp	(revision 22255)
+++ src/genworld_gui.cpp	(working copy)
@@ -603,8 +603,6 @@
 				break;
 
 			case GLAND_GENERATE_BUTTON: // Generate
-				MakeNewgameSettingsLive();
-
 				if (mode == GLWM_HEIGHTMAP &&
 						(this->x * 2 < (1U << _settings_newgame.game_creation.map_x) ||
 						this->x / 2 > (1U << _settings_newgame.game_creation.map_x) ||
@@ -616,6 +614,7 @@
 						this,
 						LandscapeGenerationCallback);
 				} else {
+					MakeNewgameSettingsLive();
 					StartGeneratingLandscape(mode);
 				}
 				break;
