Index: src/station_cmd.cpp
===================================================================
--- src/station_cmd.cpp	(Revision 24178)
+++ src/station_cmd.cpp	(Arbeitskopie)
@@ -3207,9 +3207,26 @@
 			}
 
 			if (!skip) {
-				int b = ge->last_speed - 85;
-				if (b >= 0) rating += b >> 2;
+				int b = ge->last_speed;
 
+				if ((st->last_vehicle_type == VEH_TRAIN) || (st->last_vehicle_type == VEH_AIRCRAFT)) {
+					b -= 85;
+					if (b >= 0) rating += b/4;
+				}
+				else	{
+					if (st->last_vehicle_type == VEH_ROAD)	{
+						b -= 60;
+						if (b >= 0) rating += b/2;
+					}
+					else	{
+						if (st->last_vehicle_type == VEH_SHIP)	{
+							b -= 20;
+							if (b >= 0) rating += b*2;
+						}
+						else NOT_REACHED();
+					}
+				}
+
 				byte days = ge->days_since_pickup;
 				if (st->last_vehicle_type == VEH_SHIP) days >>= 2;
 				(days > 21) ||
