Index: src/bridge.h
===================================================================
--- src/bridge.h	(revision 21000)
+++ src/bridge.h	(working copy)
@@ -75,4 +75,6 @@
 static const uint MAX_BRIDGE_LENGTH             =  16; ///<Maximum length of the bridge
 static const uint MAX_BRIDGE_LENGTH_LONGBRIDGES = 100; ///<Maximum length of the bridge with enabled longbridges
 
+TileIndex GetBridgeHighlightTile(TileIndex tile, DiagDirection dir);
+
 #endif /* BRIDGE_H */
Index: src/tunnelbridge_cmd.cpp
===================================================================
--- src/tunnelbridge_cmd.cpp	(revision 21000)
+++ src/tunnelbridge_cmd.cpp	(working copy)
@@ -256,6 +256,49 @@
 	}
 }
 
+TileIndex GetBridgeHighlightTile(TileIndex tile, DiagDirection dir)
+{
+	assert(tile < MapSize());
+	if (!IsValidDiagDirection(dir)) dir = ReverseDiagDir(GetInclinedSlopeDirection(GetTileSlope(tile, NULL)));
+	if (!IsValidDiagDirection(dir)) return INVALID_TILE;
+
+	if (!AllowBridgeRamp(tile)) return INVALID_TILE;
+	
+	uint z;
+	Slope tileh = GetTileSlope(tile, &z);
+	if (!CanBuildBridgeRamp(dir, tileh, &z)) return INVALID_TILE;
+
+	if (MayHaveBridgeAbove(tile) && IsBridgeAbove(tile)) {
+		if (DiagDirToAxis(dir) == GetBridgeAxis(tile) || z >= (GetBridgeHeight(GetSouthernBridgeEnd(tile)) - TILE_HEIGHT)) return INVALID_TILE;
+	}
+
+	TileIndexDiff step = TileOffsByDiagDir(dir);
+	uint map_edge = DistanceFromEdgeDir(tile, dir);
+	uint max_length = (_settings_game.construction.longbridges ? MAX_BRIDGE_LENGTH_LONGBRIDGES : MAX_BRIDGE_LENGTH);
+	max_length = min(max_length, map_edge);
+	bool last = false;
+
+	for (uint length = 1;; length++) {
+		tile = TILE_ADD(tile, step);
+
+		if (!IsTileType(tile, MP_TREES)) {
+			if (!AllowBridgeAbove(tile, dir, z)) return INVALID_TILE;
+			if (IsBridgeAbove(tile)) {
+				if (z >= GetBridgeHeight(GetSouthernBridgeEnd(tile)) - TILE_HEIGHT) return INVALID_TILE;
+				last = true;
+			}
+		}
+		if (z < GetTileMaxZ(tile)) last = true;
+
+		if (AllowBridgeRamp(tile)) {
+			uint cur_z;
+			Slope tileh = GetTileSlope(tile, &cur_z);
+			if (CanBuildBridgeRamp(ReverseDiagDir(dir), tileh, &cur_z) && cur_z == z) return tile;
+		}
+		if (length > max_length || last) return INVALID_TILE;
+	}
+}
+
 /**
  * Build a Bridge
  * @param end_tile end tile
