Index: src/pathfinder/follow_track.hpp
===================================================================
--- src/pathfinder/follow_track.hpp	(revision 27409)
+++ src/pathfinder/follow_track.hpp	(working copy)
@@ -144,7 +144,7 @@
 			 * missing road bit, or inability to connect the
 			 * different bits due to slopes. */
 			if (IsRoadTT() && !IsTram() && TryReverse()) return true;
-			m_err = EC_NO_WAY;
+			if (m_new_td_bits == TRACKDIR_BIT_NONE) m_err = EC_NO_WAY;
 			return false;
 		}
 		if (!Allow90degTurns()) {
