diff --git a/src/widgets/ai_widget.h b/src/widgets/ai_widget.h
index 76b22b8..b6ef02f 100644
--- a/src/widgets/ai_widget.h
+++ b/src/widgets/ai_widget.h
@@ -60,7 +60,7 @@ enum AIDebugWidgets {
 	WID_AID_SCRIPT_GAME,          ///< Game Script button.
 	WID_AID_RELOAD_TOGGLE,        ///< Reload button.
 	WID_AID_LOG_PANEL,            ///< Panel where the log is in.
-	WID_AID_SCROLLBAR,            ///< Scrollbar of the log pannel.
+	WID_AID_SCROLLBAR,            ///< Scrollbar of the log panel.
 	WID_AID_COMPANY_BUTTON_START, ///< Buttons in the VIEW.
 	WID_AID_COMPANY_BUTTON_END = WID_AID_COMPANY_BUTTON_START + MAX_COMPANIES - 1, ///< Last possible button in the VIEW.
 	WID_AID_BREAK_STRING_WIDGETS, ///< The panel to handle the breaking on string.
diff --git a/src/widgets/autoreplace_widget.h b/src/widgets/autoreplace_widget.h
index 5e610a7..a302eea 100644
--- a/src/widgets/autoreplace_widget.h
+++ b/src/widgets/autoreplace_widget.h
@@ -20,7 +20,7 @@ enum ReplaceVehicleWidgets {
 	WID_RV_LEFT_MATRIX,              ///< The matrix on the left.
 	WID_RV_LEFT_SCROLLBAR,           ///< The scrollbar for the matrix on the left.
 	WID_RV_RIGHT_MATRIX,             ///< The matrix on the right.
-	WID_RV_RIGHT_SCROLLBAR,          ///< The scrollbar for the matrxi on the right.
+	WID_RV_RIGHT_SCROLLBAR,          ///< The scrollbar for the matrix on the right.
 	WID_RV_LEFT_DETAILS,             ///< Details of the entry on the left.
 	WID_RV_RIGHT_DETAILS,            ///< Details of the entry on the right.
 
diff --git a/src/widgets/dropdown.cpp b/src/widgets/dropdown.cpp
index dd23499..ffba1a6 100644
--- a/src/widgets/dropdown.cpp
+++ b/src/widgets/dropdown.cpp
@@ -452,7 +452,7 @@ void ShowDropDownList(Window *w, DropDownList *list, int selected, int button, u
  * @param strings       Menu list, end with #INVALID_STRING_ID
  * @param selected      Index of initial selected item.
  * @param button        Button widget number of the parent window \a w that wants the dropdown menu.
- * @param disabled_mask Bitmask for diabled items (items with their bit set are not copied to the dropdown list).
+ * @param disabled_mask Bitmask for disabled items (items with their bit set are not copied to the dropdown list).
  * @param hidden_mask   Bitmask for hidden items (items with their bit set are displayed, but not selectable in the dropdown list).
  * @param width         Width of the dropdown menu. If \c 0, use the width of parent widget \a button.
  */
diff --git a/src/widgets/osk_widget.h b/src/widgets/osk_widget.h
index 6eb977f..25cd928 100644
--- a/src/widgets/osk_widget.h
+++ b/src/widgets/osk_widget.h
@@ -19,7 +19,7 @@ enum OnScreenKeyboardWidgets {
 	WID_OSK_CANCEL,    ///< Cancel key.
 	WID_OSK_OK,        ///< Ok key.
 	WID_OSK_BACKSPACE, ///< Backspace key.
-	WID_OSK_SPECIAL,   ///< Special key (at keyborads often used for tab key).
+	WID_OSK_SPECIAL,   ///< Special key (at keyboards often used for tab key).
 	WID_OSK_CAPS,      ///< Capslock key.
 	WID_OSK_SHIFT,     ///< Shift(lock) key.
 	WID_OSK_SPACE,     ///< Space bar.
diff --git a/src/widgets/rail_widget.h b/src/widgets/rail_widget.h
index a8b933b..e468a3e 100644
--- a/src/widgets/rail_widget.h
+++ b/src/widgets/rail_widget.h
@@ -12,7 +12,7 @@
 #ifndef WIDGETS_RAIL_WIDGET_H
 #define WIDGETS_RAIL_WIDGET_H
 
-/** Widgets of the #BuildRailToolbarWindow ckass. */
+/** Widgets of the #BuildRailToolbarWindow class. */
 enum RailToolbarWidgets {
 	/* Name starts with RA instead of R, because of collision with RoadToolbarWidgets */
 	WID_RAT_CAPTION,        ///< Caption of the window.
@@ -98,10 +98,10 @@ enum BuildSignalWidgets {
 /** Widgets of the #BuildRailDepotWindow class. */
 enum BuildRailDepotWidgets {
 	/* Name starts with BRA instead of BR, because of collision with BuildRoadDepotWidgets */
-	WID_BRAD_DEPOT_NE, ///< Build a depot with the entrace in the north east.
-	WID_BRAD_DEPOT_SE, ///< Build a depot with the entrace in the south east.
-	WID_BRAD_DEPOT_SW, ///< Build a depot with the entrace in the south west.
-	WID_BRAD_DEPOT_NW, ///< Build a depot with the entrace in the north west.
+	WID_BRAD_DEPOT_NE, ///< Build a depot with the entrance in the north east.
+	WID_BRAD_DEPOT_SE, ///< Build a depot with the entrance in the south east.
+	WID_BRAD_DEPOT_SW, ///< Build a depot with the entrance in the south west.
+	WID_BRAD_DEPOT_NW, ///< Build a depot with the entrance in the north west.
 };
 
 /** Widgets of the #BuildRailWaypointWindow class. */
diff --git a/src/widgets/settings_widget.h b/src/widgets/settings_widget.h
index ad9c406..4b3c6c4 100644
--- a/src/widgets/settings_widget.h
+++ b/src/widgets/settings_widget.h
@@ -54,8 +54,8 @@ enum CustomCurrencyWidgets {
 	WID_CC_RATE_DOWN,      ///< Down button.
 	WID_CC_RATE_UP,        ///< Up button.
 	WID_CC_RATE,           ///< Rate of currency.
-	WID_CC_SEPARATOR_EDIT, ///< Seperator edit button.
-	WID_CC_SEPARATOR,      ///< Current seperator.
+	WID_CC_SEPARATOR_EDIT, ///< Separator edit button.
+	WID_CC_SEPARATOR,      ///< Current separator.
 	WID_CC_PREFIX_EDIT,    ///< Prefix edit button.
 	WID_CC_PREFIX,         ///< Current prefix.
 	WID_CC_SUFFIX_EDIT,    ///< Suffix edit button.
diff --git a/src/widgets/terraform_widget.h b/src/widgets/terraform_widget.h
index 7429f41..7f8a4c4 100644
--- a/src/widgets/terraform_widget.h
+++ b/src/widgets/terraform_widget.h
@@ -21,7 +21,7 @@ enum TerraformToolbarWidgets {
 	WID_TT_LEVEL_LAND,                        ///< Level land button.
 	WID_TT_DEMOLISH,                          ///< Demolish aka dynamite button.
 	WID_TT_BUY_LAND,                          ///< Buy land button.
-	WID_TT_PLANT_TREES,                       ///< Plant trees button (note: opens seperate window, no place-push-button).
+	WID_TT_PLANT_TREES,                       ///< Plant trees button (note: opens separate window, no place-push-button).
 	WID_TT_PLACE_SIGN,                        ///< Place sign button.
 	WID_TT_PLACE_OBJECT,                      ///< Place object button.
 };
diff --git a/src/widgets/vehicle_widget.h b/src/widgets/vehicle_widget.h
index ac15d65..3a68b00 100644
--- a/src/widgets/vehicle_widget.h
+++ b/src/widgets/vehicle_widget.h
@@ -56,7 +56,7 @@ enum VehicleDetailsWidgets {
 	WID_VD_DETAILS_CARGO_CARRIED,       ///< Show carried cargo per part of the train.
 	WID_VD_DETAILS_TRAIN_VEHICLES,      ///< Show all parts of the train with their description.
 	WID_VD_DETAILS_CAPACITY_OF_EACH,    ///< Show the capacity of all train parts.
-	WID_VD_DETAILS_TOTAL_CARGO,         ///< Show the capacity and carried cargo amounts aggregrated per cargo of the train.
+	WID_VD_DETAILS_TOTAL_CARGO,         ///< Show the capacity and carried cargo amounts aggregated per cargo of the train.
 };
 
 /** Widgets of the #VehicleListWindow class. */
