diff -r a1e148317984 -r 7c4cb55a8ba1 src/settings_gui.cpp
--- a/src/settings_gui.cpp	Sun Sep 28 14:09:59 2008 +0200
+++ b/src/settings_gui.cpp	Sun Sep 28 14:09:59 2008 +0200
@@ -642,17 +642,18 @@
 
 static const char *_patches_stations[] = {
 	"station.join_stations",
+	"station.adjacent_stations",
+	"station.nonuniform_stations",
+	"station.station_spread",
+	"station.modified_catchment",
+	"construction.road_stop_on_town_road",
+	"",
 	"order.improved_load",
 	"order.selectgoods",
-	"gui.new_nonstop",
-	"station.nonuniform_stations",
-	"station.station_spread",
+	"order.gradual_loading",
+	"",
+	"station.always_small_airport",
 	"order.serviceathelipad",
-	"station.modified_catchment",
-	"order.gradual_loading",
-	"construction.road_stop_on_town_road",
-	"station.adjacent_stations",
-	"station.always_small_airport",
 	"economy.station_noise_level",
 	NULL
 };
@@ -694,6 +695,7 @@
 };
 
 static const char *_patches_vehicles[] = {
+	"gui.new_nonstop",
 	"gui.timetable_in_ticks",
 	"vehicle.realistic_acceleration",
 	"pf.forbid_90_deg",
