Index: currency.c
===================================================================
--- currency.c	(revision 4586)
+++ currency.c	(working copy)
@@ -34,6 +34,7 @@
 	{    6, ' ', CF_NOEURO, "",     " zl"  }, // polish zloty
 	{    6, '.', CF_NOEURO, "",     " Lei" }, // romanian Lei
 	{    5, ' ', CF_NOEURO, "",     " p"   }, // russian rouble
+	{  350, '.', CF_NOEURO, "",     " SIT" }, // slovenian tolar
 	{   13, '.', CF_NOEURO, "",     " Kr"  }, // swedish krona
 	{    1, ' ', CF_NOEURO, "",     ""     }, // custom currency
 };
@@ -61,6 +62,7 @@
 	STR_CURR_PLN,
 	STR_CURR_ROL,
 	STR_CURR_RUR,
+	STR_CURR_SIT,
 	STR_CURR_SEK,
 	STR_CURR_CUSTOM,
 	INVALID_STRING_ID
Index: currency.h
===================================================================
--- currency.h	(revision 4586)
+++ currency.h	(working copy)
@@ -20,7 +20,7 @@
 extern const StringID _currency_string_list[];
 
 // XXX small hack, but makes the rest of the code a bit nicer to read
-#define CUSTOM_CURRENCY_ID 23
+#define CUSTOM_CURRENCY_ID 24
 #define _custom_currency (_currency_specs[CUSTOM_CURRENCY_ID])
 #define _currency ((const CurrencySpec*)&_currency_specs[_opt_ptr->currency])
 
Index: lang/english.txt
===================================================================
--- lang/english.txt	(revision 4586)
+++ lang/english.txt	(working copy)
@@ -894,6 +894,7 @@
 STR_CURR_PLN                                                    :Polish Zloty (PLN)
 STR_CURR_ROL                                                    :Romanian Leu (ROL)
 STR_CURR_RUR                                                    :Russian Rubles (RUR)
+STR_CURR_SIT                                                    :Slovenian Tolar (SIT)
 STR_CURR_SEK                                                    :Swedish Krona (SEK)
 
 STR_CURR_CUSTOM                                                 :Custom...
Index: settings.c
===================================================================
--- settings.c	(revision 4586)
+++ settings.c	(working copy)
@@ -1200,7 +1200,7 @@
 	SDT_GENERAL("diff_custom", SDT_INTLIST, SL_ARR, (SLE_FILE_I16 | SLE_VAR_I32), 0, 0, GameOptions, diff, 17, 0, 0, 0, NULL, STR_NULL, NULL, 0, 3),
 	SDT_GENERAL("diff_custom", SDT_INTLIST, SL_ARR, (SLE_FILE_I16 | SLE_VAR_I32), 0, 0, GameOptions, diff, 18, 0, 0, 0, NULL, STR_NULL, NULL, 4, SL_MAX_VERSION),
 	    SDT_VAR(GameOptions, diff_level,SLE_UINT8, 0, 0, 9,0, 9, STR_NULL, NULL),
-	  SDT_OMANY(GameOptions, currency,  SLE_UINT8, N, 0, 0, CUSTOM_CURRENCY_ID, "GBP|USD|EUR|YEN|ATS|BEF|CHF|CZK|DEM|DKK|ESP|FIM|FRF|GRD|HUF|ISK|ITL|NLG|NOK|PLN|ROL|RUR|SEK|custom", STR_NULL, NULL),
+	  SDT_OMANY(GameOptions, currency,  SLE_UINT8, N, 0, 0, CUSTOM_CURRENCY_ID, "GBP|USD|EUR|YEN|ATS|BEF|CHF|CZK|DEM|DKK|ESP|FIM|FRF|GRD|HUF|ISK|ITL|NLG|NOK|PLN|ROL|RUR|SIT|SEK|custom", STR_NULL, NULL),
 	  SDT_OMANY(GameOptions, units,     SLE_UINT8, N, 0, 1,   2, "imperial|metric|si", STR_NULL, NULL),
 	  SDT_OMANY(GameOptions, town_name, SLE_UINT8, 0, 0, 0,  17, "english|french|german|american|latin|silly|swedish|dutch|finnish|polish|slovakish|norwegian|hungarian|austrian|romanian|czech|swiss|danish", STR_NULL, NULL),
 	  SDT_OMANY(GameOptions, landscape, SLE_UINT8, 0, 0, 0,   3, "normal|hilly|desert|candy", STR_NULL, NULL),
