Index: src/town.h
===================================================================
--- src/town.h	(revision 11621)
+++ src/town.h	(working copy)
@@ -82,7 +82,7 @@
 	TileIndex xy;
 
 	/* Current population of people and amount of houses. */
-	uint16 num_houses;
+	uint32 num_houses;
 	uint32 population;
 
 	/* Town name */
Index: src/saveload.cpp
===================================================================
--- src/saveload.cpp	(revision 11621)
+++ src/saveload.cpp	(working copy)
@@ -29,7 +29,7 @@
 #include "strings.h"
 #include <list>
 
-extern const uint16 SAVEGAME_VERSION = 83;
+extern const uint16 SAVEGAME_VERSION = 84;
 uint16 _sl_version;       ///< the major savegame version identifier
 byte   _sl_minor_version; ///< the minor savegame version, DO NOT USE!
 
Index: src/town_cmd.cpp
===================================================================
--- src/town_cmd.cpp	(revision 11621)
+++ src/town_cmd.cpp	(working copy)
@@ -1929,7 +1929,7 @@
 	_generating_world = true;
 
 	/* The more houses, the faster we grow */
-	amount = RandomRange(t->num_houses / 10) + 3;
+	amount = RandomRange(min(0xffffU, t->num_houses / 10)) + 3;
 	t->num_houses += amount;
 	UpdateTownRadius(t);
 
@@ -2400,7 +2400,8 @@
 	SLE_CONDVAR(Town, population,            SLE_UINT32,                 3, SL_MAX_VERSION),
 
 
-	    SLE_VAR(Town, num_houses,            SLE_UINT16),
+	SLE_CONDVAR(Town, num_houses,            SLE_UINT16,  0, 83),
+	SLE_CONDVAR(Town, num_houses,            SLE_UINT32, 84, SL_MAX_VERSION),
 	SLE_CONDVAR(Town, townnamegrfid,         SLE_UINT32, 66, SL_MAX_VERSION),
 	    SLE_VAR(Town, townnametype,          SLE_UINT16),
 	    SLE_VAR(Town, townnameparts,         SLE_UINT32),
