Index: src/rail_cmd.cpp
===================================================================
--- src/rail_cmd.cpp	(revision 12325)
+++ src/rail_cmd.cpp	(working copy)
@@ -360,6 +360,8 @@
 			if (IsNormalRoad(tile)) {
 				if (HasRoadWorks(tile)) return_cmd_error(STR_ROAD_WORKS_IN_PROGRESS);
 
+				if (GetDisallowedRoadDirections(tile) != DRD_NONE) return_cmd_error(STR_ERR_ONEWAY_ROADS_CAN_T_HAVE_JUNCTION);
+
 				RoadTypes roadtypes = GetRoadTypes(tile);
 				RoadBits road = GetRoadBits(tile, ROADTYPE_ROAD);
 				RoadBits tram = GetRoadBits(tile, ROADTYPE_TRAM);
