diff -r 4173a4a89178 findversion.sh
--- a/findversion.sh	Tue Oct 21 19:26:36 2008 +0000
+++ b/findversion.sh	Wed Oct 22 17:45:19 2008 +0200
@@ -82,8 +82,8 @@ elif [ -d "$ROOT_DIR/.git" ]; then
 	if [ -n "`git diff-index HEAD \"$SRC_DIR\"`" ]; then
 		MODIFIED="2"
 	fi
-	HASH=`LC_ALL=C git rev-parse --verify HEAD 2>/dev/null | cut -c1-8`
-	REV="g$HASH"
+	HASH=`LC_ALL=C git rev-parse --verify HEAD 2>/dev/null`
+	REV="g`echo $HASH | cut -c1-8`"
 	BRANCH=`git branch|grep '[*]' | sed 's~\* ~~;s~^master$~~'`
 	REV_NR=`LC_ALL=C git log --pretty=format:%s "$SRC_DIR" | grep "^(svn r[0-9]*)" | head -n 1 | sed "s~.*(svn r\([0-9]*\)).*~\1~"`
 elif [ -d "$ROOT_DIR/.hg" ]; then
@@ -91,8 +91,8 @@ elif [ -d "$ROOT_DIR/.hg" ]; then
 	if [ -n "`hg status \"$SRC_DIR\" | grep -v '^?'`" ]; then
 		MODIFIED="2"
 	fi
-	HASH=`LC_ALL=C hg parents 2>/dev/null | head -n 1 | cut -d: -f3 | cut -c1-8`
-	REV="h$HASH"
+	HASH=`LC_ALL=C hg parents 2>/dev/null | head -n 1 | cut -d: -f3`
+	REV="h`echo $HASH | cut -c1-8`"
 	BRANCH=`hg branch | sed 's~^default$~~'`
 	REV_NR=`LC_ALL=C hg log -r $HASH:0 -k "svn" -l 1 --template "{desc}\n" "$SRC_DIR" | grep "^(svn r[0-9]*)" | head -n 1 | sed "s~.*(svn r\([0-9]*\)).*~\1~"`
 else
