diff --git a/src/vehicle_gui.cpp b/src/vehicle_gui.cpp
index 1f03319..6e81fec 100644
--- a/src/vehicle_gui.cpp
+++ b/src/vehicle_gui.cpp
@@ -2372,8 +2372,8 @@ static const uint32 _vehicle_command_translation_table[][4] = {
 };
 
 /**
- * This is the Callback method after the cloning attempt of a vehicle
- * @param result the result of the cloning command
+ * This is the Callback method after attempting to start/stop a vehicle
+ * @param result the result of the start/stop command
  * @param tile unused
  * @param p1 vehicle ID
  * @param p2 unused
