Index: src/network/network_server.cpp
===================================================================
--- src/network/network_server.cpp	(revision 23947)
+++ src/network/network_server.cpp	(working copy)
@@ -701,7 +701,7 @@
  */
 NetworkRecvStatus ServerNetworkGameSocketHandler::SendChat(NetworkAction action, ClientID client_id, bool self_send, const char *msg, int64 data)
 {
-	if (this->status != STATUS_ACTIVE) return NETWORK_RECV_STATUS_OKAY;
+	if (this->status < STATUS_AUTHORIZED) return NETWORK_RECV_STATUS_OKAY;
 
 	Packet *p = new Packet(PACKET_SERVER_CHAT);
 
