Index: lang/english.txt
===================================================================
--- lang/english.txt	(revision 5625)
+++ lang/english.txt	(working copy)
@@ -881,6 +881,7 @@
 STR_TOWNNAME_DANISH                                             :Danish
 STR_TOWNNAME_TURKISH                                            :Turkish
 STR_TOWNNAME_ITALIAN                                            :Italian
+STR_TOWNNAME_CATALAN                                            :Catalan
 ############ end of	townname region
 
 STR_CURR_GBP                                                    :Pounds (£)
Index: namegen.c
===================================================================
--- namegen.c	(revision 5625)
+++ namegen.c	(working copy)
@@ -692,6 +692,43 @@
 	return 0;
 }
 
+static byte MakeCatalanTownName(char *buf, uint32 seed) {
+
+	strcpy(buf, "");
+
+	if (SeedModChance(0, 3, seed) == 0) { // real city names
+		strcat(buf, name_catalan_real[SeedModChance(4, lengthof(name_catalan_real), seed)]);
+	} else {
+		uint i;
+		if (SeedModChance(0, 2, seed) == 0) { // prefix
+			strcat(buf, name_catalan_pref[SeedModChance(11, lengthof(name_catalan_pref), seed)]);
+		}
+		else {
+			i = SeedChance(0, 2, seed);
+			if (i == 0) { // masculine form
+				strcat(buf, name_catalan_1m[SeedModChance(4, lengthof(name_catalan_1m), seed)]);
+				strcat(buf, name_catalan_2m[SeedModChance(11, lengthof(name_catalan_2m), seed)]);
+			} else { // feminine form
+				strcat(buf, name_catalan_1f[SeedModChance(4, lengthof(name_catalan_1f), seed)]);
+				strcat(buf, name_catalan_2f[SeedModChance(11, lengthof(name_catalan_2f), seed)]);
+			}
+	
+	
+			if (SeedModChance(15, 5, seed) == 0) {
+				if (SeedModChance(5, 2, seed) == 0) { // generic suffix
+					strcat(buf, name_catalan_3[SeedModChance(4, lengthof(name_catalan_3), seed)]);
+				} else { // river name suffix
+					strcat(buf, name_catalan_river1[SeedModChance(4, lengthof(name_catalan_river1), seed)]);
+				}
+			}
+		}
+	}
+
+	return 0;
+}
+
+
+
 TownNameGenerator * const _town_name_generators[] =
 {
 	MakeEnglishOriginalTownName,
@@ -714,6 +751,7 @@
 	MakeDanishTownName,
 	MakeTurkishTownName,
 	MakeItalianTownName,
+	MakeCatalanTownName,
 };
 
 // DO WE NEED THIS ANY MORE?
Index: openttd.h
===================================================================
--- openttd.h	(revision 5625)
+++ openttd.h	(working copy)
@@ -473,7 +473,8 @@
 	SPECSTR_TOWNNAME_DANISH,
 	SPECSTR_TOWNNAME_TURKISH,
 	SPECSTR_TOWNNAME_ITALIAN,
-	SPECSTR_TOWNNAME_LAST = SPECSTR_TOWNNAME_ITALIAN,
+	SPECSTR_TOWNNAME_CATALAN,
+	SPECSTR_TOWNNAME_LAST = SPECSTR_TOWNNAME_CATALAN,
 
 	// special strings for player names on the form "TownName transport".
 	SPECSTR_PLAYERNAME_START = 0x70EA,
Index: settings.c
===================================================================
--- settings.c	(revision 5625)
+++ settings.c	(working copy)
@@ -1213,7 +1213,7 @@
 	    SDT_VAR(GameOptions, diff_level,SLE_UINT8, 0, 0, 9,0, 9, STR_NULL, NULL),
 	  SDT_OMANY(GameOptions, currency,  SLE_UINT8, N, 0, 0, CUSTOM_CURRENCY_ID, "GBP|USD|EUR|YEN|ATS|BEF|CHF|CZK|DEM|DKK|ESP|FIM|FRF|GRD|HUF|ISK|ITL|NLG|NOK|PLN|ROL|RUR|SEK|custom", STR_NULL, NULL),
 	  SDT_OMANY(GameOptions, units,     SLE_UINT8, N, 0, 1,   2, "imperial|metric|si", STR_NULL, NULL),
-	  SDT_OMANY(GameOptions, town_name, SLE_UINT8, 0, 0, 0,  19, "english|french|german|american|latin|silly|swedish|dutch|finnish|polish|slovakish|norwegian|hungarian|austrian|romanian|czech|swiss|danish|turkish|italian", STR_NULL, NULL),
+	  SDT_OMANY(GameOptions, town_name, SLE_UINT8, 0, 0, 0,  20, "english|french|german|american|latin|silly|swedish|dutch|finnish|polish|slovakish|norwegian|hungarian|austrian|romanian|czech|swiss|danish|turkish|italian|catalan", STR_NULL, NULL),
 	  SDT_OMANY(GameOptions, landscape, SLE_UINT8, 0, 0, 0,   3, "normal|hilly|desert|candy", STR_NULL, NULL),
 	    SDT_VAR(GameOptions, snow_line, SLE_UINT8, 0, 0, 1,0,56, STR_NULL, NULL),
 	SDT_CONDOMANY(GameOptions,autosave, SLE_UINT8, 0, 22,             N, 0, 0, 0, "", STR_NULL, NULL),
Index: table/namegen.h
===================================================================
--- table/namegen.h	(revision 5625)
+++ table/namegen.h	(working copy)
@@ -3131,3 +3131,172 @@
 	" Ticino",
 	" Tevere",
 };
+
+static const char *name_catalan_real[] = {
+	"Barcelona",
+	"Hospitalet",
+	"Cerdanyola",
+	"Martorell",
+	"Badalona",
+	"Tarragona",
+	"Lleida",
+	"Girona",
+	"Sabadell",
+	"Terrassa",
+	"Reus",
+	"Valls",
+	"Vic",
+	"Vielha",
+	"Amposta",
+	"Tortosa",
+	"Berga",
+	"Olot",
+	"Mollerussa",
+	"Banyoles",
+	"Figueres",
+	"Balaguer",
+	"Vilafranca del Penedès",
+	"La seu d'Urgell",
+	"Pont de Suert",
+	"Igualada",
+	"Manresa",
+	"Solsona",
+	"Les borges blanques",
+	"Tremp",
+	"Sort",
+	"Colera",
+	"Portbou",
+	"El Vendrell",
+	"Falset",
+	"Ripoll",
+	"Cervera",
+	"Gandesa",
+	"Mataró",
+	"Montblanc",
+	"Vilanova i la Geltrú",
+	"Tàrrega",
+	"Camprodon",
+	"Campdevànol",
+	"Cambrils",
+	"Begur",
+	"Set Cases",
+	"Palafrugell",
+	"Begues",
+	"El Bruc",
+	"Cadaqués",
+	"Collbató",
+	"Cervelló",
+	"Esparraguera",
+	"Abrera",
+	"Alp",
+	"Das",
+	"Cercs",
+	"Manlleu",
+	"Masnou",
+	"Molins de rei",
+	"Monistrol",
+	"Rocallaura",
+	"Rubí",
+	"Ripollet",
+	"Sitges",
+	"Roses",
+};
+
+static const char *name_catalan_pref[] = {
+	"Pont de ",
+	"Parets de ",
+	"Canet de ",
+	"Castellar de ",
+	"Corbera de ",
+	"Arenys de ",
+	"Calella de ",
+	"La seu de ",
+	"La bisbal de ",
+	"Torroella de ",
+	"Port de ",
+	"Vilafranca de ",
+	"Vilanova de ",
+	"Caldes de ",
+	"La Conca de ",
+	"Olesa de",
+	"La Roca de ",
+	"Sant Esteve de ",
+	"Sant Andreu de ",
+	"Sant Jordi de ",
+	"Sant Joan de ",
+	"Sant Feliu de ",
+	"Sant Quirze de ",
+	"Sant Sadurní de ",
+	"Santa Coloma de ",
+	"Santa Margarida de ",
+	"Santa Maria de ",
+	"Sant Martí de ",
+	"Sant Pere de ",
+	"Sant Julià de ",
+	"Sant Vicenç de ",
+
+};
+
+static const char *name_catalan_1m[] = {
+	"Torrent",
+	"Cami",
+	"Mont",
+	"Bell",
+	"Puig",
+	"Riu",
+};
+
+static const char *name_catalan_1f[] = {
+	"Aigua",
+	"Selva ",
+	"Vall",
+	"Serra",
+	"Torre",
+	"Riba",
+	"Vall",
+	"Terra",
+};
+
+static const char *name_catalan_2m[] = {
+	"alt",
+	"baix",
+	"fosc",
+	"pelat",
+	"ventós",
+	"negre",
+	"roig",
+	"gris",
+};
+
+static const char *name_catalan_2f[] = {
+	"baixa",
+	"alta",
+	"fosca",
+	"clara",
+	"negra",
+	"roja",
+	"grisa",
+	"freda",
+};
+
+static const char *name_catalan_3[] = {
+	" despí",
+	" desvern",
+	" del camí",
+	" de Mar",
+	" de Dalt",
+	" de Baix",
+	" del Vallès",
+	" de Berguedà",
+	" de Conflent",
+	" de la plana",
+};
+
+static const char *name_catalan_river1[] = {
+	" d'Anoia",
+	" de Ter",
+	" de Llobregat",
+	" d'Ebre",
+	" de Segre",
+	" de Francolí",
+};
