Index: misc_gui.cpp
===================================================================
--- misc_gui.cpp	(revision 14813)
+++ misc_gui.cpp	(working copy)
@@ -1046,7 +1046,7 @@
 	_cur_dpi = old_dpi;
 }
 
-int QueryStringBaseWindow::HandleEditBoxKey(int wid, uint16 key, uint16 keycode, EventState &state)
+HandleEditBoxResult QueryStringBaseWindow::HandleEditBoxKey(int wid, uint16 key, uint16 keycode, EventState &state)
 {
 	return this->QueryString::HandleEditBoxKey(this, wid, key, keycode, state);
 }
Index: querystring_gui.h
===================================================================
--- querystring_gui.h	(revision 14813)
+++ querystring_gui.h	(working copy)
@@ -66,7 +66,7 @@
 
 	void DrawEditBox(int wid);
 	void HandleEditBox(int wid);
-	int HandleEditBoxKey(int wid, uint16 key, uint16 keycode, EventState &state);
+	HandleEditBoxResult HandleEditBoxKey(int wid, uint16 key, uint16 keycode, EventState &state);
 	virtual void OnOpenOSKWindow(int wid);
 };
 
