Index: video/win32_v.c
===================================================================
--- video/win32_v.c	(revision 7057)
+++ video/win32_v.c	(working copy)
@@ -255,7 +255,6 @@
 
 		case WM_CLOSE:
 			HandleExitGameRequest();
-			_window_maximize = IsZoomed(_wnd.main_wnd);
 			return 0;
 
 		case WM_LBUTTONDOWN:
@@ -390,6 +389,7 @@
 
 		case WM_SIZE:
 			if (wParam != SIZE_MINIMIZED) {
+				_window_maximize = (wParam == SIZE_MAXIMIZED);
 				ClientSizeChanged(LOWORD(lParam), HIWORD(lParam));
 			}
 			return 0;
@@ -557,6 +557,8 @@
 			SetRect(&r, 0, 0, _wnd.width_org, _wnd.height_org);
 		} else {
 			style = WS_OVERLAPPEDWINDOW | WS_VISIBLE;
+			/* On startup let's see if we quit maximized the last time, restore that */
+			if (_window_maximize) style |= WS_MAXIMIZE;
 			SetRect(&r, 0, 0, _wnd.width, _wnd.height);
 		}
 
@@ -577,12 +579,6 @@
 
 			_wnd.main_wnd = CreateWindow("OTTD", Windowtitle, style, x, y, w, h, 0, 0, GetModuleHandle(NULL), 0);
 			if (_wnd.main_wnd == NULL) error("CreateWindow failed");
-
-			/* On startup let's see if we quit maximized the last time, restore that */
-			if (_window_maximize) {
-				ShowWindow(_wnd.main_wnd, SW_MAXIMIZE);
-				_window_maximize = false;
-			}
 		}
 	}
 	GameSizeChanged(); // invalidate all windows, force redraw
