diff -r d2094721ed07 src/rail_cmd.cpp
--- a/src/rail_cmd.cpp	Tue Jul 23 08:18:22 2013 +0200
+++ b/src/rail_cmd.cpp	Tue Jul 23 08:27:49 2013 +0200
@@ -872,8 +872,6 @@
 	CommandCost ret = ValidateAutoDrag(&trackdir, tile, end_tile);
 	if (ret.Failed()) return ret;
 
-	if ((flags & DC_EXEC) && _settings_client.sound.confirm) SndPlayTileFx(SND_20_SPLAT_2, tile);
-
 	bool had_success = false;
 	CommandCost last_error = CMD_ERROR;
 	for (;;) {
diff -r d2094721ed07 src/rail_gui.cpp
--- a/src/rail_gui.cpp	Tue Jul 23 08:18:22 2013 +0200
+++ b/src/rail_gui.cpp	Tue Jul 23 08:27:49 2013 +0200
@@ -352,7 +352,8 @@
 	DoCommandP(TileVirtXY(_thd.selstart.x, _thd.selstart.y), TileVirtXY(_thd.selend.x, _thd.selend.y), _cur_railtype | (mode << 4),
 			_remove_button_clicked ?
 			CMD_REMOVE_RAILROAD_TRACK | CMD_MSG(STR_ERROR_CAN_T_REMOVE_RAILROAD_TRACK) :
-			CMD_BUILD_RAILROAD_TRACK  | CMD_MSG(STR_ERROR_CAN_T_BUILD_RAILROAD_TRACK));
+			CMD_BUILD_RAILROAD_TRACK  | CMD_MSG(STR_ERROR_CAN_T_BUILD_RAILROAD_TRACK),
+			CcPlaySound1E);
 }
 
 static void HandleAutodirPlacement()
