diff -r 69c327f7b2db -r 5dd72d594bf6 src/settings_gui.cpp
--- a/src/settings_gui.cpp	Sun Sep 28 14:09:59 2008 +0200
+++ b/src/settings_gui.cpp	Sun Sep 28 14:09:59 2008 +0200
@@ -586,17 +586,24 @@
 static const char *_patches_ui[] = {
 	"gui.vehicle_speed",
 	"gui.status_long_date",
+	"gui.population_in_label",
+	"gui.liveries",
+	"gui.loading_indicators",
+	"",
 	"gui.autoscroll",
 	"gui.reverse_scroll",
 	"gui.smooth_scroll",
 	"gui.errmsg_duration",
 	"gui.toolbar_pos",
+	"gui.window_snap_radius",
+	"",
 	"gui.measure_tooltip",
-	"gui.window_snap_radius",
-	"gui.population_in_label",
 	"gui.link_terraform_toolbar",
-	"gui.liveries",
+	"gui.advanced_vehicle_list",
+	"gui.always_build_infrastructure",
+	"gui.default_rail_type",
 	"gui.prefer_teamchat",
+	"",
 	/* While the horizontal scrollwheel scrolling is written as general code, only
 	 *  the cocoa (OSX) driver generates input for it.
 	 *  Since it's also able to completely disable the scrollwheel will we display it on all platforms anyway */
@@ -606,12 +613,8 @@
 	/* We might need to emulate a right mouse button on mac */
 	"gui.right_mouse_btn_emulation",
 #endif
+	"",
 	"gui.pause_on_newgame",
-	"gui.advanced_vehicle_list",
-	"gui.loading_indicators",
-	"gui.timetable_in_ticks",
-	"gui.default_rail_type",
-	"gui.always_build_infrastructure",
 	"gui.show_track_reservation",
 	NULL
 };
@@ -684,6 +687,7 @@
 };
 
 static const char *_patches_vehicles[] = {
+	"gui.timetable_in_ticks",
 	"vehicle.realistic_acceleration",
 	"pf.forbid_90_deg",
 	"vehicle.mammoth_trains",
