Index: src/tree_gui.cpp
===================================================================
--- src/tree_gui.cpp	(revision 22867)
+++ src/tree_gui.cpp	(working copy)
@@ -162,7 +162,7 @@
 	virtual void OnPlaceObject(Point pt, TileIndex tile)
 	{
 		VpStartPlaceSizing(tile, VPM_X_AND_Y_LIMITED, DDSP_PLANT_TREES);
-		VpSetPlaceSizingLimit(20);
+		VpSetPlaceSizingLimit(128);
 	}
 
 	virtual void OnPlaceDrag(ViewportPlaceMethod select_method, ViewportDragDropSelectionProcess select_proc, Point pt)
