diff -r 4c9d49e5589a src/engine.cpp
--- a/src/engine.cpp	Mon Dec 08 16:29:36 2008 +0000
+++ b/src/engine.cpp	Tue Dec 09 13:55:19 2008 +0100
@@ -205,6 +205,7 @@ static void CalcEngineReliability(Engine
 		}
 	}
 
+	/* Reliability does not really play a role for wagons, but expiration does */
 	if (age < e->duration_phase_1) {
 		uint start = e->reliability_start;
 		e->reliability = age * (e->reliability_max - start) / e->duration_phase_1 + start;
@@ -264,11 +265,7 @@ void StartupEngines()
 
 		e->reliability_spd_dec = ei->decay_speed << 2;
 
-		if (IsWagon(e->index)) {
-			e->age = 0xFFFF;
-		} else {
-			CalcEngineReliability(e);
-		}
+		CalcEngineReliability(e);
 
 		e->lifelength = ei->lifelength + _settings_game.vehicle.extend_vehicle_life;
 
diff -r 4c9d49e5589a src/oldloader.cpp
--- a/src/oldloader.cpp	Mon Dec 08 16:29:36 2008 +0000
+++ b/src/oldloader.cpp	Tue Dec 09 13:55:19 2008 +0100
@@ -1348,11 +1348,6 @@ static bool LoadOldEngine(LoadgameState 
 {
 	Engine *e = GetTempDataEngine(num);
 	if (!LoadChunk(ls, e, engine_chunk)) return false;
-
-	/* Make sure wagons are marked as do-not-age */
-	if ((num >= 27 && num < 54) || (num >= 57 && num < 84) || (num >= 89 && num < 116))
-		e->age = 0xFFFF;
-
 	return true;
 }
 
