diff -r 66461973091a src/order_cmd.cpp
--- a/src/order_cmd.cpp	Wed Oct 29 16:18:47 2008 +0100
+++ b/src/order_cmd.cpp	Wed Oct 29 16:18:52 2008 +0100
@@ -1202,8 +1202,7 @@ CommandCost CmdCloneOrder(TileIndex tile
 				/* If the destination vehicle had a OrderList, destroy it */
 				DeleteVehicleOrders(dst);
 
-				dst->orders = src->orders;
-				dst->num_orders = src->GetNumOrders();
+				dst->orders.list = src->orders.list;
 
 				/* Link this vehicle in the shared-list */
 				dst->AddToShared(src);
@@ -1249,24 +1248,29 @@ CommandCost CmdCloneOrder(TileIndex tile
 
 			/* make sure there are orders available */
 			delta = dst->IsOrderListShared() ? src->GetNumOrders() + 1 : src->GetNumOrders() - dst->GetNumOrders();
-			if (!HasOrderPoolFree(delta))
+			if (!HasOrderPoolFree(delta) ||
+			    ((dst->orders.list == NULL || dst->IsOrderListShared()) && !OrderList::CanAllocateItem()))
 				return_cmd_error(STR_8831_NO_MORE_SPACE_FOR_ORDERS);
 
 			if (flags & DC_EXEC) {
 				const Order *order;
+				Order *first = NULL;
 				Order **order_dst;
 
-				/* If the destination vehicle had a OrderList, destroy it */
-				DeleteVehicleOrders(dst);
-
-				order_dst = &dst->orders;
+				/* If the destination vehicle had an order list, destroy the chain but keep the OrderList */
+				DeleteVehicleOrders(dst, true);
+
+				order_dst = &first;
 				FOR_VEHICLE_ORDERS(src, order) {
 					*order_dst = new Order();
 					(*order_dst)->AssignOrder(*order);
 					order_dst = &(*order_dst)->next;
 				}
-
-				dst->num_orders = src->GetNumOrders();
+				if (dst->orders.list == NULL) dst->orders.list = new OrderList(first, dst);
+				else {
+					assert(dst->orders.list->GetFirstOrder() == NULL);
+					new (dst->orders.list) OrderList(first, dst);
+				}
 
 				InvalidateVehicleOrder(dst, -1);
 
