Index: src/viewport.cpp
===================================================================
--- src/viewport.cpp	(revision 13481)
+++ src/viewport.cpp	(working copy)
@@ -1870,8 +1870,7 @@
 						y <  si->sign.top + 12 &&
 						x >= si->sign.left &&
 						x <  si->sign.left + si->sign.width_1) {
-					ShowRenameSignWindow(si);
-					return true;
+					return ShowRenameSignWindow(si);
 				}
 			}
 			break;
@@ -1884,8 +1883,7 @@
 						y <  si->sign.top + 24 &&
 						x >= si->sign.left &&
 						x <  si->sign.left + si->sign.width_1 * 2) {
-					ShowRenameSignWindow(si);
-					return true;
+					return ShowRenameSignWindow(si);
 				}
 			}
 			break;
@@ -1900,8 +1898,7 @@
 						y <  si->sign.top + ScaleByZoom(12, vp->zoom) &&
 						x >= si->sign.left &&
 						x <  si->sign.left + ScaleByZoom(si->sign.width_2, vp->zoom)) {
-					ShowRenameSignWindow(si);
-					return true;
+					return ShowRenameSignWindow(si);
 				}
 			}
 			break;
Index: src/signs_gui.cpp
===================================================================
--- src/signs_gui.cpp	(revision 13481)
+++ src/signs_gui.cpp	(working copy)
@@ -8,6 +8,7 @@
 #include "textbuf_gui.h"
 #include "window_gui.h"
 #include "player_gui.h"
+#include "player_func.h"
 #include "signs_base.h"
 #include "signs_func.h"
 #include "debug.h"
@@ -301,11 +302,17 @@
 	_query_sign_edit_widgets,
 };
 
-void ShowRenameSignWindow(const Sign *si)
+bool ShowRenameSignWindow(const Sign *si)
 {
+	if (_ctrl_pressed && si->owner == _current_player) {
+		RenameSign(si->index, "");
+		return false;
+	}
+
 	/* Delete all other edit windows and the save window */
 	DeleteWindowById(WC_QUERY_STRING, 0);
 	DeleteWindowById(WC_SAVELOAD, 0);
 
 	new SignWindow(&_query_sign_edit_desc, si);
+	return true;
 }
Index: src/signs_func.h
===================================================================
--- src/signs_func.h	(revision 13481)
+++ src/signs_func.h	(working copy)
@@ -14,7 +14,7 @@
 void PlaceProc_Sign(TileIndex tile);
 
 /* signs_gui.cpp */
-void ShowRenameSignWindow(const Sign *si);
+bool ShowRenameSignWindow(const Sign *si);
 
 void ShowSignList();
 
