Index: src/driver.h
===================================================================
--- src/driver.h	(r‚vision 13071)
+++ src/driver.h	(copie de travail)
@@ -71,7 +71,12 @@
 	 */
 	virtual ~DriverFactoryBase() {
 		if (this->name == NULL) return;
-		GetDrivers().erase(this->name);
+		for (Drivers::iterator it = GetDrivers().begin(); it != GetDrivers().end(); it++) {
+			if (it->first == Drivers::key_type(this->name)) {
+				GetDrivers().erase(it);
+				break;
+			}
+		}
 		free(this->name);
 	}
 
