Index: src/video/dedicated_v.cpp
===================================================================
--- src/video/dedicated_v.cpp	(revision 15537)
+++ src/video/dedicated_v.cpp	(working copy)
@@ -141,8 +141,6 @@
 	_screen.height = _cur_resolution.height;
 	ScreenSizeChanged();
 
-	SetDebugString("net=6");
-
 #if defined(WINCE)
 	/* WinCE doesn't support console stuff */
 #elif defined(WIN32)
