consistently show selector window

diff -r d30d43b19240 src/station_gui.cpp
--- a/src/station_gui.cpp	Tue Jan 13 13:46:00 2009 +0100
+++ b/src/station_gui.cpp	Tue Jan 13 13:46:08 2009 +0100
@@ -1226,16 +1226,10 @@ static bool StationJoinerNeeded(CommandC
 	/* Now check if we could build there */
 	if (CmdFailed(DoCommand(&cmd, CommandFlagsToDCFlags(GetCommandFlags(cmd.cmd))))) return false;
 
-	/* First test for adjacent station */
-	FindStationsNearby(cmd.tile, w, h, false);
-	int neighbour_station_count = _stations_nearby_list.Length();
-	/* Now test for stations fully within station spread */
-	const Station *st = FindStationsNearby(cmd.tile, w, h, true);
-	if (_settings_game.station.adjacent_stations) {
-		return (neighbour_station_count == 0 || _stations_nearby_list.Length() > 1) && st == NULL;
-	} else {
-		return neighbour_station_count == 0 && _stations_nearby_list.Length() > 0 && st == NULL;
-	}
+	/* Test for adjacent station or station below selection */
+	const Station *st = FindStationsNearby(cmd.tile, w, h, false);
+
+	return st == NULL && (_settings_game.station.adjacent_stations || _stations_nearby_list.Length() == 0);
 }
 
 /**
