diff --git a/src/openttd.cpp b/src/openttd.cpp
index 1dd69b9..62b0bad 100644
--- a/src/openttd.cpp
+++ b/src/openttd.cpp
@@ -1170,6 +1170,7 @@ void StateGameLoop()
 			}
 		}
 
+#ifdef _DEBUG
 		/* Check whether the caches are still valid */
 		Vehicle *v;
 		FOR_ALL_VEHICLES(v) {
@@ -1188,6 +1189,7 @@ void StateGameLoop()
 				assert(memcmp(&st->goods[c].cargo, buff, sizeof(StationCargoList)) == 0);
 			}
 		}
+#endif
 
 		/* All these actions has to be done from OWNER_NONE
 		 *  for multiplayer compatibility */
