Index: src/main_gui.cpp
===================================================================
--- src/main_gui.cpp	(revision 11731)
+++ src/main_gui.cpp	(arbetskopia)
@@ -2231,7 +2231,7 @@
 			case 'Z': {
 				Point pt = GetTileBelowCursor();
 				if (pt.x != -1) {
-					ScrollMainWindowTo(pt.x, pt.y);
+					ScrollMainWindowTo(pt.x, pt.y, e->we.keypress.keycode == 'Z');
 					if (e->we.keypress.keycode == 'Z') MaxZoomInOut(ZOOM_IN, w);
 				}
 				break;
