Index: src/town_cmd.cpp
===================================================================
--- src/town_cmd.cpp	(Revision 11178)
+++ src/town_cmd.cpp	(Arbeitskopie)
@@ -894,10 +894,10 @@
  *
  * @param t The current town
  * @param tile The current tile
- * @param rcmd The RoadBits which are possible on this tile
+ * @param target_dir The current town growth dir
  * @return true if a bridge has been build else false
  */
-static bool GrowTownWithBridge(const Town *t, TileIndex tile, RoadBits rcmd)
+static bool GrowTownWithBridge(const Town *t, TileIndex tile, DiagDirection target_dir)
 {
 	DiagDirection bridge_dir; // The direction of a bridge we maybe want to build
 
@@ -912,8 +912,8 @@
 		default: return false;
 	}
 
-	/* Check if the bridge will be compatible to the RoadBits */
-	if (!(rcmd & DiagDirToRoadBits(ReverseDiagDir(bridge_dir)))) return false;
+	/* Check if the bridge will be compatible to the town growth dir */
+	if (target_dir != bridge_dir) return false;
 
 	/* We are in the right direction */
 	uint32 bridge_length = 0;     // This value stores the length of the possible bridge
@@ -1119,7 +1119,7 @@
 	rcmd = CleanUpRoadBits(tile, rcmd);
 	if (rcmd == ROAD_NONE) return;
 
-	if (GrowTownWithBridge(t1, tile, rcmd)) return;
+	if (GrowTownWithBridge(t1, tile, target_dir)) return;
 
 	GrowTownWithRoad(t1, tile, rcmd);
 }
