Index: src/road_cmd.cpp
===================================================================
--- src/road_cmd.cpp	(revision 10499)
+++ src/road_cmd.cpp	(working copy)
@@ -471,9 +471,9 @@
 
 		default:
 do_clear:;
-			ret = DoCommand(tile, 0, 0, flags, CMD_LANDSCAPE_CLEAR);
+/*			ret = DoCommand(tile, 0, 0, flags, CMD_LANDSCAPE_CLEAR);
 			if (CmdFailed(ret)) return ret;
-			cost.AddCost(ret);
+			cost.AddCost(ret);*/
 	}
 
 	if (all_bits != pieces) {
@@ -481,10 +481,10 @@
 		ret = CheckRoadSlope(tileh, &pieces, all_bits | existing);
 		/* Return an error if we need to build a foundation (ret != 0) but the
 		 * current patch-setting is turned off (or stupid AI@work) */
-		if (CmdFailed(ret) || (ret.GetCost() != 0 && (!_patches.build_on_slopes || _is_old_ai_player))) {
+/*		if (CmdFailed(ret) || (ret.GetCost() != 0 && (!_patches.build_on_slopes || _is_old_ai_player))) {
 			return_cmd_error(STR_1000_LAND_SLOPED_IN_WRONG_DIRECTION);
 		}
-		cost.AddCost(ret);
+		cost.AddCost(ret);*/
 	}
 
 	if (IsTileType(tile, MP_STREET)) {
