Index: src/network/network_gui.cpp
===================================================================
--- src/network/network_gui.cpp	(revision 10496)
+++ src/network/network_gui.cpp	(working copy)
@@ -745,7 +745,6 @@
 			if (HandleEditBoxKey(w, &WP(w, network_ql_d).q, 3, e) == 1) break; // enter pressed
 
 			ttd_strlcpy(_network_server_name, WP(w, network_ql_d).q.text.buf, sizeof(_network_server_name));
-			UpdateTextBufferSize(&WP(w, network_ql_d).q.text);
 		}
 		break;
 
