Index: src/video/sdl_v.cpp
===================================================================
--- src/video/sdl_v.cpp	(revision 17559)
+++ src/video/sdl_v.cpp	(working copy)
@@ -302,6 +302,8 @@
 		}
 	}
 
+	printf("scancode: %02x [%i] -> %02x [%i]\n", sym->scancode, sym->scancode, sym->unicode, sym->unicode);
+
 	/* check scancode for BACKQUOTE key, because we want the key left of "1", not anything else (on non-US keyboards) */
 #if defined(WIN32) || defined(__OS2__)
 	if (sym->scancode == 41) key = WKC_BACKQUOTE;
Index: src/video/allegro_v.cpp
===================================================================
--- src/video/allegro_v.cpp	(revision 17559)
+++ src/video/allegro_v.cpp	(working copy)
@@ -297,6 +297,8 @@
 	const VkMapping *map;
 	uint key = 0;
 
+	printf("scancode: %02x [%i] -> %02x [%i]\n", scancode, scancode, unicode, unicode);
+
 	for (map = _vk_mapping; map != endof(_vk_mapping); ++map) {
 		if ((uint)(scancode - map->vk_from) <= map->vk_count) {
 			key = scancode - map->vk_from + map->map_to;
