diff -r 40804f43c01c -r ab2ce62f1eb6 src/settings_gui.cpp
--- a/src/settings_gui.cpp	Fri Nov 07 11:01:12 2008 +0100
+++ b/src/settings_gui.cpp	Fri Nov 07 11:23:15 2008 +0100
@@ -742,6 +742,7 @@
 /** Widget numbers of config patches window */
 enum PatchesSelectionWidgets {
 	PATCHSEL_OPTIONSPANEL = 3, ///< Panel widget containing the option lists
+	PATCHSEL_SCROLLBAR,        ///< Scrollbar
 	PATCHSEL_INTERFACE,        ///< Button 'Interface'
 	PATCHSEL_CONSTRUCTION,     ///< Button 'Construction'
 	PATCHSEL_VEHICLES,         ///< Button 'Vehicles'
@@ -755,7 +756,6 @@
 	static const int SETTINGTREE_TOP_OFFSET;  ///< Position of top edge of patch values
 
 	static GameSettings *patches_ptr;  ///< Pointer to the game settings being displayed and modified
-	static int patches_max;  ///< Maximal number of patches on a single page
 
 	int page;
 	int entry;
@@ -779,9 +779,6 @@
 			PatchPage *page;
 			for (page = &_patches_page[0]; page != endof(_patches_page); page++) {
 				uint i;
-
-				if (patches_max < page->num) patches_max = page->num;
-
 				page->entries = MallocT<PatchEntry>(page->num);
 				for (i = 0; i != page->num; i++) {
 					uint index;
@@ -795,13 +792,13 @@
 			first_time = false;
 		}
 
-		/* Resize the window to fit the largest patch tab */
-		ResizeWindowForWidget(this, PATCHSEL_OPTIONSPANEL, 0, patches_max * SETTING_HEIGHT);
+		page = 0;
+		this->vscroll.pos = 0;
+		this->vscroll.cap = (this->widget[PATCHSEL_OPTIONSPANEL].bottom - this->widget[PATCHSEL_OPTIONSPANEL].top - 8) / SETTING_HEIGHT;
+		SetVScrollCount(this, _patches_page[page].num);
 
-		/* Recentre the window for the new size */
-		this->top = this->top - (patches_max * SETTING_HEIGHT) / 2;
 
-		this->LowerWidget(PATCHSEL_INTERFACE);
+		this->LowerWidget(page + PATCHSEL_INTERFACE); // Depress button of currently selected page
 
 		this->FindWindowPlacementAndResize(desc);
 	}
@@ -816,7 +813,7 @@
 
 		int x = SETTINGTREE_LEFT_OFFSET;
 		int y = SETTINGTREE_TOP_OFFSET;
-		for (i = 0; i != page->num; i++) {
+		for (i = vscroll.pos; i != page->num && i < vscroll.pos + vscroll.cap; i++) {
 			const SettingDesc *sd = page->entries[i].setting;
 			DrawPatch(patches_ptr, sd, x, y, this->click - (i * 2));
 			y += SETTING_HEIGHT;
@@ -877,7 +874,7 @@
 				int x = pt.x - SETTINGTREE_LEFT_OFFSET;  // Shift x coordinate
 				if (x < 0) return;  // Clicked left of the entry
 
-				byte btn = y / SETTING_HEIGHT;  // Compute which setting is selected
+				byte btn = this->vscroll.pos + y / SETTING_HEIGHT;  // Compute which setting is selected
 				if (y % SETTING_HEIGHT > SETTING_HEIGHT - 2) return;  // Clicked too low at the setting
 
 				const PatchPage *page = &_patches_page[this->page];
@@ -956,6 +953,7 @@
 				this->RaiseWidget(this->page + PATCHSEL_INTERFACE);
 				this->page = widget - PATCHSEL_INTERFACE;
 				this->LowerWidget(this->page + PATCHSEL_INTERFACE);
+				SetVScrollCount(this, _patches_page[page].num);
 				DeleteWindowById(WC_QUERY_STRING, 0);
 				this->SetDirty();
 				break;
@@ -987,25 +985,25 @@
 GameSettings *PatchesSelectionWindow::patches_ptr = NULL;
 const int PatchesSelectionWindow::SETTINGTREE_LEFT_OFFSET = 5;
 const int PatchesSelectionWindow::SETTINGTREE_TOP_OFFSET = 47;
-int PatchesSelectionWindow::patches_max = 0;
 
 static const Widget _patches_selection_widgets[] = {
 {   WWT_CLOSEBOX,   RESIZE_NONE,  COLOUR_MAUVE,     0,    10,     0,    13, STR_00C5,                        STR_018B_CLOSE_WINDOW},
-{    WWT_CAPTION,   RESIZE_NONE,  COLOUR_MAUVE,    11,   369,     0,    13, STR_CONFIG_PATCHES_CAPTION,      STR_018C_WINDOW_TITLE_DRAG_THIS},
-{      WWT_PANEL,   RESIZE_NONE,  COLOUR_MAUVE,     0,   369,    14,    41, 0x0,                             STR_NULL},
-{      WWT_PANEL,   RESIZE_NONE,  COLOUR_MAUVE,     0,   369,    42,    50, 0x0,                             STR_NULL}, // PATCHSEL_OPTIONSPANEL
+{    WWT_CAPTION,   RESIZE_NONE,  COLOUR_MAUVE,    11,   381,     0,    13, STR_CONFIG_PATCHES_CAPTION,      STR_018C_WINDOW_TITLE_DRAG_THIS},
+{      WWT_PANEL,   RESIZE_NONE,  COLOUR_MAUVE,     0,   381,    14,    41, 0x0,                             STR_NULL},
+{      WWT_PANEL,   RESIZE_NONE,  COLOUR_MAUVE,     0,   369,    42,   215, 0x0,                             STR_NULL}, // PATCHSEL_OPTIONSPANEL
+{  WWT_SCROLLBAR,   RESIZE_NONE,  COLOUR_MAUVE,   370,   381,    42,   215, 0x0,                             STR_0190_SCROLL_BAR_SCROLLS_LIST}, // PATCHSEL_SCROLLBAR
 
-{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,   10,    96,    16,    27, STR_CONFIG_PATCHES_GUI,          STR_NULL}, // PATCHSEL_INTERFACE
-{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,   97,   183,    16,    27, STR_CONFIG_PATCHES_CONSTRUCTION, STR_NULL}, // PATCHSEL_CONSTRUCTION
-{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,  184,   270,    16,    27, STR_CONFIG_PATCHES_VEHICLES,     STR_NULL}, // PATCHSEL_VEHICLES
-{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,  271,   357,    16,    27, STR_CONFIG_PATCHES_STATIONS,     STR_NULL}, // PATCHSEL_STATIONS
-{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,   10,    96,    28,    39, STR_CONFIG_PATCHES_ECONOMY,      STR_NULL}, // PATCHSEL_ECONOMY
-{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,   97,   183,    28,    39, STR_CONFIG_PATCHES_AI,           STR_NULL}, // PATCHSEL_COMPETITORS
+{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,   10,   100,    16,    27, STR_CONFIG_PATCHES_GUI,          STR_NULL}, // PATCHSEL_INTERFACE
+{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,  101,   191,    16,    27, STR_CONFIG_PATCHES_CONSTRUCTION, STR_NULL}, // PATCHSEL_CONSTRUCTION
+{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,  192,   283,    16,    27, STR_CONFIG_PATCHES_VEHICLES,     STR_NULL}, // PATCHSEL_VEHICLES
+{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,  284,   375,    16,    27, STR_CONFIG_PATCHES_STATIONS,     STR_NULL}, // PATCHSEL_STATIONS
+{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,   10,   100,    28,    39, STR_CONFIG_PATCHES_ECONOMY,      STR_NULL}, // PATCHSEL_ECONOMY
+{    WWT_TEXTBTN,   RESIZE_NONE,  COLOUR_YELLOW,  101,   191,    28,    39, STR_CONFIG_PATCHES_AI,           STR_NULL}, // PATCHSEL_COMPETITORS
 {   WIDGETS_END},
 };
 
 static const WindowDesc _patches_selection_desc = {
-	WDP_CENTER, WDP_CENTER, 370, 51, 370, 51,
+	WDP_CENTER, WDP_CENTER, 382, 216, 382, 216,
 	WC_GAME_OPTIONS, WC_NONE,
 	WDF_STD_TOOLTIPS | WDF_STD_BTN | WDF_DEF_WIDGET,
 	_patches_selection_widgets,
