Index: src/lang/english.txt
===================================================================
--- src/lang/english.txt	(revision 8418)
+++ src/lang/english.txt	(working copy)
@@ -1039,6 +1039,8 @@
 STR_CONFIG_PATCHES_NONUNIFORM_STATIONS                          :{LTBLUE}Nonuniform stations: {ORANGE}{STRING1}
 STR_CONFIG_PATCHES_NEW_PATHFINDING_ALL                          :{LTBLUE}New global pathfinding (NPF, overrides NTP): {ORANGE}{STRING1}
 STR_CONFIG_PATCHES_FREIGHT_TRAINS                               :{LTBLUE}Weight multiplier for freight to simulate heavy trains: {ORANGE}{STRING}
+STR_CONFIG_PATCHES_FARCFT_BARPT									:{LTBLUE}Fast aircrafts can only use big airports
+STR_CONFIG_PATCHES_FARCFT_BARPT2								:{WHITE}Fast aircrafts can only use big airports...
 
 STR_CONFIG_PATCHES_SMALL_AIRPORTS                               :{LTBLUE}Always allow small airports: {ORANGE}{STRING1}
 
Index: src/lang/french.txt
===================================================================
--- src/lang/french.txt	(revision 8418)
+++ src/lang/french.txt	(working copy)
@@ -1040,6 +1040,8 @@
 STR_CONFIG_PATCHES_NONUNIFORM_STATIONS                          :{LTBLUE}Gares non uniformes: {ORANGE}{STRING}
 STR_CONFIG_PATCHES_NEW_PATHFINDING_ALL                          :{LTBLUE}Nouvelle recherche de chemin globale (NPF, écrase NTP): {ORANGE}{STRING}
 STR_CONFIG_PATCHES_FREIGHT_TRAINS                               :{LTBLUE}Multiplieur de poids pour le fret (simule les trains lourds): {ORANGE}{STRING}
+STR_CONFIG_PATCHES_FARCFT_BARPT									:{LTBLUE}Les avions rapides ne peuvent utiliser que les grands aéroports
+STR_CONFIG_PATCHES_FARCFT_BARPT2								:{WHITE}Les avions rapides ne peuvent utiliser que les grands aéroports...
 
 STR_CONFIG_PATCHES_SMALL_AIRPORTS                               :{LTBLUE}Toujours permettre les petits aéroports: {ORANGE}{STRING}
 
Index: src/order_gui.cpp
===================================================================
--- src/order_gui.cpp	(revision 8418)
+++ src/order_gui.cpp	(working copy)
@@ -252,10 +252,21 @@
 		case MP_STATION:
 			if (v->type != VEH_Aircraft) break;
 			if (IsHangar(tile) && IsTileOwner(tile, _local_player)) {
-				order.type = OT_GOTO_DEPOT;
-				order.flags = OF_PART_OF_ORDERS;
-				order.dest = GetStationIndex(tile);
-				return order;
+				if (IsTileType(tile, MP_STATION)) {
+					StationID st_index = GetStationIndex(tile);
+					const Station *st = GetStation(st_index);
+					if (_patches.fast_arcft_big_arpt && (st->airport_type == AT_SMALL || st->airport_type == AT_COMMUTER) && (AircraftVehInfo(v->engine_type)->subtype & AIR_FAST))
+					{
+						ShowErrorMessage(INVALID_STRING_ID, STR_CONFIG_PATCHES_FARCFT_BARPT2, 0, 0);
+					}
+					else
+					{
+						order.type = OT_GOTO_STATION;
+						order.flags = 0;
+						order.dest = st_index;
+						return order;
+					}
+				}
 			}
 			break;
 
@@ -299,10 +310,17 @@
 			(facil=FACIL_BUS_STOP, v->type == VEH_Road && v->cargo_type == CT_PASSENGERS) ||
 			(facil=FACIL_TRUCK_STOP, 1);
 			if (st->facilities & facil) {
-				order.type = OT_GOTO_STATION;
-				order.flags = 0;
-				order.dest = st_index;
-				return order;
+				if (_patches.fast_arcft_big_arpt && v->type == VEH_Aircraft && (st->airport_type == AT_SMALL || st->airport_type == AT_COMMUTER) && (AircraftVehInfo(v->engine_type)->subtype & AIR_FAST))
+				{
+					ShowErrorMessage(INVALID_STRING_ID, STR_CONFIG_PATCHES_FARCFT_BARPT2, 0, 0);
+				}
+				else
+				{
+					order.type = OT_GOTO_STATION;
+					order.flags = 0;
+					order.dest = st_index;
+					return order;
+				}
 			}
 		}
 	}
Index: src/settings.cpp
===================================================================
--- src/settings.cpp	(revision 8418)
+++ src/settings.cpp	(working copy)
@@ -1340,6 +1340,7 @@
 	SDT_BOOL(Patches, serviceathelipad,        0, 0,  true,        STR_CONFIG_PATCHES_SERVICEATHELIPAD,   NULL),
 	SDT_BOOL(Patches, modified_catchment,      0, 0,  true,        STR_CONFIG_PATCHES_CATCHMENT,          NULL),
 	SDT_CONDBOOL(Patches, gradual_loading, 40, SL_MAX_VERSION, 0, 0,  true, STR_CONFIG_PATCHES_GRADUAL_LOADING,    NULL),
+	SDT_CONDBOOL(Patches, fast_arcft_big_arpt,38,SL_MAX_VERSION,0,0,false,           STR_CONFIG_PATCHES_FARCFT_BARPT,         NULL),
 
 	/***************************************************************************/
 	/* Economy section of the GUI-configure patches window */
Index: src/settings_gui.cpp
===================================================================
--- src/settings_gui.cpp	(revision 8418)
+++ src/settings_gui.cpp	(working copy)
@@ -599,6 +599,7 @@
 	"serviceathelipad",
 	"modified_catchment",
 	"gradual_loading",
+	"fast_arcft_big_arpt",
 };
 
 static const char *_patches_economy[] = {
Index: src/variables.h
===================================================================
--- src/variables.h	(revision 8418)
+++ src/variables.h	(working copy)
@@ -218,6 +218,7 @@
 	bool population_in_label; // Show the population of a town in his label?
 
 	uint8 freight_trains; ///< Value to multiply the weight of cargo by
+	bool fast_arcft_big_arpt; // Fast aircrafts can only use big airports
 
 	/** YAPF settings */
 	YapfSettings  yapf;
