Index: src/company_cmd.cpp
===================================================================
--- src/company_cmd.cpp	(révision 23799)
+++ src/company_cmd.cpp	(copie de travail)
@@ -843,6 +843,10 @@
 
 			if (_network_server) {
 				ci->client_playas = c->index;
+
+				/* Announce new company on network. */
+				NetworkAdminCompanyInfo(c, true);
+
 				NetworkUpdateClientInfo(ci->client_id);
 
 				if (Company::IsValidID(ci->client_playas)) {
@@ -863,9 +867,6 @@
 					 * For example in network_client.c:534? */
 					NetworkSendCommand(0, 0, 0, CMD_RENAME_PRESIDENT, NULL, ci->client_name, ci->client_playas);
 				}
-
-				/* Announce new company on network. */
-				NetworkAdminCompanyInfo(c, true);
 				NetworkServerSendChat(NETWORK_ACTION_COMPANY_NEW, DESTTYPE_BROADCAST, 0, "", ci->client_id, ci->client_playas + 1);
 			}
 #endif /* ENABLE_NETWORK */
Index: src/network/network_admin.cpp
===================================================================
--- src/network/network_admin.cpp	(révision 23799)
+++ src/network/network_admin.cpp	(copie de travail)
@@ -785,10 +785,10 @@
 	ServerNetworkAdminSocketHandler *as;
 	FOR_ALL_ACTIVE_ADMIN_SOCKETS(as) {
 		if (as->update_frequency[ADMIN_UPDATE_CLIENT_INFO] & ADMIN_FREQUENCY_AUTOMATIC) {
-			as->SendClientInfo(cs, cs->GetInfo());
 			if (new_client) {
 				as->SendClientJoin(cs->client_id);
 			}
+			as->SendClientInfo(cs, cs->GetInfo());
 		}
 	}
 }
@@ -851,11 +851,11 @@
 	ServerNetworkAdminSocketHandler *as;
 	FOR_ALL_ACTIVE_ADMIN_SOCKETS(as) {
 		if (as->update_frequency[ADMIN_UPDATE_COMPANY_INFO] != ADMIN_FREQUENCY_AUTOMATIC) continue;
-
-		as->SendCompanyInfo(company);
+		
 		if (new_company) {
 			as->SendCompanyNew(company->index);
 		}
+		as->SendCompanyInfo(company);
 	}
 }
 
Index: src/network/network_server.cpp
===================================================================
--- src/network/network_server.cpp	(révision 23799)
+++ src/network/network_server.cpp	(copie de travail)
@@ -1037,8 +1037,6 @@
 			}
 		}
 
-		NetworkAdminClientInfo(this, true);
-
 		/* also update the new client with our max values */
 		this->SendConfigUpdate();
 
@@ -1199,6 +1197,10 @@
 
 		/* Execute script for, e.g. MOTD */
 		IConsoleCmdExec("exec scripts/on_server_connect.scr 0");
+		
+		/* Inform Admin Protocol for new Client */
+		NetworkAdminClientInfo(this, true);
+		
 	}
 
 	/* Get, and validate the token. */
