Index: src/openttd.cpp
===================================================================
--- src/openttd.cpp	(revision 22971)
+++ src/openttd.cpp	(working copy)
@@ -361,7 +361,6 @@
 
 		/* initialize the ingame console */
 		IConsoleInit();
-		_cursor.in_window = true;
 		InitializeGUI();
 		IConsoleCmdExec("exec scripts/autoexec.scr 0");
 
@@ -665,6 +664,8 @@
 	}
 	free(sounddriver);
 
+	_cursor.in_window = true;
+
 	if (videodriver == NULL && _ini_videodriver != NULL) videodriver = strdup(_ini_videodriver);
 	_video_driver = (VideoDriver*)VideoDriverFactoryBase::SelectDriver(videodriver, Driver::DT_VIDEO);
 	if (_video_driver == NULL) {
