Index: rail_gui.c
===================================================================
--- rail_gui.c	(revision 4356)
+++ rail_gui.c	(working copy)
@@ -154,6 +154,7 @@
 
 static void GenericPlaceSignals(TileIndex tile)
 {
+	static const uint Date1950_1_1 = 10958;
 	uint trackstat;
 	uint i;
 
@@ -172,7 +173,7 @@
 	}
 
 	if (!_remove_button_clicked) {
-		DoCommandP(tile, i + (_ctrl_pressed ? 8 : 0), 0, CcPlaySound1E,
+		DoCommandP(tile, i + (_ctrl_pressed ^ (_date < Date1950_1_1) ? 8 : 0), 0, CcPlaySound1E,
 			CMD_BUILD_SIGNALS | CMD_AUTO | CMD_MSG(STR_1010_CAN_T_BUILD_SIGNALS_HERE));
 	} else {
 		DoCommandP(tile, i, 0, CcPlaySound1E,
