diff --git a/src/core/pool_type.hpp b/src/core/pool_type.hpp
index 9c32a81..f267c12 100644
--- a/src/core/pool_type.hpp
+++ b/src/core/pool_type.hpp
@@ -83,6 +83,24 @@ struct Pool {
 		Tindex index; ///< Index of this pool item
 
 		/**
+		 * Don't initialize the index on construction - it's done by operator new.
+		 */
+		FORCEINLINE PoolItem() {}
+
+		/**
+		 * Don't copy the index on copy-constructing.
+		 */
+		FORCEINLINE PoolItem(const PoolItem &other) {}
+
+		/**
+		 * Don't copy the index on assignment.
+		 */
+		FORCEINLINE PoolItem &operator=(const PoolItem &other)
+		{
+			return *this;
+		}
+
+		/**
 		 * Allocates space for new Titem
 		 * @param size size of Titem
 		 * @return pointer to allocated memory
