Index: src/currency.cpp
===================================================================
--- src/currency.cpp	(revision 24347)
+++ src/currency.cpp	(working copy)
@@ -55,6 +55,7 @@
 	{    4, "", 2014,      "",             NBSP "Lt",        1, STR_GAME_OPTIONS_CURRENCY_LTL    }, ///< lithuanian litas
 	{ 1850, "", CF_NOEURO, "\xE2\x82\xA9", "",               0, STR_GAME_OPTIONS_CURRENCY_KRW    }, ///< south korean won
 	{   13, "", CF_NOEURO, "R" NBSP,       "",               0, STR_GAME_OPTIONS_CURRENCY_ZAR    }, ///< south african rand
+	{ 2400, "", CF_NOEURO, "",			   NBSP "Rls",       0, STR_GAME_OPTIONS_CURRENCY_IRN    }, ///< iranian rial
 	{    1, "", CF_NOEURO, "",             "",               2, STR_GAME_OPTIONS_CURRENCY_CUSTOM }, ///< custom currency
 };
 
Index: src/currency.h
===================================================================
--- src/currency.h	(revision 24347)
+++ src/currency.h	(working copy)
@@ -17,7 +17,7 @@
 
 static const int CF_NOEURO = 0; ///< Currency never switches to the Euro (as far as known).
 static const int CF_ISEURO = 1; ///< Currency _is_ the Euro.
-static const uint NUM_CURRENCY = 32; ///< Number of currencies.
+static const uint NUM_CURRENCY = 33; ///< Number of currencies.
 static const int CUSTOM_CURRENCY_ID = NUM_CURRENCY - 1; ///< Index of the custom currency.
 
 /** Specification of a currency. */
Index: src/lang/english.txt
===================================================================
--- src/lang/english.txt	(revision 24347)
+++ src/lang/english.txt	(working copy)
@@ -899,6 +899,7 @@
 STR_GAME_OPTIONS_CURRENCY_LTL                                   :Lithuanian Litas (LTL)
 STR_GAME_OPTIONS_CURRENCY_KRW                                   :South Korean Won (KRW)
 STR_GAME_OPTIONS_CURRENCY_ZAR                                   :South African Rand (ZAR)
+STR_GAME_OPTIONS_CURRENCY_IRN 									 :Iranian Rial (Rls)
 STR_GAME_OPTIONS_CURRENCY_CUSTOM                                :Custom...
 ############ end of currency region
 
