Index: src/bridge.h
===================================================================
--- src/bridge.h	(revision 21000)
+++ src/bridge.h	(working copy)
@@ -75,6 +75,6 @@
 static const uint MAX_BRIDGE_LENGTH             =  16; ///<Maximum length of the bridge
 static const uint MAX_BRIDGE_LENGTH_LONGBRIDGES = 100; ///<Maximum length of the bridge with enabled longbridges
 
-TileIndex GetBridgeHighlightTile(TileIndex tile, DiagDirection dir);
+TileIndex GetBridgeHighlightTile(TileIndex tile, DiagDirection dir, uint skip_length);
 
 #endif /* BRIDGE_H */
Index: src/tunnelbridge_cmd.cpp
===================================================================
--- src/tunnelbridge_cmd.cpp	(revision 21000)
+++ src/tunnelbridge_cmd.cpp	(working copy)
@@ -256,7 +256,7 @@
 	}
 }
 
-TileIndex GetBridgeHighlightTile(TileIndex tile, DiagDirection dir)
+TileIndex GetBridgeHighlightTile(TileIndex tile, DiagDirection dir, uint skip_length)
 {
 	assert(tile < MapSize());
 	if (!IsValidDiagDirection(dir)) dir = ReverseDiagDir(GetInclinedSlopeDirection(GetTileSlope(tile, NULL)));
@@ -290,7 +290,7 @@
 		}
 		if (z < GetTileMaxZ(tile)) last = true;
 
-		if (AllowBridgeRamp(tile)) {
+		if (AllowBridgeRamp(tile) && length > skip_length) {
 			uint cur_z;
 			Slope tileh = GetTileSlope(tile, &cur_z);
 			if (CanBuildBridgeRamp(ReverseDiagDir(dir), tileh, &cur_z) && cur_z == z) return tile;
Index: src/viewport.cpp
===================================================================
--- src/viewport.cpp	(revision 21000)
+++ src/viewport.cpp	(working copy)
@@ -2027,14 +2027,17 @@
 			int y1 = _thd.selstart.y & ~TILE_UNIT_MASK;
 			int x2 = _thd.selend.x   & ~TILE_UNIT_MASK;
 			int y2 = _thd.selend.y   & ~TILE_UNIT_MASK;
+			uint length = 0;
 			if (x1 != x2) {
 				dir = (x1 > x2) ? DIAGDIR_NE : DIAGDIR_SW;
+				length = abs(x1 - x2) / TILE_SIZE;
 			} else if (y1 != y2) {
 				dir = (y1 > y2) ? DIAGDIR_NW : DIAGDIR_SE;
+				length = abs(y1 - y2) / TILE_SIZE;
 			}
 
 			TileIndex tile = TileVirtXY(x1, y1);
-			tile = (tile < MapSize() ? GetBridgeHighlightTile(tile, dir) : INVALID_TILE);
+			tile = (tile < MapSize() ? GetBridgeHighlightTile(tile, dir, length) : INVALID_TILE);
 			SetRedErrorSquare(tile);
 		}
 
@@ -2067,7 +2070,7 @@
 {
 	if (process == DDSP_BUILD_BRIDGE) {
 		_thd.bridge_highlight = true;
-		SetRedErrorSquare(GetBridgeHighlightTile(tile, INVALID_DIAGDIR));
+		SetRedErrorSquare(GetBridgeHighlightTile(tile, INVALID_DIAGDIR, 0));
 	}
 
 	_thd.select_method = method;
