Index: config.lib
===================================================================
--- config.lib	(revision 10455)
+++ config.lib	(working copy)
@@ -756,7 +756,7 @@
 				LDFLAGS="$LDFLAGS -noixemul"
 			fi
 
-			CFLAGS="$CFLAGS -O2 -fomit-frame-pointer"
+			CFLAGS="-O2 -fomit-frame-pointer $CFLAGS"
 		fi
 	else
 		OBJS_SUBDIR="debug"
@@ -765,7 +765,7 @@
 		if [ $enable_debug -ge 1 ]; then
 			CFLAGS="$CFLAGS -g -D_DEBUG"
 			if [ "$os" = "PSP" ]; then
-				CFLAGS="$CFLAGS -G0"
+				CFLAGS="-G0 $CFLAGS"
 			fi
 		fi
 		if [ $enable_debug -ge 2 ]; then
