Index: src/core/random_func.hpp
===================================================================
--- src/core/random_func.hpp	(revision 16057)
+++ src/core/random_func.hpp	(working copy)
@@ -129,13 +129,13 @@
  * Flips a coin with a given probability.
  *
  * This macro can be used to get true or false randomized according to a
- * given probability. The parameter a and b create a percent value with
- * (a/b). The macro returns true in (a/b) percent.
+ * given probability. The parameter a and b create a value with
+ * (a/b). The macro returns true with (a/b) probability.
  *
  * @see Chance16I()
  * @param a The numerator of the fraction
  * @param b The denominator of the fraction
- * @return True in (a/b) percent
+ * @return True with (a/b) probability
  */
 static FORCEINLINE bool Chance16(const uint a, const uint b)
 {
