Index: train_cmd.c
===================================================================
--- train_cmd.c	(revision 10373)
+++ train_cmd.c	(working copy)
@@ -2749,7 +2749,11 @@
 static byte AfterSetTrainPos(Vehicle *v, bool new_tile)
 {
 	byte old_z = v->z_pos;
+
+	// need this hint so it returns the right z coordinate on bridges.
+	_get_z_hint = v->z_pos;
 	v->z_pos = GetSlopeZ(v->x_pos, v->y_pos);
+	_get_z_hint = 0;
 
 	if (new_tile) {
 		CLRBIT(v->u.rail.flags, VRF_GOINGUP);
