commit 090a468dcc2e39f35e567b82ec99d17bbe3e9fb5
Author: fonsinchen <ulfonk_mennhar@gmx.de>
Date:   Sun Feb 17 19:03:46 2013 +0100

    -Fix: integer overflow when truncating cargo

diff --git a/src/station_cmd.cpp b/src/station_cmd.cpp
index 1285286..b2f6ba3 100644
--- a/src/station_cmd.cpp
+++ b/src/station_cmd.cpp
@@ -3307,7 +3307,9 @@ static void UpdateStationRating(Station *st)
 					waiting_changed = true;
 				}
 
-				if (waiting_changed) ge->cargo.Truncate(ge->cargo.Count() - waiting);
+				if (waiting_changed && waiting < ge->cargo.Count()) {
+					ge->cargo.Truncate(ge->cargo.Count() - waiting);
+				}
 			}
 		}
 	}
