"Show group name for unnamed vehicles in vehicle lists."

diff -r 3a43f7f011ec src/lang/english.txt
--- a/src/lang/english.txt	Tue Nov 18 23:58:24 2008 +0000
+++ b/src/lang/english.txt	Wed Nov 19 15:48:35 2008 +0100
@@ -3557,6 +3557,8 @@ STR_SIGN_NAME                           
 STR_SIGN_NAME                                                   :{SIGN}
 STR_VEHICLE_NAME                                                :{VEHICLE}
 
+STR_GROUP_NAME_TINY                                             :{BLACK}{TINYFONT}{GROUP}
+
 STR_NAME_MUST_BE_UNIQUE                                         :{WHITE}Name must be unique
 
 #### Improved sign GUI
diff -r 3a43f7f011ec src/vehicle_gui.cpp
--- a/src/vehicle_gui.cpp	Tue Nov 18 23:58:24 2008 +0000
+++ b/src/vehicle_gui.cpp	Wed Nov 19 15:48:35 2008 +0100
@@ -772,6 +772,10 @@ void BaseVehicleListWindow::DrawVehicleL
 			/* The vehicle got a name so we will print it */
 			SetDParam(0, v->index);
 			DrawString(x + 19, y, STR_01AB, TC_FROMSTRING);
+		} else if (v->group_id != DEFAULT_GROUP) {
+			/* The vehicle has no name, but is member of a group, so print group name */
+			SetDParam(0, v->group_id);
+			DrawString(x + 19, y, STR_GROUP_NAME_TINY, TC_FROMSTRING);
 		}
 
 		if (this->resize.step_height == PLY_WND_PRC__SIZE_OF_ROW_BIG) DrawSmallOrderList(v, x + 138, y);
