diff -ruN openttd-0.7.0-RC1.orig/src/sound.cpp openttd-0.7.0-RC1/src/sound.cpp
--- openttd-0.7.0-RC1.orig/src/sound.cpp	2009-03-15 01:32:18.000000000 +0100
+++ openttd-0.7.0-RC1/src/sound.cpp	2009-03-22 14:30:51.246897600 +0100
@@ -29,6 +29,10 @@
 	_file_count = ORIGINAL_SAMPLE_COUNT;
 	_files = fe;
 
+	if(!FioCheckFileExists(filename)) {
+		return;
+	}
+
 	FioOpenFile(SOUND_SLOT, filename);
 	size_t pos = FioGetPos();
 	uint count = FioReadDword() / 8;
