Index: src/misc_gui.cpp
===================================================================
--- src/misc_gui.cpp	(Revision 20806)
+++ src/misc_gui.cpp	(Arbeitskopie)
@@ -167,6 +167,11 @@
 		GetString(this->landinfo_data[line_nr], td.str, lastof(this->landinfo_data[line_nr]));
 		line_nr++;
 
+		/* Height */
+		SetDParam(0, TileHeight(tile));
+		GetString(this->landinfo_data[line_nr], STR_LAND_AREA_INFORMATION_LANDINFO_HEIGHT, lastof(this->landinfo_data[line_nr]));
+		line_nr++;
+
 		/* Up to four owners */
 		for (uint i = 0; i < 4; i++) {
 			if (td.owner_type[i] == STR_NULL) continue;
