Remove now unneeded functions/methods.

diff -r 93b88da34acb src/order_base.h
--- a/src/order_base.h	Wed Sep 24 22:56:12 2008 +0200
+++ b/src/order_base.h	Wed Sep 24 22:56:39 2008 +0200
@@ -117,12 +117,6 @@ public:
 	 * @param order the order to jump to.
 	 */
 	void MakeConditional(VehicleOrderID order);
-
-	/**
-	 * Free a complete order chain.
-	 * @note do not use on "current_order" vehicle orders!
-	 */
-	void FreeChain();
 
 	/**
 	 * Gets the destination of this order.
diff -r 93b88da34acb src/order_cmd.cpp
--- a/src/order_cmd.cpp	Wed Sep 24 22:56:12 2008 +0200
+++ b/src/order_cmd.cpp	Wed Sep 24 22:56:39 2008 +0200
@@ -109,12 +109,6 @@ void Order::SetRefit(CargoID cargo, byte
 	this->refit_subtype = subtype;
 }
 
-void Order::FreeChain()
-{
-	if (next != NULL) next->FreeChain();
-	delete this;
-}
-
 bool Order::Equals(const Order &other) const
 {
 	return
@@ -246,22 +240,6 @@ void InvalidateVehicleOrder(const Vehicl
 
 	InvalidateWindow(WC_VEHICLE_ORDERS,    v->index);
 	InvalidateWindow(WC_VEHICLE_TIMETABLE, v->index);
-}
-
-/**
- *
- * Swap two orders
- *
- */
-static void SwapOrders(Order *order1, Order *order2)
-{
-	Order temp_order;
-
-	temp_order = *order1;
-	order1->AssignOrder(*order2);
-	order1->next = order2->next;
-	order2->AssignOrder(temp_order);
-	order2->next = temp_order.next;
 }
 
 /**
