Index: config.lib
===================================================================
--- config.lib	(revision 24463)
+++ config.lib	(working copy)
@@ -218,6 +218,9 @@
 			--endian=*)                   endian="$optarg";;
 
 
+			# Be kind to people who are used to GNU autotools.
+			--prefix)                     prev_p="prefix-dir";;
+			--prefix=*)                   prefix_dir="$optarg";;
 
 			--prefix-dir)                 prevp_p="prefix-dir";;
 			--prefix-dir=*)               prefix_dir="$optarg";;
