diff --git a/src/linkgraph/demands.cpp b/src/linkgraph/demands.cpp
index 767bec2..7856b1a 100644
--- a/src/linkgraph/demands.cpp
+++ b/src/linkgraph/demands.cpp
@@ -2,11 +2,11 @@
 
 #include "../stdafx.h"
 #include "demands.h"
-#include <list>
+#include <deque>
 
 #include "../safeguards.h"
 
-typedef std::list<NodeID> NodeList;
+typedef std::deque<NodeID> NodeList;
 
 /**
  * Scale various things according to symmetric/asymmetric distribution.
diff --git a/src/linkgraph/linkgraph_gui.h b/src/linkgraph/linkgraph_gui.h
index fcf8181..1e306a4 100644
--- a/src/linkgraph/linkgraph_gui.h
+++ b/src/linkgraph/linkgraph_gui.h
@@ -17,7 +17,7 @@
 #include "../widget_type.h"
 #include "linkgraph_base.h"
 #include <map>
-#include <list>
+#include <vector>
 
 /**
  * Properties of a link between two stations.
@@ -39,7 +39,7 @@ class LinkGraphOverlay {
 public:
 	typedef std::map<StationID, LinkProperties> StationLinkMap;
 	typedef std::map<StationID, StationLinkMap> LinkMap;
-	typedef std::list<std::pair<StationID, uint> > StationSupplyList;
+	typedef std::vector<std::pair<StationID, uint> > StationSupplyList;
 
 	static const uint8 LINK_COLOURS[];
 
diff --git a/src/linkgraph/refresh.h b/src/linkgraph/refresh.h
index 496729d..d2f92e3 100644
--- a/src/linkgraph/refresh.h
+++ b/src/linkgraph/refresh.h
@@ -14,7 +14,7 @@
 
 #include "../cargo_type.h"
 #include "../vehicle_base.h"
-#include <list>
+#include <deque>
 #include <map>
 #include <set>
 
@@ -79,7 +79,7 @@ protected:
 		bool operator<(const Hop &other) const;
 	};
 
-	typedef std::list<RefitDesc> RefitList;
+	typedef std::deque<RefitDesc> RefitList;
 	typedef std::map<CargoID, uint> CapacitiesMap;
 	typedef std::set<Hop> HopSet;
 
