Index: src/rail_cmd.cpp
===================================================================
--- src/rail_cmd.cpp	(revision 11166)
+++ src/rail_cmd.cpp	(working copy)
@@ -1066,6 +1066,7 @@
 
 		if (IsTileDepotType(tile, TRANSPORT_RAIL)) {
 			/* Update build vehicle window related to this depot */
+			InvalidateWindowData(WC_VEHICLE_DEPOT, tile);
 			InvalidateWindowData(WC_BUILD_VEHICLE, tile);
 		}
 
