Index: src/dock_gui.cpp
===================================================================
--- src/dock_gui.cpp	(revision 21000)
+++ src/dock_gui.cpp	(working copy)
@@ -290,9 +290,13 @@
 	virtual void OnPlacePresize(Point pt, TileIndex tile_from)
 	{
 		DiagDirection dir = GetInclinedSlopeDirection(GetTileSlope(tile_from, NULL));
-		TileIndex tile_to = (dir != INVALID_DIAGDIR ? TileAddByDiagDir(tile_from, ReverseDiagDir(dir)) : tile_from);
-		tile_from = this->last_clicked_widget == DTW_LOCK && dir != INVALID_DIAGDIR ? TileAddByDiagDir(tile_from, dir) : tile_from;
+		TileIndex tile_to = tile_from;
 
+		if (IsValidDiagDirection(dir)) {
+			tile_to = TileAddByDiagDir(tile_from, ReverseDiagDir(dir));
+			tile_from = this->last_clicked_widget == DTW_LOCK ? TileAddByDiagDir(tile_from, dir) : tile_from;
+		}
+
 		VpSetPresizeRange(tile_from, tile_to);
 	}
 
