Index: train_cmd.c
===================================================================
--- train_cmd.c	(revision 7493)
+++ train_cmd.c	(working copy)
@@ -2514,6 +2514,8 @@
 {
 	const Order *order;
 	bool at_waypoint = false;
+	Vehicle *v2 = v;
+	bool all_full = true;
 
 	switch (v->current_order.type) {
 		case OT_GOTO_DEPOT:
@@ -2558,6 +2560,31 @@
 		return false;
 	}
 
+	if (order->flags & OF_FULL_LOAD) {
+		// check if the train is completly full
+		// TODO: check for fullness by cargo type, and compare to what's waiting at the next station?
+		for(; v2 != NULL; v2 = v2->next) {
+			if (v2->cargo_count != v2->cargo_cap) {
+				all_full = false;
+			}
+		}
+		if (all_full) {
+			v->cur_order_index++;
+			// Get the current order
+			if (v->cur_order_index >= v->num_orders) v->cur_order_index = 0;
+
+			order = GetVehicleOrder(v, v->cur_order_index);
+	
+			// If no order, do nothing.
+			if (order == NULL) {
+				v->current_order.type = OT_NOTHING;
+				v->current_order.flags = 0;
+				v->dest_tile = 0;
+				return false;
+			}
+		}
+	}
+
 	// If it is unchanged, keep it.
 	if (order->type  == v->current_order.type &&
 			order->flags == v->current_order.flags &&
