diff --git a/findversion.sh b/findversion.sh
--- a/findversion.sh
+++ b/findversion.sh
@@ -105,7 +105,12 @@
 		MODIFIED="2"
 	fi
 	HASH=`LC_ALL=C hg id -i | cut -c1-12`
-	REV="h`echo $HASH | cut -c1-8`"
+	TAG="$(hg id -t)"
+	if [ -n "$TAG" ]  && [ $TAG != "tip" ]; then
+		REV=$TAG
+	else
+		REV="h`echo $HASH | cut -c1-8`"
+	fi
 	BRANCH=`hg branch | sed 's@^default$@@'`
 	REV_NR=`LC_ALL=C hg log -f -k "(svn r" -l 1 --template "{desc}\n" | head -n 1 | sed "s@.*(svn r\([0-9]*\)).*@\1@"`
 elif [ -f "$ROOT_DIR/.ottdrev" ]; then
