--- C:/Users/Guy/AppData/Local/Temp/win32_v.cpp-revBASE.svn002.tmp.cpp	Tue Jul 10 13:47:03 2012
+++ C:/Code Repository/openttd/src/video/win32_v.cpp	Sat Nov 24 13:59:01 2012
@@ -48,6 +48,33 @@
 uint _codepage;
 #endif
 
+/* as of Windows 8 (version 6.2), windows no longer supports 8bbp */
+static bool WinSupports8BPP() { 
+#if defined(_M_X64)
+	OSVERSIONINFOEX osvi;
+	DWORDLONG dwlConditionMask = 0;
+
+	// Initialize the OSVERSIONINFOEX structure.
+	ZeroMemory(&osvi, sizeof(OSVERSIONINFOEX));
+	osvi.dwOSVersionInfoSize = sizeof(OSVERSIONINFOEX);
+	osvi.dwMajorVersion = 6;
+	osvi.dwMinorVersion = 2;
+	osvi.wServicePackMajor = 2;
+	osvi.wServicePackMinor = 0;
+
+	// Initialize the condition mask.
+	VER_SET_CONDITION( dwlConditionMask, VER_MAJORVERSION, VER_GREATER_EQUAL );
+	VER_SET_CONDITION( dwlConditionMask, VER_MINORVERSION, VER_GREATER_EQUAL );
+
+	return !((BOOL) VerifyVersionInfo(
+      &osvi, 
+      VER_MAJORVERSION | VER_MINORVERSION,
+      dwlConditionMask));
+#else
+	return true; //32bit does not have this limitation
+#endif
+}
+
 /** Whether the drawing is/may be done in a separate thread. */
 static bool _draw_threaded;
 /** Thread used to 'draw' to the screen, i.e. push data to the screen. */
@@ -277,7 +304,12 @@
 			DM_PELSWIDTH |
 			DM_PELSHEIGHT |
 			(_display_hz != 0 ? DM_DISPLAYFREQUENCY : 0);
-		settings.dmBitsPerPel = _fullscreen_bpp;
+
+		if (WinSupports8BPP() == false)
+			settings.dmBitsPerPel = 32; //must be
+		else
+			settings.dmBitsPerPel = _fullscreen_bpp;
+
 		settings.dmPelsWidth  = _wnd.width_org;
 		settings.dmPelsHeight = _wnd.height_org;
 		settings.dmDisplayFrequency = _display_hz;
