diff -r 49e3bccad1f2 -r 965e0148135b src/settings_gui.cpp
--- a/src/settings_gui.cpp	Sat Nov 15 11:14:16 2008 +0100
+++ b/src/settings_gui.cpp	Sat Nov 15 12:34:04 2008 +0100
@@ -585,6 +585,26 @@
 
 static const int SETTING_HEIGHT = 11;         ///< Height of a single patch setting in the tree view
 
+/** Data structure describing a single patch in a tab */
+struct PatchEntry {
+	const SettingDesc *setting; ///< Setting description of the patch
+	uint index;                 ///< Index of the setting in the settings table
+};
+
+/**
+ * Data structure describing one page of patches in the patch settings window.
+ *
+ * The names of the patches to display are statically defined, and from this
+ * information, a dynamic array (with length \a num) of PatchEntry entries is
+ * constructed.
+ */
+struct PatchPage {
+	const char **names;  ///< Static list of strings with patch names that are settable from the tab
+	PatchEntry *entries; ///< Array of patch entries of the page. Initially \c NULL, filled in at run time
+	byte num;            ///< Number of entries on the page (statically filled).
+};
+
+
 static const char *_patches_ui[] = {
 	"gui.vehicle_speed",
 	"gui.status_long_date",
@@ -708,25 +728,6 @@
 	"vehicle.plane_speed",
 	"order.timetabling",
 	"vehicle.dynamic_engines",
-};
-
-/** Data structure describing a single patch in a tab */
-struct PatchEntry {
-	const SettingDesc *setting; ///< Setting description of the patch
-	uint index;                 ///< Index of the setting in the settings table
-};
-
-/**
- * Data structure describing one page of patches in the patch settings window.
- *
- * The names of the patches to display are statically defined, and from this
- * information, a dynamic array (with length \a num) of PatchEntry entries is
- * constructed.
- */
-struct PatchPage {
-	const char **names;  ///< Static list of strings with patch names that are settable from the tab
-	PatchEntry *entries; ///< Array of patch entries of the page. Initially \c NULL, filled in at run time
-	byte num;            ///< Number of entries on the page (statically filled).
 };
 
 /** Array of pages (tabs), where each page holds a number of advanced settings. */
