--- C:\DOCUME~1\X\LOKALA~1\Temp\gfx.cpp-revBASE.svn000.tmp.cpp	to maj 14 22:13:33 2009
+++ D:\D\Projects\OpenTTD\trunk\src\gfx.cpp	to maj 14 22:14:21 2009
@@ -30,6 +30,7 @@
 CursorVars _cursor;
 bool _ctrl_pressed;   ///< Is Ctrl pressed?
 bool _shift_pressed;  ///< Is Shift pressed?
+bool _alt_pressed;  ///< Is Alt pressed?
 byte _fast_forward;
 bool _left_button_down;     ///< Is left mouse button pressed?
 bool _left_button_clicked;  ///< Is left mouse button clicked?
