Index: src/saveload/afterload.cpp
===================================================================
--- src/saveload/afterload.cpp	(revision 20890)
+++ src/saveload/afterload.cpp	(working copy)
@@ -2320,6 +2320,53 @@
 		}
 	}
 
+	if (CheckSavegameVersion(149)) {
+		Trackdir _road_exit_tunnel_state[DIAGDIR_END] = {TRACKDIR_X_SW, TRACKDIR_Y_NW, TRACKDIR_X_NE, TRACKDIR_Y_SE};
+		Vehicle *v;
+		FOR_ALL_VEHICLES(v) {
+			if (v->type != VEH_TRAIN && v->type != VEH_ROAD) continue;
+			if (!IsTunnelTile(v->tile)) continue;
+
+			TileIndex tile = TileVirtXY(v->x_pos, v->y_pos);
+			if (v->tile != tile && GetOtherTunnelBridgeEnd(v->tile) != tile) continue;
+
+			DiagDirection dir = GetTunnelBridgeDirection(tile);
+
+			v->tile = tile;
+			switch (v->type) {
+				case VEH_TRAIN: {
+					if (Train::From(v)->track == TRACK_BIT_WORMHOLE) {
+						Train::From(v)->track = (DiagDirToAxis(dir) == AXIS_X ? TRACK_BIT_X : TRACK_BIT_Y);
+					} else {
+						continue;
+					}
+				} break;
+
+				case VEH_ROAD: {
+					if (RoadVehicle::From(v)->state == RVSB_WORMHOLE) {
+						RoadVehicle::From(v)->state = _road_exit_tunnel_state[dir];
+						RoadVehicle::From(v)->frame = 0; // not 100% shure here but it works
+					} else {
+						continue;
+					}
+				} break;
+
+				default: NOT_REACHED();
+			}
+
+			if (dir != DirToDiagDir(v->direction)) continue;
+
+			switch (dir) {
+				default: NOT_REACHED();
+				case DIAGDIR_NE: if ((v->x_pos & 0xF) !=  0)            continue; break;
+				case DIAGDIR_SE: if ((v->y_pos & 0xF) != TILE_SIZE - 1) continue; break;
+				case DIAGDIR_SW: if ((v->x_pos & 0xF) != TILE_SIZE - 1) continue; break;
+				case DIAGDIR_NW: if ((v->y_pos & 0xF) !=  0)            continue; break;
+			}
+			VehicleEnterTile(v, tile, v->x_pos, v->y_pos); // just before leaving ramp
+		}
+	}
+
 	/* Road stops is 'only' updating some caches */
 	AfterLoadRoadStops();
 	AfterLoadLabelMaps();
