Index: src/saveload/saveload.cpp
===================================================================
--- src/saveload/saveload.cpp	(revision 26643)
+++ src/saveload/saveload.cpp	(working copy)
@@ -674,7 +674,7 @@
 			if (HasBit(i, 5)) {
 				i &= ~0x20;
 				if (HasBit(i, 4)) {
-					SlErrorCorrupt("Unsupported gamma");
+//					SlErrorCorrupt("Unsupported gamma");
 				}
 				i = (i << 8) | SlReadByte();
 			}
Index: src/vehicle.cpp
===================================================================
--- src/vehicle.cpp	(revision 26643)
+++ src/vehicle.cpp	(working copy)
@@ -235,6 +235,8 @@
 	const Engine *e = Engine::Get(engine);
 	GRFConfig *grfconfig = GetGRFConfig(e->GetGRFID());
 
+	if (grfconfig == NULL) return;
+
 	if (!HasBit(grfconfig->grf_bugs, bug_type)) {
 		SetBit(grfconfig->grf_bugs, bug_type);
 		SetDParamStr(0, grfconfig->GetName());
