diff -r 11011d9efb86 src/landscape.cpp
--- a/src/landscape.cpp	Fri Oct 17 18:04:33 2008 +0200
+++ b/src/landscape.cpp	Fri Oct 17 18:04:58 2008 +0200
@@ -1,6 +1,8 @@
 /* $Id$ */
 
 /** @file landscape.cpp Functions related to the landscape (slopes etc.). */
+
+/** @defgroup SnowLineGroup Snowline functions and data structures */
 
 #include "stdafx.h"
 #include "openttd.h"
@@ -45,6 +47,7 @@
 	_tile_type_unmovable_procs;
 
 /** Tile callback functions for each type of tile.
+ * @ingroup TileCallbackGroup
  * @see TileType */
 const TileTypeProcs * const _tile_type_procs[16] = {
 	&_tile_type_clear_procs,        ///< Callback functions for MP_CLEAR tiles
@@ -71,6 +74,7 @@
  *
  * If it is \c NULL, a static snowline height is used, as set by \c _settings_game.game_creation.snow_line.
  * Otherwise it points to a table loaded from a newGRF file, that describes the variable snowline
+ * @ingroup SnowLineGroup
  * @see GetSnowLine() GameCreationSettings */
 SnowLine *_snow_line = NULL;
 
@@ -516,6 +520,7 @@
 /**
  * Has a snow line table already been loaded.
  * @return true if the table has been loaded already.
+ * @ingroup SnowLineGroup
  */
 bool IsSnowLineSet(void)
 {
@@ -525,6 +530,7 @@
 /**
  * Set a variable snow line, as loaded from a newgrf file.
  * @param table the 12 * 32 byte table containing the snowline for each day
+ * @ingroup SnowLineGroup
  */
 void SetSnowLine(byte table[SNOW_LINE_MONTHS][SNOW_LINE_DAYS])
 {
@@ -541,6 +547,7 @@
 /**
  * Get the current snow line, either variable or static.
  * @return the snow line height.
+ * @ingroup SnowLineGroup
  */
 byte GetSnowLine(void)
 {
@@ -554,6 +561,7 @@
 /**
  * Get the highest possible snow line height, either variable or static.
  * @return the highest snow line height.
+ * @ingroup SnowLineGroup
  */
 byte HighestSnowLine(void)
 {
@@ -562,6 +570,7 @@
 
 /**
  * Clear the variable snow line table and free the memory.
+ * @ingroup SnowLineGroup
  */
 void ClearSnowLine(void)
 {
diff -r 11011d9efb86 src/landscape.h
--- a/src/landscape.h	Fri Oct 17 18:04:33 2008 +0200
+++ b/src/landscape.h	Fri Oct 17 18:04:58 2008 +0200
@@ -12,15 +12,16 @@
 
 /** Number of months in the snow line table.
  * @see SnowLine
- */
+ * @ingroup SnowLineGroup */
 static const uint SNOW_LINE_MONTHS = 12;
 
 /** Number of days in each month in the snow line table.
  * @see SnowLine
- */
+ * @ingroup SnowLineGroup */
 static const uint SNOW_LINE_DAYS   = 32;
 
-/** Structure describing the height of the snow line each day of the year */
+/** Structure describing the height of the snow line each day of the year
+ * @ingroup SnowLineGroup */
 struct SnowLine {
 	byte table[SNOW_LINE_MONTHS][SNOW_LINE_DAYS]; ///< Height of the snow line each day of the year
 	byte highest_value; ///< Highest snow line of the year
