Index: src/town_cmd.cpp
===================================================================
--- src/town_cmd.cpp	(revision 10727)
+++ src/town_cmd.cpp	(working copy)
@@ -1960,7 +1960,8 @@
 	PlayerID old;
 	CommandCost r;
 
-	if (GetTileSlope(tile, NULL) != SLOPE_FLAT) return false;
+	/* Statues can be build on slopes, just like houses. Only the steep slopes is a no go. */
+	if (IsSteepSlope(GetTileSlope(tile, NULL))) return false;
 
 	if (!IsTileType(tile, MP_HOUSE) &&
 			!IsTileType(tile, MP_CLEAR) &&
Index: src/unmovable_cmd.cpp
===================================================================
--- src/unmovable_cmd.cpp	(revision 10727)
+++ src/unmovable_cmd.cpp	(working copy)
@@ -110,6 +110,8 @@
 	return cost;
 }
 
+static Foundation GetFoundation_Unmovable(TileIndex tile, Slope tileh);
+
 static void DrawTile_Unmovable(TileInfo *ti)
 {
 
@@ -132,6 +134,9 @@
 		}
 
 		case UNMOVABLE_STATUE:
+			/* This should prevent statues from sinking into the ground when on a slope. */			 
+			if (ti->tileh != SLOPE_FLAT) DrawFoundation(ti, GetFoundation_Unmovable(ti->tile, ti->tileh));
+
 			DrawGroundSprite(SPR_CONCRETE_GROUND, PAL_NONE);
 
 			AddSortableSpriteToDraw(SPR_STATUE_COMPANY, PLAYER_SPRITE_COLOR(GetTileOwner(ti->tile)), ti->x, ti->y, 16, 16, 25, ti->z, HASBIT(_transparent_opt, TO_STRUCTURES));
