Index: src/timetable_cmd.cpp
===================================================================
--- src/timetable_cmd.cpp	(revision 27298)
+++ src/timetable_cmd.cpp	(working copy)
@@ -154,15 +154,28 @@
 	if (max_speed != order->GetMaxSpeed() && (order->IsType(OT_CONDITIONAL) || v->type == VEH_AIRCRAFT)) return CMD_ERROR;
 
 	if (flags & DC_EXEC) {
-		if (wait_time != order->GetWaitTime() || (wait_time > 0 && !order->IsWaitTimetabled())) {
-			ChangeTimetable(v, order_number, wait_time, MTF_WAIT_TIME, wait_time > 0);
+		switch (mtf) {
+			case MTF_WAIT_TIME:
+				if (wait_time != order->GetWaitTime() || (wait_time > 0 && !order->IsWaitTimetabled())) {
+					ChangeTimetable(v, order_number, wait_time, MTF_WAIT_TIME, wait_time > 0);
+				}
+				break;
+
+			case MTF_TRAVEL_TIME:
+				if (travel_time != order->GetTravelTime() || (travel_time > 0 && !order->IsTravelTimetabled())) {
+					ChangeTimetable(v, order_number, travel_time, MTF_TRAVEL_TIME, travel_time > 0);
+				}
+				break;
+
+			case MTF_TRAVEL_SPEED:
+				if (max_speed != order->GetMaxSpeed()) {
+					ChangeTimetable(v, order_number, max_speed, MTF_TRAVEL_SPEED, max_speed != UINT16_MAX);
+				}
+				break;
+
+			default:
+				NOT_REACHED();
 		}
-		if (travel_time != order->GetTravelTime() || (travel_time > 0 && !order->IsTravelTimetabled())) {
-			ChangeTimetable(v, order_number, travel_time, MTF_TRAVEL_TIME, travel_time > 0);
-		}
-		if (max_speed != order->GetMaxSpeed()) {
-			ChangeTimetable(v, order_number, max_speed, MTF_TRAVEL_SPEED, max_speed != UINT16_MAX);
-		}
 	}
 
 	return CommandCost();
