Index: src/rail_gui.cpp
===================================================================
--- src/rail_gui.cpp	(revision 11127)
+++ src/rail_gui.cpp	(working copy)
@@ -881,6 +881,28 @@
 		case 19:
 			_railstation.dragdrop ^= true;
 			ToggleWidgetLoweredState(w, 19);
+			
+			/* get the first allowed length/number of platforms */
+			const StationSpec *statspec = _railstation.newstations ? GetCustomStationSpec(_railstation.station_class, _railstation.station_type) : NULL;
+			if (statspec != NULL && HASBIT(statspec->disallowed_lengths, _railstation.platlength - 1)) {
+				for (uint i = 0; i < 7; i++) {
+					if (!HASBIT(statspec->disallowed_lengths, i)) {
+						RaiseWindowWidget(w, _railstation.platlength + 11);
+						_railstation.platlength = i + 1;
+						break;
+					}
+				}
+			}
+			if (statspec != NULL && HASBIT(statspec->disallowed_platforms, _railstation.numtracks - 1)) {
+				for (uint i = 0; i < 7; i++) {
+					if (!HASBIT(statspec->disallowed_platforms, i)) {
+						RaiseWindowWidget(w, _railstation.numtracks + 4);
+						_railstation.numtracks = i + 1;
+						break;
+					}
+				}
+			}
+
 			SetWindowWidgetLoweredState(w, _railstation.numtracks + 4, !_railstation.dragdrop);
 			SetWindowWidgetLoweredState(w, _railstation.platlength + 11, !_railstation.dragdrop);
 			SndPlayFx(SND_15_BEEP);
