Index: projects/openttd_vs140.vcxproj
===================================================================
--- projects/openttd_vs140.vcxproj	(revision 27528)
+++ projects/openttd_vs140.vcxproj	(working copy)
@@ -198,6 +198,10 @@
       <TargetMachine>MachineX86</TargetMachine>
       <MinimumRequiredVersion>5.01</MinimumRequiredVersion>
     </Link>
+    <Manifest />
+    <Manifest>
+      <EnableDpiAwareness>PerMonitorHighDPIAware</EnableDpiAwareness>
+    </Manifest>
   </ItemDefinitionGroup>
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'">
     <Midl>
