Index: src/table/sprites.h
===================================================================
--- src/table/sprites.h	(Revision 20854)
+++ src/table/sprites.h	(Arbeitskopie)
@@ -56,7 +56,7 @@
 
 /* Extra graphic spritenumbers */
 static const SpriteID SPR_OPENTTD_BASE   = 4896;
-static const uint16 OPENTTD_SPRITE_COUNT = 160;
+static const uint16 OPENTTD_SPRITE_COUNT = 179;
 
 /* Halftile-selection sprites */
 static const SpriteID SPR_HALFTILE_SELECTION_FLAT = SPR_OPENTTD_BASE;
@@ -560,6 +560,8 @@
 static const SpriteID SPR_FLAT_2_QUART_SNOW_DESERT_TILE = 4512;
 static const SpriteID SPR_FLAT_3_QUART_SNOW_DESERT_TILE = 4531;
 static const SpriteID SPR_FLAT_SNOW_DESERT_TILE         = 4550;
+/* Blacktiles all slopes. */
+static const SpriteID SPR_FLAT_BLACKTILES               = SPR_OPENTTD_BASE + 160;
 
 /* Hedge, Farmland-fence sprites */
 static const SpriteID SPR_HEDGE_BUSHES            = 4090;
Index: src/void_cmd.cpp
===================================================================
--- src/void_cmd.cpp	(Revision 20854)
+++ src/void_cmd.cpp	(Arbeitskopie)
@@ -13,13 +13,14 @@
 #include "tile_cmd.h"
 #include "command_func.h"
 #include "viewport_func.h"
+#include "slope_func.h"
 
 #include "table/strings.h"
 #include "table/sprites.h"
 
 static void DrawTile_Void(TileInfo *ti)
 {
-	DrawGroundSprite(SPR_SHADOW_CELL, PAL_NONE);
+	DrawGroundSprite(SPR_FLAT_BLACKTILES + SlopeToSpriteOffset(ti->tileh), PAL_NONE);
 }
 
 
Index: src/gfxinit.cpp
===================================================================
--- src/gfxinit.cpp	(Revision 20854)
+++ src/gfxinit.cpp	(Arbeitskopie)
@@ -154,6 +154,9 @@
 	_palette_remap_grf[i] = (_use_palette != used_set->palette);
 	LoadGrfFile(used_set->files[GFT_LOGOS].filename, 4793, i++);
 
+	/* Blacktileset comlete all slopes. */
+	LoadGrfFile("flat_blacktiles.grf", SPR_FLAT_BLACKTILES, i++);
+
 	/*
 	 * Load additional sprites for climates other than temperate.
 	 * This overwrites some of the temperate sprites, such as foundations
