Index: src/station_cmd.cpp
===================================================================
--- src/station_cmd.cpp	(Revision 10606)
+++ src/station_cmd.cpp	(Arbeitskopie)
@@ -362,7 +362,7 @@
 static void ShowRejectOrAcceptNews(const Station *st, uint num_items, CargoID *cargo, StringID msg)
 {
 	for (uint i = 0; i < num_items; i++) {
-		SetDParam(i + 1, GetCargo(cargo[i])->name);
+		SetDParam(i + 1, GetCargo(cargo[i])->name_plural);
 	}
 
 	SetDParam(0, st->index);
