Index: src/newgrf_canal.cpp
===================================================================
--- src/newgrf_canal.cpp	(revision 11039)
+++ src/newgrf_canal.cpp	(working copy)
@@ -42,7 +42,7 @@
 
 	switch (variable) {
 		case 0x80:
-			return TileHeight(tile);
+			return GetTileZ(tile) / TILE_HEIGHT;
 
 		case 0x81:
 			return GetTerrainType(tile);
Index: src/newgrf_industrytiles.cpp
===================================================================
--- src/newgrf_industrytiles.cpp	(revision 11039)
+++ src/newgrf_industrytiles.cpp	(working copy)
@@ -39,7 +39,9 @@
 	is_same_industry = (IsTileType(tile, MP_INDUSTRY) && GetIndustryIndex(tile) == index);
 	tile_type = GetTerrainType(tile) << 2 | (IsTileType(tile, MP_WATER) ? 1 : 0) << 1 | (is_same_industry ? 1 : 0);
 
-	return GetTileType(tile) << 24 | (TileHeight(tile) * 8) << 16 | tile_type << 8 | GetTileSlope(tile, NULL);
+	uint z;
+	Slope tileh = GetTileSlope(tile, &z);
+	return GetTileType(tile) << 24 | z << 16 | tile_type << 8 | tileh;
 }
 
 /** This is the position of the tile relative to the northernmost tile of the industry.
Index: src/newgrf_house.cpp
===================================================================
--- src/newgrf_house.cpp	(revision 11039)
+++ src/newgrf_house.cpp	(working copy)
@@ -185,7 +185,9 @@
 	tile = GetNearbyTile(parameter, tile);
 	tile_type = GetTerrainType(tile) << 2 | (IsTileType(tile, MP_WATER) ? 1 : 0) << 1;
 
-	return GetTileType(tile) << 24 | (TileHeight(tile) * 8) << 16 | tile_type << 8 | GetTileSlope(tile, NULL);
+	uint z;
+	Slope tileh = GetTileSlope(tile, &z);
+	return GetTileType(tile) << 24 | z << 16 | tile_type << 8 | tileh;
 }
 
 /**
