diff --git a/.gitignore b/.gitignore
index 5aa8cb7..e1743b5 100644
--- a/.gitignore
+++ b/.gitignore
@@ -36,6 +36,7 @@ projects/*.vcproj.*.user
 projects/*.vcxproj.user
 projects/*.VC.db
 projects/*.VC.opendb
+projects/.vs
 src/rev.cpp
 src/os/windows/ottdres.rc
 
diff --git a/.hgignore b/.hgignore
index debab32..426a45c 100644
--- a/.hgignore
+++ b/.hgignore
@@ -25,5 +25,6 @@ projects/*.sdf
 projects/*.opensdf
 projects/*.vcproj.*.user
 projects/*.vcxproj.user
+projects/.vs
 src/rev.cpp
 src/os/windows/ottdres.rc
