From 8f38bf3e1b96da231a33254d9fe2723f2bfbead3 Mon Sep 17 00:00:00 2001
From: Michael Lutz <michi@icosahedron.de>
Date: Mon, 5 Jan 2009 17:29:58 +0100
Subject: -Fix: A train reversing in a non-PBS block can't be stuck.

---
 src/train_cmd.cpp |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/src/train_cmd.cpp b/src/train_cmd.cpp
index e71ef9e..ea4a806 100644
--- a/src/train_cmd.cpp
+++ b/src/train_cmd.cpp
@@ -1941,6 +1941,10 @@ static void ReverseTrainDirection(Vehicle *v)
 			/* Do not wait for a way out when we're still loading */
 			MarkTrainAsStuck(v);
 		}
+	} else if (HasBit(v->u.rail.flags, VRF_TRAIN_STUCK)) {
+		/* A train not inside a PBS block can't be stuck. */
+		ClrBit(v->u.rail.flags, VRF_TRAIN_STUCK);
+		v->load_unload_time_rem = 0;
 	}
 }
 
-- 
1.6.0.4

