diff --git a/findversion.sh b/findversion.sh
index c792ec8..b08c44d 100755
--- a/findversion.sh
+++ b/findversion.sh
@@ -93,7 +93,12 @@ elif [ -d "$ROOT_DIR/.git" ]; then
 	HASH=`LC_ALL=C git rev-parse --verify HEAD 2>/dev/null`
 	REV="g`echo $HASH | cut -c1-8`"
 	BRANCH=`git branch|grep '[*]' | sed 's@\* @@;s@^master$@@'`
-	REV_NR=`LC_ALL=C git log --pretty=format:%s "$SRC_DIR" | grep "^(svn r[0-9]*)" | head -n 1 | sed "s@.*(svn r\([0-9]*\)).*@\1@"`
+        if [ -d "$ROOT_DIR/.git/svn" ]; then
+                # Cloned directly via 'git svn'
+	        REV_NR=`LC_ALL=C git log "$SRC_DIR" | grep 'git-svn-id: http://svn.openttd.org/trunk@\([0-9]*\)\s.*' | head -n 1 | sed "s@.*trunk\@\([0-9]*\).*@\1@"`
+        else
+                REV_NR=`LC_ALL=C git log --pretty=format:%s "$SRC_DIR" | grep "^(svn r[0-9]*)" | head -n 1 | sed "s@.*(svn r\([0-9]*\)).*@\1@"`
+        fi
 elif [ -d "$ROOT_DIR/.hg" ]; then
 	# We are a hg checkout
 	if [ -n "`hg status \"$SRC_DIR\" | grep -v '^?'`" ]; then

