Index: src/tunnelbridge_cmd.cpp
===================================================================
--- src/tunnelbridge_cmd.cpp	(revision 20896)
+++ src/tunnelbridge_cmd.cpp	(working copy)
@@ -874,6 +874,7 @@
 	int front_height = ti->z;
 	const byte *p = tileh_bits[axis];
 	bool draw_odd_pillar = false;
+	bool cantilever_pillars_fix = false;
 
 	switch(type) {
 		case 0: case 1: case 2: case 9: // wooden, girder and concrete bridges
@@ -886,6 +887,9 @@
 				p = tileh_bits[HasBit(image, 0) * 2 + axis];
 			}
 			break;
+		case 6: case 7: case 8: // cantilever bridges
+			cantilever_pillars_fix = true;
+			break;
 
 		default:;
 	}
@@ -947,6 +951,20 @@
 			if (!(ti->tileh & p[2])) front_height += TILE_HEIGHT;
 			if (!(ti->tileh & p[3]))  back_height += TILE_HEIGHT;
 		}
+
+		if (cantilever_pillars_fix) {
+			/* Draw part directly under the bridge at default position */
+			DRAW_PILLAR(image, x, y, z_bridge, 0);
+			z_bridge -= TILE_HEIGHT;
+
+			if (axis == AXIS_X) {
+				x -= 3;
+				x_back -= 3;
+			}	else {
+				y -= 3;
+				y_back -= 3;
+			}
+		}
 	}
 
 	for (int cur_z = z_bridge; cur_z >= front_height; cur_z -= TILE_HEIGHT) {
