Index: src/highscore_gui.cpp
===================================================================
--- src/highscore_gui.cpp	(revision 22927)
+++ src/highscore_gui.cpp	(working copy)
@@ -34,6 +34,7 @@
 	EndGameHighScoreBaseWindow(const WindowDesc *desc) : Window()
 	{
 		this->InitNested(desc);
+		this->flags4 &= ~WF_WHITE_BORDER_MASK;
 		ResizeWindow(this, _screen.width - this->width, _screen.height - this->height);
 	}
 
Index: src/main_gui.cpp
===================================================================
--- src/main_gui.cpp	(revision 22927)
+++ src/main_gui.cpp	(working copy)
@@ -244,6 +244,7 @@
 	MainWindow() : Window()
 	{
 		this->InitNested(&_main_window_desc, 0);
+		this->flags4 &= ~WF_WHITE_BORDER_MASK;
 		ResizeWindow(this, _screen.width, _screen.height);
 
 		NWidgetViewport *nvp = this->GetWidget<NWidgetViewport>(MW_VIEWPORT);
