diff -r 44d201d223f9 src/roadveh_cmd.cpp
--- a/src/roadveh_cmd.cpp	Tue Nov 04 18:44:15 2008 +0000
+++ b/src/roadveh_cmd.cpp	Wed Nov 05 04:42:03 2008 +0100
@@ -1922,7 +1922,7 @@
 	}
 
 	/* update destination */
-	if (!(this->vehstatus & VS_STOPPED) && this->current_order.IsType(OT_GOTO_STATION) && this->u.road.slot == NULL && !(this->vehstatus & VS_CRASHED)) {
+	if (!(this->vehstatus & VS_STOPPED) && this->current_order.IsType(OT_GOTO_STATION) && !(this->current_order.GetNonStopType() & ONSF_NO_STOP_AT_DESTINATION_STATION) && this->u.road.slot == NULL && !(this->vehstatus & VS_CRASHED)) {
 		Station *st = GetStation(this->current_order.GetDestination());
 		RoadStop *rs = st->GetPrimaryRoadStop(this);
 		RoadStop *best = NULL;
