diff -r 07a5d3127cf5 src/lang/english.txt
--- a/src/lang/english.txt	Sun Nov 23 14:22:48 2008 +0100
+++ b/src/lang/english.txt	Sun Nov 23 15:35:07 2008 +0100
@@ -1214,6 +1214,7 @@
 
 STR_CONFIG_PATCHES_TITLE_INTERFACE                              :{ORANGE}Interface
 STR_CONFIG_PATCHES_TITLE_INTERFACE_DISPLAY_OPTIONS              :{ORANGE}Display options
+STR_CONFIG_PATCHES_TITLE_INTERFACE_INTERACTION                  :{ORANGE}Interaction
 
 STR_CONFIG_PATCHES_TITLE_CONSTRUCTION                           :{ORANGE}Construction
 STR_CONFIG_PATCHES_TITLE_CONSTRUCTION_SIGNALS                   :{ORANGE}Signals
@@ -1223,8 +1224,12 @@
 STR_CONFIG_PATCHES_TITLE_VEHICLES_SERVICING                     :{ORANGE}Servicing
 
 STR_CONFIG_PATCHES_TITLE_STATIONS                               :{ORANGE}Stations
+STR_CONFIG_PATCHES_TITLE_STATIONS_AIRPORT                       :{ORANGE}Airports
 STR_CONFIG_PATCHES_TITLE_ECONOMY                                :{ORANGE}Economy
+STR_CONFIG_PATCHES_TITLE_TOWNS                                  :{ORANGE}Towns
+
 STR_CONFIG_PATCHES_TITLE_COMPETITORS                            :{ORANGE}Competitors
+STR_CONFIG_PATCHES_TITLE_COMPETITORS_AI                         :{ORANGE}Computer players
 
 STR_CONFIG_PATCHES_DISABLED                                     :disabled
 STR_CONFIG_PATCHES_INT32                                        :{NUM}
diff -r 07a5d3127cf5 src/settings_gui.cpp
--- a/src/settings_gui.cpp	Sun Nov 23 14:22:48 2008 +0100
+++ b/src/settings_gui.cpp	Sun Nov 23 15:35:07 2008 +0100
@@ -1000,19 +1000,11 @@
 };
 static PatchPage _patches_ui_display_options_page = { _patches_ui_display_options, lengthof(_patches_ui_display_options)};
 
-
-static PatchEntry _patches_ui[] = {
-	PatchEntry(&_patches_ui_display_options_page, STR_CONFIG_PATCHES_TITLE_INTERFACE_DISPLAY_OPTIONS),
-	PatchEntry("gui.show_finances"),
+static PatchEntry _patches_ui_interaction[]= {
 	PatchEntry("gui.autoscroll"),
 	PatchEntry("gui.reverse_scroll"),
 	PatchEntry("gui.smooth_scroll"),
-	PatchEntry("gui.errmsg_duration"),
-	PatchEntry("gui.toolbar_pos"),
 	PatchEntry("gui.window_snap_radius"),
-	PatchEntry("gui.link_terraform_toolbar"),
-	PatchEntry("gui.liveries"),
-	PatchEntry("gui.prefer_teamchat"),
 	/* While the horizontal scrollwheel scrolling is written as general code, only
 	 *  the cocoa (OSX) driver generates input for it.
 	 *  Since it's also able to completely disable the scrollwheel will we display it on all platforms anyway */
@@ -1022,6 +1014,17 @@
 	/* We might need to emulate a right mouse button on mac */
 	PatchEntry("gui.right_mouse_btn_emulation"),
 #endif
+};
+static PatchPage _patches_ui_interaction_page = {_patches_ui_interaction, lengthof(_patches_ui_interaction)};
+
+static PatchEntry _patches_ui[] = {
+	PatchEntry(&_patches_ui_display_options_page, STR_CONFIG_PATCHES_TITLE_INTERFACE_DISPLAY_OPTIONS),
+	PatchEntry(&_patches_ui_interaction_page, STR_CONFIG_PATCHES_TITLE_INTERFACE_INTERACTION),
+	PatchEntry("gui.show_finances"),
+	PatchEntry("gui.errmsg_duration"),
+	PatchEntry("gui.toolbar_pos"),
+	PatchEntry("gui.link_terraform_toolbar"),
+	PatchEntry("gui.liveries"),
 	PatchEntry("gui.pause_on_newgame"),
 	PatchEntry("gui.advanced_vehicle_list"),
 	PatchEntry("gui.timetable_in_ticks"),
@@ -1050,28 +1053,46 @@
 	PatchEntry("construction.autoslope"),
 	PatchEntry("construction.extra_dynamite"),
 	PatchEntry("construction.longbridges"),
-	PatchEntry("station.always_small_airport"),
 	PatchEntry("game_creation.oil_refinery_limit"),
 };
 static PatchPage _patches_construction_page = {_patches_construction, lengthof(_patches_construction)};
 
 /* == STATIONS PAGE == */
 
+static PatchEntry _patches_stations_airport[] = {
+	PatchEntry("station.always_small_airport"),
+	PatchEntry("order.serviceathelipad"),
+	PatchEntry("economy.station_noise_level"),
+};
+static PatchPage _patches_stations_airport_page = {_patches_stations_airport, lengthof(_patches_stations_airport)};
+
 static PatchEntry _patches_stations[] = {
+	PatchEntry(&_patches_stations_airport_page, STR_CONFIG_PATCHES_TITLE_STATIONS_AIRPORT),
 	PatchEntry("station.join_stations"),
 	PatchEntry("order.improved_load"),
 	PatchEntry("order.selectgoods"),
 	PatchEntry("gui.new_nonstop"),
 	PatchEntry("station.nonuniform_stations"),
 	PatchEntry("station.station_spread"),
-	PatchEntry("order.serviceathelipad"),
 	PatchEntry("station.modified_catchment"),
 	PatchEntry("order.gradual_loading"),
 	PatchEntry("construction.road_stop_on_town_road"),
 	PatchEntry("station.adjacent_stations"),
-	PatchEntry("economy.station_noise_level"),
 };
 static PatchPage _patches_stations_page = {_patches_stations, lengthof(_patches_stations)};
+
+/* == TOWNS PAGE == */
+
+static PatchEntry _patches_town[] = {
+	PatchEntry("economy.bribe"),
+	PatchEntry("economy.exclusive_rights"),
+	PatchEntry("economy.town_layout"),
+	PatchEntry("economy.mod_road_rebuild"),
+	PatchEntry("economy.town_growth_rate"),
+	PatchEntry("economy.larger_towns"),
+	PatchEntry("economy.initial_city_size"),
+};
+static PatchPage _patches_town_page = {_patches_town, lengthof(_patches_town)};
 
 /* == ECONOMY PAGE == */
 
@@ -1080,18 +1101,9 @@
 	PatchEntry("construction.raw_industry_construction"),
 	PatchEntry("economy.multiple_industry_per_town"),
 	PatchEntry("economy.same_industry_close"),
-	PatchEntry("economy.bribe"),
-	PatchEntry("economy.exclusive_rights"),
-	PatchEntry("economy.give_money"),
 	PatchEntry("gui.colored_news_year"),
 	PatchEntry("gui.ending_year"),
 	PatchEntry("economy.smooth_economy"),
-	PatchEntry("economy.allow_shares"),
-	PatchEntry("economy.town_layout"),
-	PatchEntry("economy.mod_road_rebuild"),
-	PatchEntry("economy.town_growth_rate"),
-	PatchEntry("economy.larger_towns"),
-	PatchEntry("economy.initial_city_size"),
 };
 static PatchPage _patches_economy_page = {_patches_economy, lengthof(_patches_economy)};
 
@@ -1106,6 +1118,14 @@
 	PatchEntry("ai.ai_disable_veh_ship"),
 };
 static PatchPage _patches_ai_page = {_patches_ai, lengthof(_patches_ai)};
+
+static PatchEntry _patches_competitors[] = {
+	PatchEntry(&_patches_ai_page, STR_CONFIG_PATCHES_TITLE_COMPETITORS_AI),
+	PatchEntry("economy.allow_shares"),
+	PatchEntry("economy.give_money"),
+	PatchEntry("gui.prefer_teamchat"),
+};
+static PatchPage _patches_competitors_page = {_patches_competitors, lengthof(_patches_competitors)};
 
 /* == VEHICLES PAGE == */
 
@@ -1161,8 +1181,9 @@
 	PatchEntry(&_patches_construction_page, STR_CONFIG_PATCHES_TITLE_CONSTRUCTION),
 	PatchEntry(&_patches_vehicles_page,     STR_CONFIG_PATCHES_TITLE_VEHICLES),
 	PatchEntry(&_patches_stations_page,     STR_CONFIG_PATCHES_TITLE_STATIONS),
+	PatchEntry(&_patches_town_page,         STR_CONFIG_PATCHES_TITLE_TOWNS),
 	PatchEntry(&_patches_economy_page,      STR_CONFIG_PATCHES_TITLE_ECONOMY),
-	PatchEntry(&_patches_ai_page,           STR_CONFIG_PATCHES_TITLE_COMPETITORS),
+	PatchEntry(&_patches_competitors_page,  STR_CONFIG_PATCHES_TITLE_COMPETITORS),
 };
 static PatchPage _patches_main_page = {_patches_main, lengthof(_patches_main)};
 
