Index: src/tunnelbridge_cmd.cpp
===================================================================
--- src/tunnelbridge_cmd.cpp	(revision 20956)
+++ src/tunnelbridge_cmd.cpp	(working copy)
@@ -913,7 +913,7 @@
 	if (z_front_south < z_front) DrawPillar(psid, x, y, bottom_z, w, h, &half_pillar_sub_sprite[axis][1]);
 
 	/* Draw back pillars, skip top two parts, which are hidden by the bridge */
-	int z_bridge_back = z_bridge - 2 * (int)TILE_HEIGHT;
+	int z_bridge_back = (HasBit(flags, 1) ? z_bridge : z_bridge - 2 * (int)TILE_HEIGHT);
 	if (drawfarpillar && (z_back_north <= z_bridge_back || z_back_south <= z_bridge_back)) {
 		bottom_z = DrawPillarColumn(z_back, z_bridge_back, psid, x_back, y_back, w, h);
 		if (z_back_north < z_back) DrawPillar(psid, x_back, y_back, bottom_z, w, h, &half_pillar_sub_sprite[axis][0]);
