diff -r da7d3c53747d src/industry_cmd.cpp
--- a/src/industry_cmd.cpp	Fri Oct 17 17:47:42 2008 +0000
+++ b/src/industry_cmd.cpp	Sat Oct 18 01:46:18 2008 +0200
@@ -2089,6 +2089,9 @@
 				case 0xE:                         // increment production
 					increment = res == 0x0D ? -1 : 1;
 					break;
+				case 0xF:                         // Set production to highest bits of register 0x100
+					i->prod_level = Clamp(GB(GetRegister(0x100), 16, 16), PRODLEVEL_MINIMUM, PRODLEVEL_MAXIMUM);
+					break;
 			}
 		}
 	}
