Index: src/network/core/address.cpp
===================================================================
--- src/network/core/address.cpp	(wersja 23508)
+++ src/network/core/address.cpp	(kopia robocza)
@@ -81,7 +81,7 @@
 	if (this->GetAddress()->ss_family == AF_INET6) buffer = strecpy(buffer, "[", last);
 	buffer = strecpy(buffer, this->GetHostname(), last);
 	if (this->GetAddress()->ss_family == AF_INET6) buffer = strecpy(buffer, "]", last);
-	buffer += seprintf(buffer, last, ":%d", this->GetPort());
+	buffer += seprintf(buffer, last, ":%d", (int)this->GetPort());
 
 	if (with_family) {
 		char family;
