provide {ISODATE} string code

diff -r c733b3e4f97b src/lang/english.txt
--- a/src/lang/english.txt	Thu Nov 20 18:44:13 2008 +0000
+++ b/src/lang/english.txt	Fri Nov 21 14:47:31 2008 +0100
@@ -3490,6 +3490,7 @@ STR_DATE_TINY                           
 STR_DATE_TINY                                                   :{RAW_STRING}-{RAW_STRING}-{NUM}
 STR_DATE_SHORT                                                  :{STRING} {NUM}
 STR_DATE_LONG                                                   :{STRING} {STRING} {NUM}
+STR_DATE_ISO                                                    :{2:NUM}-{1:RAW_STRING}-{0:RAW_STRING}
 
 ########
 
diff -r c733b3e4f97b src/strgen/strgen.cpp
--- a/src/strgen/strgen.cpp	Thu Nov 20 18:44:13 2008 +0000
+++ b/src/strgen/strgen.cpp	Fri Nov 21 14:47:31 2008 +0100
@@ -489,6 +489,7 @@ static const CmdStruct _cmd_structs[] = 
 	{"DATE_TINY",  EmitSingleChar, SCC_DATE_TINY, 1, 0},
 	{"DATE_SHORT", EmitSingleChar, SCC_DATE_SHORT, 1, 0},
 	{"DATE_LONG",  EmitSingleChar, SCC_DATE_LONG, 1, 0},
+	{"ISO_DATE",   EmitSingleChar, SCC_DATE_ISO, 1, 0},
 
 	{"SKIP", EmitSingleChar, SCC_SKIP, 1, 0},
 
diff -r c733b3e4f97b src/strings.cpp
--- a/src/strings.cpp	Thu Nov 20 18:44:13 2008 +0000
+++ b/src/strings.cpp	Fri Nov 21 14:47:31 2008 +0100
@@ -292,7 +292,7 @@ static char *FormatMonthAndYear(char *bu
 	return FormatString(buff, GetStringPtr(STR_DATE_SHORT), args, 0, last);
 }
 
-static char *FormatTinyDate(char *buff, Date date, const char* last)
+static char *FormatTinyOrISODate(char *buff, Date date, StringID str, const char* last)
 {
 	YearMonthDay ymd;
 	ConvertDateToYMD(date, &ymd);
@@ -304,7 +304,7 @@ static char *FormatTinyDate(char *buff, 
 	snprintf(month, lengthof(month), "%02i", ymd.month + 1);
 
 	int64 args[3] = { (int64)(size_t)day, (int64)(size_t)month, ymd.year };
-	return FormatString(buff, GetStringPtr(STR_DATE_TINY), args, 0, last);
+	return FormatString(buff, GetStringPtr(str), args, 0, last);
 }
 
 static char *FormatGenericCurrency(char *buff, const CurrencySpec *spec, Money number, bool compact, const char* last)
@@ -722,7 +722,12 @@ static char* FormatString(char* buff, co
 			}
 
 			case SCC_DATE_TINY: { // {DATE_TINY}
-				buff = FormatTinyDate(buff, GetInt32(&argv), last);
+				buff = FormatTinyOrISODate(buff, GetInt32(&argv), STR_DATE_TINY, last);
+				break;
+			}
+
+			case SCC_DATE_ISO: { // {ISO_DATE}
+				buff = FormatTinyOrISODate(buff, GetInt32(&argv), STR_DATE_ISO, last);
 				break;
 			}
 
diff -r c733b3e4f97b src/table/control_codes.h
--- a/src/table/control_codes.h	Thu Nov 20 18:44:13 2008 +0000
+++ b/src/table/control_codes.h	Fri Nov 21 14:47:31 2008 +0100
@@ -53,6 +53,7 @@ enum StringControlCode {
 	SCC_DATE_TINY,
 	SCC_DATE_SHORT,
 	SCC_DATE_LONG,
+	SCC_DATE_ISO,
 
 	SCC_STRING1,
 	SCC_STRING2,
