From c66afb49d0b52a869e9fd4b779f472e8641e1f34 Mon Sep 17 00:00:00 2001
From: Michael Lutz <michi@icosahedron.de>
Date: Fri, 22 Aug 2008 19:06:45 +0200
Subject: -Fix (r13935) [FS#2247]: Signal state for PBS signals was not updated when cycling the signal side.

---
 src/rail_cmd.cpp |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/src/rail_cmd.cpp b/src/rail_cmd.cpp
index 2b5227c..e324c73 100644
--- a/src/rail_cmd.cpp
+++ b/src/rail_cmd.cpp
@@ -931,6 +931,8 @@ CommandCost CmdBuildSingleSignal(TileIndex tile, uint32 flags, uint32 p1, uint32
 				} else {
 					/* cycle the signal side: both -> left -> right -> both -> ... */
 					CycleSignalSide(tile, track);
+					/* Query current signal type so the check for PBS signals below works. */
+					sigtype = GetSignalType(tile, track);
 				}
 			}
 		} else {
@@ -942,6 +944,7 @@ CommandCost CmdBuildSingleSignal(TileIndex tile, uint32 flags, uint32 p1, uint32
 		}
 
 		if (IsPbsSignal(sigtype)) {
+			/* PBS signals should show red unless they are on a reservation. */
 			uint mask = GetPresentSignals(tile) & SignalOnTrack(track);
 			SetSignalStates(tile, (GetSignalStates(tile) & ~mask) | ((HasBit(GetTrackReservation(tile), track) ? (uint)-1 : 0) & mask));
 		}
-- 
1.5.6.4

