Index: src/video/cocoa/event.mm
===================================================================
--- src/video/cocoa/event.mm	(revision 25997)
+++ src/video/cocoa/event.mm	(working copy)
@@ -293,7 +293,7 @@
 	if (down) {
 		uint32 pressed_key = QZ_MapKey(keycode);
 		/* Don't handle normal characters if an edit box has the focus. */
-		if (!EditBoxInGlobalFocus() || (!IsInsideMM(pressed_key, 'A', 'Z' + 1) && !IsInsideMM(pressed_key, '0', '9' + 1))) {
+		if (!EditBoxInGlobalFocus() || (!IsInsideMM(pressed_key & ~(WKC_SHIFT | WKC_ALT), 'A', 'Z' + 1) && !IsInsideMM(pressed_key & ~(WKC_SHIFT | WKC_ALT), '0', '9' + 1))) {
 			HandleKeypress(pressed_key, unicode);
 		}
 		DEBUG(driver, 2, "cocoa_v: QZ_KeyEvent: %x (%x), down, mapping: %x", keycode, unicode, pressed_key);
