diff -r 7e8b1e8069d4 src/lang/english.txt
--- a/src/lang/english.txt	Sun Nov 23 14:04:33 2008 +0100
+++ b/src/lang/english.txt	Sun Nov 23 14:22:41 2008 +0100
@@ -1213,10 +1213,15 @@
 STR_CONFIG_MODIFIED_ROAD_REBUILD                                :{LTBLUE}Remove absurd road-elements during the road construction
 
 STR_CONFIG_PATCHES_TITLE_INTERFACE                              :{ORANGE}Interface
+STR_CONFIG_PATCHES_TITLE_INTERFACE_DISPLAY_OPTIONS              :{ORANGE}Display options
+
 STR_CONFIG_PATCHES_TITLE_CONSTRUCTION                           :{ORANGE}Construction
 STR_CONFIG_PATCHES_TITLE_CONSTRUCTION_SIGNALS                   :{ORANGE}Signals
 
 STR_CONFIG_PATCHES_TITLE_VEHICLES                               :{ORANGE}Vehicles
+STR_CONFIG_PATCHES_TITLE_VEHICLES_AUTORENEW                     :{ORANGE}Autorenew
+STR_CONFIG_PATCHES_TITLE_VEHICLES_SERVICING                     :{ORANGE}Servicing
+
 STR_CONFIG_PATCHES_TITLE_STATIONS                               :{ORANGE}Stations
 STR_CONFIG_PATCHES_TITLE_ECONOMY                                :{ORANGE}Economy
 STR_CONFIG_PATCHES_TITLE_COMPETITORS                            :{ORANGE}Competitors
diff -r 7e8b1e8069d4 src/settings_gui.cpp
--- a/src/settings_gui.cpp	Sun Nov 23 14:04:33 2008 +0100
+++ b/src/settings_gui.cpp	Sun Nov 23 14:22:42 2008 +0100
@@ -990,18 +990,26 @@
 
 /* == INTERFACE PAGE == */
 
-static PatchEntry _patches_ui[] = {
+static PatchEntry _patches_ui_display_options[] = {
 	PatchEntry("gui.vehicle_speed"),
 	PatchEntry("gui.status_long_date"),
+	PatchEntry("gui.population_in_label"),
+	PatchEntry("gui.measure_tooltip"),
+	PatchEntry("gui.loading_indicators"),
+	PatchEntry("gui.show_track_reservation"),
+};
+static PatchPage _patches_ui_display_options_page = { _patches_ui_display_options, lengthof(_patches_ui_display_options)};
+
+
+static PatchEntry _patches_ui[] = {
+	PatchEntry(&_patches_ui_display_options_page, STR_CONFIG_PATCHES_TITLE_INTERFACE_DISPLAY_OPTIONS),
 	PatchEntry("gui.show_finances"),
 	PatchEntry("gui.autoscroll"),
 	PatchEntry("gui.reverse_scroll"),
 	PatchEntry("gui.smooth_scroll"),
 	PatchEntry("gui.errmsg_duration"),
 	PatchEntry("gui.toolbar_pos"),
-	PatchEntry("gui.measure_tooltip"),
 	PatchEntry("gui.window_snap_radius"),
-	PatchEntry("gui.population_in_label"),
 	PatchEntry("gui.link_terraform_toolbar"),
 	PatchEntry("gui.liveries"),
 	PatchEntry("gui.prefer_teamchat"),
@@ -1016,17 +1024,16 @@
 #endif
 	PatchEntry("gui.pause_on_newgame"),
 	PatchEntry("gui.advanced_vehicle_list"),
-	PatchEntry("gui.loading_indicators"),
 	PatchEntry("gui.timetable_in_ticks"),
 	PatchEntry("gui.default_rail_type"),
 	PatchEntry("gui.always_build_infrastructure"),
-	PatchEntry("gui.show_track_reservation"),
 };
 static PatchPage _patches_ui_page = {_patches_ui, lengthof(_patches_ui)};
 
 /* == CONSTRUCTION PAGE == */
 
 static PatchEntry _patches_construction_signals[] = {
+	PatchEntry("construction.signal_side"),
 	PatchEntry("gui.enable_signal_gui"),
 	PatchEntry("gui.drag_signals_density"),
 	PatchEntry("gui.semaphore_build_before"),
@@ -1043,7 +1050,6 @@
 	PatchEntry("construction.autoslope"),
 	PatchEntry("construction.extra_dynamite"),
 	PatchEntry("construction.longbridges"),
-	PatchEntry("construction.signal_side"),
 	PatchEntry("station.always_small_airport"),
 	PatchEntry("game_creation.oil_refinery_limit"),
 };
@@ -1103,7 +1109,26 @@
 
 /* == VEHICLES PAGE == */
 
+static PatchEntry _patches_vehicles_autorenew[] = {
+	PatchEntry("gui.autorenew"),
+	PatchEntry("gui.autorenew_months"),
+	PatchEntry("gui.autorenew_money"),
+};
+static PatchPage _patches_vehicles_autorenew_page = {_patches_vehicles_autorenew, lengthof(_patches_vehicles_autorenew)};
+
+static PatchEntry _patches_vehicles_servicing[] = {
+	PatchEntry("vehicle.servint_ispercent"),
+	PatchEntry("vehicle.servint_trains"),
+	PatchEntry("vehicle.servint_roadveh"),
+	PatchEntry("vehicle.servint_ships"),
+	PatchEntry("vehicle.servint_aircraft"),
+	PatchEntry("order.no_servicing_if_no_breakdowns"),
+};
+static PatchPage _patches_vehicles_servicing_page = {_patches_vehicles_servicing, lengthof(_patches_vehicles_servicing)};
+
 static PatchEntry _patches_vehicles[] = {
+	PatchEntry(&_patches_vehicles_autorenew_page, STR_CONFIG_PATCHES_TITLE_VEHICLES_AUTORENEW),
+	PatchEntry(&_patches_vehicles_servicing_page, STR_CONFIG_PATCHES_TITLE_VEHICLES_SERVICING),
 	PatchEntry("vehicle.realistic_acceleration"),
 	PatchEntry("pf.forbid_90_deg"),
 	PatchEntry("vehicle.mammoth_trains"),
@@ -1116,19 +1141,10 @@
 	PatchEntry("gui.order_review_system"),
 	PatchEntry("vehicle.never_expire_vehicles"),
 	PatchEntry("gui.lost_train_warn"),
-	PatchEntry("gui.autorenew"),
-	PatchEntry("gui.autorenew_months"),
-	PatchEntry("gui.autorenew_money"),
 	PatchEntry("vehicle.max_trains"),
 	PatchEntry("vehicle.max_roadveh"),
 	PatchEntry("vehicle.max_aircraft"),
 	PatchEntry("vehicle.max_ships"),
-	PatchEntry("vehicle.servint_ispercent"),
-	PatchEntry("vehicle.servint_trains"),
-	PatchEntry("vehicle.servint_roadveh"),
-	PatchEntry("vehicle.servint_ships"),
-	PatchEntry("vehicle.servint_aircraft"),
-	PatchEntry("order.no_servicing_if_no_breakdowns"),
 	PatchEntry("vehicle.wagon_speed_limits"),
 	PatchEntry("vehicle.disable_elrails"),
 	PatchEntry("vehicle.freight_trains"),
