diff -r 256e56822a58 src/group_cmd.cpp
--- a/src/group_cmd.cpp	Sun Mar 14 14:01:53 2010 +0000
+++ b/src/group_cmd.cpp	Sun Mar 14 15:18:05 2010 +0100
@@ -150,12 +150,12 @@
 	return CommandCost();
 }
 
-static bool IsUniqueGroupName(const char *name)
+static bool IsUniqueGroupNameForVehicleType(const char *name, const VehicleTypeByte vehicle_type)
 {
 	const Group *g;
 
 	FOR_ALL_GROUPS(g) {
-		if (g->name != NULL && strcmp(g->name, name) == 0) return false;
+		if (g->name != NULL && strcmp(g->name, name) == 0 && g->vehicle_type == vehicle_type) return false;
 	}
 
 	return true;
@@ -180,7 +180,7 @@
 
 	if (!reset) {
 		if (strlen(text) >= MAX_LENGTH_GROUP_NAME_BYTES) return CMD_ERROR;
-		if (!IsUniqueGroupName(text)) return_cmd_error(STR_ERROR_NAME_MUST_BE_UNIQUE);
+		if (!IsUniqueGroupNameForVehicleType(text, g->vehicle_type)) return_cmd_error(STR_ERROR_NAME_MUST_BE_UNIQUE);
 	}
 
 	if (flags & DC_EXEC) {
