Index: src/tunnelbridge_cmd.cpp
===================================================================
--- src/tunnelbridge_cmd.cpp	(revision 17491)
+++ src/tunnelbridge_cmd.cpp	(working copy)
@@ -39,6 +39,8 @@
 #include "cheat_type.h"
 #include "elrail_func.h"
 #include "landscape_type.h"
+#include "window_func.h"
+#include "gui.h"
 
 #include "table/sprites.h"
 #include "table/strings.h"
@@ -525,6 +527,18 @@
 	cost.AddCost(_price.build_tunnel);
 	cost.AddCost(ret);
 
+	/* Show end of tunnel in ViewPort if Ctrl pressed */
+
+
+	if (_ctrl_pressed && IsLocalCompany()) {
+		int i = 0;
+		/* find next free window number for extra viewport */
+		while (FindWindowById(WC_EXTRA_VIEW_PORT, i) != NULL) i++;
+		i--;
+		Window *w = FindWindowById(WC_EXTRA_VIEW_PORT, i);
+		if (w != NULL) ScrollWindowTo((TileX(end_tile) * TILE_SIZE + TILE_SIZE / 2),(TileY(end_tile) * TILE_SIZE + TILE_SIZE / 2), end_z, w);
+	}
+
 	/* if the command fails from here on we want the end tile to be highlighted */
 	_build_tunnel_endtile = end_tile;
 
