Index: src/openttd.cpp
===================================================================
--- src/openttd.cpp	(revision 27603)
+++ src/openttd.cpp	(working copy)
@@ -366,6 +366,7 @@
 		_settings_game.ai_config[c] = NULL;
 		if (_settings_newgame.ai_config[c] != NULL) {
 			_settings_game.ai_config[c] = new AIConfig(_settings_newgame.ai_config[c]);
+			if (!AIConfig::GetConfig(c, AIConfig::SSS_FORCE_GAME)->HasScript()) AIConfig::GetConfig(c, AIConfig::SSS_FORCE_GAME)->Change(NULL);
 		}
 	}
 	_settings_game.game_config = NULL;
