Index: saveload.c
===================================================================
--- saveload.c	(revision 4531)
+++ saveload.c	(working copy)
@@ -1537,6 +1537,7 @@
 		SlSaveChunks();
 		SlWriteFill(); // flush the save buffer
 
+		_ts.saveinprogress = true;
 		if (_network_server ||
 					(save_thread = OTTDCreateThread(&SaveFileToDisk, (void*)"")) == NULL) {
 			DEBUG(misc, 1) ("[Sl] Cannot create savegame thread, reverting to single-threaded mode...");
