diff -r 531716190738 config.lib
--- a/config.lib	Tue Oct 14 12:07:14 2008 +0000
+++ b/config.lib	Tue Oct 14 14:32:50 2008 +0200
@@ -982,7 +982,7 @@ make_cflags_and_ldflags() {
 
 		# Each debug level reduces the optimization by a bit
 		if [ $enable_debug -ge 1 ]; then
-			CFLAGS="$CFLAGS -g -D_DEBUG -D_FORTIFY_SOURCE=2"
+			CFLAGS="$CFLAGS -g -D_DEBUG=$enable_debug -D_FORTIFY_SOURCE=2"
 			if [ "$os" = "PSP" ]; then
 				CFLAGS="$CFLAGS -G0"
 			fi
diff -r 531716190738 src/map_func.h
--- a/src/map_func.h	Tue Oct 14 12:07:14 2008 +0000
+++ b/src/map_func.h	Tue Oct 14 14:32:50 2008 +0200
@@ -211,7 +211,7 @@ static inline TileIndexDiff ToTileIndexD
 }
 
 
-#ifndef _DEBUG
+#if !defined(_DEBUG) || _DEBUG <= 1
 	/**
 	 * Adds to tiles together.
 	 *
