Index: station_cmd.cpp
===================================================================
--- station_cmd.cpp	(revision 10925)
+++ station_cmd.cpp	(working copy)
@@ -2761,6 +2761,13 @@
 {
 	Station* st = GetStationByTile(tile);
 
+	std::list<Vehicle *> loading_vehicles_copy = st->loading_vehicles;
+	std::list<Vehicle *>::iterator iter;
+	for (iter = loading_vehicles_copy.begin(); iter != loading_vehicles_copy.end(); ++iter) {
+		Vehicle *v = *iter;
+		v->LeaveStation();
+	}
+
 	MakeWater(tile);
 
 	st->dock_tile = 0;
