diff -r 748dc4c946eb src/industry_cmd.cpp
--- a/src/industry_cmd.cpp	Sun Jul 17 17:45:15 2011 +0000
+++ b/src/industry_cmd.cpp	Tue Jul 19 16:14:48 2011 +0200
@@ -1091,7 +1091,7 @@
 		if ((indbehav & INDUSTRYBEH_CUT_TREES) != 0) {
 			bool cut = ((i->counter % INDUSTRY_CUT_TREE_TICKS) == 0);
 			if (HasBit(indsp->callback_mask, CBM_IND_SPECIAL_EFFECT)) {
-				cut = (GetIndustryCallback(CBID_INDUSTRY_SPECIAL_EFFECT, 0, 1, i, i->type, i->location.tile) != 0);
+				cut = (GetIndustryCallback(CBID_INDUSTRY_SPECIAL_EFFECT, Random(), 1, i, i->type, i->location.tile) != 0);
 			}
 
 			if (cut) ChopLumberMillTrees(i);
