Index: src/build_vehicle_gui.cpp
===================================================================
--- src/build_vehicle_gui.cpp	(revision 20537)
+++ src/build_vehicle_gui.cpp	(working copy)
@@ -1087,7 +1087,7 @@
 				EngineID sel_eng = this->sel_engine;
 				if (sel_eng != INVALID_ENGINE) {
 					CommandCallback *callback = (this->vehicle_type == VEH_TRAIN && RailVehInfo(sel_eng)->railveh_type == RAILVEH_WAGON) ? CcBuildWagon : CcBuildPrimaryVehicle;
-					DoCommandP(this->window_number, sel_eng, 0, GetCmdBuildVeh(this->vehicle_type), callback);
+					DoCommandP(this->window_number + 1, sel_eng, 0, GetCmdBuildVeh(this->vehicle_type), callback);
 				}
 				break;
 			}
