Index: src/video/win32_v.cpp
===================================================================
--- src/video/win32_v.cpp	(revision 10929)
+++ src/video/win32_v.cpp	(working copy)
@@ -483,17 +483,27 @@
 
 			/* No matter the keyboard layout, we will map the '~' to the console */
 			scancode = scancode == 41 ? (int)WKC_BACKQUOTE : keycode;
+
 			HandleKeypress(GB(charcode, 0, 16) | (scancode << 16));
 			return 0;
 		}
 
 		case WM_KEYDOWN: {
 			keycode = MapWindowsKey(wParam);
+			uint32 unshifted_keycode = keycode & ~WKC_SHIFT;
 
 			/* Silently drop all text messages as those will be handled by WM_CHAR
 			 * WM_KEYDOWN only handles CTRL+ commands and special keys like VK_LEFT, etc. */
 			if (keycode == 0 || (keycode > WKC_PAUSE && GB(keycode, 13, 4) == 0)) return 0;
 
+			/* New keys already handled by WM_CHAR */
+			if (unshifted_keycode == WKC_SLASH) return 0;
+			if (unshifted_keycode >= WKC_SINGLEQUOTE && unshifted_keycode <= WKC_PERIOD) return 0;
+			if (unshifted_keycode >= WKC_L_BRACKET && unshifted_keycode <= WKC_R_BRACKET) return 0;
+			if (unshifted_keycode == WKC_MINUS) return 0;
+			/* Numpad keys also handled in char */
+			if (unshifted_keycode >= WKC_NUM_DIV && unshifted_keycode <= WKC_NUM_PLUS) return 0;
+
 			HandleKeypress(0 | (keycode << 16));
 			return 0;
 		}
