Index: src/fileio.cpp
===================================================================
--- src/fileio.cpp	(revision 17666)
+++ src/fileio.cpp	(working copy)
@@ -906,7 +906,9 @@
 	FOR_ALL_SEARCHPATHS(sp) DEBUG(misc, 4, "%s added as search path", _searchpaths[sp]);
 
 	if (_config_file != NULL) {
-		_personal_dir = strdup(_config_file);
+		char tmp[1024];
+		snprintf(tmp, 1024, "%s/%s", _searchpaths[SP_WORKING_DIR], _config_file);
+		_personal_dir = strdup(tmp);
 		char *end = strrchr(_personal_dir , PATHSEPCHAR);
 		if (end == NULL) {
 			_personal_dir[0] = '\0';
