diff -r e390a8a344e0 -r 037b80e8b486 src/network/network.cpp
--- a/src/network/network.cpp	Wed Jan 19 00:00:04 2011 +0100
+++ b/src/network/network.cpp	Wed Jan 19 00:05:47 2011 +0100
@@ -163,7 +163,7 @@
 	if (strcmp(password, "*") == 0) password = "";
 
 	if (!_network_server) {
-		NetworkClientSetPassword(password);
+		NetworkClientSetCompanyPassword(password);
 	} else {
 		HashCurrentCompanyPassword(password);
 	}
diff -r e390a8a344e0 -r 037b80e8b486 src/network/network_client.cpp
--- a/src/network/network_client.cpp	Wed Jan 19 00:00:04 2011 +0100
+++ b/src/network/network_client.cpp	Wed Jan 19 00:05:47 2011 +0100
@@ -1178,7 +1178,7 @@
 	MyClient::SendChat(action, type, dest, msg, data);
 }
 
-void NetworkClientSetPassword(const char *password)
+void NetworkClientSetCompanyPassword(const char *password)
 {
 	MyClient::SendSetPassword(password);
 }
diff -r e390a8a344e0 -r 037b80e8b486 src/network/network_client.h
--- a/src/network/network_client.h	Wed Jan 19 00:00:04 2011 +0100
+++ b/src/network/network_client.h	Wed Jan 19 00:05:47 2011 +0100
@@ -108,7 +108,7 @@
 typedef ClientNetworkGameSocketHandler MyClient;
 
 void NetworkClient_Connected();
-void NetworkClientSetPassword(const char *password);
+void NetworkClientSetCompanyPassword(const char *password);
 
 extern CompanyID _network_join_as;
 
