diff -r 77a7d6d4f649 src/os/unix/unix.cpp
--- a/src/os/unix/unix.cpp	Sun Aug 03 17:45:30 2014 +0000
+++ b/src/os/unix/unix.cpp	Tue Aug 05 15:32:42 2014 -0400
@@ -368,10 +368,10 @@
 	if (child_pid != 0) return;
 
 	const char *args[3];
-	args[0] = "/usr/bin/xdg-open";
+	args[0] = "xdg-open";
 	args[1] = url;
 	args[2] = NULL;
-	execv(args[0], const_cast<char * const *>(args));
+	execvp(args[0], const_cast<char * const *>(args));
 	DEBUG(misc, 0, "Failed to open url: %s", url);
 	exit(0);
 }
