Index: src/station_cmd.cpp
===================================================================
--- src/station_cmd.cpp	(revision 27967)
+++ src/station_cmd.cpp	(working copy)
@@ -3846,7 +3846,7 @@
 		Station *st = *st_iter;
 
 		/* Is the station reserved exclusively for somebody else? */
-		if (st->town->exclusive_counter > 0 && st->town->exclusivity != st->owner) continue;
+		if (st->owner != OWNER_NONE && st->town->exclusive_counter > 0 && st->town->exclusivity != st->owner) continue;
 
 		if (st->goods[type].rating == 0) continue; // Lowest possible rating, better not to give cargo anymore
 
Index: src/station_gui.cpp
===================================================================
--- src/station_gui.cpp	(revision 27967)
+++ src/station_gui.cpp	(working copy)
@@ -1818,7 +1818,7 @@
 		const Station *st = Station::Get(this->window_number);
 		int y = r.top + WD_FRAMERECT_TOP;
 
-		if (st->town->exclusive_counter > 0) {
+		if (st->owner != OWNER_NONE && st->town->exclusive_counter > 0) {
 			SetDParam(0, st->town->exclusivity);
 			y = DrawStringMultiLine(r.left + WD_FRAMERECT_LEFT, r.right - WD_FRAMERECT_RIGHT, y, r.bottom, st->town->exclusivity == st->owner ? STR_STATION_VIEW_EXCLUSIVE_RIGHTS_SELF : STR_STATION_VIEW_EXCLUSIVE_RIGHTS_COMPANY);
 			y += WD_PAR_VSEP_WIDE;
