Index: airport_gui.cpp
===================================================================
--- airport_gui.cpp	(revision 26627)
+++ airport_gui.cpp	(working copy)
@@ -476,6 +476,13 @@
 	{
 		/* First try to select an airport in the selected class. */
 		AirportClass *sel_apclass = AirportClass::Get(_selected_airport_class);
+		if(_selected_airport_index >= 0 && _selected_airport_index <= (int)sel_apclass->GetSpecCount()){
+			const AirportSpec *as = sel_apclass->GetSpec(_selected_airport_index);
+			if(as->IsAvailable()){ 
+				SelectOtherAirport(_selected_airport_index);
+				return;
+			}
+		}
 		for (uint i = 0; i < sel_apclass->GetSpecCount(); i++) {
 			const AirportSpec *as = sel_apclass->GetSpec(i);
 			if (as->IsAvailable()) {
