Index: src/window.cpp
===================================================================
--- src/window.cpp	(revision 12026)
+++ src/window.cpp	(working copy)
@@ -1720,13 +1720,6 @@
 	int x = _cursor.pos.x;
 	int y = _cursor.pos.y;
 
-	if (_input_events_this_tick != 0) {
-		/* HandleAutoscroll is called only once per GameLoop() - so we can clear the counter here */
-		_input_events_this_tick = 0;
-		/* there were some inputs this tick, don't scroll ??? */
-		return;
-	}
-
 	if (_patches.autoscroll && _game_mode != GM_MENU && !IsGeneratingWorld()) {
 		w = FindWindowFromPt(x, y);
 		if (w == NULL || w->flags4 & WF_DISABLE_VP_SCROLL) return;
@@ -1905,6 +1898,12 @@
 
 void InputLoop()
 {
+	if (_input_events_this_tick != 0) {
+		/* InputLoop is called only once per GameLoop() - so we can clear the counter here */
+		_input_events_this_tick = 0;
+		/* there were some inputs this tick, don't scroll neither move */
+		return;
+	}
 	HandleMouseEvents();
 	HandleAutoscroll();
 }
