Index: src/saveload/afterload.cpp
===================================================================
--- src/saveload/afterload.cpp	(revision 22643)
+++ src/saveload/afterload.cpp	(working copy)
@@ -1514,7 +1514,7 @@
 	 */
 	if (IsSavegameVersionBefore(83)) {
 		for (TileIndex t = 0; t < map_size; t++) {
-			if (IsTileType(t, MP_WATER) && IsShipDepot(t)) {
+			if (IsShipDepotTile(t)) {
 				_m[t].m4 = (TileHeight(t) == 0) ? OWNER_WATER : OWNER_NONE;
 			}
 		}
