Index: src/town_cmd.cpp
===================================================================
--- src/town_cmd.cpp	(Revision 8605)
+++ src/town_cmd.cpp	(Arbeitskopie)
@@ -812,8 +812,8 @@
 	// clearing some land and then building a road there.
 	tile = t->xy;
 	for (ptr = _town_coord_mod; ptr != endof(_town_coord_mod); ++ptr) {
-		/* Only work with plain land that not already has a house */
-		if (!IsTileType(tile, MP_HOUSE) && GetTileSlope(tile, NULL) == SLOPE_FLAT) {
+		/* Only work with plain land that not already has a house and it is not a water tile */
+		if (!IsTileType(tile, MP_HOUSE) && !IsTileType(tile, MP_WATER) && GetTileSlope(tile, NULL) == SLOPE_FLAT) {
 			if (!CmdFailed(DoCommand(tile, 0, 0, DC_AUTO, CMD_LANDSCAPE_CLEAR))) {
 				DoCommand(tile, GenRandomRoadBits(), t->index, DC_EXEC | DC_AUTO, CMD_BUILD_ROAD);
 				_current_player = old_player;
