diff -r 694ba4fb232f src/vehicle.cpp
--- a/src/vehicle.cpp	Sat Jan 15 09:53:21 2011 +0000
+++ b/src/vehicle.cpp	Sat Jan 15 12:31:53 2011 +0100
@@ -2181,6 +2181,7 @@
 	FOR_ALL_VEHICLES(v) {
 		/* Code ripped from CmdStartStopTrain. Can't call it, because of
 		 * ownership problems, so we'll duplicate some code, for now */
+		if (!v->IsPrimaryVehicle()) continue;
 		v->vehstatus |= VS_STOPPED;
 		v->MarkDirty();
 		SetWindowWidgetDirty(WC_VEHICLE_VIEW, v->index, VVW_WIDGET_START_STOP_VEH);
