Index: src/town_gui.cpp
===================================================================
--- src/town_gui.cpp	(revision 20006)
+++ src/town_gui.cpp	(working copy)
@@ -1188,7 +1187,7 @@
 
 	virtual EventState OnKeyPress(uint16 key, uint16 keycode)
 	{
-		EventState state;
+		EventState state = ES_NOT_HANDLED;
 		if (this->HandleEditBoxKey(TSEW_TOWNNAME_EDITBOX, key, keycode, state) == HEBR_CANCEL) {
 			this->UnfocusFocusedWidget();
 		}
