Index: src/sound/sdl_s.cpp
===================================================================
--- src/sound/sdl_s.cpp	(revision 17960)
+++ src/sound/sdl_s.cpp	(working copy)
@@ -32,7 +32,7 @@
 	const char *s = SdlOpen(SDL_INIT_AUDIO);
 	if (s != NULL) return s;
 
-	spec.freq = GetDriverParamInt(parm, "hz", 44100);
+	spec.freq = GetDriverParamInt(parm, "hz", 11025);
 	spec.format = AUDIO_S16SYS;
 	spec.channels = 2;
 	spec.samples = 512;
