Index: src/driver.h
===================================================================
--- src/driver.h	(revision 13065)
+++ src/driver.h	(working copy)
@@ -71,8 +71,14 @@
 	 */
 	virtual ~DriverFactoryBase() {
 		if (this->name == NULL) return;
-		GetDrivers().erase(this->name);
+
+		/* Prefix the name with driver type, to get the key we put the driver in */
+		char buf[32];
+		strecpy(buf, GetDriverTypeName(this->type), lastof(buf));
+		strecpy(buf + 5, this->name, lastof(buf));
+
 		free(this->name);
+		GetDrivers().erase(buf);
 	}
 
 	/** Shuts down all active drivers
