Index: src/network/network_server.cpp
===================================================================
--- src/network/network_server.cpp	(revision 20988)
+++ src/network/network_server.cpp	(working copy)
@@ -221,9 +221,9 @@
 	Packet *p;
 
 	FOR_ALL_COMPANIES(company) {
-		p = new Packet(PACKET_SERVER_COMPANY_INFO);
+		p = new Packet(PACKET_SERVER_MAP);
 
-		p->Send_uint8 (NETWORK_COMPANY_INFO_VERSION);
+		p->Send_uint8 (2);
 		p->Send_bool  (true);
 		this->Send_CompanyInformation(p, company, &company_stats[company->index]);
 
