Index: src/settings.cpp
===================================================================
--- src/settings.cpp	(wersja 22296)
+++ src/settings.cpp	(kopia robocza)
@@ -1076,8 +1076,8 @@
 				return false;
 			}
 		}
-		Station *st;
-		FOR_ALL_STATIONS(st) {
+		BaseStation *st;
+		FOR_ALL_ITEMS(BaseStation, id, st) {
 			if (TileX(st->xy) == 0 || TileY(st->xy) == 0) {
 				ShowErrorMessage(STR_CONFIG_SETTING_EDGES_NOT_EMPTY, INVALID_STRING_ID, WL_ERROR);
 				return false;
