Index: src/ai/ai_gui.cpp
===================================================================
--- src/ai/ai_gui.cpp	(revision 18415)
+++ src/ai/ai_gui.cpp	(working copy)
@@ -588,6 +588,8 @@
 				} else {
 					new_value = min(MAX_COMPANIES - 1, _settings_newgame.difficulty.max_no_competitors + 1);
 				}
+				if (this->selected_slot > _settings_newgame.difficulty.max_no_competitors)
+					this->selected_slot = INVALID_COMPANY; // Deselect if current selected slot gets out of the allowed range
 				IConsoleSetSetting("difficulty.max_no_competitors", new_value);
 				this->InvalidateData();
 				break;
