diff -r 6a3e9b2988c5 src/window.cpp
--- a/src/window.cpp	Thu Mar 05 18:27:25 2009 +0100
+++ b/src/window.cpp	Thu Mar 05 20:45:46 2009 +0100
@@ -1347,7 +1347,7 @@
 	w->SetDirty();
 }
 
-static bool _dragging_window;
+static bool _dragging_window; ///< A window is being dragged or resized.
 
 static bool HandleWindowDragging()
 {
diff -r 6a3e9b2988c5 src/window_gui.h
--- a/src/window_gui.h	Thu Mar 05 18:27:25 2009 +0100
+++ b/src/window_gui.h	Thu Mar 05 20:45:46 2009 +0100
@@ -584,7 +584,7 @@
 	WF_SCROLL_DOWN       = 1 <<  5, ///< Lower scroll button has been pressed, @see ScrollbarClickHandler()
 	WF_SCROLL_MIDDLE     = 1 <<  6, ///< Scrollbar scrolling, @see ScrollbarClickHandler()
 	WF_HSCROLL           = 1 <<  7,
-	WF_SIZING            = 1 <<  8,
+	WF_SIZING            = 1 <<  8, ///< Window is being resized.
 	WF_STICKY            = 1 <<  9, ///< Window is made sticky by user
 
 	WF_DISABLE_VP_SCROLL = 1 << 10, ///< Window does not do autoscroll, @see HandleAutoscroll()
