Index: src/aircraft_cmd.cpp
===================================================================
--- src/aircraft_cmd.cpp	(revision 27325)
+++ src/aircraft_cmd.cpp	(working copy)
@@ -37,6 +37,7 @@
 #include "core/backup_type.hpp"
 #include "zoom_func.h"
 #include "disaster_vehicle.h"
+#include "newgrf_airporttiles.h"
 
 #include "table/strings.h"
 
@@ -1347,8 +1348,14 @@
  */
 static void AircraftLandAirplane(Aircraft *v)
 {
+	Station *st = Station::Get(v->targetairport);
+
+	TileIndex vt = TileVirtXY(v->x_pos, v->y_pos);
+
 	v->UpdateDeltaXY(INVALID_DIR);
 
+	AirportTileAnimationTrigger(st, vt, AAT_STATION_AIRPLANE_LAND);
+
 	if (!PlayVehicleSound(v, VSE_TOUCHDOWN)) {
 		SndPlayVehicleFx(SND_17_SKID_PLANE, v);
 	}
Index: src/newgrf_animation_type.h
===================================================================
--- src/newgrf_animation_type.h	(revision 27325)
+++ src/newgrf_animation_type.h	(working copy)
@@ -46,11 +46,12 @@
 
 /** Animation triggers for airport tiles */
 enum AirpAnimationTrigger {
-	AAT_BUILT,               ///< Triggered when the airport is built (for all tiles at the same time).
-	AAT_TILELOOP,            ///< Triggered in the periodic tile loop.
-	AAT_STATION_NEW_CARGO,   ///< Triggered when new cargo arrives at the station (for all tiles at the same time).
-	AAT_STATION_CARGO_TAKEN, ///< Triggered when a cargo type is completely removed from the station (for all tiles at the same time).
-	AAT_STATION_250_TICKS,   ///< Triggered every 250 ticks (for all tiles at the same time).
+	AAT_BUILT,                   ///< Triggered when the airport is built (for all tiles at the same time).
+	AAT_TILELOOP,                ///< Triggered in the periodic tile loop.
+	AAT_STATION_NEW_CARGO,       ///< Triggered when new cargo arrives at the station (for all tiles at the same time).
+	AAT_STATION_CARGO_TAKEN,     ///< Triggered when a cargo type is completely removed from the station (for all tiles at the same time).
+	AAT_STATION_250_TICKS,       ///< Triggered every 250 ticks (for all tiles at the same time).
+	AAT_STATION_AIRPLANE_LAND,   ///< Triggered when an airplane (not a helicopter) touches down at the specific airport tile.
 };
 
 /** Animation triggers for objects. */
