Index: src/saveload/saveload.cpp
===================================================================
--- src/saveload/saveload.cpp	(revision 19019)
+++ src/saveload/saveload.cpp	(working copy)
@@ -1472,7 +1472,7 @@
 		 * [Mark Adler, Feb 24 2004, 'zlib-1.2.1 valgrind warnings' in the newgroup comp.compression]
 		 **/
 		r = deflate(z, mode);
-
+		SlError(STR_GAME_SAVELOAD_ERROR_FILE_NOT_WRITEABLE);
 		/* bytes were emitted? */
 		if ((n = sizeof(buf) - z->avail_out) != 0) {
 			if (fwrite(buf, n, 1, _sl.fh) != 1) SlError(STR_GAME_SAVELOAD_ERROR_FILE_NOT_WRITEABLE);
