Index: src/video/sdl_v.cpp
===================================================================
--- src/video/sdl_v.cpp	(revision 15232)
+++ src/video/sdl_v.cpp	(working copy)
@@ -17,6 +17,7 @@
 #include "../core/random_func.hpp"
 #include "../functions.h"
 #include "../texteff.hpp"
+#include "../fileio_func.h"
 #include "sdl_v.h"
 #include <SDL.h>
 
@@ -148,8 +149,15 @@
 
 static void GetAvailableVideoMode(int *w, int *h)
 {
-	/* All modes available? */
-	if (_all_modes || _num_resolutions == 0) return;
+	/* When there are no resolutions we can't change anything.
+	 * Furthermore when all modes are correct we don't need to change
+	 * anything, but when there isn't a config file yet we don't use
+	 * the given resolution, but we try to find the best matching
+	 * resolution. When for example the device can't show 640x480 even
+	 * when SDL says so. */
+	static bool first_run = !FileExists(_config_file);
+	if ((!first_run && _all_modes) || _num_resolutions == 0) return;
+	first_run = false;
 
 	/* Is the wanted mode among the available modes? */
 	for (int i = 0; i != _num_resolutions; i++) {
