diff -r c9b9d6c2e63f findversion.sh
--- a/findversion.sh	Thu Jul 15 17:14:36 2010 +0000
+++ b/findversion.sh	Fri Jul 30 23:21:51 2010 +0200
@@ -108,7 +108,7 @@
 	HASH=`LC_ALL=C hg id -i | cut -c1-12`
 	REV="h`echo $HASH | cut -c1-8`"
 	BRANCH=`hg branch | sed 's@^default$@@'`
-	REV_NR=`LC_ALL=C hg log -r $HASH:0 -k "svn" -l 1 --template "{desc}\n" "$SRC_DIR" | grep "^(svn r[0-9]*)" | head -n 1 | sed "s@.*(svn r\([0-9]*\)).*@\1@"`
+	REV_NR=`LC_ALL=C hg log -f -k "(svn r" --template "{desc}\n" --stat | awk 'BEGIN { s = "" } /^\(svn/ { s = $2 } /^$/ { s = 0 } / src/ { if (s != "") { sub(")", "", s); print s; exit 0 } }'`
 else
 	# We don't know
 	MODIFIED="1"
