diff --git a/src/newgrf_object.cpp b/src/newgrf_object.cpp
--- a/src/newgrf_object.cpp
+++ b/src/newgrf_object.cpp
@@ -140,6 +140,11 @@
 
 	const ObjectSpec *spec = ObjectSpec::GetByTile(tile);
 
+	/* Default objects have no associated NewGRF file */
+	if (spec->grf_prop.grffile == NULL) {
+		return 0xFFFE; // Defined in another grf file
+	}
+	
 	if (spec->grf_prop.grffile->grfid == cur_grfid) { // same object, same grf ?
 		return spec->grf_prop.local_id;
 	}
