Index: src/os/unix/unix.cpp
===================================================================
--- src/os/unix/unix.cpp	(revision 19771)
+++ src/os/unix/unix.cpp	(working copy)
@@ -99,6 +99,9 @@
 		snprintf(filename, lengthof(filename), "%s:%s", path, ent->d_name);
 	} else // XXX - only next line!
 #else
+	/* The base path isn't even valid. */
+	if (lstat(path, sb) == -1) return false;
+
 	assert(path[strlen(path) - 1] == PATHSEPCHAR);
 	if (strlen(path) > 2) assert(path[strlen(path) - 2] != PATHSEPCHAR);
 #endif
