From 71b410eb094b6662f88dcb45b822f309f4e569cc Mon Sep 17 00:00:00 2001
From: Michael Lutz <michi@icosahedron.de>
Date: Fri, 22 Aug 2008 20:25:18 +0200
Subject: -Fix (r13961) [FS#2248]: Clear the stuck state of a train that is reversing with the first vehicle inside a depot.

---
 src/train_cmd.cpp |    7 ++++++-
 1 files changed, 6 insertions(+), 1 deletions(-)

diff --git a/src/train_cmd.cpp b/src/train_cmd.cpp
index dc22232..344d359 100644
--- a/src/train_cmd.cpp
+++ b/src/train_cmd.cpp
@@ -1887,7 +1887,12 @@ static void ReverseTrainDirection(Vehicle *v)
 	if (crossing != INVALID_TILE) MaybeBarCrossingWithSound(crossing);
 
 	/* If we are inside a depot after reversing, don't bother with path reserving. */
-	if (v->u.rail.track & TRACK_BIT_DEPOT) return;
+	if (v->u.rail.track & TRACK_BIT_DEPOT) {
+		/* Can't be stuck here as inside a depot is always a safe tile. */
+		if (HasBit(v->u.rail.flags, VRF_TRAIN_STUCK)) InvalidateWindowWidget(WC_VEHICLE_VIEW, v->index, VVW_WIDGET_START_STOP_VEH);
+		ClrBit(v->u.rail.flags, VRF_TRAIN_STUCK);
+		return;
+	}
 
 	/* TrainExitDir does not always produce the desired dir for depots and
 	 * tunnels/bridges that is needed for UpdateSignalsOnSegment. */
-- 
1.5.6.4

