diff --git a/src/settings.cpp b/src/settings.cpp
index 5a80b9b..4dbeda7 100644
--- a/src/settings.cpp
+++ b/src/settings.cpp
@@ -809,9 +809,10 @@ static bool InvalidateDetailsWindow(int32 p1)
 	return true;
 }
 
-static bool InvalidateStationBuildWindow(int32 p1)
+static bool StationSpreadChanged(int32 p1)
 {
 	SetWindowDirty(WC_BUILD_STATION, 0);
+	InvalidateWindowData(WC_SELECT_STATION, 0);
 	return true;
 }
 
diff --git a/src/table/settings.ini b/src/table/settings.ini
index d1c1a1e..eafbee3 100644
--- a/src/table/settings.ini
+++ b/src/table/settings.ini
@@ -13,7 +13,7 @@ static bool v_PositionStatusbar(int32 p1);
 static bool PopulationInLabelActive(int32 p1);
 static bool RedrawScreen(int32 p1);
 static bool RedrawSmallmap(int32 p1);
-static bool InvalidateStationBuildWindow(int32 p1);
+static bool StationSpreadChanged(int32 p1);
 static bool InvalidateBuildIndustryWindow(int32 p1);
 static bool CloseSignalGUI(int32 p1);
 static bool InvalidateTownViewWindow(int32 p1);
@@ -1010,7 +1010,7 @@ max      = 64
 str      = STR_CONFIG_SETTING_STATION_SPREAD
 strhelp  = STR_CONFIG_SETTING_STATION_SPREAD_HELPTEXT
 strval   = STR_CONFIG_SETTING_TILE_LENGTH
-proc     = InvalidateStationBuildWindow
+proc     = StationSpreadChanged
 cat      = SC_BASIC
 
 [SDT_BOOL]
