diff --git a/src/pathfinder/follow_track.hpp b/src/pathfinder/follow_track.hpp
index c322d4c..cbbc102 100644
--- a/src/pathfinder/follow_track.hpp
+++ b/src/pathfinder/follow_track.hpp
@@ -259,8 +259,10 @@ protected:
 	/** return true if we can leave m_old_tile in m_exitdir */
 	inline bool CanExitOldTile()
 	{
+		if (!IsRoadTT()) return true;
+
 		/* road stop can be left at one direction only unless it's a drive-through stop */
-		if (IsRoadTT() && IsStandardRoadStopTile(m_old_tile)) {
+		if (IsStandardRoadStopTile(m_old_tile)) {
 			DiagDirection exitdir = GetRoadStopDir(m_old_tile);
 			if (exitdir != m_exitdir) {
 				m_err = EC_NO_WAY;
@@ -278,7 +280,7 @@ protected:
 		}

 		/* road depots can be also left in one direction only */
-		if (IsRoadTT() && IsRoadDepotTile(m_old_tile)) {
+		if (IsRoadDepotTile(m_old_tile)) {
 			DiagDirection exitdir = GetRoadDepotDirection(m_old_tile);
 			if (exitdir != m_exitdir) {
 				m_err = EC_NO_WAY;
