diff -r 69f5f9e51119 src/vehicle.cpp
--- a/src/vehicle.cpp	Thu Nov 18 18:50:17 2010 +0000
+++ b/src/vehicle.cpp	Thu Nov 18 23:11:11 2010 +0100
@@ -1871,7 +1871,7 @@
 		default: break;
 	}
 	if (default_effect == VE_DEFAULT) {
-		if (this->type == VEH_TRAIN && Train::From(this)->IsEngine()) {
+		if (this->type == VEH_TRAIN && !(Train::From(this)->IsWagon() || Train::From(this)->IsArticulatedPart())) {
 			if (e->u.rail.engclass == 0) {
 				/* Steam is offset by -4 units */
 				SB(this->vcache.cached_vis_effect, VE_OFFSET_START, VE_OFFSET_COUNT, VE_OFFSET_CENTRE - 4);
@@ -1955,7 +1955,7 @@
 		}
 
 		if (effect_type == VE_TYPE_DEFAULT) {
-			if (v->type == VEH_TRAIN && Train::From(v)->IsEngine()) {
+			if (v->type == VEH_TRAIN) {
 				/* Use default effect type for engine class. */
 				effect_type = RailVehInfo(v->engine_type)->engclass + 1;
 			} else {
