Index: newgrf_config.cpp
===================================================================
--- newgrf_config.cpp	(revision 27932)
+++ newgrf_config.cpp	(working copy)
@@ -787,15 +787,11 @@
 	/* Only then can we really start, especially by marking the whole screen dirty. Get those other windows hidden!. */
 	MarkWholeScreenDirty();
 
-	if (!VideoDriver::GetInstance()->HasGUI() || !ThreadObject::New(&DoScanNewGRFFiles, callback, NULL, "ottd:newgrf-scan")) {
-		_modal_progress_work_mutex->EndCritical();
-		_modal_progress_paint_mutex->EndCritical();
-		DoScanNewGRFFiles(callback);
-		_modal_progress_paint_mutex->BeginCritical();
-		_modal_progress_work_mutex->BeginCritical();
-	} else {
-		UpdateNewGRFScanStatus(0, NULL);
-	}
+    _modal_progress_work_mutex->EndCritical();
+	_modal_progress_paint_mutex->EndCritical();
+	DoScanNewGRFFiles(callback);
+	_modal_progress_paint_mutex->BeginCritical();
+	_modal_progress_work_mutex->BeginCritical();
 }
 
 /**
