diff -r 89b45241f4f1 src/goal.cpp
--- a/src/goal.cpp	Sat Jun 29 16:11:06 2013 +0200
+++ b/src/goal.cpp	Sat Jun 29 18:12:25 2013 +0200
@@ -165,7 +165,7 @@
 			g->progress = strdup(text);
 		}
 
-		InvalidateWindowData(WC_GOALS_LIST, 0);
+		InvalidateWindowData(WC_GOALS_LIST, 0, -1);
 	}
 
 	return CommandCost();
diff -r 89b45241f4f1 src/goal_gui.cpp
--- a/src/goal_gui.cpp	Sat Jun 29 16:11:06 2013 +0200
+++ b/src/goal_gui.cpp	Sat Jun 29 18:12:25 2013 +0200
@@ -245,12 +245,15 @@
 	/**
 	 * Some data on this window has become invalid.
 	 * @param data Information about the changed data.
+	 *        This is -1 if progress text changed, otherwise 0.
 	 * @param gui_scope Whether the call is done from GUI scope. You may not do everything when not in GUI scope. See #InvalidateWindowData() for details.
 	 */
 	virtual void OnInvalidateData(int data = 0, bool gui_scope = true)
 	{
 		if (!gui_scope) return;
 		this->vscroll->SetCount(this->CountLines());
+		/* Update widget min sizes if progress text was changed. */
+		if (data == -1) this->ReInit();
 	}
 };
 
