Index: src/window.cpp
===================================================================
--- src/window.cpp	(revision 18978)
+++ src/window.cpp	(working copy)
@@ -2109,6 +2109,10 @@
 				if (!(w->flags4 & WF_DISABLE_VP_SCROLL)) {
 					_scrolling_viewport = true;
 					_cursor.fix_at = true;
+					
+					// clear 2D scrolling caches before we start a 2D scroll
+					_cursor.h_wheel = 0;
+					_cursor.v_wheel = 0;
 				}
 				break;
 
