Index: src/lang/english.txt
===================================================================
--- src/lang/english.txt	(revision 15815)
+++ src/lang/english.txt	(working copy)
@@ -1107,6 +1107,7 @@
 STR_CONFIG_SETTING_TOWN_GROWTH_NORMAL                           :Normal
 STR_CONFIG_SETTING_TOWN_GROWTH_FAST                             :Fast
 STR_CONFIG_SETTING_TOWN_GROWTH_VERY_FAST                        :Very fast
+STR_CONFIG_PATCHES_TOWN_MAX_SIZE                                :{LTBLUE}Town maximum population (0 = disabled): {ORANGE}{STRING1}
 STR_CONFIG_SETTING_LARGER_TOWNS                                 :{LTBLUE}Proportion of towns that will become cities: {ORANGE}1 in {STRING1}
 STR_CONFIG_SETTING_LARGER_TOWNS_DISABLED                        :{LTBLUE}Proportion of towns that will become cities: {ORANGE}None
 STR_CONFIG_SETTING_CITY_SIZE_MULTIPLIER                         :{LTBLUE}Initial city size multiplier: {ORANGE}{STRING1}
Index: src/saveload/saveload.cpp
===================================================================
--- src/saveload/saveload.cpp	(revision 15815)
+++ src/saveload/saveload.cpp	(working copy)
@@ -40,7 +40,7 @@
 
 #include "saveload_internal.h"
 
-extern const uint16 SAVEGAME_VERSION = 115;
+extern const uint16 SAVEGAME_VERSION = 116;
 
 SavegameType _savegame_type; ///< type of savegame we are loading
 
Index: src/settings_gui.cpp
===================================================================
--- src/settings_gui.cpp	(revision 15815)
+++ src/settings_gui.cpp	(working copy)
@@ -1110,6 +1110,7 @@
 	SettingEntry("economy.town_growth_rate"),
 	SettingEntry("economy.larger_towns"),
 	SettingEntry("economy.initial_city_size"),
+	SettingEntry("economy.town_max_size"),
 };
 /** Towns sub-page */
 static SettingsPage _settings_economy_towns_page = {_settings_economy_towns, lengthof(_settings_economy_towns)};
Index: src/settings_type.h
===================================================================
--- src/settings_type.h	(revision 15815)
+++ src/settings_type.h	(working copy)
@@ -321,6 +321,7 @@
 	uint8  town_growth_rate;                 ///< town growth rate
 	uint8  larger_towns;                     ///< the number of cities to build. These start off larger and grow twice as fast
 	uint8  initial_city_size;                ///< multiplier for the initial size of the cities compared to towns
+	uint32 town_max_size;                    ///< limit of city population.
 	TownLayoutByte town_layout;              ///< select town layout
 	bool   allow_town_roads;                 ///< towns are allowed to build roads (always allowed when generating world / in SE)
 	bool   station_noise_level;              ///< build new airports when the town noise level is still within accepted limits
Index: src/table/settings.h
===================================================================
--- src/table/settings.h	(revision 15815)
+++ src/table/settings.h	(working copy)
@@ -412,6 +412,7 @@
 	 SDT_CONDVAR(GameSettings, economy.larger_towns,                 SLE_UINT8, 54, SL_MAX_VERSION, 0, D0,    4,     0,     255, 1, STR_CONFIG_SETTING_LARGER_TOWNS,           NULL),
 	 SDT_CONDVAR(GameSettings, economy.initial_city_size,            SLE_UINT8, 56, SL_MAX_VERSION, 0, 0,     2,     1,      10, 1, STR_CONFIG_SETTING_CITY_SIZE_MULTIPLIER,   NULL),
 	SDT_CONDBOOL(GameSettings, economy.mod_road_rebuild,                        77, SL_MAX_VERSION, 0, 0, false,                    STR_CONFIG_SETTING_MODIFIED_ROAD_REBUILD,  NULL),
+	 SDT_CONDVAR(GameSettings, economy.town_max_size,               SLE_UINT32,116, SL_MAX_VERSION, 0, 0,     0,     0,25000000, 0, STR_CONFIG_PATCHES_TOWN_MAX_SIZE,          NULL),
 
 	SDT_CONDNULL(1, 0, 106), // previously ai-new setting.
 	    SDT_BOOL(GameSettings, ai.ai_in_multiplayer,                                                0, 0, true,                     STR_CONFIG_SETTING_AI_IN_MULTIPLAYER,      NULL),
Index: src/town_cmd.cpp
===================================================================
--- src/town_cmd.cpp	(revision 15815)
+++ src/town_cmd.cpp	(working copy)
@@ -1272,6 +1272,11 @@
  */
 static bool GrowTown(Town *t)
 {
+	/* Limited city size by _settings_game.economy.town_max_size */
+	if((_settings_game.economy.town_max_size != 0) && (t->population >= _settings_game.economy.town_max_size)) {
+		return false;
+	}
+
 	static const TileIndexDiffC _town_coord_mod[] = {
 		{-1,  0},
 		{ 1,  1},
