--- C:\DOCUME~1\X\LOKALA~1\Temp\win32_v.cpp-revBASE.svn000.tmp.cpp	to maj 14 22:05:47 2009
+++ D:\D\Projects\OpenTTD\trunk\src\video\win32_v.cpp	to maj 14 22:06:18 2009
@@ -857,6 +857,7 @@
 
 			_ctrl_pressed = _wnd.has_focus && GetAsyncKeyState(VK_CONTROL)<0;
 			_shift_pressed = _wnd.has_focus && GetAsyncKeyState(VK_SHIFT)<0;
+			_alt_pressed = _wnd.has_focus && GetAsyncKeyState(VK_MENU)<0;
 
 			/* determine which directional keys are down */
 			if (_wnd.has_focus) {
