Index: src/newgrf.cpp
===================================================================
--- src/newgrf.cpp	(revision 19051)
+++ src/newgrf.cpp	(working copy)
@@ -2160,7 +2160,7 @@
 				break;
 
 			case 0x12: // Special flags
-				tsp->animation_special_flags = buf->ReadByte();
+				tsp->special_flags = buf->ReadByte();
 				break;
 
 			default:
Index: src/industrytype.h
===================================================================
--- src/industrytype.h	(revision 19051)
+++ src/industrytype.h	(working copy)
@@ -167,7 +167,7 @@
 	uint16 animation_info;                ///< Information about the animation (is it looping, how many loops etc)
 	uint8 animation_speed;                ///< The speed of the animation
 	uint8 animation_triggers;             ///< When to start the animation
-	uint8 animation_special_flags;        ///< Extra flags to influence the animation
+	uint8 special_flags;                  ///< Bitmask of extra flags used by the tile
 	bool enabled;                         ///< entity still avaible (by default true).newgrf can disable it, though
 	struct GRFFileProps grf_prop;
 };
Index: src/newgrf_industrytiles.cpp
===================================================================
--- src/newgrf_industrytiles.cpp	(revision 19051)
+++ src/newgrf_industrytiles.cpp	(working copy)
@@ -291,7 +291,7 @@
 	uint16 num_frames = GB(itspec->animation_info, 0, 8);
 
 	if (HasBit(itspec->callback_mask, CBM_INDT_ANIM_NEXT_FRAME)) {
-		uint16 callback_res = GetIndustryTileCallback(CBID_INDTILE_ANIM_NEXT_FRAME, HasBit(itspec->animation_special_flags, 0) ? Random() : 0, 0, gfx, ind, tile);
+		uint16 callback_res = GetIndustryTileCallback(CBID_INDTILE_ANIM_NEXT_FRAME, HasBit(itspec->special_flags, 0) ? Random() : 0, 0, gfx, ind, tile);
 
 		if (callback_res != CALLBACK_FAILED) {
 			frame_set_by_callback = true;
