Index: src/core/enum_type.hpp
===================================================================
--- src/core/enum_type.hpp	(revision 22021)
+++ src/core/enum_type.hpp	(working copy)
@@ -172,6 +172,13 @@
 		this->m_val = (storage_type)((enum_type)this->m_val & e);
 		return *this;
 	}
+
+	/** Bit math (xor) assignment operator (from enum_type) */
+	FORCEINLINE SimpleTinyEnumT &operator ^= (enum_type e)
+	{
+		this->m_val = (storage_type)((enum_type)this->m_val ^ e);
+		return *this;
+	}
 };
 
 #endif /* ENUM_TYPE_HPP */
