Index: src/cargomonitor.cpp
===================================================================
--- src/cargomonitor.cpp	(revision 27929)
+++ src/cargomonitor.cpp	(working copy)
@@ -153,7 +153,13 @@
 	for (const Industry * const *ip = st->industries_near.Begin(); ip != st->industries_near.End(); ip++) {
 		CargoMonitorID num = EncodeCargoIndustryMonitor(company, cargo_type, (*ip)->index);
 		CargoMonitorMap::iterator iter = _cargo_deliveries.find(num);
-		if (iter != _cargo_deliveries.end()) iter->second += amount;
+		if (iter != _cargo_deliveries.end()) {
+			uint cargo_index;
+			for (cargo_index = 0; cargo_index < lengthof((*ip)->accepts_cargo); cargo_index++) {
+				if (cargo_type == (*ip)->accepts_cargo[cargo_index]) break;
+			}
+			iter->second = (*ip)->incoming_cargo_waiting[cargo_index];
+		}
 	}
 }
 
