Index: station_gui.c
===================================================================
--- station_gui.c	(revision 7068)
+++ station_gui.c	(working copy)
@@ -373,7 +373,7 @@ static void PlayerStationsWndProc(Window
 		}
 		case 27: {
 			int i;
-			for (i = 0; i < NUM_CARGO; i++) {
+			for (i = 0; i <= NUM_CARGO; i++) {
 				LowerWindowWidget(w, i + 12);
 			}
 			LowerWindowWidget(w, 27);
@@ -433,7 +433,7 @@ static void PlayerStationsWndProc(Window
 		for (i = 0; i < 5; i++) {
 			if (HASBIT(facilities, i)) LowerWindowWidget(w, i + 6);
 		}
-		for (i = 0; i < NUM_CARGO; i++) {
+		for (i = 0; i <= NUM_CARGO; i++) {
 			if (HASBIT(cargo_filter, i)) LowerWindowWidget(w, i + 12);
 		}
 		SetWindowWidgetLoweredState(w, 26, facilities == (FACIL_TRAIN | FACIL_TRUCK_STOP | FACIL_BUS_STOP | FACIL_AIRPORT | FACIL_DOCK));
