Index: src/station_cmd.cpp
===================================================================
--- src/station_cmd.cpp	(revision 21229)
+++ src/station_cmd.cpp	(working copy)
@@ -1385,7 +1385,6 @@
 {
 	/* Count of the number of tiles removed */
 	int quantity = 0;
-	CommandCost total_cost(EXPENSES_CONSTRUCTION);
 
 	/* Do the action for every tile into the area */
 	TILE_AREA_LOOP(tile, ta) {
@@ -1455,10 +1454,6 @@
 				if (IsRailStationTile(v->tile)) SetRailStationPlatformReservation(v->tile, TrackdirToExitdir(ReverseTrackdir(v->GetVehicleTrackdir())), true);
 			}
 		}
-		if (keep_rail) {
-			/* Don't refund the 'steel' of the track! */
-			total_cost.AddCost(-_price[PR_CLEAR_RAIL]);
-		}
 	}
 
 	if (quantity == 0) return_cmd_error(STR_ERROR_THERE_IS_NO_STATION);
@@ -1481,8 +1476,11 @@
 		}
 	}
 
-	total_cost.AddCost(quantity * removal_cost);
-	return total_cost;
+	if (keep_rail) {
+		/* Don't refund the 'steel' of the track! */
+		removal_cost -= _price[PR_CLEAR_RAIL];
+	}
+	return CommandCost(EXPENSES_CONSTRUCTION, quantity * removal_cost);
 }
 
 /**
