diff -r 2b3f4de6ba32 src/map_func.h
--- a/src/map_func.h	Thu Mar 05 11:50:42 2009 +1100
+++ b/src/map_func.h	Thu Mar 05 13:55:08 2009 +1100
@@ -102,8 +102,8 @@
 }
 
 /**
- * Gets the maximum X coordinate within the map, including MP_VOID
- * @return the maximum X coordinate
+ * Gets the maximum Y coordinate within the map, including MP_VOID
+ * @return the maximum Y coordinate
  */
 static inline uint MapMaxY()
 {
