Index: src/video/win32_v.cpp
===================================================================
--- src/video/win32_v.cpp	(r‚vision 10834)
+++ src/video/win32_v.cpp	(copie de travail)
@@ -621,6 +621,9 @@
 
 #if !defined(WINCE)
 		case WM_ACTIVATE: {
+			/* Don't do anything if we are closing openttd */
+			if (_exit_game) break;
+
 			bool active = (LOWORD(wParam) != WA_INACTIVE);
 			bool minimized = (HIWORD(wParam) != 0);
 			if (_wnd.fullscreen) {
