Index: src/saveload/vehicle_sl.cpp
===================================================================
--- src/saveload/vehicle_sl.cpp	(Revision 20894)
+++ src/saveload/vehicle_sl.cpp	(Arbeitskopie)
@@ -423,7 +423,8 @@
 		 SLE_CONDVAR(Vehicle, x_pos,                 SLE_UINT32,                   6, SL_MAX_VERSION),
 		 SLE_CONDVAR(Vehicle, y_pos,                 SLE_FILE_U16 | SLE_VAR_U32,   0,   5),
 		 SLE_CONDVAR(Vehicle, y_pos,                 SLE_UINT32,                   6, SL_MAX_VERSION),
-		     SLE_VAR(Vehicle, z_pos,                 SLE_UINT8),
+		 SLE_CONDVAR(Vehicle, z_pos,                 SLE_FILE_U8 | SLE_VAR_U32,    0, MORE_HEIGHTLEVEL_SAVEGAME_VERSION - 1),
+		 SLE_CONDVAR(Vehicle, z_pos,                 SLE_UINT32,                   MORE_HEIGHTLEVEL_SAVEGAME_VERSION, SL_MAX_VERSION),
 		     SLE_VAR(Vehicle, direction,             SLE_UINT8),
 
 		SLE_CONDNULL(2,                                                            0,  57),
@@ -616,7 +617,8 @@
 		 SLE_CONDVAR(Vehicle, x_pos,                 SLE_INT32,                    6, SL_MAX_VERSION),
 		 SLE_CONDVAR(Vehicle, y_pos,                 SLE_FILE_I16 | SLE_VAR_I32,   0,   5),
 		 SLE_CONDVAR(Vehicle, y_pos,                 SLE_INT32,                    6, SL_MAX_VERSION),
-		     SLE_VAR(Vehicle, z_pos,                 SLE_UINT8),
+		 SLE_CONDVAR(Vehicle, z_pos,                 SLE_FILE_U8 | SLE_VAR_U32,    0, MORE_HEIGHTLEVEL_SAVEGAME_VERSION - 1),
+		 SLE_CONDVAR(Vehicle, z_pos,                 SLE_UINT32,                   MORE_HEIGHTLEVEL_SAVEGAME_VERSION, SL_MAX_VERSION),
 
 		     SLE_VAR(Vehicle, cur_image,             SLE_UINT16),
 		SLE_CONDNULL(5,                                                            0,  57),
@@ -648,7 +650,8 @@
 		 SLE_CONDVAR(Vehicle, x_pos,                 SLE_INT32,                    6, SL_MAX_VERSION),
 		 SLE_CONDVAR(Vehicle, y_pos,                 SLE_FILE_I16 | SLE_VAR_I32,   0,   5),
 		 SLE_CONDVAR(Vehicle, y_pos,                 SLE_INT32,                    6, SL_MAX_VERSION),
-		     SLE_VAR(Vehicle, z_pos,                 SLE_UINT8),
+		 SLE_CONDVAR(Vehicle, z_pos,                 SLE_FILE_U8 | SLE_VAR_U32,    0, MORE_HEIGHTLEVEL_SAVEGAME_VERSION - 1),
+		 SLE_CONDVAR(Vehicle, z_pos,                 SLE_UINT32,                   MORE_HEIGHTLEVEL_SAVEGAME_VERSION, SL_MAX_VERSION),
 		     SLE_VAR(Vehicle, direction,             SLE_UINT8),
 
 		SLE_CONDNULL(5,                                                            0,  57),
Index: src/vehicle_base.h
===================================================================
--- src/vehicle_base.h	(Revision 20894)
+++ src/vehicle_base.h	(Arbeitskopie)
@@ -126,7 +126,7 @@
 
 	int32 x_pos;             // coordinates
 	int32 y_pos;
-	byte z_pos;
+	int32 z_pos;
 	DirectionByte direction; // facing
 
 	OwnerByte owner;         // which company owns the vehicle?
