Index: strings.cpp
===================================================================
--- strings.cpp	(revision 15633)
+++ strings.cpp	(working copy)
@@ -1238,8 +1238,12 @@
 	/* resolution size? */
 	if (IsInsideMM(ind, (SPECSTR_RESOLUTION_START - 0x70E4), (SPECSTR_RESOLUTION_END - 0x70E4) + 1)) {
 		int i = ind - (SPECSTR_RESOLUTION_START - 0x70E4);
+		bool is16_9 = ((_resolutions[i].width / 16) == (_resolutions[i].height / 9));
+		bool is16_10 = ((_resolutions[i].width / 16) == (_resolutions[i].height / 10));
+		bool is4_3 = ((_resolutions[i].width / 4) == (_resolutions[i].height / 3));
 		buff += seprintf(
-			buff, last, "%dx%d", _resolutions[i].width, _resolutions[i].height
+			buff, last, "%dx%d%s", _resolutions[i].width, _resolutions[i].height,
+			( is4_3 ? " (4:3)" : ( is16_9 ? " (16:9)" : ( is16_10 ? " (16:10)" : "" )))
 		);
 		return buff;
 	}
