Index: src/industry_cmd.cpp
===================================================================
--- src/industry_cmd.cpp	(revision 27931)
+++ src/industry_cmd.cpp	(working copy)
@@ -810,11 +810,14 @@
 
 	if (_game_mode == GM_EDITOR) return;
 
+	/* Get IndustryGfx before TransportIndustryGoods changes it */
+	IndustryGfx gfx = GetIndustryGfx(tile);
+
 	TransportIndustryGoods(tile);
 
 	if (StartStopIndustryTileAnimation(tile, IAT_TILELOOP)) return;
 
-	IndustryGfx newgfx = GetIndustryTileSpec(GetIndustryGfx(tile))->anim_next;
+	IndustryGfx newgfx = GetIndustryTileSpec(gfx)->anim_next;
 	if (newgfx != INDUSTRYTILE_NOANIM) {
 		ResetIndustryConstructionStage(tile);
 		SetIndustryGfx(tile, newgfx);
@@ -822,7 +825,6 @@
 		return;
 	}
 
-	IndustryGfx gfx = GetIndustryGfx(tile);
 	switch (gfx) {
 	case GFX_COAL_MINE_TOWER_NOT_ANIMATED:
 	case GFX_COPPER_MINE_TOWER_NOT_ANIMATED:
