Index: src/network/network_gamelist.cpp
===================================================================
--- src/network/network_gamelist.cpp	(revision 22021)
+++ src/network/network_gamelist.cpp	(working copy)
@@ -16,6 +16,7 @@
 
 #include "../stdafx.h"
 #include "../debug.h"
+#include "../window_func.h"
 #include "../thread/thread.h"
 #include "network_internal.h"
 #include "network_udp.h"
@@ -195,6 +196,8 @@
 			}
 		}
 	}
+	
+	InvalidateWindowClassesData(WC_NETWORK_WINDOW);
 }
 
 #endif /* ENABLE_NETWORK */
