Index: src/town_cmd.cpp
===================================================================
--- src/town_cmd.cpp	(Revision 13876)
+++ src/town_cmd.cpp	(Arbeitskopie)
@@ -1022,7 +1022,7 @@
 		_grow_town_result = GROWTH_SEARCH_STOPPED;
 
 		/* Remove hills etc */
-		LevelTownLand(tile);
+		if (Chance16(1, 6)) LevelTownLand(tile);
 
 		/* Is a road allowed here? */
 		switch (t1->GetActiveLayout()) {
