Index: src/town_gui.cpp
===================================================================
--- src/town_gui.cpp	(revisión: 17618)
+++ src/town_gui.cpp	(copia de trabajo)
@@ -1098,6 +1098,11 @@
 		this->ExecuteFoundTownCommand(tile, false, STR_ERROR_CAN_T_FOUND_TOWN_HERE, CcFoundTown);
 	}
 
+	virtual void OnOpenOSKWindow(int wid)
+	{
+		if (_cursor.sprite != SPR_CURSOR_TOWN) ShowOnScreenKeyboard(this, wid, 0, 0);
+	}
+
 	virtual void OnPlaceObjectAbort()
 	{
 		this->RaiseButtons();
