Index: src/video/cocoa/event.mm
===================================================================
--- src/video/cocoa/event.mm	(revision 25039)
+++ src/video/cocoa/event.mm	(working copy)
@@ -542,6 +542,12 @@
 			_cursor.v_wheel -= (int)([ event deltaY ] * 5 * _settings_client.gui.scrollwheel_multiplier);
 			break;
 
+		case NSCursorUpdate:
+		case NSMouseEntered:
+		case NSMouseExited:
+			/* Catching those events when the cursor is dragging. Otherwise, osx update itself the cursor with the arrow system one while the cursor position is virtually out of the window. Moreover OTTD hide its own cursor. */
+			if (_cursor.fix_at) break;
+
 		default:
 			[ NSApp sendEvent:event ];
 	}
