Index: src/network/core/tcp_connect.cpp
===================================================================
--- src/network/core/tcp_connect.cpp	(revision 20035)
+++ src/network/core/tcp_connect.cpp	(working copy)
@@ -15,6 +15,7 @@
 
 #include "../../stdafx.h"
 #include "../../thread/thread.h"
+#include "../../gfx_func.h"
 
 #include "tcp.h"
 
@@ -47,6 +48,7 @@
 
 /* static */ void TCPConnecter::ThreadEntry(void *param)
 {
+	CSleep(10000);
 	static_cast<TCPConnecter*>(param)->Connect();
 }
 
