Index: src/string_func.h
===================================================================
--- src/string_func.h	(revision 22851)
+++ src/string_func.h	(working copy)
@@ -195,7 +195,7 @@
 #endif
 
 /* strndup is a GNU extension */
-#if defined(_GNU_SOURCE) || (defined(__NetBSD_Version__) && 400000000 <= __NetBSD_Version__) || (defined(__FreeBSD_version) && 701101 <= __FreeBSD_version)
+#if defined(_GNU_SOURCE) || (defined(__NetBSD_Version__) && 400000000 <= __NetBSD_Version__) || (defined(__FreeBSD_version) && 701101 <= __FreeBSD_version) || (defined(__DARWIN_C_LEVEL) && __DARWIN_C_LEVEL >= 200809L)
 #	undef DEFINE_STRNDUP
 #else
 #	define DEFINE_STRNDUP
