Fix missing include in routing.cpp

diff -r 7940ecee9754 src/routing.cpp
--- a/src/routing.cpp	Mon Dec 08 16:28:56 2008 +0100
+++ b/src/routing.cpp	Tue Dec 09 15:01:01 2008 +0100
@@ -11,6 +11,7 @@
 #include "routing_classes.h"
 #include "station_base.h"
 #include "map_func.h"
+#include "newgrf_cargo.h"
 #include "cargo_type.h"
 #include "cargotype.h"
 #include "settings_type.h"
@@ -133,7 +134,7 @@ int32 RoutingBase::ModifyGoldRoutingSyst
 	for (CargoID c = CT_BEGIN; c != NUM_CARGO; c++) {
 		const CargoSpec *cs = GetCargo(c);
 
-		if (IsCargoInClass(c, CC_ARMOURED))
+		if (IsCargoInClass(c, CC_ARMOURED)) {
 			ModifyRoutingSystem(c, p1);
 		}
 	}
