--- /home/micro/Downloads/openttd-1.1.0-beta1/config.lib	2010-12-08 17:39:31.000000000 +0300
+++ /home/micro/ports/config.lib	2011-01-05 23:26:15.000000000 +0300
@@ -1538,7 +1538,7 @@
 		CFLAGS="$CFLAGS -DWITH_ZLIB"
 	fi
 
-	if [ -n "$lzma_config" ]; then
+	if [ -n "$lzma_config" ] && [ "$os" != "FREEBSD" ]; then
 		CFLAGS="$CFLAGS -DWITH_LZMA"
 		CFLAGS="$CFLAGS `$lzma_config --cflags | tr '\n\r' '  '`"
 
@@ -2561,13 +2561,23 @@
 
 	if [ "$with_lzma" = "1" ] || [ "$with_lzma" = "" ] || [ "$with_lzma" = "2" ]; then
 		lzma_config="pkg-config liblzma"
+		if [ "$os" = "FREEBSD" ]; then
+		    lzma_config="lzma"
+		fi
 	else
 		lzma_config="$with_lzma"
 	fi
 
 	version=`$lzma_config --modversion 2>/dev/null`
+	if [ "$os" = "FREEBSD" ]; then
+	    version=`$lzma_config --version | grep lzma | awk '{print $2}' 2>/dev/null`
+	fi
 	ret=$?
-	log 2 "executing $lzma_config --modversion"
+	if [ "$os" = "FREEBSD" ]; then
+	    log 2 "executing $lzma_config --version | grep lzma | awk '{print $2}'"
+	else
+	    log 2 "executing $lzma_config --modversion"
+	fi
 	log 2 "  returned $version"
 	log 2 "  exit code $ret"
 
@@ -2576,7 +2586,7 @@
 
 		# It was forced, so it should be found.
 		if [ "$with_lzma" != "1" ]; then
-			log 1 "configure: error: pkg-config liblzma couldn't be found"
+			log 1 "configure: error: '$lzma_config' couldn't be found"
 			log 1 "configure: error: you supplied '$with_lzma', but it seems invalid"
 			exit 1
 		fi
