Index: src/newgrf_airport.h
===================================================================
--- src/newgrf_airport.h	(wersja 23645)
+++ src/newgrf_airport.h	(kopia robocza)
@@ -51,7 +51,7 @@
 		if (this->att->ti.x == -0x80) {
 			this->tile = INVALID_TILE;
 		} else {
-			this->tile = base_tile + ToTileIndexDiff(att->ti);
+			this->tile = this->base_tile + ToTileIndexDiff(this->att->ti);
 		}
 		return *this;
 	}
