Index: src/music_gui.cpp
===================================================================
--- src/music_gui.cpp	(revision 11268)
+++ src/music_gui.cpp	(working copy)
@@ -96,6 +96,7 @@
 	FioFindFullPath(filename, lengthof(filename), GM_DIR,
 			origin_songs_specs[_music_wnd_cursong - 1].filename);
 	_music_driver->PlaySong(filename);
+	_song_is_active = true;
 }
 
 static void DoStopMusic()
@@ -158,7 +159,6 @@
 	}
 	_music_wnd_cursong = _cur_playlist[0];
 	DoPlaySong();
-	_song_is_active = true;
 
 	InvalidateWindowWidget(WC_MUSIC_WINDOW, 0, 9);
 }
