Index: src/road_cmd.cpp
===================================================================
--- src/road_cmd.cpp	(revision 20211)
+++ src/road_cmd.cpp	(working copy)
@@ -860,10 +860,8 @@
 				cost.AddCost(ret);
 				had_success = true;
 			} else {
-				last_error = ret;
-
 				/* Ownership errors are more important. */
-				if (last_error.GetErrorMessage() == STR_ERROR_OWNED_BY) break;
+				if (last_error.GetErrorMessage() != STR_ERROR_OWNED_BY) last_error = ret;
 			}
 		}
 
