Index: src/sound.cpp
===================================================================
--- src/sound.cpp	(revision 11943)
+++ src/sound.cpp	(working copy)
@@ -208,8 +240,8 @@
 		const ViewPort *vp = (*wz)->viewport;
 
 		if (vp != NULL &&
-				IsInsideBS(x, vp->virtual_left, vp->virtual_width) &&
-				IsInsideBS(y, vp->virtual_top, vp->virtual_height)) {
+				IsInsideBS(x, vp->virtual_left-TILE_PIXELS, vp->virtual_width+TILE_PIXELS*2) &&
+				IsInsideBS(y, vp->virtual_top-TILE_SIZE, vp->virtual_height+TILE_SIZE*2)) {
 			int left = (x - vp->virtual_left);
 
 			StartSound(
