diff --git a/projects/generate_vs140.vcxproj b/projects/generate_vs140.vcxproj
index 67b1b77..354a2ac 100644
--- a/projects/generate_vs140.vcxproj
+++ b/projects/generate_vs140.vcxproj
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup Label="ProjectConfigurations">
     <ProjectConfiguration Include="Debug|Win32">
diff --git a/projects/generate_vs141.vcxproj b/projects/generate_vs141.vcxproj
index 2db93cf..2cf9e5a 100644
--- a/projects/generate_vs141.vcxproj
+++ b/projects/generate_vs141.vcxproj
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup Label="ProjectConfigurations">
     <ProjectConfiguration Include="Debug|Win32">
diff --git a/projects/langs_vs100.vcxproj b/projects/langs_vs100.vcxproj
index fc77e03..88cce9e 100644
--- a/projects/langs_vs100.vcxproj
+++ b/projects/langs_vs100.vcxproj
@@ -317,18 +317,18 @@
       <AdditionalInputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\src\lang\english.txt;..\objs\strgen\strgen.exe;%(AdditionalInputs)</AdditionalInputs>
       <Outputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\bin\lang\slovenian.lng;%(Outputs)</Outputs>
     </CustomBuild>
-    <CustomBuild Include="..\src\lang\spanish_MX.txt">
-      <Message Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">Generating spanish_MX language file</Message>
-      <Command Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\objs\strgen\strgen.exe -s ..\src\lang -d ..\bin\lang "%(FullPath)"</Command>
-      <AdditionalInputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\src\lang\english.txt;..\objs\strgen\strgen.exe;%(AdditionalInputs)</AdditionalInputs>
-      <Outputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\bin\lang\spanish_MX.lng;%(Outputs)</Outputs>
-    </CustomBuild>
     <CustomBuild Include="..\src\lang\spanish.txt">
       <Message Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">Generating spanish language file</Message>
       <Command Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\objs\strgen\strgen.exe -s ..\src\lang -d ..\bin\lang "%(FullPath)"</Command>
       <AdditionalInputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\src\lang\english.txt;..\objs\strgen\strgen.exe;%(AdditionalInputs)</AdditionalInputs>
       <Outputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\bin\lang\spanish.lng;%(Outputs)</Outputs>
     </CustomBuild>
+    <CustomBuild Include="..\src\lang\spanish_MX.txt">
+      <Message Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">Generating spanish_MX language file</Message>
+      <Command Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\objs\strgen\strgen.exe -s ..\src\lang -d ..\bin\lang "%(FullPath)"</Command>
+      <AdditionalInputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\src\lang\english.txt;..\objs\strgen\strgen.exe;%(AdditionalInputs)</AdditionalInputs>
+      <Outputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\bin\lang\spanish_MX.lng;%(Outputs)</Outputs>
+    </CustomBuild>
     <CustomBuild Include="..\src\lang\swedish.txt">
       <Message Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">Generating swedish language file</Message>
       <Command Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\objs\strgen\strgen.exe -s ..\src\lang -d ..\bin\lang "%(FullPath)"</Command>
diff --git a/projects/langs_vs100.vcxproj.filters b/projects/langs_vs100.vcxproj.filters
index 0afc8c9..bf5257e 100644
--- a/projects/langs_vs100.vcxproj.filters
+++ b/projects/langs_vs100.vcxproj.filters
@@ -142,10 +142,10 @@
     <CustomBuild Include="..\src\lang\slovenian.txt">
       <Filter>Translations</Filter>
     </CustomBuild>
-    <CustomBuild Include="..\src\lang\spanish_MX.txt">
+    <CustomBuild Include="..\src\lang\spanish.txt">
       <Filter>Translations</Filter>
     </CustomBuild>
-    <CustomBuild Include="..\src\lang\spanish.txt">
+    <CustomBuild Include="..\src\lang\spanish_MX.txt">
       <Filter>Translations</Filter>
     </CustomBuild>
     <CustomBuild Include="..\src\lang\swedish.txt">
diff --git a/projects/langs_vs140.vcxproj b/projects/langs_vs140.vcxproj
index 317fd34..25e7160 100644
--- a/projects/langs_vs140.vcxproj
+++ b/projects/langs_vs140.vcxproj
@@ -317,18 +317,18 @@
       <AdditionalInputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\src\lang\english.txt;..\objs\strgen\strgen.exe;%(AdditionalInputs)</AdditionalInputs>
       <Outputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\bin\lang\slovenian.lng;%(Outputs)</Outputs>
     </CustomBuild>
-    <CustomBuild Include="..\src\lang\spanish_MX.txt">
-      <Message Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">Generating spanish_MX language file</Message>
-      <Command Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\objs\strgen\strgen.exe -s ..\src\lang -d ..\bin\lang "%(FullPath)"</Command>
-      <AdditionalInputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\src\lang\english.txt;..\objs\strgen\strgen.exe;%(AdditionalInputs)</AdditionalInputs>
-      <Outputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\bin\lang\spanish_MX.lng;%(Outputs)</Outputs>
-    </CustomBuild>
     <CustomBuild Include="..\src\lang\spanish.txt">
       <Message Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">Generating spanish language file</Message>
       <Command Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\objs\strgen\strgen.exe -s ..\src\lang -d ..\bin\lang "%(FullPath)"</Command>
       <AdditionalInputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\src\lang\english.txt;..\objs\strgen\strgen.exe;%(AdditionalInputs)</AdditionalInputs>
       <Outputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\bin\lang\spanish.lng;%(Outputs)</Outputs>
     </CustomBuild>
+    <CustomBuild Include="..\src\lang\spanish_MX.txt">
+      <Message Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">Generating spanish_MX language file</Message>
+      <Command Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\objs\strgen\strgen.exe -s ..\src\lang -d ..\bin\lang "%(FullPath)"</Command>
+      <AdditionalInputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\src\lang\english.txt;..\objs\strgen\strgen.exe;%(AdditionalInputs)</AdditionalInputs>
+      <Outputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\bin\lang\spanish_MX.lng;%(Outputs)</Outputs>
+    </CustomBuild>
     <CustomBuild Include="..\src\lang\swedish.txt">
       <Message Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">Generating swedish language file</Message>
       <Command Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\objs\strgen\strgen.exe -s ..\src\lang -d ..\bin\lang "%(FullPath)"</Command>
diff --git a/projects/langs_vs140.vcxproj.filters b/projects/langs_vs140.vcxproj.filters
index 0afc8c9..bf5257e 100644
--- a/projects/langs_vs140.vcxproj.filters
+++ b/projects/langs_vs140.vcxproj.filters
@@ -142,10 +142,10 @@
     <CustomBuild Include="..\src\lang\slovenian.txt">
       <Filter>Translations</Filter>
     </CustomBuild>
-    <CustomBuild Include="..\src\lang\spanish_MX.txt">
+    <CustomBuild Include="..\src\lang\spanish.txt">
       <Filter>Translations</Filter>
     </CustomBuild>
-    <CustomBuild Include="..\src\lang\spanish.txt">
+    <CustomBuild Include="..\src\lang\spanish_MX.txt">
       <Filter>Translations</Filter>
     </CustomBuild>
     <CustomBuild Include="..\src\lang\swedish.txt">
diff --git a/projects/langs_vs141.vcxproj b/projects/langs_vs141.vcxproj
index 2444b0e..c554c6e 100644
--- a/projects/langs_vs141.vcxproj
+++ b/projects/langs_vs141.vcxproj
@@ -317,18 +317,18 @@
       <AdditionalInputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\src\lang\english.txt;..\objs\strgen\strgen.exe;%(AdditionalInputs)</AdditionalInputs>
       <Outputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\bin\lang\slovenian.lng;%(Outputs)</Outputs>
     </CustomBuild>
-    <CustomBuild Include="..\src\lang\spanish_MX.txt">
-      <Message Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">Generating spanish_MX language file</Message>
-      <Command Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\objs\strgen\strgen.exe -s ..\src\lang -d ..\bin\lang "%(FullPath)"</Command>
-      <AdditionalInputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\src\lang\english.txt;..\objs\strgen\strgen.exe;%(AdditionalInputs)</AdditionalInputs>
-      <Outputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\bin\lang\spanish_MX.lng;%(Outputs)</Outputs>
-    </CustomBuild>
     <CustomBuild Include="..\src\lang\spanish.txt">
       <Message Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">Generating spanish language file</Message>
       <Command Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\objs\strgen\strgen.exe -s ..\src\lang -d ..\bin\lang "%(FullPath)"</Command>
       <AdditionalInputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\src\lang\english.txt;..\objs\strgen\strgen.exe;%(AdditionalInputs)</AdditionalInputs>
       <Outputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\bin\lang\spanish.lng;%(Outputs)</Outputs>
     </CustomBuild>
+    <CustomBuild Include="..\src\lang\spanish_MX.txt">
+      <Message Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">Generating spanish_MX language file</Message>
+      <Command Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\objs\strgen\strgen.exe -s ..\src\lang -d ..\bin\lang "%(FullPath)"</Command>
+      <AdditionalInputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\src\lang\english.txt;..\objs\strgen\strgen.exe;%(AdditionalInputs)</AdditionalInputs>
+      <Outputs Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\bin\lang\spanish_MX.lng;%(Outputs)</Outputs>
+    </CustomBuild>
     <CustomBuild Include="..\src\lang\swedish.txt">
       <Message Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">Generating swedish language file</Message>
       <Command Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">..\objs\strgen\strgen.exe -s ..\src\lang -d ..\bin\lang "%(FullPath)"</Command>
diff --git a/projects/langs_vs141.vcxproj.filters b/projects/langs_vs141.vcxproj.filters
index 0afc8c9..bf5257e 100644
--- a/projects/langs_vs141.vcxproj.filters
+++ b/projects/langs_vs141.vcxproj.filters
@@ -142,10 +142,10 @@
     <CustomBuild Include="..\src\lang\slovenian.txt">
       <Filter>Translations</Filter>
     </CustomBuild>
-    <CustomBuild Include="..\src\lang\spanish_MX.txt">
+    <CustomBuild Include="..\src\lang\spanish.txt">
       <Filter>Translations</Filter>
     </CustomBuild>
-    <CustomBuild Include="..\src\lang\spanish.txt">
+    <CustomBuild Include="..\src\lang\spanish_MX.txt">
       <Filter>Translations</Filter>
     </CustomBuild>
     <CustomBuild Include="..\src\lang\swedish.txt">
diff --git a/projects/langs_vs80.vcproj b/projects/langs_vs80.vcproj
index e683d57..909613b 100644
--- a/projects/langs_vs80.vcproj
+++ b/projects/langs_vs80.vcproj
@@ -739,32 +739,32 @@
 				</FileConfiguration>
 			</File>
 			<File
-				RelativePath="..\src\lang\spanish_MX.txt"
+				RelativePath="..\src\lang\spanish.txt"
 				>
 				<FileConfiguration
 					Name="Debug|Win32"
 					>
 					<Tool
 						Name="VCCustomBuildTool"
-						Description="Generating spanish_MX language file"
+						Description="Generating spanish language file"
 						CommandLine="..\objs\strgen\strgen.exe -s ..\src\lang -d ..\bin\lang &quot;$(InputPath)&quot;&#x0D;&#x0A;exit 0&#x0D;&#x0A;"
 						AdditionalDependencies="..\src\lang\english.txt;..\objs\strgen\strgen.exe"
-						Outputs="..\bin\lang\spanish_MX.lng"
+						Outputs="..\bin\lang\spanish.lng"
 					/>
 				</FileConfiguration>
 			</File>
 			<File
-				RelativePath="..\src\lang\spanish.txt"
+				RelativePath="..\src\lang\spanish_MX.txt"
 				>
 				<FileConfiguration
 					Name="Debug|Win32"
 					>
 					<Tool
 						Name="VCCustomBuildTool"
-						Description="Generating spanish language file"
+						Description="Generating spanish_MX language file"
 						CommandLine="..\objs\strgen\strgen.exe -s ..\src\lang -d ..\bin\lang &quot;$(InputPath)&quot;&#x0D;&#x0A;exit 0&#x0D;&#x0A;"
 						AdditionalDependencies="..\src\lang\english.txt;..\objs\strgen\strgen.exe"
-						Outputs="..\bin\lang\spanish.lng"
+						Outputs="..\bin\lang\spanish_MX.lng"
 					/>
 				</FileConfiguration>
 			</File>
diff --git a/projects/langs_vs90.vcproj b/projects/langs_vs90.vcproj
index c1f0282..583e0f9 100644
--- a/projects/langs_vs90.vcproj
+++ b/projects/langs_vs90.vcproj
@@ -740,32 +740,32 @@
 				</FileConfiguration>
 			</File>
 			<File
-				RelativePath="..\src\lang\spanish_MX.txt"
+				RelativePath="..\src\lang\spanish.txt"
 				>
 				<FileConfiguration
 					Name="Debug|Win32"
 					>
 					<Tool
 						Name="VCCustomBuildTool"
-						Description="Generating spanish_MX language file"
+						Description="Generating spanish language file"
 						CommandLine="..\objs\strgen\strgen.exe -s ..\src\lang -d ..\bin\lang &quot;$(InputPath)&quot;&#x0D;&#x0A;exit 0&#x0D;&#x0A;"
 						AdditionalDependencies="..\src\lang\english.txt;..\objs\strgen\strgen.exe"
-						Outputs="..\bin\lang\spanish_MX.lng"
+						Outputs="..\bin\lang\spanish.lng"
 					/>
 				</FileConfiguration>
 			</File>
 			<File
-				RelativePath="..\src\lang\spanish.txt"
+				RelativePath="..\src\lang\spanish_MX.txt"
 				>
 				<FileConfiguration
 					Name="Debug|Win32"
 					>
 					<Tool
 						Name="VCCustomBuildTool"
-						Description="Generating spanish language file"
+						Description="Generating spanish_MX language file"
 						CommandLine="..\objs\strgen\strgen.exe -s ..\src\lang -d ..\bin\lang &quot;$(InputPath)&quot;&#x0D;&#x0A;exit 0&#x0D;&#x0A;"
 						AdditionalDependencies="..\src\lang\english.txt;..\objs\strgen\strgen.exe"
-						Outputs="..\bin\lang\spanish.lng"
+						Outputs="..\bin\lang\spanish_MX.lng"
 					/>
 				</FileConfiguration>
 			</File>
diff --git a/projects/openttd_vs100.vcxproj b/projects/openttd_vs100.vcxproj
index 9c3b107..29a37f4 100644
--- a/projects/openttd_vs100.vcxproj
+++ b/projects/openttd_vs100.vcxproj
@@ -557,6 +557,7 @@
     <ClInclude Include="..\src\news_func.h" />
     <ClInclude Include="..\src\news_gui.h" />
     <ClInclude Include="..\src\news_type.h" />
+    <ClInclude Include="..\src\music\midifile.hpp" />
     <ClInclude Include="..\src\music\null_m.h" />
     <ClInclude Include="..\src\sound\null_s.h" />
     <ClInclude Include="..\src\video\null_v.h" />
@@ -1274,6 +1275,7 @@
     <ClCompile Include="..\src\video\win32_v.cpp" />
     <ClCompile Include="..\src\music\dmusic.cpp" />
     <ClCompile Include="..\src\music\null_m.cpp" />
+    <ClCompile Include="..\src\music\midifile.cpp" />
     <ClCompile Include="..\src\music\win32_m.cpp" />
     <ClCompile Include="..\src\sound\null_s.cpp" />
     <ClCompile Include="..\src\sound\sdl_s.cpp" />
diff --git a/projects/openttd_vs100.vcxproj.filters b/projects/openttd_vs100.vcxproj.filters
index 06800ff..a77ef9b 100644
--- a/projects/openttd_vs100.vcxproj.filters
+++ b/projects/openttd_vs100.vcxproj.filters
@@ -864,6 +864,9 @@
     <ClInclude Include="..\src\news_type.h">
       <Filter>Header Files</Filter>
     </ClInclude>
+    <ClInclude Include="..\src\music\midifile.hpp">
+      <Filter>Header Files</Filter>
+    </ClInclude>
     <ClInclude Include="..\src\music\null_m.h">
       <Filter>Header Files</Filter>
     </ClInclude>
@@ -3015,6 +3018,9 @@
     <ClCompile Include="..\src\music\null_m.cpp">
       <Filter>Music</Filter>
     </ClCompile>
+    <ClCompile Include="..\src\music\midifile.cpp">
+      <Filter>Music</Filter>
+    </ClCompile>
     <ClCompile Include="..\src\music\win32_m.cpp">
       <Filter>Music</Filter>
     </ClCompile>
diff --git a/projects/openttd_vs140.vcxproj b/projects/openttd_vs140.vcxproj
index 2aefabd..0a4587a 100644
--- a/projects/openttd_vs140.vcxproj
+++ b/projects/openttd_vs140.vcxproj
@@ -574,6 +574,7 @@
     <ClInclude Include="..\src\news_func.h" />
     <ClInclude Include="..\src\news_gui.h" />
     <ClInclude Include="..\src\news_type.h" />
+    <ClInclude Include="..\src\music\midifile.hpp" />
     <ClInclude Include="..\src\music\null_m.h" />
     <ClInclude Include="..\src\sound\null_s.h" />
     <ClInclude Include="..\src\video\null_v.h" />
@@ -1291,6 +1292,7 @@
     <ClCompile Include="..\src\video\win32_v.cpp" />
     <ClCompile Include="..\src\music\dmusic.cpp" />
     <ClCompile Include="..\src\music\null_m.cpp" />
+    <ClCompile Include="..\src\music\midifile.cpp" />
     <ClCompile Include="..\src\music\win32_m.cpp" />
     <ClCompile Include="..\src\sound\null_s.cpp" />
     <ClCompile Include="..\src\sound\sdl_s.cpp" />
diff --git a/projects/openttd_vs140.vcxproj.filters b/projects/openttd_vs140.vcxproj.filters
index 06800ff..a77ef9b 100644
--- a/projects/openttd_vs140.vcxproj.filters
+++ b/projects/openttd_vs140.vcxproj.filters
@@ -864,6 +864,9 @@
     <ClInclude Include="..\src\news_type.h">
       <Filter>Header Files</Filter>
     </ClInclude>
+    <ClInclude Include="..\src\music\midifile.hpp">
+      <Filter>Header Files</Filter>
+    </ClInclude>
     <ClInclude Include="..\src\music\null_m.h">
       <Filter>Header Files</Filter>
     </ClInclude>
@@ -3015,6 +3018,9 @@
     <ClCompile Include="..\src\music\null_m.cpp">
       <Filter>Music</Filter>
     </ClCompile>
+    <ClCompile Include="..\src\music\midifile.cpp">
+      <Filter>Music</Filter>
+    </ClCompile>
     <ClCompile Include="..\src\music\win32_m.cpp">
       <Filter>Music</Filter>
     </ClCompile>
diff --git a/projects/openttd_vs141.vcxproj b/projects/openttd_vs141.vcxproj
index c735569..38746db 100644
--- a/projects/openttd_vs141.vcxproj
+++ b/projects/openttd_vs141.vcxproj
@@ -574,6 +574,7 @@
     <ClInclude Include="..\src\news_func.h" />
     <ClInclude Include="..\src\news_gui.h" />
     <ClInclude Include="..\src\news_type.h" />
+    <ClInclude Include="..\src\music\midifile.hpp" />
     <ClInclude Include="..\src\music\null_m.h" />
     <ClInclude Include="..\src\sound\null_s.h" />
     <ClInclude Include="..\src\video\null_v.h" />
@@ -1291,6 +1292,7 @@
     <ClCompile Include="..\src\video\win32_v.cpp" />
     <ClCompile Include="..\src\music\dmusic.cpp" />
     <ClCompile Include="..\src\music\null_m.cpp" />
+    <ClCompile Include="..\src\music\midifile.cpp" />
     <ClCompile Include="..\src\music\win32_m.cpp" />
     <ClCompile Include="..\src\sound\null_s.cpp" />
     <ClCompile Include="..\src\sound\sdl_s.cpp" />
diff --git a/projects/openttd_vs141.vcxproj.filters b/projects/openttd_vs141.vcxproj.filters
index 06800ff..a77ef9b 100644
--- a/projects/openttd_vs141.vcxproj.filters
+++ b/projects/openttd_vs141.vcxproj.filters
@@ -864,6 +864,9 @@
     <ClInclude Include="..\src\news_type.h">
       <Filter>Header Files</Filter>
     </ClInclude>
+    <ClInclude Include="..\src\music\midifile.hpp">
+      <Filter>Header Files</Filter>
+    </ClInclude>
     <ClInclude Include="..\src\music\null_m.h">
       <Filter>Header Files</Filter>
     </ClInclude>
@@ -3015,6 +3018,9 @@
     <ClCompile Include="..\src\music\null_m.cpp">
       <Filter>Music</Filter>
     </ClCompile>
+    <ClCompile Include="..\src\music\midifile.cpp">
+      <Filter>Music</Filter>
+    </ClCompile>
     <ClCompile Include="..\src\music\win32_m.cpp">
       <Filter>Music</Filter>
     </ClCompile>
diff --git a/projects/openttd_vs80.vcproj b/projects/openttd_vs80.vcproj
index 70dcab2..7ccf6ea 100644
--- a/projects/openttd_vs80.vcproj
+++ b/projects/openttd_vs80.vcproj
@@ -1455,6 +1455,10 @@
 				>
 			</File>
 			<File
+				RelativePath=".\..\src\music\midifile.hpp"
+				>
+			</File>
+			<File
 				RelativePath=".\..\src\music\null_m.h"
 				>
 			</File>
@@ -4435,6 +4439,10 @@
 				>
 			</File>
 			<File
+				RelativePath=".\..\src\music\midifile.cpp"
+				>
+			</File>
+			<File
 				RelativePath=".\..\src\music\win32_m.cpp"
 				>
 			</File>
diff --git a/projects/openttd_vs90.vcproj b/projects/openttd_vs90.vcproj
index dd722d0..d42d69b 100644
--- a/projects/openttd_vs90.vcproj
+++ b/projects/openttd_vs90.vcproj
@@ -1452,6 +1452,10 @@
 				>
 			</File>
 			<File
+				RelativePath=".\..\src\music\midifile.hpp"
+				>
+			</File>
+			<File
 				RelativePath=".\..\src\music\null_m.h"
 				>
 			</File>
@@ -4432,6 +4436,10 @@
 				>
 			</File>
 			<File
+				RelativePath=".\..\src\music\midifile.cpp"
+				>
+			</File>
+			<File
 				RelativePath=".\..\src\music\win32_m.cpp"
 				>
 			</File>
diff --git a/projects/settings_vs100.vcxproj b/projects/settings_vs100.vcxproj
index 61b2897..20e799d 100644
--- a/projects/settings_vs100.vcxproj
+++ b/projects/settings_vs100.vcxproj
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup Label="ProjectConfigurations">
     <ProjectConfiguration Include="Debug|Win32">
diff --git a/projects/settings_vs100.vcxproj.filters b/projects/settings_vs100.vcxproj.filters
index 8caf9fa..40fe848 100644
--- a/projects/settings_vs100.vcxproj.filters
+++ b/projects/settings_vs100.vcxproj.filters
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup>
     <Filter Include="INI">
diff --git a/projects/settings_vs140.vcxproj b/projects/settings_vs140.vcxproj
index 4fe07db..65563f3 100644
--- a/projects/settings_vs140.vcxproj
+++ b/projects/settings_vs140.vcxproj
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup Label="ProjectConfigurations">
     <ProjectConfiguration Include="Debug|Win32">
diff --git a/projects/settings_vs140.vcxproj.filters b/projects/settings_vs140.vcxproj.filters
index 8caf9fa..40fe848 100644
--- a/projects/settings_vs140.vcxproj.filters
+++ b/projects/settings_vs140.vcxproj.filters
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup>
     <Filter Include="INI">
diff --git a/projects/settings_vs141.vcxproj b/projects/settings_vs141.vcxproj
index ed6f5ba..26e532c 100644
--- a/projects/settings_vs141.vcxproj
+++ b/projects/settings_vs141.vcxproj
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup Label="ProjectConfigurations">
     <ProjectConfiguration Include="Debug|Win32">
diff --git a/projects/settings_vs141.vcxproj.filters b/projects/settings_vs141.vcxproj.filters
index 8caf9fa..40fe848 100644
--- a/projects/settings_vs141.vcxproj.filters
+++ b/projects/settings_vs141.vcxproj.filters
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup>
     <Filter Include="INI">
diff --git a/projects/settingsgen_vs100.vcxproj b/projects/settingsgen_vs100.vcxproj
index 30454d8..5c45b64 100644
--- a/projects/settingsgen_vs100.vcxproj
+++ b/projects/settingsgen_vs100.vcxproj
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup Label="ProjectConfigurations">
     <ProjectConfiguration Include="Debug|Win32">
diff --git a/projects/settingsgen_vs100.vcxproj.filters b/projects/settingsgen_vs100.vcxproj.filters
index c8afe0c..f4a69ad 100644
--- a/projects/settingsgen_vs100.vcxproj.filters
+++ b/projects/settingsgen_vs100.vcxproj.filters
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup>
     <Filter Include="Source Files">
diff --git a/projects/settingsgen_vs140.vcxproj b/projects/settingsgen_vs140.vcxproj
index fc6d079..705b219 100644
--- a/projects/settingsgen_vs140.vcxproj
+++ b/projects/settingsgen_vs140.vcxproj
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup Label="ProjectConfigurations">
     <ProjectConfiguration Include="Debug|Win32">
diff --git a/projects/settingsgen_vs140.vcxproj.filters b/projects/settingsgen_vs140.vcxproj.filters
index c8afe0c..f4a69ad 100644
--- a/projects/settingsgen_vs140.vcxproj.filters
+++ b/projects/settingsgen_vs140.vcxproj.filters
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup>
     <Filter Include="Source Files">
diff --git a/projects/settingsgen_vs141.vcxproj b/projects/settingsgen_vs141.vcxproj
index 2b356f2..ad6a0dc 100644
--- a/projects/settingsgen_vs141.vcxproj
+++ b/projects/settingsgen_vs141.vcxproj
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup Label="ProjectConfigurations">
     <ProjectConfiguration Include="Debug|Win32">
diff --git a/projects/settingsgen_vs141.vcxproj.filters b/projects/settingsgen_vs141.vcxproj.filters
index c8afe0c..f4a69ad 100644
--- a/projects/settingsgen_vs141.vcxproj.filters
+++ b/projects/settingsgen_vs141.vcxproj.filters
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup>
     <Filter Include="Source Files">
diff --git a/projects/strgen_vs140.vcxproj b/projects/strgen_vs140.vcxproj
index ebf56a4..624c025 100644
--- a/projects/strgen_vs140.vcxproj
+++ b/projects/strgen_vs140.vcxproj
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup Label="ProjectConfigurations">
     <ProjectConfiguration Include="Debug|Win32">
diff --git a/projects/strgen_vs141.vcxproj b/projects/strgen_vs141.vcxproj
index addc6f1..b8fb4da 100644
--- a/projects/strgen_vs141.vcxproj
+++ b/projects/strgen_vs141.vcxproj
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup Label="ProjectConfigurations">
     <ProjectConfiguration Include="Debug|Win32">
diff --git a/projects/version_vs100.vcxproj b/projects/version_vs100.vcxproj
index 008e6dd..b3ab78d 100644
--- a/projects/version_vs100.vcxproj
+++ b/projects/version_vs100.vcxproj
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup Label="ProjectConfigurations">
     <ProjectConfiguration Include="Debug|Win32">
diff --git a/projects/version_vs140.vcxproj b/projects/version_vs140.vcxproj
index baf374b..a514d84 100644
--- a/projects/version_vs140.vcxproj
+++ b/projects/version_vs140.vcxproj
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup Label="ProjectConfigurations">
     <ProjectConfiguration Include="Debug|Win32">
diff --git a/projects/version_vs141.vcxproj b/projects/version_vs141.vcxproj
index 5d8ee8c..adc1405 100644
--- a/projects/version_vs141.vcxproj
+++ b/projects/version_vs141.vcxproj
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup Label="ProjectConfigurations">
     <ProjectConfiguration Include="Debug|Win32">
