Index: src/rail_cmd.cpp
===================================================================
--- src/rail_cmd.cpp	(revision 27653)
+++ src/rail_cmd.cpp	(working copy)
@@ -557,12 +557,11 @@
 			if (ret.Failed()) return ret;
 			cost.AddCost(ret);
 
-			ret = DoCommand(tile, 0, 0, flags, CMD_LANDSCAPE_CLEAR);
-			if (ret.Failed()) return ret;
-			cost.AddCost(ret);
-
-			if (water_ground) {
-				cost.AddCost(-_price[PR_CLEAR_WATER]);
+			if (!water_ground) {
+				ret = DoCommand(tile, 0, 0, flags, CMD_LANDSCAPE_CLEAR);
+				if (ret.Failed()) return ret;
+				cost.AddCost(ret);
+			} else {
 				cost.AddCost(_price[PR_CLEAR_ROUGH]);
 			}
 
