diff --git a/src/currency.cpp b/src/currency.cpp
index 8cc385f..5dcc0cc 100644
--- a/src/currency.cpp
+++ b/src/currency.cpp
@@ -62,6 +62,7 @@
 	{    1, "", CF_NOEURO, "",             "",               2, STR_GAME_OPTIONS_CURRENCY_CUSTOM }, ///< custom currency (add further languages below)
 	{    3, "", CF_NOEURO, "",             NBSP "GEL",       1, STR_GAME_OPTIONS_CURRENCY_GEL    }, ///< Georgian Lari
 	{ 4901, "", CF_NOEURO, "",             NBSP "Rls",       1, STR_GAME_OPTIONS_CURRENCY_IRR    }, ///< Iranian Rial
+	{    3, "", CF_NOEURO, "",             NBSP "mBTC",      1, STR_GAME_OPTIONS_CURRENCY_BTC    }, ///< Bitcoin cryptocurrency
 };
 
 /** Array of currencies used by the system */
diff --git a/src/currency.h b/src/currency.h
index ab80260..84d09e9 100644
--- a/src/currency.h
+++ b/src/currency.h
@@ -58,6 +58,7 @@
 	CURRENCY_CUSTOM,    ///< Custom currency
 	CURRENCY_GEL,       ///< Georgian Lari
 	CURRENCY_IRR,       ///< Iranian Rial
+	CURRENCY_BTC,       ///< Bitcoin cryptocurrency
 	CURRENCY_END,       ///< always the last item
 };
 
diff --git a/src/lang/english.txt b/src/lang/english.txt
index 0bd430b..8778300 100644
--- a/src/lang/english.txt
+++ b/src/lang/english.txt
@@ -925,6 +925,7 @@
 STR_GAME_OPTIONS_CURRENCY_CUSTOM                                :Custom...
 STR_GAME_OPTIONS_CURRENCY_GEL                                   :Georgian Lari (GEL)
 STR_GAME_OPTIONS_CURRENCY_IRR                                   :Iranian Rial (IRR)
+STR_GAME_OPTIONS_CURRENCY_BTC                                   :Bitcoin (BTC)
 ############ end of currency region
 
 STR_GAME_OPTIONS_ROAD_VEHICLES_FRAME                            :{BLACK}Road vehicles
