--- openttd-1.2.2-base/src/video/sdl_v.cpp	2012-08-29 15:49:16.000000000 -0600
+++ openttd-1.2.2/src/video/sdl_v.cpp	2012-08-29 17:30:25.000000000 -0600
@@ -25,6 +25,7 @@
 #include "../fileio_func.h"
 #include "sdl_v.h"
 #include <SDL.h>
+#include <SDL_syswm.h>
 
 static FVideoDriver_SDL iFVideoDriver_SDL;
 
@@ -382,7 +383,26 @@
 #elif defined(__sgi__)
 	if (sym->scancode == 22) key = WKC_BACKQUOTE;
 #else
-	if (sym->scancode == 49) key = WKC_BACKQUOTE;
+#ifdef SDL_VIDEO_DRIVER_X11
+	static bool is_X11 = false, checked_X11 = false;
+	if (!checked_X11) {
+		SDL_SysWMinfo info;
+		SDL_VERSION(&info.version);
+		if (SDL_GetWMInfo(&info)) {
+			is_X11 = (info.subsystem == SDL_SYSWM_X11);
+		}
+		else is_X11 = false;
+		checked_X11 = true;
+	}
+	if (is_X11) {
+		if(sym->scancode == 49) key = WKC_BACKQUOTE;
+	}
+	else
+	// if X11 is supported by the version of SDL we compile against, use
+	// the keysym-based code only if we are NOT using X11
+#endif
+	// some drivers (like DirectFB) don't have predictable scancodes :/
+	if(sym->sym == SDLK_BACKQUOTE) key = WKC_BACKQUOTE;
 #endif
 
 	/* META are the command keys on mac */
@@ -476,6 +496,13 @@
 			if ((ev.key.keysym.mod & (KMOD_ALT | KMOD_META)) &&
 					(ev.key.keysym.sym == SDLK_RETURN || ev.key.keysym.sym == SDLK_f)) {
 				ToggleFullScreen(!_fullscreen);
+			} else if (((ev.key.keysym.sym == SDLK_LCTRL || ev.key.keysym.sym == SDLK_RCTRL) && ev.key.keysym.mod & (KMOD_ALT)) ||
+					((ev.key.keysym.sym == SDLK_LALT || ev.key.keysym.sym == SDLK_RALT) && ev.key.keysym.mod & (KMOD_CTRL))) {
+				// force a redraw of the entire screen when control
+				// and alt are pressed to work around a bug in
+				// the DirectFB backend and possibly others
+				_num_dirty_rects = MAX_DIRTY_RECTS + 1;
+				HandleKeypress(ConvertSdlKeyIntoMy(&ev.key.keysym));
 			} else {
 				HandleKeypress(ConvertSdlKeyIntoMy(&ev.key.keysym));
 			}
