Index: src/ai/ai_gui.cpp
===================================================================
--- src/ai/ai_gui.cpp	(revision 19763)
+++ src/ai/ai_gui.cpp	(working copy)
@@ -1014,7 +1014,6 @@
 			CompanyID old_company = _current_company;
 			_current_company = ai_debug_company;
 			AILog::LogData *log = (AILog::LogData *)AIObject::GetLogPointer();
-			_current_company = old_company;
 
 			if (log != NULL && case_sensitive_break_check?
 					strstr(log->lines[log->pos], this->edit_str_buf) != 0 :
@@ -1032,6 +1031,8 @@
 				/* Highlight row that matched */
 				this->highlight_row = log->pos;
 			}
+
+			_current_company = old_company;
 		}
 	}
 
