Index: src/video/win32_v.cpp
===================================================================
--- src/video/win32_v.cpp	(revision 10783)
+++ src/video/win32_v.cpp	(working copy)
@@ -14,6 +14,8 @@
 #include <windows.h>
 #include <tchar.h>
 
+//static bool _stupid_fs998 = false;
+
 static struct {
 	HWND main_wnd;
 	HBITMAP dib_sect;
@@ -145,6 +147,7 @@
 
 static void ClientSizeChanged(int w, int h)
 {
+	printf("csc: %i, %i\n", w, h);
 	// allocate new dib section of the new size
 	if (AllocateDibSection(w, h)) {
 		// mark all palette colors dirty
@@ -301,6 +304,10 @@
 	static uint32 keycode = 0;
 	static bool console = false;
 
+	//if (_stupid_fs998) {
+		printf("window: %i, %x, %x\n", msg, wParam, lParam);
+	//}
+
 	switch (msg) {
 		case WM_CREATE:
 			SetTimer(hwnd, TID_POLLMOUSE, MOUSE_POLL_DELAY, (TIMERPROC)TrackMouseTimerProc);
@@ -795,7 +802,10 @@
 	DestroyWindow(_wnd.main_wnd);
 
 #if !defined(WINCE)
-	if (_wnd.fullscreen) ChangeDisplaySettings(NULL, 0);
+	if (_wnd.fullscreen) {
+//		_stupid_fs998 = true;
+		ChangeDisplaySettings(NULL, 0);
+	}
 #endif
 	MyShowCursor(true);
 }
@@ -889,6 +899,7 @@
 
 bool VideoDriver_Win32::ChangeResolution(int w, int h)
 {
+	printf("cr: %i, %i\n", w, h);
 	_wnd.width = _wnd.width_org = w;
 	_wnd.height = _wnd.height_org = h;
 
