Index: src/lang/english.txt
===================================================================
--- src/lang/english.txt	(revision 27220)
+++ src/lang/english.txt	(working copy)
@@ -3660,7 +3660,7 @@
 STR_VEHICLE_INFO_WEIGHT_POWER_MAX_SPEED                         :{BLACK}Weight: {LTBLUE}{WEIGHT_SHORT} {BLACK}Power: {LTBLUE}{POWER}{BLACK} Max. speed: {LTBLUE}{VELOCITY}
 STR_VEHICLE_INFO_WEIGHT_POWER_MAX_SPEED_MAX_TE                  :{BLACK}Weight: {LTBLUE}{WEIGHT_SHORT} {BLACK}Power: {LTBLUE}{POWER}{BLACK} Max. speed: {LTBLUE}{VELOCITY} {BLACK}Max. T.E.: {LTBLUE}{FORCE}
 
-STR_VEHICLE_INFO_PROFIT_THIS_YEAR_LAST_YEAR                     :{BLACK}Profit this year: {LTBLUE}{CURRENCY_LONG} (last year: {CURRENCY_LONG})
+STR_VEHICLE_INFO_PROFIT_THIS_YEAR_LAST_YEAR                     :{BLACK}Profit this year: {LTBLUE}{CURRENCY_LONG} (last year: {CURRENCY_LONG}) (lifetime: {CURRENCY_LONG})
 STR_VEHICLE_INFO_RELIABILITY_BREAKDOWNS                         :{BLACK}Reliability: {LTBLUE}{COMMA}%  {BLACK}Breakdowns since last service: {LTBLUE}{COMMA}
 
 STR_VEHICLE_INFO_BUILT_VALUE                                    :{LTBLUE}{ENGINE} {BLACK}Built: {LTBLUE}{NUM}{BLACK} Value: {LTBLUE}{CURRENCY_LONG}
Index: src/saveload/afterload.cpp
===================================================================
--- src/saveload/afterload.cpp	(revision 27220)
+++ src/saveload/afterload.cpp	(working copy)
@@ -2972,6 +2972,12 @@
 		FOR_ALL_STATIONS(st) UpdateStationAcceptance(st, false);
 	}
 
+	/* Set lifetime vehicle profit to 0 if save game before 195 */
+	if (IsSavegameVersionBefore(195)) {
+		Vehicle *v;
+		FOR_ALL_VEHICLES(v) v->profit_lifetime = 0;
+	}
+
 	/* Road stops is 'only' updating some caches */
 	AfterLoadRoadStops();
 	AfterLoadLabelMaps();
Index: src/saveload/saveload.cpp
===================================================================
--- src/saveload/saveload.cpp	(revision 27220)
+++ src/saveload/saveload.cpp	(working copy)
@@ -263,7 +263,7 @@
  *  193   26802
  *  194   26881   1.5.x
  */
-extern const uint16 SAVEGAME_VERSION = 194; ///< Current savegame version of OpenTTD.
+extern const uint16 SAVEGAME_VERSION = 195; ///< Current savegame version of OpenTTD.
 
 SavegameType _savegame_type; ///< type of savegame we are loading
 
Index: src/saveload/vehicle_sl.cpp
===================================================================
--- src/saveload/vehicle_sl.cpp	(revision 27220)
+++ src/saveload/vehicle_sl.cpp	(working copy)
@@ -683,6 +683,7 @@
 		 SLE_CONDVAR(Vehicle, profit_this_year,      SLE_INT64,                   65, SL_MAX_VERSION),
 		 SLE_CONDVAR(Vehicle, profit_last_year,      SLE_FILE_I32 | SLE_VAR_I64,   0,  64),
 		 SLE_CONDVAR(Vehicle, profit_last_year,      SLE_INT64,                   65, SL_MAX_VERSION),
+		 SLE_CONDVAR(Vehicle, profit_lifetime,       SLE_INT64,                   195, SL_MAX_VERSION),
 		SLEG_CONDVAR(         _cargo_feeder_share,   SLE_FILE_I32 | SLE_VAR_I64,  51,  64),
 		SLEG_CONDVAR(         _cargo_feeder_share,   SLE_INT64,                   65,  67),
 		SLEG_CONDVAR(         _cargo_loaded_at_xy,   SLE_UINT32,                  51,  67),
Index: src/vehicle.cpp
===================================================================
--- src/vehicle.cpp	(revision 27220)
+++ src/vehicle.cpp	(working copy)
@@ -2650,6 +2650,7 @@
 			}
 
 			v->profit_last_year = v->profit_this_year;
+			v->profit_lifetime += v->profit_this_year;
 			v->profit_this_year = 0;
 			SetWindowDirty(WC_VEHICLE_DETAILS, v->index);
 		}
Index: src/vehicle_base.h
===================================================================
--- src/vehicle_base.h	(revision 27220)
+++ src/vehicle_base.h	(working copy)
@@ -180,6 +180,7 @@
 
 	Money profit_this_year;             ///< Profit this year << 8, low 8 bits are fract
 	Money profit_last_year;             ///< Profit last year << 8, low 8 bits are fract
+	Money profit_lifetime;              ///< Profit lifetime << 8, low 8 bits are fract
 	Money value;                        ///< Value of the vehicle
 
 	CargoPayment *cargo_payment;        ///< The cargo payment we're currently in
@@ -513,6 +514,12 @@
 	 */
 	Money GetDisplayProfitLastYear() const { return (this->profit_last_year >> 8); }
 
+	/**
+	 * Gets the lifetime profit of vehicle. It can be sent into SetDParam for string processing.
+	 * @return the vehicle's lifetime profit
+	 */
+	Money GetDisplayProfitLifetime() const { return (this->profit_lifetime >> 8); }
+	
 	void SetNext(Vehicle *next);
 
 	/**
Index: src/vehicle_gui.cpp
===================================================================
--- src/vehicle_gui.cpp	(revision 27220)
+++ src/vehicle_gui.cpp	(working copy)
@@ -2075,6 +2075,7 @@
 				/* Draw profit */
 				SetDParam(0, v->GetDisplayProfitThisYear());
 				SetDParam(1, v->GetDisplayProfitLastYear());
+				SetDParam(2, v->GetDisplayProfitLifetime());
 				DrawString(r.left + WD_FRAMERECT_LEFT, r.right - WD_FRAMERECT_RIGHT, y, STR_VEHICLE_INFO_PROFIT_THIS_YEAR_LAST_YEAR);
 				y += FONT_HEIGHT_NORMAL;
 
