Index: src/genworld.cpp
===================================================================
--- src/genworld.cpp	(Revision 11546)
+++ src/genworld.cpp	(Arbeitskopie)
@@ -260,6 +260,9 @@
 	/* Set the date before loading sprites as some newgrfs check it */
 	SetDate(ConvertYMDToDate(_patches.starting_year, 0, 1));
 
+	/* The player has got 100 years to get enough points for the highscore */
+	_patches.ending_year = _patches.starting_year + 100;
+
 	/* Load the right landscape stuff */
 	GfxLoadSprites();
 	LoadStringWidthTable();
