Index: src/texteff.cpp
===================================================================
--- src/texteff.cpp	(revision 11051)
+++ src/texteff.cpp	(working copy)
@@ -259,6 +259,7 @@
 /** Text Effects */
 static void MarkTextEffectAreaDirty(TextEffect *te)
 {
+	/* Width and height of the text effect are doubled, so they are correct in both zoom out levels 1x and 2x. */
 	MarkAllViewportsDirty(
 		te->x,
 		te->y - 1,
@@ -321,6 +322,15 @@
 	te->params_1 = GetDParam(0);
 	te->params_2 = GetDParam(4);
 
+	/* Update width of text effect */
+	char buffer[100];
+	GetString(buffer, msg, lastof(buffer));
+	int w = GetStringBoundingBox(buffer).width;
+
+	/* Only allow to make it broader, so it completely covers the old text. That avoids remnants of the old text. */
+	int right_new = te->x + w;
+	if (te->right < right_new) te->right = right_new;
+
 	MarkTextEffectAreaDirty(te);
 }
 
