Index: Makefile.src.in
===================================================================
--- Makefile.src.in	(revision 12309)
+++ Makefile.src.in	(working copy)
@@ -125,14 +125,14 @@
 endif
 
 # This helps to recompile if flags change
-RES := $(shell if [ "`cat $(CONFIG_CACHE_COMPILER) 2>/dev/null`" != "$(CC_CFLAGS) $(CFLAGS)" ]; then echo "$(CC_CFLAGS) $(CFLAGS)" > $(CONFIG_CACHE_COMPILER); fi )
-RES := $(shell if [ "`cat $(CONFIG_CACHE_LINKER) 2>/dev/null`" != "$(LDFLAGS) $(LIBS)" ]; then echo "$(LDFLAGS) $(LIBS)" > $(CONFIG_CACHE_LINKER); fi )
-RES := $(shell if [ "`cat $(CONFIG_CACHE_ENDIAN) 2>/dev/null`" != "$(ENDIAN_FORCE)" ]; then echo "$(ENDIAN_FORCE)" > $(CONFIG_CACHE_ENDIAN); fi )
+RES := $(shell if [ '`cat $(CONFIG_CACHE_COMPILER) 2>/dev/null`' != '$(CC_CFLAGS) $(CFLAGS)' ]; then echo '$(CC_CFLAGS) $(CFLAGS)' > $(CONFIG_CACHE_COMPILER); fi )
+RES := $(shell if [ '`cat $(CONFIG_CACHE_LINKER) 2>/dev/null`' != '$(LDFLAGS) $(LIBS)' ]; then echo '$(LDFLAGS) $(LIBS)' > $(CONFIG_CACHE_LINKER); fi )
+RES := $(shell if [ '`cat $(CONFIG_CACHE_ENDIAN) 2>/dev/null`' != '$(ENDIAN_FORCE)' ]; then echo '$(ENDIAN_FORCE)' > $(CONFIG_CACHE_ENDIAN); fi )
 
 # If there is a change in the source-file-list, make sure we recheck the deps
-RES := $(shell if [ "`cat $(CONFIG_CACHE_SOURCE) 2>/dev/null`" != "$(SRCS)" ]; then echo "$(SRCS)" > $(CONFIG_CACHE_SOURCE); fi )
+RES := $(shell if [ '`cat $(CONFIG_CACHE_SOURCE) 2>/dev/null`' != '$(SRCS)' ]; then echo '$(SRCS)' > $(CONFIG_CACHE_SOURCE); fi )
 # If there is a change in the revision, make sure we recompile rev.cpp
-RES := $(shell if [ "`cat $(CONFIG_CACHE_VERSION) 2>/dev/null`" != "$(REV)" ]; then echo "$(REV)" > $(CONFIG_CACHE_VERSION); fi )
+RES := $(shell if [ '`cat $(CONFIG_CACHE_VERSION) 2>/dev/null`' != '$(REV)' ]; then echo '$(REV)' > $(CONFIG_CACHE_VERSION); fi )
 
 ifndef MAKEDEPEND
 # The slow, but always correct, dep-check
Index: config.lib
===================================================================
--- config.lib	(revision 12309)
+++ config.lib	(working copy)
@@ -1146,10 +1146,10 @@
 	fi
 
 	if [ -n "$with_midi" ]; then
-		CFLAGS="$CFLAGS -DEXTERNAL_PLAYER=\\\\\"$with_midi\\\\\""
+		CFLAGS="$CFLAGS -DEXTERNAL_PLAYER=\"\\\\\"$with_midi\\\\\"\""
 	fi
 	if [ -n "$with_midi_arg" ]; then
-		CFLAGS="$CFLAGS -DMIDI_ARG=\\\\\"$with_midi_arg\\\\\""
+		CFLAGS="$CFLAGS -DMIDI_ARG=\"\\\\\"$with_midi_arg\\\\\"\""
 	fi
 
 	if [ "$enable_dedicated" != "0" ]; then
@@ -1200,14 +1200,14 @@
 	fi
 
 	if [ -n "$personal_dir" ]; then
-		CFLAGS="$CFLAGS -DWITH_PERSONAL_DIR -DPERSONAL_DIR=\\\\\"$personal_dir\\\\\""
+		CFLAGS="$CFLAGS -DWITH_PERSONAL_DIR -DPERSONAL_DIR=\"\\\\\"$personal_dir\\\\\"\""
 	fi
 
 	if [ -n "$shared_dir" ]; then
-		CFLAGS="$CFLAGS -DWITH_SHARED_DIR -DSHARED_DIR=\\\\\"$shared_dir\\\\\""
+		CFLAGS="$CFLAGS -DWITH_SHARED_DIR -DSHARED_DIR=\"\\\\\"$shared_dir\\\\\"\""
 	fi
 
-	CFLAGS="$CFLAGS -DGLOBAL_DATA_DIR=\\\\\"$prefix_dir/$data_dir\\\\\""
+	CFLAGS="$CFLAGS -DGLOBAL_DATA_DIR=\"\\\\\"$prefix_dir/$data_dir\\\\\"\""
 
 	log 1 "using CFLAGS... $CFLAGS $CC_CFLAGS"
 	log 1 "using LDFLAGS... $LIBS $LDFLAGS"
@@ -1219,7 +1219,7 @@
 	# Btw, this almost always comes from outside the configure, so it is
 	#  not something we can control.
 	if [ "$with_makedepend" != "0" ]; then
-		cflags_makedep="` echo "$CFLAGS" | sed 's# /# -#g'`"
+		cflags_makedep="` echo '$(CFLAGS)' | sed 's# /# -#g'`"
 	else
 		makedepend=""
 	fi
