Index: src/video/sdl_v.cpp
===================================================================
--- src/video/sdl_v.cpp	(Revision 11101)
+++ src/video/sdl_v.cpp	(Arbeitskopie)
@@ -288,6 +288,7 @@
 	AS(SDLK_QUOTE,   WKC_SINGLEQUOTE),
 	AS(SDLK_COMMA,   WKC_COMMA),
 	AS(SDLK_MINUS,   WKC_MINUS),
+	AS(SDLK_PLUS,    WKC_PLUS),
 	AS(SDLK_PERIOD,  WKC_PERIOD)
 };
 
Index: src/gfx.h
===================================================================
--- src/gfx.h	(Revision 11101)
+++ src/gfx.h	(Arbeitskopie)
@@ -122,6 +122,7 @@
 	WKC_COMMA       = 151, ///< , Comma
 	WKC_PERIOD      = 152, ///< . Period
 	WKC_MINUS       = 153, ///< - Minus
+	WKC_PLUS        = 154, ///< + Plus
 };
 
 enum GameModes {
 
Index: src/main_gui.cpp
===================================================================
--- src/main_gui.cpp	(Revision 11101)
+++ src/main_gui.cpp	(Arbeitskopie)
@@ -1650,6 +1650,7 @@
 		case WKC_SHIFT | WKC_F3: ShowVehicleListWindow(_local_player, VEH_SHIP); break;
 		case WKC_SHIFT | WKC_F4: ShowVehicleListWindow(_local_player, VEH_AIRCRAFT); break;
 		case WKC_NUM_PLUS: // Fall through
+		case WKC_PLUS: // Fall through
 		case WKC_EQUALS: // Fall through
 		case WKC_SHIFT | WKC_EQUALS: // Fall through
 		case WKC_SHIFT | WKC_F5: ToolbarZoomInClick(w); break;
