Index: depot_gui.cpp
===================================================================
--- depot_gui.cpp	(revision 27441)
+++ depot_gui.cpp	(working copy)
@@ -229,6 +229,7 @@
 	VehicleID vehicle_over; ///< Rail vehicle over which another one is dragged, \c INVALID_VEHICLE if none.
 	VehicleType type;
 	bool generate_list;
+	bool sell_hovered;
 	VehicleList vehicle_list;
 	VehicleList wagon_list;
 	uint unitnumber_digits;
@@ -243,6 +244,7 @@
 		this->sel = INVALID_VEHICLE;
 		this->vehicle_over = INVALID_VEHICLE;
 		this->generate_list = true;
+		this->sell_hovered = false;
 		this->type = type;
 		this->num_columns = 1; // for non-trains this gets set in FinishInitNested()
 		this->unitnumber_digits = 2;
@@ -871,7 +873,18 @@
 
 	virtual void OnMouseDrag(Point pt, int widget)
 	{
-		if (this->type != VEH_TRAIN || this->sel == INVALID_VEHICLE) return;
+		if (this->sel == INVALID_VEHICLE) return;
+		if (widget == WID_D_SELL && this->sell_hovered == false) {
+			this->sell_hovered = true;
+			this->SetWidgetLoweredState(WID_D_SELL, true);
+			this->SetWidgetDirty(WID_D_SELL);
+		}
+		if (widget != WID_D_SELL && this->sell_hovered == true) {
+			this->sell_hovered = false;
+			this->SetWidgetLoweredState(WID_D_SELL, false);
+			this->SetWidgetDirty(WID_D_SELL);
+		}
+		if (this->type != VEH_TRAIN) return;
 
 		/* A rail vehicle is dragged.. */
 		if (widget != WID_D_MATRIX) { // ..outside of the depot matrix.
