Index: src/vehicle_gui.cpp
===================================================================
--- src/vehicle_gui.cpp	(revision 14044)
+++ src/vehicle_gui.cpp	(working copy)
@@ -443,7 +443,10 @@
 
 	/* STR_02BD is used to start the string with {BLACK} */
 	SetDParam(0, GetGRFStringID(GetEngineGRFID(engine), 0xD000 + callback));
-	return DrawStringMultiLine(x, y, STR_02BD, w);
+	PrepareTextRefStackUsage(0);
+	uint result = DrawStringMultiLine(x, y, STR_02BD, w);
+	StopTextRefStackUsage();
+	return result;
 }
 
 /** Display list of cargo types of the engine, for the purchase information window */
