From c6d9b63f76d47cff4c8c9062b357f0e4ffae8c8c Mon Sep 17 00:00:00 2001
From: Michael Lutz <michi@icosahedron.de>
Date: Wed, 3 Sep 2008 14:08:08 +0200
Subject: -Fix (r13958): Trains would not look ahead and extend their reservation on a waypoint tile in all cases.

---
 src/train_cmd.cpp |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/train_cmd.cpp b/src/train_cmd.cpp
index e66271a..64f3904 100644
--- a/src/train_cmd.cpp
+++ b/src/train_cmd.cpp
@@ -2327,8 +2327,8 @@ static void CheckNextTrainTile(Vehicle *v)
 	/* Don't do any look-ahead if path_backoff_interval is 255. */
 	if (_settings_game.pf.path_backoff_interval == 255) return;
 
-	/* Exit if we reached our destination or are inside a depot. */
-	if ((v->tile == v->dest_tile && !v->current_order.IsType(OT_GOTO_STATION)) || v->u.rail.track & TRACK_BIT_DEPOT) return;
+	/* Exit if we reached our destination depot or are inside a depot. */
+	if ((v->tile == v->dest_tile && v->current_order.IsType(OT_GOTO_DEPOT)) || v->u.rail.track & TRACK_BIT_DEPOT) return;
 	/* Exit if we are on a station tile and are going to stop. */
 	if (IsRailwayStationTile(v->tile) && v->current_order.ShouldStopAtStation(v, GetStationIndex(v->tile))) return;
 	/* Exit if the current order doesn't have a destination, but the train has orders. */
-- 
1.5.6.4

