Index: src/network/network_client.cpp
===================================================================
--- src/network/network_client.cpp	(revision 20794)
+++ src/network/network_client.cpp	(working copy)
@@ -868,7 +868,7 @@
 	 *  his _local_company value before reconnecting
 	 * COMPANY_SPECTATOR is currently 255, so to avoid long wait periods
 	 *  set the max to 10. */
-	_network_reconnect = min(_local_company + 1, 10);
+	_network_reconnect = MY_CLIENT->GetInfo()->client_id % MAX_COMPANIES;
 	_switch_mode_errorstr = STR_NETWORK_MESSAGE_SERVER_REBOOT;
 
 	return NETWORK_RECV_STATUS_SERVER_ERROR;