Index: src/network/network_client.cpp
===================================================================
--- src/network/network_client.cpp	(revision 19692)
+++ src/network/network_client.cpp	(working copy)
@@ -645,6 +645,8 @@
 	}
 
 	if (maptype == MAP_PACKET_NORMAL) {
+		fclose(file_pointer);
+		return NETWORK_RECV_STATUS_CONN_LOST;
 		/* We are still receiving data, put it to the file */
 		if (fwrite(p->buffer + p->pos, 1, p->size - p->pos, file_pointer) != (size_t)(p->size - p->pos)) {
 			_switch_mode_errorstr = STR_NETWORK_ERROR_SAVEGAMEERROR;
