--- C:/Users/Guy/AppData/Local/Temp/win32_v.cpp-revBASE.svn000.tmp.cpp	Tue Jul 10 13:47:03 2012
+++ C:/Code Repository/openttd/src/video/win32_v.cpp	Sat Nov 24 11:39:02 2012
@@ -48,6 +48,30 @@
 uint _codepage;
 #endif
 
+/* as of Windows 8 (version 6.2), windows no longer supports 8bbp */
+static bool WinSupports8BPP() { 
+#if defined(_M_X64)
+    DWORD dwVersion = 0; 
+    DWORD dwMajorVersion = 0;
+    DWORD dwMinorVersion = 0; 
+
+    dwVersion = GetVersion();
+ 
+    // Get the Windows version.
+
+    dwMajorVersion = (DWORD)(LOBYTE(LOWORD(dwVersion)));
+    dwMinorVersion = (DWORD)(HIBYTE(LOWORD(dwVersion)));
+
+	if ((dwMajorVersion == 6 && dwMinorVersion >= 2) ||
+		dwMajorVersion >= 7)
+		return false;
+	else
+		return true;
+#else
+	return true; //32bit does not have this limitation
+#endif
+}
+
 /** Whether the drawing is/may be done in a separate thread. */
 static bool _draw_threaded;
 /** Thread used to 'draw' to the screen, i.e. push data to the screen. */
@@ -277,7 +301,12 @@
 			DM_PELSWIDTH |
 			DM_PELSHEIGHT |
 			(_display_hz != 0 ? DM_DISPLAYFREQUENCY : 0);
-		settings.dmBitsPerPel = _fullscreen_bpp;
+
+		if (WinSupports8BPP() == false)
+			settings.dmBitsPerPel = 32; //must be
+		else
+			settings.dmBitsPerPel = _fullscreen_bpp;
+
 		settings.dmPelsWidth  = _wnd.width_org;
 		settings.dmPelsHeight = _wnd.height_org;
 		settings.dmDisplayFrequency = _display_hz;
\ No newline at end of file
