Index: src/company_base.h
===================================================================
--- src/company_base.h	(revision 22476)
+++ src/company_base.h	(working copy)
@@ -78,10 +78,10 @@
 	 */
 	bool is_ai;
 
-	Money yearly_expenses[3][EXPENSES_END];              ///< Expenses of the company for the last three years, in every #Expenses category.
-	CompanyEconomyEntry cur_economy;                     ///< Economic data of the company of this quarter.
-	CompanyEconomyEntry old_economy[MAX_HISTORY_MONTHS]; ///< Economic data of the company of the last #MAX_HISTORY_MONTHS months.
-	byte num_valid_stat_ent;                             ///< Number of valid statistical entries in #old_economy.
+	Money yearly_expenses[3][EXPENSES_END];                ///< Expenses of the company for the last three years, in every #Expenses category.
+	CompanyEconomyEntry cur_economy;                       ///< Economic data of the company of this quarter.
+	CompanyEconomyEntry old_economy[MAX_HISTORY_QUARTERS]; ///< Economic data of the company of the last #MAX_HISTORY_QUARTERS quarters.
+	byte num_valid_stat_ent;                               ///< Number of valid statistical entries in #old_economy.
 
 	CompanyProperties() : name(NULL), president_name(NULL) {}
 
Index: src/company_type.h
===================================================================
--- src/company_type.h	(revision 22476)
+++ src/company_type.h	(working copy)
@@ -40,7 +40,7 @@
 static const uint MAX_LENGTH_PRESIDENT_NAME_CHARS = 32; ///< The maximum length of a president name in characters including '\0'
 static const uint MAX_LENGTH_COMPANY_NAME_CHARS   = 32; ///< The maximum length of a company name in characters including '\0'
 
-static const uint MAX_HISTORY_MONTHS              = 24; ///< The maximum number of months kept as performance's history
+static const uint MAX_HISTORY_QUARTERS            = 24; ///< The maximum number of quarters kept as performance's history
 
 /** Define basic enum properties */
 template <> struct EnumPropsT<Owner> : MakeEnumPropsT<Owner, byte, OWNER_BEGIN, OWNER_END, INVALID_OWNER> {};
Index: src/economy.cpp
===================================================================
--- src/economy.cpp	(revision 22476)
+++ src/economy.cpp	(working copy)
@@ -565,7 +565,7 @@
 		c->old_economy[0] = c->cur_economy;
 		memset(&c->cur_economy, 0, sizeof(c->cur_economy));
 
-		if (c->num_valid_stat_ent != MAX_HISTORY_MONTHS) c->num_valid_stat_ent++;
+		if (c->num_valid_stat_ent != MAX_HISTORY_QUARTERS) c->num_valid_stat_ent++;
 
 		UpdateCompanyRatingAndValue(c, true);
 		if (c->block_preview != 0) c->block_preview--;
