Index: src/ai/api/ai_order.cpp
===================================================================
--- src/ai/api/ai_order.cpp	(revision 22025)
+++ src/ai/api/ai_order.cpp	(working copy)
@@ -397,6 +397,7 @@
 	EnforcePrecondition(false, AreOrderFlagsValid(destination, order_flags));
 
 	Order order;
+	memset(&order, 0, sizeof(order));
 	OrderType ot = (order_flags & AIOF_GOTO_NEAREST_DEPOT) ? OT_GOTO_DEPOT : ::GetOrderTypeByTile(destination);
 	switch (ot) {
 		case OT_GOTO_DEPOT: {
