Index: src/command.cpp
===================================================================
--- src/command.cpp	(wersja 22663)
+++ src/command.cpp	(kopia robocza)
@@ -498,15 +498,21 @@
 bool DoCommandP(TileIndex tile, uint32 p1, uint32 p2, uint32 cmd, CommandCallback *callback, const char *text, bool my_cmd)
 {
 	/* Cost estimation is generally only done when the
-	 * local user presses shift while doing somthing.
-	 * However, in case of incoming network commands,
-	 * map generation of the pause button we do want
-	 * to execute. */
-	bool estimate_only = _shift_pressed && IsLocalCompany() &&
-			!IsGeneratingWorld() &&
-			!(cmd & CMD_NETWORK_COMMAND) &&
-			(cmd & CMD_ID_MASK) != CMD_PAUSE;
+	 * local user presses shift while constructing somthing.
+	 * However, in case of incoming network commands or
+	 * map generation we do want to execute. */
+	bool estimate_only = false;
+	switch (_command_proc_table[cmd & CMD_ID_MASK].type) {
+		case CMDT_LANDSCAPE_CONSTRUCTION:
+		case CMDT_VEHICLE_CONSTRUCTION:
+			estimate_only = _shift_pressed && IsLocalCompany() &&
+					!IsGeneratingWorld() && !(cmd & CMD_NETWORK_COMMAND);
+			break;
 
+		default:
+			break; // just to silence warnings
+	}
+
 	/* We're only sending the command, so don't do
 	 * fancy things for 'success'. */
 	bool only_sending = _networking && !(cmd & CMD_NETWORK_COMMAND);
