Index: src/vehicle.cpp
===================================================================
--- src/vehicle.cpp	(revision 11235)
+++ src/vehicle.cpp	(working copy)
@@ -564,13 +564,10 @@
 /**
  * Deletes all vehicles in a chain.
  * @param v The first vehicle in the chain.
- *
- * @warning This function is not valid for any vehicle containing articulated
- * parts.
  */
 void DeleteVehicleChain(Vehicle *v)
 {
-	assert(v->type != VEH_TRAIN && v->type != VEH_ROAD);
+	assert(v->First() == v);
 
 	do {
 		Vehicle *u = v;
