Index: src/misc/dbg_helpers.cpp
===================================================================
--- src/misc/dbg_helpers.cpp	(Revision 11415)
+++ src/misc/dbg_helpers.cpp	(Arbeitskopie)
@@ -53,7 +53,8 @@
 CStrA ValueStr(SignalType t)
 {
 	CStrA out;
-	out.Format("%d (%s)", t, ItemAtT(t, signal_type_names, "UNK"));
+	/* The cast of t is to silence gcc on 64bit machines */
+	out.Format("%d (%s)", t, ItemAtT((uint8)t, signal_type_names, "UNK"));
 	return out.Transfer();
 }
 
