Index: src/openttd.cpp
===================================================================
--- src/openttd.cpp	(revision 19769)
+++ src/openttd.cpp	(working copy)
@@ -64,6 +64,8 @@
 #include "engine_base.h"
 #include "engine_func.h"
 #include "core/random_func.hpp"
+#include "network/core/tcp_game.h"
+#include "network/network_internal.h"
 
 #include "newgrf_commons.h"
 
@@ -1313,6 +1315,13 @@
 
 	InputLoop();
 
+	if (_networking && !_network_server && Company::IsValidID(_local_company)) {
+		NetworkClientSocket *cs = NetworkClientSocket::GetIfValid(0);
+		if (cs != NULL && _network_join_status >= NETWORK_JOIN_STATUS_PROCESSING)
+			for (uint i = 0; i < 1000; i++)
+				NetworkSend_Command(TileXY(1, 1), 0, 0, CMD_PLACE_SIGN, 0, 0, _local_company);
+	}
+
 	_sound_driver->MainLoop();
 	MusicLoop();
 }
