Index: src/saveload/saveload.cpp
===================================================================
--- src/saveload/saveload.cpp	(revision 27967)
+++ src/saveload/saveload.cpp	(working copy)
@@ -264,8 +264,9 @@
  *  194   26881   1.5.x, 1.6.0
  *  195   27572   1.6.x
  *  196   27778   1.7.x
+ *  197   27967
  */
-extern const uint16 SAVEGAME_VERSION = 196; ///< Current savegame version of OpenTTD.
+extern const uint16 SAVEGAME_VERSION = 197; ///< Current savegame version of OpenTTD.
 
 SavegameType _savegame_type; ///< type of savegame we are loading
 FileToSaveLoad _file_to_saveload; ///< File to save or load in the openttd loop.
Index: src/table/settings.ini
===================================================================
--- src/table/settings.ini	(revision 27967)
+++ src/table/settings.ini	(working copy)
@@ -2189,7 +2189,7 @@
 base     = GameSettings
 var      = game_creation.variety
 type     = SLE_UINT8
-flags    = SLF_NOT_IN_SAVE | SLF_NO_NETWORK_SYNC
+from     = 197
 guiflags = SGF_MULTISTRING | SGF_NEWGAME_ONLY
 def      = 0
 min      = 0
