diff --git a/Makefile.grf.in b/Makefile.grf.in
--- a/Makefile.grf.in
+++ b/Makefile.grf.in
@@ -22,6 +22,7 @@
 GRF_DIR  = $(ROOT_DIR)/media/extra_grf
 BIN_DIR  = !!BIN_DIR!!/data
 OBJS_DIR = !!GRF_OBJS_DIR!!
+OS       = !!OS!!
 
 # Check if we want to show what we are doing
 ifdef VERBOSE
@@ -35,6 +36,7 @@
 # Some configurational settings for your environment.
 GRFCODEC := grfcodec
 NFORENUM := $(shell [ `which nforenum 2>/dev/null` ] && echo "nforenum" || echo "renum")
+MD5SUM   := $(shell [ "$(OS)" = "OSX" ] && echo "md5 -r" || echo "md5sum")
 
 # Some "should not be changed" settings.
 NFO_FILES    := $(GRF_DIR)/*.nfo
@@ -58,7 +60,7 @@
 	$(Q) $(NFORENUM) $(OBJS_DIR)/sprites/openttd.nfo
 	$(Q) $(GRFCODEC) -e -m1 $(OBJS_DIR)/openttd.grf
 	$(Q) cp $(OBJS_DIR)/openttd.grf $(BIN_DIR)/openttd.grf
-	$(Q) for grf in $(BIN_DIR)/orig*.obg; do sed 's/^OPENTTD.GRF  = [0-9a-f]*$$/OPENTTD.GRF  = '`md5sum $(BIN_DIR)/openttd.grf | sed 's@ .*@@'`'/' -i $$grf; done
+	$(Q) for grf in $(BIN_DIR)/orig*.obg; do sed -i "" 's/^OPENTTD.GRF  = [0-9a-f]*$$/OPENTTD.GRF  = '`$(MD5SUM) $(BIN_DIR)/openttd.grf | sed 's@ .*@@'`'/' $$grf; done
 
 # Clean up temporary files.
 clean:
