Index: src/bootstrap_gui.cpp
===================================================================
--- src/bootstrap_gui.cpp	(revision 24979)
+++ src/bootstrap_gui.cpp	(working copy)
@@ -84,6 +84,7 @@
 	/** Simple call the constructor of the superclass. */
 	BootstrapContentDownloadStatusWindow() : BaseNetworkContentDownloadStatusWindow(&_bootstrap_download_status_window_desc)
 	{
+		OnDownloadComplete((ContentID)0);
 	}
 
 	virtual void OnDownloadComplete(ContentID cid)
@@ -175,7 +176,7 @@
 		switch (widget) {
 			case WID_BAFD_YES:
 				/* We got permission to connect! Yay! */
-				_network_content_client.Connect();
+				OnReceiveContentInfo(NULL);
 				break;
 
 			case WID_BAFD_NO:
@@ -196,7 +197,6 @@
 	virtual void OnReceiveContentInfo(const ContentInfo *ci)
 	{
 		/* And once the meta data is received, start downloading it. */
-		_network_content_client.Select(ci->id);
 		new BootstrapContentDownloadStatusWindow();
 		delete this;
 	}
@@ -212,8 +212,6 @@
  */
 bool HandleBootstrap()
 {
-	if (BaseGraphics::GetUsedSet() != NULL) return true;
-
 	/* No user interface, bail out with an error. */
 	if (BlitterFactoryBase::GetCurrentBlitter()->GetScreenDepth() == 0) goto failure;
 
