diff -r 07362bcbcfa6 src/autoreplace_gui.cpp
--- a/src/autoreplace_gui.cpp	Sat Jan 03 14:16:18 2009 +0000
+++ b/src/autoreplace_gui.cpp	Sat Jan 03 15:49:16 2009 +0100
@@ -253,13 +253,6 @@
 
 	virtual void OnPaint()
 	{
-		static const StringID _vehicle_type_names[] = {
-			STR_019F_TRAIN,
-			STR_019C_ROAD_VEHICLE,
-			STR_019E_SHIP,
-			STR_019D_AIRCRAFT
-		};
-
 		if (this->update_left || this->update_right) this->GenerateLists();
 
 		Company *c = GetCompany(_local_company);
@@ -287,7 +280,7 @@
 										!EngineHasReplacementForCompany(c, selected_id[0], selected_group));
 
 		/* now the actual drawing of the window itself takes place */
-		SetDParam(0, _vehicle_type_names[this->window_number]);
+		SetDParam(0, STR_019F_TRAIN + this->window_number);
 
 		if (this->window_number == VEH_TRAIN) {
 			/* set on/off for renew_keep_length */
diff -r 07362bcbcfa6 src/lang/english.txt
--- a/src/lang/english.txt	Sat Jan 03 14:16:18 2009 +0000
+++ b/src/lang/english.txt	Sat Jan 03 15:49:16 2009 +0100
@@ -478,10 +478,10 @@
 STR_AGE_RED                                                     :{RED}{COMMA} year{P "" s} ({COMMA})
 ############ range for service numbers ends
 
+STR_019F_TRAIN                                                  :Train
 STR_019C_ROAD_VEHICLE                                           :Road vehicle
+STR_019E_SHIP                                                   :Ship
 STR_019D_AIRCRAFT                                               :Aircraft
-STR_019E_SHIP                                                   :Ship
-STR_019F_TRAIN                                                  :Train
 STR_01A0_IS_GETTING_OLD                                         :{WHITE}{STRING} {COMMA} is getting old
 STR_01A1_IS_GETTING_VERY_OLD                                    :{WHITE}{STRING} {COMMA} is getting very old
 STR_01A2_IS_GETTING_VERY_OLD_AND                                :{WHITE}{STRING} {COMMA} is getting very old and urgently needs replacing
diff -r 07362bcbcfa6 src/vehicle.cpp
--- a/src/vehicle.cpp	Sat Jan 03 14:16:18 2009 +0000
+++ b/src/vehicle.cpp	Sat Jan 03 15:49:16 2009 +0100
@@ -1066,13 +1066,6 @@
 	}
 }
 
-static const StringID _vehicle_type_names[4] = {
-	STR_019F_TRAIN,
-	STR_019C_ROAD_VEHICLE,
-	STR_019E_SHIP,
-	STR_019D_AIRCRAFT,
-};
-
 static void ShowVehicleGettingOld(Vehicle *v, StringID msg)
 {
 	if (v->owner != _local_company) return;
@@ -1080,7 +1073,7 @@
 	/* Do not show getting-old message if autorenew is active (and it can replace the vehicle) */
 	if (GetCompany(v->owner)->engine_renew && GetEngine(v->engine_type)->company_avail != 0) return;
 
-	SetDParam(0, _vehicle_type_names[v->type]);
+	SetDParam(0, STR_019F_TRAIN + v->type);
 	SetDParam(1, v->unitnumber);
 	AddNewsItem(msg, NS_ADVICE, v->index, 0);
 }
@@ -1620,7 +1613,7 @@
 				_vehicles_to_autoreplace[v] = false;
 				if (v->owner == _local_company) {
 					/* Notify the user that we stopped the vehicle */
-					SetDParam(0, _vehicle_type_names[v->type]);
+					SetDParam(0, STR_019F_TRAIN + v->type);
 					SetDParam(1, v->unitnumber);
 					AddNewsItem(STR_ORDER_REFIT_FAILED, NS_ADVICE, v->index, 0);
 				}
