Index: src/network/core/udp.cpp
===================================================================
--- src/network/core/udp.cpp	(revision 22743)
+++ src/network/core/udp.cpp	(working copy)
@@ -88,7 +88,7 @@
 		NetworkAddress send(*recv);

 		/* Not the same type */
-		if (!send.IsFamily(s->first.GetAddress()->ss_family)) continue;
+		if (!recv->IsFamily(s->first.GetAddress()->ss_family)) continue;

 		p->PrepareToSend();

@@ -101,11 +101,11 @@
 #endif

 		/* Send the buffer */
-		int res = sendto(s->second, (const char*)p->buffer, p->size, 0, (struct sockaddr *)send.GetAddress(), send.GetAddressLength());
-		DEBUG(net, 7, "[udp] sendto(%s)", send.GetAddressAsString());
+		int res = sendto(s->second, (const char*)p->buffer, p->size, 0, (struct sockaddr *)recv->GetAddress(), recv->GetAddressLength());
+		DEBUG(net, 7, "[udp] sendto(%s)", recv->GetAddressAsString());

 		/* Check for any errors, but ignore it otherwise */
-		if (res == -1) DEBUG(net, 1, "[udp] sendto(%s) failed with: %i", send.GetAddressAsString(), GET_LAST_ERROR());
+		if (res == -1) DEBUG(net, 1, "[udp] sendto(%s) failed with: %i", recv->GetAddressAsString(), GET_LAST_ERROR());

 		if (!all) break;
 	}
