diff -r a03447555c68 src/cargo_type.h
--- a/src/cargo_type.h	Fri Oct 17 18:03:22 2008 +0200
+++ b/src/cargo_type.h	Fri Oct 17 18:03:55 2008 +0200
@@ -9,7 +9,7 @@
 
 /** Available types of cargo */
 enum CargoTypes {
-	// Temperate
+	/* Temperate */
 	CT_PASSENGERS   =  0,
 	CT_COAL         =  1,
 	CT_MAIL         =  2,
@@ -22,14 +22,14 @@
 	CT_STEEL        =  9,
 	CT_VALUABLES    = 10,
 
-	// Arctic
+	/* Arctic */
 	CT_WHEAT        =  6,
 	CT_HILLY_UNUSED =  8,
 	CT_PAPER        =  9,
 	CT_GOLD         = 10,
 	CT_FOOD         = 11,
 
-	// Tropic
+	/* Tropic */
 	CT_RUBBER       =  1,
 	CT_FRUIT        =  4,
 	CT_MAIZE        =  6,
@@ -37,7 +37,7 @@
 	CT_WATER        =  9,
 	CT_DIAMONDS     = 10,
 
-	// Toyland
+	/* Toyland */
 	CT_SUGAR        =  1,
 	CT_TOYS         =  3,
 	CT_BATTERIES    =  4,
diff -r a03447555c68 src/town_cmd.cpp
--- a/src/town_cmd.cpp	Fri Oct 17 18:03:22 2008 +0200
+++ b/src/town_cmd.cpp	Fri Oct 17 18:03:55 2008 +0200
@@ -2611,20 +2611,20 @@
 
 /** Tile callback functions for a town */
 extern const TileTypeProcs _tile_type_town_procs = {
-	DrawTile_Town,           /* draw_tile_proc */
-	GetSlopeZ_Town,          /* get_slope_z_proc */
-	ClearTile_Town,          /* clear_tile_proc */
-	GetAcceptedCargo_Town,   /* get_accepted_cargo_proc */
-	GetTileDesc_Town,        /* get_tile_desc_proc */
-	GetTileTrackStatus_Town, /* get_tile_track_status_proc */
-	ClickTile_Town,          /* click_tile_proc */
-	AnimateTile_Town,        /* animate_tile_proc */
-	TileLoop_Town,           /* tile_loop_clear */
-	ChangeTileOwner_Town,    /* change_tile_owner_clear */
-	GetProducedCargo_Town,   /* get_produced_cargo_proc */
-	NULL,                    /* vehicle_enter_tile_proc */
-	GetFoundation_Town,      /* get_foundation_proc */
-	TerraformTile_Town,      /* terraform_tile_proc */
+	DrawTile_Town,           // draw_tile_proc
+	GetSlopeZ_Town,          // get_slope_z_proc
+	ClearTile_Town,          // clear_tile_proc
+	GetAcceptedCargo_Town,   // get_accepted_cargo_proc
+	GetTileDesc_Town,        // get_tile_desc_proc
+	GetTileTrackStatus_Town, // get_tile_track_status_proc
+	ClickTile_Town,          // click_tile_proc
+	AnimateTile_Town,        // animate_tile_proc
+	TileLoop_Town,           // tile_loop_clear
+	ChangeTileOwner_Town,    // change_tile_owner_clear
+	GetProducedCargo_Town,   // get_produced_cargo_proc
+	NULL,                    // vehicle_enter_tile_proc
+	GetFoundation_Town,      // get_foundation_proc
+	TerraformTile_Town,      // terraform_tile_proc
 };
 
 
