Index: src/newgrf_commons.cpp
===================================================================
--- src/newgrf_commons.cpp	(Revision 11051)
+++ src/newgrf_commons.cpp	(Arbeitskopie)
@@ -252,7 +252,7 @@
 {
 	switch (_opt.landscape) {
 		case LT_TROPIC: return GetTropicZone(tile) == TROPICZONE_DESERT ? 1 : 2;
-		case LT_ARCTIC: return GetTileZ(tile) >= GetSnowLine() ? 4 : 0;
+		case LT_ARCTIC: return GetTileZ(tile) > GetSnowLine() ? 4 : 0;
 		default:        return 0;
 	}
 }
