diff -r 487e6b6103fb src/station_gui.cpp
--- a/src/station_gui.cpp	Mon Jan 12 00:06:22 2009 +0000
+++ b/src/station_gui.cpp	Mon Jan 12 12:01:12 2009 +0100
@@ -1206,7 +1206,14 @@ static const WindowDesc _select_station_
  */
 static bool StationJoinerNeeded(CommandContainer cmd, int w, int h)
 {
-	if (CmdFailed(DoCommand(&cmd, DC_NO_WATER | DC_AUTO))) return false;
+	/* Flags get send to the DoCommand */
+	uint32 flags = 0;
+	/* Command flags are used internally */
+	uint cmd_flags = GetCommandFlags(cmd.cmd);
+	if (cmd_flags & CMD_NO_WATER) flags |= DC_NO_WATER;
+	if (cmd_flags & CMD_AUTO) flags |= DC_AUTO;
+
+	if (CmdFailed(DoCommand(&cmd, flags))) return false;
 
 	/* Only show selection if distant join is enabled in the settings */
 	if (!_settings_game.station.distant_join_stations) return false;
