Index: src/settings.cpp
===================================================================
--- src/settings.cpp	(revision 12382)
+++ src/settings.cpp	(working copy)
@@ -1323,7 +1323,7 @@
 	  SDTG_VAR("max_spectators",       SLE_UINT8, S, 0, _network_game_info.spectators_max, 10, 0, MAX_CLIENTS, 0, STR_NULL, NULL),
 	  SDTG_VAR("restart_game_year",    SLE_INT32, S,D0, _network_restart_game_year,    0, MIN_YEAR, MAX_YEAR, 1, STR_NULL, NULL),
 	  SDTG_VAR("min_players",          SLE_UINT8, S, 0, _network_min_players,               0, 0, 10,   0, STR_NULL, NULL),
-	SDTG_OMANY("server_lang",          SLE_UINT8, S, 0, _network_game_info.server_lang,     0, 28, "ANY|ENGLISH|GERMAN|FRENCH|BRAZILIAN|BULGARIAN|CHINESE|CZECH|DANISH|DUTCH|ESPERANTO|FINNISH|HUNGARIAN|ICELANDIC|ITALIAN|JAPANESE|KOREAN|LITHUANIAN|NORWEGIAN|POLISH|PORTUGUESE|ROMANIAN|RUSSIAN|SLOVAK|SLOVENIAN|SPANISH|SWEDISH|TURKISH|UKRAINIAN", STR_NULL, NULL),
+	SDTG_OMANY("server_lang",          SLE_UINT8, S, 0, _network_game_info.server_lang,     0, 35, "ANY|ENGLISH|GERMAN|FRENCH|BRAZILIAN|BULGARIAN|CHINESE|CZECH|DANISH|DUTCH|ESPERANTO|FINNISH|HUNGARIAN|ICELANDIC|ITALIAN|JAPANESE|KOREAN|LITHUANIAN|NORWEGIAN|POLISH|PORTUGUESE|ROMANIAN|RUSSIAN|SLOVAK|SLOVENIAN|SPANISH|SWEDISH|TURKISH|UKRAINIAN|AFRIKAANS|CROATIAN|CATALAN|ESTONIAN|GALICIAN|GREEK|LATVIAN", STR_NULL, NULL),
 	 SDTG_BOOL("reload_cfg",                      S, 0, _network_reload_cfg,           false,              STR_NULL, NULL),
 	  SDTG_END()
 };
Index: src/lang/english.txt
===================================================================
--- src/lang/english.txt	(revision 12382)
+++ src/lang/english.txt	(working copy)
@@ -1451,6 +1451,13 @@
 STR_NETWORK_LANG_SWEDISH                                        :Swedish
 STR_NETWORK_LANG_TURKISH                                        :Turkish
 STR_NETWORK_LANG_UKRAINIAN                                      :Ukrainian
+STR_NETWORK_LANG_AFRIKAANS                                      :Afrikaans
+STR_NETWORK_LANG_CROATIAN                                       :Croatian
+STR_NETWORK_LANG_CATALAN                                        :Catalan
+STR_NETWORK_LANG_ESTONIAN                                       :Estonian
+STR_NETWORK_LANG_GALICIAN                                       :Galician
+STR_NETWORK_LANG_GREEK                                          :Greek
+STR_NETWORK_LANG_LATVIAN                                        :Latvian
 ############ End of leave-in-this-order
 
 STR_NETWORK_GAME_LOBBY                                          :{WHITE}Multiplayer game lobby
Index: src/table/sprites.h
===================================================================
--- src/table/sprites.h	(revision 12382)
+++ src/table/sprites.h	(working copy)
@@ -49,7 +49,7 @@
 
 	/* Extra graphic spritenumbers */
 	SPR_OPENTTD_BASE     = 4896,
-	OPENTTD_SPRITE_COUNT = 138,
+	OPENTTD_SPRITE_COUNT = 145,
 
 	/* Halftile-selection sprites */
 	SPR_HALFTILE_SELECTION_FLAT = SPR_OPENTTD_BASE,
@@ -201,7 +201,7 @@
 
 	/* Flags sprites (in same order as enum NetworkLanguage) */
 	SPR_FLAGS_BASE = SPR_ONEWAY_BASE + ONEWAY_SPRITE_COUNT,
-	FLAGS_SPRITE_COUNT = 29,
+	FLAGS_SPRITE_COUNT = 36,
 
 	/* Not really a sprite, but an empty bounding box. Used to construct bounding boxes, that help sorting the sprites, but do not have a sprite associated. */
 	SPR_EMPTY_BOUNDING_BOX = SPR_FLAGS_BASE + FLAGS_SPRITE_COUNT,
Index: src/table/files.h
===================================================================
--- src/table/files.h	(revision 12382)
+++ src/table/files.h	(working copy)
@@ -33,9 +33,10 @@
 		{ "TRGT.GRF",      {0xfc, 0xde, 0x1d, 0x7e, 0x8a, 0x74, 0x19, 0x7d, 0x72, 0xa6, 0x26, 0x95, 0x88, 0x4b, 0x90, 0x9e} }
 	},
 	{ "SAMPLE.CAT",    {0x42, 0x2e, 0xa3, 0xdd, 0x07, 0x4d, 0x28, 0x59, 0xbb, 0x51, 0x63, 0x9a, 0x6e, 0x0e, 0x85, 0xda} },
-	{ "OPENTTDD.GRF",  {0x9e, 0xb2, 0xf4, 0xd8, 0xf3, 0x63, 0xd4, 0x22, 0x5b, 0x70, 0xa6, 0xd4, 0xbe, 0x55, 0x1c, 0xd9} }
+	{ "OPENTTDD.GRF",  {0x61, 0x47, 0x47, 0x15, 0xa9, 0x06, 0x10, 0xb6, 0xed, 0x1f, 0xe1, 0x4d, 0x03, 0x6c, 0xa8, 0x02} }
 };
 
+
 static FileList files_win = {
 	{
 		{ "TRG1R.GRF",     {0xb0, 0x4c, 0xe5, 0x93, 0xd8, 0xc5, 0x01, 0x6e, 0x07, 0x47, 0x3a, 0x74, 0x3d, 0x7d, 0x33, 0x58} }, //    0 - 4792 inclusive
@@ -46,5 +47,5 @@
 		{ "TRGTR.GRF",     {0xde, 0x53, 0x65, 0x05, 0x17, 0xfe, 0x66, 0x1c, 0xea, 0xa3, 0x13, 0x8c, 0x6e, 0xdb, 0x0e, 0xb8} }
 	},
 	{ "SAMPLE.CAT",    {0x92, 0x12, 0xe8, 0x1e, 0x72, 0xba, 0xdd, 0x4b, 0xbe, 0x1e, 0xae, 0xae, 0x66, 0x45, 0x8e, 0x10} },
-	{ "OPENTTDW.GRF",  {0x1d, 0x9b, 0x03, 0x5e, 0x5d, 0x87, 0xef, 0x21, 0x0f, 0xe3, 0x6e, 0x6d, 0x12, 0xa1, 0x85, 0xfb} }
+	{ "OPENTTDW.GRF",  {0xc8, 0xc0, 0x01, 0x73, 0x60, 0xa7, 0x5e, 0xb1, 0xef, 0x9b, 0xe6, 0xba, 0xad, 0x3a, 0x7b, 0xd4} }
 };
Index: src/network/core/config.h
===================================================================
--- src/network/core/config.h	(revision 12382)
+++ src/network/core/config.h	(working copy)
@@ -47,7 +47,7 @@
 	 */
 	NETWORK_MAX_GRF_COUNT         =   55,
 
-	NETWORK_NUM_LANGUAGES         =   29, ///< Number of known languages (to the network protocol) + 1 for 'any'.
+	NETWORK_NUM_LANGUAGES         =   36, ///< Number of known languages (to the network protocol) + 1 for 'any'.
 	/**
 	 * The number of landscapes in OpenTTD.
 	 * This number must be equal to NUM_LANDSCAPE, but as this number is used
Index: src/network/network_internal.h
===================================================================
--- src/network/network_internal.h	(revision 12382)
+++ src/network/network_internal.h	(working copy)
@@ -75,7 +75,9 @@
 	NETWORK_JOIN_STATUS_GETTING_COMPANY_INFO,
 };
 
+#endif /* ENABLE_NETWORK */
 /* Language ids for server_lang and client_lang. Do NOT modify the order. */
+/* They're always defined, because of the future usage in newgrf_text.cpp */
 enum NetworkLanguage {
 	NETLANG_ANY = 0,
 	NETLANG_ENGLISH,
@@ -106,9 +108,18 @@
 	NETLANG_SWEDISH,
 	NETLANG_TURKISH,
 	NETLANG_UKRAINIAN,
+	NETLANG_AFRIKAANS,
+	NETLANG_CROATIAN,
+	NETLANG_CATALAN,
+	NETLANG_ESTONIAN,
+	NETLANG_GALICIAN,
+	NETLANG_GREEK,
+	NETLANG_LATVIAN,
 	NETLANG_COUNT
 };
+#ifdef ENABLE_NETWORK
 
+
 VARDEF NetworkGameInfo _network_game_info;
 VARDEF NetworkPlayerInfo _network_player_info[MAX_PLAYERS];
 VARDEF NetworkClientInfo _network_client_info[MAX_CLIENT_INFO];
