Index: src/saveload/saveload.cpp
===================================================================
--- src/saveload/saveload.cpp	(Revision 20894)
+++ src/saveload/saveload.cpp	(Arbeitskopie)
@@ -214,7 +214,7 @@
  *  149   20832
  *  150   20857
  */
-extern const uint16 SAVEGAME_VERSION = 150; ///< current savegame version of OpenTTD
+extern const uint16 SAVEGAME_VERSION = MORE_HEIGHTLEVEL_SAVEGAME_VERSION; ///< current savegame version of patched OpenTTD
 
 SavegameType _savegame_type; ///< type of savegame we are loading
 
Index: src/saveload/saveload.h
===================================================================
--- src/saveload/saveload.h	(Revision 20894)
+++ src/saveload/saveload.h	(Arbeitskopie)
@@ -84,6 +84,7 @@
 };
 
 #define SL_MAX_VERSION 255
+#define MORE_HEIGHTLEVEL_SAVEGAME_VERSION 250
 
 enum ChunkType {
 	CH_RIFF         =  0,
