Index: src/industry_cmd.cpp
===================================================================
--- src/industry_cmd.cpp	(revision 19286)
+++ src/industry_cmd.cpp	(working copy)
@@ -1366,7 +1366,7 @@
 				/* Clear the tiles as OWNER_TOWN to not affect town rating, and to not clear protected buildings */
 				CompanyID old_company = _current_company;
 				_current_company = OWNER_TOWN;
-				CommandCost ret = DoCommand(cur_tile, 0, 0, DC_NONE, CMD_LANDSCAPE_CLEAR).Failed();
+				CommandCost ret = DoCommand(cur_tile, 0, 0, DC_NONE, CMD_LANDSCAPE_CLEAR);
 				_current_company = old_company;
 
 				if (ret.Failed()) return ret;
