Index: src/station_cmd.cpp
===================================================================
--- src/station_cmd.cpp	(wersja 11376)
+++ src/station_cmd.cpp	(kopia robocza)
@@ -1800,7 +1800,7 @@
 /** Build a buoy.
  * @param tile tile where to place the bouy
  * @param flags operation to perform
- * @param p1 unused
+ * @param p1 (bit 0) - buoy doubles as a dock
  * @param p2 unused
  */
 CommandCost CmdBuildBuoy(TileIndex tile, uint32 flags, uint32 p1, uint32 p2)
@@ -1825,9 +1825,12 @@
 	if (flags & DC_EXEC) {
 		st->dock_tile = tile;
 		st->facilities |= FACIL_DOCK;
-		/* Buoys are marked in the Station struct by this flag. Yes, it is this
-		 * braindead.. */
-		st->had_vehicle_of_type |= HVOT_BUOY;
+		if (!HASBIT(p1, 0))
+			/* Buoys are marked in the Station struct by this flag. Yes, it is this
+			 * braindead.. */
+			st->had_vehicle_of_type |= HVOT_BUOY;
+		else
+			st->had_vehicle_of_type &= ~HVOT_BUOY;
 		st->owner = OWNER_NONE;
 
 		st->build_date = _date;
Index: src/ship_cmd.cpp
===================================================================
--- src/ship_cmd.cpp	(wersja 11376)
+++ src/ship_cmd.cpp	(kopia robocza)
@@ -682,7 +682,7 @@
 			} else if (v->dest_tile != 0) {
 				/* We have a target, let's see if we reached it... */
 				if (v->current_order.type == OT_GOTO_STATION &&
-						IsBuoyTile(v->dest_tile) &&
+						GetStation(v->current_order.dest)->IsBuoy() &&
 						DistanceManhattan(v->dest_tile, gp.new_tile) <= 3) {
 					/* We got within 3 tiles of our target buoy, so let's skip to our
 					 * next order */
Index: src/dock_gui.cpp
===================================================================
--- src/dock_gui.cpp	(wersja 11376)
+++ src/dock_gui.cpp	(kopia robocza)
@@ -48,7 +48,7 @@
 
 static void PlaceDocks_Buoy(TileIndex tile)
 {
-	DoCommandP(tile, 0, 0, CcBuildDocks, CMD_BUILD_BUOY | CMD_MSG(STR_9835_CAN_T_POSITION_BUOY_HERE));
+	DoCommandP(tile, _ctrl_pressed, 0, CcBuildDocks, CMD_BUILD_BUOY | CMD_MSG(STR_9835_CAN_T_POSITION_BUOY_HERE));
 }
 
 static void PlaceDocks_DemolishArea(TileIndex tile)
