diff a/src/hotkeys.cpp b/src/hotkeys.cpp
--- a/src/hotkeys.cpp
+++ b/src/hotkeys.cpp
@@ -292,13 +292,24 @@
 
 /**
  * Check if a keycode is bound to something.
- * @param keycode The keycode that was pressed
+ * @param keycode The keycode that was pressed.
+ * @param prev The hotkey number to start from, or -1 to get the first match.
  * @param global_only Limit the search to hotkeys defined as 'global'.
  * @return The number of the matching hotkey or -1.
  */
-int HotkeyList::CheckMatch(uint16 keycode, bool global_only) const
+int HotkeyList::CheckMatch(uint16 keycode, int prev, bool global_only) const
 {
-	for (const Hotkey *list = this->items; list->name != NULL; ++list) {
+	const Hotkey *list = this->items;
+	if (prev != -1) {
+		for (; list->name != NULL; ++list) {
+			if (list->num == prev) {
+				++list;
+				break;
+			}
+		}
+	}
+
+	for (; list->name != NULL; ++list) {
 		if (list->keycodes.Contains(keycode | WKC_GLOBAL_HOTKEY) || (!global_only && list->keycodes.Contains(keycode))) {
 			return list->num;
 		}
@@ -342,7 +353,8 @@
 	for (HotkeyList **list = _hotkey_lists->Begin(); list != _hotkey_lists->End(); ++list) {
 		if ((*list)->global_hotkey_handler == NULL) continue;
 
-		int hotkey = (*list)->CheckMatch(keycode, true);
+		int hotkey = -1;
+		hotkey = (*list)->CheckMatch(keycode, hotkey, true);
 		if (hotkey >= 0 && ((*list)->global_hotkey_handler(hotkey) == ES_HANDLED)) return;
 	}
 }
diff a/src/hotkeys.h b/src/hotkeys.h
--- a/src/hotkeys.h
+++ b/src/hotkeys.h
@@ -47,7 +47,7 @@
 	void Load(IniFile *ini);
 	void Save(IniFile *ini) const;
 
-	int CheckMatch(uint16 keycode, bool global_only = false) const;
+	int CheckMatch(uint16 keycode, int prev, bool global_only = false) const;
 
 	GlobalHotkeyHandlerFunc global_hotkey_handler;
 private:
diff a/src/main_gui.cpp b/src/main_gui.cpp
--- a/src/main_gui.cpp
+++ b/src/main_gui.cpp
@@ -528,8 +528,9 @@
  */
 bool IsQuitKey(uint16 keycode)
 {
-	int num = MainWindow::hotkeys.CheckMatch(keycode);
-	return num == GHK_QUIT;
+	int hotkey = -1;
+	hotkey = MainWindow::hotkeys.CheckMatch(keycode, hotkey);
+	return hotkey == GHK_QUIT;
 }
 
 
diff a/src/window.cpp b/src/window.cpp
--- a/src/window.cpp
+++ b/src/window.cpp
@@ -2484,7 +2484,8 @@
 	FOR_ALL_WINDOWS_FROM_FRONT(w) {
 		if (w->window_class == WC_MAIN_TOOLBAR) continue;
 		if (w->window_desc->hotkeys != NULL) {
-			int hotkey = w->window_desc->hotkeys->CheckMatch(keycode);
+			int hotkey = -1;
+			hotkey = w->window_desc->hotkeys->CheckMatch(keycode, hotkey);
 			if (hotkey >= 0 && w->OnHotkey(hotkey) == ES_HANDLED) return;
 		}
 		if (w->OnKeyPress(key, keycode) == ES_HANDLED) return;
@@ -2494,7 +2495,8 @@
 	/* When there is no toolbar w is null, check for that */
 	if (w != NULL) {
 		if (w->window_desc->hotkeys != NULL) {
-			int hotkey = w->window_desc->hotkeys->CheckMatch(keycode);
+			int hotkey = -1;
+			hotkey = w->window_desc->hotkeys->CheckMatch(keycode, hotkey);
 			if (hotkey >= 0 && w->OnHotkey(hotkey) == ES_HANDLED) return;
 		}
 		if (w->OnKeyPress(key, keycode) == ES_HANDLED) return;
