Index: src/saveload/station_sl.cpp
===================================================================
--- src/saveload/station_sl.cpp	(revision 27765)
+++ src/saveload/station_sl.cpp	(working copy)
@@ -325,6 +325,10 @@
 
 static void Load_STNS()
 {
+	_cargo_source_xy = 0;
+	_cargo_days = 0;
+	_cargo_feeder_share = 0;
+
 	int index;
 	while ((index = SlIterateArray()) != -1) {
 		Station *st = new (index) Station();
@@ -514,8 +518,9 @@
 
 static void Load_STNN()
 {
-	int index;
+	_num_flows = 0;
 
+	int index;
 	while ((index = SlIterateArray()) != -1) {
 		bool waypoint = (SlReadByte() & FACIL_WAYPOINT) != 0;
 
