Index: src/order_cmd.cpp
===================================================================
--- src/order_cmd.cpp	(revision 27632)
+++ src/order_cmd.cpp	(working copy)
@@ -897,7 +897,7 @@
 	if (!Order::CanAllocateItem()) return_cmd_error(STR_ERROR_NO_MORE_SPACE_FOR_ORDERS);
 	if (v->orders.list == NULL && !OrderList::CanAllocateItem()) return_cmd_error(STR_ERROR_NO_MORE_SPACE_FOR_ORDERS);
 
-	if (v->type == VEH_SHIP && _settings_game.pf.pathfinder_for_ships != VPF_NPF) {
+	if (v->type == VEH_SHIP && _settings_game.pf.pathfinder_for_ships == VPF_OPF) {
 		/* Make sure the new destination is not too far away from the previous */
 		const Order *prev = NULL;
 		uint n = 0;
