Index: src/main_gui.cpp
===================================================================
--- src/main_gui.cpp	(revision 11616)
+++ src/main_gui.cpp	(working copy)
@@ -1889,7 +1889,17 @@
 			case WKC_F11: PlaceLandBlockInfo(); break;
 			case WKC_CTRL | 'S': MenuClickSmallScreenshot(); break;
 			case WKC_CTRL | 'G': MenuClickWorldScreenshot(); break;
+			case WKC_NUM_PLUS: // Fall through
+			case WKC_EQUALS: // Fall through
+			case WKC_SHIFT | WKC_EQUALS: // Fall through
+			case WKC_SHIFT | WKC_F5: ToolbarZoomInClick(w); break;
+			case WKC_NUM_MINUS: // Fall through
+			case WKC_MINUS: // Fall through
+			case WKC_SHIFT | WKC_MINUS: // Fall through
+			case WKC_SHIFT | WKC_F6: ToolbarZoomOutClick(w); break;
 			case 'L': ShowEditorTerraformToolBar(); break;
+			case 'M': ShowSmallMap(); break;
+			case 'V': ShowExtraViewPortWindow(); break;
 			default: return;
 		}
 		e->we.keypress.cont = false;
