Index: src/linkgraph/linkgraph_gui.cpp
===================================================================
--- a/src/linkgraph/linkgraph_gui.cpp
+++ b/src/linkgraph/linkgraph_gui.cpp
@@ -220,11 +220,12 @@ void LinkGraphOverlay::DrawContent(Point pta, Point ptb, const LinkProperties &c
 
 	/* Move line a bit 90° against its dominant direction to prevent it from
 	 * being hidden below the grey line. */
+	int side = _settings_game.vehicle.road_side ? 1 : -1;
 	if (abs(pta.x - ptb.x) < abs(pta.y - ptb.y)) {
-		int offset_x = (pta.y > ptb.y ? 1 : -1) * this->scale;
+		int offset_x = (pta.y > ptb.y ? 1 : -1) * side * this->scale;
 		GfxDrawLine(pta.x + offset_x, pta.y, ptb.x + offset_x, ptb.y, colour, this->scale, dash);
 	} else {
-		int offset_y = (pta.x < ptb.x ? 1 : -1) * this->scale;
+		int offset_y = (pta.x < ptb.x ? 1 : -1) * side * this->scale;
 		GfxDrawLine(pta.x, pta.y + offset_y, ptb.x, ptb.y + offset_y, colour, this->scale, dash);
 	}
