Index: src/articulated_vehicles.cpp
===================================================================
--- src/articulated_vehicles.cpp	(revision 11115)
+++ src/articulated_vehicles.cpp	(working copy)
@@ -14,13 +14,12 @@
 #include "newgrf_callbacks.h"
 #include "newgrf_engine.h"
 
-
 uint CountArticulatedParts(EngineID engine_type)
 {
 	if (!HASBIT(EngInfo(engine_type)->callbackmask, CBM_ARTIC_ENGINE)) return 0;
 
 	uint i;
-	for (i = 1; i < 10; i++) {
+	for (i = 1; i < MAX_ARTICULATED_VEH_PARTS; i++) {
 		uint16 callback = GetVehicleCallback(CBID_VEHICLE_ARTIC_ENGINE, i, 0, engine_type, NULL);
 		if (callback == CALLBACK_FAILED || callback == 0xFF) break;
 	}
@@ -35,7 +34,7 @@
 
 	if (!HASBIT(EngInfo(v->engine_type)->callbackmask, CBM_ARTIC_ENGINE)) return;
 
-	for (uint i = 1; i < 10; i++) {
+	for (uint i = 1; i < MAX_ARTICULATED_VEH_PARTS; i++) {
 		uint16 callback = GetVehicleCallback(CBID_VEHICLE_ARTIC_ENGINE, i, 0, v->engine_type, v);
 		if (callback == CALLBACK_FAILED || callback == 0xFF) return;
 
Index: src/articulated_vehicles.h
===================================================================
--- src/articulated_vehicles.h	(revision 11115)
+++ src/articulated_vehicles.h	(working copy)
@@ -5,6 +5,8 @@
 #ifndef ARTICULATED_VEHICLES_H
 #define ARTICULATED_VEHICLES_H
 
+#define MAX_ARTICULATED_VEH_PARTS 32
+
 uint CountArticulatedParts(EngineID engine_type);
 void AddArticulatedParts(Vehicle **vl, VehicleType type);
 
Index: src/roadveh_cmd.cpp
===================================================================
--- src/roadveh_cmd.cpp	(revision 11115)
+++ src/roadveh_cmd.cpp	(working copy)
@@ -181,8 +181,8 @@
 
 	uint num_vehicles = 1 + CountArticulatedParts(p1);
 
-	/* Allow for the front and up to 10 articulated parts. */
-	Vehicle *vl[11];
+	/* Allow for the front and limited number of articulated parts. */
+	Vehicle *vl[MAX_ARTICULATED_VEH_PARTS + 1];
 	memset(&vl, 0, sizeof(vl));
 
 	if (!Vehicle::AllocateList(vl, num_vehicles)) {
Index: src/train_cmd.cpp
===================================================================
--- src/train_cmd.cpp	(revision 11115)
+++ src/train_cmd.cpp	(working copy)
@@ -517,7 +517,7 @@
 	uint num_vehicles = 1 + CountArticulatedParts(engine);
 
 	if (!(flags & DC_QUERY_COST)) {
-		Vehicle *vl[11]; // Allow for wagon and upto 10 artic parts.
+		Vehicle *vl[MAX_ARTICULATED_VEH_PARTS + 1]; // Allow for wagon and limited number of artic parts.
 
 		memset(&vl, 0, sizeof(vl));
 
@@ -685,7 +685,7 @@
 		CountArticulatedParts(p1);
 
 	if (!(flags & DC_QUERY_COST)) {
-		Vehicle *vl[12]; // Allow for upto 10 artic parts and dual-heads
+		Vehicle *vl[MAX_ARTICULATED_VEH_PARTS + 2]; // Limit number of artic parts and dual-heads
 
 		memset(&vl, 0, sizeof(vl));
 
