Index: settings.c
===================================================================
--- settings.c	(revision 4466)
+++ settings.c	(working copy)
@@ -15,6 +15,7 @@
 #include "console.h"
 #include "saveload.h"
 #include "npf.h"
+#include "hal.h"
 
 /** The patch values that are used for new games and/or modified in config file */
 Patches _patches_newgame;
@@ -1068,6 +1069,12 @@
 	DoCommandP(0, 2, _patches.autorenew_money, NULL, CMD_REPLACE_VEHICLE);
 	return 0;
 }
+
+static int32 MusicVolumeUpdate(int32 p1)
+{
+	_music_driver->set_volume(msf.music_vol);
+	return 0;
+}
 /* End - Callback Functions */
 
 #ifndef EXTERNAL_PLAYER
@@ -1076,7 +1083,7 @@
 
 static const SettingDesc _music_settings[] = {
 	 SDT_VAR(MusicFileSettings, playlist,  SLE_UINT8, S, 0,   0, 0,   5, STR_NULL, NULL),
-	 SDT_VAR(MusicFileSettings, music_vol, SLE_UINT8, S, 0, 128, 0, 100, STR_NULL, NULL),
+	 SDT_VAR(MusicFileSettings, music_vol, SLE_UINT8, S, 0, 128, 0, 100, STR_NULL, MusicVolumeUpdate),
 	 SDT_VAR(MusicFileSettings, effect_vol,SLE_UINT8, S, 0, 128, 0, 100, STR_NULL, NULL),
 	SDT_LIST(MusicFileSettings, custom_1,  SLE_UINT8, S, 0, NULL,        STR_NULL, NULL),
 	SDT_LIST(MusicFileSettings, custom_2,  SLE_UINT8, S, 0, NULL,        STR_NULL, NULL),
