Index: src/newgrf.cpp
===================================================================
--- src/newgrf.cpp	(revision 16836)
+++ src/newgrf.cpp	(working copy)
@@ -5786,9 +5786,9 @@
 
 			if (hs == NULL) continue;
 
-			const HouseSpec *next1 = (i + 1 < HOUSE_MAX ? file->housespec[i + 1] : NULL);
-			const HouseSpec *next2 = (i + 2 < HOUSE_MAX ? file->housespec[i + 2] : NULL);
-			const HouseSpec *next3 = (i + 3 < HOUSE_MAX ? file->housespec[i + 3] : NULL);
+			HouseSpec *next1 = (i + 1 < HOUSE_MAX ? file->housespec[i + 1] : NULL);
+			HouseSpec *next2 = (i + 2 < HOUSE_MAX ? file->housespec[i + 2] : NULL);
+			HouseSpec *next3 = (i + 3 < HOUSE_MAX ? file->housespec[i + 3] : NULL);
 
 			if (((hs->building_flags & BUILDING_HAS_2_TILES) != 0 &&
 						(next1 == NULL || !next1->enabled || (next1->building_flags & BUILDING_HAS_1_TILE) != 0)) ||
@@ -5796,6 +5796,9 @@
 						(next2 == NULL || !next2->enabled || (next2->building_flags & BUILDING_HAS_1_TILE) != 0 ||
 						next3 == NULL || !next3->enabled || (next3->building_flags & BUILDING_HAS_1_TILE) != 0))) {
 				hs->enabled = false;
+				if (hs->building_flags & BUILDING_HAS_2_TILES && next1 != NULL) next1->enabled = false;
+				if (hs->building_flags & BUILDING_HAS_4_TILES && next2 != NULL) next2->enabled = false;
+				if (hs->building_flags & BUILDING_HAS_4_TILES && next3 != NULL) next3->enabled = false;
 				DEBUG(grf, 1, "FinaliseHouseArray: %s defines house %d as multitile, but no suitable tiles follow. Disabling house.", file->filename, hs->local_id);
 				continue;
 			}
