Index: src/newgrf_house.cpp
===================================================================
--- src/newgrf_house.cpp	(Revision 14569)
+++ src/newgrf_house.cpp	(Arbeitskopie)
@@ -481,11 +481,22 @@
 	const HouseSpec *hs = GetHouseSpecs(house_id);
 	const SpriteGroup *group;
 	ResolverObject object;
+	Town * town = GetTownByTile(ti->tile);
 
-	if (ti->tileh != SLOPE_FLAT) DrawFoundation(ti, FOUNDATION_LEVELED);
+	if (ti->tileh != SLOPE_FLAT) {
+		bool draw_old_one = true;
+		if (HasBit(hs->callback_mask, CBM_HOUSE_DRAW_FOUNDATIONS)) {
+			/* Called to determine the type (if any) of foundation to draw for industry tile */
+			uint32 callback_res =
+				GetHouseCallback(CBID_HOUSE_DRAW_FOUNDATIONS, 0, 0, house_id, town, ti->tile);
+			draw_old_one = callback_res != 0;
+		}
 
-	NewHouseResolver(&object, house_id, ti->tile, GetTownByTile(ti->tile));
+		if (draw_old_one) DrawFoundation(ti, FOUNDATION_LEVELED);
+	}
 
+	NewHouseResolver(&object, house_id, ti->tile, town);
+
 	group = Resolve(hs->spritegroup, &object);
 	if (group == NULL || group->type != SGT_TILELAYOUT) {
 		/* XXX: This is for debugging purposes really, and shouldn't stay. */
