diff -r 216149dbc2fc src/ai/api/ai_industry.cpp
--- a/src/ai/api/ai_industry.cpp	Mon Jan 12 18:15:18 2009 +0000
+++ b/src/ai/api/ai_industry.cpp	Mon Jan 12 20:03:12 2009 +0100
@@ -10,6 +10,7 @@
 #include "../../industry.h"
 #include "../../strings_func.h"
 #include "../../station_func.h"
+#include "../../station_base.h"
 #include "table/strings.h"
 
 /* static */ IndustryID AIIndustry::GetMaxIndustryID()
