Index: src/station_cmd.cpp
===================================================================
--- src/station_cmd.cpp	(revision 21229)
+++ src/station_cmd.cpp	(working copy)
@@ -1544,9 +1544,9 @@
 
 
 /**
- * Remove a rail road station/waypoint
+ * Remove a rail station/waypoint
  * @param st The station/waypoint to remove the rail part from
  * @param flags operation to perform
  * @tparam T the type of station to remove
 template <class T>
 CommandCost RemoveRailStation(T *st, DoCommandFlag flags)
@@ -1612,8 +1612,8 @@
 }
 
 /**
- * Remove a rail road station
+ * Remove a rail station
  * @param tile TileIndex been queried
  * @param flags operation to perform
  * @return cost or failure of operation
 static CommandCost RemoveRailStation(TileIndex tile, DoCommandFlag flags)