Index: src/economy.cpp
===================================================================
--- src/economy.cpp	(revision 27657)
+++ src/economy.cpp	(working copy)
@@ -654,10 +654,17 @@
  */
 static void CompaniesGenStatistics()
 {
+	Company *c;
+
+	/* Check for bankruptcy each month before infrastructure
+	 * costs to prevent false bankrupts for AIs. */
+	FOR_ALL_COMPANIES(c) {
+		CompanyCheckBankrupt(c);
+	}
+
 	Station *st;
 
 	Backup<CompanyByte> cur_company(_current_company, FILE_LINE);
-	Company *c;
 
 	if (!_settings_game.economy.infrastructure_maintenance) {
 		FOR_ALL_STATIONS(st) {
@@ -688,11 +695,6 @@
 	}
 	cur_company.Restore();
 
-	/* Check for bankruptcy each month */
-	FOR_ALL_COMPANIES(c) {
-		CompanyCheckBankrupt(c);
-	}
-
 	/* Only run the economic statics and update company stats every 3rd month (1st of quarter). */
 	if (!HasBit(1 << 0 | 1 << 3 | 1 << 6 | 1 << 9, _cur_month)) return;
 
