diff --git a/src/ai/ai_core.cpp b/src/ai/ai_core.cpp
index fcba313..42d5d89 100644
--- a/src/ai/ai_core.cpp
+++ b/src/ai/ai_core.cpp
@@ -70,7 +70,8 @@
 
 	const Company *c;
 	FOR_ALL_COMPANIES(c) {
-		if (c->is_ai) {
+		/* the second test is needed when we're watching a network game with AIs */
+		if (c->is_ai && c->ai_instance) {
 			_current_company = c->index;
 			c->ai_instance->GameLoop();
 		}
