Index: src/company_cmd.cpp
===================================================================
--- src/company_cmd.cpp	(revision 20022)
+++ src/company_cmd.cpp	(working copy)
@@ -612,7 +612,7 @@
 
 #ifdef ENABLE_AI
 	if (_next_competitor_start == 0) {
-		_next_competitor_start = AI::GetStartNextTime() * DAY_TICKS;
+		_next_competitor_start = /*AI::GetStartNextTime() */ DAY_TICKS;
 	}
 
 	if (AI::CanStartNew() && _game_mode != GM_MENU && --_next_competitor_start == 0) {
Index: src/company_gui.cpp
===================================================================
--- src/company_gui.cpp	(revision 20022)
+++ src/company_gui.cpp	(working copy)
@@ -792,6 +792,7 @@
 	{
 		for (LiveryScheme scheme = LS_DEFAULT; scheme < LS_END; scheme++) {
 			if (HasBit(this->sel, scheme)) {
+				printf("Current company %d, local %d\n", (int)_current_company, (int)_local_company);
 				DoCommandP(0, scheme | (widget == SCLW_WIDGET_PRI_COL_DROPDOWN ? 0 : 256), index, CMD_SET_COMPANY_COLOUR);
 			}
 		}
