diff -r 4e8b1c754f33 src/music_gui.cpp
--- a/src/music_gui.cpp	Thu Oct 16 13:20:35 2008 +0000
+++ b/src/music_gui.cpp	Fri Oct 17 10:06:40 2008 +0200
@@ -114,7 +114,7 @@
 	do {
 		/* We are now checking for the existence of that file prior
 		 * to add it to the list of available songs */
-		if (FioCheckFileExists(origin_songs_specs[_playlists[msf.playlist][i]].filename, GM_DIR)) {
+		if (FioCheckFileExists(origin_songs_specs[_playlists[msf.playlist][i] - 1].filename, GM_DIR)) {
 			_cur_playlist[j] = _playlists[msf.playlist][i];
 			j++;
 		}
