Index: src/spritecache.cpp
===================================================================
--- src/spritecache.cpp	(revision 12002)
+++ src/spritecache.cpp	(working copy)
@@ -262,8 +262,8 @@
 
 void DupSprite(SpriteID old_spr, SpriteID new_spr)
 {
+	SpriteCache *scnew = AllocateSpriteCache(new_spr); // may reallocate: so put it first
 	SpriteCache *scold = GetSpriteCache(old_spr);
-	SpriteCache *scnew = AllocateSpriteCache(new_spr);
 
 	scnew->file_slot = scold->file_slot;
 	scnew->file_pos = scold->file_pos;
