Index: src/video/null_v.cpp
===================================================================
--- src/video/null_v.cpp	(revision 16312)
+++ src/video/null_v.cpp	(working copy)
@@ -7,11 +7,14 @@
 #include "../blitter/factory.hpp"
 #include "null_v.h"
 
+#include "../company_func.h"
+#include "../command_func.h"
+
 static FVideoDriver_Null iFVideoDriver_Null;
 
 const char *VideoDriver_Null::Start(const char * const *parm)
 {
-	this->ticks = GetDriverParamInt(parm, "ticks", 1000);
+	this->ticks = GetDriverParamInt(parm, "ticks", 10000);
 	_screen.width  = _screen.pitch = _cur_resolution.width;
 	_screen.height = _cur_resolution.height;
 	ScreenSizeChanged();
@@ -32,6 +35,12 @@
 
 	for (i = 0; i < this->ticks; i++) {
 		GameLoop();
+		if (i == 189) {
+			CompanyID cid = _current_company;
+			_current_company = 0;
+			DoCommandP(0, 0, 608, CMD_MASS_START_STOP);
+			_current_company = cid;
+		}
 		_screen.dst_ptr = NULL;
 		UpdateWindows();
 	}
Index: src/misc/str.hpp
===================================================================
--- src/misc/str.hpp	(revision 16312)
+++ src/misc/str.hpp	(working copy)
@@ -119,7 +119,7 @@
 	int AddFormatL(const Tchar *format, va_list args)
 	{
 		bsize_t addSize = Api::StrLen(format);
-		if (addSize < 16) addSize = 16;
+		if (addSize < 32) addSize = 32;
 		addSize += addSize / 2;
 		int ret;
 		int err = 0;
Index: src/yapf/yapf_rail.cpp
===================================================================
--- src/yapf/yapf_rail.cpp	(revision 16312)
+++ src/yapf/yapf_rail.cpp	(working copy)
@@ -11,7 +11,7 @@
 #include "../vehicle_func.h"
 #include "../functions.h"
 
-#define DEBUG_YAPF_CACHE 0
+#define DEBUG_YAPF_CACHE 1
 
 #if DEBUG_YAPF_CACHE
 template <typename Tpf> void DumpState(Tpf &pf1, Tpf &pf2)
