--- src/water_map.h
+++ src/water_map.h
@@ -100,6 +100,7 @@
  * Is it a plain water tile?
  * @param t Water tile to query.
  * @return \c true if any type of clear water like ocean, river, or canal.
+ * @pre IsTileType(t, MP_WATER)
  */
 static inline bool IsWater(TileIndex t)
 {
@@ -111,6 +112,7 @@
  * Is it a sea water tile?
  * @param t Water tile to query.
  * @return \c true if it is a sea water tile.
+ * @pre IsTileType(t, MP_WATER)
  */
 static inline bool IsSea(TileIndex t)
 {
@@ -121,6 +123,7 @@
  * Is it a canal tile?
  * @param t Water tile to query.
  * @return \c true if it is a canal tile.
+ * @pre IsTileType(t, MP_WATER)
  */
 static inline bool IsCanal(TileIndex t)
 {
@@ -131,6 +134,7 @@
  * Is it a river water tile?
  * @param t Water tile to query.
  * @return \c true if it is a river water tile.
+ * @pre IsTileType(t, MP_WATER)
  */
 static inline bool IsRiver(TileIndex t)
 {
@@ -151,6 +155,7 @@
  * Is it a coast tile?
  * @param t Water tile to query.
  * @return \c true if it is a sea water tile.
+ * @pre IsTileType(t, MP_WATER)
  */
 static inline bool IsCoast(TileIndex t)
 {
@@ -227,6 +232,7 @@
  * Get the direction of the ship depot.
  * @param t Water tile to query.
  * @return Direction of the depot.
+ * @pre IsShipDepotTile(t)
  */
 static inline DiagDirection GetShipDepotDirection(TileIndex t)
 {
@@ -237,6 +243,7 @@
  * Get the most northern tile of a ship depot.
  * @param t One of the tiles of the ship depot.
  * @return The northern tile of the depot.
+ * @pre IsShipDepotTile(t)
  */
 static inline TileIndex GetShipDepotNorthTile(TileIndex t)
 {
@@ -244,7 +251,7 @@
 }
 
 /**
- * Is it a water lock tile?
+ * Is there a lock on a given water tile?
  * @param t Water tile to query.
  * @return \c true if it is a water lock tile.
  */
@@ -289,8 +296,17 @@
 
 /**
  * Get a section of a depot or a lock.
+ *
+ * Return value for depot:
+ *    \li bit 0 - depot part (DepotPart)
+ *    \li bit 1 - depot axis (Axis)
+ * * Return value for lock:
+ *    \li bits 0..1 - lock direction (DiagDirection)
+ *    \li bits 2..3 - lock part (LockPart)
+ *
  * @param t Water tile to query.
  * @return The section.
+ * @pre IsShipDepotTile(t) || IsLockTile(t)
  */
 static inline byte GetSection(TileIndex t)
 {
