Index: src/waypoint_cmd.cpp
===================================================================
--- src/waypoint_cmd.cpp	(revision 27655)
+++ src/waypoint_cmd.cpp	(working copy)
@@ -324,6 +324,7 @@
 		if (wp->town == NULL) MakeDefaultName(wp);
 
 		MakeBuoy(tile, wp->index, GetWaterClass(tile));
+		MarkTileDirtyByTile(tile);
 
 		wp->UpdateVirtCoord();
 		InvalidateWindowData(WC_WAYPOINT_VIEW, wp->index);
