Index: src/station_gui.cpp
===================================================================
--- src/station_gui.cpp	(revision 15072)
+++ src/station_gui.cpp	(working copy)
@@ -1111,6 +1111,10 @@
 	{
 		this->vscroll.cap = 6;
 		this->resize.step_height = 10;
+		_thd.pos.x = TileX(tile) * TILE_SIZE;
+		_thd.pos.y = TileY(tile) * TILE_SIZE;
+		_thd.size.x = w * TILE_SIZE;
+		_thd.size.y = h * TILE_SIZE;
 		_thd.lock_pos = true;
 		_thd.lock_size = true;
 
