Index: main_gui.c
===================================================================
--- main_gui.c	(revision 3419)
+++ main_gui.c	(working copy)
@@ -932,9 +931,11 @@
 static void ToolbarBuildRailClick(Window *w)
 {
 	const Player* p = GetPlayer(_local_player);
+	int i;
 	Window *w2;
+	for(i = 0; (!HasRailtypeAvail(p, _last_built_railtype - i) && i != INVALID_RAILTYPE); i++);
 	w2 = PopupMainToolbMenu(w, 457, 19, STR_1015_RAILROAD_CONSTRUCTION, RAILTYPE_END, ~p->avail_railtypes);
-	WP(w2,menu_d).sel_index = _last_built_railtype;
+	WP(w2,menu_d).sel_index = (i == INVALID_RAILTYPE) ? RAILTYPE_RAIL : (_last_built_railtype - i);
 }
 
 static void ToolbarBuildRoadClick(Window *w)
