Index: src/roadveh_cmd.cpp
===================================================================
--- src/roadveh_cmd.cpp	(revision 13619)
+++ src/roadveh_cmd.cpp	(working copy)
@@ -279,6 +279,18 @@
 
 		VehiclePositionChanged(v);
 
+		CargoID ct = CT_INVALID;
+		for (Vehicle *u = v->Next(); u != NULL; /* Nothing */) {
+			if (u->cargo_cap == 0) {
+				if (ct != CT_INVALID) u->cargo_type = ct;
+			} else if (ct == CT_INVALID) {
+				ct = u->cargo_type;
+				u = v;
+				continue;
+			}
+			u = u->Next();
+		}
+
 		InvalidateWindowData(WC_VEHICLE_DEPOT, v->tile);
 		InvalidateWindowClassesData(WC_ROADVEH_LIST, 0);
 		InvalidateWindow(WC_COMPANY, v->owner);
Index: src/roadveh_gui.cpp
===================================================================
--- src/roadveh_gui.cpp	(revision 13619)
+++ src/roadveh_gui.cpp	(working copy)
@@ -63,7 +63,9 @@
 
 		for (const Vehicle *u = v; u != NULL; u = u->Next()) {
 			str = STR_8812_EMPTY;
-			if (!u->cargo.Empty()) {
+			if (u->cargo_cap == 0) {
+				str = STR_01A9_NONE;
+			} else if (!u->cargo.Empty()) {
 				SetDParam(0, u->cargo_type);
 				SetDParam(1, u->cargo.Count());
 				SetDParam(2, u->cargo.Source());
