Index: openttd.c
===================================================================
--- openttd.c	(revision 3305)
+++ openttd.c	(working copy)
@@ -1169,6 +1169,11 @@
 	// Load the sprites
 	GfxLoadSprites();
 
+	/* Connect front and rear engines of multiheaded trains and converts subtype to the new format */
+	if (CheckSavegameVersionOldStyle(17, 1)) {
+		ConvertOldMultiheadToNew();
+	}
+
 	// Update current year
 	SetDate(_date);
 
Index: vehicle.c
===================================================================
--- vehicle.c	(revision 3305)
+++ vehicle.c	(working copy)
@@ -2252,7 +2252,7 @@
  *  Converts all trains to the new subtype format introduced in savegame 16.2
  *  It also links multiheaded engines or make them forget they are multiheaded if no suitable partner is found
  */
-static inline void ConvertOldMultiheadToNew(void)
+void ConvertOldMultiheadToNew(void)
 {
 	Vehicle *v;
 	FOR_ALL_VEHICLES(v) {
@@ -2396,11 +2396,6 @@
 			}
 		}
 	}
-
-	/* Connect front and rear engines of multiheaded trains and converts subtype to the new format */
-	if (CheckSavegameVersionOldStyle(17, 1)) {
-		ConvertOldMultiheadToNew();
-	}
 }
 
 const ChunkHandler _veh_chunk_handlers[] = {
Index: vehicle.h
===================================================================
--- vehicle.h	(revision 3305)
+++ vehicle.h	(working copy)
@@ -446,4 +446,6 @@
  * Best is to have a virtual value for it when it needs to change again */
 #define STATUS_BAR 5
 
+void ConvertOldMultiheadToNew(void);
+
 #endif /* VEHICLE_H */
