Index: src/console_cmds.cpp
===================================================================
--- src/console_cmds.cpp	(wersja 23508)
+++ src/console_cmds.cpp	(kopia robocza)
@@ -881,7 +881,7 @@
 	}
 
 	/* Don't resolve the address first, just print it directly as it comes from the config file. */
-	IConsolePrintF(CC_DEFAULT, "Reconnecting to %s:%d...", _settings_client.network.last_host, _settings_client.network.last_port);
+	IConsolePrintF(CC_DEFAULT, "Reconnecting to %s:%d...", _settings_client.network.last_host, (int)_settings_client.network.last_port);
 
 	NetworkClientConnectGame(NetworkAddress(_settings_client.network.last_host, _settings_client.network.last_port), playas);
 	return true;
Index: src/network/core/address.cpp
===================================================================
--- src/network/core/address.cpp	(wersja 23508)
+++ src/network/core/address.cpp	(kopia robocza)
@@ -81,7 +81,7 @@
 	if (this->GetAddress()->ss_family == AF_INET6) buffer = strecpy(buffer, "[", last);
 	buffer = strecpy(buffer, this->GetHostname(), last);
 	if (this->GetAddress()->ss_family == AF_INET6) buffer = strecpy(buffer, "]", last);
-	buffer += seprintf(buffer, last, ":%d", this->GetPort());
+	buffer += seprintf(buffer, last, ":%d", (int)this->GetPort());
 
 	if (with_family) {
 		char family;
Index: src/network/network.cpp
===================================================================
--- src/network/network.cpp	(wersja 23508)
+++ src/network/network.cpp	(kopia robocza)
@@ -1021,7 +1021,7 @@
 {
 	extern SOCKET _debug_socket;  // Comes from debug.c
 
-	DEBUG(net, 0, "Redirecting DEBUG() to %s:%d", address.GetHostname(), address.GetPort());
+	DEBUG(net, 0, "Redirecting DEBUG() to %s:%d", address.GetHostname(), (int)address.GetPort());
 
 	SOCKET s = address.Connect();
 	if (s == INVALID_SOCKET) {
