Index: trunk/src/blitter/8bpp_base.cpp
===================================================================
--- trunk/src/blitter/8bpp_base.cpp	(revision 10765)
+++ trunk/src/blitter/8bpp_base.cpp	(working copy)
@@ -60,7 +60,7 @@
 		stepx = 1;
 	}
 
-	this->SetPixel(video, x, y, color);
+	if (x > 0 && y > 0 && x < screen_width && y < screen_height) this->SetPixel(video, x, y, color);
 	if (dx > dy) {
 		frac = dy - (dx / 2);
 		while (x != x2) {
Index: trunk/src/blitter/32bpp_base.cpp
===================================================================
--- trunk/src/blitter/32bpp_base.cpp	(revision 10765)
+++ trunk/src/blitter/32bpp_base.cpp	(working copy)
@@ -56,7 +56,7 @@
 		stepx = 1;
 	}
 
-	this->SetPixel(video, x, y, color);
+	if (x > 0 && y > 0 && x < screen_width && y < screen_height) this->SetPixel(video, x, y, color);
 	if (dx > dy) {
 		frac = dy - (dx / 2);
 		while (x != x2) {
