Index: src/dock_gui.cpp
===================================================================
--- src/dock_gui.cpp	(revision 10048)
+++ src/dock_gui.cpp	(working copy)
@@ -242,7 +242,11 @@
 
 		rad = (_patches.modified_catchment) ? CA_DOCK : 4;
 
-		if (_station_show_coverage) SetTileSelectBigSize(-rad, -rad, 2 * rad, 2 * rad);
+		if (_station_show_coverage) {
+			SetTileSelectBigSize(-rad, -rad, 2 * rad, 2 * rad);
+		} else {
+			SetTileSelectSize(1, 1);
+		}
 
 		DrawStationCoverageAreaText(4, 50, (uint)-1, rad);
 		break;
