diff --git a/src/vehicle_gui.cpp b/src/vehicle_gui.cpp
index 1f03319..2e8b0ec 100644
--- a/src/vehicle_gui.cpp
+++ b/src/vehicle_gui.cpp
@@ -2372,7 +2372,7 @@ static const uint32 _vehicle_command_translation_table[][4] = {
 };
 
 /**
- * This is the Callback method after the cloning attempt of a vehicle
+ * This is the Callback method after attempting to start/stop a vehicle
  * @param result the result of the cloning command
  * @param tile unused
  * @param p1 vehicle ID
