Index: src/rail_gui.cpp
===================================================================
--- src/rail_gui.cpp	(revision 22375)
+++ src/rail_gui.cpp	(working copy)
@@ -1546,6 +1546,10 @@
 		this->InitNested(desc, TRANSPORT_RAIL);
 		this->OnInvalidateData();
 	}
+	virtual ~BuildSignalWindow()
+	{
+	    _convert_signal_button = false;
+	}
 
 	virtual void SetStringParameters(int widget) const
 	{
