Index: os/macosx/plistgen.sh
===================================================================
--- os/macosx/plistgen.sh	(revision 12217)
+++ os/macosx/plistgen.sh	(working copy)
@@ -19,7 +19,7 @@
         <key>CFBundleDevelopmentRegion</key>
         <string>English</string>
         <key>CFBundleDisplayName</key>
-        <string>Open Transport Tycoon</string>
+        <string>OpenTTD</string>
         <key>CFBundleExecutable</key>
         <string>openttd</string>
         <key>CFBundleGetInfoString</key>
