diff -r 01b207fc81b0 src/rail_gui.cpp
--- a/src/rail_gui.cpp	Tue Jan 13 19:48:33 2009 +0000
+++ b/src/rail_gui.cpp	Tue Jan 13 21:47:34 2009 +0100
@@ -759,6 +759,7 @@ struct BuildRailToolbarWindow : Window {
 		delete FindWindowById(WC_BUILD_STATION, 0);
 		delete FindWindowById(WC_BUILD_DEPOT, 0);
 		delete FindWindowById(WC_SELECT_STATION, 0);
+		delete FindWindowById(WC_BUILD_BRIDGE, 0);
 	}
 
 	virtual void OnPlacePresize(Point pt, TileIndex tile)
diff -r 01b207fc81b0 src/road_gui.cpp
--- a/src/road_gui.cpp	Tue Jan 13 19:48:33 2009 +0000
+++ b/src/road_gui.cpp	Tue Jan 13 21:47:34 2009 +0100
@@ -531,6 +531,7 @@ struct BuildRoadToolbarWindow : Window {
 		delete FindWindowById(WC_TRUCK_STATION, 0);
 		delete FindWindowById(WC_BUILD_DEPOT, 0);
 		delete FindWindowById(WC_SELECT_STATION, 0);
+		delete FindWindowById(WC_BUILD_BRIDGE, 0);
 	}
 
 	virtual void OnPlaceDrag(ViewportPlaceMethod select_method, ViewportDragDropSelectionProcess select_proc, Point pt)
