diff -r 920105c0553e src/road_cmd.cpp
--- a/src/road_cmd.cpp	Sun Jan 02 18:45:24 2011 +0000
+++ b/src/road_cmd.cpp	Sun Jan 02 19:21:11 2011 +0000
@@ -1240,7 +1240,10 @@
 				DrawGroundSprite(road, pal);
 
 				SpriteID rail = GetCustomRailSprite(rti, ti->tile, RTSG_CROSSING) + axis;
-				DrawGroundSprite(rail, PAL_NONE);
+				/* Draw tracks, draw PBS reserved tracks darker */
+				PaletteID pal = (_game_mode != GM_MENU && _settings_client.gui.show_track_reservation && HasCrossingReservation(ti->tile)) ? PALETTE_CRASH : PAL_NONE; 
+				DrawGroundSprite(rail, pal);
+				
 				DrawRailTileSeq(ti, &_crossing_layout, TO_CATENARY, rail, 0, PAL_NONE);
 			} else {
 				SpriteID image = rti->base_sprites.crossing;
