diff -r 7dd629d55898 src/group_gui.cpp
--- a/src/group_gui.cpp	Tue Mar 16 18:45:44 2010 +0000
+++ b/src/group_gui.cpp	Tue Mar 16 20:22:53 2010 +0100
@@ -400,7 +400,7 @@
 
 					/* draw the selected group in white, else we draw it in black */
 					SetDParam(0, g->index);
-					DrawString(r.left + WD_FRAMERECT_LEFT + 8, r.right - WD_FRAMERECT_RIGHT, y1, STR_GROUP_NAME, (this->group_sel == g->index) ? TC_WHITE : TC_BLACK);
+					DrawString(r.left + WD_FRAMERECT_LEFT + 8, r.right - WD_FRAMERECT_RIGHT - 8, y1, STR_GROUP_NAME, (this->group_sel == g->index) ? TC_WHITE : TC_BLACK);
 
 					/* draw the number of vehicles of the group */
 					SetDParam(0, g->num_vehicle);
