Index: src/lang/english.txt
===================================================================
--- src/lang/english.txt	(revision 9730)
+++ src/lang/english.txt	(working copy)
@@ -2534,6 +2534,15 @@
 STR_HEADING_FOR_TRAIN_DEPOT_SERVICE                             :{LTBLUE}Service at {TOWN} Train Depot
 STR_HEADING_FOR_TRAIN_DEPOT_SERVICE_VEL                         :{LTBLUE}Service at {TOWN} Train Depot, {VELOCITY}
 
+STR_NEW_GO_TO                                                   :Go to {STATION} {SKIP}{STRING}{STRING}
+STR_NEW_GO_NON_STOP_TO                                          :Go non-stop to {STATION} {SKIP}{STRING}{STRING}
+STR_LOADING_BTN                                                 :{BLACK}Loading
+STR_ORDER_TRANSFER                                              :(Transfer and take cargo){STRING}
+STR_ORDER_LOAD                                                  :(Load){STRING}
+STR_ORDER_TRANSFER_LOAD                                         :(Transfer and wait for load){STRING}
+STR_ORDER_UNLOAD                                                :(Unload){STRING}
+STR_ORDER_TRANSFER_UNLOAD                                       :(Transfer and leave empty){STRING}
+
 STR_INVALID_ORDER                                               :{RED} (Invalid Order)
 
 STR_UNKNOWN_DESTINATION                                         :unknown destination
Index: src/order_gui.cpp
===================================================================
--- src/order_gui.cpp	(revision 9730)
+++ src/order_gui.cpp	(working copy)
@@ -156,10 +156,42 @@
 			SetDParam(1, 6);
 
 			switch (order->type) {
-				case OT_GOTO_STATION:
+				case OT_GOTO_STATION: {
+					if(order->flags <= 16) { /* Old Display to be removed*/
 					SetDParam(1, StationOrderStrings[order->flags]);
 					SetDParam(2, order->dest);
+					}
+
+					else { 
+						/*New Display System.  This new display system allows for better expandability.  
+						* A new Display system is need to simplify the order system when adding new Order Flags
+						* This display system was originaly in MinIN. 
+						*/
+
+					    SetDParam(1, STR_NEW_GO_TO);
+						SetDParam(4, STR_EMPTY);
+						SetDParam(5, STR_EMPTY);
+
+					    // Non-Stop
+					    if(HASBIT(order->flags, OFB_NON_STOP))
+							SetDParam(1, STR_NEW_GO_NON_STOP_TO);
+
+					    SetDParam(2, order->dest);
+
+						// Transfer
+						if(HASBIT(order->flags, OFB_TRANSFER))
+							SetDParam(4, STR_ORDER_TRANSFER);
+
+					    // Load
+					    if(HASBIT(order->flags, OFB_FULL_LOAD))
+							SetDParam(4, STR_ORDER_LOAD + (HASBIT(order->flags, OFB_TRANSFER)?1:0));
+
+					    // Unload
+					    else if(HASBIT(order->flags, OFB_UNLOAD))
+							SetDParam(4, STR_ORDER_UNLOAD  + (HASBIT(order->flags, OFB_TRANSFER)?1:0));
+					}
 					break;
+				}
 
 				case OT_GOTO_DEPOT: {
 					StringID s = STR_NULL;
