Index: src/console_cmds.cpp
===================================================================
--- src/console_cmds.cpp	(Revision 13486)
+++ src/console_cmds.cpp	(Arbeitskopie)
@@ -119,6 +119,19 @@
 	return true;
 }
 
+extern void StartupEconomy();
+DEF_CONSOLE_CMD(ConResetEconomy)
+{
+	if (argc == 0) {
+		IConsoleHelp("Reset cargo payment rates and prices. It might solve issues with unpaid cargo transports. Usage: 'reseteconomy'");
+		return true;
+	}
+
+	StartupEconomy();
+	InitializeLandscapeVariables(false);
+	return true;
+}
+
 #ifdef _DEBUG
 DEF_CONSOLE_CMD(ConResetTile)
 {
@@ -1357,6 +1370,7 @@
 	IConsoleCmdRegister("getdate",      ConGetDate);
 	IConsoleCmdRegister("quit",         ConExit);
 	IConsoleCmdRegister("resetengines", ConResetEngines);
+	IConsoleCmdRegister("reseteconomy", ConResetEconomy);
 	IConsoleCmdRegister("return",       ConReturn);
 	IConsoleCmdRegister("screenshot",   ConScreenShot);
 	IConsoleCmdRegister("script",       ConScript);
