Index: src/widget.cpp
===================================================================
--- src/widget.cpp	(revision 11834)
+++ src/widget.cpp	(working copy)
@@ -7,6 +7,7 @@
 #include "core/math_func.hpp"
 #include "player_func.h"
 #include "gfx_func.h"
+#include "strings_func.h"
 #include "window_gui.h"
 #include "window_func.h"
 
@@ -522,6 +523,7 @@
 		case WE_PAINT: {
 			int x,y,i,sel;
 			int width, height;
+			int cur_argument;
 
 			DrawWindowWidgets(w);
 
@@ -531,8 +533,17 @@
 			sel    = WP(w, dropdown_d).selected_index;
 			width  = w->widget[0].right - 3;
 			height = w->widget[0].bottom - 3;
+			
+			cur_argument = 0;
 
 			for (i = 0; WP(w, dropdown_d).items[i] != INVALID_STRING_ID; i++, sel--) {
+				if (WP(w, dropdown_d).num_arguments != NULL) {
+					for (int j = 0; j<WP(w, dropdown_d).num_arguments[i]; j++) {
+						SetDParam(j, WP(w, dropdown_d).arguments[cur_argument]);
+						cur_argument++;
+					}
+				}
+				
 				if (HasBit(WP(w, dropdown_d).hidden_state, i)) continue;
 
 				if (y >= 0 && y <= height) {
@@ -612,6 +623,11 @@
 
 void ShowDropDownMenu(Window *w, const StringID *strings, int selected, int button, uint32 disabled_mask, uint32 hidden_mask)
 {
+	ShowDropDownMenuWithArguments(w, strings, NULL, NULL, selected, button, disabled_mask, hidden_mask);
+}
+
+void ShowDropDownMenuWithArguments(Window *w, const StringID *strings, int *num_arguments, uint64 *arguments, int selected, int button, uint32 disabled_mask, uint32 hidden_mask)
+{
 	int i;
 	const Widget *wi;
 	Window *w2;
@@ -710,6 +726,9 @@
 
 	WP(w2, dropdown_d).click_delay = 0;
 	WP(w2, dropdown_d).drag_mode = true;
+	
+	WP(w2, dropdown_d).num_arguments = num_arguments;
+	WP(w2, dropdown_d).arguments = arguments;
 }
 
 
Index: src/window_gui.h
===================================================================
--- src/window_gui.h	(revision 11834)
+++ src/window_gui.h	(working copy)
@@ -450,6 +466,8 @@
 	const StringID *items;
 	byte click_delay;
 	bool drag_mode;
+	int *num_arguments;
+	uint64 *arguments;
 };
 assert_compile(WINDOW_CUSTOM_SIZE >= sizeof(dropdown_d));
 
@@ -593,8 +611,8 @@
 int GetWidgetFromPos(const Window *w, int x, int y);
 void DrawWindowWidgets(const Window *w);
 void ShowDropDownMenu(Window *w, const StringID *strings, int selected, int button, uint32 disabled_mask, uint32 hidden_mask);
+void ShowDropDownMenuWithArguments(Window *w, const StringID *strings, int *num_arguments, uint64 *arguments, int selected, int button, uint32 disabled_mask, uint32 hidden_mask);
 
-
 Window *GetCallbackWnd();
 void DeleteNonVitalWindows();
 void DeleteAllNonVitalWindows();
