Index: waypoint_gui.cpp
===================================================================
--- waypoint_gui.cpp	(revision 21023)
+++ waypoint_gui.cpp	(working copy)
@@ -46,7 +46,9 @@
 	 */
 	TileIndex GetCenterTile() const
 	{
-		return this->wp->IsInUse() ? this->wp->train_station.GetCenterTile() : this->wp->xy;
+		TileArea ta;
+		this->wp->GetTileArea(&ta, this->vt == VEH_TRAIN ? STATION_WAYPOINT : STATION_BUOY);
+		return this->wp->IsInUse() ? ta.GetCenterTile() : this->wp->xy;
 	}
 
 public:
