Index: rail.h
===================================================================
--- rail.h	(revision 5333)
+++ rail.h	(working copy)
@@ -269,7 +269,7 @@
  * Discards all directional information from the given TrackdirBits. Any
  * Track which is present in either direction will be present in the result.
  */
-static inline TrackBits TrackdirBitsToTrackBits(TrackdirBits bits) { return (TrackBits)(bits | (bits >> 8)); }
+static inline TrackBits TrackdirBitsToTrackBits(TrackdirBits bits) { return (TrackBits)((bits | (bits >> 8)) & TRACK_BIT_MASK); }
 
 /**
  * Maps a trackdir to the trackdir that you will end up on if you go straight
