Index: src/lang/english.txt
===================================================================
--- src/lang/english.txt	(revision 19753)
+++ src/lang/english.txt	(working copy)
@@ -665,11 +665,14 @@
 STR_SMALLMAP_LEGENDA_STATIONS_AIRPORTS_DOCKS                    :{TINYFONT}{BLACK}Stations/Airports/Docks
 STR_SMALLMAP_LEGENDA_BUILDINGS_INDUSTRIES                       :{TINYFONT}{BLACK}Buildings/Industries
 STR_SMALLMAP_LEGENDA_VEHICLES                                   :{TINYFONT}{BLACK}Vehicles
+STR_SMALLMAP_LEGENDA_0M                                         :{TINYFONT}{BLACK}0m
+STR_SMALLMAP_LEGENDA_50M                                        :{TINYFONT}{BLACK}50m
 STR_SMALLMAP_LEGENDA_100M                                       :{TINYFONT}{BLACK}100m
+STR_SMALLMAP_LEGENDA_150M                                       :{TINYFONT}{BLACK}150m
 STR_SMALLMAP_LEGENDA_200M                                       :{TINYFONT}{BLACK}200m
+STR_SMALLMAP_LEGENDA_250M                                       :{TINYFONT}{BLACK}250m
 STR_SMALLMAP_LEGENDA_300M                                       :{TINYFONT}{BLACK}300m
-STR_SMALLMAP_LEGENDA_400M                                       :{TINYFONT}{BLACK}400m
-STR_SMALLMAP_LEGENDA_500M                                       :{TINYFONT}{BLACK}500m
+STR_SMALLMAP_LEGENDA_350M                                       :{TINYFONT}{BLACK}350m
 STR_SMALLMAP_LEGENDA_TRAINS                                     :{TINYFONT}{BLACK}Trains
 STR_SMALLMAP_LEGENDA_ROAD_VEHICLES                              :{TINYFONT}{BLACK}Road Vehicles
 STR_SMALLMAP_LEGENDA_SHIPS                                      :{TINYFONT}{BLACK}Ships
Index: src/smallmap_gui.cpp
===================================================================
--- src/smallmap_gui.cpp	(revision 19753)
+++ src/smallmap_gui.cpp	(working copy)
@@ -77,11 +77,14 @@
 /** Legend text giving the colours to look for on the minimap */
 static LegendAndColour _legend_land_contours[] = {
 	/* The colours for the following values are set at BuildLandLegend() based on each colour scheme. */
+	MC(STR_SMALLMAP_LEGENDA_0M),
+	MC(STR_SMALLMAP_LEGENDA_50M),
 	MC(STR_SMALLMAP_LEGENDA_100M),
-	MC(STR_SMALLMAP_LEGENDA_200M),
+	MC(STR_SMALLMAP_LEGENDA_150M),
+	MS(_legend_land_contours[4].colour, STR_SMALLMAP_LEGENDA_200M),
+	MC(STR_SMALLMAP_LEGENDA_250M),
 	MC(STR_SMALLMAP_LEGENDA_300M),
-	MC(STR_SMALLMAP_LEGENDA_400M),
-	MC(STR_SMALLMAP_LEGENDA_500M),
+	MC(STR_SMALLMAP_LEGENDA_350M),
 
 	MS(0xD7, STR_SMALLMAP_LEGENDA_ROADS),
 	MK(0x0A, STR_SMALLMAP_LEGENDA_RAILROADS),
@@ -272,10 +275,13 @@
 void BuildLandLegend()
 {
 	_legend_land_contours[0].colour = _heightmap_schemes[_settings_client.gui.smallmap_land_colour].height_colours[0];
-	_legend_land_contours[1].colour = _heightmap_schemes[_settings_client.gui.smallmap_land_colour].height_colours[4];
-	_legend_land_contours[2].colour = _heightmap_schemes[_settings_client.gui.smallmap_land_colour].height_colours[8];
-	_legend_land_contours[3].colour = _heightmap_schemes[_settings_client.gui.smallmap_land_colour].height_colours[12];
-	_legend_land_contours[4].colour = _heightmap_schemes[_settings_client.gui.smallmap_land_colour].height_colours[14];
+	_legend_land_contours[1].colour = _heightmap_schemes[_settings_client.gui.smallmap_land_colour].height_colours[2];
+	_legend_land_contours[2].colour = _heightmap_schemes[_settings_client.gui.smallmap_land_colour].height_colours[4];
+	_legend_land_contours[3].colour = _heightmap_schemes[_settings_client.gui.smallmap_land_colour].height_colours[6];
+	_legend_land_contours[4].colour = _heightmap_schemes[_settings_client.gui.smallmap_land_colour].height_colours[8];
+	_legend_land_contours[5].colour = _heightmap_schemes[_settings_client.gui.smallmap_land_colour].height_colours[10];
+	_legend_land_contours[6].colour = _heightmap_schemes[_settings_client.gui.smallmap_land_colour].height_colours[12];
+	_legend_land_contours[7].colour = _heightmap_schemes[_settings_client.gui.smallmap_land_colour].height_colours[14];
 }
 
 struct AndOr {
