diff --git a/src/network/core/os_abstraction.h b/src/network/core/os_abstraction.h
--- a/src/network/core/os_abstraction.h
+++ b/src/network/core/os_abstraction.h
@@ -190,10 +190,6 @@
 	typedef int socklen_t;
 #endif
 
-#ifdef __HAIKU__
-	#define IPV6_V6ONLY 27
-#endif
-
 #if defined(PSP)
 #	include <sys/socket.h>
 #	include <netinet/in.h>
diff --git a/src/stdafx.h b/src/stdafx.h
--- a/src/stdafx.h
+++ b/src/stdafx.h
@@ -32,7 +32,7 @@
  * INT64_MAX for them ourselves. */
 #if defined(__APPLE__)
 	/* Already done in osx_stdafx.h */
-#elif !defined(_MSC_VER) && !defined( __MORPHOS__) && !defined(_STDINT_H_)
+#elif !defined(_MSC_VER) && !defined( __MORPHOS__) && !defined(_STDINT_H_) || defined(__HAIKU__)
 	#if defined(SUNOS)
 		/* SunOS/Solaris does not have stdint.h, but inttypes.h defines everything
 		 * stdint.h defines and we need. */
