diff -r 3adbc1f6c0db src/train_cmd.cpp
--- a/src/train_cmd.cpp	Sun Jan 04 14:19:54 2009 +0100
+++ b/src/train_cmd.cpp	Sun Jan 04 15:07:59 2009 +0100
@@ -2849,6 +2849,8 @@
 	{
 		++this->index;
 
+		uint16 conditional_depth = 0;
+
 		do {
 			/* Wrap around. */
 			if (this->index >= this->v->GetNumOrders()) this->index = 0;
@@ -2866,8 +2868,10 @@
 					UpdateOrderDest(this->v, order);
 					return true;
 				case OT_CONDITIONAL: {
+					if (conditional_depth > this->v->GetNumOrders()) return false;
 					VehicleOrderID next = ProcessConditionalOrder(order, this->v);
 					if (next != INVALID_VEH_ORDER_ID) {
+						conditional_depth++;
 						this->index = next;
 						/* Don't increment next, so no break here. */
 						continue;
