diff -r 99f2fcfb3d74 src/town_cmd.cpp
--- a/src/town_cmd.cpp	Tue Jun 02 03:31:03 2015 +0300
+++ b/src/town_cmd.cpp	Wed Jun 17 01:13:19 2015 +0300
@@ -1315,10 +1315,10 @@
 		/* Check whether a road connection exists or can be build. */
 		switch (GetTileType(target_tile)) {
 			case MP_ROAD:
+			case MP_STATION:
 				return target_rb != ROAD_NONE;
 
 			case MP_HOUSE:
-			case MP_STATION:
 			case MP_INDUSTRY:
 			case MP_OBJECT:
 				return false;
