diff -r 237997d34419 -r 69c327f7b2db src/settings_gui.cpp
--- a/src/settings_gui.cpp	Sun Sep 28 14:09:54 2008 +0200
+++ b/src/settings_gui.cpp	Sun Sep 28 14:09:59 2008 +0200
@@ -586,7 +586,6 @@
 static const char *_patches_ui[] = {
 	"gui.vehicle_speed",
 	"gui.status_long_date",
-	"gui.show_finances",
 	"gui.autoscroll",
 	"gui.reverse_scroll",
 	"gui.smooth_scroll",
@@ -626,7 +625,6 @@
 	"station.always_small_airport",
 	"gui.enable_signal_gui",
 	"gui.drag_signals_density",
-	"game_creation.oil_refinery_limit",
 	"gui.semaphore_build_before",
 	"gui.default_signal_type",
 	"gui.cycle_signal_types",
@@ -650,28 +648,34 @@
 };
 
 static const char *_patches_economy[] = {
+	"gui.show_finances",
 	"economy.inflation",
+	"economy.smooth_economy",
+	"economy.bribe",
+	"economy.exclusive_rights",
+	"economy.allow_shares",
+	"economy.give_money",
+	"",
 	"construction.raw_industry_construction",
 	"economy.multiple_industry_per_town",
 	"economy.same_industry_close",
-	"economy.bribe",
-	"economy.exclusive_rights",
-	"economy.give_money",
-	"gui.colored_news_year",
-	"gui.ending_year",
-	"economy.smooth_economy",
-	"economy.allow_shares",
+	"game_creation.oil_refinery_limit",
+	"",
 	"economy.town_layout",
 	"economy.mod_road_rebuild",
 	"economy.town_growth_rate",
 	"economy.larger_towns",
 	"economy.initial_city_size",
+	"",
+	"gui.colored_news_year",
+	"gui.ending_year",
 	NULL
 };
 
 static const char *_patches_ai[] = {
 	"ai.ainew_active",
 	"ai.ai_in_multiplayer",
+	"",
 	"ai.ai_disable_veh_train",
 	"ai.ai_disable_veh_roadveh",
 	"ai.ai_disable_veh_aircraft",
