Index: src/genworld_gui.cpp
===================================================================
--- src/genworld_gui.cpp	(revision 10649)
+++ src/genworld_gui.cpp	(working copy)
@@ -255,7 +255,7 @@
 		SetWindowWidgetDisabledState(w, GLAND_START_DATE_DOWN, _patches_newgame.starting_year <= MIN_YEAR);
 		SetWindowWidgetDisabledState(w, GLAND_START_DATE_UP,   _patches_newgame.starting_year >= MAX_YEAR);
 		SetWindowWidgetDisabledState(w, GLAND_SNOW_LEVEL_DOWN, _patches_newgame.snow_line_height <= 2 || _opt_newgame.landscape != LT_ARCTIC);
-		SetWindowWidgetDisabledState(w, GLAND_SNOW_LEVEL_UP,   _patches_newgame.snow_line_height >= 13 || _opt_newgame.landscape != LT_ARCTIC);
+		SetWindowWidgetDisabledState(w, GLAND_SNOW_LEVEL_UP,   _patches_newgame.snow_line_height >= MAX_TILE_HEIGHT - 2 || _opt_newgame.landscape != LT_ARCTIC);
 
 		SetWindowWidgetLoweredState(w, GLAND_TEMPERATE, _opt_newgame.landscape == LT_TEMPERATE);
 		SetWindowWidgetLoweredState(w, GLAND_ARCTIC,    _opt_newgame.landscape == LT_ARCTIC);
@@ -402,7 +402,7 @@
 				HandleButtonClick(w, e->we.click.widget);
 				SetWindowDirty(w);
 
-				_patches_newgame.snow_line_height = clamp(_patches_newgame.snow_line_height + e->we.click.widget - GLAND_SNOW_LEVEL_TEXT, 2, 13);
+				_patches_newgame.snow_line_height = clamp(_patches_newgame.snow_line_height + e->we.click.widget - GLAND_SNOW_LEVEL_TEXT, 2, MAX_TILE_HEIGHT - 2);
 			}
 			_left_button_clicked = false;
 			break;
@@ -498,7 +498,7 @@
 				break;
 			case GLAND_SNOW_LEVEL_TEXT:
 				InvalidateWidget(w, GLAND_SNOW_LEVEL_TEXT);
-				_patches_newgame.snow_line_height = clamp(value, 2, 13);
+				_patches_newgame.snow_line_height = clamp(value, 2, MAX_TILE_HEIGHT - 2);
 				break;
 			}
 
@@ -611,7 +611,7 @@
 		SetWindowWidgetDisabledState(w, CSCEN_START_DATE_DOWN,       _patches_newgame.starting_year <= MIN_YEAR);
 		SetWindowWidgetDisabledState(w, CSCEN_START_DATE_UP,         _patches_newgame.starting_year >= MAX_YEAR);
 		SetWindowWidgetDisabledState(w, CSCEN_FLAT_LAND_HEIGHT_DOWN, _patches_newgame.se_flat_world_height <= 0);
-		SetWindowWidgetDisabledState(w, CSCEN_FLAT_LAND_HEIGHT_UP,   _patches_newgame.se_flat_world_height >= 15);
+		SetWindowWidgetDisabledState(w, CSCEN_FLAT_LAND_HEIGHT_UP,   _patches_newgame.se_flat_world_height >= MAX_TILE_HEIGHT);
 
 		SetWindowWidgetLoweredState(w, CSCEN_TEMPERATE, _opt_newgame.landscape == LT_TEMPERATE);
 		SetWindowWidgetLoweredState(w, CSCEN_ARCTIC,    _opt_newgame.landscape == LT_ARCTIC);
@@ -672,7 +672,7 @@
 				HandleButtonClick(w, e->we.click.widget);
 				SetWindowDirty(w);
 
-				_patches_newgame.se_flat_world_height = clamp(_patches_newgame.se_flat_world_height + e->we.click.widget - CSCEN_FLAT_LAND_HEIGHT_TEXT, 0, 15);
+				_patches_newgame.se_flat_world_height = clamp(_patches_newgame.se_flat_world_height + e->we.click.widget - CSCEN_FLAT_LAND_HEIGHT_TEXT, 0, MAX_TILE_HEIGHT);
 			}
 			_left_button_clicked = false;
 			break;
@@ -703,7 +703,7 @@
 				break;
 			case CSCEN_FLAT_LAND_HEIGHT_TEXT:
 				InvalidateWidget(w, CSCEN_FLAT_LAND_HEIGHT_TEXT);
-				_patches_newgame.se_flat_world_height = clamp(value, 0, 15);
+				_patches_newgame.se_flat_world_height = clamp(value, 0, MAX_TILE_HEIGHT);
 				break;
 			}
 
Index: src/smallmap_gui.cpp
===================================================================
--- src/smallmap_gui.cpp	(revision 10649)
+++ src/smallmap_gui.cpp	(working copy)
@@ -193,8 +193,11 @@
 
 #define MKCOLOR(x) TO_LE32X(x)
 
-/* Height encodings; 16 levels XXX - needs updating for more/finer heights! */
-static const uint32 _map_height_bits[16] = {
+/*
+ * Height encodings; MAX_TILE_HEIGHT + 1 levels, from 0 to MAX_TILE_HEIGHT
+ * XXX - needs updating for correct number of heights if MAX_TILE_HEIGHT changes!
+ */
+static const uint32 _map_height_bits[MAX_TILE_HEIGHT + 1] = {
 	MKCOLOR(0x5A5A5A5A),
 	MKCOLOR(0x5A5B5A5B),
 	MKCOLOR(0x5B5B5B5B),
Index: src/tile.h
===================================================================
--- src/tile.h	(revision 10649)
+++ src/tile.h	(working copy)
@@ -10,6 +10,9 @@
 #include "map.h"
 #include "slope.h"
 
+/** maximum allowed tile height */
+#define MAX_TILE_HEIGHT 15
+
 enum TileType {
 	MP_CLEAR,
 	MP_RAILWAY,
Index: src/clear_cmd.cpp
===================================================================
--- src/clear_cmd.cpp	(revision 10649)
+++ src/clear_cmd.cpp	(working copy)
@@ -179,7 +179,7 @@
 
 	_error_message = STR_1004_TOO_HIGH;
 
-	if (height > 15) return false;
+	if (height > MAX_TILE_HEIGHT) return false;
 
 	nh = TerraformGetHeightOfTile(ts, tile);
 	if (nh < 0 || height == nh) return false;
